// @cdktf/provider-aws/lib/data-aws-codeartifact-authorization-token/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsCodeartifactAuthorizationToken = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/codeartifact_authorization_token aws_codeartifact_authorization_token}
*/
class DataAwsCodeartifactAuthorizationToken extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/codeartifact_authorization_token aws_codeartifact_authorization_token} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsCodeartifactAuthorizationTokenConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_codeartifact_authorization_token',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._domain = config.domain;
        this._domainOwner = config.domainOwner;
        this._durationSeconds = config.durationSeconds;
        this._id = config.id;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // authorization_token - computed: true, optional: false, required: false
    get authorizationToken() {
        return this.getStringAttribute('authorization_token');
    }
    get domain() {
        return this.getStringAttribute('domain');
    }
    set domain(value) {
        this._domain = value;
    }
    // Temporarily expose input value. Use with caution.
    get domainInput() {
        return this._domain;
    }
    get domainOwner() {
        return this.getStringAttribute('domain_owner');
    }
    set domainOwner(value) {
        this._domainOwner = value;
    }
    resetDomainOwner() {
        this._domainOwner = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get domainOwnerInput() {
        return this._domainOwner;
    }
    get durationSeconds() {
        return this.getNumberAttribute('duration_seconds');
    }
    set durationSeconds(value) {
        this._durationSeconds = value;
    }
    resetDurationSeconds() {
        this._durationSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get durationSecondsInput() {
        return this._durationSeconds;
    }
    // expiration - computed: true, optional: false, required: false
    get expiration() {
        return this.getStringAttribute('expiration');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            domain: cdktf.stringToTerraform(this._domain),
            domain_owner: cdktf.stringToTerraform(this._domainOwner),
            duration_seconds: cdktf.numberToTerraform(this._durationSeconds),
            id: cdktf.stringToTerraform(this._id),
        };
    }
}
exports.DataAwsCodeartifactAuthorizationToken = DataAwsCodeartifactAuthorizationToken;
_a = JSII_RTTI_SYMBOL_1;
DataAwsCodeartifactAuthorizationToken[_a] = { fqn: "@cdktf/provider-aws.dataAwsCodeartifactAuthorizationToken.DataAwsCodeartifactAuthorizationToken", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsCodeartifactAuthorizationToken.tfResourceType = "aws_codeartifact_authorization_token";
