// @cdktf/provider-aws/lib/autoscaling-attachment/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AutoscalingAttachment = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/autoscaling_attachment aws_autoscaling_attachment}
*/
class AutoscalingAttachment extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/autoscaling_attachment aws_autoscaling_attachment} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options AutoscalingAttachmentConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_autoscaling_attachment',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._albTargetGroupArn = config.albTargetGroupArn;
        this._autoscalingGroupName = config.autoscalingGroupName;
        this._elb = config.elb;
        this._id = config.id;
        this._lbTargetGroupArn = config.lbTargetGroupArn;
    }
    get albTargetGroupArn() {
        return this.getStringAttribute('alb_target_group_arn');
    }
    set albTargetGroupArn(value) {
        this._albTargetGroupArn = value;
    }
    resetAlbTargetGroupArn() {
        this._albTargetGroupArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get albTargetGroupArnInput() {
        return this._albTargetGroupArn;
    }
    get autoscalingGroupName() {
        return this.getStringAttribute('autoscaling_group_name');
    }
    set autoscalingGroupName(value) {
        this._autoscalingGroupName = value;
    }
    // Temporarily expose input value. Use with caution.
    get autoscalingGroupNameInput() {
        return this._autoscalingGroupName;
    }
    get elb() {
        return this.getStringAttribute('elb');
    }
    set elb(value) {
        this._elb = value;
    }
    resetElb() {
        this._elb = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get elbInput() {
        return this._elb;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get lbTargetGroupArn() {
        return this.getStringAttribute('lb_target_group_arn');
    }
    set lbTargetGroupArn(value) {
        this._lbTargetGroupArn = value;
    }
    resetLbTargetGroupArn() {
        this._lbTargetGroupArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lbTargetGroupArnInput() {
        return this._lbTargetGroupArn;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            alb_target_group_arn: cdktf.stringToTerraform(this._albTargetGroupArn),
            autoscaling_group_name: cdktf.stringToTerraform(this._autoscalingGroupName),
            elb: cdktf.stringToTerraform(this._elb),
            id: cdktf.stringToTerraform(this._id),
            lb_target_group_arn: cdktf.stringToTerraform(this._lbTargetGroupArn),
        };
    }
}
exports.AutoscalingAttachment = AutoscalingAttachment;
_a = JSII_RTTI_SYMBOL_1;
AutoscalingAttachment[_a] = { fqn: "@cdktf/provider-aws.autoscalingAttachment.AutoscalingAttachment", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
AutoscalingAttachment.tfResourceType = "aws_autoscaling_attachment";
