// @cdktf/provider-aws/lib/ecs-service/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EcsService = exports.EcsServiceTimeoutsOutputReference = exports.ecsServiceTimeoutsToTerraform = exports.EcsServiceServiceRegistriesOutputReference = exports.ecsServiceServiceRegistriesToTerraform = exports.EcsServicePlacementConstraintsList = exports.EcsServicePlacementConstraintsOutputReference = exports.ecsServicePlacementConstraintsToTerraform = exports.EcsServiceOrderedPlacementStrategyList = exports.EcsServiceOrderedPlacementStrategyOutputReference = exports.ecsServiceOrderedPlacementStrategyToTerraform = exports.EcsServiceNetworkConfigurationOutputReference = exports.ecsServiceNetworkConfigurationToTerraform = exports.EcsServiceLoadBalancerList = exports.EcsServiceLoadBalancerOutputReference = exports.ecsServiceLoadBalancerToTerraform = exports.EcsServiceDeploymentControllerOutputReference = exports.ecsServiceDeploymentControllerToTerraform = exports.EcsServiceDeploymentCircuitBreakerOutputReference = exports.ecsServiceDeploymentCircuitBreakerToTerraform = exports.EcsServiceCapacityProviderStrategyList = exports.EcsServiceCapacityProviderStrategyOutputReference = exports.ecsServiceCapacityProviderStrategyToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function ecsServiceCapacityProviderStrategyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        base: cdktf.numberToTerraform(struct.base),
        capacity_provider: cdktf.stringToTerraform(struct.capacityProvider),
        weight: cdktf.numberToTerraform(struct.weight),
    };
}
exports.ecsServiceCapacityProviderStrategyToTerraform = ecsServiceCapacityProviderStrategyToTerraform;
class EcsServiceCapacityProviderStrategyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._base !== undefined) {
            hasAnyValues = true;
            internalValueResult.base = this._base;
        }
        if (this._capacityProvider !== undefined) {
            hasAnyValues = true;
            internalValueResult.capacityProvider = this._capacityProvider;
        }
        if (this._weight !== undefined) {
            hasAnyValues = true;
            internalValueResult.weight = this._weight;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._base = undefined;
            this._capacityProvider = undefined;
            this._weight = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._base = value.base;
            this._capacityProvider = value.capacityProvider;
            this._weight = value.weight;
        }
    }
    get base() {
        return this.getNumberAttribute('base');
    }
    set base(value) {
        this._base = value;
    }
    resetBase() {
        this._base = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get baseInput() {
        return this._base;
    }
    get capacityProvider() {
        return this.getStringAttribute('capacity_provider');
    }
    set capacityProvider(value) {
        this._capacityProvider = value;
    }
    // Temporarily expose input value. Use with caution.
    get capacityProviderInput() {
        return this._capacityProvider;
    }
    get weight() {
        return this.getNumberAttribute('weight');
    }
    set weight(value) {
        this._weight = value;
    }
    resetWeight() {
        this._weight = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get weightInput() {
        return this._weight;
    }
}
exports.EcsServiceCapacityProviderStrategyOutputReference = EcsServiceCapacityProviderStrategyOutputReference;
_a = JSII_RTTI_SYMBOL_1;
EcsServiceCapacityProviderStrategyOutputReference[_a] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceCapacityProviderStrategyOutputReference", version: "10.0.12" };
class EcsServiceCapacityProviderStrategyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EcsServiceCapacityProviderStrategyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EcsServiceCapacityProviderStrategyList = EcsServiceCapacityProviderStrategyList;
_b = JSII_RTTI_SYMBOL_1;
EcsServiceCapacityProviderStrategyList[_b] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceCapacityProviderStrategyList", version: "10.0.12" };
function ecsServiceDeploymentCircuitBreakerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enable: cdktf.booleanToTerraform(struct.enable),
        rollback: cdktf.booleanToTerraform(struct.rollback),
    };
}
exports.ecsServiceDeploymentCircuitBreakerToTerraform = ecsServiceDeploymentCircuitBreakerToTerraform;
class EcsServiceDeploymentCircuitBreakerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enable !== undefined) {
            hasAnyValues = true;
            internalValueResult.enable = this._enable;
        }
        if (this._rollback !== undefined) {
            hasAnyValues = true;
            internalValueResult.rollback = this._rollback;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enable = undefined;
            this._rollback = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enable = value.enable;
            this._rollback = value.rollback;
        }
    }
    get enable() {
        return this.getBooleanAttribute('enable');
    }
    set enable(value) {
        this._enable = value;
    }
    // Temporarily expose input value. Use with caution.
    get enableInput() {
        return this._enable;
    }
    get rollback() {
        return this.getBooleanAttribute('rollback');
    }
    set rollback(value) {
        this._rollback = value;
    }
    // Temporarily expose input value. Use with caution.
    get rollbackInput() {
        return this._rollback;
    }
}
exports.EcsServiceDeploymentCircuitBreakerOutputReference = EcsServiceDeploymentCircuitBreakerOutputReference;
_c = JSII_RTTI_SYMBOL_1;
EcsServiceDeploymentCircuitBreakerOutputReference[_c] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceDeploymentCircuitBreakerOutputReference", version: "10.0.12" };
function ecsServiceDeploymentControllerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.ecsServiceDeploymentControllerToTerraform = ecsServiceDeploymentControllerToTerraform;
class EcsServiceDeploymentControllerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._type = value.type;
        }
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.EcsServiceDeploymentControllerOutputReference = EcsServiceDeploymentControllerOutputReference;
_d = JSII_RTTI_SYMBOL_1;
EcsServiceDeploymentControllerOutputReference[_d] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceDeploymentControllerOutputReference", version: "10.0.12" };
function ecsServiceLoadBalancerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        container_name: cdktf.stringToTerraform(struct.containerName),
        container_port: cdktf.numberToTerraform(struct.containerPort),
        elb_name: cdktf.stringToTerraform(struct.elbName),
        target_group_arn: cdktf.stringToTerraform(struct.targetGroupArn),
    };
}
exports.ecsServiceLoadBalancerToTerraform = ecsServiceLoadBalancerToTerraform;
class EcsServiceLoadBalancerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._containerName !== undefined) {
            hasAnyValues = true;
            internalValueResult.containerName = this._containerName;
        }
        if (this._containerPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.containerPort = this._containerPort;
        }
        if (this._elbName !== undefined) {
            hasAnyValues = true;
            internalValueResult.elbName = this._elbName;
        }
        if (this._targetGroupArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetGroupArn = this._targetGroupArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._containerName = undefined;
            this._containerPort = undefined;
            this._elbName = undefined;
            this._targetGroupArn = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._containerName = value.containerName;
            this._containerPort = value.containerPort;
            this._elbName = value.elbName;
            this._targetGroupArn = value.targetGroupArn;
        }
    }
    get containerName() {
        return this.getStringAttribute('container_name');
    }
    set containerName(value) {
        this._containerName = value;
    }
    // Temporarily expose input value. Use with caution.
    get containerNameInput() {
        return this._containerName;
    }
    get containerPort() {
        return this.getNumberAttribute('container_port');
    }
    set containerPort(value) {
        this._containerPort = value;
    }
    // Temporarily expose input value. Use with caution.
    get containerPortInput() {
        return this._containerPort;
    }
    get elbName() {
        return this.getStringAttribute('elb_name');
    }
    set elbName(value) {
        this._elbName = value;
    }
    resetElbName() {
        this._elbName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get elbNameInput() {
        return this._elbName;
    }
    get targetGroupArn() {
        return this.getStringAttribute('target_group_arn');
    }
    set targetGroupArn(value) {
        this._targetGroupArn = value;
    }
    resetTargetGroupArn() {
        this._targetGroupArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetGroupArnInput() {
        return this._targetGroupArn;
    }
}
exports.EcsServiceLoadBalancerOutputReference = EcsServiceLoadBalancerOutputReference;
_e = JSII_RTTI_SYMBOL_1;
EcsServiceLoadBalancerOutputReference[_e] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceLoadBalancerOutputReference", version: "10.0.12" };
class EcsServiceLoadBalancerList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EcsServiceLoadBalancerOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EcsServiceLoadBalancerList = EcsServiceLoadBalancerList;
_f = JSII_RTTI_SYMBOL_1;
EcsServiceLoadBalancerList[_f] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceLoadBalancerList", version: "10.0.12" };
function ecsServiceNetworkConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        assign_public_ip: cdktf.booleanToTerraform(struct.assignPublicIp),
        security_groups: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.securityGroups),
        subnets: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.subnets),
    };
}
exports.ecsServiceNetworkConfigurationToTerraform = ecsServiceNetworkConfigurationToTerraform;
class EcsServiceNetworkConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._assignPublicIp !== undefined) {
            hasAnyValues = true;
            internalValueResult.assignPublicIp = this._assignPublicIp;
        }
        if (this._securityGroups !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroups = this._securityGroups;
        }
        if (this._subnets !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnets = this._subnets;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._assignPublicIp = undefined;
            this._securityGroups = undefined;
            this._subnets = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._assignPublicIp = value.assignPublicIp;
            this._securityGroups = value.securityGroups;
            this._subnets = value.subnets;
        }
    }
    get assignPublicIp() {
        return this.getBooleanAttribute('assign_public_ip');
    }
    set assignPublicIp(value) {
        this._assignPublicIp = value;
    }
    resetAssignPublicIp() {
        this._assignPublicIp = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get assignPublicIpInput() {
        return this._assignPublicIp;
    }
    get securityGroups() {
        return cdktf.Fn.tolist(this.getListAttribute('security_groups'));
    }
    set securityGroups(value) {
        this._securityGroups = value;
    }
    resetSecurityGroups() {
        this._securityGroups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupsInput() {
        return this._securityGroups;
    }
    get subnets() {
        return cdktf.Fn.tolist(this.getListAttribute('subnets'));
    }
    set subnets(value) {
        this._subnets = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetsInput() {
        return this._subnets;
    }
}
exports.EcsServiceNetworkConfigurationOutputReference = EcsServiceNetworkConfigurationOutputReference;
_g = JSII_RTTI_SYMBOL_1;
EcsServiceNetworkConfigurationOutputReference[_g] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceNetworkConfigurationOutputReference", version: "10.0.12" };
function ecsServiceOrderedPlacementStrategyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        field: cdktf.stringToTerraform(struct.field),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.ecsServiceOrderedPlacementStrategyToTerraform = ecsServiceOrderedPlacementStrategyToTerraform;
class EcsServiceOrderedPlacementStrategyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._field !== undefined) {
            hasAnyValues = true;
            internalValueResult.field = this._field;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._field = undefined;
            this._type = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._field = value.field;
            this._type = value.type;
        }
    }
    get field() {
        return this.getStringAttribute('field');
    }
    set field(value) {
        this._field = value;
    }
    resetField() {
        this._field = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fieldInput() {
        return this._field;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.EcsServiceOrderedPlacementStrategyOutputReference = EcsServiceOrderedPlacementStrategyOutputReference;
_h = JSII_RTTI_SYMBOL_1;
EcsServiceOrderedPlacementStrategyOutputReference[_h] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceOrderedPlacementStrategyOutputReference", version: "10.0.12" };
class EcsServiceOrderedPlacementStrategyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EcsServiceOrderedPlacementStrategyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EcsServiceOrderedPlacementStrategyList = EcsServiceOrderedPlacementStrategyList;
_j = JSII_RTTI_SYMBOL_1;
EcsServiceOrderedPlacementStrategyList[_j] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceOrderedPlacementStrategyList", version: "10.0.12" };
function ecsServicePlacementConstraintsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        expression: cdktf.stringToTerraform(struct.expression),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.ecsServicePlacementConstraintsToTerraform = ecsServicePlacementConstraintsToTerraform;
class EcsServicePlacementConstraintsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._expression !== undefined) {
            hasAnyValues = true;
            internalValueResult.expression = this._expression;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._expression = undefined;
            this._type = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._expression = value.expression;
            this._type = value.type;
        }
    }
    get expression() {
        return this.getStringAttribute('expression');
    }
    set expression(value) {
        this._expression = value;
    }
    resetExpression() {
        this._expression = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get expressionInput() {
        return this._expression;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.EcsServicePlacementConstraintsOutputReference = EcsServicePlacementConstraintsOutputReference;
_k = JSII_RTTI_SYMBOL_1;
EcsServicePlacementConstraintsOutputReference[_k] = { fqn: "@cdktf/provider-aws.ecsService.EcsServicePlacementConstraintsOutputReference", version: "10.0.12" };
class EcsServicePlacementConstraintsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EcsServicePlacementConstraintsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EcsServicePlacementConstraintsList = EcsServicePlacementConstraintsList;
_l = JSII_RTTI_SYMBOL_1;
EcsServicePlacementConstraintsList[_l] = { fqn: "@cdktf/provider-aws.ecsService.EcsServicePlacementConstraintsList", version: "10.0.12" };
function ecsServiceServiceRegistriesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        container_name: cdktf.stringToTerraform(struct.containerName),
        container_port: cdktf.numberToTerraform(struct.containerPort),
        port: cdktf.numberToTerraform(struct.port),
        registry_arn: cdktf.stringToTerraform(struct.registryArn),
    };
}
exports.ecsServiceServiceRegistriesToTerraform = ecsServiceServiceRegistriesToTerraform;
class EcsServiceServiceRegistriesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._containerName !== undefined) {
            hasAnyValues = true;
            internalValueResult.containerName = this._containerName;
        }
        if (this._containerPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.containerPort = this._containerPort;
        }
        if (this._port !== undefined) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        if (this._registryArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.registryArn = this._registryArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._containerName = undefined;
            this._containerPort = undefined;
            this._port = undefined;
            this._registryArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._containerName = value.containerName;
            this._containerPort = value.containerPort;
            this._port = value.port;
            this._registryArn = value.registryArn;
        }
    }
    get containerName() {
        return this.getStringAttribute('container_name');
    }
    set containerName(value) {
        this._containerName = value;
    }
    resetContainerName() {
        this._containerName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerNameInput() {
        return this._containerName;
    }
    get containerPort() {
        return this.getNumberAttribute('container_port');
    }
    set containerPort(value) {
        this._containerPort = value;
    }
    resetContainerPort() {
        this._containerPort = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerPortInput() {
        return this._containerPort;
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
    get registryArn() {
        return this.getStringAttribute('registry_arn');
    }
    set registryArn(value) {
        this._registryArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get registryArnInput() {
        return this._registryArn;
    }
}
exports.EcsServiceServiceRegistriesOutputReference = EcsServiceServiceRegistriesOutputReference;
_m = JSII_RTTI_SYMBOL_1;
EcsServiceServiceRegistriesOutputReference[_m] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceServiceRegistriesOutputReference", version: "10.0.12" };
function ecsServiceTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.ecsServiceTimeoutsToTerraform = ecsServiceTimeoutsToTerraform;
class EcsServiceTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.EcsServiceTimeoutsOutputReference = EcsServiceTimeoutsOutputReference;
_o = JSII_RTTI_SYMBOL_1;
EcsServiceTimeoutsOutputReference[_o] = { fqn: "@cdktf/provider-aws.ecsService.EcsServiceTimeoutsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/ecs_service aws_ecs_service}
*/
class EcsService extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ecs_service aws_ecs_service} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options EcsServiceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ecs_service',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // capacity_provider_strategy - computed: false, optional: true, required: false
        this._capacityProviderStrategy = new EcsServiceCapacityProviderStrategyList(this, "capacity_provider_strategy", true);
        // deployment_circuit_breaker - computed: false, optional: true, required: false
        this._deploymentCircuitBreaker = new EcsServiceDeploymentCircuitBreakerOutputReference(this, "deployment_circuit_breaker");
        // deployment_controller - computed: false, optional: true, required: false
        this._deploymentController = new EcsServiceDeploymentControllerOutputReference(this, "deployment_controller");
        // load_balancer - computed: false, optional: true, required: false
        this._loadBalancer = new EcsServiceLoadBalancerList(this, "load_balancer", true);
        // network_configuration - computed: false, optional: true, required: false
        this._networkConfiguration = new EcsServiceNetworkConfigurationOutputReference(this, "network_configuration");
        // ordered_placement_strategy - computed: false, optional: true, required: false
        this._orderedPlacementStrategy = new EcsServiceOrderedPlacementStrategyList(this, "ordered_placement_strategy", false);
        // placement_constraints - computed: false, optional: true, required: false
        this._placementConstraints = new EcsServicePlacementConstraintsList(this, "placement_constraints", true);
        // service_registries - computed: false, optional: true, required: false
        this._serviceRegistries = new EcsServiceServiceRegistriesOutputReference(this, "service_registries");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new EcsServiceTimeoutsOutputReference(this, "timeouts");
        this._cluster = config.cluster;
        this._deploymentMaximumPercent = config.deploymentMaximumPercent;
        this._deploymentMinimumHealthyPercent = config.deploymentMinimumHealthyPercent;
        this._desiredCount = config.desiredCount;
        this._enableEcsManagedTags = config.enableEcsManagedTags;
        this._enableExecuteCommand = config.enableExecuteCommand;
        this._forceNewDeployment = config.forceNewDeployment;
        this._healthCheckGracePeriodSeconds = config.healthCheckGracePeriodSeconds;
        this._iamRole = config.iamRole;
        this._id = config.id;
        this._launchType = config.launchType;
        this._name = config.name;
        this._platformVersion = config.platformVersion;
        this._propagateTags = config.propagateTags;
        this._schedulingStrategy = config.schedulingStrategy;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._taskDefinition = config.taskDefinition;
        this._waitForSteadyState = config.waitForSteadyState;
        this._capacityProviderStrategy.internalValue = config.capacityProviderStrategy;
        this._deploymentCircuitBreaker.internalValue = config.deploymentCircuitBreaker;
        this._deploymentController.internalValue = config.deploymentController;
        this._loadBalancer.internalValue = config.loadBalancer;
        this._networkConfiguration.internalValue = config.networkConfiguration;
        this._orderedPlacementStrategy.internalValue = config.orderedPlacementStrategy;
        this._placementConstraints.internalValue = config.placementConstraints;
        this._serviceRegistries.internalValue = config.serviceRegistries;
        this._timeouts.internalValue = config.timeouts;
    }
    get cluster() {
        return this.getStringAttribute('cluster');
    }
    set cluster(value) {
        this._cluster = value;
    }
    resetCluster() {
        this._cluster = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clusterInput() {
        return this._cluster;
    }
    get deploymentMaximumPercent() {
        return this.getNumberAttribute('deployment_maximum_percent');
    }
    set deploymentMaximumPercent(value) {
        this._deploymentMaximumPercent = value;
    }
    resetDeploymentMaximumPercent() {
        this._deploymentMaximumPercent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deploymentMaximumPercentInput() {
        return this._deploymentMaximumPercent;
    }
    get deploymentMinimumHealthyPercent() {
        return this.getNumberAttribute('deployment_minimum_healthy_percent');
    }
    set deploymentMinimumHealthyPercent(value) {
        this._deploymentMinimumHealthyPercent = value;
    }
    resetDeploymentMinimumHealthyPercent() {
        this._deploymentMinimumHealthyPercent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deploymentMinimumHealthyPercentInput() {
        return this._deploymentMinimumHealthyPercent;
    }
    get desiredCount() {
        return this.getNumberAttribute('desired_count');
    }
    set desiredCount(value) {
        this._desiredCount = value;
    }
    resetDesiredCount() {
        this._desiredCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get desiredCountInput() {
        return this._desiredCount;
    }
    get enableEcsManagedTags() {
        return this.getBooleanAttribute('enable_ecs_managed_tags');
    }
    set enableEcsManagedTags(value) {
        this._enableEcsManagedTags = value;
    }
    resetEnableEcsManagedTags() {
        this._enableEcsManagedTags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableEcsManagedTagsInput() {
        return this._enableEcsManagedTags;
    }
    get enableExecuteCommand() {
        return this.getBooleanAttribute('enable_execute_command');
    }
    set enableExecuteCommand(value) {
        this._enableExecuteCommand = value;
    }
    resetEnableExecuteCommand() {
        this._enableExecuteCommand = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableExecuteCommandInput() {
        return this._enableExecuteCommand;
    }
    get forceNewDeployment() {
        return this.getBooleanAttribute('force_new_deployment');
    }
    set forceNewDeployment(value) {
        this._forceNewDeployment = value;
    }
    resetForceNewDeployment() {
        this._forceNewDeployment = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get forceNewDeploymentInput() {
        return this._forceNewDeployment;
    }
    get healthCheckGracePeriodSeconds() {
        return this.getNumberAttribute('health_check_grace_period_seconds');
    }
    set healthCheckGracePeriodSeconds(value) {
        this._healthCheckGracePeriodSeconds = value;
    }
    resetHealthCheckGracePeriodSeconds() {
        this._healthCheckGracePeriodSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get healthCheckGracePeriodSecondsInput() {
        return this._healthCheckGracePeriodSeconds;
    }
    get iamRole() {
        return this.getStringAttribute('iam_role');
    }
    set iamRole(value) {
        this._iamRole = value;
    }
    resetIamRole() {
        this._iamRole = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iamRoleInput() {
        return this._iamRole;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get launchType() {
        return this.getStringAttribute('launch_type');
    }
    set launchType(value) {
        this._launchType = value;
    }
    resetLaunchType() {
        this._launchType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get launchTypeInput() {
        return this._launchType;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get platformVersion() {
        return this.getStringAttribute('platform_version');
    }
    set platformVersion(value) {
        this._platformVersion = value;
    }
    resetPlatformVersion() {
        this._platformVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get platformVersionInput() {
        return this._platformVersion;
    }
    get propagateTags() {
        return this.getStringAttribute('propagate_tags');
    }
    set propagateTags(value) {
        this._propagateTags = value;
    }
    resetPropagateTags() {
        this._propagateTags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get propagateTagsInput() {
        return this._propagateTags;
    }
    get schedulingStrategy() {
        return this.getStringAttribute('scheduling_strategy');
    }
    set schedulingStrategy(value) {
        this._schedulingStrategy = value;
    }
    resetSchedulingStrategy() {
        this._schedulingStrategy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get schedulingStrategyInput() {
        return this._schedulingStrategy;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get taskDefinition() {
        return this.getStringAttribute('task_definition');
    }
    set taskDefinition(value) {
        this._taskDefinition = value;
    }
    resetTaskDefinition() {
        this._taskDefinition = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get taskDefinitionInput() {
        return this._taskDefinition;
    }
    get waitForSteadyState() {
        return this.getBooleanAttribute('wait_for_steady_state');
    }
    set waitForSteadyState(value) {
        this._waitForSteadyState = value;
    }
    resetWaitForSteadyState() {
        this._waitForSteadyState = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get waitForSteadyStateInput() {
        return this._waitForSteadyState;
    }
    get capacityProviderStrategy() {
        return this._capacityProviderStrategy;
    }
    putCapacityProviderStrategy(value) {
        this._capacityProviderStrategy.internalValue = value;
    }
    resetCapacityProviderStrategy() {
        this._capacityProviderStrategy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get capacityProviderStrategyInput() {
        return this._capacityProviderStrategy.internalValue;
    }
    get deploymentCircuitBreaker() {
        return this._deploymentCircuitBreaker;
    }
    putDeploymentCircuitBreaker(value) {
        this._deploymentCircuitBreaker.internalValue = value;
    }
    resetDeploymentCircuitBreaker() {
        this._deploymentCircuitBreaker.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deploymentCircuitBreakerInput() {
        return this._deploymentCircuitBreaker.internalValue;
    }
    get deploymentController() {
        return this._deploymentController;
    }
    putDeploymentController(value) {
        this._deploymentController.internalValue = value;
    }
    resetDeploymentController() {
        this._deploymentController.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deploymentControllerInput() {
        return this._deploymentController.internalValue;
    }
    get loadBalancer() {
        return this._loadBalancer;
    }
    putLoadBalancer(value) {
        this._loadBalancer.internalValue = value;
    }
    resetLoadBalancer() {
        this._loadBalancer.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get loadBalancerInput() {
        return this._loadBalancer.internalValue;
    }
    get networkConfiguration() {
        return this._networkConfiguration;
    }
    putNetworkConfiguration(value) {
        this._networkConfiguration.internalValue = value;
    }
    resetNetworkConfiguration() {
        this._networkConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkConfigurationInput() {
        return this._networkConfiguration.internalValue;
    }
    get orderedPlacementStrategy() {
        return this._orderedPlacementStrategy;
    }
    putOrderedPlacementStrategy(value) {
        this._orderedPlacementStrategy.internalValue = value;
    }
    resetOrderedPlacementStrategy() {
        this._orderedPlacementStrategy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get orderedPlacementStrategyInput() {
        return this._orderedPlacementStrategy.internalValue;
    }
    get placementConstraints() {
        return this._placementConstraints;
    }
    putPlacementConstraints(value) {
        this._placementConstraints.internalValue = value;
    }
    resetPlacementConstraints() {
        this._placementConstraints.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get placementConstraintsInput() {
        return this._placementConstraints.internalValue;
    }
    get serviceRegistries() {
        return this._serviceRegistries;
    }
    putServiceRegistries(value) {
        this._serviceRegistries.internalValue = value;
    }
    resetServiceRegistries() {
        this._serviceRegistries.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceRegistriesInput() {
        return this._serviceRegistries.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster: cdktf.stringToTerraform(this._cluster),
            deployment_maximum_percent: cdktf.numberToTerraform(this._deploymentMaximumPercent),
            deployment_minimum_healthy_percent: cdktf.numberToTerraform(this._deploymentMinimumHealthyPercent),
            desired_count: cdktf.numberToTerraform(this._desiredCount),
            enable_ecs_managed_tags: cdktf.booleanToTerraform(this._enableEcsManagedTags),
            enable_execute_command: cdktf.booleanToTerraform(this._enableExecuteCommand),
            force_new_deployment: cdktf.booleanToTerraform(this._forceNewDeployment),
            health_check_grace_period_seconds: cdktf.numberToTerraform(this._healthCheckGracePeriodSeconds),
            iam_role: cdktf.stringToTerraform(this._iamRole),
            id: cdktf.stringToTerraform(this._id),
            launch_type: cdktf.stringToTerraform(this._launchType),
            name: cdktf.stringToTerraform(this._name),
            platform_version: cdktf.stringToTerraform(this._platformVersion),
            propagate_tags: cdktf.stringToTerraform(this._propagateTags),
            scheduling_strategy: cdktf.stringToTerraform(this._schedulingStrategy),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            task_definition: cdktf.stringToTerraform(this._taskDefinition),
            wait_for_steady_state: cdktf.booleanToTerraform(this._waitForSteadyState),
            capacity_provider_strategy: cdktf.listMapper(ecsServiceCapacityProviderStrategyToTerraform, true)(this._capacityProviderStrategy.internalValue),
            deployment_circuit_breaker: ecsServiceDeploymentCircuitBreakerToTerraform(this._deploymentCircuitBreaker.internalValue),
            deployment_controller: ecsServiceDeploymentControllerToTerraform(this._deploymentController.internalValue),
            load_balancer: cdktf.listMapper(ecsServiceLoadBalancerToTerraform, true)(this._loadBalancer.internalValue),
            network_configuration: ecsServiceNetworkConfigurationToTerraform(this._networkConfiguration.internalValue),
            ordered_placement_strategy: cdktf.listMapper(ecsServiceOrderedPlacementStrategyToTerraform, true)(this._orderedPlacementStrategy.internalValue),
            placement_constraints: cdktf.listMapper(ecsServicePlacementConstraintsToTerraform, true)(this._placementConstraints.internalValue),
            service_registries: ecsServiceServiceRegistriesToTerraform(this._serviceRegistries.internalValue),
            timeouts: ecsServiceTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.EcsService = EcsService;
_p = JSII_RTTI_SYMBOL_1;
EcsService[_p] = { fqn: "@cdktf/provider-aws.ecsService.EcsService", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
EcsService.tfResourceType = "aws_ecs_service";
