// @cdktf/provider-aws/lib/data-aws-kms-key/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsKmsKey = exports.DataAwsKmsKeyMultiRegionConfigurationList = exports.DataAwsKmsKeyMultiRegionConfigurationOutputReference = exports.dataAwsKmsKeyMultiRegionConfigurationToTerraform = exports.DataAwsKmsKeyMultiRegionConfigurationReplicaKeysList = exports.DataAwsKmsKeyMultiRegionConfigurationReplicaKeysOutputReference = exports.dataAwsKmsKeyMultiRegionConfigurationReplicaKeysToTerraform = exports.DataAwsKmsKeyMultiRegionConfigurationPrimaryKeyList = exports.DataAwsKmsKeyMultiRegionConfigurationPrimaryKeyOutputReference = exports.dataAwsKmsKeyMultiRegionConfigurationPrimaryKeyToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsKmsKeyMultiRegionConfigurationPrimaryKeyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsKmsKeyMultiRegionConfigurationPrimaryKeyToTerraform = dataAwsKmsKeyMultiRegionConfigurationPrimaryKeyToTerraform;
class DataAwsKmsKeyMultiRegionConfigurationPrimaryKeyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // region - computed: true, optional: false, required: false
    get region() {
        return this.getStringAttribute('region');
    }
}
exports.DataAwsKmsKeyMultiRegionConfigurationPrimaryKeyOutputReference = DataAwsKmsKeyMultiRegionConfigurationPrimaryKeyOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsKmsKeyMultiRegionConfigurationPrimaryKeyOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsKmsKey.DataAwsKmsKeyMultiRegionConfigurationPrimaryKeyOutputReference", version: "10.0.12" };
class DataAwsKmsKeyMultiRegionConfigurationPrimaryKeyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsKmsKeyMultiRegionConfigurationPrimaryKeyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsKmsKeyMultiRegionConfigurationPrimaryKeyList = DataAwsKmsKeyMultiRegionConfigurationPrimaryKeyList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsKmsKeyMultiRegionConfigurationPrimaryKeyList[_b] = { fqn: "@cdktf/provider-aws.dataAwsKmsKey.DataAwsKmsKeyMultiRegionConfigurationPrimaryKeyList", version: "10.0.12" };
function dataAwsKmsKeyMultiRegionConfigurationReplicaKeysToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsKmsKeyMultiRegionConfigurationReplicaKeysToTerraform = dataAwsKmsKeyMultiRegionConfigurationReplicaKeysToTerraform;
class DataAwsKmsKeyMultiRegionConfigurationReplicaKeysOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // region - computed: true, optional: false, required: false
    get region() {
        return this.getStringAttribute('region');
    }
}
exports.DataAwsKmsKeyMultiRegionConfigurationReplicaKeysOutputReference = DataAwsKmsKeyMultiRegionConfigurationReplicaKeysOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsKmsKeyMultiRegionConfigurationReplicaKeysOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsKmsKey.DataAwsKmsKeyMultiRegionConfigurationReplicaKeysOutputReference", version: "10.0.12" };
class DataAwsKmsKeyMultiRegionConfigurationReplicaKeysList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsKmsKeyMultiRegionConfigurationReplicaKeysOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsKmsKeyMultiRegionConfigurationReplicaKeysList = DataAwsKmsKeyMultiRegionConfigurationReplicaKeysList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsKmsKeyMultiRegionConfigurationReplicaKeysList[_d] = { fqn: "@cdktf/provider-aws.dataAwsKmsKey.DataAwsKmsKeyMultiRegionConfigurationReplicaKeysList", version: "10.0.12" };
function dataAwsKmsKeyMultiRegionConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsKmsKeyMultiRegionConfigurationToTerraform = dataAwsKmsKeyMultiRegionConfigurationToTerraform;
class DataAwsKmsKeyMultiRegionConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // primary_key - computed: true, optional: false, required: false
        this._primaryKey = new DataAwsKmsKeyMultiRegionConfigurationPrimaryKeyList(this, "primary_key", false);
        // replica_keys - computed: true, optional: false, required: false
        this._replicaKeys = new DataAwsKmsKeyMultiRegionConfigurationReplicaKeysList(this, "replica_keys", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // multi_region_key_type - computed: true, optional: false, required: false
    get multiRegionKeyType() {
        return this.getStringAttribute('multi_region_key_type');
    }
    get primaryKey() {
        return this._primaryKey;
    }
    get replicaKeys() {
        return this._replicaKeys;
    }
}
exports.DataAwsKmsKeyMultiRegionConfigurationOutputReference = DataAwsKmsKeyMultiRegionConfigurationOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsKmsKeyMultiRegionConfigurationOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsKmsKey.DataAwsKmsKeyMultiRegionConfigurationOutputReference", version: "10.0.12" };
class DataAwsKmsKeyMultiRegionConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsKmsKeyMultiRegionConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsKmsKeyMultiRegionConfigurationList = DataAwsKmsKeyMultiRegionConfigurationList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsKmsKeyMultiRegionConfigurationList[_f] = { fqn: "@cdktf/provider-aws.dataAwsKmsKey.DataAwsKmsKeyMultiRegionConfigurationList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/kms_key aws_kms_key}
*/
class DataAwsKmsKey extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/kms_key aws_kms_key} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsKmsKeyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_kms_key',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // multi_region_configuration - computed: true, optional: false, required: false
        this._multiRegionConfiguration = new DataAwsKmsKeyMultiRegionConfigurationList(this, "multi_region_configuration", false);
        this._grantTokens = config.grantTokens;
        this._id = config.id;
        this._keyId = config.keyId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // aws_account_id - computed: true, optional: false, required: false
    get awsAccountId() {
        return this.getStringAttribute('aws_account_id');
    }
    // creation_date - computed: true, optional: false, required: false
    get creationDate() {
        return this.getStringAttribute('creation_date');
    }
    // customer_master_key_spec - computed: true, optional: false, required: false
    get customerMasterKeySpec() {
        return this.getStringAttribute('customer_master_key_spec');
    }
    // deletion_date - computed: true, optional: false, required: false
    get deletionDate() {
        return this.getStringAttribute('deletion_date');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    // expiration_model - computed: true, optional: false, required: false
    get expirationModel() {
        return this.getStringAttribute('expiration_model');
    }
    get grantTokens() {
        return this.getListAttribute('grant_tokens');
    }
    set grantTokens(value) {
        this._grantTokens = value;
    }
    resetGrantTokens() {
        this._grantTokens = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get grantTokensInput() {
        return this._grantTokens;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get keyId() {
        return this.getStringAttribute('key_id');
    }
    set keyId(value) {
        this._keyId = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyIdInput() {
        return this._keyId;
    }
    // key_manager - computed: true, optional: false, required: false
    get keyManager() {
        return this.getStringAttribute('key_manager');
    }
    // key_state - computed: true, optional: false, required: false
    get keyState() {
        return this.getStringAttribute('key_state');
    }
    // key_usage - computed: true, optional: false, required: false
    get keyUsage() {
        return this.getStringAttribute('key_usage');
    }
    // multi_region - computed: true, optional: false, required: false
    get multiRegion() {
        return this.getBooleanAttribute('multi_region');
    }
    get multiRegionConfiguration() {
        return this._multiRegionConfiguration;
    }
    // origin - computed: true, optional: false, required: false
    get origin() {
        return this.getStringAttribute('origin');
    }
    // valid_to - computed: true, optional: false, required: false
    get validTo() {
        return this.getStringAttribute('valid_to');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            grant_tokens: cdktf.listMapper(cdktf.stringToTerraform, false)(this._grantTokens),
            id: cdktf.stringToTerraform(this._id),
            key_id: cdktf.stringToTerraform(this._keyId),
        };
    }
}
exports.DataAwsKmsKey = DataAwsKmsKey;
_g = JSII_RTTI_SYMBOL_1;
DataAwsKmsKey[_g] = { fqn: "@cdktf/provider-aws.dataAwsKmsKey.DataAwsKmsKey", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsKmsKey.tfResourceType = "aws_kms_key";
