// @cdktf/provider-aws/lib/route53-key-signing-key/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Route53KeySigningKey = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/route53_key_signing_key aws_route53_key_signing_key}
*/
class Route53KeySigningKey extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/route53_key_signing_key aws_route53_key_signing_key} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options Route53KeySigningKeyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_route53_key_signing_key',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._hostedZoneId = config.hostedZoneId;
        this._id = config.id;
        this._keyManagementServiceArn = config.keyManagementServiceArn;
        this._name = config.name;
        this._status = config.status;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // digest_algorithm_mnemonic - computed: true, optional: false, required: false
    get digestAlgorithmMnemonic() {
        return this.getStringAttribute('digest_algorithm_mnemonic');
    }
    // digest_algorithm_type - computed: true, optional: false, required: false
    get digestAlgorithmType() {
        return this.getNumberAttribute('digest_algorithm_type');
    }
    // digest_value - computed: true, optional: false, required: false
    get digestValue() {
        return this.getStringAttribute('digest_value');
    }
    // dnskey_record - computed: true, optional: false, required: false
    get dnskeyRecord() {
        return this.getStringAttribute('dnskey_record');
    }
    // ds_record - computed: true, optional: false, required: false
    get dsRecord() {
        return this.getStringAttribute('ds_record');
    }
    // flag - computed: true, optional: false, required: false
    get flag() {
        return this.getNumberAttribute('flag');
    }
    get hostedZoneId() {
        return this.getStringAttribute('hosted_zone_id');
    }
    set hostedZoneId(value) {
        this._hostedZoneId = value;
    }
    // Temporarily expose input value. Use with caution.
    get hostedZoneIdInput() {
        return this._hostedZoneId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get keyManagementServiceArn() {
        return this.getStringAttribute('key_management_service_arn');
    }
    set keyManagementServiceArn(value) {
        this._keyManagementServiceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyManagementServiceArnInput() {
        return this._keyManagementServiceArn;
    }
    // key_tag - computed: true, optional: false, required: false
    get keyTag() {
        return this.getNumberAttribute('key_tag');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // public_key - computed: true, optional: false, required: false
    get publicKey() {
        return this.getStringAttribute('public_key');
    }
    // signing_algorithm_mnemonic - computed: true, optional: false, required: false
    get signingAlgorithmMnemonic() {
        return this.getStringAttribute('signing_algorithm_mnemonic');
    }
    // signing_algorithm_type - computed: true, optional: false, required: false
    get signingAlgorithmType() {
        return this.getNumberAttribute('signing_algorithm_type');
    }
    get status() {
        return this.getStringAttribute('status');
    }
    set status(value) {
        this._status = value;
    }
    resetStatus() {
        this._status = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get statusInput() {
        return this._status;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            hosted_zone_id: cdktf.stringToTerraform(this._hostedZoneId),
            id: cdktf.stringToTerraform(this._id),
            key_management_service_arn: cdktf.stringToTerraform(this._keyManagementServiceArn),
            name: cdktf.stringToTerraform(this._name),
            status: cdktf.stringToTerraform(this._status),
        };
    }
}
exports.Route53KeySigningKey = Route53KeySigningKey;
_a = JSII_RTTI_SYMBOL_1;
Route53KeySigningKey[_a] = { fqn: "@cdktf/provider-aws.route53KeySigningKey.Route53KeySigningKey", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
Route53KeySigningKey.tfResourceType = "aws_route53_key_signing_key";
