// @cdktf/provider-aws/lib/data-aws-efs-file-system/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsEfsFileSystem = exports.DataAwsEfsFileSystemLifecyclePolicyList = exports.DataAwsEfsFileSystemLifecyclePolicyOutputReference = exports.dataAwsEfsFileSystemLifecyclePolicyToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsEfsFileSystemLifecyclePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEfsFileSystemLifecyclePolicyToTerraform = dataAwsEfsFileSystemLifecyclePolicyToTerraform;
class DataAwsEfsFileSystemLifecyclePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // transition_to_ia - computed: true, optional: false, required: false
    get transitionToIa() {
        return this.getStringAttribute('transition_to_ia');
    }
    // transition_to_primary_storage_class - computed: true, optional: false, required: false
    get transitionToPrimaryStorageClass() {
        return this.getStringAttribute('transition_to_primary_storage_class');
    }
}
exports.DataAwsEfsFileSystemLifecyclePolicyOutputReference = DataAwsEfsFileSystemLifecyclePolicyOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsEfsFileSystemLifecyclePolicyOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsEfsFileSystem.DataAwsEfsFileSystemLifecyclePolicyOutputReference", version: "10.0.12" };
class DataAwsEfsFileSystemLifecyclePolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEfsFileSystemLifecyclePolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEfsFileSystemLifecyclePolicyList = DataAwsEfsFileSystemLifecyclePolicyList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsEfsFileSystemLifecyclePolicyList[_b] = { fqn: "@cdktf/provider-aws.dataAwsEfsFileSystem.DataAwsEfsFileSystemLifecyclePolicyList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/efs_file_system aws_efs_file_system}
*/
class DataAwsEfsFileSystem extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/efs_file_system aws_efs_file_system} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsEfsFileSystemConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_efs_file_system',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // lifecycle_policy - computed: true, optional: false, required: false
        this._lifecyclePolicy = new DataAwsEfsFileSystemLifecyclePolicyList(this, "lifecycle_policy", false);
        this._creationToken = config.creationToken;
        this._fileSystemId = config.fileSystemId;
        this._id = config.id;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // availability_zone_id - computed: true, optional: false, required: false
    get availabilityZoneId() {
        return this.getStringAttribute('availability_zone_id');
    }
    // availability_zone_name - computed: true, optional: false, required: false
    get availabilityZoneName() {
        return this.getStringAttribute('availability_zone_name');
    }
    get creationToken() {
        return this.getStringAttribute('creation_token');
    }
    set creationToken(value) {
        this._creationToken = value;
    }
    resetCreationToken() {
        this._creationToken = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get creationTokenInput() {
        return this._creationToken;
    }
    // dns_name - computed: true, optional: false, required: false
    get dnsName() {
        return this.getStringAttribute('dns_name');
    }
    // encrypted - computed: true, optional: false, required: false
    get encrypted() {
        return this.getBooleanAttribute('encrypted');
    }
    get fileSystemId() {
        return this.getStringAttribute('file_system_id');
    }
    set fileSystemId(value) {
        this._fileSystemId = value;
    }
    resetFileSystemId() {
        this._fileSystemId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fileSystemIdInput() {
        return this._fileSystemId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // kms_key_id - computed: true, optional: false, required: false
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    get lifecyclePolicy() {
        return this._lifecyclePolicy;
    }
    // performance_mode - computed: true, optional: false, required: false
    get performanceMode() {
        return this.getStringAttribute('performance_mode');
    }
    // provisioned_throughput_in_mibps - computed: true, optional: false, required: false
    get provisionedThroughputInMibps() {
        return this.getNumberAttribute('provisioned_throughput_in_mibps');
    }
    // size_in_bytes - computed: true, optional: false, required: false
    get sizeInBytes() {
        return this.getNumberAttribute('size_in_bytes');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // throughput_mode - computed: true, optional: false, required: false
    get throughputMode() {
        return this.getStringAttribute('throughput_mode');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            creation_token: cdktf.stringToTerraform(this._creationToken),
            file_system_id: cdktf.stringToTerraform(this._fileSystemId),
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
}
exports.DataAwsEfsFileSystem = DataAwsEfsFileSystem;
_c = JSII_RTTI_SYMBOL_1;
DataAwsEfsFileSystem[_c] = { fqn: "@cdktf/provider-aws.dataAwsEfsFileSystem.DataAwsEfsFileSystem", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsEfsFileSystem.tfResourceType = "aws_efs_file_system";
