// @cdktf/provider-aws/lib/db-instance-automated-backups-replication/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DbInstanceAutomatedBackupsReplication = exports.DbInstanceAutomatedBackupsReplicationTimeoutsOutputReference = exports.dbInstanceAutomatedBackupsReplicationTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dbInstanceAutomatedBackupsReplicationTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.dbInstanceAutomatedBackupsReplicationTimeoutsToTerraform = dbInstanceAutomatedBackupsReplicationTimeoutsToTerraform;
class DbInstanceAutomatedBackupsReplicationTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
}
exports.DbInstanceAutomatedBackupsReplicationTimeoutsOutputReference = DbInstanceAutomatedBackupsReplicationTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DbInstanceAutomatedBackupsReplicationTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.dbInstanceAutomatedBackupsReplication.DbInstanceAutomatedBackupsReplicationTimeoutsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/db_instance_automated_backups_replication aws_db_instance_automated_backups_replication}
*/
class DbInstanceAutomatedBackupsReplication extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/db_instance_automated_backups_replication aws_db_instance_automated_backups_replication} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DbInstanceAutomatedBackupsReplicationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_db_instance_automated_backups_replication',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new DbInstanceAutomatedBackupsReplicationTimeoutsOutputReference(this, "timeouts");
        this._id = config.id;
        this._kmsKeyId = config.kmsKeyId;
        this._preSignedUrl = config.preSignedUrl;
        this._retentionPeriod = config.retentionPeriod;
        this._sourceDbInstanceArn = config.sourceDbInstanceArn;
        this._timeouts.internalValue = config.timeouts;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    resetKmsKeyId() {
        this._kmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
    get preSignedUrl() {
        return this.getStringAttribute('pre_signed_url');
    }
    set preSignedUrl(value) {
        this._preSignedUrl = value;
    }
    resetPreSignedUrl() {
        this._preSignedUrl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get preSignedUrlInput() {
        return this._preSignedUrl;
    }
    get retentionPeriod() {
        return this.getNumberAttribute('retention_period');
    }
    set retentionPeriod(value) {
        this._retentionPeriod = value;
    }
    resetRetentionPeriod() {
        this._retentionPeriod = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get retentionPeriodInput() {
        return this._retentionPeriod;
    }
    get sourceDbInstanceArn() {
        return this.getStringAttribute('source_db_instance_arn');
    }
    set sourceDbInstanceArn(value) {
        this._sourceDbInstanceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get sourceDbInstanceArnInput() {
        return this._sourceDbInstanceArn;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            kms_key_id: cdktf.stringToTerraform(this._kmsKeyId),
            pre_signed_url: cdktf.stringToTerraform(this._preSignedUrl),
            retention_period: cdktf.numberToTerraform(this._retentionPeriod),
            source_db_instance_arn: cdktf.stringToTerraform(this._sourceDbInstanceArn),
            timeouts: dbInstanceAutomatedBackupsReplicationTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.DbInstanceAutomatedBackupsReplication = DbInstanceAutomatedBackupsReplication;
_b = JSII_RTTI_SYMBOL_1;
DbInstanceAutomatedBackupsReplication[_b] = { fqn: "@cdktf/provider-aws.dbInstanceAutomatedBackupsReplication.DbInstanceAutomatedBackupsReplication", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DbInstanceAutomatedBackupsReplication.tfResourceType = "aws_db_instance_automated_backups_replication";
