// @cdktf/provider-aws/lib/sqs-queue-redrive-policy/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SqsQueueRedrivePolicy = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/sqs_queue_redrive_policy aws_sqs_queue_redrive_policy}
*/
class SqsQueueRedrivePolicy extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/sqs_queue_redrive_policy aws_sqs_queue_redrive_policy} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SqsQueueRedrivePolicyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_sqs_queue_redrive_policy',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._queueUrl = config.queueUrl;
        this._redrivePolicy = config.redrivePolicy;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get queueUrl() {
        return this.getStringAttribute('queue_url');
    }
    set queueUrl(value) {
        this._queueUrl = value;
    }
    // Temporarily expose input value. Use with caution.
    get queueUrlInput() {
        return this._queueUrl;
    }
    get redrivePolicy() {
        return this.getStringAttribute('redrive_policy');
    }
    set redrivePolicy(value) {
        this._redrivePolicy = value;
    }
    // Temporarily expose input value. Use with caution.
    get redrivePolicyInput() {
        return this._redrivePolicy;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            queue_url: cdktf.stringToTerraform(this._queueUrl),
            redrive_policy: cdktf.stringToTerraform(this._redrivePolicy),
        };
    }
}
exports.SqsQueueRedrivePolicy = SqsQueueRedrivePolicy;
_a = JSII_RTTI_SYMBOL_1;
SqsQueueRedrivePolicy[_a] = { fqn: "@cdktf/provider-aws.sqsQueueRedrivePolicy.SqsQueueRedrivePolicy", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
SqsQueueRedrivePolicy.tfResourceType = "aws_sqs_queue_redrive_policy";
