// @cdktf/provider-aws/lib/redshift-usage-limit/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedshiftUsageLimit = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/redshift_usage_limit aws_redshift_usage_limit}
*/
class RedshiftUsageLimit extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/redshift_usage_limit aws_redshift_usage_limit} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options RedshiftUsageLimitConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_redshift_usage_limit',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._amount = config.amount;
        this._breachAction = config.breachAction;
        this._clusterIdentifier = config.clusterIdentifier;
        this._featureType = config.featureType;
        this._id = config.id;
        this._limitType = config.limitType;
        this._period = config.period;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    get amount() {
        return this.getNumberAttribute('amount');
    }
    set amount(value) {
        this._amount = value;
    }
    // Temporarily expose input value. Use with caution.
    get amountInput() {
        return this._amount;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get breachAction() {
        return this.getStringAttribute('breach_action');
    }
    set breachAction(value) {
        this._breachAction = value;
    }
    resetBreachAction() {
        this._breachAction = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get breachActionInput() {
        return this._breachAction;
    }
    get clusterIdentifier() {
        return this.getStringAttribute('cluster_identifier');
    }
    set clusterIdentifier(value) {
        this._clusterIdentifier = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterIdentifierInput() {
        return this._clusterIdentifier;
    }
    get featureType() {
        return this.getStringAttribute('feature_type');
    }
    set featureType(value) {
        this._featureType = value;
    }
    // Temporarily expose input value. Use with caution.
    get featureTypeInput() {
        return this._featureType;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get limitType() {
        return this.getStringAttribute('limit_type');
    }
    set limitType(value) {
        this._limitType = value;
    }
    // Temporarily expose input value. Use with caution.
    get limitTypeInput() {
        return this._limitType;
    }
    get period() {
        return this.getStringAttribute('period');
    }
    set period(value) {
        this._period = value;
    }
    resetPeriod() {
        this._period = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get periodInput() {
        return this._period;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            amount: cdktf.numberToTerraform(this._amount),
            breach_action: cdktf.stringToTerraform(this._breachAction),
            cluster_identifier: cdktf.stringToTerraform(this._clusterIdentifier),
            feature_type: cdktf.stringToTerraform(this._featureType),
            id: cdktf.stringToTerraform(this._id),
            limit_type: cdktf.stringToTerraform(this._limitType),
            period: cdktf.stringToTerraform(this._period),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
        };
    }
}
exports.RedshiftUsageLimit = RedshiftUsageLimit;
_a = JSII_RTTI_SYMBOL_1;
RedshiftUsageLimit[_a] = { fqn: "@cdktf/provider-aws.redshiftUsageLimit.RedshiftUsageLimit", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
RedshiftUsageLimit.tfResourceType = "aws_redshift_usage_limit";
