// @cdktf/provider-aws/lib/data-aws-iam-session-context/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsIamSessionContext = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/iam_session_context aws_iam_session_context}
*/
class DataAwsIamSessionContext extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/iam_session_context aws_iam_session_context} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsIamSessionContextConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iam_session_context',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._arn = config.arn;
        this._id = config.id;
    }
    get arn() {
        return this.getStringAttribute('arn');
    }
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    get arnInput() {
        return this._arn;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // issuer_arn - computed: true, optional: false, required: false
    get issuerArn() {
        return this.getStringAttribute('issuer_arn');
    }
    // issuer_id - computed: true, optional: false, required: false
    get issuerId() {
        return this.getStringAttribute('issuer_id');
    }
    // issuer_name - computed: true, optional: false, required: false
    get issuerName() {
        return this.getStringAttribute('issuer_name');
    }
    // session_name - computed: true, optional: false, required: false
    get sessionName() {
        return this.getStringAttribute('session_name');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            arn: cdktf.stringToTerraform(this._arn),
            id: cdktf.stringToTerraform(this._id),
        };
    }
}
exports.DataAwsIamSessionContext = DataAwsIamSessionContext;
_a = JSII_RTTI_SYMBOL_1;
DataAwsIamSessionContext[_a] = { fqn: "@cdktf/provider-aws.dataAwsIamSessionContext.DataAwsIamSessionContext", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsIamSessionContext.tfResourceType = "aws_iam_session_context";
