// @cdktf/provider-aws/lib/msk-cluster/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MskCluster = exports.MskClusterTimeoutsOutputReference = exports.mskClusterTimeoutsToTerraform = exports.MskClusterOpenMonitoringOutputReference = exports.mskClusterOpenMonitoringToTerraform = exports.MskClusterOpenMonitoringPrometheusOutputReference = exports.mskClusterOpenMonitoringPrometheusToTerraform = exports.MskClusterOpenMonitoringPrometheusNodeExporterOutputReference = exports.mskClusterOpenMonitoringPrometheusNodeExporterToTerraform = exports.MskClusterOpenMonitoringPrometheusJmxExporterOutputReference = exports.mskClusterOpenMonitoringPrometheusJmxExporterToTerraform = exports.MskClusterLoggingInfoOutputReference = exports.mskClusterLoggingInfoToTerraform = exports.MskClusterLoggingInfoBrokerLogsOutputReference = exports.mskClusterLoggingInfoBrokerLogsToTerraform = exports.MskClusterLoggingInfoBrokerLogsS3OutputReference = exports.mskClusterLoggingInfoBrokerLogsS3ToTerraform = exports.MskClusterLoggingInfoBrokerLogsFirehoseOutputReference = exports.mskClusterLoggingInfoBrokerLogsFirehoseToTerraform = exports.MskClusterLoggingInfoBrokerLogsCloudwatchLogsOutputReference = exports.mskClusterLoggingInfoBrokerLogsCloudwatchLogsToTerraform = exports.MskClusterEncryptionInfoOutputReference = exports.mskClusterEncryptionInfoToTerraform = exports.MskClusterEncryptionInfoEncryptionInTransitOutputReference = exports.mskClusterEncryptionInfoEncryptionInTransitToTerraform = exports.MskClusterConfigurationInfoOutputReference = exports.mskClusterConfigurationInfoToTerraform = exports.MskClusterClientAuthenticationOutputReference = exports.mskClusterClientAuthenticationToTerraform = exports.MskClusterClientAuthenticationTlsOutputReference = exports.mskClusterClientAuthenticationTlsToTerraform = exports.MskClusterClientAuthenticationSaslOutputReference = exports.mskClusterClientAuthenticationSaslToTerraform = exports.MskClusterBrokerNodeGroupInfoOutputReference = exports.mskClusterBrokerNodeGroupInfoToTerraform = exports.MskClusterBrokerNodeGroupInfoStorageInfoOutputReference = exports.mskClusterBrokerNodeGroupInfoStorageInfoToTerraform = exports.MskClusterBrokerNodeGroupInfoStorageInfoEbsStorageInfoOutputReference = exports.mskClusterBrokerNodeGroupInfoStorageInfoEbsStorageInfoToTerraform = exports.MskClusterBrokerNodeGroupInfoStorageInfoEbsStorageInfoProvisionedThroughputOutputReference = exports.mskClusterBrokerNodeGroupInfoStorageInfoEbsStorageInfoProvisionedThroughputToTerraform = exports.MskClusterBrokerNodeGroupInfoConnectivityInfoOutputReference = exports.mskClusterBrokerNodeGroupInfoConnectivityInfoToTerraform = exports.MskClusterBrokerNodeGroupInfoConnectivityInfoPublicAccessOutputReference = exports.mskClusterBrokerNodeGroupInfoConnectivityInfoPublicAccessToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function mskClusterBrokerNodeGroupInfoConnectivityInfoPublicAccessToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.mskClusterBrokerNodeGroupInfoConnectivityInfoPublicAccessToTerraform = mskClusterBrokerNodeGroupInfoConnectivityInfoPublicAccessToTerraform;
class MskClusterBrokerNodeGroupInfoConnectivityInfoPublicAccessOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._type = value.type;
        }
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.MskClusterBrokerNodeGroupInfoConnectivityInfoPublicAccessOutputReference = MskClusterBrokerNodeGroupInfoConnectivityInfoPublicAccessOutputReference;
_a = JSII_RTTI_SYMBOL_1;
MskClusterBrokerNodeGroupInfoConnectivityInfoPublicAccessOutputReference[_a] = { fqn: "@cdktf/provider-aws.mskCluster.MskClusterBrokerNodeGroupInfoConnectivityInfoPublicAccessOutputReference", version: "10.0.12" };
function mskClusterBrokerNodeGroupInfoConnectivityInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        public_access: mskClusterBrokerNodeGroupInfoConnectivityInfoPublicAccessToTerraform(struct.publicAccess),
    };
}
exports.mskClusterBrokerNodeGroupInfoConnectivityInfoToTerraform = mskClusterBrokerNodeGroupInfoConnectivityInfoToTerraform;
class MskClusterBrokerNodeGroupInfoConnectivityInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // public_access - computed: false, optional: true, required: false
        this._publicAccess = new MskClusterBrokerNodeGroupInfoConnectivityInfoPublicAccessOutputReference(this, "public_access");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._publicAccess?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.publicAccess = this._publicAccess?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._publicAccess.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._publicAccess.internalValue = value.publicAccess;
        }
    }
    get publicAccess() {
        return this._publicAccess;
    }
    putPublicAccess(value) {
        this._publicAccess.internalValue = value;
    }
    resetPublicAccess() {
        this._publicAccess.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get publicAccessInput() {
        return this._publicAccess.internalValue;
    }
}
exports.MskClusterBrokerNodeGroupInfoConnectivityInfoOutputReference = MskClusterBrokerNodeGroupInfoConnectivityInfoOutputReference;
_b = JSII_RTTI_SYMBOL_1;
MskClusterBrokerNodeGroupInfoConnectivityInfoOutputReference[_b] = { fqn: "@cdktf/provider-aws.mskCluster.MskClusterBrokerNodeGroupInfoConnectivityInfoOutputReference", version: "10.0.12" };
function mskClusterBrokerNodeGroupInfoStorageInfoEbsStorageInfoProvisionedThroughputToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        volume_throughput: cdktf.numberToTerraform(struct.volumeThroughput),
    };
}
exports.mskClusterBrokerNodeGroupInfoStorageInfoEbsStorageInfoProvisionedThroughputToTerraform = mskClusterBrokerNodeGroupInfoStorageInfoEbsStorageInfoProvisionedThroughputToTerraform;
class MskClusterBrokerNodeGroupInfoStorageInfoEbsStorageInfoProvisionedThroughputOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._volumeThroughput !== undefined) {
            hasAnyValues = true;
            internalValueResult.volumeThroughput = this._volumeThroughput;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._volumeThroughput = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._volumeThroughput = value.volumeThroughput;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get volumeThroughput() {
        return this.getNumberAttribute('volume_throughput');
    }
    set volumeThroughput(value) {
        this._volumeThroughput = value;
    }
    resetVolumeThroughput() {
        this._volumeThroughput = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumeThroughputInput() {
        return this._volumeThroughput;
    }
}
exports.MskClusterBrokerNodeGroupInfoStorageInfoEbsStorageInfoProvisionedThroughputOutputReference = MskClusterBrokerNodeGroupInfoStorageInfoEbsStorageInfoProvisionedThroughputOutputReference;
_c = JSII_RTTI_SYMBOL_1;
MskClusterBrokerNodeGroupInfoStorageInfoEbsStorageInfoProvisionedThroughputOutputReference[_c] = { fqn: "@cdktf/provider-aws.mskCluster.MskClusterBrokerNodeGroupInfoStorageInfoEbsStorageInfoProvisionedThroughputOutputReference", version: "10.0.12" };
function mskClusterBrokerNodeGroupInfoStorageInfoEbsStorageInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        volume_size: cdktf.numberToTerraform(struct.volumeSize),
        provisioned_throughput: mskClusterBrokerNodeGroupInfoStorageInfoEbsStorageInfoProvisionedThroughputToTerraform(struct.provisionedThroughput),
    };
}
exports.mskClusterBrokerNodeGroupInfoStorageInfoEbsStorageInfoToTerraform = mskClusterBrokerNodeGroupInfoStorageInfoEbsStorageInfoToTerraform;
class MskClusterBrokerNodeGroupInfoStorageInfoEbsStorageInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // provisioned_throughput - computed: false, optional: true, required: false
        this._provisionedThroughput = new MskClusterBrokerNodeGroupInfoStorageInfoEbsStorageInfoProvisionedThroughputOutputReference(this, "provisioned_throughput");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._volumeSize !== undefined) {
            hasAnyValues = true;
            internalValueResult.volumeSize = this._volumeSize;
        }
        if (this._provisionedThroughput?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.provisionedThroughput = this._provisionedThroughput?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._volumeSize = undefined;
            this._provisionedThroughput.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._volumeSize = value.volumeSize;
            this._provisionedThroughput.internalValue = value.provisionedThroughput;
        }
    }
    get volumeSize() {
        return this.getNumberAttribute('volume_size');
    }
    set volumeSize(value) {
        this._volumeSize = value;
    }
    resetVolumeSize() {
        this._volumeSize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumeSizeInput() {
        return this._volumeSize;
    }
    get provisionedThroughput() {
        return this._provisionedThroughput;
    }
    putProvisionedThroughput(value) {
        this._provisionedThroughput.internalValue = value;
    }
    resetProvisionedThroughput() {
        this._provisionedThroughput.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get provisionedThroughputInput() {
        return this._provisionedThroughput.internalValue;
    }
}
exports.MskClusterBrokerNodeGroupInfoStorageInfoEbsStorageInfoOutputReference = MskClusterBrokerNodeGroupInfoStorageInfoEbsStorageInfoOutputReference;
_d = JSII_RTTI_SYMBOL_1;
MskClusterBrokerNodeGroupInfoStorageInfoEbsStorageInfoOutputReference[_d] = { fqn: "@cdktf/provider-aws.mskCluster.MskClusterBrokerNodeGroupInfoStorageInfoEbsStorageInfoOutputReference", version: "10.0.12" };
function mskClusterBrokerNodeGroupInfoStorageInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ebs_storage_info: mskClusterBrokerNodeGroupInfoStorageInfoEbsStorageInfoToTerraform(struct.ebsStorageInfo),
    };
}
exports.mskClusterBrokerNodeGroupInfoStorageInfoToTerraform = mskClusterBrokerNodeGroupInfoStorageInfoToTerraform;
class MskClusterBrokerNodeGroupInfoStorageInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // ebs_storage_info - computed: false, optional: true, required: false
        this._ebsStorageInfo = new MskClusterBrokerNodeGroupInfoStorageInfoEbsStorageInfoOutputReference(this, "ebs_storage_info");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._ebsStorageInfo?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.ebsStorageInfo = this._ebsStorageInfo?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._ebsStorageInfo.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._ebsStorageInfo.internalValue = value.ebsStorageInfo;
        }
    }
    get ebsStorageInfo() {
        return this._ebsStorageInfo;
    }
    putEbsStorageInfo(value) {
        this._ebsStorageInfo.internalValue = value;
    }
    resetEbsStorageInfo() {
        this._ebsStorageInfo.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ebsStorageInfoInput() {
        return this._ebsStorageInfo.internalValue;
    }
}
exports.MskClusterBrokerNodeGroupInfoStorageInfoOutputReference = MskClusterBrokerNodeGroupInfoStorageInfoOutputReference;
_e = JSII_RTTI_SYMBOL_1;
MskClusterBrokerNodeGroupInfoStorageInfoOutputReference[_e] = { fqn: "@cdktf/provider-aws.mskCluster.MskClusterBrokerNodeGroupInfoStorageInfoOutputReference", version: "10.0.12" };
function mskClusterBrokerNodeGroupInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        az_distribution: cdktf.stringToTerraform(struct.azDistribution),
        client_subnets: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.clientSubnets),
        ebs_volume_size: cdktf.numberToTerraform(struct.ebsVolumeSize),
        instance_type: cdktf.stringToTerraform(struct.instanceType),
        security_groups: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.securityGroups),
        connectivity_info: mskClusterBrokerNodeGroupInfoConnectivityInfoToTerraform(struct.connectivityInfo),
        storage_info: mskClusterBrokerNodeGroupInfoStorageInfoToTerraform(struct.storageInfo),
    };
}
exports.mskClusterBrokerNodeGroupInfoToTerraform = mskClusterBrokerNodeGroupInfoToTerraform;
class MskClusterBrokerNodeGroupInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // connectivity_info - computed: false, optional: true, required: false
        this._connectivityInfo = new MskClusterBrokerNodeGroupInfoConnectivityInfoOutputReference(this, "connectivity_info");
        // storage_info - computed: false, optional: true, required: false
        this._storageInfo = new MskClusterBrokerNodeGroupInfoStorageInfoOutputReference(this, "storage_info");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._azDistribution !== undefined) {
            hasAnyValues = true;
            internalValueResult.azDistribution = this._azDistribution;
        }
        if (this._clientSubnets !== undefined) {
            hasAnyValues = true;
            internalValueResult.clientSubnets = this._clientSubnets;
        }
        if (this._ebsVolumeSize !== undefined) {
            hasAnyValues = true;
            internalValueResult.ebsVolumeSize = this._ebsVolumeSize;
        }
        if (this._instanceType !== undefined) {
            hasAnyValues = true;
            internalValueResult.instanceType = this._instanceType;
        }
        if (this._securityGroups !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroups = this._securityGroups;
        }
        if (this._connectivityInfo?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.connectivityInfo = this._connectivityInfo?.internalValue;
        }
        if (this._storageInfo?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.storageInfo = this._storageInfo?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._azDistribution = undefined;
            this._clientSubnets = undefined;
            this._ebsVolumeSize = undefined;
            this._instanceType = undefined;
            this._securityGroups = undefined;
            this._connectivityInfo.internalValue = undefined;
            this._storageInfo.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._azDistribution = value.azDistribution;
            this._clientSubnets = value.clientSubnets;
            this._ebsVolumeSize = value.ebsVolumeSize;
            this._instanceType = value.instanceType;
            this._securityGroups = value.securityGroups;
            this._connectivityInfo.internalValue = value.connectivityInfo;
            this._storageInfo.internalValue = value.storageInfo;
        }
    }
    get azDistribution() {
        return this.getStringAttribute('az_distribution');
    }
    set azDistribution(value) {
        this._azDistribution = value;
    }
    resetAzDistribution() {
        this._azDistribution = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get azDistributionInput() {
        return this._azDistribution;
    }
    get clientSubnets() {
        return cdktf.Fn.tolist(this.getListAttribute('client_subnets'));
    }
    set clientSubnets(value) {
        this._clientSubnets = value;
    }
    // Temporarily expose input value. Use with caution.
    get clientSubnetsInput() {
        return this._clientSubnets;
    }
    get ebsVolumeSize() {
        return this.getNumberAttribute('ebs_volume_size');
    }
    set ebsVolumeSize(value) {
        this._ebsVolumeSize = value;
    }
    resetEbsVolumeSize() {
        this._ebsVolumeSize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ebsVolumeSizeInput() {
        return this._ebsVolumeSize;
    }
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    set instanceType(value) {
        this._instanceType = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceTypeInput() {
        return this._instanceType;
    }
    get securityGroups() {
        return cdktf.Fn.tolist(this.getListAttribute('security_groups'));
    }
    set securityGroups(value) {
        this._securityGroups = value;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupsInput() {
        return this._securityGroups;
    }
    get connectivityInfo() {
        return this._connectivityInfo;
    }
    putConnectivityInfo(value) {
        this._connectivityInfo.internalValue = value;
    }
    resetConnectivityInfo() {
        this._connectivityInfo.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get connectivityInfoInput() {
        return this._connectivityInfo.internalValue;
    }
    get storageInfo() {
        return this._storageInfo;
    }
    putStorageInfo(value) {
        this._storageInfo.internalValue = value;
    }
    resetStorageInfo() {
        this._storageInfo.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get storageInfoInput() {
        return this._storageInfo.internalValue;
    }
}
exports.MskClusterBrokerNodeGroupInfoOutputReference = MskClusterBrokerNodeGroupInfoOutputReference;
_f = JSII_RTTI_SYMBOL_1;
MskClusterBrokerNodeGroupInfoOutputReference[_f] = { fqn: "@cdktf/provider-aws.mskCluster.MskClusterBrokerNodeGroupInfoOutputReference", version: "10.0.12" };
function mskClusterClientAuthenticationSaslToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        iam: cdktf.booleanToTerraform(struct.iam),
        scram: cdktf.booleanToTerraform(struct.scram),
    };
}
exports.mskClusterClientAuthenticationSaslToTerraform = mskClusterClientAuthenticationSaslToTerraform;
class MskClusterClientAuthenticationSaslOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._iam !== undefined) {
            hasAnyValues = true;
            internalValueResult.iam = this._iam;
        }
        if (this._scram !== undefined) {
            hasAnyValues = true;
            internalValueResult.scram = this._scram;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._iam = undefined;
            this._scram = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._iam = value.iam;
            this._scram = value.scram;
        }
    }
    get iam() {
        return this.getBooleanAttribute('iam');
    }
    set iam(value) {
        this._iam = value;
    }
    resetIam() {
        this._iam = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iamInput() {
        return this._iam;
    }
    get scram() {
        return this.getBooleanAttribute('scram');
    }
    set scram(value) {
        this._scram = value;
    }
    resetScram() {
        this._scram = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scramInput() {
        return this._scram;
    }
}
exports.MskClusterClientAuthenticationSaslOutputReference = MskClusterClientAuthenticationSaslOutputReference;
_g = JSII_RTTI_SYMBOL_1;
MskClusterClientAuthenticationSaslOutputReference[_g] = { fqn: "@cdktf/provider-aws.mskCluster.MskClusterClientAuthenticationSaslOutputReference", version: "10.0.12" };
function mskClusterClientAuthenticationTlsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        certificate_authority_arns: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.certificateAuthorityArns),
    };
}
exports.mskClusterClientAuthenticationTlsToTerraform = mskClusterClientAuthenticationTlsToTerraform;
class MskClusterClientAuthenticationTlsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._certificateAuthorityArns !== undefined) {
            hasAnyValues = true;
            internalValueResult.certificateAuthorityArns = this._certificateAuthorityArns;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._certificateAuthorityArns = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._certificateAuthorityArns = value.certificateAuthorityArns;
        }
    }
    get certificateAuthorityArns() {
        return cdktf.Fn.tolist(this.getListAttribute('certificate_authority_arns'));
    }
    set certificateAuthorityArns(value) {
        this._certificateAuthorityArns = value;
    }
    resetCertificateAuthorityArns() {
        this._certificateAuthorityArns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get certificateAuthorityArnsInput() {
        return this._certificateAuthorityArns;
    }
}
exports.MskClusterClientAuthenticationTlsOutputReference = MskClusterClientAuthenticationTlsOutputReference;
_h = JSII_RTTI_SYMBOL_1;
MskClusterClientAuthenticationTlsOutputReference[_h] = { fqn: "@cdktf/provider-aws.mskCluster.MskClusterClientAuthenticationTlsOutputReference", version: "10.0.12" };
function mskClusterClientAuthenticationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        unauthenticated: cdktf.booleanToTerraform(struct.unauthenticated),
        sasl: mskClusterClientAuthenticationSaslToTerraform(struct.sasl),
        tls: mskClusterClientAuthenticationTlsToTerraform(struct.tls),
    };
}
exports.mskClusterClientAuthenticationToTerraform = mskClusterClientAuthenticationToTerraform;
class MskClusterClientAuthenticationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // sasl - computed: false, optional: true, required: false
        this._sasl = new MskClusterClientAuthenticationSaslOutputReference(this, "sasl");
        // tls - computed: false, optional: true, required: false
        this._tls = new MskClusterClientAuthenticationTlsOutputReference(this, "tls");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._unauthenticated !== undefined) {
            hasAnyValues = true;
            internalValueResult.unauthenticated = this._unauthenticated;
        }
        if (this._sasl?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.sasl = this._sasl?.internalValue;
        }
        if (this._tls?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.tls = this._tls?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._unauthenticated = undefined;
            this._sasl.internalValue = undefined;
            this._tls.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._unauthenticated = value.unauthenticated;
            this._sasl.internalValue = value.sasl;
            this._tls.internalValue = value.tls;
        }
    }
    get unauthenticated() {
        return this.getBooleanAttribute('unauthenticated');
    }
    set unauthenticated(value) {
        this._unauthenticated = value;
    }
    resetUnauthenticated() {
        this._unauthenticated = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get unauthenticatedInput() {
        return this._unauthenticated;
    }
    get sasl() {
        return this._sasl;
    }
    putSasl(value) {
        this._sasl.internalValue = value;
    }
    resetSasl() {
        this._sasl.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get saslInput() {
        return this._sasl.internalValue;
    }
    get tls() {
        return this._tls;
    }
    putTls(value) {
        this._tls.internalValue = value;
    }
    resetTls() {
        this._tls.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tlsInput() {
        return this._tls.internalValue;
    }
}
exports.MskClusterClientAuthenticationOutputReference = MskClusterClientAuthenticationOutputReference;
_j = JSII_RTTI_SYMBOL_1;
MskClusterClientAuthenticationOutputReference[_j] = { fqn: "@cdktf/provider-aws.mskCluster.MskClusterClientAuthenticationOutputReference", version: "10.0.12" };
function mskClusterConfigurationInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        arn: cdktf.stringToTerraform(struct.arn),
        revision: cdktf.numberToTerraform(struct.revision),
    };
}
exports.mskClusterConfigurationInfoToTerraform = mskClusterConfigurationInfoToTerraform;
class MskClusterConfigurationInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._arn !== undefined) {
            hasAnyValues = true;
            internalValueResult.arn = this._arn;
        }
        if (this._revision !== undefined) {
            hasAnyValues = true;
            internalValueResult.revision = this._revision;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._arn = undefined;
            this._revision = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._arn = value.arn;
            this._revision = value.revision;
        }
    }
    get arn() {
        return this.getStringAttribute('arn');
    }
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    get arnInput() {
        return this._arn;
    }
    get revision() {
        return this.getNumberAttribute('revision');
    }
    set revision(value) {
        this._revision = value;
    }
    // Temporarily expose input value. Use with caution.
    get revisionInput() {
        return this._revision;
    }
}
exports.MskClusterConfigurationInfoOutputReference = MskClusterConfigurationInfoOutputReference;
_k = JSII_RTTI_SYMBOL_1;
MskClusterConfigurationInfoOutputReference[_k] = { fqn: "@cdktf/provider-aws.mskCluster.MskClusterConfigurationInfoOutputReference", version: "10.0.12" };
function mskClusterEncryptionInfoEncryptionInTransitToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        client_broker: cdktf.stringToTerraform(struct.clientBroker),
        in_cluster: cdktf.booleanToTerraform(struct.inCluster),
    };
}
exports.mskClusterEncryptionInfoEncryptionInTransitToTerraform = mskClusterEncryptionInfoEncryptionInTransitToTerraform;
class MskClusterEncryptionInfoEncryptionInTransitOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._clientBroker !== undefined) {
            hasAnyValues = true;
            internalValueResult.clientBroker = this._clientBroker;
        }
        if (this._inCluster !== undefined) {
            hasAnyValues = true;
            internalValueResult.inCluster = this._inCluster;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._clientBroker = undefined;
            this._inCluster = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._clientBroker = value.clientBroker;
            this._inCluster = value.inCluster;
        }
    }
    get clientBroker() {
        return this.getStringAttribute('client_broker');
    }
    set clientBroker(value) {
        this._clientBroker = value;
    }
    resetClientBroker() {
        this._clientBroker = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clientBrokerInput() {
        return this._clientBroker;
    }
    get inCluster() {
        return this.getBooleanAttribute('in_cluster');
    }
    set inCluster(value) {
        this._inCluster = value;
    }
    resetInCluster() {
        this._inCluster = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get inClusterInput() {
        return this._inCluster;
    }
}
exports.MskClusterEncryptionInfoEncryptionInTransitOutputReference = MskClusterEncryptionInfoEncryptionInTransitOutputReference;
_l = JSII_RTTI_SYMBOL_1;
MskClusterEncryptionInfoEncryptionInTransitOutputReference[_l] = { fqn: "@cdktf/provider-aws.mskCluster.MskClusterEncryptionInfoEncryptionInTransitOutputReference", version: "10.0.12" };
function mskClusterEncryptionInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        encryption_at_rest_kms_key_arn: cdktf.stringToTerraform(struct.encryptionAtRestKmsKeyArn),
        encryption_in_transit: mskClusterEncryptionInfoEncryptionInTransitToTerraform(struct.encryptionInTransit),
    };
}
exports.mskClusterEncryptionInfoToTerraform = mskClusterEncryptionInfoToTerraform;
class MskClusterEncryptionInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // encryption_in_transit - computed: false, optional: true, required: false
        this._encryptionInTransit = new MskClusterEncryptionInfoEncryptionInTransitOutputReference(this, "encryption_in_transit");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._encryptionAtRestKmsKeyArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.encryptionAtRestKmsKeyArn = this._encryptionAtRestKmsKeyArn;
        }
        if (this._encryptionInTransit?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.encryptionInTransit = this._encryptionInTransit?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._encryptionAtRestKmsKeyArn = undefined;
            this._encryptionInTransit.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._encryptionAtRestKmsKeyArn = value.encryptionAtRestKmsKeyArn;
            this._encryptionInTransit.internalValue = value.encryptionInTransit;
        }
    }
    get encryptionAtRestKmsKeyArn() {
        return this.getStringAttribute('encryption_at_rest_kms_key_arn');
    }
    set encryptionAtRestKmsKeyArn(value) {
        this._encryptionAtRestKmsKeyArn = value;
    }
    resetEncryptionAtRestKmsKeyArn() {
        this._encryptionAtRestKmsKeyArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encryptionAtRestKmsKeyArnInput() {
        return this._encryptionAtRestKmsKeyArn;
    }
    get encryptionInTransit() {
        return this._encryptionInTransit;
    }
    putEncryptionInTransit(value) {
        this._encryptionInTransit.internalValue = value;
    }
    resetEncryptionInTransit() {
        this._encryptionInTransit.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encryptionInTransitInput() {
        return this._encryptionInTransit.internalValue;
    }
}
exports.MskClusterEncryptionInfoOutputReference = MskClusterEncryptionInfoOutputReference;
_m = JSII_RTTI_SYMBOL_1;
MskClusterEncryptionInfoOutputReference[_m] = { fqn: "@cdktf/provider-aws.mskCluster.MskClusterEncryptionInfoOutputReference", version: "10.0.12" };
function mskClusterLoggingInfoBrokerLogsCloudwatchLogsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        log_group: cdktf.stringToTerraform(struct.logGroup),
    };
}
exports.mskClusterLoggingInfoBrokerLogsCloudwatchLogsToTerraform = mskClusterLoggingInfoBrokerLogsCloudwatchLogsToTerraform;
class MskClusterLoggingInfoBrokerLogsCloudwatchLogsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._logGroup !== undefined) {
            hasAnyValues = true;
            internalValueResult.logGroup = this._logGroup;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._logGroup = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._logGroup = value.logGroup;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get logGroup() {
        return this.getStringAttribute('log_group');
    }
    set logGroup(value) {
        this._logGroup = value;
    }
    resetLogGroup() {
        this._logGroup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get logGroupInput() {
        return this._logGroup;
    }
}
exports.MskClusterLoggingInfoBrokerLogsCloudwatchLogsOutputReference = MskClusterLoggingInfoBrokerLogsCloudwatchLogsOutputReference;
_o = JSII_RTTI_SYMBOL_1;
MskClusterLoggingInfoBrokerLogsCloudwatchLogsOutputReference[_o] = { fqn: "@cdktf/provider-aws.mskCluster.MskClusterLoggingInfoBrokerLogsCloudwatchLogsOutputReference", version: "10.0.12" };
function mskClusterLoggingInfoBrokerLogsFirehoseToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        delivery_stream: cdktf.stringToTerraform(struct.deliveryStream),
        enabled: cdktf.booleanToTerraform(struct.enabled),
    };
}
exports.mskClusterLoggingInfoBrokerLogsFirehoseToTerraform = mskClusterLoggingInfoBrokerLogsFirehoseToTerraform;
class MskClusterLoggingInfoBrokerLogsFirehoseOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._deliveryStream !== undefined) {
            hasAnyValues = true;
            internalValueResult.deliveryStream = this._deliveryStream;
        }
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._deliveryStream = undefined;
            this._enabled = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._deliveryStream = value.deliveryStream;
            this._enabled = value.enabled;
        }
    }
    get deliveryStream() {
        return this.getStringAttribute('delivery_stream');
    }
    set deliveryStream(value) {
        this._deliveryStream = value;
    }
    resetDeliveryStream() {
        this._deliveryStream = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deliveryStreamInput() {
        return this._deliveryStream;
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
}
exports.MskClusterLoggingInfoBrokerLogsFirehoseOutputReference = MskClusterLoggingInfoBrokerLogsFirehoseOutputReference;
_p = JSII_RTTI_SYMBOL_1;
MskClusterLoggingInfoBrokerLogsFirehoseOutputReference[_p] = { fqn: "@cdktf/provider-aws.mskCluster.MskClusterLoggingInfoBrokerLogsFirehoseOutputReference", version: "10.0.12" };
function mskClusterLoggingInfoBrokerLogsS3ToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket: cdktf.stringToTerraform(struct.bucket),
        enabled: cdktf.booleanToTerraform(struct.enabled),
        prefix: cdktf.stringToTerraform(struct.prefix),
    };
}
exports.mskClusterLoggingInfoBrokerLogsS3ToTerraform = mskClusterLoggingInfoBrokerLogsS3ToTerraform;
class MskClusterLoggingInfoBrokerLogsS3OutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucket !== undefined) {
            hasAnyValues = true;
            internalValueResult.bucket = this._bucket;
        }
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._prefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucket = undefined;
            this._enabled = undefined;
            this._prefix = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucket = value.bucket;
            this._enabled = value.enabled;
            this._prefix = value.prefix;
        }
    }
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    set bucket(value) {
        this._bucket = value;
    }
    resetBucket() {
        this._bucket = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bucketInput() {
        return this._bucket;
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
}
exports.MskClusterLoggingInfoBrokerLogsS3OutputReference = MskClusterLoggingInfoBrokerLogsS3OutputReference;
_q = JSII_RTTI_SYMBOL_1;
MskClusterLoggingInfoBrokerLogsS3OutputReference[_q] = { fqn: "@cdktf/provider-aws.mskCluster.MskClusterLoggingInfoBrokerLogsS3OutputReference", version: "10.0.12" };
function mskClusterLoggingInfoBrokerLogsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cloudwatch_logs: mskClusterLoggingInfoBrokerLogsCloudwatchLogsToTerraform(struct.cloudwatchLogs),
        firehose: mskClusterLoggingInfoBrokerLogsFirehoseToTerraform(struct.firehose),
        s3: mskClusterLoggingInfoBrokerLogsS3ToTerraform(struct.s3),
    };
}
exports.mskClusterLoggingInfoBrokerLogsToTerraform = mskClusterLoggingInfoBrokerLogsToTerraform;
class MskClusterLoggingInfoBrokerLogsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // cloudwatch_logs - computed: false, optional: true, required: false
        this._cloudwatchLogs = new MskClusterLoggingInfoBrokerLogsCloudwatchLogsOutputReference(this, "cloudwatch_logs");
        // firehose - computed: false, optional: true, required: false
        this._firehose = new MskClusterLoggingInfoBrokerLogsFirehoseOutputReference(this, "firehose");
        // s3 - computed: false, optional: true, required: false
        this._s3 = new MskClusterLoggingInfoBrokerLogsS3OutputReference(this, "s3");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cloudwatchLogs?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.cloudwatchLogs = this._cloudwatchLogs?.internalValue;
        }
        if (this._firehose?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.firehose = this._firehose?.internalValue;
        }
        if (this._s3?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3 = this._s3?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cloudwatchLogs.internalValue = undefined;
            this._firehose.internalValue = undefined;
            this._s3.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cloudwatchLogs.internalValue = value.cloudwatchLogs;
            this._firehose.internalValue = value.firehose;
            this._s3.internalValue = value.s3;
        }
    }
    get cloudwatchLogs() {
        return this._cloudwatchLogs;
    }
    putCloudwatchLogs(value) {
        this._cloudwatchLogs.internalValue = value;
    }
    resetCloudwatchLogs() {
        this._cloudwatchLogs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudwatchLogsInput() {
        return this._cloudwatchLogs.internalValue;
    }
    get firehose() {
        return this._firehose;
    }
    putFirehose(value) {
        this._firehose.internalValue = value;
    }
    resetFirehose() {
        this._firehose.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get firehoseInput() {
        return this._firehose.internalValue;
    }
    get s3() {
        return this._s3;
    }
    putS3(value) {
        this._s3.internalValue = value;
    }
    resetS3() {
        this._s3.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get s3Input() {
        return this._s3.internalValue;
    }
}
exports.MskClusterLoggingInfoBrokerLogsOutputReference = MskClusterLoggingInfoBrokerLogsOutputReference;
_r = JSII_RTTI_SYMBOL_1;
MskClusterLoggingInfoBrokerLogsOutputReference[_r] = { fqn: "@cdktf/provider-aws.mskCluster.MskClusterLoggingInfoBrokerLogsOutputReference", version: "10.0.12" };
function mskClusterLoggingInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        broker_logs: mskClusterLoggingInfoBrokerLogsToTerraform(struct.brokerLogs),
    };
}
exports.mskClusterLoggingInfoToTerraform = mskClusterLoggingInfoToTerraform;
class MskClusterLoggingInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // broker_logs - computed: false, optional: false, required: true
        this._brokerLogs = new MskClusterLoggingInfoBrokerLogsOutputReference(this, "broker_logs");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._brokerLogs?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.brokerLogs = this._brokerLogs?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._brokerLogs.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._brokerLogs.internalValue = value.brokerLogs;
        }
    }
    get brokerLogs() {
        return this._brokerLogs;
    }
    putBrokerLogs(value) {
        this._brokerLogs.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get brokerLogsInput() {
        return this._brokerLogs.internalValue;
    }
}
exports.MskClusterLoggingInfoOutputReference = MskClusterLoggingInfoOutputReference;
_s = JSII_RTTI_SYMBOL_1;
MskClusterLoggingInfoOutputReference[_s] = { fqn: "@cdktf/provider-aws.mskCluster.MskClusterLoggingInfoOutputReference", version: "10.0.12" };
function mskClusterOpenMonitoringPrometheusJmxExporterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled_in_broker: cdktf.booleanToTerraform(struct.enabledInBroker),
    };
}
exports.mskClusterOpenMonitoringPrometheusJmxExporterToTerraform = mskClusterOpenMonitoringPrometheusJmxExporterToTerraform;
class MskClusterOpenMonitoringPrometheusJmxExporterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabledInBroker !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabledInBroker = this._enabledInBroker;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabledInBroker = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabledInBroker = value.enabledInBroker;
        }
    }
    get enabledInBroker() {
        return this.getBooleanAttribute('enabled_in_broker');
    }
    set enabledInBroker(value) {
        this._enabledInBroker = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInBrokerInput() {
        return this._enabledInBroker;
    }
}
exports.MskClusterOpenMonitoringPrometheusJmxExporterOutputReference = MskClusterOpenMonitoringPrometheusJmxExporterOutputReference;
_t = JSII_RTTI_SYMBOL_1;
MskClusterOpenMonitoringPrometheusJmxExporterOutputReference[_t] = { fqn: "@cdktf/provider-aws.mskCluster.MskClusterOpenMonitoringPrometheusJmxExporterOutputReference", version: "10.0.12" };
function mskClusterOpenMonitoringPrometheusNodeExporterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled_in_broker: cdktf.booleanToTerraform(struct.enabledInBroker),
    };
}
exports.mskClusterOpenMonitoringPrometheusNodeExporterToTerraform = mskClusterOpenMonitoringPrometheusNodeExporterToTerraform;
class MskClusterOpenMonitoringPrometheusNodeExporterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabledInBroker !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabledInBroker = this._enabledInBroker;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabledInBroker = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabledInBroker = value.enabledInBroker;
        }
    }
    get enabledInBroker() {
        return this.getBooleanAttribute('enabled_in_broker');
    }
    set enabledInBroker(value) {
        this._enabledInBroker = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInBrokerInput() {
        return this._enabledInBroker;
    }
}
exports.MskClusterOpenMonitoringPrometheusNodeExporterOutputReference = MskClusterOpenMonitoringPrometheusNodeExporterOutputReference;
_u = JSII_RTTI_SYMBOL_1;
MskClusterOpenMonitoringPrometheusNodeExporterOutputReference[_u] = { fqn: "@cdktf/provider-aws.mskCluster.MskClusterOpenMonitoringPrometheusNodeExporterOutputReference", version: "10.0.12" };
function mskClusterOpenMonitoringPrometheusToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        jmx_exporter: mskClusterOpenMonitoringPrometheusJmxExporterToTerraform(struct.jmxExporter),
        node_exporter: mskClusterOpenMonitoringPrometheusNodeExporterToTerraform(struct.nodeExporter),
    };
}
exports.mskClusterOpenMonitoringPrometheusToTerraform = mskClusterOpenMonitoringPrometheusToTerraform;
class MskClusterOpenMonitoringPrometheusOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // jmx_exporter - computed: false, optional: true, required: false
        this._jmxExporter = new MskClusterOpenMonitoringPrometheusJmxExporterOutputReference(this, "jmx_exporter");
        // node_exporter - computed: false, optional: true, required: false
        this._nodeExporter = new MskClusterOpenMonitoringPrometheusNodeExporterOutputReference(this, "node_exporter");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._jmxExporter?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.jmxExporter = this._jmxExporter?.internalValue;
        }
        if (this._nodeExporter?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.nodeExporter = this._nodeExporter?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._jmxExporter.internalValue = undefined;
            this._nodeExporter.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._jmxExporter.internalValue = value.jmxExporter;
            this._nodeExporter.internalValue = value.nodeExporter;
        }
    }
    get jmxExporter() {
        return this._jmxExporter;
    }
    putJmxExporter(value) {
        this._jmxExporter.internalValue = value;
    }
    resetJmxExporter() {
        this._jmxExporter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get jmxExporterInput() {
        return this._jmxExporter.internalValue;
    }
    get nodeExporter() {
        return this._nodeExporter;
    }
    putNodeExporter(value) {
        this._nodeExporter.internalValue = value;
    }
    resetNodeExporter() {
        this._nodeExporter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nodeExporterInput() {
        return this._nodeExporter.internalValue;
    }
}
exports.MskClusterOpenMonitoringPrometheusOutputReference = MskClusterOpenMonitoringPrometheusOutputReference;
_v = JSII_RTTI_SYMBOL_1;
MskClusterOpenMonitoringPrometheusOutputReference[_v] = { fqn: "@cdktf/provider-aws.mskCluster.MskClusterOpenMonitoringPrometheusOutputReference", version: "10.0.12" };
function mskClusterOpenMonitoringToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        prometheus: mskClusterOpenMonitoringPrometheusToTerraform(struct.prometheus),
    };
}
exports.mskClusterOpenMonitoringToTerraform = mskClusterOpenMonitoringToTerraform;
class MskClusterOpenMonitoringOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // prometheus - computed: false, optional: false, required: true
        this._prometheus = new MskClusterOpenMonitoringPrometheusOutputReference(this, "prometheus");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._prometheus?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.prometheus = this._prometheus?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._prometheus.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._prometheus.internalValue = value.prometheus;
        }
    }
    get prometheus() {
        return this._prometheus;
    }
    putPrometheus(value) {
        this._prometheus.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get prometheusInput() {
        return this._prometheus.internalValue;
    }
}
exports.MskClusterOpenMonitoringOutputReference = MskClusterOpenMonitoringOutputReference;
_w = JSII_RTTI_SYMBOL_1;
MskClusterOpenMonitoringOutputReference[_w] = { fqn: "@cdktf/provider-aws.mskCluster.MskClusterOpenMonitoringOutputReference", version: "10.0.12" };
function mskClusterTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.mskClusterTimeoutsToTerraform = mskClusterTimeoutsToTerraform;
class MskClusterTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.MskClusterTimeoutsOutputReference = MskClusterTimeoutsOutputReference;
_x = JSII_RTTI_SYMBOL_1;
MskClusterTimeoutsOutputReference[_x] = { fqn: "@cdktf/provider-aws.mskCluster.MskClusterTimeoutsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/msk_cluster aws_msk_cluster}
*/
class MskCluster extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/msk_cluster aws_msk_cluster} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options MskClusterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_msk_cluster',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // broker_node_group_info - computed: false, optional: false, required: true
        this._brokerNodeGroupInfo = new MskClusterBrokerNodeGroupInfoOutputReference(this, "broker_node_group_info");
        // client_authentication - computed: false, optional: true, required: false
        this._clientAuthentication = new MskClusterClientAuthenticationOutputReference(this, "client_authentication");
        // configuration_info - computed: false, optional: true, required: false
        this._configurationInfo = new MskClusterConfigurationInfoOutputReference(this, "configuration_info");
        // encryption_info - computed: false, optional: true, required: false
        this._encryptionInfo = new MskClusterEncryptionInfoOutputReference(this, "encryption_info");
        // logging_info - computed: false, optional: true, required: false
        this._loggingInfo = new MskClusterLoggingInfoOutputReference(this, "logging_info");
        // open_monitoring - computed: false, optional: true, required: false
        this._openMonitoring = new MskClusterOpenMonitoringOutputReference(this, "open_monitoring");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new MskClusterTimeoutsOutputReference(this, "timeouts");
        this._clusterName = config.clusterName;
        this._enhancedMonitoring = config.enhancedMonitoring;
        this._id = config.id;
        this._kafkaVersion = config.kafkaVersion;
        this._numberOfBrokerNodes = config.numberOfBrokerNodes;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._brokerNodeGroupInfo.internalValue = config.brokerNodeGroupInfo;
        this._clientAuthentication.internalValue = config.clientAuthentication;
        this._configurationInfo.internalValue = config.configurationInfo;
        this._encryptionInfo.internalValue = config.encryptionInfo;
        this._loggingInfo.internalValue = config.loggingInfo;
        this._openMonitoring.internalValue = config.openMonitoring;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // bootstrap_brokers - computed: true, optional: false, required: false
    get bootstrapBrokers() {
        return this.getStringAttribute('bootstrap_brokers');
    }
    // bootstrap_brokers_public_sasl_iam - computed: true, optional: false, required: false
    get bootstrapBrokersPublicSaslIam() {
        return this.getStringAttribute('bootstrap_brokers_public_sasl_iam');
    }
    // bootstrap_brokers_public_sasl_scram - computed: true, optional: false, required: false
    get bootstrapBrokersPublicSaslScram() {
        return this.getStringAttribute('bootstrap_brokers_public_sasl_scram');
    }
    // bootstrap_brokers_public_tls - computed: true, optional: false, required: false
    get bootstrapBrokersPublicTls() {
        return this.getStringAttribute('bootstrap_brokers_public_tls');
    }
    // bootstrap_brokers_sasl_iam - computed: true, optional: false, required: false
    get bootstrapBrokersSaslIam() {
        return this.getStringAttribute('bootstrap_brokers_sasl_iam');
    }
    // bootstrap_brokers_sasl_scram - computed: true, optional: false, required: false
    get bootstrapBrokersSaslScram() {
        return this.getStringAttribute('bootstrap_brokers_sasl_scram');
    }
    // bootstrap_brokers_tls - computed: true, optional: false, required: false
    get bootstrapBrokersTls() {
        return this.getStringAttribute('bootstrap_brokers_tls');
    }
    get clusterName() {
        return this.getStringAttribute('cluster_name');
    }
    set clusterName(value) {
        this._clusterName = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterNameInput() {
        return this._clusterName;
    }
    // current_version - computed: true, optional: false, required: false
    get currentVersion() {
        return this.getStringAttribute('current_version');
    }
    get enhancedMonitoring() {
        return this.getStringAttribute('enhanced_monitoring');
    }
    set enhancedMonitoring(value) {
        this._enhancedMonitoring = value;
    }
    resetEnhancedMonitoring() {
        this._enhancedMonitoring = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enhancedMonitoringInput() {
        return this._enhancedMonitoring;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get kafkaVersion() {
        return this.getStringAttribute('kafka_version');
    }
    set kafkaVersion(value) {
        this._kafkaVersion = value;
    }
    // Temporarily expose input value. Use with caution.
    get kafkaVersionInput() {
        return this._kafkaVersion;
    }
    get numberOfBrokerNodes() {
        return this.getNumberAttribute('number_of_broker_nodes');
    }
    set numberOfBrokerNodes(value) {
        this._numberOfBrokerNodes = value;
    }
    // Temporarily expose input value. Use with caution.
    get numberOfBrokerNodesInput() {
        return this._numberOfBrokerNodes;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // zookeeper_connect_string - computed: true, optional: false, required: false
    get zookeeperConnectString() {
        return this.getStringAttribute('zookeeper_connect_string');
    }
    // zookeeper_connect_string_tls - computed: true, optional: false, required: false
    get zookeeperConnectStringTls() {
        return this.getStringAttribute('zookeeper_connect_string_tls');
    }
    get brokerNodeGroupInfo() {
        return this._brokerNodeGroupInfo;
    }
    putBrokerNodeGroupInfo(value) {
        this._brokerNodeGroupInfo.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get brokerNodeGroupInfoInput() {
        return this._brokerNodeGroupInfo.internalValue;
    }
    get clientAuthentication() {
        return this._clientAuthentication;
    }
    putClientAuthentication(value) {
        this._clientAuthentication.internalValue = value;
    }
    resetClientAuthentication() {
        this._clientAuthentication.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clientAuthenticationInput() {
        return this._clientAuthentication.internalValue;
    }
    get configurationInfo() {
        return this._configurationInfo;
    }
    putConfigurationInfo(value) {
        this._configurationInfo.internalValue = value;
    }
    resetConfigurationInfo() {
        this._configurationInfo.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get configurationInfoInput() {
        return this._configurationInfo.internalValue;
    }
    get encryptionInfo() {
        return this._encryptionInfo;
    }
    putEncryptionInfo(value) {
        this._encryptionInfo.internalValue = value;
    }
    resetEncryptionInfo() {
        this._encryptionInfo.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encryptionInfoInput() {
        return this._encryptionInfo.internalValue;
    }
    get loggingInfo() {
        return this._loggingInfo;
    }
    putLoggingInfo(value) {
        this._loggingInfo.internalValue = value;
    }
    resetLoggingInfo() {
        this._loggingInfo.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get loggingInfoInput() {
        return this._loggingInfo.internalValue;
    }
    get openMonitoring() {
        return this._openMonitoring;
    }
    putOpenMonitoring(value) {
        this._openMonitoring.internalValue = value;
    }
    resetOpenMonitoring() {
        this._openMonitoring.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get openMonitoringInput() {
        return this._openMonitoring.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_name: cdktf.stringToTerraform(this._clusterName),
            enhanced_monitoring: cdktf.stringToTerraform(this._enhancedMonitoring),
            id: cdktf.stringToTerraform(this._id),
            kafka_version: cdktf.stringToTerraform(this._kafkaVersion),
            number_of_broker_nodes: cdktf.numberToTerraform(this._numberOfBrokerNodes),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            broker_node_group_info: mskClusterBrokerNodeGroupInfoToTerraform(this._brokerNodeGroupInfo.internalValue),
            client_authentication: mskClusterClientAuthenticationToTerraform(this._clientAuthentication.internalValue),
            configuration_info: mskClusterConfigurationInfoToTerraform(this._configurationInfo.internalValue),
            encryption_info: mskClusterEncryptionInfoToTerraform(this._encryptionInfo.internalValue),
            logging_info: mskClusterLoggingInfoToTerraform(this._loggingInfo.internalValue),
            open_monitoring: mskClusterOpenMonitoringToTerraform(this._openMonitoring.internalValue),
            timeouts: mskClusterTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.MskCluster = MskCluster;
_y = JSII_RTTI_SYMBOL_1;
MskCluster[_y] = { fqn: "@cdktf/provider-aws.mskCluster.MskCluster", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
MskCluster.tfResourceType = "aws_msk_cluster";
