// @cdktf/provider-aws/lib/inspector-assessment-target/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.InspectorAssessmentTarget = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/inspector_assessment_target aws_inspector_assessment_target}
*/
class InspectorAssessmentTarget extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/inspector_assessment_target aws_inspector_assessment_target} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options InspectorAssessmentTargetConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_inspector_assessment_target',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._name = config.name;
        this._resourceGroupArn = config.resourceGroupArn;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get resourceGroupArn() {
        return this.getStringAttribute('resource_group_arn');
    }
    set resourceGroupArn(value) {
        this._resourceGroupArn = value;
    }
    resetResourceGroupArn() {
        this._resourceGroupArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourceGroupArnInput() {
        return this._resourceGroupArn;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            resource_group_arn: cdktf.stringToTerraform(this._resourceGroupArn),
        };
    }
}
exports.InspectorAssessmentTarget = InspectorAssessmentTarget;
_a = JSII_RTTI_SYMBOL_1;
InspectorAssessmentTarget[_a] = { fqn: "@cdktf/provider-aws.inspectorAssessmentTarget.InspectorAssessmentTarget", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
InspectorAssessmentTarget.tfResourceType = "aws_inspector_assessment_target";
