// @cdktf/provider-aws/lib/data-aws-serverlessapplicationrepository-application/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsServerlessapplicationrepositoryApplication = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/serverlessapplicationrepository_application aws_serverlessapplicationrepository_application}
*/
class DataAwsServerlessapplicationrepositoryApplication extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/serverlessapplicationrepository_application aws_serverlessapplicationrepository_application} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsServerlessapplicationrepositoryApplicationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_serverlessapplicationrepository_application',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._applicationId = config.applicationId;
        this._id = config.id;
        this._semanticVersion = config.semanticVersion;
    }
    get applicationId() {
        return this.getStringAttribute('application_id');
    }
    set applicationId(value) {
        this._applicationId = value;
    }
    // Temporarily expose input value. Use with caution.
    get applicationIdInput() {
        return this._applicationId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // required_capabilities - computed: true, optional: false, required: false
    get requiredCapabilities() {
        return cdktf.Fn.tolist(this.getListAttribute('required_capabilities'));
    }
    get semanticVersion() {
        return this.getStringAttribute('semantic_version');
    }
    set semanticVersion(value) {
        this._semanticVersion = value;
    }
    resetSemanticVersion() {
        this._semanticVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get semanticVersionInput() {
        return this._semanticVersion;
    }
    // source_code_url - computed: true, optional: false, required: false
    get sourceCodeUrl() {
        return this.getStringAttribute('source_code_url');
    }
    // template_url - computed: true, optional: false, required: false
    get templateUrl() {
        return this.getStringAttribute('template_url');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            application_id: cdktf.stringToTerraform(this._applicationId),
            id: cdktf.stringToTerraform(this._id),
            semantic_version: cdktf.stringToTerraform(this._semanticVersion),
        };
    }
}
exports.DataAwsServerlessapplicationrepositoryApplication = DataAwsServerlessapplicationrepositoryApplication;
_a = JSII_RTTI_SYMBOL_1;
DataAwsServerlessapplicationrepositoryApplication[_a] = { fqn: "@cdktf/provider-aws.dataAwsServerlessapplicationrepositoryApplication.DataAwsServerlessapplicationrepositoryApplication", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsServerlessapplicationrepositoryApplication.tfResourceType = "aws_serverlessapplicationrepository_application";
