// @cdktf/provider-aws/lib/cloudwatch-event-api-destination/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudwatchEventApiDestination = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/cloudwatch_event_api_destination aws_cloudwatch_event_api_destination}
*/
class CloudwatchEventApiDestination extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/cloudwatch_event_api_destination aws_cloudwatch_event_api_destination} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CloudwatchEventApiDestinationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cloudwatch_event_api_destination',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._connectionArn = config.connectionArn;
        this._description = config.description;
        this._httpMethod = config.httpMethod;
        this._id = config.id;
        this._invocationEndpoint = config.invocationEndpoint;
        this._invocationRateLimitPerSecond = config.invocationRateLimitPerSecond;
        this._name = config.name;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get connectionArn() {
        return this.getStringAttribute('connection_arn');
    }
    set connectionArn(value) {
        this._connectionArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get connectionArnInput() {
        return this._connectionArn;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get httpMethod() {
        return this.getStringAttribute('http_method');
    }
    set httpMethod(value) {
        this._httpMethod = value;
    }
    // Temporarily expose input value. Use with caution.
    get httpMethodInput() {
        return this._httpMethod;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get invocationEndpoint() {
        return this.getStringAttribute('invocation_endpoint');
    }
    set invocationEndpoint(value) {
        this._invocationEndpoint = value;
    }
    // Temporarily expose input value. Use with caution.
    get invocationEndpointInput() {
        return this._invocationEndpoint;
    }
    get invocationRateLimitPerSecond() {
        return this.getNumberAttribute('invocation_rate_limit_per_second');
    }
    set invocationRateLimitPerSecond(value) {
        this._invocationRateLimitPerSecond = value;
    }
    resetInvocationRateLimitPerSecond() {
        this._invocationRateLimitPerSecond = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get invocationRateLimitPerSecondInput() {
        return this._invocationRateLimitPerSecond;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            connection_arn: cdktf.stringToTerraform(this._connectionArn),
            description: cdktf.stringToTerraform(this._description),
            http_method: cdktf.stringToTerraform(this._httpMethod),
            id: cdktf.stringToTerraform(this._id),
            invocation_endpoint: cdktf.stringToTerraform(this._invocationEndpoint),
            invocation_rate_limit_per_second: cdktf.numberToTerraform(this._invocationRateLimitPerSecond),
            name: cdktf.stringToTerraform(this._name),
        };
    }
}
exports.CloudwatchEventApiDestination = CloudwatchEventApiDestination;
_a = JSII_RTTI_SYMBOL_1;
CloudwatchEventApiDestination[_a] = { fqn: "@cdktf/provider-aws.cloudwatchEventApiDestination.CloudwatchEventApiDestination", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
CloudwatchEventApiDestination.tfResourceType = "aws_cloudwatch_event_api_destination";
