// @cdktf/provider-aws/lib/kinesis-analytics-application/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7;
Object.defineProperty(exports, "__esModule", { value: true });
exports.KinesisAnalyticsApplication = exports.KinesisAnalyticsApplicationReferenceDataSourcesOutputReference = exports.kinesisAnalyticsApplicationReferenceDataSourcesToTerraform = exports.KinesisAnalyticsApplicationReferenceDataSourcesSchemaOutputReference = exports.kinesisAnalyticsApplicationReferenceDataSourcesSchemaToTerraform = exports.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatOutputReference = exports.kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatToTerraform = exports.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersOutputReference = exports.kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersToTerraform = exports.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersJsonOutputReference = exports.kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersJsonToTerraform = exports.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersCsvOutputReference = exports.kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersCsvToTerraform = exports.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordColumnsList = exports.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordColumnsOutputReference = exports.kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordColumnsToTerraform = exports.KinesisAnalyticsApplicationReferenceDataSourcesS3OutputReference = exports.kinesisAnalyticsApplicationReferenceDataSourcesS3ToTerraform = exports.KinesisAnalyticsApplicationOutputsList = exports.KinesisAnalyticsApplicationOutputsOutputReference = exports.kinesisAnalyticsApplicationOutputsToTerraform = exports.KinesisAnalyticsApplicationOutputsSchemaOutputReference = exports.kinesisAnalyticsApplicationOutputsSchemaToTerraform = exports.KinesisAnalyticsApplicationOutputsLambdaOutputReference = exports.kinesisAnalyticsApplicationOutputsLambdaToTerraform = exports.KinesisAnalyticsApplicationOutputsKinesisStreamOutputReference = exports.kinesisAnalyticsApplicationOutputsKinesisStreamToTerraform = exports.KinesisAnalyticsApplicationOutputsKinesisFirehoseOutputReference = exports.kinesisAnalyticsApplicationOutputsKinesisFirehoseToTerraform = exports.KinesisAnalyticsApplicationInputsOutputReference = exports.kinesisAnalyticsApplicationInputsToTerraform = exports.KinesisAnalyticsApplicationInputsStartingPositionConfigurationList = exports.KinesisAnalyticsApplicationInputsStartingPositionConfigurationOutputReference = exports.kinesisAnalyticsApplicationInputsStartingPositionConfigurationToTerraform = exports.KinesisAnalyticsApplicationInputsSchemaOutputReference = exports.kinesisAnalyticsApplicationInputsSchemaToTerraform = exports.KinesisAnalyticsApplicationInputsSchemaRecordFormatOutputReference = exports.kinesisAnalyticsApplicationInputsSchemaRecordFormatToTerraform = exports.KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersOutputReference = exports.kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersToTerraform = exports.KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersJsonOutputReference = exports.kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersJsonToTerraform = exports.KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersCsvOutputReference = exports.kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersCsvToTerraform = exports.KinesisAnalyticsApplicationInputsSchemaRecordColumnsList = exports.KinesisAnalyticsApplicationInputsSchemaRecordColumnsOutputReference = exports.kinesisAnalyticsApplicationInputsSchemaRecordColumnsToTerraform = exports.KinesisAnalyticsApplicationInputsProcessingConfigurationOutputReference = exports.kinesisAnalyticsApplicationInputsProcessingConfigurationToTerraform = exports.KinesisAnalyticsApplicationInputsProcessingConfigurationLambdaOutputReference = exports.kinesisAnalyticsApplicationInputsProcessingConfigurationLambdaToTerraform = exports.KinesisAnalyticsApplicationInputsParallelismOutputReference = exports.kinesisAnalyticsApplicationInputsParallelismToTerraform = exports.KinesisAnalyticsApplicationInputsKinesisStreamOutputReference = exports.kinesisAnalyticsApplicationInputsKinesisStreamToTerraform = exports.KinesisAnalyticsApplicationInputsKinesisFirehoseOutputReference = exports.kinesisAnalyticsApplicationInputsKinesisFirehoseToTerraform = exports.KinesisAnalyticsApplicationCloudwatchLoggingOptionsOutputReference = exports.kinesisAnalyticsApplicationCloudwatchLoggingOptionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function kinesisAnalyticsApplicationCloudwatchLoggingOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        log_stream_arn: cdktf.stringToTerraform(struct.logStreamArn),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
    };
}
exports.kinesisAnalyticsApplicationCloudwatchLoggingOptionsToTerraform = kinesisAnalyticsApplicationCloudwatchLoggingOptionsToTerraform;
class KinesisAnalyticsApplicationCloudwatchLoggingOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._logStreamArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.logStreamArn = this._logStreamArn;
        }
        if (this._roleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._logStreamArn = undefined;
            this._roleArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._logStreamArn = value.logStreamArn;
            this._roleArn = value.roleArn;
        }
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get logStreamArn() {
        return this.getStringAttribute('log_stream_arn');
    }
    set logStreamArn(value) {
        this._logStreamArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get logStreamArnInput() {
        return this._logStreamArn;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
}
exports.KinesisAnalyticsApplicationCloudwatchLoggingOptionsOutputReference = KinesisAnalyticsApplicationCloudwatchLoggingOptionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationCloudwatchLoggingOptionsOutputReference[_a] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationCloudwatchLoggingOptionsOutputReference", version: "10.0.12" };
function kinesisAnalyticsApplicationInputsKinesisFirehoseToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resource_arn: cdktf.stringToTerraform(struct.resourceArn),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
    };
}
exports.kinesisAnalyticsApplicationInputsKinesisFirehoseToTerraform = kinesisAnalyticsApplicationInputsKinesisFirehoseToTerraform;
class KinesisAnalyticsApplicationInputsKinesisFirehoseOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resourceArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourceArn = this._resourceArn;
        }
        if (this._roleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._resourceArn = undefined;
            this._roleArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._resourceArn = value.resourceArn;
            this._roleArn = value.roleArn;
        }
    }
    get resourceArn() {
        return this.getStringAttribute('resource_arn');
    }
    set resourceArn(value) {
        this._resourceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourceArnInput() {
        return this._resourceArn;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
}
exports.KinesisAnalyticsApplicationInputsKinesisFirehoseOutputReference = KinesisAnalyticsApplicationInputsKinesisFirehoseOutputReference;
_b = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsKinesisFirehoseOutputReference[_b] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationInputsKinesisFirehoseOutputReference", version: "10.0.12" };
function kinesisAnalyticsApplicationInputsKinesisStreamToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resource_arn: cdktf.stringToTerraform(struct.resourceArn),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
    };
}
exports.kinesisAnalyticsApplicationInputsKinesisStreamToTerraform = kinesisAnalyticsApplicationInputsKinesisStreamToTerraform;
class KinesisAnalyticsApplicationInputsKinesisStreamOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resourceArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourceArn = this._resourceArn;
        }
        if (this._roleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._resourceArn = undefined;
            this._roleArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._resourceArn = value.resourceArn;
            this._roleArn = value.roleArn;
        }
    }
    get resourceArn() {
        return this.getStringAttribute('resource_arn');
    }
    set resourceArn(value) {
        this._resourceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourceArnInput() {
        return this._resourceArn;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
}
exports.KinesisAnalyticsApplicationInputsKinesisStreamOutputReference = KinesisAnalyticsApplicationInputsKinesisStreamOutputReference;
_c = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsKinesisStreamOutputReference[_c] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationInputsKinesisStreamOutputReference", version: "10.0.12" };
function kinesisAnalyticsApplicationInputsParallelismToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        count: cdktf.numberToTerraform(struct.count),
    };
}
exports.kinesisAnalyticsApplicationInputsParallelismToTerraform = kinesisAnalyticsApplicationInputsParallelismToTerraform;
class KinesisAnalyticsApplicationInputsParallelismOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._count !== undefined) {
            hasAnyValues = true;
            internalValueResult.count = this._count;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._count = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._count = value.count;
        }
    }
    get count() {
        return this.getNumberAttribute('count');
    }
    set count(value) {
        this._count = value;
    }
    resetCount() {
        this._count = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get countInput() {
        return this._count;
    }
}
exports.KinesisAnalyticsApplicationInputsParallelismOutputReference = KinesisAnalyticsApplicationInputsParallelismOutputReference;
_d = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsParallelismOutputReference[_d] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationInputsParallelismOutputReference", version: "10.0.12" };
function kinesisAnalyticsApplicationInputsProcessingConfigurationLambdaToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resource_arn: cdktf.stringToTerraform(struct.resourceArn),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
    };
}
exports.kinesisAnalyticsApplicationInputsProcessingConfigurationLambdaToTerraform = kinesisAnalyticsApplicationInputsProcessingConfigurationLambdaToTerraform;
class KinesisAnalyticsApplicationInputsProcessingConfigurationLambdaOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resourceArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourceArn = this._resourceArn;
        }
        if (this._roleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._resourceArn = undefined;
            this._roleArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._resourceArn = value.resourceArn;
            this._roleArn = value.roleArn;
        }
    }
    get resourceArn() {
        return this.getStringAttribute('resource_arn');
    }
    set resourceArn(value) {
        this._resourceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourceArnInput() {
        return this._resourceArn;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
}
exports.KinesisAnalyticsApplicationInputsProcessingConfigurationLambdaOutputReference = KinesisAnalyticsApplicationInputsProcessingConfigurationLambdaOutputReference;
_e = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsProcessingConfigurationLambdaOutputReference[_e] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationInputsProcessingConfigurationLambdaOutputReference", version: "10.0.12" };
function kinesisAnalyticsApplicationInputsProcessingConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        lambda: kinesisAnalyticsApplicationInputsProcessingConfigurationLambdaToTerraform(struct.lambda),
    };
}
exports.kinesisAnalyticsApplicationInputsProcessingConfigurationToTerraform = kinesisAnalyticsApplicationInputsProcessingConfigurationToTerraform;
class KinesisAnalyticsApplicationInputsProcessingConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // lambda - computed: false, optional: false, required: true
        this._lambda = new KinesisAnalyticsApplicationInputsProcessingConfigurationLambdaOutputReference(this, "lambda");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._lambda?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.lambda = this._lambda?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._lambda.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._lambda.internalValue = value.lambda;
        }
    }
    get lambda() {
        return this._lambda;
    }
    putLambda(value) {
        this._lambda.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get lambdaInput() {
        return this._lambda.internalValue;
    }
}
exports.KinesisAnalyticsApplicationInputsProcessingConfigurationOutputReference = KinesisAnalyticsApplicationInputsProcessingConfigurationOutputReference;
_f = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsProcessingConfigurationOutputReference[_f] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationInputsProcessingConfigurationOutputReference", version: "10.0.12" };
function kinesisAnalyticsApplicationInputsSchemaRecordColumnsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        mapping: cdktf.stringToTerraform(struct.mapping),
        name: cdktf.stringToTerraform(struct.name),
        sql_type: cdktf.stringToTerraform(struct.sqlType),
    };
}
exports.kinesisAnalyticsApplicationInputsSchemaRecordColumnsToTerraform = kinesisAnalyticsApplicationInputsSchemaRecordColumnsToTerraform;
class KinesisAnalyticsApplicationInputsSchemaRecordColumnsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._mapping !== undefined) {
            hasAnyValues = true;
            internalValueResult.mapping = this._mapping;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._sqlType !== undefined) {
            hasAnyValues = true;
            internalValueResult.sqlType = this._sqlType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._mapping = undefined;
            this._name = undefined;
            this._sqlType = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._mapping = value.mapping;
            this._name = value.name;
            this._sqlType = value.sqlType;
        }
    }
    get mapping() {
        return this.getStringAttribute('mapping');
    }
    set mapping(value) {
        this._mapping = value;
    }
    resetMapping() {
        this._mapping = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get mappingInput() {
        return this._mapping;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get sqlType() {
        return this.getStringAttribute('sql_type');
    }
    set sqlType(value) {
        this._sqlType = value;
    }
    // Temporarily expose input value. Use with caution.
    get sqlTypeInput() {
        return this._sqlType;
    }
}
exports.KinesisAnalyticsApplicationInputsSchemaRecordColumnsOutputReference = KinesisAnalyticsApplicationInputsSchemaRecordColumnsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsSchemaRecordColumnsOutputReference[_g] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationInputsSchemaRecordColumnsOutputReference", version: "10.0.12" };
class KinesisAnalyticsApplicationInputsSchemaRecordColumnsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new KinesisAnalyticsApplicationInputsSchemaRecordColumnsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.KinesisAnalyticsApplicationInputsSchemaRecordColumnsList = KinesisAnalyticsApplicationInputsSchemaRecordColumnsList;
_h = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsSchemaRecordColumnsList[_h] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationInputsSchemaRecordColumnsList", version: "10.0.12" };
function kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersCsvToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        record_column_delimiter: cdktf.stringToTerraform(struct.recordColumnDelimiter),
        record_row_delimiter: cdktf.stringToTerraform(struct.recordRowDelimiter),
    };
}
exports.kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersCsvToTerraform = kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersCsvToTerraform;
class KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersCsvOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._recordColumnDelimiter !== undefined) {
            hasAnyValues = true;
            internalValueResult.recordColumnDelimiter = this._recordColumnDelimiter;
        }
        if (this._recordRowDelimiter !== undefined) {
            hasAnyValues = true;
            internalValueResult.recordRowDelimiter = this._recordRowDelimiter;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._recordColumnDelimiter = undefined;
            this._recordRowDelimiter = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._recordColumnDelimiter = value.recordColumnDelimiter;
            this._recordRowDelimiter = value.recordRowDelimiter;
        }
    }
    get recordColumnDelimiter() {
        return this.getStringAttribute('record_column_delimiter');
    }
    set recordColumnDelimiter(value) {
        this._recordColumnDelimiter = value;
    }
    // Temporarily expose input value. Use with caution.
    get recordColumnDelimiterInput() {
        return this._recordColumnDelimiter;
    }
    get recordRowDelimiter() {
        return this.getStringAttribute('record_row_delimiter');
    }
    set recordRowDelimiter(value) {
        this._recordRowDelimiter = value;
    }
    // Temporarily expose input value. Use with caution.
    get recordRowDelimiterInput() {
        return this._recordRowDelimiter;
    }
}
exports.KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersCsvOutputReference = KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersCsvOutputReference;
_j = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersCsvOutputReference[_j] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersCsvOutputReference", version: "10.0.12" };
function kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersJsonToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        record_row_path: cdktf.stringToTerraform(struct.recordRowPath),
    };
}
exports.kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersJsonToTerraform = kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersJsonToTerraform;
class KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersJsonOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._recordRowPath !== undefined) {
            hasAnyValues = true;
            internalValueResult.recordRowPath = this._recordRowPath;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._recordRowPath = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._recordRowPath = value.recordRowPath;
        }
    }
    get recordRowPath() {
        return this.getStringAttribute('record_row_path');
    }
    set recordRowPath(value) {
        this._recordRowPath = value;
    }
    // Temporarily expose input value. Use with caution.
    get recordRowPathInput() {
        return this._recordRowPath;
    }
}
exports.KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersJsonOutputReference = KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersJsonOutputReference;
_k = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersJsonOutputReference[_k] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersJsonOutputReference", version: "10.0.12" };
function kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        csv: kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersCsvToTerraform(struct.csv),
        json: kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersJsonToTerraform(struct.json),
    };
}
exports.kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersToTerraform = kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersToTerraform;
class KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // csv - computed: false, optional: true, required: false
        this._csv = new KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersCsvOutputReference(this, "csv");
        // json - computed: false, optional: true, required: false
        this._json = new KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersJsonOutputReference(this, "json");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._csv?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.csv = this._csv?.internalValue;
        }
        if (this._json?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.json = this._json?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._csv.internalValue = undefined;
            this._json.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._csv.internalValue = value.csv;
            this._json.internalValue = value.json;
        }
    }
    get csv() {
        return this._csv;
    }
    putCsv(value) {
        this._csv.internalValue = value;
    }
    resetCsv() {
        this._csv.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get csvInput() {
        return this._csv.internalValue;
    }
    get json() {
        return this._json;
    }
    putJson(value) {
        this._json.internalValue = value;
    }
    resetJson() {
        this._json.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get jsonInput() {
        return this._json.internalValue;
    }
}
exports.KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersOutputReference = KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersOutputReference;
_l = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersOutputReference[_l] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersOutputReference", version: "10.0.12" };
function kinesisAnalyticsApplicationInputsSchemaRecordFormatToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        mapping_parameters: kinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersToTerraform(struct.mappingParameters),
    };
}
exports.kinesisAnalyticsApplicationInputsSchemaRecordFormatToTerraform = kinesisAnalyticsApplicationInputsSchemaRecordFormatToTerraform;
class KinesisAnalyticsApplicationInputsSchemaRecordFormatOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // mapping_parameters - computed: false, optional: true, required: false
        this._mappingParameters = new KinesisAnalyticsApplicationInputsSchemaRecordFormatMappingParametersOutputReference(this, "mapping_parameters");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._mappingParameters?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.mappingParameters = this._mappingParameters?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._mappingParameters.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._mappingParameters.internalValue = value.mappingParameters;
        }
    }
    // record_format_type - computed: true, optional: false, required: false
    get recordFormatType() {
        return this.getStringAttribute('record_format_type');
    }
    get mappingParameters() {
        return this._mappingParameters;
    }
    putMappingParameters(value) {
        this._mappingParameters.internalValue = value;
    }
    resetMappingParameters() {
        this._mappingParameters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get mappingParametersInput() {
        return this._mappingParameters.internalValue;
    }
}
exports.KinesisAnalyticsApplicationInputsSchemaRecordFormatOutputReference = KinesisAnalyticsApplicationInputsSchemaRecordFormatOutputReference;
_m = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsSchemaRecordFormatOutputReference[_m] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationInputsSchemaRecordFormatOutputReference", version: "10.0.12" };
function kinesisAnalyticsApplicationInputsSchemaToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        record_encoding: cdktf.stringToTerraform(struct.recordEncoding),
        record_columns: cdktf.listMapper(kinesisAnalyticsApplicationInputsSchemaRecordColumnsToTerraform, true)(struct.recordColumns),
        record_format: kinesisAnalyticsApplicationInputsSchemaRecordFormatToTerraform(struct.recordFormat),
    };
}
exports.kinesisAnalyticsApplicationInputsSchemaToTerraform = kinesisAnalyticsApplicationInputsSchemaToTerraform;
class KinesisAnalyticsApplicationInputsSchemaOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // record_columns - computed: false, optional: false, required: true
        this._recordColumns = new KinesisAnalyticsApplicationInputsSchemaRecordColumnsList(this, "record_columns", false);
        // record_format - computed: false, optional: false, required: true
        this._recordFormat = new KinesisAnalyticsApplicationInputsSchemaRecordFormatOutputReference(this, "record_format");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._recordEncoding !== undefined) {
            hasAnyValues = true;
            internalValueResult.recordEncoding = this._recordEncoding;
        }
        if (this._recordColumns?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.recordColumns = this._recordColumns?.internalValue;
        }
        if (this._recordFormat?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.recordFormat = this._recordFormat?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._recordEncoding = undefined;
            this._recordColumns.internalValue = undefined;
            this._recordFormat.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._recordEncoding = value.recordEncoding;
            this._recordColumns.internalValue = value.recordColumns;
            this._recordFormat.internalValue = value.recordFormat;
        }
    }
    get recordEncoding() {
        return this.getStringAttribute('record_encoding');
    }
    set recordEncoding(value) {
        this._recordEncoding = value;
    }
    resetRecordEncoding() {
        this._recordEncoding = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get recordEncodingInput() {
        return this._recordEncoding;
    }
    get recordColumns() {
        return this._recordColumns;
    }
    putRecordColumns(value) {
        this._recordColumns.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get recordColumnsInput() {
        return this._recordColumns.internalValue;
    }
    get recordFormat() {
        return this._recordFormat;
    }
    putRecordFormat(value) {
        this._recordFormat.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get recordFormatInput() {
        return this._recordFormat.internalValue;
    }
}
exports.KinesisAnalyticsApplicationInputsSchemaOutputReference = KinesisAnalyticsApplicationInputsSchemaOutputReference;
_o = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsSchemaOutputReference[_o] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationInputsSchemaOutputReference", version: "10.0.12" };
function kinesisAnalyticsApplicationInputsStartingPositionConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        starting_position: cdktf.stringToTerraform(struct.startingPosition),
    };
}
exports.kinesisAnalyticsApplicationInputsStartingPositionConfigurationToTerraform = kinesisAnalyticsApplicationInputsStartingPositionConfigurationToTerraform;
class KinesisAnalyticsApplicationInputsStartingPositionConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._startingPosition !== undefined) {
            hasAnyValues = true;
            internalValueResult.startingPosition = this._startingPosition;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._startingPosition = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._startingPosition = value.startingPosition;
        }
    }
    get startingPosition() {
        return this.getStringAttribute('starting_position');
    }
    set startingPosition(value) {
        this._startingPosition = value;
    }
    resetStartingPosition() {
        this._startingPosition = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startingPositionInput() {
        return this._startingPosition;
    }
}
exports.KinesisAnalyticsApplicationInputsStartingPositionConfigurationOutputReference = KinesisAnalyticsApplicationInputsStartingPositionConfigurationOutputReference;
_p = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsStartingPositionConfigurationOutputReference[_p] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationInputsStartingPositionConfigurationOutputReference", version: "10.0.12" };
class KinesisAnalyticsApplicationInputsStartingPositionConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new KinesisAnalyticsApplicationInputsStartingPositionConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.KinesisAnalyticsApplicationInputsStartingPositionConfigurationList = KinesisAnalyticsApplicationInputsStartingPositionConfigurationList;
_q = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsStartingPositionConfigurationList[_q] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationInputsStartingPositionConfigurationList", version: "10.0.12" };
function kinesisAnalyticsApplicationInputsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name_prefix: cdktf.stringToTerraform(struct.namePrefix),
        kinesis_firehose: kinesisAnalyticsApplicationInputsKinesisFirehoseToTerraform(struct.kinesisFirehose),
        kinesis_stream: kinesisAnalyticsApplicationInputsKinesisStreamToTerraform(struct.kinesisStream),
        parallelism: kinesisAnalyticsApplicationInputsParallelismToTerraform(struct.parallelism),
        processing_configuration: kinesisAnalyticsApplicationInputsProcessingConfigurationToTerraform(struct.processingConfiguration),
        schema: kinesisAnalyticsApplicationInputsSchemaToTerraform(struct.schema),
        starting_position_configuration: cdktf.listMapper(kinesisAnalyticsApplicationInputsStartingPositionConfigurationToTerraform, true)(struct.startingPositionConfiguration),
    };
}
exports.kinesisAnalyticsApplicationInputsToTerraform = kinesisAnalyticsApplicationInputsToTerraform;
class KinesisAnalyticsApplicationInputsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // kinesis_firehose - computed: false, optional: true, required: false
        this._kinesisFirehose = new KinesisAnalyticsApplicationInputsKinesisFirehoseOutputReference(this, "kinesis_firehose");
        // kinesis_stream - computed: false, optional: true, required: false
        this._kinesisStream = new KinesisAnalyticsApplicationInputsKinesisStreamOutputReference(this, "kinesis_stream");
        // parallelism - computed: false, optional: true, required: false
        this._parallelism = new KinesisAnalyticsApplicationInputsParallelismOutputReference(this, "parallelism");
        // processing_configuration - computed: false, optional: true, required: false
        this._processingConfiguration = new KinesisAnalyticsApplicationInputsProcessingConfigurationOutputReference(this, "processing_configuration");
        // schema - computed: false, optional: false, required: true
        this._schema = new KinesisAnalyticsApplicationInputsSchemaOutputReference(this, "schema");
        // starting_position_configuration - computed: false, optional: true, required: false
        this._startingPositionConfiguration = new KinesisAnalyticsApplicationInputsStartingPositionConfigurationList(this, "starting_position_configuration", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._namePrefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.namePrefix = this._namePrefix;
        }
        if (this._kinesisFirehose?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.kinesisFirehose = this._kinesisFirehose?.internalValue;
        }
        if (this._kinesisStream?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.kinesisStream = this._kinesisStream?.internalValue;
        }
        if (this._parallelism?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.parallelism = this._parallelism?.internalValue;
        }
        if (this._processingConfiguration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.processingConfiguration = this._processingConfiguration?.internalValue;
        }
        if (this._schema?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.schema = this._schema?.internalValue;
        }
        if (this._startingPositionConfiguration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.startingPositionConfiguration = this._startingPositionConfiguration?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._namePrefix = undefined;
            this._kinesisFirehose.internalValue = undefined;
            this._kinesisStream.internalValue = undefined;
            this._parallelism.internalValue = undefined;
            this._processingConfiguration.internalValue = undefined;
            this._schema.internalValue = undefined;
            this._startingPositionConfiguration.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._namePrefix = value.namePrefix;
            this._kinesisFirehose.internalValue = value.kinesisFirehose;
            this._kinesisStream.internalValue = value.kinesisStream;
            this._parallelism.internalValue = value.parallelism;
            this._processingConfiguration.internalValue = value.processingConfiguration;
            this._schema.internalValue = value.schema;
            this._startingPositionConfiguration.internalValue = value.startingPositionConfiguration;
        }
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get namePrefix() {
        return this.getStringAttribute('name_prefix');
    }
    set namePrefix(value) {
        this._namePrefix = value;
    }
    // Temporarily expose input value. Use with caution.
    get namePrefixInput() {
        return this._namePrefix;
    }
    // stream_names - computed: true, optional: false, required: false
    get streamNames() {
        return this.getListAttribute('stream_names');
    }
    get kinesisFirehose() {
        return this._kinesisFirehose;
    }
    putKinesisFirehose(value) {
        this._kinesisFirehose.internalValue = value;
    }
    resetKinesisFirehose() {
        this._kinesisFirehose.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kinesisFirehoseInput() {
        return this._kinesisFirehose.internalValue;
    }
    get kinesisStream() {
        return this._kinesisStream;
    }
    putKinesisStream(value) {
        this._kinesisStream.internalValue = value;
    }
    resetKinesisStream() {
        this._kinesisStream.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kinesisStreamInput() {
        return this._kinesisStream.internalValue;
    }
    get parallelism() {
        return this._parallelism;
    }
    putParallelism(value) {
        this._parallelism.internalValue = value;
    }
    resetParallelism() {
        this._parallelism.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parallelismInput() {
        return this._parallelism.internalValue;
    }
    get processingConfiguration() {
        return this._processingConfiguration;
    }
    putProcessingConfiguration(value) {
        this._processingConfiguration.internalValue = value;
    }
    resetProcessingConfiguration() {
        this._processingConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get processingConfigurationInput() {
        return this._processingConfiguration.internalValue;
    }
    get schema() {
        return this._schema;
    }
    putSchema(value) {
        this._schema.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get schemaInput() {
        return this._schema.internalValue;
    }
    get startingPositionConfiguration() {
        return this._startingPositionConfiguration;
    }
    putStartingPositionConfiguration(value) {
        this._startingPositionConfiguration.internalValue = value;
    }
    resetStartingPositionConfiguration() {
        this._startingPositionConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startingPositionConfigurationInput() {
        return this._startingPositionConfiguration.internalValue;
    }
}
exports.KinesisAnalyticsApplicationInputsOutputReference = KinesisAnalyticsApplicationInputsOutputReference;
_r = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationInputsOutputReference[_r] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationInputsOutputReference", version: "10.0.12" };
function kinesisAnalyticsApplicationOutputsKinesisFirehoseToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resource_arn: cdktf.stringToTerraform(struct.resourceArn),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
    };
}
exports.kinesisAnalyticsApplicationOutputsKinesisFirehoseToTerraform = kinesisAnalyticsApplicationOutputsKinesisFirehoseToTerraform;
class KinesisAnalyticsApplicationOutputsKinesisFirehoseOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resourceArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourceArn = this._resourceArn;
        }
        if (this._roleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._resourceArn = undefined;
            this._roleArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._resourceArn = value.resourceArn;
            this._roleArn = value.roleArn;
        }
    }
    get resourceArn() {
        return this.getStringAttribute('resource_arn');
    }
    set resourceArn(value) {
        this._resourceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourceArnInput() {
        return this._resourceArn;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
}
exports.KinesisAnalyticsApplicationOutputsKinesisFirehoseOutputReference = KinesisAnalyticsApplicationOutputsKinesisFirehoseOutputReference;
_s = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationOutputsKinesisFirehoseOutputReference[_s] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationOutputsKinesisFirehoseOutputReference", version: "10.0.12" };
function kinesisAnalyticsApplicationOutputsKinesisStreamToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resource_arn: cdktf.stringToTerraform(struct.resourceArn),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
    };
}
exports.kinesisAnalyticsApplicationOutputsKinesisStreamToTerraform = kinesisAnalyticsApplicationOutputsKinesisStreamToTerraform;
class KinesisAnalyticsApplicationOutputsKinesisStreamOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resourceArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourceArn = this._resourceArn;
        }
        if (this._roleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._resourceArn = undefined;
            this._roleArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._resourceArn = value.resourceArn;
            this._roleArn = value.roleArn;
        }
    }
    get resourceArn() {
        return this.getStringAttribute('resource_arn');
    }
    set resourceArn(value) {
        this._resourceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourceArnInput() {
        return this._resourceArn;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
}
exports.KinesisAnalyticsApplicationOutputsKinesisStreamOutputReference = KinesisAnalyticsApplicationOutputsKinesisStreamOutputReference;
_t = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationOutputsKinesisStreamOutputReference[_t] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationOutputsKinesisStreamOutputReference", version: "10.0.12" };
function kinesisAnalyticsApplicationOutputsLambdaToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resource_arn: cdktf.stringToTerraform(struct.resourceArn),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
    };
}
exports.kinesisAnalyticsApplicationOutputsLambdaToTerraform = kinesisAnalyticsApplicationOutputsLambdaToTerraform;
class KinesisAnalyticsApplicationOutputsLambdaOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resourceArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourceArn = this._resourceArn;
        }
        if (this._roleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._resourceArn = undefined;
            this._roleArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._resourceArn = value.resourceArn;
            this._roleArn = value.roleArn;
        }
    }
    get resourceArn() {
        return this.getStringAttribute('resource_arn');
    }
    set resourceArn(value) {
        this._resourceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourceArnInput() {
        return this._resourceArn;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
}
exports.KinesisAnalyticsApplicationOutputsLambdaOutputReference = KinesisAnalyticsApplicationOutputsLambdaOutputReference;
_u = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationOutputsLambdaOutputReference[_u] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationOutputsLambdaOutputReference", version: "10.0.12" };
function kinesisAnalyticsApplicationOutputsSchemaToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        record_format_type: cdktf.stringToTerraform(struct.recordFormatType),
    };
}
exports.kinesisAnalyticsApplicationOutputsSchemaToTerraform = kinesisAnalyticsApplicationOutputsSchemaToTerraform;
class KinesisAnalyticsApplicationOutputsSchemaOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._recordFormatType !== undefined) {
            hasAnyValues = true;
            internalValueResult.recordFormatType = this._recordFormatType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._recordFormatType = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._recordFormatType = value.recordFormatType;
        }
    }
    get recordFormatType() {
        return this.getStringAttribute('record_format_type');
    }
    set recordFormatType(value) {
        this._recordFormatType = value;
    }
    // Temporarily expose input value. Use with caution.
    get recordFormatTypeInput() {
        return this._recordFormatType;
    }
}
exports.KinesisAnalyticsApplicationOutputsSchemaOutputReference = KinesisAnalyticsApplicationOutputsSchemaOutputReference;
_v = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationOutputsSchemaOutputReference[_v] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationOutputsSchemaOutputReference", version: "10.0.12" };
function kinesisAnalyticsApplicationOutputsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        kinesis_firehose: kinesisAnalyticsApplicationOutputsKinesisFirehoseToTerraform(struct.kinesisFirehose),
        kinesis_stream: kinesisAnalyticsApplicationOutputsKinesisStreamToTerraform(struct.kinesisStream),
        lambda: kinesisAnalyticsApplicationOutputsLambdaToTerraform(struct.lambda),
        schema: kinesisAnalyticsApplicationOutputsSchemaToTerraform(struct.schema),
    };
}
exports.kinesisAnalyticsApplicationOutputsToTerraform = kinesisAnalyticsApplicationOutputsToTerraform;
class KinesisAnalyticsApplicationOutputsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // kinesis_firehose - computed: false, optional: true, required: false
        this._kinesisFirehose = new KinesisAnalyticsApplicationOutputsKinesisFirehoseOutputReference(this, "kinesis_firehose");
        // kinesis_stream - computed: false, optional: true, required: false
        this._kinesisStream = new KinesisAnalyticsApplicationOutputsKinesisStreamOutputReference(this, "kinesis_stream");
        // lambda - computed: false, optional: true, required: false
        this._lambda = new KinesisAnalyticsApplicationOutputsLambdaOutputReference(this, "lambda");
        // schema - computed: false, optional: false, required: true
        this._schema = new KinesisAnalyticsApplicationOutputsSchemaOutputReference(this, "schema");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._kinesisFirehose?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.kinesisFirehose = this._kinesisFirehose?.internalValue;
        }
        if (this._kinesisStream?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.kinesisStream = this._kinesisStream?.internalValue;
        }
        if (this._lambda?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.lambda = this._lambda?.internalValue;
        }
        if (this._schema?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.schema = this._schema?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._kinesisFirehose.internalValue = undefined;
            this._kinesisStream.internalValue = undefined;
            this._lambda.internalValue = undefined;
            this._schema.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._kinesisFirehose.internalValue = value.kinesisFirehose;
            this._kinesisStream.internalValue = value.kinesisStream;
            this._lambda.internalValue = value.lambda;
            this._schema.internalValue = value.schema;
        }
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get kinesisFirehose() {
        return this._kinesisFirehose;
    }
    putKinesisFirehose(value) {
        this._kinesisFirehose.internalValue = value;
    }
    resetKinesisFirehose() {
        this._kinesisFirehose.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kinesisFirehoseInput() {
        return this._kinesisFirehose.internalValue;
    }
    get kinesisStream() {
        return this._kinesisStream;
    }
    putKinesisStream(value) {
        this._kinesisStream.internalValue = value;
    }
    resetKinesisStream() {
        this._kinesisStream.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kinesisStreamInput() {
        return this._kinesisStream.internalValue;
    }
    get lambda() {
        return this._lambda;
    }
    putLambda(value) {
        this._lambda.internalValue = value;
    }
    resetLambda() {
        this._lambda.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lambdaInput() {
        return this._lambda.internalValue;
    }
    get schema() {
        return this._schema;
    }
    putSchema(value) {
        this._schema.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get schemaInput() {
        return this._schema.internalValue;
    }
}
exports.KinesisAnalyticsApplicationOutputsOutputReference = KinesisAnalyticsApplicationOutputsOutputReference;
_w = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationOutputsOutputReference[_w] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationOutputsOutputReference", version: "10.0.12" };
class KinesisAnalyticsApplicationOutputsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new KinesisAnalyticsApplicationOutputsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.KinesisAnalyticsApplicationOutputsList = KinesisAnalyticsApplicationOutputsList;
_x = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationOutputsList[_x] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationOutputsList", version: "10.0.12" };
function kinesisAnalyticsApplicationReferenceDataSourcesS3ToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket_arn: cdktf.stringToTerraform(struct.bucketArn),
        file_key: cdktf.stringToTerraform(struct.fileKey),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
    };
}
exports.kinesisAnalyticsApplicationReferenceDataSourcesS3ToTerraform = kinesisAnalyticsApplicationReferenceDataSourcesS3ToTerraform;
class KinesisAnalyticsApplicationReferenceDataSourcesS3OutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucketArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.bucketArn = this._bucketArn;
        }
        if (this._fileKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.fileKey = this._fileKey;
        }
        if (this._roleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucketArn = undefined;
            this._fileKey = undefined;
            this._roleArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucketArn = value.bucketArn;
            this._fileKey = value.fileKey;
            this._roleArn = value.roleArn;
        }
    }
    get bucketArn() {
        return this.getStringAttribute('bucket_arn');
    }
    set bucketArn(value) {
        this._bucketArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get bucketArnInput() {
        return this._bucketArn;
    }
    get fileKey() {
        return this.getStringAttribute('file_key');
    }
    set fileKey(value) {
        this._fileKey = value;
    }
    // Temporarily expose input value. Use with caution.
    get fileKeyInput() {
        return this._fileKey;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
}
exports.KinesisAnalyticsApplicationReferenceDataSourcesS3OutputReference = KinesisAnalyticsApplicationReferenceDataSourcesS3OutputReference;
_y = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationReferenceDataSourcesS3OutputReference[_y] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationReferenceDataSourcesS3OutputReference", version: "10.0.12" };
function kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordColumnsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        mapping: cdktf.stringToTerraform(struct.mapping),
        name: cdktf.stringToTerraform(struct.name),
        sql_type: cdktf.stringToTerraform(struct.sqlType),
    };
}
exports.kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordColumnsToTerraform = kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordColumnsToTerraform;
class KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordColumnsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._mapping !== undefined) {
            hasAnyValues = true;
            internalValueResult.mapping = this._mapping;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._sqlType !== undefined) {
            hasAnyValues = true;
            internalValueResult.sqlType = this._sqlType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._mapping = undefined;
            this._name = undefined;
            this._sqlType = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._mapping = value.mapping;
            this._name = value.name;
            this._sqlType = value.sqlType;
        }
    }
    get mapping() {
        return this.getStringAttribute('mapping');
    }
    set mapping(value) {
        this._mapping = value;
    }
    resetMapping() {
        this._mapping = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get mappingInput() {
        return this._mapping;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get sqlType() {
        return this.getStringAttribute('sql_type');
    }
    set sqlType(value) {
        this._sqlType = value;
    }
    // Temporarily expose input value. Use with caution.
    get sqlTypeInput() {
        return this._sqlType;
    }
}
exports.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordColumnsOutputReference = KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordColumnsOutputReference;
_z = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordColumnsOutputReference[_z] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordColumnsOutputReference", version: "10.0.12" };
class KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordColumnsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordColumnsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordColumnsList = KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordColumnsList;
_0 = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordColumnsList[_0] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordColumnsList", version: "10.0.12" };
function kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersCsvToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        record_column_delimiter: cdktf.stringToTerraform(struct.recordColumnDelimiter),
        record_row_delimiter: cdktf.stringToTerraform(struct.recordRowDelimiter),
    };
}
exports.kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersCsvToTerraform = kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersCsvToTerraform;
class KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersCsvOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._recordColumnDelimiter !== undefined) {
            hasAnyValues = true;
            internalValueResult.recordColumnDelimiter = this._recordColumnDelimiter;
        }
        if (this._recordRowDelimiter !== undefined) {
            hasAnyValues = true;
            internalValueResult.recordRowDelimiter = this._recordRowDelimiter;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._recordColumnDelimiter = undefined;
            this._recordRowDelimiter = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._recordColumnDelimiter = value.recordColumnDelimiter;
            this._recordRowDelimiter = value.recordRowDelimiter;
        }
    }
    get recordColumnDelimiter() {
        return this.getStringAttribute('record_column_delimiter');
    }
    set recordColumnDelimiter(value) {
        this._recordColumnDelimiter = value;
    }
    // Temporarily expose input value. Use with caution.
    get recordColumnDelimiterInput() {
        return this._recordColumnDelimiter;
    }
    get recordRowDelimiter() {
        return this.getStringAttribute('record_row_delimiter');
    }
    set recordRowDelimiter(value) {
        this._recordRowDelimiter = value;
    }
    // Temporarily expose input value. Use with caution.
    get recordRowDelimiterInput() {
        return this._recordRowDelimiter;
    }
}
exports.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersCsvOutputReference = KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersCsvOutputReference;
_1 = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersCsvOutputReference[_1] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersCsvOutputReference", version: "10.0.12" };
function kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersJsonToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        record_row_path: cdktf.stringToTerraform(struct.recordRowPath),
    };
}
exports.kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersJsonToTerraform = kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersJsonToTerraform;
class KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersJsonOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._recordRowPath !== undefined) {
            hasAnyValues = true;
            internalValueResult.recordRowPath = this._recordRowPath;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._recordRowPath = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._recordRowPath = value.recordRowPath;
        }
    }
    get recordRowPath() {
        return this.getStringAttribute('record_row_path');
    }
    set recordRowPath(value) {
        this._recordRowPath = value;
    }
    // Temporarily expose input value. Use with caution.
    get recordRowPathInput() {
        return this._recordRowPath;
    }
}
exports.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersJsonOutputReference = KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersJsonOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersJsonOutputReference[_2] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersJsonOutputReference", version: "10.0.12" };
function kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        csv: kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersCsvToTerraform(struct.csv),
        json: kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersJsonToTerraform(struct.json),
    };
}
exports.kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersToTerraform = kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersToTerraform;
class KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // csv - computed: false, optional: true, required: false
        this._csv = new KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersCsvOutputReference(this, "csv");
        // json - computed: false, optional: true, required: false
        this._json = new KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersJsonOutputReference(this, "json");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._csv?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.csv = this._csv?.internalValue;
        }
        if (this._json?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.json = this._json?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._csv.internalValue = undefined;
            this._json.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._csv.internalValue = value.csv;
            this._json.internalValue = value.json;
        }
    }
    get csv() {
        return this._csv;
    }
    putCsv(value) {
        this._csv.internalValue = value;
    }
    resetCsv() {
        this._csv.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get csvInput() {
        return this._csv.internalValue;
    }
    get json() {
        return this._json;
    }
    putJson(value) {
        this._json.internalValue = value;
    }
    resetJson() {
        this._json.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get jsonInput() {
        return this._json.internalValue;
    }
}
exports.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersOutputReference = KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersOutputReference;
_3 = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersOutputReference[_3] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersOutputReference", version: "10.0.12" };
function kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        mapping_parameters: kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersToTerraform(struct.mappingParameters),
    };
}
exports.kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatToTerraform = kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatToTerraform;
class KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // mapping_parameters - computed: false, optional: true, required: false
        this._mappingParameters = new KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatMappingParametersOutputReference(this, "mapping_parameters");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._mappingParameters?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.mappingParameters = this._mappingParameters?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._mappingParameters.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._mappingParameters.internalValue = value.mappingParameters;
        }
    }
    // record_format_type - computed: true, optional: false, required: false
    get recordFormatType() {
        return this.getStringAttribute('record_format_type');
    }
    get mappingParameters() {
        return this._mappingParameters;
    }
    putMappingParameters(value) {
        this._mappingParameters.internalValue = value;
    }
    resetMappingParameters() {
        this._mappingParameters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get mappingParametersInput() {
        return this._mappingParameters.internalValue;
    }
}
exports.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatOutputReference = KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatOutputReference[_4] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatOutputReference", version: "10.0.12" };
function kinesisAnalyticsApplicationReferenceDataSourcesSchemaToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        record_encoding: cdktf.stringToTerraform(struct.recordEncoding),
        record_columns: cdktf.listMapper(kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordColumnsToTerraform, true)(struct.recordColumns),
        record_format: kinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatToTerraform(struct.recordFormat),
    };
}
exports.kinesisAnalyticsApplicationReferenceDataSourcesSchemaToTerraform = kinesisAnalyticsApplicationReferenceDataSourcesSchemaToTerraform;
class KinesisAnalyticsApplicationReferenceDataSourcesSchemaOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // record_columns - computed: false, optional: false, required: true
        this._recordColumns = new KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordColumnsList(this, "record_columns", false);
        // record_format - computed: false, optional: false, required: true
        this._recordFormat = new KinesisAnalyticsApplicationReferenceDataSourcesSchemaRecordFormatOutputReference(this, "record_format");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._recordEncoding !== undefined) {
            hasAnyValues = true;
            internalValueResult.recordEncoding = this._recordEncoding;
        }
        if (this._recordColumns?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.recordColumns = this._recordColumns?.internalValue;
        }
        if (this._recordFormat?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.recordFormat = this._recordFormat?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._recordEncoding = undefined;
            this._recordColumns.internalValue = undefined;
            this._recordFormat.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._recordEncoding = value.recordEncoding;
            this._recordColumns.internalValue = value.recordColumns;
            this._recordFormat.internalValue = value.recordFormat;
        }
    }
    get recordEncoding() {
        return this.getStringAttribute('record_encoding');
    }
    set recordEncoding(value) {
        this._recordEncoding = value;
    }
    resetRecordEncoding() {
        this._recordEncoding = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get recordEncodingInput() {
        return this._recordEncoding;
    }
    get recordColumns() {
        return this._recordColumns;
    }
    putRecordColumns(value) {
        this._recordColumns.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get recordColumnsInput() {
        return this._recordColumns.internalValue;
    }
    get recordFormat() {
        return this._recordFormat;
    }
    putRecordFormat(value) {
        this._recordFormat.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get recordFormatInput() {
        return this._recordFormat.internalValue;
    }
}
exports.KinesisAnalyticsApplicationReferenceDataSourcesSchemaOutputReference = KinesisAnalyticsApplicationReferenceDataSourcesSchemaOutputReference;
_5 = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationReferenceDataSourcesSchemaOutputReference[_5] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationReferenceDataSourcesSchemaOutputReference", version: "10.0.12" };
function kinesisAnalyticsApplicationReferenceDataSourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        table_name: cdktf.stringToTerraform(struct.tableName),
        s3: kinesisAnalyticsApplicationReferenceDataSourcesS3ToTerraform(struct.s3),
        schema: kinesisAnalyticsApplicationReferenceDataSourcesSchemaToTerraform(struct.schema),
    };
}
exports.kinesisAnalyticsApplicationReferenceDataSourcesToTerraform = kinesisAnalyticsApplicationReferenceDataSourcesToTerraform;
class KinesisAnalyticsApplicationReferenceDataSourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // s3 - computed: false, optional: false, required: true
        this._s3 = new KinesisAnalyticsApplicationReferenceDataSourcesS3OutputReference(this, "s3");
        // schema - computed: false, optional: false, required: true
        this._schema = new KinesisAnalyticsApplicationReferenceDataSourcesSchemaOutputReference(this, "schema");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._tableName !== undefined) {
            hasAnyValues = true;
            internalValueResult.tableName = this._tableName;
        }
        if (this._s3?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3 = this._s3?.internalValue;
        }
        if (this._schema?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.schema = this._schema?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._tableName = undefined;
            this._s3.internalValue = undefined;
            this._schema.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._tableName = value.tableName;
            this._s3.internalValue = value.s3;
            this._schema.internalValue = value.schema;
        }
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get tableName() {
        return this.getStringAttribute('table_name');
    }
    set tableName(value) {
        this._tableName = value;
    }
    // Temporarily expose input value. Use with caution.
    get tableNameInput() {
        return this._tableName;
    }
    get s3() {
        return this._s3;
    }
    putS3(value) {
        this._s3.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get s3Input() {
        return this._s3.internalValue;
    }
    get schema() {
        return this._schema;
    }
    putSchema(value) {
        this._schema.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get schemaInput() {
        return this._schema.internalValue;
    }
}
exports.KinesisAnalyticsApplicationReferenceDataSourcesOutputReference = KinesisAnalyticsApplicationReferenceDataSourcesOutputReference;
_6 = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplicationReferenceDataSourcesOutputReference[_6] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplicationReferenceDataSourcesOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/kinesis_analytics_application aws_kinesis_analytics_application}
*/
class KinesisAnalyticsApplication extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/kinesis_analytics_application aws_kinesis_analytics_application} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options KinesisAnalyticsApplicationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_kinesis_analytics_application',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // cloudwatch_logging_options - computed: false, optional: true, required: false
        this._cloudwatchLoggingOptions = new KinesisAnalyticsApplicationCloudwatchLoggingOptionsOutputReference(this, "cloudwatch_logging_options");
        // inputs - computed: false, optional: true, required: false
        this._inputs = new KinesisAnalyticsApplicationInputsOutputReference(this, "inputs");
        // outputs - computed: false, optional: true, required: false
        this._outputs = new KinesisAnalyticsApplicationOutputsList(this, "outputs", true);
        // reference_data_sources - computed: false, optional: true, required: false
        this._referenceDataSources = new KinesisAnalyticsApplicationReferenceDataSourcesOutputReference(this, "reference_data_sources");
        this._code = config.code;
        this._description = config.description;
        this._id = config.id;
        this._name = config.name;
        this._startApplication = config.startApplication;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._cloudwatchLoggingOptions.internalValue = config.cloudwatchLoggingOptions;
        this._inputs.internalValue = config.inputs;
        this._outputs.internalValue = config.outputs;
        this._referenceDataSources.internalValue = config.referenceDataSources;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get code() {
        return this.getStringAttribute('code');
    }
    set code(value) {
        this._code = value;
    }
    resetCode() {
        this._code = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get codeInput() {
        return this._code;
    }
    // create_timestamp - computed: true, optional: false, required: false
    get createTimestamp() {
        return this.getStringAttribute('create_timestamp');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // last_update_timestamp - computed: true, optional: false, required: false
    get lastUpdateTimestamp() {
        return this.getStringAttribute('last_update_timestamp');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get startApplication() {
        return this.getBooleanAttribute('start_application');
    }
    set startApplication(value) {
        this._startApplication = value;
    }
    resetStartApplication() {
        this._startApplication = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startApplicationInput() {
        return this._startApplication;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getNumberAttribute('version');
    }
    get cloudwatchLoggingOptions() {
        return this._cloudwatchLoggingOptions;
    }
    putCloudwatchLoggingOptions(value) {
        this._cloudwatchLoggingOptions.internalValue = value;
    }
    resetCloudwatchLoggingOptions() {
        this._cloudwatchLoggingOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudwatchLoggingOptionsInput() {
        return this._cloudwatchLoggingOptions.internalValue;
    }
    get inputs() {
        return this._inputs;
    }
    putInputs(value) {
        this._inputs.internalValue = value;
    }
    resetInputs() {
        this._inputs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get inputsInput() {
        return this._inputs.internalValue;
    }
    get outputs() {
        return this._outputs;
    }
    putOutputs(value) {
        this._outputs.internalValue = value;
    }
    resetOutputs() {
        this._outputs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get outputsInput() {
        return this._outputs.internalValue;
    }
    get referenceDataSources() {
        return this._referenceDataSources;
    }
    putReferenceDataSources(value) {
        this._referenceDataSources.internalValue = value;
    }
    resetReferenceDataSources() {
        this._referenceDataSources.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get referenceDataSourcesInput() {
        return this._referenceDataSources.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            code: cdktf.stringToTerraform(this._code),
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            start_application: cdktf.booleanToTerraform(this._startApplication),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            cloudwatch_logging_options: kinesisAnalyticsApplicationCloudwatchLoggingOptionsToTerraform(this._cloudwatchLoggingOptions.internalValue),
            inputs: kinesisAnalyticsApplicationInputsToTerraform(this._inputs.internalValue),
            outputs: cdktf.listMapper(kinesisAnalyticsApplicationOutputsToTerraform, true)(this._outputs.internalValue),
            reference_data_sources: kinesisAnalyticsApplicationReferenceDataSourcesToTerraform(this._referenceDataSources.internalValue),
        };
    }
}
exports.KinesisAnalyticsApplication = KinesisAnalyticsApplication;
_7 = JSII_RTTI_SYMBOL_1;
KinesisAnalyticsApplication[_7] = { fqn: "@cdktf/provider-aws.kinesisAnalyticsApplication.KinesisAnalyticsApplication", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
KinesisAnalyticsApplication.tfResourceType = "aws_kinesis_analytics_application";
