// @cdktf/provider-aws/lib/codebuild-webhook/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodebuildWebhook = exports.CodebuildWebhookFilterGroupList = exports.CodebuildWebhookFilterGroupOutputReference = exports.codebuildWebhookFilterGroupToTerraform = exports.CodebuildWebhookFilterGroupFilterList = exports.CodebuildWebhookFilterGroupFilterOutputReference = exports.codebuildWebhookFilterGroupFilterToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function codebuildWebhookFilterGroupFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        exclude_matched_pattern: cdktf.booleanToTerraform(struct.excludeMatchedPattern),
        pattern: cdktf.stringToTerraform(struct.pattern),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.codebuildWebhookFilterGroupFilterToTerraform = codebuildWebhookFilterGroupFilterToTerraform;
class CodebuildWebhookFilterGroupFilterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._excludeMatchedPattern !== undefined) {
            hasAnyValues = true;
            internalValueResult.excludeMatchedPattern = this._excludeMatchedPattern;
        }
        if (this._pattern !== undefined) {
            hasAnyValues = true;
            internalValueResult.pattern = this._pattern;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._excludeMatchedPattern = undefined;
            this._pattern = undefined;
            this._type = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._excludeMatchedPattern = value.excludeMatchedPattern;
            this._pattern = value.pattern;
            this._type = value.type;
        }
    }
    get excludeMatchedPattern() {
        return this.getBooleanAttribute('exclude_matched_pattern');
    }
    set excludeMatchedPattern(value) {
        this._excludeMatchedPattern = value;
    }
    resetExcludeMatchedPattern() {
        this._excludeMatchedPattern = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get excludeMatchedPatternInput() {
        return this._excludeMatchedPattern;
    }
    get pattern() {
        return this.getStringAttribute('pattern');
    }
    set pattern(value) {
        this._pattern = value;
    }
    // Temporarily expose input value. Use with caution.
    get patternInput() {
        return this._pattern;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.CodebuildWebhookFilterGroupFilterOutputReference = CodebuildWebhookFilterGroupFilterOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CodebuildWebhookFilterGroupFilterOutputReference[_a] = { fqn: "@cdktf/provider-aws.codebuildWebhook.CodebuildWebhookFilterGroupFilterOutputReference", version: "10.0.12" };
class CodebuildWebhookFilterGroupFilterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CodebuildWebhookFilterGroupFilterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CodebuildWebhookFilterGroupFilterList = CodebuildWebhookFilterGroupFilterList;
_b = JSII_RTTI_SYMBOL_1;
CodebuildWebhookFilterGroupFilterList[_b] = { fqn: "@cdktf/provider-aws.codebuildWebhook.CodebuildWebhookFilterGroupFilterList", version: "10.0.12" };
function codebuildWebhookFilterGroupToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        filter: cdktf.listMapper(codebuildWebhookFilterGroupFilterToTerraform, true)(struct.filter),
    };
}
exports.codebuildWebhookFilterGroupToTerraform = codebuildWebhookFilterGroupToTerraform;
class CodebuildWebhookFilterGroupOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // filter - computed: false, optional: true, required: false
        this._filter = new CodebuildWebhookFilterGroupFilterList(this, "filter", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._filter?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.filter = this._filter?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._filter.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._filter.internalValue = value.filter;
        }
    }
    get filter() {
        return this._filter;
    }
    putFilter(value) {
        this._filter.internalValue = value;
    }
    resetFilter() {
        this._filter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter.internalValue;
    }
}
exports.CodebuildWebhookFilterGroupOutputReference = CodebuildWebhookFilterGroupOutputReference;
_c = JSII_RTTI_SYMBOL_1;
CodebuildWebhookFilterGroupOutputReference[_c] = { fqn: "@cdktf/provider-aws.codebuildWebhook.CodebuildWebhookFilterGroupOutputReference", version: "10.0.12" };
class CodebuildWebhookFilterGroupList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CodebuildWebhookFilterGroupOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CodebuildWebhookFilterGroupList = CodebuildWebhookFilterGroupList;
_d = JSII_RTTI_SYMBOL_1;
CodebuildWebhookFilterGroupList[_d] = { fqn: "@cdktf/provider-aws.codebuildWebhook.CodebuildWebhookFilterGroupList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/codebuild_webhook aws_codebuild_webhook}
*/
class CodebuildWebhook extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/codebuild_webhook aws_codebuild_webhook} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CodebuildWebhookConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_codebuild_webhook',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // filter_group - computed: false, optional: true, required: false
        this._filterGroup = new CodebuildWebhookFilterGroupList(this, "filter_group", true);
        this._branchFilter = config.branchFilter;
        this._buildType = config.buildType;
        this._id = config.id;
        this._projectName = config.projectName;
        this._filterGroup.internalValue = config.filterGroup;
    }
    get branchFilter() {
        return this.getStringAttribute('branch_filter');
    }
    set branchFilter(value) {
        this._branchFilter = value;
    }
    resetBranchFilter() {
        this._branchFilter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get branchFilterInput() {
        return this._branchFilter;
    }
    get buildType() {
        return this.getStringAttribute('build_type');
    }
    set buildType(value) {
        this._buildType = value;
    }
    resetBuildType() {
        this._buildType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get buildTypeInput() {
        return this._buildType;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // payload_url - computed: true, optional: false, required: false
    get payloadUrl() {
        return this.getStringAttribute('payload_url');
    }
    get projectName() {
        return this.getStringAttribute('project_name');
    }
    set projectName(value) {
        this._projectName = value;
    }
    // Temporarily expose input value. Use with caution.
    get projectNameInput() {
        return this._projectName;
    }
    // secret - computed: true, optional: false, required: false
    get secret() {
        return this.getStringAttribute('secret');
    }
    // url - computed: true, optional: false, required: false
    get url() {
        return this.getStringAttribute('url');
    }
    get filterGroup() {
        return this._filterGroup;
    }
    putFilterGroup(value) {
        this._filterGroup.internalValue = value;
    }
    resetFilterGroup() {
        this._filterGroup.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterGroupInput() {
        return this._filterGroup.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            branch_filter: cdktf.stringToTerraform(this._branchFilter),
            build_type: cdktf.stringToTerraform(this._buildType),
            id: cdktf.stringToTerraform(this._id),
            project_name: cdktf.stringToTerraform(this._projectName),
            filter_group: cdktf.listMapper(codebuildWebhookFilterGroupToTerraform, true)(this._filterGroup.internalValue),
        };
    }
}
exports.CodebuildWebhook = CodebuildWebhook;
_e = JSII_RTTI_SYMBOL_1;
CodebuildWebhook[_e] = { fqn: "@cdktf/provider-aws.codebuildWebhook.CodebuildWebhook", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
CodebuildWebhook.tfResourceType = "aws_codebuild_webhook";
