// @cdktf/provider-aws/lib/s3-bucket-acl/index.js
"use strict";
var _a, _b, _c, _d, _e, _f;
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3BucketAcl = exports.S3BucketAclAccessControlPolicyOutputReference = exports.s3BucketAclAccessControlPolicyToTerraform = exports.S3BucketAclAccessControlPolicyOwnerOutputReference = exports.s3BucketAclAccessControlPolicyOwnerToTerraform = exports.S3BucketAclAccessControlPolicyGrantList = exports.S3BucketAclAccessControlPolicyGrantOutputReference = exports.s3BucketAclAccessControlPolicyGrantToTerraform = exports.S3BucketAclAccessControlPolicyGrantGranteeOutputReference = exports.s3BucketAclAccessControlPolicyGrantGranteeToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function s3BucketAclAccessControlPolicyGrantGranteeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        email_address: cdktf.stringToTerraform(struct.emailAddress),
        id: cdktf.stringToTerraform(struct.id),
        type: cdktf.stringToTerraform(struct.type),
        uri: cdktf.stringToTerraform(struct.uri),
    };
}
exports.s3BucketAclAccessControlPolicyGrantGranteeToTerraform = s3BucketAclAccessControlPolicyGrantGranteeToTerraform;
class S3BucketAclAccessControlPolicyGrantGranteeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._emailAddress !== undefined) {
            hasAnyValues = true;
            internalValueResult.emailAddress = this._emailAddress;
        }
        if (this._id !== undefined) {
            hasAnyValues = true;
            internalValueResult.id = this._id;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.uri = this._uri;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._emailAddress = undefined;
            this._id = undefined;
            this._type = undefined;
            this._uri = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._emailAddress = value.emailAddress;
            this._id = value.id;
            this._type = value.type;
            this._uri = value.uri;
        }
    }
    // display_name - computed: true, optional: false, required: false
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    get emailAddress() {
        return this.getStringAttribute('email_address');
    }
    set emailAddress(value) {
        this._emailAddress = value;
    }
    resetEmailAddress() {
        this._emailAddress = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get emailAddressInput() {
        return this._emailAddress;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get uri() {
        return this.getStringAttribute('uri');
    }
    set uri(value) {
        this._uri = value;
    }
    resetUri() {
        this._uri = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get uriInput() {
        return this._uri;
    }
}
exports.S3BucketAclAccessControlPolicyGrantGranteeOutputReference = S3BucketAclAccessControlPolicyGrantGranteeOutputReference;
_a = JSII_RTTI_SYMBOL_1;
S3BucketAclAccessControlPolicyGrantGranteeOutputReference[_a] = { fqn: "@cdktf/provider-aws.s3BucketAcl.S3BucketAclAccessControlPolicyGrantGranteeOutputReference", version: "10.0.12" };
function s3BucketAclAccessControlPolicyGrantToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        permission: cdktf.stringToTerraform(struct.permission),
        grantee: s3BucketAclAccessControlPolicyGrantGranteeToTerraform(struct.grantee),
    };
}
exports.s3BucketAclAccessControlPolicyGrantToTerraform = s3BucketAclAccessControlPolicyGrantToTerraform;
class S3BucketAclAccessControlPolicyGrantOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // grantee - computed: false, optional: true, required: false
        this._grantee = new S3BucketAclAccessControlPolicyGrantGranteeOutputReference(this, "grantee");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._permission !== undefined) {
            hasAnyValues = true;
            internalValueResult.permission = this._permission;
        }
        if (this._grantee?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.grantee = this._grantee?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._permission = undefined;
            this._grantee.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._permission = value.permission;
            this._grantee.internalValue = value.grantee;
        }
    }
    get permission() {
        return this.getStringAttribute('permission');
    }
    set permission(value) {
        this._permission = value;
    }
    // Temporarily expose input value. Use with caution.
    get permissionInput() {
        return this._permission;
    }
    get grantee() {
        return this._grantee;
    }
    putGrantee(value) {
        this._grantee.internalValue = value;
    }
    resetGrantee() {
        this._grantee.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get granteeInput() {
        return this._grantee.internalValue;
    }
}
exports.S3BucketAclAccessControlPolicyGrantOutputReference = S3BucketAclAccessControlPolicyGrantOutputReference;
_b = JSII_RTTI_SYMBOL_1;
S3BucketAclAccessControlPolicyGrantOutputReference[_b] = { fqn: "@cdktf/provider-aws.s3BucketAcl.S3BucketAclAccessControlPolicyGrantOutputReference", version: "10.0.12" };
class S3BucketAclAccessControlPolicyGrantList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new S3BucketAclAccessControlPolicyGrantOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.S3BucketAclAccessControlPolicyGrantList = S3BucketAclAccessControlPolicyGrantList;
_c = JSII_RTTI_SYMBOL_1;
S3BucketAclAccessControlPolicyGrantList[_c] = { fqn: "@cdktf/provider-aws.s3BucketAcl.S3BucketAclAccessControlPolicyGrantList", version: "10.0.12" };
function s3BucketAclAccessControlPolicyOwnerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        display_name: cdktf.stringToTerraform(struct.displayName),
        id: cdktf.stringToTerraform(struct.id),
    };
}
exports.s3BucketAclAccessControlPolicyOwnerToTerraform = s3BucketAclAccessControlPolicyOwnerToTerraform;
class S3BucketAclAccessControlPolicyOwnerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._displayName !== undefined) {
            hasAnyValues = true;
            internalValueResult.displayName = this._displayName;
        }
        if (this._id !== undefined) {
            hasAnyValues = true;
            internalValueResult.id = this._id;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._displayName = undefined;
            this._id = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._displayName = value.displayName;
            this._id = value.id;
        }
    }
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    set displayName(value) {
        this._displayName = value;
    }
    resetDisplayName() {
        this._displayName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get displayNameInput() {
        return this._displayName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
}
exports.S3BucketAclAccessControlPolicyOwnerOutputReference = S3BucketAclAccessControlPolicyOwnerOutputReference;
_d = JSII_RTTI_SYMBOL_1;
S3BucketAclAccessControlPolicyOwnerOutputReference[_d] = { fqn: "@cdktf/provider-aws.s3BucketAcl.S3BucketAclAccessControlPolicyOwnerOutputReference", version: "10.0.12" };
function s3BucketAclAccessControlPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        grant: cdktf.listMapper(s3BucketAclAccessControlPolicyGrantToTerraform, true)(struct.grant),
        owner: s3BucketAclAccessControlPolicyOwnerToTerraform(struct.owner),
    };
}
exports.s3BucketAclAccessControlPolicyToTerraform = s3BucketAclAccessControlPolicyToTerraform;
class S3BucketAclAccessControlPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // grant - computed: false, optional: true, required: false
        this._grant = new S3BucketAclAccessControlPolicyGrantList(this, "grant", true);
        // owner - computed: false, optional: false, required: true
        this._owner = new S3BucketAclAccessControlPolicyOwnerOutputReference(this, "owner");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._grant?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.grant = this._grant?.internalValue;
        }
        if (this._owner?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.owner = this._owner?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._grant.internalValue = undefined;
            this._owner.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._grant.internalValue = value.grant;
            this._owner.internalValue = value.owner;
        }
    }
    get grant() {
        return this._grant;
    }
    putGrant(value) {
        this._grant.internalValue = value;
    }
    resetGrant() {
        this._grant.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get grantInput() {
        return this._grant.internalValue;
    }
    get owner() {
        return this._owner;
    }
    putOwner(value) {
        this._owner.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get ownerInput() {
        return this._owner.internalValue;
    }
}
exports.S3BucketAclAccessControlPolicyOutputReference = S3BucketAclAccessControlPolicyOutputReference;
_e = JSII_RTTI_SYMBOL_1;
S3BucketAclAccessControlPolicyOutputReference[_e] = { fqn: "@cdktf/provider-aws.s3BucketAcl.S3BucketAclAccessControlPolicyOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/s3_bucket_acl aws_s3_bucket_acl}
*/
class S3BucketAcl extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/s3_bucket_acl aws_s3_bucket_acl} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options S3BucketAclConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_s3_bucket_acl',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // access_control_policy - computed: false, optional: true, required: false
        this._accessControlPolicy = new S3BucketAclAccessControlPolicyOutputReference(this, "access_control_policy");
        this._acl = config.acl;
        this._bucket = config.bucket;
        this._expectedBucketOwner = config.expectedBucketOwner;
        this._id = config.id;
        this._accessControlPolicy.internalValue = config.accessControlPolicy;
    }
    get acl() {
        return this.getStringAttribute('acl');
    }
    set acl(value) {
        this._acl = value;
    }
    resetAcl() {
        this._acl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get aclInput() {
        return this._acl;
    }
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    set bucket(value) {
        this._bucket = value;
    }
    // Temporarily expose input value. Use with caution.
    get bucketInput() {
        return this._bucket;
    }
    get expectedBucketOwner() {
        return this.getStringAttribute('expected_bucket_owner');
    }
    set expectedBucketOwner(value) {
        this._expectedBucketOwner = value;
    }
    resetExpectedBucketOwner() {
        this._expectedBucketOwner = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get expectedBucketOwnerInput() {
        return this._expectedBucketOwner;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get accessControlPolicy() {
        return this._accessControlPolicy;
    }
    putAccessControlPolicy(value) {
        this._accessControlPolicy.internalValue = value;
    }
    resetAccessControlPolicy() {
        this._accessControlPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessControlPolicyInput() {
        return this._accessControlPolicy.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            acl: cdktf.stringToTerraform(this._acl),
            bucket: cdktf.stringToTerraform(this._bucket),
            expected_bucket_owner: cdktf.stringToTerraform(this._expectedBucketOwner),
            id: cdktf.stringToTerraform(this._id),
            access_control_policy: s3BucketAclAccessControlPolicyToTerraform(this._accessControlPolicy.internalValue),
        };
    }
}
exports.S3BucketAcl = S3BucketAcl;
_f = JSII_RTTI_SYMBOL_1;
S3BucketAcl[_f] = { fqn: "@cdktf/provider-aws.s3BucketAcl.S3BucketAcl", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
S3BucketAcl.tfResourceType = "aws_s3_bucket_acl";
