// @cdktf/provider-aws/lib/batch-scheduling-policy/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BatchSchedulingPolicy = exports.BatchSchedulingPolicyFairSharePolicyOutputReference = exports.batchSchedulingPolicyFairSharePolicyToTerraform = exports.BatchSchedulingPolicyFairSharePolicyShareDistributionList = exports.BatchSchedulingPolicyFairSharePolicyShareDistributionOutputReference = exports.batchSchedulingPolicyFairSharePolicyShareDistributionToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function batchSchedulingPolicyFairSharePolicyShareDistributionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        share_identifier: cdktf.stringToTerraform(struct.shareIdentifier),
        weight_factor: cdktf.numberToTerraform(struct.weightFactor),
    };
}
exports.batchSchedulingPolicyFairSharePolicyShareDistributionToTerraform = batchSchedulingPolicyFairSharePolicyShareDistributionToTerraform;
class BatchSchedulingPolicyFairSharePolicyShareDistributionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._shareIdentifier !== undefined) {
            hasAnyValues = true;
            internalValueResult.shareIdentifier = this._shareIdentifier;
        }
        if (this._weightFactor !== undefined) {
            hasAnyValues = true;
            internalValueResult.weightFactor = this._weightFactor;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._shareIdentifier = undefined;
            this._weightFactor = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._shareIdentifier = value.shareIdentifier;
            this._weightFactor = value.weightFactor;
        }
    }
    get shareIdentifier() {
        return this.getStringAttribute('share_identifier');
    }
    set shareIdentifier(value) {
        this._shareIdentifier = value;
    }
    // Temporarily expose input value. Use with caution.
    get shareIdentifierInput() {
        return this._shareIdentifier;
    }
    get weightFactor() {
        return this.getNumberAttribute('weight_factor');
    }
    set weightFactor(value) {
        this._weightFactor = value;
    }
    resetWeightFactor() {
        this._weightFactor = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get weightFactorInput() {
        return this._weightFactor;
    }
}
exports.BatchSchedulingPolicyFairSharePolicyShareDistributionOutputReference = BatchSchedulingPolicyFairSharePolicyShareDistributionOutputReference;
_a = JSII_RTTI_SYMBOL_1;
BatchSchedulingPolicyFairSharePolicyShareDistributionOutputReference[_a] = { fqn: "@cdktf/provider-aws.batchSchedulingPolicy.BatchSchedulingPolicyFairSharePolicyShareDistributionOutputReference", version: "10.0.12" };
class BatchSchedulingPolicyFairSharePolicyShareDistributionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new BatchSchedulingPolicyFairSharePolicyShareDistributionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.BatchSchedulingPolicyFairSharePolicyShareDistributionList = BatchSchedulingPolicyFairSharePolicyShareDistributionList;
_b = JSII_RTTI_SYMBOL_1;
BatchSchedulingPolicyFairSharePolicyShareDistributionList[_b] = { fqn: "@cdktf/provider-aws.batchSchedulingPolicy.BatchSchedulingPolicyFairSharePolicyShareDistributionList", version: "10.0.12" };
function batchSchedulingPolicyFairSharePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        compute_reservation: cdktf.numberToTerraform(struct.computeReservation),
        share_decay_seconds: cdktf.numberToTerraform(struct.shareDecaySeconds),
        share_distribution: cdktf.listMapper(batchSchedulingPolicyFairSharePolicyShareDistributionToTerraform, true)(struct.shareDistribution),
    };
}
exports.batchSchedulingPolicyFairSharePolicyToTerraform = batchSchedulingPolicyFairSharePolicyToTerraform;
class BatchSchedulingPolicyFairSharePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // share_distribution - computed: false, optional: true, required: false
        this._shareDistribution = new BatchSchedulingPolicyFairSharePolicyShareDistributionList(this, "share_distribution", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._computeReservation !== undefined) {
            hasAnyValues = true;
            internalValueResult.computeReservation = this._computeReservation;
        }
        if (this._shareDecaySeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.shareDecaySeconds = this._shareDecaySeconds;
        }
        if (this._shareDistribution?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.shareDistribution = this._shareDistribution?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._computeReservation = undefined;
            this._shareDecaySeconds = undefined;
            this._shareDistribution.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._computeReservation = value.computeReservation;
            this._shareDecaySeconds = value.shareDecaySeconds;
            this._shareDistribution.internalValue = value.shareDistribution;
        }
    }
    get computeReservation() {
        return this.getNumberAttribute('compute_reservation');
    }
    set computeReservation(value) {
        this._computeReservation = value;
    }
    resetComputeReservation() {
        this._computeReservation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get computeReservationInput() {
        return this._computeReservation;
    }
    get shareDecaySeconds() {
        return this.getNumberAttribute('share_decay_seconds');
    }
    set shareDecaySeconds(value) {
        this._shareDecaySeconds = value;
    }
    resetShareDecaySeconds() {
        this._shareDecaySeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get shareDecaySecondsInput() {
        return this._shareDecaySeconds;
    }
    get shareDistribution() {
        return this._shareDistribution;
    }
    putShareDistribution(value) {
        this._shareDistribution.internalValue = value;
    }
    resetShareDistribution() {
        this._shareDistribution.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get shareDistributionInput() {
        return this._shareDistribution.internalValue;
    }
}
exports.BatchSchedulingPolicyFairSharePolicyOutputReference = BatchSchedulingPolicyFairSharePolicyOutputReference;
_c = JSII_RTTI_SYMBOL_1;
BatchSchedulingPolicyFairSharePolicyOutputReference[_c] = { fqn: "@cdktf/provider-aws.batchSchedulingPolicy.BatchSchedulingPolicyFairSharePolicyOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/batch_scheduling_policy aws_batch_scheduling_policy}
*/
class BatchSchedulingPolicy extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/batch_scheduling_policy aws_batch_scheduling_policy} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options BatchSchedulingPolicyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_batch_scheduling_policy',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // fair_share_policy - computed: false, optional: true, required: false
        this._fairSharePolicy = new BatchSchedulingPolicyFairSharePolicyOutputReference(this, "fair_share_policy");
        this._id = config.id;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._fairSharePolicy.internalValue = config.fairSharePolicy;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get fairSharePolicy() {
        return this._fairSharePolicy;
    }
    putFairSharePolicy(value) {
        this._fairSharePolicy.internalValue = value;
    }
    resetFairSharePolicy() {
        this._fairSharePolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fairSharePolicyInput() {
        return this._fairSharePolicy.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            fair_share_policy: batchSchedulingPolicyFairSharePolicyToTerraform(this._fairSharePolicy.internalValue),
        };
    }
}
exports.BatchSchedulingPolicy = BatchSchedulingPolicy;
_d = JSII_RTTI_SYMBOL_1;
BatchSchedulingPolicy[_d] = { fqn: "@cdktf/provider-aws.batchSchedulingPolicy.BatchSchedulingPolicy", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
BatchSchedulingPolicy.tfResourceType = "aws_batch_scheduling_policy";
