// @cdktf/provider-aws/lib/opsworks-rds-db-instance/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpsworksRdsDbInstance = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/opsworks_rds_db_instance aws_opsworks_rds_db_instance}
*/
class OpsworksRdsDbInstance extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/opsworks_rds_db_instance aws_opsworks_rds_db_instance} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options OpsworksRdsDbInstanceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_opsworks_rds_db_instance',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._dbPassword = config.dbPassword;
        this._dbUser = config.dbUser;
        this._id = config.id;
        this._rdsDbInstanceArn = config.rdsDbInstanceArn;
        this._stackId = config.stackId;
    }
    get dbPassword() {
        return this.getStringAttribute('db_password');
    }
    set dbPassword(value) {
        this._dbPassword = value;
    }
    // Temporarily expose input value. Use with caution.
    get dbPasswordInput() {
        return this._dbPassword;
    }
    get dbUser() {
        return this.getStringAttribute('db_user');
    }
    set dbUser(value) {
        this._dbUser = value;
    }
    // Temporarily expose input value. Use with caution.
    get dbUserInput() {
        return this._dbUser;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get rdsDbInstanceArn() {
        return this.getStringAttribute('rds_db_instance_arn');
    }
    set rdsDbInstanceArn(value) {
        this._rdsDbInstanceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get rdsDbInstanceArnInput() {
        return this._rdsDbInstanceArn;
    }
    get stackId() {
        return this.getStringAttribute('stack_id');
    }
    set stackId(value) {
        this._stackId = value;
    }
    // Temporarily expose input value. Use with caution.
    get stackIdInput() {
        return this._stackId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            db_password: cdktf.stringToTerraform(this._dbPassword),
            db_user: cdktf.stringToTerraform(this._dbUser),
            id: cdktf.stringToTerraform(this._id),
            rds_db_instance_arn: cdktf.stringToTerraform(this._rdsDbInstanceArn),
            stack_id: cdktf.stringToTerraform(this._stackId),
        };
    }
}
exports.OpsworksRdsDbInstance = OpsworksRdsDbInstance;
_a = JSII_RTTI_SYMBOL_1;
OpsworksRdsDbInstance[_a] = { fqn: "@cdktf/provider-aws.opsworksRdsDbInstance.OpsworksRdsDbInstance", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
OpsworksRdsDbInstance.tfResourceType = "aws_opsworks_rds_db_instance";
