// @cdktf/provider-aws/lib/data-aws-workspaces-workspace/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsWorkspacesWorkspace = exports.DataAwsWorkspacesWorkspaceWorkspacePropertiesList = exports.DataAwsWorkspacesWorkspaceWorkspacePropertiesOutputReference = exports.dataAwsWorkspacesWorkspaceWorkspacePropertiesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsWorkspacesWorkspaceWorkspacePropertiesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsWorkspacesWorkspaceWorkspacePropertiesToTerraform = dataAwsWorkspacesWorkspaceWorkspacePropertiesToTerraform;
class DataAwsWorkspacesWorkspaceWorkspacePropertiesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // compute_type_name - computed: true, optional: false, required: false
    get computeTypeName() {
        return this.getStringAttribute('compute_type_name');
    }
    // root_volume_size_gib - computed: true, optional: false, required: false
    get rootVolumeSizeGib() {
        return this.getNumberAttribute('root_volume_size_gib');
    }
    // running_mode - computed: true, optional: false, required: false
    get runningMode() {
        return this.getStringAttribute('running_mode');
    }
    // running_mode_auto_stop_timeout_in_minutes - computed: true, optional: false, required: false
    get runningModeAutoStopTimeoutInMinutes() {
        return this.getNumberAttribute('running_mode_auto_stop_timeout_in_minutes');
    }
    // user_volume_size_gib - computed: true, optional: false, required: false
    get userVolumeSizeGib() {
        return this.getNumberAttribute('user_volume_size_gib');
    }
}
exports.DataAwsWorkspacesWorkspaceWorkspacePropertiesOutputReference = DataAwsWorkspacesWorkspaceWorkspacePropertiesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsWorkspacesWorkspaceWorkspacePropertiesOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsWorkspacesWorkspace.DataAwsWorkspacesWorkspaceWorkspacePropertiesOutputReference", version: "10.0.12" };
class DataAwsWorkspacesWorkspaceWorkspacePropertiesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsWorkspacesWorkspaceWorkspacePropertiesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsWorkspacesWorkspaceWorkspacePropertiesList = DataAwsWorkspacesWorkspaceWorkspacePropertiesList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsWorkspacesWorkspaceWorkspacePropertiesList[_b] = { fqn: "@cdktf/provider-aws.dataAwsWorkspacesWorkspace.DataAwsWorkspacesWorkspaceWorkspacePropertiesList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/workspaces_workspace aws_workspaces_workspace}
*/
class DataAwsWorkspacesWorkspace extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/workspaces_workspace aws_workspaces_workspace} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsWorkspacesWorkspaceConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_workspaces_workspace',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // workspace_properties - computed: true, optional: false, required: false
        this._workspaceProperties = new DataAwsWorkspacesWorkspaceWorkspacePropertiesList(this, "workspace_properties", false);
        this._directoryId = config.directoryId;
        this._id = config.id;
        this._tags = config.tags;
        this._userName = config.userName;
        this._workspaceId = config.workspaceId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // bundle_id - computed: true, optional: false, required: false
    get bundleId() {
        return this.getStringAttribute('bundle_id');
    }
    // computer_name - computed: true, optional: false, required: false
    get computerName() {
        return this.getStringAttribute('computer_name');
    }
    get directoryId() {
        return this.getStringAttribute('directory_id');
    }
    set directoryId(value) {
        this._directoryId = value;
    }
    resetDirectoryId() {
        this._directoryId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get directoryIdInput() {
        return this._directoryId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // ip_address - computed: true, optional: false, required: false
    get ipAddress() {
        return this.getStringAttribute('ip_address');
    }
    // root_volume_encryption_enabled - computed: true, optional: false, required: false
    get rootVolumeEncryptionEnabled() {
        return this.getBooleanAttribute('root_volume_encryption_enabled');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get userName() {
        return this.getStringAttribute('user_name');
    }
    set userName(value) {
        this._userName = value;
    }
    resetUserName() {
        this._userName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get userNameInput() {
        return this._userName;
    }
    // user_volume_encryption_enabled - computed: true, optional: false, required: false
    get userVolumeEncryptionEnabled() {
        return this.getBooleanAttribute('user_volume_encryption_enabled');
    }
    // volume_encryption_key - computed: true, optional: false, required: false
    get volumeEncryptionKey() {
        return this.getStringAttribute('volume_encryption_key');
    }
    get workspaceId() {
        return this.getStringAttribute('workspace_id');
    }
    set workspaceId(value) {
        this._workspaceId = value;
    }
    resetWorkspaceId() {
        this._workspaceId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get workspaceIdInput() {
        return this._workspaceId;
    }
    get workspaceProperties() {
        return this._workspaceProperties;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            directory_id: cdktf.stringToTerraform(this._directoryId),
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            user_name: cdktf.stringToTerraform(this._userName),
            workspace_id: cdktf.stringToTerraform(this._workspaceId),
        };
    }
}
exports.DataAwsWorkspacesWorkspace = DataAwsWorkspacesWorkspace;
_c = JSII_RTTI_SYMBOL_1;
DataAwsWorkspacesWorkspace[_c] = { fqn: "@cdktf/provider-aws.dataAwsWorkspacesWorkspace.DataAwsWorkspacesWorkspace", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsWorkspacesWorkspace.tfResourceType = "aws_workspaces_workspace";
