// @cdktf/provider-aws/lib/s3-bucket-replication-configuration/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t;
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3BucketReplicationConfigurationA = exports.S3BucketReplicationConfigurationRuleList = exports.S3BucketReplicationConfigurationRuleOutputReference = exports.s3BucketReplicationConfigurationRuleToTerraform = exports.S3BucketReplicationConfigurationRuleSourceSelectionCriteriaOutputReference = exports.s3BucketReplicationConfigurationRuleSourceSelectionCriteriaToTerraform = exports.S3BucketReplicationConfigurationRuleSourceSelectionCriteriaSseKmsEncryptedObjectsOutputReference = exports.s3BucketReplicationConfigurationRuleSourceSelectionCriteriaSseKmsEncryptedObjectsToTerraform = exports.S3BucketReplicationConfigurationRuleSourceSelectionCriteriaReplicaModificationsOutputReference = exports.s3BucketReplicationConfigurationRuleSourceSelectionCriteriaReplicaModificationsToTerraform = exports.S3BucketReplicationConfigurationRuleFilterOutputReference = exports.s3BucketReplicationConfigurationRuleFilterToTerraform = exports.S3BucketReplicationConfigurationRuleFilterTagOutputReference = exports.s3BucketReplicationConfigurationRuleFilterTagToTerraform = exports.S3BucketReplicationConfigurationRuleFilterAndOutputReference = exports.s3BucketReplicationConfigurationRuleFilterAndToTerraform = exports.S3BucketReplicationConfigurationRuleExistingObjectReplicationOutputReference = exports.s3BucketReplicationConfigurationRuleExistingObjectReplicationToTerraform = exports.S3BucketReplicationConfigurationRuleDestinationOutputReference = exports.s3BucketReplicationConfigurationRuleDestinationToTerraform = exports.S3BucketReplicationConfigurationRuleDestinationReplicationTimeOutputReference = exports.s3BucketReplicationConfigurationRuleDestinationReplicationTimeToTerraform = exports.S3BucketReplicationConfigurationRuleDestinationReplicationTimeTimeOutputReference = exports.s3BucketReplicationConfigurationRuleDestinationReplicationTimeTimeToTerraform = exports.S3BucketReplicationConfigurationRuleDestinationMetricsOutputReference = exports.s3BucketReplicationConfigurationRuleDestinationMetricsToTerraform = exports.S3BucketReplicationConfigurationRuleDestinationMetricsEventThresholdOutputReference = exports.s3BucketReplicationConfigurationRuleDestinationMetricsEventThresholdToTerraform = exports.S3BucketReplicationConfigurationRuleDestinationEncryptionConfigurationOutputReference = exports.s3BucketReplicationConfigurationRuleDestinationEncryptionConfigurationToTerraform = exports.S3BucketReplicationConfigurationRuleDestinationAccessControlTranslationOutputReference = exports.s3BucketReplicationConfigurationRuleDestinationAccessControlTranslationToTerraform = exports.S3BucketReplicationConfigurationRuleDeleteMarkerReplicationOutputReference = exports.s3BucketReplicationConfigurationRuleDeleteMarkerReplicationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function s3BucketReplicationConfigurationRuleDeleteMarkerReplicationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        status: cdktf.stringToTerraform(struct.status),
    };
}
exports.s3BucketReplicationConfigurationRuleDeleteMarkerReplicationToTerraform = s3BucketReplicationConfigurationRuleDeleteMarkerReplicationToTerraform;
class S3BucketReplicationConfigurationRuleDeleteMarkerReplicationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._status !== undefined) {
            hasAnyValues = true;
            internalValueResult.status = this._status;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._status = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._status = value.status;
        }
    }
    get status() {
        return this.getStringAttribute('status');
    }
    set status(value) {
        this._status = value;
    }
    // Temporarily expose input value. Use with caution.
    get statusInput() {
        return this._status;
    }
}
exports.S3BucketReplicationConfigurationRuleDeleteMarkerReplicationOutputReference = S3BucketReplicationConfigurationRuleDeleteMarkerReplicationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleDeleteMarkerReplicationOutputReference[_a] = { fqn: "@cdktf/provider-aws.s3BucketReplicationConfiguration.S3BucketReplicationConfigurationRuleDeleteMarkerReplicationOutputReference", version: "10.0.12" };
function s3BucketReplicationConfigurationRuleDestinationAccessControlTranslationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        owner: cdktf.stringToTerraform(struct.owner),
    };
}
exports.s3BucketReplicationConfigurationRuleDestinationAccessControlTranslationToTerraform = s3BucketReplicationConfigurationRuleDestinationAccessControlTranslationToTerraform;
class S3BucketReplicationConfigurationRuleDestinationAccessControlTranslationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._owner !== undefined) {
            hasAnyValues = true;
            internalValueResult.owner = this._owner;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._owner = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._owner = value.owner;
        }
    }
    get owner() {
        return this.getStringAttribute('owner');
    }
    set owner(value) {
        this._owner = value;
    }
    // Temporarily expose input value. Use with caution.
    get ownerInput() {
        return this._owner;
    }
}
exports.S3BucketReplicationConfigurationRuleDestinationAccessControlTranslationOutputReference = S3BucketReplicationConfigurationRuleDestinationAccessControlTranslationOutputReference;
_b = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleDestinationAccessControlTranslationOutputReference[_b] = { fqn: "@cdktf/provider-aws.s3BucketReplicationConfiguration.S3BucketReplicationConfigurationRuleDestinationAccessControlTranslationOutputReference", version: "10.0.12" };
function s3BucketReplicationConfigurationRuleDestinationEncryptionConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        replica_kms_key_id: cdktf.stringToTerraform(struct.replicaKmsKeyId),
    };
}
exports.s3BucketReplicationConfigurationRuleDestinationEncryptionConfigurationToTerraform = s3BucketReplicationConfigurationRuleDestinationEncryptionConfigurationToTerraform;
class S3BucketReplicationConfigurationRuleDestinationEncryptionConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._replicaKmsKeyId !== undefined) {
            hasAnyValues = true;
            internalValueResult.replicaKmsKeyId = this._replicaKmsKeyId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._replicaKmsKeyId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._replicaKmsKeyId = value.replicaKmsKeyId;
        }
    }
    get replicaKmsKeyId() {
        return this.getStringAttribute('replica_kms_key_id');
    }
    set replicaKmsKeyId(value) {
        this._replicaKmsKeyId = value;
    }
    // Temporarily expose input value. Use with caution.
    get replicaKmsKeyIdInput() {
        return this._replicaKmsKeyId;
    }
}
exports.S3BucketReplicationConfigurationRuleDestinationEncryptionConfigurationOutputReference = S3BucketReplicationConfigurationRuleDestinationEncryptionConfigurationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleDestinationEncryptionConfigurationOutputReference[_c] = { fqn: "@cdktf/provider-aws.s3BucketReplicationConfiguration.S3BucketReplicationConfigurationRuleDestinationEncryptionConfigurationOutputReference", version: "10.0.12" };
function s3BucketReplicationConfigurationRuleDestinationMetricsEventThresholdToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        minutes: cdktf.numberToTerraform(struct.minutes),
    };
}
exports.s3BucketReplicationConfigurationRuleDestinationMetricsEventThresholdToTerraform = s3BucketReplicationConfigurationRuleDestinationMetricsEventThresholdToTerraform;
class S3BucketReplicationConfigurationRuleDestinationMetricsEventThresholdOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._minutes !== undefined) {
            hasAnyValues = true;
            internalValueResult.minutes = this._minutes;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._minutes = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._minutes = value.minutes;
        }
    }
    get minutes() {
        return this.getNumberAttribute('minutes');
    }
    set minutes(value) {
        this._minutes = value;
    }
    // Temporarily expose input value. Use with caution.
    get minutesInput() {
        return this._minutes;
    }
}
exports.S3BucketReplicationConfigurationRuleDestinationMetricsEventThresholdOutputReference = S3BucketReplicationConfigurationRuleDestinationMetricsEventThresholdOutputReference;
_d = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleDestinationMetricsEventThresholdOutputReference[_d] = { fqn: "@cdktf/provider-aws.s3BucketReplicationConfiguration.S3BucketReplicationConfigurationRuleDestinationMetricsEventThresholdOutputReference", version: "10.0.12" };
function s3BucketReplicationConfigurationRuleDestinationMetricsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        status: cdktf.stringToTerraform(struct.status),
        event_threshold: s3BucketReplicationConfigurationRuleDestinationMetricsEventThresholdToTerraform(struct.eventThreshold),
    };
}
exports.s3BucketReplicationConfigurationRuleDestinationMetricsToTerraform = s3BucketReplicationConfigurationRuleDestinationMetricsToTerraform;
class S3BucketReplicationConfigurationRuleDestinationMetricsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // event_threshold - computed: false, optional: true, required: false
        this._eventThreshold = new S3BucketReplicationConfigurationRuleDestinationMetricsEventThresholdOutputReference(this, "event_threshold");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._status !== undefined) {
            hasAnyValues = true;
            internalValueResult.status = this._status;
        }
        if (this._eventThreshold?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.eventThreshold = this._eventThreshold?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._status = undefined;
            this._eventThreshold.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._status = value.status;
            this._eventThreshold.internalValue = value.eventThreshold;
        }
    }
    get status() {
        return this.getStringAttribute('status');
    }
    set status(value) {
        this._status = value;
    }
    // Temporarily expose input value. Use with caution.
    get statusInput() {
        return this._status;
    }
    get eventThreshold() {
        return this._eventThreshold;
    }
    putEventThreshold(value) {
        this._eventThreshold.internalValue = value;
    }
    resetEventThreshold() {
        this._eventThreshold.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get eventThresholdInput() {
        return this._eventThreshold.internalValue;
    }
}
exports.S3BucketReplicationConfigurationRuleDestinationMetricsOutputReference = S3BucketReplicationConfigurationRuleDestinationMetricsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleDestinationMetricsOutputReference[_e] = { fqn: "@cdktf/provider-aws.s3BucketReplicationConfiguration.S3BucketReplicationConfigurationRuleDestinationMetricsOutputReference", version: "10.0.12" };
function s3BucketReplicationConfigurationRuleDestinationReplicationTimeTimeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        minutes: cdktf.numberToTerraform(struct.minutes),
    };
}
exports.s3BucketReplicationConfigurationRuleDestinationReplicationTimeTimeToTerraform = s3BucketReplicationConfigurationRuleDestinationReplicationTimeTimeToTerraform;
class S3BucketReplicationConfigurationRuleDestinationReplicationTimeTimeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._minutes !== undefined) {
            hasAnyValues = true;
            internalValueResult.minutes = this._minutes;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._minutes = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._minutes = value.minutes;
        }
    }
    get minutes() {
        return this.getNumberAttribute('minutes');
    }
    set minutes(value) {
        this._minutes = value;
    }
    // Temporarily expose input value. Use with caution.
    get minutesInput() {
        return this._minutes;
    }
}
exports.S3BucketReplicationConfigurationRuleDestinationReplicationTimeTimeOutputReference = S3BucketReplicationConfigurationRuleDestinationReplicationTimeTimeOutputReference;
_f = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleDestinationReplicationTimeTimeOutputReference[_f] = { fqn: "@cdktf/provider-aws.s3BucketReplicationConfiguration.S3BucketReplicationConfigurationRuleDestinationReplicationTimeTimeOutputReference", version: "10.0.12" };
function s3BucketReplicationConfigurationRuleDestinationReplicationTimeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        status: cdktf.stringToTerraform(struct.status),
        time: s3BucketReplicationConfigurationRuleDestinationReplicationTimeTimeToTerraform(struct.time),
    };
}
exports.s3BucketReplicationConfigurationRuleDestinationReplicationTimeToTerraform = s3BucketReplicationConfigurationRuleDestinationReplicationTimeToTerraform;
class S3BucketReplicationConfigurationRuleDestinationReplicationTimeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // time - computed: false, optional: false, required: true
        this._time = new S3BucketReplicationConfigurationRuleDestinationReplicationTimeTimeOutputReference(this, "time");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._status !== undefined) {
            hasAnyValues = true;
            internalValueResult.status = this._status;
        }
        if (this._time?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.time = this._time?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._status = undefined;
            this._time.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._status = value.status;
            this._time.internalValue = value.time;
        }
    }
    get status() {
        return this.getStringAttribute('status');
    }
    set status(value) {
        this._status = value;
    }
    // Temporarily expose input value. Use with caution.
    get statusInput() {
        return this._status;
    }
    get time() {
        return this._time;
    }
    putTime(value) {
        this._time.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get timeInput() {
        return this._time.internalValue;
    }
}
exports.S3BucketReplicationConfigurationRuleDestinationReplicationTimeOutputReference = S3BucketReplicationConfigurationRuleDestinationReplicationTimeOutputReference;
_g = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleDestinationReplicationTimeOutputReference[_g] = { fqn: "@cdktf/provider-aws.s3BucketReplicationConfiguration.S3BucketReplicationConfigurationRuleDestinationReplicationTimeOutputReference", version: "10.0.12" };
function s3BucketReplicationConfigurationRuleDestinationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        account: cdktf.stringToTerraform(struct.account),
        bucket: cdktf.stringToTerraform(struct.bucket),
        storage_class: cdktf.stringToTerraform(struct.storageClass),
        access_control_translation: s3BucketReplicationConfigurationRuleDestinationAccessControlTranslationToTerraform(struct.accessControlTranslation),
        encryption_configuration: s3BucketReplicationConfigurationRuleDestinationEncryptionConfigurationToTerraform(struct.encryptionConfiguration),
        metrics: s3BucketReplicationConfigurationRuleDestinationMetricsToTerraform(struct.metrics),
        replication_time: s3BucketReplicationConfigurationRuleDestinationReplicationTimeToTerraform(struct.replicationTime),
    };
}
exports.s3BucketReplicationConfigurationRuleDestinationToTerraform = s3BucketReplicationConfigurationRuleDestinationToTerraform;
class S3BucketReplicationConfigurationRuleDestinationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // access_control_translation - computed: false, optional: true, required: false
        this._accessControlTranslation = new S3BucketReplicationConfigurationRuleDestinationAccessControlTranslationOutputReference(this, "access_control_translation");
        // encryption_configuration - computed: false, optional: true, required: false
        this._encryptionConfiguration = new S3BucketReplicationConfigurationRuleDestinationEncryptionConfigurationOutputReference(this, "encryption_configuration");
        // metrics - computed: false, optional: true, required: false
        this._metrics = new S3BucketReplicationConfigurationRuleDestinationMetricsOutputReference(this, "metrics");
        // replication_time - computed: false, optional: true, required: false
        this._replicationTime = new S3BucketReplicationConfigurationRuleDestinationReplicationTimeOutputReference(this, "replication_time");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._account !== undefined) {
            hasAnyValues = true;
            internalValueResult.account = this._account;
        }
        if (this._bucket !== undefined) {
            hasAnyValues = true;
            internalValueResult.bucket = this._bucket;
        }
        if (this._storageClass !== undefined) {
            hasAnyValues = true;
            internalValueResult.storageClass = this._storageClass;
        }
        if (this._accessControlTranslation?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.accessControlTranslation = this._accessControlTranslation?.internalValue;
        }
        if (this._encryptionConfiguration?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.encryptionConfiguration = this._encryptionConfiguration?.internalValue;
        }
        if (this._metrics?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.metrics = this._metrics?.internalValue;
        }
        if (this._replicationTime?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.replicationTime = this._replicationTime?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._account = undefined;
            this._bucket = undefined;
            this._storageClass = undefined;
            this._accessControlTranslation.internalValue = undefined;
            this._encryptionConfiguration.internalValue = undefined;
            this._metrics.internalValue = undefined;
            this._replicationTime.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._account = value.account;
            this._bucket = value.bucket;
            this._storageClass = value.storageClass;
            this._accessControlTranslation.internalValue = value.accessControlTranslation;
            this._encryptionConfiguration.internalValue = value.encryptionConfiguration;
            this._metrics.internalValue = value.metrics;
            this._replicationTime.internalValue = value.replicationTime;
        }
    }
    get account() {
        return this.getStringAttribute('account');
    }
    set account(value) {
        this._account = value;
    }
    resetAccount() {
        this._account = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accountInput() {
        return this._account;
    }
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    set bucket(value) {
        this._bucket = value;
    }
    // Temporarily expose input value. Use with caution.
    get bucketInput() {
        return this._bucket;
    }
    get storageClass() {
        return this.getStringAttribute('storage_class');
    }
    set storageClass(value) {
        this._storageClass = value;
    }
    resetStorageClass() {
        this._storageClass = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get storageClassInput() {
        return this._storageClass;
    }
    get accessControlTranslation() {
        return this._accessControlTranslation;
    }
    putAccessControlTranslation(value) {
        this._accessControlTranslation.internalValue = value;
    }
    resetAccessControlTranslation() {
        this._accessControlTranslation.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessControlTranslationInput() {
        return this._accessControlTranslation.internalValue;
    }
    get encryptionConfiguration() {
        return this._encryptionConfiguration;
    }
    putEncryptionConfiguration(value) {
        this._encryptionConfiguration.internalValue = value;
    }
    resetEncryptionConfiguration() {
        this._encryptionConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encryptionConfigurationInput() {
        return this._encryptionConfiguration.internalValue;
    }
    get metrics() {
        return this._metrics;
    }
    putMetrics(value) {
        this._metrics.internalValue = value;
    }
    resetMetrics() {
        this._metrics.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get metricsInput() {
        return this._metrics.internalValue;
    }
    get replicationTime() {
        return this._replicationTime;
    }
    putReplicationTime(value) {
        this._replicationTime.internalValue = value;
    }
    resetReplicationTime() {
        this._replicationTime.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get replicationTimeInput() {
        return this._replicationTime.internalValue;
    }
}
exports.S3BucketReplicationConfigurationRuleDestinationOutputReference = S3BucketReplicationConfigurationRuleDestinationOutputReference;
_h = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleDestinationOutputReference[_h] = { fqn: "@cdktf/provider-aws.s3BucketReplicationConfiguration.S3BucketReplicationConfigurationRuleDestinationOutputReference", version: "10.0.12" };
function s3BucketReplicationConfigurationRuleExistingObjectReplicationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        status: cdktf.stringToTerraform(struct.status),
    };
}
exports.s3BucketReplicationConfigurationRuleExistingObjectReplicationToTerraform = s3BucketReplicationConfigurationRuleExistingObjectReplicationToTerraform;
class S3BucketReplicationConfigurationRuleExistingObjectReplicationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._status !== undefined) {
            hasAnyValues = true;
            internalValueResult.status = this._status;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._status = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._status = value.status;
        }
    }
    get status() {
        return this.getStringAttribute('status');
    }
    set status(value) {
        this._status = value;
    }
    // Temporarily expose input value. Use with caution.
    get statusInput() {
        return this._status;
    }
}
exports.S3BucketReplicationConfigurationRuleExistingObjectReplicationOutputReference = S3BucketReplicationConfigurationRuleExistingObjectReplicationOutputReference;
_j = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleExistingObjectReplicationOutputReference[_j] = { fqn: "@cdktf/provider-aws.s3BucketReplicationConfiguration.S3BucketReplicationConfigurationRuleExistingObjectReplicationOutputReference", version: "10.0.12" };
function s3BucketReplicationConfigurationRuleFilterAndToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        prefix: cdktf.stringToTerraform(struct.prefix),
        tags: cdktf.hashMapper(cdktf.stringToTerraform)(struct.tags),
    };
}
exports.s3BucketReplicationConfigurationRuleFilterAndToTerraform = s3BucketReplicationConfigurationRuleFilterAndToTerraform;
class S3BucketReplicationConfigurationRuleFilterAndOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._prefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        if (this._tags !== undefined) {
            hasAnyValues = true;
            internalValueResult.tags = this._tags;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._prefix = undefined;
            this._tags = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._prefix = value.prefix;
            this._tags = value.tags;
        }
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
}
exports.S3BucketReplicationConfigurationRuleFilterAndOutputReference = S3BucketReplicationConfigurationRuleFilterAndOutputReference;
_k = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleFilterAndOutputReference[_k] = { fqn: "@cdktf/provider-aws.s3BucketReplicationConfiguration.S3BucketReplicationConfigurationRuleFilterAndOutputReference", version: "10.0.12" };
function s3BucketReplicationConfigurationRuleFilterTagToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.s3BucketReplicationConfigurationRuleFilterTagToTerraform = s3BucketReplicationConfigurationRuleFilterTagToTerraform;
class S3BucketReplicationConfigurationRuleFilterTagOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._key = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._key = value.key;
            this._value = value.value;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.S3BucketReplicationConfigurationRuleFilterTagOutputReference = S3BucketReplicationConfigurationRuleFilterTagOutputReference;
_l = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleFilterTagOutputReference[_l] = { fqn: "@cdktf/provider-aws.s3BucketReplicationConfiguration.S3BucketReplicationConfigurationRuleFilterTagOutputReference", version: "10.0.12" };
function s3BucketReplicationConfigurationRuleFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        prefix: cdktf.stringToTerraform(struct.prefix),
        and: s3BucketReplicationConfigurationRuleFilterAndToTerraform(struct.and),
        tag: s3BucketReplicationConfigurationRuleFilterTagToTerraform(struct.tag),
    };
}
exports.s3BucketReplicationConfigurationRuleFilterToTerraform = s3BucketReplicationConfigurationRuleFilterToTerraform;
class S3BucketReplicationConfigurationRuleFilterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // and - computed: false, optional: true, required: false
        this._and = new S3BucketReplicationConfigurationRuleFilterAndOutputReference(this, "and");
        // tag - computed: false, optional: true, required: false
        this._tag = new S3BucketReplicationConfigurationRuleFilterTagOutputReference(this, "tag");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._prefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        if (this._and?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.and = this._and?.internalValue;
        }
        if (this._tag?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.tag = this._tag?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._prefix = undefined;
            this._and.internalValue = undefined;
            this._tag.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._prefix = value.prefix;
            this._and.internalValue = value.and;
            this._tag.internalValue = value.tag;
        }
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
    get and() {
        return this._and;
    }
    putAnd(value) {
        this._and.internalValue = value;
    }
    resetAnd() {
        this._and.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get andInput() {
        return this._and.internalValue;
    }
    get tag() {
        return this._tag;
    }
    putTag(value) {
        this._tag.internalValue = value;
    }
    resetTag() {
        this._tag.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagInput() {
        return this._tag.internalValue;
    }
}
exports.S3BucketReplicationConfigurationRuleFilterOutputReference = S3BucketReplicationConfigurationRuleFilterOutputReference;
_m = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleFilterOutputReference[_m] = { fqn: "@cdktf/provider-aws.s3BucketReplicationConfiguration.S3BucketReplicationConfigurationRuleFilterOutputReference", version: "10.0.12" };
function s3BucketReplicationConfigurationRuleSourceSelectionCriteriaReplicaModificationsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        status: cdktf.stringToTerraform(struct.status),
    };
}
exports.s3BucketReplicationConfigurationRuleSourceSelectionCriteriaReplicaModificationsToTerraform = s3BucketReplicationConfigurationRuleSourceSelectionCriteriaReplicaModificationsToTerraform;
class S3BucketReplicationConfigurationRuleSourceSelectionCriteriaReplicaModificationsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._status !== undefined) {
            hasAnyValues = true;
            internalValueResult.status = this._status;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._status = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._status = value.status;
        }
    }
    get status() {
        return this.getStringAttribute('status');
    }
    set status(value) {
        this._status = value;
    }
    // Temporarily expose input value. Use with caution.
    get statusInput() {
        return this._status;
    }
}
exports.S3BucketReplicationConfigurationRuleSourceSelectionCriteriaReplicaModificationsOutputReference = S3BucketReplicationConfigurationRuleSourceSelectionCriteriaReplicaModificationsOutputReference;
_o = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleSourceSelectionCriteriaReplicaModificationsOutputReference[_o] = { fqn: "@cdktf/provider-aws.s3BucketReplicationConfiguration.S3BucketReplicationConfigurationRuleSourceSelectionCriteriaReplicaModificationsOutputReference", version: "10.0.12" };
function s3BucketReplicationConfigurationRuleSourceSelectionCriteriaSseKmsEncryptedObjectsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        status: cdktf.stringToTerraform(struct.status),
    };
}
exports.s3BucketReplicationConfigurationRuleSourceSelectionCriteriaSseKmsEncryptedObjectsToTerraform = s3BucketReplicationConfigurationRuleSourceSelectionCriteriaSseKmsEncryptedObjectsToTerraform;
class S3BucketReplicationConfigurationRuleSourceSelectionCriteriaSseKmsEncryptedObjectsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._status !== undefined) {
            hasAnyValues = true;
            internalValueResult.status = this._status;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._status = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._status = value.status;
        }
    }
    get status() {
        return this.getStringAttribute('status');
    }
    set status(value) {
        this._status = value;
    }
    // Temporarily expose input value. Use with caution.
    get statusInput() {
        return this._status;
    }
}
exports.S3BucketReplicationConfigurationRuleSourceSelectionCriteriaSseKmsEncryptedObjectsOutputReference = S3BucketReplicationConfigurationRuleSourceSelectionCriteriaSseKmsEncryptedObjectsOutputReference;
_p = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleSourceSelectionCriteriaSseKmsEncryptedObjectsOutputReference[_p] = { fqn: "@cdktf/provider-aws.s3BucketReplicationConfiguration.S3BucketReplicationConfigurationRuleSourceSelectionCriteriaSseKmsEncryptedObjectsOutputReference", version: "10.0.12" };
function s3BucketReplicationConfigurationRuleSourceSelectionCriteriaToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        replica_modifications: s3BucketReplicationConfigurationRuleSourceSelectionCriteriaReplicaModificationsToTerraform(struct.replicaModifications),
        sse_kms_encrypted_objects: s3BucketReplicationConfigurationRuleSourceSelectionCriteriaSseKmsEncryptedObjectsToTerraform(struct.sseKmsEncryptedObjects),
    };
}
exports.s3BucketReplicationConfigurationRuleSourceSelectionCriteriaToTerraform = s3BucketReplicationConfigurationRuleSourceSelectionCriteriaToTerraform;
class S3BucketReplicationConfigurationRuleSourceSelectionCriteriaOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // replica_modifications - computed: false, optional: true, required: false
        this._replicaModifications = new S3BucketReplicationConfigurationRuleSourceSelectionCriteriaReplicaModificationsOutputReference(this, "replica_modifications");
        // sse_kms_encrypted_objects - computed: false, optional: true, required: false
        this._sseKmsEncryptedObjects = new S3BucketReplicationConfigurationRuleSourceSelectionCriteriaSseKmsEncryptedObjectsOutputReference(this, "sse_kms_encrypted_objects");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._replicaModifications?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.replicaModifications = this._replicaModifications?.internalValue;
        }
        if (this._sseKmsEncryptedObjects?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.sseKmsEncryptedObjects = this._sseKmsEncryptedObjects?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._replicaModifications.internalValue = undefined;
            this._sseKmsEncryptedObjects.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._replicaModifications.internalValue = value.replicaModifications;
            this._sseKmsEncryptedObjects.internalValue = value.sseKmsEncryptedObjects;
        }
    }
    get replicaModifications() {
        return this._replicaModifications;
    }
    putReplicaModifications(value) {
        this._replicaModifications.internalValue = value;
    }
    resetReplicaModifications() {
        this._replicaModifications.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get replicaModificationsInput() {
        return this._replicaModifications.internalValue;
    }
    get sseKmsEncryptedObjects() {
        return this._sseKmsEncryptedObjects;
    }
    putSseKmsEncryptedObjects(value) {
        this._sseKmsEncryptedObjects.internalValue = value;
    }
    resetSseKmsEncryptedObjects() {
        this._sseKmsEncryptedObjects.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sseKmsEncryptedObjectsInput() {
        return this._sseKmsEncryptedObjects.internalValue;
    }
}
exports.S3BucketReplicationConfigurationRuleSourceSelectionCriteriaOutputReference = S3BucketReplicationConfigurationRuleSourceSelectionCriteriaOutputReference;
_q = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleSourceSelectionCriteriaOutputReference[_q] = { fqn: "@cdktf/provider-aws.s3BucketReplicationConfiguration.S3BucketReplicationConfigurationRuleSourceSelectionCriteriaOutputReference", version: "10.0.12" };
function s3BucketReplicationConfigurationRuleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        id: cdktf.stringToTerraform(struct.id),
        prefix: cdktf.stringToTerraform(struct.prefix),
        priority: cdktf.numberToTerraform(struct.priority),
        status: cdktf.stringToTerraform(struct.status),
        delete_marker_replication: s3BucketReplicationConfigurationRuleDeleteMarkerReplicationToTerraform(struct.deleteMarkerReplication),
        destination: s3BucketReplicationConfigurationRuleDestinationToTerraform(struct.destination),
        existing_object_replication: s3BucketReplicationConfigurationRuleExistingObjectReplicationToTerraform(struct.existingObjectReplication),
        filter: s3BucketReplicationConfigurationRuleFilterToTerraform(struct.filter),
        source_selection_criteria: s3BucketReplicationConfigurationRuleSourceSelectionCriteriaToTerraform(struct.sourceSelectionCriteria),
    };
}
exports.s3BucketReplicationConfigurationRuleToTerraform = s3BucketReplicationConfigurationRuleToTerraform;
class S3BucketReplicationConfigurationRuleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // delete_marker_replication - computed: false, optional: true, required: false
        this._deleteMarkerReplication = new S3BucketReplicationConfigurationRuleDeleteMarkerReplicationOutputReference(this, "delete_marker_replication");
        // destination - computed: false, optional: false, required: true
        this._destination = new S3BucketReplicationConfigurationRuleDestinationOutputReference(this, "destination");
        // existing_object_replication - computed: false, optional: true, required: false
        this._existingObjectReplication = new S3BucketReplicationConfigurationRuleExistingObjectReplicationOutputReference(this, "existing_object_replication");
        // filter - computed: false, optional: true, required: false
        this._filter = new S3BucketReplicationConfigurationRuleFilterOutputReference(this, "filter");
        // source_selection_criteria - computed: false, optional: true, required: false
        this._sourceSelectionCriteria = new S3BucketReplicationConfigurationRuleSourceSelectionCriteriaOutputReference(this, "source_selection_criteria");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._id !== undefined) {
            hasAnyValues = true;
            internalValueResult.id = this._id;
        }
        if (this._prefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        if (this._priority !== undefined) {
            hasAnyValues = true;
            internalValueResult.priority = this._priority;
        }
        if (this._status !== undefined) {
            hasAnyValues = true;
            internalValueResult.status = this._status;
        }
        if (this._deleteMarkerReplication?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.deleteMarkerReplication = this._deleteMarkerReplication?.internalValue;
        }
        if (this._destination?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.destination = this._destination?.internalValue;
        }
        if (this._existingObjectReplication?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.existingObjectReplication = this._existingObjectReplication?.internalValue;
        }
        if (this._filter?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.filter = this._filter?.internalValue;
        }
        if (this._sourceSelectionCriteria?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceSelectionCriteria = this._sourceSelectionCriteria?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._id = undefined;
            this._prefix = undefined;
            this._priority = undefined;
            this._status = undefined;
            this._deleteMarkerReplication.internalValue = undefined;
            this._destination.internalValue = undefined;
            this._existingObjectReplication.internalValue = undefined;
            this._filter.internalValue = undefined;
            this._sourceSelectionCriteria.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._id = value.id;
            this._prefix = value.prefix;
            this._priority = value.priority;
            this._status = value.status;
            this._deleteMarkerReplication.internalValue = value.deleteMarkerReplication;
            this._destination.internalValue = value.destination;
            this._existingObjectReplication.internalValue = value.existingObjectReplication;
            this._filter.internalValue = value.filter;
            this._sourceSelectionCriteria.internalValue = value.sourceSelectionCriteria;
        }
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
    get priority() {
        return this.getNumberAttribute('priority');
    }
    set priority(value) {
        this._priority = value;
    }
    resetPriority() {
        this._priority = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get priorityInput() {
        return this._priority;
    }
    get status() {
        return this.getStringAttribute('status');
    }
    set status(value) {
        this._status = value;
    }
    // Temporarily expose input value. Use with caution.
    get statusInput() {
        return this._status;
    }
    get deleteMarkerReplication() {
        return this._deleteMarkerReplication;
    }
    putDeleteMarkerReplication(value) {
        this._deleteMarkerReplication.internalValue = value;
    }
    resetDeleteMarkerReplication() {
        this._deleteMarkerReplication.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteMarkerReplicationInput() {
        return this._deleteMarkerReplication.internalValue;
    }
    get destination() {
        return this._destination;
    }
    putDestination(value) {
        this._destination.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get destinationInput() {
        return this._destination.internalValue;
    }
    get existingObjectReplication() {
        return this._existingObjectReplication;
    }
    putExistingObjectReplication(value) {
        this._existingObjectReplication.internalValue = value;
    }
    resetExistingObjectReplication() {
        this._existingObjectReplication.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get existingObjectReplicationInput() {
        return this._existingObjectReplication.internalValue;
    }
    get filter() {
        return this._filter;
    }
    putFilter(value) {
        this._filter.internalValue = value;
    }
    resetFilter() {
        this._filter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter.internalValue;
    }
    get sourceSelectionCriteria() {
        return this._sourceSelectionCriteria;
    }
    putSourceSelectionCriteria(value) {
        this._sourceSelectionCriteria.internalValue = value;
    }
    resetSourceSelectionCriteria() {
        this._sourceSelectionCriteria.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceSelectionCriteriaInput() {
        return this._sourceSelectionCriteria.internalValue;
    }
}
exports.S3BucketReplicationConfigurationRuleOutputReference = S3BucketReplicationConfigurationRuleOutputReference;
_r = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleOutputReference[_r] = { fqn: "@cdktf/provider-aws.s3BucketReplicationConfiguration.S3BucketReplicationConfigurationRuleOutputReference", version: "10.0.12" };
class S3BucketReplicationConfigurationRuleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new S3BucketReplicationConfigurationRuleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.S3BucketReplicationConfigurationRuleList = S3BucketReplicationConfigurationRuleList;
_s = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationRuleList[_s] = { fqn: "@cdktf/provider-aws.s3BucketReplicationConfiguration.S3BucketReplicationConfigurationRuleList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/s3_bucket_replication_configuration aws_s3_bucket_replication_configuration}
*/
class S3BucketReplicationConfigurationA extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/s3_bucket_replication_configuration aws_s3_bucket_replication_configuration} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options S3BucketReplicationConfigurationAConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_s3_bucket_replication_configuration',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // rule - computed: false, optional: false, required: true
        this._rule = new S3BucketReplicationConfigurationRuleList(this, "rule", false);
        this._bucket = config.bucket;
        this._id = config.id;
        this._role = config.role;
        this._token = config.token;
        this._rule.internalValue = config.rule;
    }
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    set bucket(value) {
        this._bucket = value;
    }
    // Temporarily expose input value. Use with caution.
    get bucketInput() {
        return this._bucket;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get role() {
        return this.getStringAttribute('role');
    }
    set role(value) {
        this._role = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleInput() {
        return this._role;
    }
    get token() {
        return this.getStringAttribute('token');
    }
    set token(value) {
        this._token = value;
    }
    resetToken() {
        this._token = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tokenInput() {
        return this._token;
    }
    get rule() {
        return this._rule;
    }
    putRule(value) {
        this._rule.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get ruleInput() {
        return this._rule.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            bucket: cdktf.stringToTerraform(this._bucket),
            id: cdktf.stringToTerraform(this._id),
            role: cdktf.stringToTerraform(this._role),
            token: cdktf.stringToTerraform(this._token),
            rule: cdktf.listMapper(s3BucketReplicationConfigurationRuleToTerraform, true)(this._rule.internalValue),
        };
    }
}
exports.S3BucketReplicationConfigurationA = S3BucketReplicationConfigurationA;
_t = JSII_RTTI_SYMBOL_1;
S3BucketReplicationConfigurationA[_t] = { fqn: "@cdktf/provider-aws.s3BucketReplicationConfiguration.S3BucketReplicationConfigurationA", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
S3BucketReplicationConfigurationA.tfResourceType = "aws_s3_bucket_replication_configuration";
