// @cdktf/provider-aws/lib/data-aws-apigatewayv2-export/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsApigatewayv2Export = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/apigatewayv2_export aws_apigatewayv2_export}
*/
class DataAwsApigatewayv2Export extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/apigatewayv2_export aws_apigatewayv2_export} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsApigatewayv2ExportConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_apigatewayv2_export',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._apiId = config.apiId;
        this._exportVersion = config.exportVersion;
        this._id = config.id;
        this._includeExtensions = config.includeExtensions;
        this._outputType = config.outputType;
        this._specification = config.specification;
        this._stageName = config.stageName;
    }
    get apiId() {
        return this.getStringAttribute('api_id');
    }
    set apiId(value) {
        this._apiId = value;
    }
    // Temporarily expose input value. Use with caution.
    get apiIdInput() {
        return this._apiId;
    }
    // body - computed: true, optional: false, required: false
    get body() {
        return this.getStringAttribute('body');
    }
    get exportVersion() {
        return this.getStringAttribute('export_version');
    }
    set exportVersion(value) {
        this._exportVersion = value;
    }
    resetExportVersion() {
        this._exportVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exportVersionInput() {
        return this._exportVersion;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get includeExtensions() {
        return this.getBooleanAttribute('include_extensions');
    }
    set includeExtensions(value) {
        this._includeExtensions = value;
    }
    resetIncludeExtensions() {
        this._includeExtensions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includeExtensionsInput() {
        return this._includeExtensions;
    }
    get outputType() {
        return this.getStringAttribute('output_type');
    }
    set outputType(value) {
        this._outputType = value;
    }
    // Temporarily expose input value. Use with caution.
    get outputTypeInput() {
        return this._outputType;
    }
    get specification() {
        return this.getStringAttribute('specification');
    }
    set specification(value) {
        this._specification = value;
    }
    // Temporarily expose input value. Use with caution.
    get specificationInput() {
        return this._specification;
    }
    get stageName() {
        return this.getStringAttribute('stage_name');
    }
    set stageName(value) {
        this._stageName = value;
    }
    resetStageName() {
        this._stageName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get stageNameInput() {
        return this._stageName;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            api_id: cdktf.stringToTerraform(this._apiId),
            export_version: cdktf.stringToTerraform(this._exportVersion),
            id: cdktf.stringToTerraform(this._id),
            include_extensions: cdktf.booleanToTerraform(this._includeExtensions),
            output_type: cdktf.stringToTerraform(this._outputType),
            specification: cdktf.stringToTerraform(this._specification),
            stage_name: cdktf.stringToTerraform(this._stageName),
        };
    }
}
exports.DataAwsApigatewayv2Export = DataAwsApigatewayv2Export;
_a = JSII_RTTI_SYMBOL_1;
DataAwsApigatewayv2Export[_a] = { fqn: "@cdktf/provider-aws.dataAwsApigatewayv2Export.DataAwsApigatewayv2Export", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsApigatewayv2Export.tfResourceType = "aws_apigatewayv2_export";
