// @cdktf/provider-aws/lib/apigatewayv2-route-response/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Apigatewayv2RouteResponse = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/apigatewayv2_route_response aws_apigatewayv2_route_response}
*/
class Apigatewayv2RouteResponse extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/apigatewayv2_route_response aws_apigatewayv2_route_response} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options Apigatewayv2RouteResponseConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_apigatewayv2_route_response',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._apiId = config.apiId;
        this._id = config.id;
        this._modelSelectionExpression = config.modelSelectionExpression;
        this._responseModels = config.responseModels;
        this._routeId = config.routeId;
        this._routeResponseKey = config.routeResponseKey;
    }
    get apiId() {
        return this.getStringAttribute('api_id');
    }
    set apiId(value) {
        this._apiId = value;
    }
    // Temporarily expose input value. Use with caution.
    get apiIdInput() {
        return this._apiId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get modelSelectionExpression() {
        return this.getStringAttribute('model_selection_expression');
    }
    set modelSelectionExpression(value) {
        this._modelSelectionExpression = value;
    }
    resetModelSelectionExpression() {
        this._modelSelectionExpression = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modelSelectionExpressionInput() {
        return this._modelSelectionExpression;
    }
    get responseModels() {
        return this.getStringMapAttribute('response_models');
    }
    set responseModels(value) {
        this._responseModels = value;
    }
    resetResponseModels() {
        this._responseModels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get responseModelsInput() {
        return this._responseModels;
    }
    get routeId() {
        return this.getStringAttribute('route_id');
    }
    set routeId(value) {
        this._routeId = value;
    }
    // Temporarily expose input value. Use with caution.
    get routeIdInput() {
        return this._routeId;
    }
    get routeResponseKey() {
        return this.getStringAttribute('route_response_key');
    }
    set routeResponseKey(value) {
        this._routeResponseKey = value;
    }
    // Temporarily expose input value. Use with caution.
    get routeResponseKeyInput() {
        return this._routeResponseKey;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            api_id: cdktf.stringToTerraform(this._apiId),
            id: cdktf.stringToTerraform(this._id),
            model_selection_expression: cdktf.stringToTerraform(this._modelSelectionExpression),
            response_models: cdktf.hashMapper(cdktf.stringToTerraform)(this._responseModels),
            route_id: cdktf.stringToTerraform(this._routeId),
            route_response_key: cdktf.stringToTerraform(this._routeResponseKey),
        };
    }
}
exports.Apigatewayv2RouteResponse = Apigatewayv2RouteResponse;
_a = JSII_RTTI_SYMBOL_1;
Apigatewayv2RouteResponse[_a] = { fqn: "@cdktf/provider-aws.apigatewayv2RouteResponse.Apigatewayv2RouteResponse", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
Apigatewayv2RouteResponse.tfResourceType = "aws_apigatewayv2_route_response";
