// @cdktf/provider-aws/lib/data-aws-connect-user-hierarchy-structure/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsConnectUserHierarchyStructure = exports.DataAwsConnectUserHierarchyStructureHierarchyStructureList = exports.DataAwsConnectUserHierarchyStructureHierarchyStructureOutputReference = exports.dataAwsConnectUserHierarchyStructureHierarchyStructureToTerraform = exports.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelTwoList = exports.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelTwoOutputReference = exports.dataAwsConnectUserHierarchyStructureHierarchyStructureLevelTwoToTerraform = exports.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelThreeList = exports.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelThreeOutputReference = exports.dataAwsConnectUserHierarchyStructureHierarchyStructureLevelThreeToTerraform = exports.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelOneList = exports.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelOneOutputReference = exports.dataAwsConnectUserHierarchyStructureHierarchyStructureLevelOneToTerraform = exports.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFourList = exports.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFourOutputReference = exports.dataAwsConnectUserHierarchyStructureHierarchyStructureLevelFourToTerraform = exports.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFiveList = exports.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFiveOutputReference = exports.dataAwsConnectUserHierarchyStructureHierarchyStructureLevelFiveToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsConnectUserHierarchyStructureHierarchyStructureLevelFiveToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsConnectUserHierarchyStructureHierarchyStructureLevelFiveToTerraform = dataAwsConnectUserHierarchyStructureHierarchyStructureLevelFiveToTerraform;
class DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFiveOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFiveOutputReference = DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFiveOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFiveOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyStructure.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFiveOutputReference", version: "10.0.12" };
class DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFiveList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFiveOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFiveList = DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFiveList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFiveList[_b] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyStructure.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFiveList", version: "10.0.12" };
function dataAwsConnectUserHierarchyStructureHierarchyStructureLevelFourToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsConnectUserHierarchyStructureHierarchyStructureLevelFourToTerraform = dataAwsConnectUserHierarchyStructureHierarchyStructureLevelFourToTerraform;
class DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFourOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFourOutputReference = DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFourOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFourOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyStructure.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFourOutputReference", version: "10.0.12" };
class DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFourList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFourOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFourList = DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFourList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFourList[_d] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyStructure.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFourList", version: "10.0.12" };
function dataAwsConnectUserHierarchyStructureHierarchyStructureLevelOneToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsConnectUserHierarchyStructureHierarchyStructureLevelOneToTerraform = dataAwsConnectUserHierarchyStructureHierarchyStructureLevelOneToTerraform;
class DataAwsConnectUserHierarchyStructureHierarchyStructureLevelOneOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelOneOutputReference = DataAwsConnectUserHierarchyStructureHierarchyStructureLevelOneOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyStructureHierarchyStructureLevelOneOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyStructure.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelOneOutputReference", version: "10.0.12" };
class DataAwsConnectUserHierarchyStructureHierarchyStructureLevelOneList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsConnectUserHierarchyStructureHierarchyStructureLevelOneOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelOneList = DataAwsConnectUserHierarchyStructureHierarchyStructureLevelOneList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyStructureHierarchyStructureLevelOneList[_f] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyStructure.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelOneList", version: "10.0.12" };
function dataAwsConnectUserHierarchyStructureHierarchyStructureLevelThreeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsConnectUserHierarchyStructureHierarchyStructureLevelThreeToTerraform = dataAwsConnectUserHierarchyStructureHierarchyStructureLevelThreeToTerraform;
class DataAwsConnectUserHierarchyStructureHierarchyStructureLevelThreeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelThreeOutputReference = DataAwsConnectUserHierarchyStructureHierarchyStructureLevelThreeOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyStructureHierarchyStructureLevelThreeOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyStructure.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelThreeOutputReference", version: "10.0.12" };
class DataAwsConnectUserHierarchyStructureHierarchyStructureLevelThreeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsConnectUserHierarchyStructureHierarchyStructureLevelThreeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelThreeList = DataAwsConnectUserHierarchyStructureHierarchyStructureLevelThreeList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyStructureHierarchyStructureLevelThreeList[_h] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyStructure.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelThreeList", version: "10.0.12" };
function dataAwsConnectUserHierarchyStructureHierarchyStructureLevelTwoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsConnectUserHierarchyStructureHierarchyStructureLevelTwoToTerraform = dataAwsConnectUserHierarchyStructureHierarchyStructureLevelTwoToTerraform;
class DataAwsConnectUserHierarchyStructureHierarchyStructureLevelTwoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelTwoOutputReference = DataAwsConnectUserHierarchyStructureHierarchyStructureLevelTwoOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyStructureHierarchyStructureLevelTwoOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyStructure.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelTwoOutputReference", version: "10.0.12" };
class DataAwsConnectUserHierarchyStructureHierarchyStructureLevelTwoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsConnectUserHierarchyStructureHierarchyStructureLevelTwoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelTwoList = DataAwsConnectUserHierarchyStructureHierarchyStructureLevelTwoList;
_k = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyStructureHierarchyStructureLevelTwoList[_k] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyStructure.DataAwsConnectUserHierarchyStructureHierarchyStructureLevelTwoList", version: "10.0.12" };
function dataAwsConnectUserHierarchyStructureHierarchyStructureToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsConnectUserHierarchyStructureHierarchyStructureToTerraform = dataAwsConnectUserHierarchyStructureHierarchyStructureToTerraform;
class DataAwsConnectUserHierarchyStructureHierarchyStructureOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // level_five - computed: true, optional: false, required: false
        this._levelFive = new DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFiveList(this, "level_five", false);
        // level_four - computed: true, optional: false, required: false
        this._levelFour = new DataAwsConnectUserHierarchyStructureHierarchyStructureLevelFourList(this, "level_four", false);
        // level_one - computed: true, optional: false, required: false
        this._levelOne = new DataAwsConnectUserHierarchyStructureHierarchyStructureLevelOneList(this, "level_one", false);
        // level_three - computed: true, optional: false, required: false
        this._levelThree = new DataAwsConnectUserHierarchyStructureHierarchyStructureLevelThreeList(this, "level_three", false);
        // level_two - computed: true, optional: false, required: false
        this._levelTwo = new DataAwsConnectUserHierarchyStructureHierarchyStructureLevelTwoList(this, "level_two", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get levelFive() {
        return this._levelFive;
    }
    get levelFour() {
        return this._levelFour;
    }
    get levelOne() {
        return this._levelOne;
    }
    get levelThree() {
        return this._levelThree;
    }
    get levelTwo() {
        return this._levelTwo;
    }
}
exports.DataAwsConnectUserHierarchyStructureHierarchyStructureOutputReference = DataAwsConnectUserHierarchyStructureHierarchyStructureOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyStructureHierarchyStructureOutputReference[_l] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyStructure.DataAwsConnectUserHierarchyStructureHierarchyStructureOutputReference", version: "10.0.12" };
class DataAwsConnectUserHierarchyStructureHierarchyStructureList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsConnectUserHierarchyStructureHierarchyStructureOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsConnectUserHierarchyStructureHierarchyStructureList = DataAwsConnectUserHierarchyStructureHierarchyStructureList;
_m = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyStructureHierarchyStructureList[_m] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyStructure.DataAwsConnectUserHierarchyStructureHierarchyStructureList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/connect_user_hierarchy_structure aws_connect_user_hierarchy_structure}
*/
class DataAwsConnectUserHierarchyStructure extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/connect_user_hierarchy_structure aws_connect_user_hierarchy_structure} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsConnectUserHierarchyStructureConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_connect_user_hierarchy_structure',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // hierarchy_structure - computed: true, optional: false, required: false
        this._hierarchyStructure = new DataAwsConnectUserHierarchyStructureHierarchyStructureList(this, "hierarchy_structure", false);
        this._id = config.id;
        this._instanceId = config.instanceId;
    }
    get hierarchyStructure() {
        return this._hierarchyStructure;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instanceId() {
        return this.getStringAttribute('instance_id');
    }
    set instanceId(value) {
        this._instanceId = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceIdInput() {
        return this._instanceId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            instance_id: cdktf.stringToTerraform(this._instanceId),
        };
    }
}
exports.DataAwsConnectUserHierarchyStructure = DataAwsConnectUserHierarchyStructure;
_o = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyStructure[_o] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyStructure.DataAwsConnectUserHierarchyStructure", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsConnectUserHierarchyStructure.tfResourceType = "aws_connect_user_hierarchy_structure";
