// @cdktf/provider-aws/lib/ec2-transit-gateway/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ec2TransitGateway = exports.Ec2TransitGatewayTimeoutsOutputReference = exports.ec2TransitGatewayTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function ec2TransitGatewayTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.ec2TransitGatewayTimeoutsToTerraform = ec2TransitGatewayTimeoutsToTerraform;
class Ec2TransitGatewayTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.Ec2TransitGatewayTimeoutsOutputReference = Ec2TransitGatewayTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
Ec2TransitGatewayTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.ec2TransitGateway.Ec2TransitGatewayTimeoutsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/ec2_transit_gateway aws_ec2_transit_gateway}
*/
class Ec2TransitGateway extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ec2_transit_gateway aws_ec2_transit_gateway} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options Ec2TransitGatewayConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_ec2_transit_gateway',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new Ec2TransitGatewayTimeoutsOutputReference(this, "timeouts");
        this._amazonSideAsn = config.amazonSideAsn;
        this._autoAcceptSharedAttachments = config.autoAcceptSharedAttachments;
        this._defaultRouteTableAssociation = config.defaultRouteTableAssociation;
        this._defaultRouteTablePropagation = config.defaultRouteTablePropagation;
        this._description = config.description;
        this._dnsSupport = config.dnsSupport;
        this._id = config.id;
        this._multicastSupport = config.multicastSupport;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._transitGatewayCidrBlocks = config.transitGatewayCidrBlocks;
        this._vpnEcmpSupport = config.vpnEcmpSupport;
        this._timeouts.internalValue = config.timeouts;
    }
    get amazonSideAsn() {
        return this.getNumberAttribute('amazon_side_asn');
    }
    set amazonSideAsn(value) {
        this._amazonSideAsn = value;
    }
    resetAmazonSideAsn() {
        this._amazonSideAsn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get amazonSideAsnInput() {
        return this._amazonSideAsn;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // association_default_route_table_id - computed: true, optional: false, required: false
    get associationDefaultRouteTableId() {
        return this.getStringAttribute('association_default_route_table_id');
    }
    get autoAcceptSharedAttachments() {
        return this.getStringAttribute('auto_accept_shared_attachments');
    }
    set autoAcceptSharedAttachments(value) {
        this._autoAcceptSharedAttachments = value;
    }
    resetAutoAcceptSharedAttachments() {
        this._autoAcceptSharedAttachments = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoAcceptSharedAttachmentsInput() {
        return this._autoAcceptSharedAttachments;
    }
    get defaultRouteTableAssociation() {
        return this.getStringAttribute('default_route_table_association');
    }
    set defaultRouteTableAssociation(value) {
        this._defaultRouteTableAssociation = value;
    }
    resetDefaultRouteTableAssociation() {
        this._defaultRouteTableAssociation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultRouteTableAssociationInput() {
        return this._defaultRouteTableAssociation;
    }
    get defaultRouteTablePropagation() {
        return this.getStringAttribute('default_route_table_propagation');
    }
    set defaultRouteTablePropagation(value) {
        this._defaultRouteTablePropagation = value;
    }
    resetDefaultRouteTablePropagation() {
        this._defaultRouteTablePropagation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultRouteTablePropagationInput() {
        return this._defaultRouteTablePropagation;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get dnsSupport() {
        return this.getStringAttribute('dns_support');
    }
    set dnsSupport(value) {
        this._dnsSupport = value;
    }
    resetDnsSupport() {
        this._dnsSupport = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dnsSupportInput() {
        return this._dnsSupport;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get multicastSupport() {
        return this.getStringAttribute('multicast_support');
    }
    set multicastSupport(value) {
        this._multicastSupport = value;
    }
    resetMulticastSupport() {
        this._multicastSupport = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get multicastSupportInput() {
        return this._multicastSupport;
    }
    // owner_id - computed: true, optional: false, required: false
    get ownerId() {
        return this.getStringAttribute('owner_id');
    }
    // propagation_default_route_table_id - computed: true, optional: false, required: false
    get propagationDefaultRouteTableId() {
        return this.getStringAttribute('propagation_default_route_table_id');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get transitGatewayCidrBlocks() {
        return cdktf.Fn.tolist(this.getListAttribute('transit_gateway_cidr_blocks'));
    }
    set transitGatewayCidrBlocks(value) {
        this._transitGatewayCidrBlocks = value;
    }
    resetTransitGatewayCidrBlocks() {
        this._transitGatewayCidrBlocks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get transitGatewayCidrBlocksInput() {
        return this._transitGatewayCidrBlocks;
    }
    get vpnEcmpSupport() {
        return this.getStringAttribute('vpn_ecmp_support');
    }
    set vpnEcmpSupport(value) {
        this._vpnEcmpSupport = value;
    }
    resetVpnEcmpSupport() {
        this._vpnEcmpSupport = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpnEcmpSupportInput() {
        return this._vpnEcmpSupport;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            amazon_side_asn: cdktf.numberToTerraform(this._amazonSideAsn),
            auto_accept_shared_attachments: cdktf.stringToTerraform(this._autoAcceptSharedAttachments),
            default_route_table_association: cdktf.stringToTerraform(this._defaultRouteTableAssociation),
            default_route_table_propagation: cdktf.stringToTerraform(this._defaultRouteTablePropagation),
            description: cdktf.stringToTerraform(this._description),
            dns_support: cdktf.stringToTerraform(this._dnsSupport),
            id: cdktf.stringToTerraform(this._id),
            multicast_support: cdktf.stringToTerraform(this._multicastSupport),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            transit_gateway_cidr_blocks: cdktf.listMapper(cdktf.stringToTerraform, false)(this._transitGatewayCidrBlocks),
            vpn_ecmp_support: cdktf.stringToTerraform(this._vpnEcmpSupport),
            timeouts: ec2TransitGatewayTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.Ec2TransitGateway = Ec2TransitGateway;
_b = JSII_RTTI_SYMBOL_1;
Ec2TransitGateway[_b] = { fqn: "@cdktf/provider-aws.ec2TransitGateway.Ec2TransitGateway", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
Ec2TransitGateway.tfResourceType = "aws_ec2_transit_gateway";
