// @cdktf/provider-aws/lib/data-aws-opensearch-domain/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsOpensearchDomain = exports.DataAwsOpensearchDomainVpcOptionsList = exports.DataAwsOpensearchDomainVpcOptionsOutputReference = exports.dataAwsOpensearchDomainVpcOptionsToTerraform = exports.DataAwsOpensearchDomainSnapshotOptionsList = exports.DataAwsOpensearchDomainSnapshotOptionsOutputReference = exports.dataAwsOpensearchDomainSnapshotOptionsToTerraform = exports.DataAwsOpensearchDomainNodeToNodeEncryptionList = exports.DataAwsOpensearchDomainNodeToNodeEncryptionOutputReference = exports.dataAwsOpensearchDomainNodeToNodeEncryptionToTerraform = exports.DataAwsOpensearchDomainLogPublishingOptionsList = exports.DataAwsOpensearchDomainLogPublishingOptionsOutputReference = exports.dataAwsOpensearchDomainLogPublishingOptionsToTerraform = exports.DataAwsOpensearchDomainEncryptionAtRestList = exports.DataAwsOpensearchDomainEncryptionAtRestOutputReference = exports.dataAwsOpensearchDomainEncryptionAtRestToTerraform = exports.DataAwsOpensearchDomainEbsOptionsList = exports.DataAwsOpensearchDomainEbsOptionsOutputReference = exports.dataAwsOpensearchDomainEbsOptionsToTerraform = exports.DataAwsOpensearchDomainCognitoOptionsList = exports.DataAwsOpensearchDomainCognitoOptionsOutputReference = exports.dataAwsOpensearchDomainCognitoOptionsToTerraform = exports.DataAwsOpensearchDomainClusterConfigList = exports.DataAwsOpensearchDomainClusterConfigOutputReference = exports.dataAwsOpensearchDomainClusterConfigToTerraform = exports.DataAwsOpensearchDomainClusterConfigZoneAwarenessConfigList = exports.DataAwsOpensearchDomainClusterConfigZoneAwarenessConfigOutputReference = exports.dataAwsOpensearchDomainClusterConfigZoneAwarenessConfigToTerraform = exports.DataAwsOpensearchDomainClusterConfigColdStorageOptionsList = exports.DataAwsOpensearchDomainClusterConfigColdStorageOptionsOutputReference = exports.dataAwsOpensearchDomainClusterConfigColdStorageOptionsToTerraform = exports.DataAwsOpensearchDomainAutoTuneOptionsList = exports.DataAwsOpensearchDomainAutoTuneOptionsOutputReference = exports.dataAwsOpensearchDomainAutoTuneOptionsToTerraform = exports.DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleList = exports.DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleOutputReference = exports.dataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleToTerraform = exports.DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationList = exports.DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationOutputReference = exports.dataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationToTerraform = exports.DataAwsOpensearchDomainAdvancedSecurityOptionsList = exports.DataAwsOpensearchDomainAdvancedSecurityOptionsOutputReference = exports.dataAwsOpensearchDomainAdvancedSecurityOptionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsOpensearchDomainAdvancedSecurityOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsOpensearchDomainAdvancedSecurityOptionsToTerraform = dataAwsOpensearchDomainAdvancedSecurityOptionsToTerraform;
class DataAwsOpensearchDomainAdvancedSecurityOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    // internal_user_database_enabled - computed: true, optional: false, required: false
    get internalUserDatabaseEnabled() {
        return this.getBooleanAttribute('internal_user_database_enabled');
    }
}
exports.DataAwsOpensearchDomainAdvancedSecurityOptionsOutputReference = DataAwsOpensearchDomainAdvancedSecurityOptionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainAdvancedSecurityOptionsOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainAdvancedSecurityOptionsOutputReference", version: "10.0.12" };
class DataAwsOpensearchDomainAdvancedSecurityOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsOpensearchDomainAdvancedSecurityOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsOpensearchDomainAdvancedSecurityOptionsList = DataAwsOpensearchDomainAdvancedSecurityOptionsList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainAdvancedSecurityOptionsList[_b] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainAdvancedSecurityOptionsList", version: "10.0.12" };
function dataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationToTerraform = dataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationToTerraform;
class DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // unit - computed: true, optional: false, required: false
    get unit() {
        return this.getStringAttribute('unit');
    }
    // value - computed: true, optional: false, required: false
    get value() {
        return this.getNumberAttribute('value');
    }
}
exports.DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationOutputReference = DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationOutputReference", version: "10.0.12" };
class DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationList = DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationList[_d] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationList", version: "10.0.12" };
function dataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleToTerraform = dataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleToTerraform;
class DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // duration - computed: true, optional: false, required: false
        this._duration = new DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleDurationList(this, "duration", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // cron_expression_for_recurrence - computed: true, optional: false, required: false
    get cronExpressionForRecurrence() {
        return this.getStringAttribute('cron_expression_for_recurrence');
    }
    get duration() {
        return this._duration;
    }
    // start_at - computed: true, optional: false, required: false
    get startAt() {
        return this.getStringAttribute('start_at');
    }
}
exports.DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleOutputReference = DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleOutputReference", version: "10.0.12" };
class DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleList = DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleList[_f] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleList", version: "10.0.12" };
function dataAwsOpensearchDomainAutoTuneOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsOpensearchDomainAutoTuneOptionsToTerraform = dataAwsOpensearchDomainAutoTuneOptionsToTerraform;
class DataAwsOpensearchDomainAutoTuneOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // maintenance_schedule - computed: true, optional: false, required: false
        this._maintenanceSchedule = new DataAwsOpensearchDomainAutoTuneOptionsMaintenanceScheduleList(this, "maintenance_schedule", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // desired_state - computed: true, optional: false, required: false
    get desiredState() {
        return this.getStringAttribute('desired_state');
    }
    get maintenanceSchedule() {
        return this._maintenanceSchedule;
    }
    // rollback_on_disable - computed: true, optional: false, required: false
    get rollbackOnDisable() {
        return this.getStringAttribute('rollback_on_disable');
    }
}
exports.DataAwsOpensearchDomainAutoTuneOptionsOutputReference = DataAwsOpensearchDomainAutoTuneOptionsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainAutoTuneOptionsOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainAutoTuneOptionsOutputReference", version: "10.0.12" };
class DataAwsOpensearchDomainAutoTuneOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsOpensearchDomainAutoTuneOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsOpensearchDomainAutoTuneOptionsList = DataAwsOpensearchDomainAutoTuneOptionsList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainAutoTuneOptionsList[_h] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainAutoTuneOptionsList", version: "10.0.12" };
function dataAwsOpensearchDomainClusterConfigColdStorageOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsOpensearchDomainClusterConfigColdStorageOptionsToTerraform = dataAwsOpensearchDomainClusterConfigColdStorageOptionsToTerraform;
class DataAwsOpensearchDomainClusterConfigColdStorageOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
}
exports.DataAwsOpensearchDomainClusterConfigColdStorageOptionsOutputReference = DataAwsOpensearchDomainClusterConfigColdStorageOptionsOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainClusterConfigColdStorageOptionsOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainClusterConfigColdStorageOptionsOutputReference", version: "10.0.12" };
class DataAwsOpensearchDomainClusterConfigColdStorageOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsOpensearchDomainClusterConfigColdStorageOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsOpensearchDomainClusterConfigColdStorageOptionsList = DataAwsOpensearchDomainClusterConfigColdStorageOptionsList;
_k = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainClusterConfigColdStorageOptionsList[_k] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainClusterConfigColdStorageOptionsList", version: "10.0.12" };
function dataAwsOpensearchDomainClusterConfigZoneAwarenessConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsOpensearchDomainClusterConfigZoneAwarenessConfigToTerraform = dataAwsOpensearchDomainClusterConfigZoneAwarenessConfigToTerraform;
class DataAwsOpensearchDomainClusterConfigZoneAwarenessConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // availability_zone_count - computed: true, optional: false, required: false
    get availabilityZoneCount() {
        return this.getNumberAttribute('availability_zone_count');
    }
}
exports.DataAwsOpensearchDomainClusterConfigZoneAwarenessConfigOutputReference = DataAwsOpensearchDomainClusterConfigZoneAwarenessConfigOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainClusterConfigZoneAwarenessConfigOutputReference[_l] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainClusterConfigZoneAwarenessConfigOutputReference", version: "10.0.12" };
class DataAwsOpensearchDomainClusterConfigZoneAwarenessConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsOpensearchDomainClusterConfigZoneAwarenessConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsOpensearchDomainClusterConfigZoneAwarenessConfigList = DataAwsOpensearchDomainClusterConfigZoneAwarenessConfigList;
_m = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainClusterConfigZoneAwarenessConfigList[_m] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainClusterConfigZoneAwarenessConfigList", version: "10.0.12" };
function dataAwsOpensearchDomainClusterConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsOpensearchDomainClusterConfigToTerraform = dataAwsOpensearchDomainClusterConfigToTerraform;
class DataAwsOpensearchDomainClusterConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // cold_storage_options - computed: true, optional: false, required: false
        this._coldStorageOptions = new DataAwsOpensearchDomainClusterConfigColdStorageOptionsList(this, "cold_storage_options", false);
        // zone_awareness_config - computed: true, optional: false, required: false
        this._zoneAwarenessConfig = new DataAwsOpensearchDomainClusterConfigZoneAwarenessConfigList(this, "zone_awareness_config", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get coldStorageOptions() {
        return this._coldStorageOptions;
    }
    // dedicated_master_count - computed: true, optional: false, required: false
    get dedicatedMasterCount() {
        return this.getNumberAttribute('dedicated_master_count');
    }
    // dedicated_master_enabled - computed: true, optional: false, required: false
    get dedicatedMasterEnabled() {
        return this.getBooleanAttribute('dedicated_master_enabled');
    }
    // dedicated_master_type - computed: true, optional: false, required: false
    get dedicatedMasterType() {
        return this.getStringAttribute('dedicated_master_type');
    }
    // instance_count - computed: true, optional: false, required: false
    get instanceCount() {
        return this.getNumberAttribute('instance_count');
    }
    // instance_type - computed: true, optional: false, required: false
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    // warm_count - computed: true, optional: false, required: false
    get warmCount() {
        return this.getNumberAttribute('warm_count');
    }
    // warm_enabled - computed: true, optional: false, required: false
    get warmEnabled() {
        return this.getBooleanAttribute('warm_enabled');
    }
    // warm_type - computed: true, optional: false, required: false
    get warmType() {
        return this.getStringAttribute('warm_type');
    }
    get zoneAwarenessConfig() {
        return this._zoneAwarenessConfig;
    }
    // zone_awareness_enabled - computed: true, optional: false, required: false
    get zoneAwarenessEnabled() {
        return this.getBooleanAttribute('zone_awareness_enabled');
    }
}
exports.DataAwsOpensearchDomainClusterConfigOutputReference = DataAwsOpensearchDomainClusterConfigOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainClusterConfigOutputReference[_o] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainClusterConfigOutputReference", version: "10.0.12" };
class DataAwsOpensearchDomainClusterConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsOpensearchDomainClusterConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsOpensearchDomainClusterConfigList = DataAwsOpensearchDomainClusterConfigList;
_p = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainClusterConfigList[_p] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainClusterConfigList", version: "10.0.12" };
function dataAwsOpensearchDomainCognitoOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsOpensearchDomainCognitoOptionsToTerraform = dataAwsOpensearchDomainCognitoOptionsToTerraform;
class DataAwsOpensearchDomainCognitoOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    // identity_pool_id - computed: true, optional: false, required: false
    get identityPoolId() {
        return this.getStringAttribute('identity_pool_id');
    }
    // role_arn - computed: true, optional: false, required: false
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    // user_pool_id - computed: true, optional: false, required: false
    get userPoolId() {
        return this.getStringAttribute('user_pool_id');
    }
}
exports.DataAwsOpensearchDomainCognitoOptionsOutputReference = DataAwsOpensearchDomainCognitoOptionsOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainCognitoOptionsOutputReference[_q] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainCognitoOptionsOutputReference", version: "10.0.12" };
class DataAwsOpensearchDomainCognitoOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsOpensearchDomainCognitoOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsOpensearchDomainCognitoOptionsList = DataAwsOpensearchDomainCognitoOptionsList;
_r = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainCognitoOptionsList[_r] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainCognitoOptionsList", version: "10.0.12" };
function dataAwsOpensearchDomainEbsOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsOpensearchDomainEbsOptionsToTerraform = dataAwsOpensearchDomainEbsOptionsToTerraform;
class DataAwsOpensearchDomainEbsOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ebs_enabled - computed: true, optional: false, required: false
    get ebsEnabled() {
        return this.getBooleanAttribute('ebs_enabled');
    }
    // iops - computed: true, optional: false, required: false
    get iops() {
        return this.getNumberAttribute('iops');
    }
    // throughput - computed: true, optional: false, required: false
    get throughput() {
        return this.getNumberAttribute('throughput');
    }
    // volume_size - computed: true, optional: false, required: false
    get volumeSize() {
        return this.getNumberAttribute('volume_size');
    }
    // volume_type - computed: true, optional: false, required: false
    get volumeType() {
        return this.getStringAttribute('volume_type');
    }
}
exports.DataAwsOpensearchDomainEbsOptionsOutputReference = DataAwsOpensearchDomainEbsOptionsOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainEbsOptionsOutputReference[_s] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainEbsOptionsOutputReference", version: "10.0.12" };
class DataAwsOpensearchDomainEbsOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsOpensearchDomainEbsOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsOpensearchDomainEbsOptionsList = DataAwsOpensearchDomainEbsOptionsList;
_t = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainEbsOptionsList[_t] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainEbsOptionsList", version: "10.0.12" };
function dataAwsOpensearchDomainEncryptionAtRestToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsOpensearchDomainEncryptionAtRestToTerraform = dataAwsOpensearchDomainEncryptionAtRestToTerraform;
class DataAwsOpensearchDomainEncryptionAtRestOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    // kms_key_id - computed: true, optional: false, required: false
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
}
exports.DataAwsOpensearchDomainEncryptionAtRestOutputReference = DataAwsOpensearchDomainEncryptionAtRestOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainEncryptionAtRestOutputReference[_u] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainEncryptionAtRestOutputReference", version: "10.0.12" };
class DataAwsOpensearchDomainEncryptionAtRestList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsOpensearchDomainEncryptionAtRestOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsOpensearchDomainEncryptionAtRestList = DataAwsOpensearchDomainEncryptionAtRestList;
_v = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainEncryptionAtRestList[_v] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainEncryptionAtRestList", version: "10.0.12" };
function dataAwsOpensearchDomainLogPublishingOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsOpensearchDomainLogPublishingOptionsToTerraform = dataAwsOpensearchDomainLogPublishingOptionsToTerraform;
class DataAwsOpensearchDomainLogPublishingOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // cloudwatch_log_group_arn - computed: true, optional: false, required: false
    get cloudwatchLogGroupArn() {
        return this.getStringAttribute('cloudwatch_log_group_arn');
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    // log_type - computed: true, optional: false, required: false
    get logType() {
        return this.getStringAttribute('log_type');
    }
}
exports.DataAwsOpensearchDomainLogPublishingOptionsOutputReference = DataAwsOpensearchDomainLogPublishingOptionsOutputReference;
_w = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainLogPublishingOptionsOutputReference[_w] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainLogPublishingOptionsOutputReference", version: "10.0.12" };
class DataAwsOpensearchDomainLogPublishingOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsOpensearchDomainLogPublishingOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsOpensearchDomainLogPublishingOptionsList = DataAwsOpensearchDomainLogPublishingOptionsList;
_x = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainLogPublishingOptionsList[_x] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainLogPublishingOptionsList", version: "10.0.12" };
function dataAwsOpensearchDomainNodeToNodeEncryptionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsOpensearchDomainNodeToNodeEncryptionToTerraform = dataAwsOpensearchDomainNodeToNodeEncryptionToTerraform;
class DataAwsOpensearchDomainNodeToNodeEncryptionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
}
exports.DataAwsOpensearchDomainNodeToNodeEncryptionOutputReference = DataAwsOpensearchDomainNodeToNodeEncryptionOutputReference;
_y = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainNodeToNodeEncryptionOutputReference[_y] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainNodeToNodeEncryptionOutputReference", version: "10.0.12" };
class DataAwsOpensearchDomainNodeToNodeEncryptionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsOpensearchDomainNodeToNodeEncryptionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsOpensearchDomainNodeToNodeEncryptionList = DataAwsOpensearchDomainNodeToNodeEncryptionList;
_z = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainNodeToNodeEncryptionList[_z] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainNodeToNodeEncryptionList", version: "10.0.12" };
function dataAwsOpensearchDomainSnapshotOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsOpensearchDomainSnapshotOptionsToTerraform = dataAwsOpensearchDomainSnapshotOptionsToTerraform;
class DataAwsOpensearchDomainSnapshotOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // automated_snapshot_start_hour - computed: true, optional: false, required: false
    get automatedSnapshotStartHour() {
        return this.getNumberAttribute('automated_snapshot_start_hour');
    }
}
exports.DataAwsOpensearchDomainSnapshotOptionsOutputReference = DataAwsOpensearchDomainSnapshotOptionsOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainSnapshotOptionsOutputReference[_0] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainSnapshotOptionsOutputReference", version: "10.0.12" };
class DataAwsOpensearchDomainSnapshotOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsOpensearchDomainSnapshotOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsOpensearchDomainSnapshotOptionsList = DataAwsOpensearchDomainSnapshotOptionsList;
_1 = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainSnapshotOptionsList[_1] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainSnapshotOptionsList", version: "10.0.12" };
function dataAwsOpensearchDomainVpcOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsOpensearchDomainVpcOptionsToTerraform = dataAwsOpensearchDomainVpcOptionsToTerraform;
class DataAwsOpensearchDomainVpcOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // availability_zones - computed: true, optional: false, required: false
    get availabilityZones() {
        return cdktf.Fn.tolist(this.getListAttribute('availability_zones'));
    }
    // security_group_ids - computed: true, optional: false, required: false
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    // subnet_ids - computed: true, optional: false, required: false
    get subnetIds() {
        return cdktf.Fn.tolist(this.getListAttribute('subnet_ids'));
    }
    // vpc_id - computed: true, optional: false, required: false
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
}
exports.DataAwsOpensearchDomainVpcOptionsOutputReference = DataAwsOpensearchDomainVpcOptionsOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainVpcOptionsOutputReference[_2] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainVpcOptionsOutputReference", version: "10.0.12" };
class DataAwsOpensearchDomainVpcOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsOpensearchDomainVpcOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsOpensearchDomainVpcOptionsList = DataAwsOpensearchDomainVpcOptionsList;
_3 = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomainVpcOptionsList[_3] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomainVpcOptionsList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/opensearch_domain aws_opensearch_domain}
*/
class DataAwsOpensearchDomain extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/opensearch_domain aws_opensearch_domain} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsOpensearchDomainConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_opensearch_domain',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // advanced_options - computed: true, optional: false, required: false
        this._advancedOptions = new cdktf.StringMap(this, "advanced_options");
        // advanced_security_options - computed: true, optional: false, required: false
        this._advancedSecurityOptions = new DataAwsOpensearchDomainAdvancedSecurityOptionsList(this, "advanced_security_options", false);
        // auto_tune_options - computed: true, optional: false, required: false
        this._autoTuneOptions = new DataAwsOpensearchDomainAutoTuneOptionsList(this, "auto_tune_options", false);
        // cluster_config - computed: true, optional: false, required: false
        this._clusterConfig = new DataAwsOpensearchDomainClusterConfigList(this, "cluster_config", false);
        // cognito_options - computed: true, optional: false, required: false
        this._cognitoOptions = new DataAwsOpensearchDomainCognitoOptionsList(this, "cognito_options", false);
        // ebs_options - computed: true, optional: false, required: false
        this._ebsOptions = new DataAwsOpensearchDomainEbsOptionsList(this, "ebs_options", false);
        // encryption_at_rest - computed: true, optional: false, required: false
        this._encryptionAtRest = new DataAwsOpensearchDomainEncryptionAtRestList(this, "encryption_at_rest", false);
        // log_publishing_options - computed: true, optional: false, required: false
        this._logPublishingOptions = new DataAwsOpensearchDomainLogPublishingOptionsList(this, "log_publishing_options", true);
        // node_to_node_encryption - computed: true, optional: false, required: false
        this._nodeToNodeEncryption = new DataAwsOpensearchDomainNodeToNodeEncryptionList(this, "node_to_node_encryption", false);
        // snapshot_options - computed: true, optional: false, required: false
        this._snapshotOptions = new DataAwsOpensearchDomainSnapshotOptionsList(this, "snapshot_options", false);
        // vpc_options - computed: true, optional: false, required: false
        this._vpcOptions = new DataAwsOpensearchDomainVpcOptionsList(this, "vpc_options", false);
        this._domainName = config.domainName;
        this._id = config.id;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // access_policies - computed: true, optional: false, required: false
    get accessPolicies() {
        return this.getStringAttribute('access_policies');
    }
    get advancedOptions() {
        return this._advancedOptions;
    }
    get advancedSecurityOptions() {
        return this._advancedSecurityOptions;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get autoTuneOptions() {
        return this._autoTuneOptions;
    }
    get clusterConfig() {
        return this._clusterConfig;
    }
    get cognitoOptions() {
        return this._cognitoOptions;
    }
    // created - computed: true, optional: false, required: false
    get created() {
        return this.getBooleanAttribute('created');
    }
    // deleted - computed: true, optional: false, required: false
    get deleted() {
        return this.getBooleanAttribute('deleted');
    }
    // domain_id - computed: true, optional: false, required: false
    get domainId() {
        return this.getStringAttribute('domain_id');
    }
    get domainName() {
        return this.getStringAttribute('domain_name');
    }
    set domainName(value) {
        this._domainName = value;
    }
    // Temporarily expose input value. Use with caution.
    get domainNameInput() {
        return this._domainName;
    }
    get ebsOptions() {
        return this._ebsOptions;
    }
    get encryptionAtRest() {
        return this._encryptionAtRest;
    }
    // endpoint - computed: true, optional: false, required: false
    get endpoint() {
        return this.getStringAttribute('endpoint');
    }
    // engine_version - computed: true, optional: false, required: false
    get engineVersion() {
        return this.getStringAttribute('engine_version');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // kibana_endpoint - computed: true, optional: false, required: false
    get kibanaEndpoint() {
        return this.getStringAttribute('kibana_endpoint');
    }
    get logPublishingOptions() {
        return this._logPublishingOptions;
    }
    get nodeToNodeEncryption() {
        return this._nodeToNodeEncryption;
    }
    // processing - computed: true, optional: false, required: false
    get processing() {
        return this.getBooleanAttribute('processing');
    }
    get snapshotOptions() {
        return this._snapshotOptions;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get vpcOptions() {
        return this._vpcOptions;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            domain_name: cdktf.stringToTerraform(this._domainName),
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
}
exports.DataAwsOpensearchDomain = DataAwsOpensearchDomain;
_4 = JSII_RTTI_SYMBOL_1;
DataAwsOpensearchDomain[_4] = { fqn: "@cdktf/provider-aws.dataAwsOpensearchDomain.DataAwsOpensearchDomain", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsOpensearchDomain.tfResourceType = "aws_opensearch_domain";
