// @cdktf/provider-aws/lib/lightsail-lb-certificate-attachment/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LightsailLbCertificateAttachment = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/lightsail_lb_certificate_attachment aws_lightsail_lb_certificate_attachment}
*/
class LightsailLbCertificateAttachment extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/lightsail_lb_certificate_attachment aws_lightsail_lb_certificate_attachment} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options LightsailLbCertificateAttachmentConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_lightsail_lb_certificate_attachment',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._certificateName = config.certificateName;
        this._id = config.id;
        this._lbName = config.lbName;
    }
    get certificateName() {
        return this.getStringAttribute('certificate_name');
    }
    set certificateName(value) {
        this._certificateName = value;
    }
    // Temporarily expose input value. Use with caution.
    get certificateNameInput() {
        return this._certificateName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get lbName() {
        return this.getStringAttribute('lb_name');
    }
    set lbName(value) {
        this._lbName = value;
    }
    // Temporarily expose input value. Use with caution.
    get lbNameInput() {
        return this._lbName;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            certificate_name: cdktf.stringToTerraform(this._certificateName),
            id: cdktf.stringToTerraform(this._id),
            lb_name: cdktf.stringToTerraform(this._lbName),
        };
    }
}
exports.LightsailLbCertificateAttachment = LightsailLbCertificateAttachment;
_a = JSII_RTTI_SYMBOL_1;
LightsailLbCertificateAttachment[_a] = { fqn: "@cdktf/provider-aws.lightsailLbCertificateAttachment.LightsailLbCertificateAttachment", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
LightsailLbCertificateAttachment.tfResourceType = "aws_lightsail_lb_certificate_attachment";
