// @cdktf/provider-aws/lib/rds-reserved-instance/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RdsReservedInstance = exports.RdsReservedInstanceTimeoutsOutputReference = exports.rdsReservedInstanceTimeoutsToTerraform = exports.RdsReservedInstanceRecurringChargesList = exports.RdsReservedInstanceRecurringChargesOutputReference = exports.rdsReservedInstanceRecurringChargesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function rdsReservedInstanceRecurringChargesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.rdsReservedInstanceRecurringChargesToTerraform = rdsReservedInstanceRecurringChargesToTerraform;
class RdsReservedInstanceRecurringChargesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // recurring_charge_amount - computed: true, optional: false, required: false
    get recurringChargeAmount() {
        return this.getNumberAttribute('recurring_charge_amount');
    }
    // recurring_charge_frequency - computed: true, optional: false, required: false
    get recurringChargeFrequency() {
        return this.getStringAttribute('recurring_charge_frequency');
    }
}
exports.RdsReservedInstanceRecurringChargesOutputReference = RdsReservedInstanceRecurringChargesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
RdsReservedInstanceRecurringChargesOutputReference[_a] = { fqn: "@cdktf/provider-aws.rdsReservedInstance.RdsReservedInstanceRecurringChargesOutputReference", version: "10.0.12" };
class RdsReservedInstanceRecurringChargesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new RdsReservedInstanceRecurringChargesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.RdsReservedInstanceRecurringChargesList = RdsReservedInstanceRecurringChargesList;
_b = JSII_RTTI_SYMBOL_1;
RdsReservedInstanceRecurringChargesList[_b] = { fqn: "@cdktf/provider-aws.rdsReservedInstance.RdsReservedInstanceRecurringChargesList", version: "10.0.12" };
function rdsReservedInstanceTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.rdsReservedInstanceTimeoutsToTerraform = rdsReservedInstanceTimeoutsToTerraform;
class RdsReservedInstanceTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.RdsReservedInstanceTimeoutsOutputReference = RdsReservedInstanceTimeoutsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
RdsReservedInstanceTimeoutsOutputReference[_c] = { fqn: "@cdktf/provider-aws.rdsReservedInstance.RdsReservedInstanceTimeoutsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/rds_reserved_instance aws_rds_reserved_instance}
*/
class RdsReservedInstance extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/rds_reserved_instance aws_rds_reserved_instance} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options RdsReservedInstanceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_rds_reserved_instance',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // recurring_charges - computed: true, optional: false, required: false
        this._recurringCharges = new RdsReservedInstanceRecurringChargesList(this, "recurring_charges", false);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new RdsReservedInstanceTimeoutsOutputReference(this, "timeouts");
        this._id = config.id;
        this._instanceCount = config.instanceCount;
        this._offeringId = config.offeringId;
        this._reservationId = config.reservationId;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // currency_code - computed: true, optional: false, required: false
    get currencyCode() {
        return this.getStringAttribute('currency_code');
    }
    // db_instance_class - computed: true, optional: false, required: false
    get dbInstanceClass() {
        return this.getStringAttribute('db_instance_class');
    }
    // duration - computed: true, optional: false, required: false
    get duration() {
        return this.getNumberAttribute('duration');
    }
    // fixed_price - computed: true, optional: false, required: false
    get fixedPrice() {
        return this.getNumberAttribute('fixed_price');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instanceCount() {
        return this.getNumberAttribute('instance_count');
    }
    set instanceCount(value) {
        this._instanceCount = value;
    }
    resetInstanceCount() {
        this._instanceCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instanceCountInput() {
        return this._instanceCount;
    }
    // lease_id - computed: true, optional: false, required: false
    get leaseId() {
        return this.getStringAttribute('lease_id');
    }
    // multi_az - computed: true, optional: false, required: false
    get multiAz() {
        return this.getBooleanAttribute('multi_az');
    }
    get offeringId() {
        return this.getStringAttribute('offering_id');
    }
    set offeringId(value) {
        this._offeringId = value;
    }
    // Temporarily expose input value. Use with caution.
    get offeringIdInput() {
        return this._offeringId;
    }
    // offering_type - computed: true, optional: false, required: false
    get offeringType() {
        return this.getStringAttribute('offering_type');
    }
    // product_description - computed: true, optional: false, required: false
    get productDescription() {
        return this.getStringAttribute('product_description');
    }
    get recurringCharges() {
        return this._recurringCharges;
    }
    get reservationId() {
        return this.getStringAttribute('reservation_id');
    }
    set reservationId(value) {
        this._reservationId = value;
    }
    resetReservationId() {
        this._reservationId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get reservationIdInput() {
        return this._reservationId;
    }
    // start_time - computed: true, optional: false, required: false
    get startTime() {
        return this.getStringAttribute('start_time');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // usage_price - computed: true, optional: false, required: false
    get usagePrice() {
        return this.getNumberAttribute('usage_price');
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            instance_count: cdktf.numberToTerraform(this._instanceCount),
            offering_id: cdktf.stringToTerraform(this._offeringId),
            reservation_id: cdktf.stringToTerraform(this._reservationId),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            timeouts: rdsReservedInstanceTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.RdsReservedInstance = RdsReservedInstance;
_d = JSII_RTTI_SYMBOL_1;
RdsReservedInstance[_d] = { fqn: "@cdktf/provider-aws.rdsReservedInstance.RdsReservedInstance", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
RdsReservedInstance.tfResourceType = "aws_rds_reserved_instance";
