// @cdktf/provider-aws/lib/securityhub-standards-subscription/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecurityhubStandardsSubscription = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/securityhub_standards_subscription aws_securityhub_standards_subscription}
*/
class SecurityhubStandardsSubscription extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/securityhub_standards_subscription aws_securityhub_standards_subscription} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SecurityhubStandardsSubscriptionConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_securityhub_standards_subscription',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._standardsArn = config.standardsArn;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get standardsArn() {
        return this.getStringAttribute('standards_arn');
    }
    set standardsArn(value) {
        this._standardsArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get standardsArnInput() {
        return this._standardsArn;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            standards_arn: cdktf.stringToTerraform(this._standardsArn),
        };
    }
}
exports.SecurityhubStandardsSubscription = SecurityhubStandardsSubscription;
_a = JSII_RTTI_SYMBOL_1;
SecurityhubStandardsSubscription[_a] = { fqn: "@cdktf/provider-aws.securityhubStandardsSubscription.SecurityhubStandardsSubscription", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
SecurityhubStandardsSubscription.tfResourceType = "aws_securityhub_standards_subscription";
