// @cdktf/provider-aws/lib/data-aws-iam-user-ssh-key/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsIamUserSshKey = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/iam_user_ssh_key aws_iam_user_ssh_key}
*/
class DataAwsIamUserSshKey extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/iam_user_ssh_key aws_iam_user_ssh_key} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsIamUserSshKeyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iam_user_ssh_key',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._encoding = config.encoding;
        this._id = config.id;
        this._sshPublicKeyId = config.sshPublicKeyId;
        this._username = config.username;
    }
    get encoding() {
        return this.getStringAttribute('encoding');
    }
    set encoding(value) {
        this._encoding = value;
    }
    // Temporarily expose input value. Use with caution.
    get encodingInput() {
        return this._encoding;
    }
    // fingerprint - computed: true, optional: false, required: false
    get fingerprint() {
        return this.getStringAttribute('fingerprint');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // public_key - computed: true, optional: false, required: false
    get publicKey() {
        return this.getStringAttribute('public_key');
    }
    get sshPublicKeyId() {
        return this.getStringAttribute('ssh_public_key_id');
    }
    set sshPublicKeyId(value) {
        this._sshPublicKeyId = value;
    }
    // Temporarily expose input value. Use with caution.
    get sshPublicKeyIdInput() {
        return this._sshPublicKeyId;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get username() {
        return this.getStringAttribute('username');
    }
    set username(value) {
        this._username = value;
    }
    // Temporarily expose input value. Use with caution.
    get usernameInput() {
        return this._username;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            encoding: cdktf.stringToTerraform(this._encoding),
            id: cdktf.stringToTerraform(this._id),
            ssh_public_key_id: cdktf.stringToTerraform(this._sshPublicKeyId),
            username: cdktf.stringToTerraform(this._username),
        };
    }
}
exports.DataAwsIamUserSshKey = DataAwsIamUserSshKey;
_a = JSII_RTTI_SYMBOL_1;
DataAwsIamUserSshKey[_a] = { fqn: "@cdktf/provider-aws.dataAwsIamUserSshKey.DataAwsIamUserSshKey", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsIamUserSshKey.tfResourceType = "aws_iam_user_ssh_key";
