// @cdktf/provider-aws/lib/iot-thing/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IotThing = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/iot_thing aws_iot_thing}
*/
class IotThing extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/iot_thing aws_iot_thing} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options IotThingConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iot_thing',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._attributes = config.attributes;
        this._id = config.id;
        this._name = config.name;
        this._thingTypeName = config.thingTypeName;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get attributes() {
        return this.getStringMapAttribute('attributes');
    }
    set attributes(value) {
        this._attributes = value;
    }
    resetAttributes() {
        this._attributes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get attributesInput() {
        return this._attributes;
    }
    // default_client_id - computed: true, optional: false, required: false
    get defaultClientId() {
        return this.getStringAttribute('default_client_id');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get thingTypeName() {
        return this.getStringAttribute('thing_type_name');
    }
    set thingTypeName(value) {
        this._thingTypeName = value;
    }
    resetThingTypeName() {
        this._thingTypeName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get thingTypeNameInput() {
        return this._thingTypeName;
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getNumberAttribute('version');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            attributes: cdktf.hashMapper(cdktf.stringToTerraform)(this._attributes),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            thing_type_name: cdktf.stringToTerraform(this._thingTypeName),
        };
    }
}
exports.IotThing = IotThing;
_a = JSII_RTTI_SYMBOL_1;
IotThing[_a] = { fqn: "@cdktf/provider-aws.iotThing.IotThing", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
IotThing.tfResourceType = "aws_iot_thing";
