// @cdktf/provider-aws/lib/data-aws-imagebuilder-container-recipe/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsImagebuilderContainerRecipe = exports.DataAwsImagebuilderContainerRecipeTargetRepositoryList = exports.DataAwsImagebuilderContainerRecipeTargetRepositoryOutputReference = exports.dataAwsImagebuilderContainerRecipeTargetRepositoryToTerraform = exports.DataAwsImagebuilderContainerRecipeInstanceConfigurationList = exports.DataAwsImagebuilderContainerRecipeInstanceConfigurationOutputReference = exports.dataAwsImagebuilderContainerRecipeInstanceConfigurationToTerraform = exports.DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingList = exports.DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingOutputReference = exports.dataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingToTerraform = exports.DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingEbsList = exports.DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingEbsOutputReference = exports.dataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingEbsToTerraform = exports.DataAwsImagebuilderContainerRecipeComponentList = exports.DataAwsImagebuilderContainerRecipeComponentOutputReference = exports.dataAwsImagebuilderContainerRecipeComponentToTerraform = exports.DataAwsImagebuilderContainerRecipeComponentParameterList = exports.DataAwsImagebuilderContainerRecipeComponentParameterOutputReference = exports.dataAwsImagebuilderContainerRecipeComponentParameterToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsImagebuilderContainerRecipeComponentParameterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsImagebuilderContainerRecipeComponentParameterToTerraform = dataAwsImagebuilderContainerRecipeComponentParameterToTerraform;
class DataAwsImagebuilderContainerRecipeComponentParameterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // value - computed: true, optional: false, required: false
    get value() {
        return this.getStringAttribute('value');
    }
}
exports.DataAwsImagebuilderContainerRecipeComponentParameterOutputReference = DataAwsImagebuilderContainerRecipeComponentParameterOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderContainerRecipeComponentParameterOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsImagebuilderContainerRecipe.DataAwsImagebuilderContainerRecipeComponentParameterOutputReference", version: "10.0.12" };
class DataAwsImagebuilderContainerRecipeComponentParameterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsImagebuilderContainerRecipeComponentParameterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsImagebuilderContainerRecipeComponentParameterList = DataAwsImagebuilderContainerRecipeComponentParameterList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderContainerRecipeComponentParameterList[_b] = { fqn: "@cdktf/provider-aws.dataAwsImagebuilderContainerRecipe.DataAwsImagebuilderContainerRecipeComponentParameterList", version: "10.0.12" };
function dataAwsImagebuilderContainerRecipeComponentToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsImagebuilderContainerRecipeComponentToTerraform = dataAwsImagebuilderContainerRecipeComponentToTerraform;
class DataAwsImagebuilderContainerRecipeComponentOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // parameter - computed: true, optional: false, required: false
        this._parameter = new DataAwsImagebuilderContainerRecipeComponentParameterList(this, "parameter", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // component_arn - computed: true, optional: false, required: false
    get componentArn() {
        return this.getStringAttribute('component_arn');
    }
    get parameter() {
        return this._parameter;
    }
}
exports.DataAwsImagebuilderContainerRecipeComponentOutputReference = DataAwsImagebuilderContainerRecipeComponentOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderContainerRecipeComponentOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsImagebuilderContainerRecipe.DataAwsImagebuilderContainerRecipeComponentOutputReference", version: "10.0.12" };
class DataAwsImagebuilderContainerRecipeComponentList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsImagebuilderContainerRecipeComponentOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsImagebuilderContainerRecipeComponentList = DataAwsImagebuilderContainerRecipeComponentList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderContainerRecipeComponentList[_d] = { fqn: "@cdktf/provider-aws.dataAwsImagebuilderContainerRecipe.DataAwsImagebuilderContainerRecipeComponentList", version: "10.0.12" };
function dataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingEbsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingEbsToTerraform = dataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingEbsToTerraform;
class DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingEbsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // delete_on_termination - computed: true, optional: false, required: false
    get deleteOnTermination() {
        return this.getBooleanAttribute('delete_on_termination');
    }
    // encrypted - computed: true, optional: false, required: false
    get encrypted() {
        return this.getBooleanAttribute('encrypted');
    }
    // iops - computed: true, optional: false, required: false
    get iops() {
        return this.getNumberAttribute('iops');
    }
    // kms_key_id - computed: true, optional: false, required: false
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    // snapshot_id - computed: true, optional: false, required: false
    get snapshotId() {
        return this.getStringAttribute('snapshot_id');
    }
    // throughput - computed: true, optional: false, required: false
    get throughput() {
        return this.getNumberAttribute('throughput');
    }
    // volume_size - computed: true, optional: false, required: false
    get volumeSize() {
        return this.getNumberAttribute('volume_size');
    }
    // volume_type - computed: true, optional: false, required: false
    get volumeType() {
        return this.getStringAttribute('volume_type');
    }
}
exports.DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingEbsOutputReference = DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingEbsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingEbsOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsImagebuilderContainerRecipe.DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingEbsOutputReference", version: "10.0.12" };
class DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingEbsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingEbsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingEbsList = DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingEbsList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingEbsList[_f] = { fqn: "@cdktf/provider-aws.dataAwsImagebuilderContainerRecipe.DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingEbsList", version: "10.0.12" };
function dataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingToTerraform = dataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingToTerraform;
class DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // ebs - computed: true, optional: false, required: false
        this._ebs = new DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingEbsList(this, "ebs", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // device_name - computed: true, optional: false, required: false
    get deviceName() {
        return this.getStringAttribute('device_name');
    }
    get ebs() {
        return this._ebs;
    }
    // no_device - computed: true, optional: false, required: false
    get noDevice() {
        return this.getStringAttribute('no_device');
    }
    // virtual_name - computed: true, optional: false, required: false
    get virtualName() {
        return this.getStringAttribute('virtual_name');
    }
}
exports.DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingOutputReference = DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsImagebuilderContainerRecipe.DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingOutputReference", version: "10.0.12" };
class DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingList = DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingList[_h] = { fqn: "@cdktf/provider-aws.dataAwsImagebuilderContainerRecipe.DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingList", version: "10.0.12" };
function dataAwsImagebuilderContainerRecipeInstanceConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsImagebuilderContainerRecipeInstanceConfigurationToTerraform = dataAwsImagebuilderContainerRecipeInstanceConfigurationToTerraform;
class DataAwsImagebuilderContainerRecipeInstanceConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // block_device_mapping - computed: true, optional: false, required: false
        this._blockDeviceMapping = new DataAwsImagebuilderContainerRecipeInstanceConfigurationBlockDeviceMappingList(this, "block_device_mapping", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get blockDeviceMapping() {
        return this._blockDeviceMapping;
    }
    // image - computed: true, optional: false, required: false
    get image() {
        return this.getStringAttribute('image');
    }
}
exports.DataAwsImagebuilderContainerRecipeInstanceConfigurationOutputReference = DataAwsImagebuilderContainerRecipeInstanceConfigurationOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderContainerRecipeInstanceConfigurationOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsImagebuilderContainerRecipe.DataAwsImagebuilderContainerRecipeInstanceConfigurationOutputReference", version: "10.0.12" };
class DataAwsImagebuilderContainerRecipeInstanceConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsImagebuilderContainerRecipeInstanceConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsImagebuilderContainerRecipeInstanceConfigurationList = DataAwsImagebuilderContainerRecipeInstanceConfigurationList;
_k = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderContainerRecipeInstanceConfigurationList[_k] = { fqn: "@cdktf/provider-aws.dataAwsImagebuilderContainerRecipe.DataAwsImagebuilderContainerRecipeInstanceConfigurationList", version: "10.0.12" };
function dataAwsImagebuilderContainerRecipeTargetRepositoryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsImagebuilderContainerRecipeTargetRepositoryToTerraform = dataAwsImagebuilderContainerRecipeTargetRepositoryToTerraform;
class DataAwsImagebuilderContainerRecipeTargetRepositoryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // repository_name - computed: true, optional: false, required: false
    get repositoryName() {
        return this.getStringAttribute('repository_name');
    }
    // service - computed: true, optional: false, required: false
    get service() {
        return this.getStringAttribute('service');
    }
}
exports.DataAwsImagebuilderContainerRecipeTargetRepositoryOutputReference = DataAwsImagebuilderContainerRecipeTargetRepositoryOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderContainerRecipeTargetRepositoryOutputReference[_l] = { fqn: "@cdktf/provider-aws.dataAwsImagebuilderContainerRecipe.DataAwsImagebuilderContainerRecipeTargetRepositoryOutputReference", version: "10.0.12" };
class DataAwsImagebuilderContainerRecipeTargetRepositoryList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsImagebuilderContainerRecipeTargetRepositoryOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsImagebuilderContainerRecipeTargetRepositoryList = DataAwsImagebuilderContainerRecipeTargetRepositoryList;
_m = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderContainerRecipeTargetRepositoryList[_m] = { fqn: "@cdktf/provider-aws.dataAwsImagebuilderContainerRecipe.DataAwsImagebuilderContainerRecipeTargetRepositoryList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/imagebuilder_container_recipe aws_imagebuilder_container_recipe}
*/
class DataAwsImagebuilderContainerRecipe extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/imagebuilder_container_recipe aws_imagebuilder_container_recipe} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsImagebuilderContainerRecipeConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_imagebuilder_container_recipe',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // component - computed: true, optional: false, required: false
        this._component = new DataAwsImagebuilderContainerRecipeComponentList(this, "component", false);
        // instance_configuration - computed: true, optional: false, required: false
        this._instanceConfiguration = new DataAwsImagebuilderContainerRecipeInstanceConfigurationList(this, "instance_configuration", false);
        // target_repository - computed: true, optional: false, required: false
        this._targetRepository = new DataAwsImagebuilderContainerRecipeTargetRepositoryList(this, "target_repository", false);
        this._arn = config.arn;
        this._id = config.id;
        this._tags = config.tags;
    }
    get arn() {
        return this.getStringAttribute('arn');
    }
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    get arnInput() {
        return this._arn;
    }
    get component() {
        return this._component;
    }
    // container_type - computed: true, optional: false, required: false
    get containerType() {
        return this.getStringAttribute('container_type');
    }
    // date_created - computed: true, optional: false, required: false
    get dateCreated() {
        return this.getStringAttribute('date_created');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // dockerfile_template_data - computed: true, optional: false, required: false
    get dockerfileTemplateData() {
        return this.getStringAttribute('dockerfile_template_data');
    }
    // encrypted - computed: true, optional: false, required: false
    get encrypted() {
        return this.getBooleanAttribute('encrypted');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instanceConfiguration() {
        return this._instanceConfiguration;
    }
    // kms_key_id - computed: true, optional: false, required: false
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // owner - computed: true, optional: false, required: false
    get owner() {
        return this.getStringAttribute('owner');
    }
    // parent_image - computed: true, optional: false, required: false
    get parentImage() {
        return this.getStringAttribute('parent_image');
    }
    // platform - computed: true, optional: false, required: false
    get platform() {
        return this.getStringAttribute('platform');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get targetRepository() {
        return this._targetRepository;
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
    // working_directory - computed: true, optional: false, required: false
    get workingDirectory() {
        return this.getStringAttribute('working_directory');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            arn: cdktf.stringToTerraform(this._arn),
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
}
exports.DataAwsImagebuilderContainerRecipe = DataAwsImagebuilderContainerRecipe;
_o = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderContainerRecipe[_o] = { fqn: "@cdktf/provider-aws.dataAwsImagebuilderContainerRecipe.DataAwsImagebuilderContainerRecipe", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsImagebuilderContainerRecipe.tfResourceType = "aws_imagebuilder_container_recipe";
