// @cdktf/provider-aws/lib/transfer-workflow/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferWorkflow = exports.TransferWorkflowStepsList = exports.TransferWorkflowStepsOutputReference = exports.transferWorkflowStepsToTerraform = exports.TransferWorkflowStepsTagStepDetailsOutputReference = exports.transferWorkflowStepsTagStepDetailsToTerraform = exports.TransferWorkflowStepsTagStepDetailsTagsList = exports.TransferWorkflowStepsTagStepDetailsTagsOutputReference = exports.transferWorkflowStepsTagStepDetailsTagsToTerraform = exports.TransferWorkflowStepsDeleteStepDetailsOutputReference = exports.transferWorkflowStepsDeleteStepDetailsToTerraform = exports.TransferWorkflowStepsCustomStepDetailsOutputReference = exports.transferWorkflowStepsCustomStepDetailsToTerraform = exports.TransferWorkflowStepsCopyStepDetailsOutputReference = exports.transferWorkflowStepsCopyStepDetailsToTerraform = exports.TransferWorkflowStepsCopyStepDetailsDestinationFileLocationOutputReference = exports.transferWorkflowStepsCopyStepDetailsDestinationFileLocationToTerraform = exports.TransferWorkflowStepsCopyStepDetailsDestinationFileLocationS3FileLocationOutputReference = exports.transferWorkflowStepsCopyStepDetailsDestinationFileLocationS3FileLocationToTerraform = exports.TransferWorkflowStepsCopyStepDetailsDestinationFileLocationEfsFileLocationOutputReference = exports.transferWorkflowStepsCopyStepDetailsDestinationFileLocationEfsFileLocationToTerraform = exports.TransferWorkflowOnExceptionStepsList = exports.TransferWorkflowOnExceptionStepsOutputReference = exports.transferWorkflowOnExceptionStepsToTerraform = exports.TransferWorkflowOnExceptionStepsTagStepDetailsOutputReference = exports.transferWorkflowOnExceptionStepsTagStepDetailsToTerraform = exports.TransferWorkflowOnExceptionStepsTagStepDetailsTagsList = exports.TransferWorkflowOnExceptionStepsTagStepDetailsTagsOutputReference = exports.transferWorkflowOnExceptionStepsTagStepDetailsTagsToTerraform = exports.TransferWorkflowOnExceptionStepsDeleteStepDetailsOutputReference = exports.transferWorkflowOnExceptionStepsDeleteStepDetailsToTerraform = exports.TransferWorkflowOnExceptionStepsCustomStepDetailsOutputReference = exports.transferWorkflowOnExceptionStepsCustomStepDetailsToTerraform = exports.TransferWorkflowOnExceptionStepsCopyStepDetailsOutputReference = exports.transferWorkflowOnExceptionStepsCopyStepDetailsToTerraform = exports.TransferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationOutputReference = exports.transferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationToTerraform = exports.TransferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationS3FileLocationOutputReference = exports.transferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationS3FileLocationToTerraform = exports.TransferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationEfsFileLocationOutputReference = exports.transferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationEfsFileLocationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function transferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationEfsFileLocationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        file_system_id: cdktf.stringToTerraform(struct.fileSystemId),
        path: cdktf.stringToTerraform(struct.path),
    };
}
exports.transferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationEfsFileLocationToTerraform = transferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationEfsFileLocationToTerraform;
class TransferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationEfsFileLocationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._fileSystemId !== undefined) {
            hasAnyValues = true;
            internalValueResult.fileSystemId = this._fileSystemId;
        }
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._fileSystemId = undefined;
            this._path = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._fileSystemId = value.fileSystemId;
            this._path = value.path;
        }
    }
    get fileSystemId() {
        return this.getStringAttribute('file_system_id');
    }
    set fileSystemId(value) {
        this._fileSystemId = value;
    }
    resetFileSystemId() {
        this._fileSystemId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fileSystemIdInput() {
        return this._fileSystemId;
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    resetPath() {
        this._path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
}
exports.TransferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationEfsFileLocationOutputReference = TransferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationEfsFileLocationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
TransferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationEfsFileLocationOutputReference[_a] = { fqn: "@cdktf/provider-aws.transferWorkflow.TransferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationEfsFileLocationOutputReference", version: "10.0.12" };
function transferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationS3FileLocationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket: cdktf.stringToTerraform(struct.bucket),
        key: cdktf.stringToTerraform(struct.key),
    };
}
exports.transferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationS3FileLocationToTerraform = transferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationS3FileLocationToTerraform;
class TransferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationS3FileLocationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucket !== undefined) {
            hasAnyValues = true;
            internalValueResult.bucket = this._bucket;
        }
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucket = undefined;
            this._key = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucket = value.bucket;
            this._key = value.key;
        }
    }
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    set bucket(value) {
        this._bucket = value;
    }
    resetBucket() {
        this._bucket = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bucketInput() {
        return this._bucket;
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
}
exports.TransferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationS3FileLocationOutputReference = TransferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationS3FileLocationOutputReference;
_b = JSII_RTTI_SYMBOL_1;
TransferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationS3FileLocationOutputReference[_b] = { fqn: "@cdktf/provider-aws.transferWorkflow.TransferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationS3FileLocationOutputReference", version: "10.0.12" };
function transferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        efs_file_location: transferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationEfsFileLocationToTerraform(struct.efsFileLocation),
        s3_file_location: transferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationS3FileLocationToTerraform(struct.s3FileLocation),
    };
}
exports.transferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationToTerraform = transferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationToTerraform;
class TransferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // efs_file_location - computed: false, optional: true, required: false
        this._efsFileLocation = new TransferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationEfsFileLocationOutputReference(this, "efs_file_location");
        // s3_file_location - computed: false, optional: true, required: false
        this._s3FileLocation = new TransferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationS3FileLocationOutputReference(this, "s3_file_location");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._efsFileLocation?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.efsFileLocation = this._efsFileLocation?.internalValue;
        }
        if (this._s3FileLocation?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3FileLocation = this._s3FileLocation?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._efsFileLocation.internalValue = undefined;
            this._s3FileLocation.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._efsFileLocation.internalValue = value.efsFileLocation;
            this._s3FileLocation.internalValue = value.s3FileLocation;
        }
    }
    get efsFileLocation() {
        return this._efsFileLocation;
    }
    putEfsFileLocation(value) {
        this._efsFileLocation.internalValue = value;
    }
    resetEfsFileLocation() {
        this._efsFileLocation.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get efsFileLocationInput() {
        return this._efsFileLocation.internalValue;
    }
    get s3FileLocation() {
        return this._s3FileLocation;
    }
    putS3FileLocation(value) {
        this._s3FileLocation.internalValue = value;
    }
    resetS3FileLocation() {
        this._s3FileLocation.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get s3FileLocationInput() {
        return this._s3FileLocation.internalValue;
    }
}
exports.TransferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationOutputReference = TransferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
TransferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationOutputReference[_c] = { fqn: "@cdktf/provider-aws.transferWorkflow.TransferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationOutputReference", version: "10.0.12" };
function transferWorkflowOnExceptionStepsCopyStepDetailsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        overwrite_existing: cdktf.stringToTerraform(struct.overwriteExisting),
        source_file_location: cdktf.stringToTerraform(struct.sourceFileLocation),
        destination_file_location: transferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationToTerraform(struct.destinationFileLocation),
    };
}
exports.transferWorkflowOnExceptionStepsCopyStepDetailsToTerraform = transferWorkflowOnExceptionStepsCopyStepDetailsToTerraform;
class TransferWorkflowOnExceptionStepsCopyStepDetailsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // destination_file_location - computed: false, optional: true, required: false
        this._destinationFileLocation = new TransferWorkflowOnExceptionStepsCopyStepDetailsDestinationFileLocationOutputReference(this, "destination_file_location");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._overwriteExisting !== undefined) {
            hasAnyValues = true;
            internalValueResult.overwriteExisting = this._overwriteExisting;
        }
        if (this._sourceFileLocation !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceFileLocation = this._sourceFileLocation;
        }
        if (this._destinationFileLocation?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.destinationFileLocation = this._destinationFileLocation?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._name = undefined;
            this._overwriteExisting = undefined;
            this._sourceFileLocation = undefined;
            this._destinationFileLocation.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._name = value.name;
            this._overwriteExisting = value.overwriteExisting;
            this._sourceFileLocation = value.sourceFileLocation;
            this._destinationFileLocation.internalValue = value.destinationFileLocation;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get overwriteExisting() {
        return this.getStringAttribute('overwrite_existing');
    }
    set overwriteExisting(value) {
        this._overwriteExisting = value;
    }
    resetOverwriteExisting() {
        this._overwriteExisting = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get overwriteExistingInput() {
        return this._overwriteExisting;
    }
    get sourceFileLocation() {
        return this.getStringAttribute('source_file_location');
    }
    set sourceFileLocation(value) {
        this._sourceFileLocation = value;
    }
    resetSourceFileLocation() {
        this._sourceFileLocation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceFileLocationInput() {
        return this._sourceFileLocation;
    }
    get destinationFileLocation() {
        return this._destinationFileLocation;
    }
    putDestinationFileLocation(value) {
        this._destinationFileLocation.internalValue = value;
    }
    resetDestinationFileLocation() {
        this._destinationFileLocation.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get destinationFileLocationInput() {
        return this._destinationFileLocation.internalValue;
    }
}
exports.TransferWorkflowOnExceptionStepsCopyStepDetailsOutputReference = TransferWorkflowOnExceptionStepsCopyStepDetailsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
TransferWorkflowOnExceptionStepsCopyStepDetailsOutputReference[_d] = { fqn: "@cdktf/provider-aws.transferWorkflow.TransferWorkflowOnExceptionStepsCopyStepDetailsOutputReference", version: "10.0.12" };
function transferWorkflowOnExceptionStepsCustomStepDetailsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        source_file_location: cdktf.stringToTerraform(struct.sourceFileLocation),
        target: cdktf.stringToTerraform(struct.target),
        timeout_seconds: cdktf.numberToTerraform(struct.timeoutSeconds),
    };
}
exports.transferWorkflowOnExceptionStepsCustomStepDetailsToTerraform = transferWorkflowOnExceptionStepsCustomStepDetailsToTerraform;
class TransferWorkflowOnExceptionStepsCustomStepDetailsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._sourceFileLocation !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceFileLocation = this._sourceFileLocation;
        }
        if (this._target !== undefined) {
            hasAnyValues = true;
            internalValueResult.target = this._target;
        }
        if (this._timeoutSeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeoutSeconds = this._timeoutSeconds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._name = undefined;
            this._sourceFileLocation = undefined;
            this._target = undefined;
            this._timeoutSeconds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._name = value.name;
            this._sourceFileLocation = value.sourceFileLocation;
            this._target = value.target;
            this._timeoutSeconds = value.timeoutSeconds;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get sourceFileLocation() {
        return this.getStringAttribute('source_file_location');
    }
    set sourceFileLocation(value) {
        this._sourceFileLocation = value;
    }
    resetSourceFileLocation() {
        this._sourceFileLocation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceFileLocationInput() {
        return this._sourceFileLocation;
    }
    get target() {
        return this.getStringAttribute('target');
    }
    set target(value) {
        this._target = value;
    }
    resetTarget() {
        this._target = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetInput() {
        return this._target;
    }
    get timeoutSeconds() {
        return this.getNumberAttribute('timeout_seconds');
    }
    set timeoutSeconds(value) {
        this._timeoutSeconds = value;
    }
    resetTimeoutSeconds() {
        this._timeoutSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutSecondsInput() {
        return this._timeoutSeconds;
    }
}
exports.TransferWorkflowOnExceptionStepsCustomStepDetailsOutputReference = TransferWorkflowOnExceptionStepsCustomStepDetailsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
TransferWorkflowOnExceptionStepsCustomStepDetailsOutputReference[_e] = { fqn: "@cdktf/provider-aws.transferWorkflow.TransferWorkflowOnExceptionStepsCustomStepDetailsOutputReference", version: "10.0.12" };
function transferWorkflowOnExceptionStepsDeleteStepDetailsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        source_file_location: cdktf.stringToTerraform(struct.sourceFileLocation),
    };
}
exports.transferWorkflowOnExceptionStepsDeleteStepDetailsToTerraform = transferWorkflowOnExceptionStepsDeleteStepDetailsToTerraform;
class TransferWorkflowOnExceptionStepsDeleteStepDetailsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._sourceFileLocation !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceFileLocation = this._sourceFileLocation;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._name = undefined;
            this._sourceFileLocation = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._name = value.name;
            this._sourceFileLocation = value.sourceFileLocation;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get sourceFileLocation() {
        return this.getStringAttribute('source_file_location');
    }
    set sourceFileLocation(value) {
        this._sourceFileLocation = value;
    }
    resetSourceFileLocation() {
        this._sourceFileLocation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceFileLocationInput() {
        return this._sourceFileLocation;
    }
}
exports.TransferWorkflowOnExceptionStepsDeleteStepDetailsOutputReference = TransferWorkflowOnExceptionStepsDeleteStepDetailsOutputReference;
_f = JSII_RTTI_SYMBOL_1;
TransferWorkflowOnExceptionStepsDeleteStepDetailsOutputReference[_f] = { fqn: "@cdktf/provider-aws.transferWorkflow.TransferWorkflowOnExceptionStepsDeleteStepDetailsOutputReference", version: "10.0.12" };
function transferWorkflowOnExceptionStepsTagStepDetailsTagsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.transferWorkflowOnExceptionStepsTagStepDetailsTagsToTerraform = transferWorkflowOnExceptionStepsTagStepDetailsTagsToTerraform;
class TransferWorkflowOnExceptionStepsTagStepDetailsTagsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._key = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._key = value.key;
            this._value = value.value;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.TransferWorkflowOnExceptionStepsTagStepDetailsTagsOutputReference = TransferWorkflowOnExceptionStepsTagStepDetailsTagsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
TransferWorkflowOnExceptionStepsTagStepDetailsTagsOutputReference[_g] = { fqn: "@cdktf/provider-aws.transferWorkflow.TransferWorkflowOnExceptionStepsTagStepDetailsTagsOutputReference", version: "10.0.12" };
class TransferWorkflowOnExceptionStepsTagStepDetailsTagsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new TransferWorkflowOnExceptionStepsTagStepDetailsTagsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.TransferWorkflowOnExceptionStepsTagStepDetailsTagsList = TransferWorkflowOnExceptionStepsTagStepDetailsTagsList;
_h = JSII_RTTI_SYMBOL_1;
TransferWorkflowOnExceptionStepsTagStepDetailsTagsList[_h] = { fqn: "@cdktf/provider-aws.transferWorkflow.TransferWorkflowOnExceptionStepsTagStepDetailsTagsList", version: "10.0.12" };
function transferWorkflowOnExceptionStepsTagStepDetailsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        source_file_location: cdktf.stringToTerraform(struct.sourceFileLocation),
        tags: cdktf.listMapper(transferWorkflowOnExceptionStepsTagStepDetailsTagsToTerraform, true)(struct.tags),
    };
}
exports.transferWorkflowOnExceptionStepsTagStepDetailsToTerraform = transferWorkflowOnExceptionStepsTagStepDetailsToTerraform;
class TransferWorkflowOnExceptionStepsTagStepDetailsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // tags - computed: false, optional: true, required: false
        this._tags = new TransferWorkflowOnExceptionStepsTagStepDetailsTagsList(this, "tags", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._sourceFileLocation !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceFileLocation = this._sourceFileLocation;
        }
        if (this._tags?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.tags = this._tags?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._name = undefined;
            this._sourceFileLocation = undefined;
            this._tags.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._name = value.name;
            this._sourceFileLocation = value.sourceFileLocation;
            this._tags.internalValue = value.tags;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get sourceFileLocation() {
        return this.getStringAttribute('source_file_location');
    }
    set sourceFileLocation(value) {
        this._sourceFileLocation = value;
    }
    resetSourceFileLocation() {
        this._sourceFileLocation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceFileLocationInput() {
        return this._sourceFileLocation;
    }
    get tags() {
        return this._tags;
    }
    putTags(value) {
        this._tags.internalValue = value;
    }
    resetTags() {
        this._tags.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags.internalValue;
    }
}
exports.TransferWorkflowOnExceptionStepsTagStepDetailsOutputReference = TransferWorkflowOnExceptionStepsTagStepDetailsOutputReference;
_j = JSII_RTTI_SYMBOL_1;
TransferWorkflowOnExceptionStepsTagStepDetailsOutputReference[_j] = { fqn: "@cdktf/provider-aws.transferWorkflow.TransferWorkflowOnExceptionStepsTagStepDetailsOutputReference", version: "10.0.12" };
function transferWorkflowOnExceptionStepsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
        copy_step_details: transferWorkflowOnExceptionStepsCopyStepDetailsToTerraform(struct.copyStepDetails),
        custom_step_details: transferWorkflowOnExceptionStepsCustomStepDetailsToTerraform(struct.customStepDetails),
        delete_step_details: transferWorkflowOnExceptionStepsDeleteStepDetailsToTerraform(struct.deleteStepDetails),
        tag_step_details: transferWorkflowOnExceptionStepsTagStepDetailsToTerraform(struct.tagStepDetails),
    };
}
exports.transferWorkflowOnExceptionStepsToTerraform = transferWorkflowOnExceptionStepsToTerraform;
class TransferWorkflowOnExceptionStepsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // copy_step_details - computed: false, optional: true, required: false
        this._copyStepDetails = new TransferWorkflowOnExceptionStepsCopyStepDetailsOutputReference(this, "copy_step_details");
        // custom_step_details - computed: false, optional: true, required: false
        this._customStepDetails = new TransferWorkflowOnExceptionStepsCustomStepDetailsOutputReference(this, "custom_step_details");
        // delete_step_details - computed: false, optional: true, required: false
        this._deleteStepDetails = new TransferWorkflowOnExceptionStepsDeleteStepDetailsOutputReference(this, "delete_step_details");
        // tag_step_details - computed: false, optional: true, required: false
        this._tagStepDetails = new TransferWorkflowOnExceptionStepsTagStepDetailsOutputReference(this, "tag_step_details");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._copyStepDetails?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.copyStepDetails = this._copyStepDetails?.internalValue;
        }
        if (this._customStepDetails?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.customStepDetails = this._customStepDetails?.internalValue;
        }
        if (this._deleteStepDetails?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.deleteStepDetails = this._deleteStepDetails?.internalValue;
        }
        if (this._tagStepDetails?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.tagStepDetails = this._tagStepDetails?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._type = undefined;
            this._copyStepDetails.internalValue = undefined;
            this._customStepDetails.internalValue = undefined;
            this._deleteStepDetails.internalValue = undefined;
            this._tagStepDetails.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._type = value.type;
            this._copyStepDetails.internalValue = value.copyStepDetails;
            this._customStepDetails.internalValue = value.customStepDetails;
            this._deleteStepDetails.internalValue = value.deleteStepDetails;
            this._tagStepDetails.internalValue = value.tagStepDetails;
        }
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get copyStepDetails() {
        return this._copyStepDetails;
    }
    putCopyStepDetails(value) {
        this._copyStepDetails.internalValue = value;
    }
    resetCopyStepDetails() {
        this._copyStepDetails.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get copyStepDetailsInput() {
        return this._copyStepDetails.internalValue;
    }
    get customStepDetails() {
        return this._customStepDetails;
    }
    putCustomStepDetails(value) {
        this._customStepDetails.internalValue = value;
    }
    resetCustomStepDetails() {
        this._customStepDetails.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get customStepDetailsInput() {
        return this._customStepDetails.internalValue;
    }
    get deleteStepDetails() {
        return this._deleteStepDetails;
    }
    putDeleteStepDetails(value) {
        this._deleteStepDetails.internalValue = value;
    }
    resetDeleteStepDetails() {
        this._deleteStepDetails.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteStepDetailsInput() {
        return this._deleteStepDetails.internalValue;
    }
    get tagStepDetails() {
        return this._tagStepDetails;
    }
    putTagStepDetails(value) {
        this._tagStepDetails.internalValue = value;
    }
    resetTagStepDetails() {
        this._tagStepDetails.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagStepDetailsInput() {
        return this._tagStepDetails.internalValue;
    }
}
exports.TransferWorkflowOnExceptionStepsOutputReference = TransferWorkflowOnExceptionStepsOutputReference;
_k = JSII_RTTI_SYMBOL_1;
TransferWorkflowOnExceptionStepsOutputReference[_k] = { fqn: "@cdktf/provider-aws.transferWorkflow.TransferWorkflowOnExceptionStepsOutputReference", version: "10.0.12" };
class TransferWorkflowOnExceptionStepsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new TransferWorkflowOnExceptionStepsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.TransferWorkflowOnExceptionStepsList = TransferWorkflowOnExceptionStepsList;
_l = JSII_RTTI_SYMBOL_1;
TransferWorkflowOnExceptionStepsList[_l] = { fqn: "@cdktf/provider-aws.transferWorkflow.TransferWorkflowOnExceptionStepsList", version: "10.0.12" };
function transferWorkflowStepsCopyStepDetailsDestinationFileLocationEfsFileLocationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        file_system_id: cdktf.stringToTerraform(struct.fileSystemId),
        path: cdktf.stringToTerraform(struct.path),
    };
}
exports.transferWorkflowStepsCopyStepDetailsDestinationFileLocationEfsFileLocationToTerraform = transferWorkflowStepsCopyStepDetailsDestinationFileLocationEfsFileLocationToTerraform;
class TransferWorkflowStepsCopyStepDetailsDestinationFileLocationEfsFileLocationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._fileSystemId !== undefined) {
            hasAnyValues = true;
            internalValueResult.fileSystemId = this._fileSystemId;
        }
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._fileSystemId = undefined;
            this._path = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._fileSystemId = value.fileSystemId;
            this._path = value.path;
        }
    }
    get fileSystemId() {
        return this.getStringAttribute('file_system_id');
    }
    set fileSystemId(value) {
        this._fileSystemId = value;
    }
    resetFileSystemId() {
        this._fileSystemId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fileSystemIdInput() {
        return this._fileSystemId;
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    resetPath() {
        this._path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
}
exports.TransferWorkflowStepsCopyStepDetailsDestinationFileLocationEfsFileLocationOutputReference = TransferWorkflowStepsCopyStepDetailsDestinationFileLocationEfsFileLocationOutputReference;
_m = JSII_RTTI_SYMBOL_1;
TransferWorkflowStepsCopyStepDetailsDestinationFileLocationEfsFileLocationOutputReference[_m] = { fqn: "@cdktf/provider-aws.transferWorkflow.TransferWorkflowStepsCopyStepDetailsDestinationFileLocationEfsFileLocationOutputReference", version: "10.0.12" };
function transferWorkflowStepsCopyStepDetailsDestinationFileLocationS3FileLocationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket: cdktf.stringToTerraform(struct.bucket),
        key: cdktf.stringToTerraform(struct.key),
    };
}
exports.transferWorkflowStepsCopyStepDetailsDestinationFileLocationS3FileLocationToTerraform = transferWorkflowStepsCopyStepDetailsDestinationFileLocationS3FileLocationToTerraform;
class TransferWorkflowStepsCopyStepDetailsDestinationFileLocationS3FileLocationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucket !== undefined) {
            hasAnyValues = true;
            internalValueResult.bucket = this._bucket;
        }
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucket = undefined;
            this._key = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucket = value.bucket;
            this._key = value.key;
        }
    }
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    set bucket(value) {
        this._bucket = value;
    }
    resetBucket() {
        this._bucket = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bucketInput() {
        return this._bucket;
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
}
exports.TransferWorkflowStepsCopyStepDetailsDestinationFileLocationS3FileLocationOutputReference = TransferWorkflowStepsCopyStepDetailsDestinationFileLocationS3FileLocationOutputReference;
_o = JSII_RTTI_SYMBOL_1;
TransferWorkflowStepsCopyStepDetailsDestinationFileLocationS3FileLocationOutputReference[_o] = { fqn: "@cdktf/provider-aws.transferWorkflow.TransferWorkflowStepsCopyStepDetailsDestinationFileLocationS3FileLocationOutputReference", version: "10.0.12" };
function transferWorkflowStepsCopyStepDetailsDestinationFileLocationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        efs_file_location: transferWorkflowStepsCopyStepDetailsDestinationFileLocationEfsFileLocationToTerraform(struct.efsFileLocation),
        s3_file_location: transferWorkflowStepsCopyStepDetailsDestinationFileLocationS3FileLocationToTerraform(struct.s3FileLocation),
    };
}
exports.transferWorkflowStepsCopyStepDetailsDestinationFileLocationToTerraform = transferWorkflowStepsCopyStepDetailsDestinationFileLocationToTerraform;
class TransferWorkflowStepsCopyStepDetailsDestinationFileLocationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // efs_file_location - computed: false, optional: true, required: false
        this._efsFileLocation = new TransferWorkflowStepsCopyStepDetailsDestinationFileLocationEfsFileLocationOutputReference(this, "efs_file_location");
        // s3_file_location - computed: false, optional: true, required: false
        this._s3FileLocation = new TransferWorkflowStepsCopyStepDetailsDestinationFileLocationS3FileLocationOutputReference(this, "s3_file_location");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._efsFileLocation?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.efsFileLocation = this._efsFileLocation?.internalValue;
        }
        if (this._s3FileLocation?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3FileLocation = this._s3FileLocation?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._efsFileLocation.internalValue = undefined;
            this._s3FileLocation.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._efsFileLocation.internalValue = value.efsFileLocation;
            this._s3FileLocation.internalValue = value.s3FileLocation;
        }
    }
    get efsFileLocation() {
        return this._efsFileLocation;
    }
    putEfsFileLocation(value) {
        this._efsFileLocation.internalValue = value;
    }
    resetEfsFileLocation() {
        this._efsFileLocation.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get efsFileLocationInput() {
        return this._efsFileLocation.internalValue;
    }
    get s3FileLocation() {
        return this._s3FileLocation;
    }
    putS3FileLocation(value) {
        this._s3FileLocation.internalValue = value;
    }
    resetS3FileLocation() {
        this._s3FileLocation.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get s3FileLocationInput() {
        return this._s3FileLocation.internalValue;
    }
}
exports.TransferWorkflowStepsCopyStepDetailsDestinationFileLocationOutputReference = TransferWorkflowStepsCopyStepDetailsDestinationFileLocationOutputReference;
_p = JSII_RTTI_SYMBOL_1;
TransferWorkflowStepsCopyStepDetailsDestinationFileLocationOutputReference[_p] = { fqn: "@cdktf/provider-aws.transferWorkflow.TransferWorkflowStepsCopyStepDetailsDestinationFileLocationOutputReference", version: "10.0.12" };
function transferWorkflowStepsCopyStepDetailsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        overwrite_existing: cdktf.stringToTerraform(struct.overwriteExisting),
        source_file_location: cdktf.stringToTerraform(struct.sourceFileLocation),
        destination_file_location: transferWorkflowStepsCopyStepDetailsDestinationFileLocationToTerraform(struct.destinationFileLocation),
    };
}
exports.transferWorkflowStepsCopyStepDetailsToTerraform = transferWorkflowStepsCopyStepDetailsToTerraform;
class TransferWorkflowStepsCopyStepDetailsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // destination_file_location - computed: false, optional: true, required: false
        this._destinationFileLocation = new TransferWorkflowStepsCopyStepDetailsDestinationFileLocationOutputReference(this, "destination_file_location");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._overwriteExisting !== undefined) {
            hasAnyValues = true;
            internalValueResult.overwriteExisting = this._overwriteExisting;
        }
        if (this._sourceFileLocation !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceFileLocation = this._sourceFileLocation;
        }
        if (this._destinationFileLocation?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.destinationFileLocation = this._destinationFileLocation?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._name = undefined;
            this._overwriteExisting = undefined;
            this._sourceFileLocation = undefined;
            this._destinationFileLocation.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._name = value.name;
            this._overwriteExisting = value.overwriteExisting;
            this._sourceFileLocation = value.sourceFileLocation;
            this._destinationFileLocation.internalValue = value.destinationFileLocation;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get overwriteExisting() {
        return this.getStringAttribute('overwrite_existing');
    }
    set overwriteExisting(value) {
        this._overwriteExisting = value;
    }
    resetOverwriteExisting() {
        this._overwriteExisting = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get overwriteExistingInput() {
        return this._overwriteExisting;
    }
    get sourceFileLocation() {
        return this.getStringAttribute('source_file_location');
    }
    set sourceFileLocation(value) {
        this._sourceFileLocation = value;
    }
    resetSourceFileLocation() {
        this._sourceFileLocation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceFileLocationInput() {
        return this._sourceFileLocation;
    }
    get destinationFileLocation() {
        return this._destinationFileLocation;
    }
    putDestinationFileLocation(value) {
        this._destinationFileLocation.internalValue = value;
    }
    resetDestinationFileLocation() {
        this._destinationFileLocation.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get destinationFileLocationInput() {
        return this._destinationFileLocation.internalValue;
    }
}
exports.TransferWorkflowStepsCopyStepDetailsOutputReference = TransferWorkflowStepsCopyStepDetailsOutputReference;
_q = JSII_RTTI_SYMBOL_1;
TransferWorkflowStepsCopyStepDetailsOutputReference[_q] = { fqn: "@cdktf/provider-aws.transferWorkflow.TransferWorkflowStepsCopyStepDetailsOutputReference", version: "10.0.12" };
function transferWorkflowStepsCustomStepDetailsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        source_file_location: cdktf.stringToTerraform(struct.sourceFileLocation),
        target: cdktf.stringToTerraform(struct.target),
        timeout_seconds: cdktf.numberToTerraform(struct.timeoutSeconds),
    };
}
exports.transferWorkflowStepsCustomStepDetailsToTerraform = transferWorkflowStepsCustomStepDetailsToTerraform;
class TransferWorkflowStepsCustomStepDetailsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._sourceFileLocation !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceFileLocation = this._sourceFileLocation;
        }
        if (this._target !== undefined) {
            hasAnyValues = true;
            internalValueResult.target = this._target;
        }
        if (this._timeoutSeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeoutSeconds = this._timeoutSeconds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._name = undefined;
            this._sourceFileLocation = undefined;
            this._target = undefined;
            this._timeoutSeconds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._name = value.name;
            this._sourceFileLocation = value.sourceFileLocation;
            this._target = value.target;
            this._timeoutSeconds = value.timeoutSeconds;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get sourceFileLocation() {
        return this.getStringAttribute('source_file_location');
    }
    set sourceFileLocation(value) {
        this._sourceFileLocation = value;
    }
    resetSourceFileLocation() {
        this._sourceFileLocation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceFileLocationInput() {
        return this._sourceFileLocation;
    }
    get target() {
        return this.getStringAttribute('target');
    }
    set target(value) {
        this._target = value;
    }
    resetTarget() {
        this._target = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetInput() {
        return this._target;
    }
    get timeoutSeconds() {
        return this.getNumberAttribute('timeout_seconds');
    }
    set timeoutSeconds(value) {
        this._timeoutSeconds = value;
    }
    resetTimeoutSeconds() {
        this._timeoutSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutSecondsInput() {
        return this._timeoutSeconds;
    }
}
exports.TransferWorkflowStepsCustomStepDetailsOutputReference = TransferWorkflowStepsCustomStepDetailsOutputReference;
_r = JSII_RTTI_SYMBOL_1;
TransferWorkflowStepsCustomStepDetailsOutputReference[_r] = { fqn: "@cdktf/provider-aws.transferWorkflow.TransferWorkflowStepsCustomStepDetailsOutputReference", version: "10.0.12" };
function transferWorkflowStepsDeleteStepDetailsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        source_file_location: cdktf.stringToTerraform(struct.sourceFileLocation),
    };
}
exports.transferWorkflowStepsDeleteStepDetailsToTerraform = transferWorkflowStepsDeleteStepDetailsToTerraform;
class TransferWorkflowStepsDeleteStepDetailsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._sourceFileLocation !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceFileLocation = this._sourceFileLocation;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._name = undefined;
            this._sourceFileLocation = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._name = value.name;
            this._sourceFileLocation = value.sourceFileLocation;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get sourceFileLocation() {
        return this.getStringAttribute('source_file_location');
    }
    set sourceFileLocation(value) {
        this._sourceFileLocation = value;
    }
    resetSourceFileLocation() {
        this._sourceFileLocation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceFileLocationInput() {
        return this._sourceFileLocation;
    }
}
exports.TransferWorkflowStepsDeleteStepDetailsOutputReference = TransferWorkflowStepsDeleteStepDetailsOutputReference;
_s = JSII_RTTI_SYMBOL_1;
TransferWorkflowStepsDeleteStepDetailsOutputReference[_s] = { fqn: "@cdktf/provider-aws.transferWorkflow.TransferWorkflowStepsDeleteStepDetailsOutputReference", version: "10.0.12" };
function transferWorkflowStepsTagStepDetailsTagsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.transferWorkflowStepsTagStepDetailsTagsToTerraform = transferWorkflowStepsTagStepDetailsTagsToTerraform;
class TransferWorkflowStepsTagStepDetailsTagsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._key = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._key = value.key;
            this._value = value.value;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.TransferWorkflowStepsTagStepDetailsTagsOutputReference = TransferWorkflowStepsTagStepDetailsTagsOutputReference;
_t = JSII_RTTI_SYMBOL_1;
TransferWorkflowStepsTagStepDetailsTagsOutputReference[_t] = { fqn: "@cdktf/provider-aws.transferWorkflow.TransferWorkflowStepsTagStepDetailsTagsOutputReference", version: "10.0.12" };
class TransferWorkflowStepsTagStepDetailsTagsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new TransferWorkflowStepsTagStepDetailsTagsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.TransferWorkflowStepsTagStepDetailsTagsList = TransferWorkflowStepsTagStepDetailsTagsList;
_u = JSII_RTTI_SYMBOL_1;
TransferWorkflowStepsTagStepDetailsTagsList[_u] = { fqn: "@cdktf/provider-aws.transferWorkflow.TransferWorkflowStepsTagStepDetailsTagsList", version: "10.0.12" };
function transferWorkflowStepsTagStepDetailsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        source_file_location: cdktf.stringToTerraform(struct.sourceFileLocation),
        tags: cdktf.listMapper(transferWorkflowStepsTagStepDetailsTagsToTerraform, true)(struct.tags),
    };
}
exports.transferWorkflowStepsTagStepDetailsToTerraform = transferWorkflowStepsTagStepDetailsToTerraform;
class TransferWorkflowStepsTagStepDetailsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // tags - computed: false, optional: true, required: false
        this._tags = new TransferWorkflowStepsTagStepDetailsTagsList(this, "tags", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._sourceFileLocation !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceFileLocation = this._sourceFileLocation;
        }
        if (this._tags?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.tags = this._tags?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._name = undefined;
            this._sourceFileLocation = undefined;
            this._tags.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._name = value.name;
            this._sourceFileLocation = value.sourceFileLocation;
            this._tags.internalValue = value.tags;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get sourceFileLocation() {
        return this.getStringAttribute('source_file_location');
    }
    set sourceFileLocation(value) {
        this._sourceFileLocation = value;
    }
    resetSourceFileLocation() {
        this._sourceFileLocation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceFileLocationInput() {
        return this._sourceFileLocation;
    }
    get tags() {
        return this._tags;
    }
    putTags(value) {
        this._tags.internalValue = value;
    }
    resetTags() {
        this._tags.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags.internalValue;
    }
}
exports.TransferWorkflowStepsTagStepDetailsOutputReference = TransferWorkflowStepsTagStepDetailsOutputReference;
_v = JSII_RTTI_SYMBOL_1;
TransferWorkflowStepsTagStepDetailsOutputReference[_v] = { fqn: "@cdktf/provider-aws.transferWorkflow.TransferWorkflowStepsTagStepDetailsOutputReference", version: "10.0.12" };
function transferWorkflowStepsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
        copy_step_details: transferWorkflowStepsCopyStepDetailsToTerraform(struct.copyStepDetails),
        custom_step_details: transferWorkflowStepsCustomStepDetailsToTerraform(struct.customStepDetails),
        delete_step_details: transferWorkflowStepsDeleteStepDetailsToTerraform(struct.deleteStepDetails),
        tag_step_details: transferWorkflowStepsTagStepDetailsToTerraform(struct.tagStepDetails),
    };
}
exports.transferWorkflowStepsToTerraform = transferWorkflowStepsToTerraform;
class TransferWorkflowStepsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // copy_step_details - computed: false, optional: true, required: false
        this._copyStepDetails = new TransferWorkflowStepsCopyStepDetailsOutputReference(this, "copy_step_details");
        // custom_step_details - computed: false, optional: true, required: false
        this._customStepDetails = new TransferWorkflowStepsCustomStepDetailsOutputReference(this, "custom_step_details");
        // delete_step_details - computed: false, optional: true, required: false
        this._deleteStepDetails = new TransferWorkflowStepsDeleteStepDetailsOutputReference(this, "delete_step_details");
        // tag_step_details - computed: false, optional: true, required: false
        this._tagStepDetails = new TransferWorkflowStepsTagStepDetailsOutputReference(this, "tag_step_details");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._copyStepDetails?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.copyStepDetails = this._copyStepDetails?.internalValue;
        }
        if (this._customStepDetails?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.customStepDetails = this._customStepDetails?.internalValue;
        }
        if (this._deleteStepDetails?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.deleteStepDetails = this._deleteStepDetails?.internalValue;
        }
        if (this._tagStepDetails?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.tagStepDetails = this._tagStepDetails?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._type = undefined;
            this._copyStepDetails.internalValue = undefined;
            this._customStepDetails.internalValue = undefined;
            this._deleteStepDetails.internalValue = undefined;
            this._tagStepDetails.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._type = value.type;
            this._copyStepDetails.internalValue = value.copyStepDetails;
            this._customStepDetails.internalValue = value.customStepDetails;
            this._deleteStepDetails.internalValue = value.deleteStepDetails;
            this._tagStepDetails.internalValue = value.tagStepDetails;
        }
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get copyStepDetails() {
        return this._copyStepDetails;
    }
    putCopyStepDetails(value) {
        this._copyStepDetails.internalValue = value;
    }
    resetCopyStepDetails() {
        this._copyStepDetails.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get copyStepDetailsInput() {
        return this._copyStepDetails.internalValue;
    }
    get customStepDetails() {
        return this._customStepDetails;
    }
    putCustomStepDetails(value) {
        this._customStepDetails.internalValue = value;
    }
    resetCustomStepDetails() {
        this._customStepDetails.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get customStepDetailsInput() {
        return this._customStepDetails.internalValue;
    }
    get deleteStepDetails() {
        return this._deleteStepDetails;
    }
    putDeleteStepDetails(value) {
        this._deleteStepDetails.internalValue = value;
    }
    resetDeleteStepDetails() {
        this._deleteStepDetails.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteStepDetailsInput() {
        return this._deleteStepDetails.internalValue;
    }
    get tagStepDetails() {
        return this._tagStepDetails;
    }
    putTagStepDetails(value) {
        this._tagStepDetails.internalValue = value;
    }
    resetTagStepDetails() {
        this._tagStepDetails.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagStepDetailsInput() {
        return this._tagStepDetails.internalValue;
    }
}
exports.TransferWorkflowStepsOutputReference = TransferWorkflowStepsOutputReference;
_w = JSII_RTTI_SYMBOL_1;
TransferWorkflowStepsOutputReference[_w] = { fqn: "@cdktf/provider-aws.transferWorkflow.TransferWorkflowStepsOutputReference", version: "10.0.12" };
class TransferWorkflowStepsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new TransferWorkflowStepsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.TransferWorkflowStepsList = TransferWorkflowStepsList;
_x = JSII_RTTI_SYMBOL_1;
TransferWorkflowStepsList[_x] = { fqn: "@cdktf/provider-aws.transferWorkflow.TransferWorkflowStepsList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/transfer_workflow aws_transfer_workflow}
*/
class TransferWorkflow extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/transfer_workflow aws_transfer_workflow} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options TransferWorkflowConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_transfer_workflow',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // on_exception_steps - computed: false, optional: true, required: false
        this._onExceptionSteps = new TransferWorkflowOnExceptionStepsList(this, "on_exception_steps", false);
        // steps - computed: false, optional: false, required: true
        this._steps = new TransferWorkflowStepsList(this, "steps", false);
        this._description = config.description;
        this._id = config.id;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._onExceptionSteps.internalValue = config.onExceptionSteps;
        this._steps.internalValue = config.steps;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get onExceptionSteps() {
        return this._onExceptionSteps;
    }
    putOnExceptionSteps(value) {
        this._onExceptionSteps.internalValue = value;
    }
    resetOnExceptionSteps() {
        this._onExceptionSteps.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get onExceptionStepsInput() {
        return this._onExceptionSteps.internalValue;
    }
    get steps() {
        return this._steps;
    }
    putSteps(value) {
        this._steps.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get stepsInput() {
        return this._steps.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            on_exception_steps: cdktf.listMapper(transferWorkflowOnExceptionStepsToTerraform, true)(this._onExceptionSteps.internalValue),
            steps: cdktf.listMapper(transferWorkflowStepsToTerraform, true)(this._steps.internalValue),
        };
    }
}
exports.TransferWorkflow = TransferWorkflow;
_y = JSII_RTTI_SYMBOL_1;
TransferWorkflow[_y] = { fqn: "@cdktf/provider-aws.transferWorkflow.TransferWorkflow", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
TransferWorkflow.tfResourceType = "aws_transfer_workflow";
