// @cdktf/provider-aws/lib/data-aws-connect-queue/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsConnectQueue = exports.DataAwsConnectQueueOutboundCallerConfigList = exports.DataAwsConnectQueueOutboundCallerConfigOutputReference = exports.dataAwsConnectQueueOutboundCallerConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsConnectQueueOutboundCallerConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsConnectQueueOutboundCallerConfigToTerraform = dataAwsConnectQueueOutboundCallerConfigToTerraform;
class DataAwsConnectQueueOutboundCallerConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // outbound_caller_id_name - computed: true, optional: false, required: false
    get outboundCallerIdName() {
        return this.getStringAttribute('outbound_caller_id_name');
    }
    // outbound_caller_id_number_id - computed: true, optional: false, required: false
    get outboundCallerIdNumberId() {
        return this.getStringAttribute('outbound_caller_id_number_id');
    }
    // outbound_flow_id - computed: true, optional: false, required: false
    get outboundFlowId() {
        return this.getStringAttribute('outbound_flow_id');
    }
}
exports.DataAwsConnectQueueOutboundCallerConfigOutputReference = DataAwsConnectQueueOutboundCallerConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsConnectQueueOutboundCallerConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsConnectQueue.DataAwsConnectQueueOutboundCallerConfigOutputReference", version: "10.0.12" };
class DataAwsConnectQueueOutboundCallerConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsConnectQueueOutboundCallerConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsConnectQueueOutboundCallerConfigList = DataAwsConnectQueueOutboundCallerConfigList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsConnectQueueOutboundCallerConfigList[_b] = { fqn: "@cdktf/provider-aws.dataAwsConnectQueue.DataAwsConnectQueueOutboundCallerConfigList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/connect_queue aws_connect_queue}
*/
class DataAwsConnectQueue extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/connect_queue aws_connect_queue} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsConnectQueueConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_connect_queue',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // outbound_caller_config - computed: true, optional: false, required: false
        this._outboundCallerConfig = new DataAwsConnectQueueOutboundCallerConfigList(this, "outbound_caller_config", false);
        this._id = config.id;
        this._instanceId = config.instanceId;
        this._name = config.name;
        this._queueId = config.queueId;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // hours_of_operation_id - computed: true, optional: false, required: false
    get hoursOfOperationId() {
        return this.getStringAttribute('hours_of_operation_id');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instanceId() {
        return this.getStringAttribute('instance_id');
    }
    set instanceId(value) {
        this._instanceId = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceIdInput() {
        return this._instanceId;
    }
    // max_contacts - computed: true, optional: false, required: false
    get maxContacts() {
        return this.getNumberAttribute('max_contacts');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get outboundCallerConfig() {
        return this._outboundCallerConfig;
    }
    get queueId() {
        return this.getStringAttribute('queue_id');
    }
    set queueId(value) {
        this._queueId = value;
    }
    resetQueueId() {
        this._queueId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get queueIdInput() {
        return this._queueId;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            instance_id: cdktf.stringToTerraform(this._instanceId),
            name: cdktf.stringToTerraform(this._name),
            queue_id: cdktf.stringToTerraform(this._queueId),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
}
exports.DataAwsConnectQueue = DataAwsConnectQueue;
_c = JSII_RTTI_SYMBOL_1;
DataAwsConnectQueue[_c] = { fqn: "@cdktf/provider-aws.dataAwsConnectQueue.DataAwsConnectQueue", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsConnectQueue.tfResourceType = "aws_connect_queue";
