// @cdktf/provider-aws/lib/ssm-parameter/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SsmParameter = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/ssm_parameter aws_ssm_parameter}
*/
class SsmParameter extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ssm_parameter aws_ssm_parameter} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SsmParameterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ssm_parameter',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._allowedPattern = config.allowedPattern;
        this._arn = config.arn;
        this._dataType = config.dataType;
        this._description = config.description;
        this._id = config.id;
        this._insecureValue = config.insecureValue;
        this._keyId = config.keyId;
        this._name = config.name;
        this._overwrite = config.overwrite;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._tier = config.tier;
        this._type = config.type;
        this._value = config.value;
    }
    get allowedPattern() {
        return this.getStringAttribute('allowed_pattern');
    }
    set allowedPattern(value) {
        this._allowedPattern = value;
    }
    resetAllowedPattern() {
        this._allowedPattern = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allowedPatternInput() {
        return this._allowedPattern;
    }
    get arn() {
        return this.getStringAttribute('arn');
    }
    set arn(value) {
        this._arn = value;
    }
    resetArn() {
        this._arn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get arnInput() {
        return this._arn;
    }
    get dataType() {
        return this.getStringAttribute('data_type');
    }
    set dataType(value) {
        this._dataType = value;
    }
    resetDataType() {
        this._dataType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataTypeInput() {
        return this._dataType;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get insecureValue() {
        return this.getStringAttribute('insecure_value');
    }
    set insecureValue(value) {
        this._insecureValue = value;
    }
    resetInsecureValue() {
        this._insecureValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get insecureValueInput() {
        return this._insecureValue;
    }
    get keyId() {
        return this.getStringAttribute('key_id');
    }
    set keyId(value) {
        this._keyId = value;
    }
    resetKeyId() {
        this._keyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyIdInput() {
        return this._keyId;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get overwrite() {
        return this.getBooleanAttribute('overwrite');
    }
    set overwrite(value) {
        this._overwrite = value;
    }
    resetOverwrite() {
        this._overwrite = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get overwriteInput() {
        return this._overwrite;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get tier() {
        return this.getStringAttribute('tier');
    }
    set tier(value) {
        this._tier = value;
    }
    resetTier() {
        this._tier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tierInput() {
        return this._tier;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getNumberAttribute('version');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            allowed_pattern: cdktf.stringToTerraform(this._allowedPattern),
            arn: cdktf.stringToTerraform(this._arn),
            data_type: cdktf.stringToTerraform(this._dataType),
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            insecure_value: cdktf.stringToTerraform(this._insecureValue),
            key_id: cdktf.stringToTerraform(this._keyId),
            name: cdktf.stringToTerraform(this._name),
            overwrite: cdktf.booleanToTerraform(this._overwrite),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            tier: cdktf.stringToTerraform(this._tier),
            type: cdktf.stringToTerraform(this._type),
            value: cdktf.stringToTerraform(this._value),
        };
    }
}
exports.SsmParameter = SsmParameter;
_a = JSII_RTTI_SYMBOL_1;
SsmParameter[_a] = { fqn: "@cdktf/provider-aws.ssmParameter.SsmParameter", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
SsmParameter.tfResourceType = "aws_ssm_parameter";
