// @cdktf/provider-aws/lib/codebuild-report-group/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodebuildReportGroup = exports.CodebuildReportGroupExportConfigOutputReference = exports.codebuildReportGroupExportConfigToTerraform = exports.CodebuildReportGroupExportConfigS3DestinationOutputReference = exports.codebuildReportGroupExportConfigS3DestinationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function codebuildReportGroupExportConfigS3DestinationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket: cdktf.stringToTerraform(struct.bucket),
        encryption_disabled: cdktf.booleanToTerraform(struct.encryptionDisabled),
        encryption_key: cdktf.stringToTerraform(struct.encryptionKey),
        packaging: cdktf.stringToTerraform(struct.packaging),
        path: cdktf.stringToTerraform(struct.path),
    };
}
exports.codebuildReportGroupExportConfigS3DestinationToTerraform = codebuildReportGroupExportConfigS3DestinationToTerraform;
class CodebuildReportGroupExportConfigS3DestinationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucket !== undefined) {
            hasAnyValues = true;
            internalValueResult.bucket = this._bucket;
        }
        if (this._encryptionDisabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.encryptionDisabled = this._encryptionDisabled;
        }
        if (this._encryptionKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.encryptionKey = this._encryptionKey;
        }
        if (this._packaging !== undefined) {
            hasAnyValues = true;
            internalValueResult.packaging = this._packaging;
        }
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucket = undefined;
            this._encryptionDisabled = undefined;
            this._encryptionKey = undefined;
            this._packaging = undefined;
            this._path = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucket = value.bucket;
            this._encryptionDisabled = value.encryptionDisabled;
            this._encryptionKey = value.encryptionKey;
            this._packaging = value.packaging;
            this._path = value.path;
        }
    }
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    set bucket(value) {
        this._bucket = value;
    }
    // Temporarily expose input value. Use with caution.
    get bucketInput() {
        return this._bucket;
    }
    get encryptionDisabled() {
        return this.getBooleanAttribute('encryption_disabled');
    }
    set encryptionDisabled(value) {
        this._encryptionDisabled = value;
    }
    resetEncryptionDisabled() {
        this._encryptionDisabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encryptionDisabledInput() {
        return this._encryptionDisabled;
    }
    get encryptionKey() {
        return this.getStringAttribute('encryption_key');
    }
    set encryptionKey(value) {
        this._encryptionKey = value;
    }
    // Temporarily expose input value. Use with caution.
    get encryptionKeyInput() {
        return this._encryptionKey;
    }
    get packaging() {
        return this.getStringAttribute('packaging');
    }
    set packaging(value) {
        this._packaging = value;
    }
    resetPackaging() {
        this._packaging = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get packagingInput() {
        return this._packaging;
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    resetPath() {
        this._path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
}
exports.CodebuildReportGroupExportConfigS3DestinationOutputReference = CodebuildReportGroupExportConfigS3DestinationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CodebuildReportGroupExportConfigS3DestinationOutputReference[_a] = { fqn: "@cdktf/provider-aws.codebuildReportGroup.CodebuildReportGroupExportConfigS3DestinationOutputReference", version: "10.0.12" };
function codebuildReportGroupExportConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
        s3_destination: codebuildReportGroupExportConfigS3DestinationToTerraform(struct.s3Destination),
    };
}
exports.codebuildReportGroupExportConfigToTerraform = codebuildReportGroupExportConfigToTerraform;
class CodebuildReportGroupExportConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // s3_destination - computed: false, optional: true, required: false
        this._s3Destination = new CodebuildReportGroupExportConfigS3DestinationOutputReference(this, "s3_destination");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._s3Destination?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3Destination = this._s3Destination?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._type = undefined;
            this._s3Destination.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._type = value.type;
            this._s3Destination.internalValue = value.s3Destination;
        }
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get s3Destination() {
        return this._s3Destination;
    }
    putS3Destination(value) {
        this._s3Destination.internalValue = value;
    }
    resetS3Destination() {
        this._s3Destination.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get s3DestinationInput() {
        return this._s3Destination.internalValue;
    }
}
exports.CodebuildReportGroupExportConfigOutputReference = CodebuildReportGroupExportConfigOutputReference;
_b = JSII_RTTI_SYMBOL_1;
CodebuildReportGroupExportConfigOutputReference[_b] = { fqn: "@cdktf/provider-aws.codebuildReportGroup.CodebuildReportGroupExportConfigOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/codebuild_report_group aws_codebuild_report_group}
*/
class CodebuildReportGroup extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/codebuild_report_group aws_codebuild_report_group} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CodebuildReportGroupConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_codebuild_report_group',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // export_config - computed: false, optional: false, required: true
        this._exportConfig = new CodebuildReportGroupExportConfigOutputReference(this, "export_config");
        this._deleteReports = config.deleteReports;
        this._id = config.id;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._type = config.type;
        this._exportConfig.internalValue = config.exportConfig;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // created - computed: true, optional: false, required: false
    get created() {
        return this.getStringAttribute('created');
    }
    get deleteReports() {
        return this.getBooleanAttribute('delete_reports');
    }
    set deleteReports(value) {
        this._deleteReports = value;
    }
    resetDeleteReports() {
        this._deleteReports = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteReportsInput() {
        return this._deleteReports;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get exportConfig() {
        return this._exportConfig;
    }
    putExportConfig(value) {
        this._exportConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get exportConfigInput() {
        return this._exportConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            delete_reports: cdktf.booleanToTerraform(this._deleteReports),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            type: cdktf.stringToTerraform(this._type),
            export_config: codebuildReportGroupExportConfigToTerraform(this._exportConfig.internalValue),
        };
    }
}
exports.CodebuildReportGroup = CodebuildReportGroup;
_c = JSII_RTTI_SYMBOL_1;
CodebuildReportGroup[_c] = { fqn: "@cdktf/provider-aws.codebuildReportGroup.CodebuildReportGroup", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
CodebuildReportGroup.tfResourceType = "aws_codebuild_report_group";
