// @cdktf/provider-aws/lib/appstream-fleet-stack-association/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppstreamFleetStackAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/appstream_fleet_stack_association aws_appstream_fleet_stack_association}
*/
class AppstreamFleetStackAssociation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/appstream_fleet_stack_association aws_appstream_fleet_stack_association} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options AppstreamFleetStackAssociationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_appstream_fleet_stack_association',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._fleetName = config.fleetName;
        this._id = config.id;
        this._stackName = config.stackName;
    }
    get fleetName() {
        return this.getStringAttribute('fleet_name');
    }
    set fleetName(value) {
        this._fleetName = value;
    }
    // Temporarily expose input value. Use with caution.
    get fleetNameInput() {
        return this._fleetName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get stackName() {
        return this.getStringAttribute('stack_name');
    }
    set stackName(value) {
        this._stackName = value;
    }
    // Temporarily expose input value. Use with caution.
    get stackNameInput() {
        return this._stackName;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            fleet_name: cdktf.stringToTerraform(this._fleetName),
            id: cdktf.stringToTerraform(this._id),
            stack_name: cdktf.stringToTerraform(this._stackName),
        };
    }
}
exports.AppstreamFleetStackAssociation = AppstreamFleetStackAssociation;
_a = JSII_RTTI_SYMBOL_1;
AppstreamFleetStackAssociation[_a] = { fqn: "@cdktf/provider-aws.appstreamFleetStackAssociation.AppstreamFleetStackAssociation", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
AppstreamFleetStackAssociation.tfResourceType = "aws_appstream_fleet_stack_association";
