// @cdktf/provider-aws/lib/mskconnect-connector/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MskconnectConnector = exports.MskconnectConnectorWorkerConfigurationOutputReference = exports.mskconnectConnectorWorkerConfigurationToTerraform = exports.MskconnectConnectorTimeoutsOutputReference = exports.mskconnectConnectorTimeoutsToTerraform = exports.MskconnectConnectorPluginList = exports.MskconnectConnectorPluginOutputReference = exports.mskconnectConnectorPluginToTerraform = exports.MskconnectConnectorPluginCustomPluginOutputReference = exports.mskconnectConnectorPluginCustomPluginToTerraform = exports.MskconnectConnectorLogDeliveryOutputReference = exports.mskconnectConnectorLogDeliveryToTerraform = exports.MskconnectConnectorLogDeliveryWorkerLogDeliveryOutputReference = exports.mskconnectConnectorLogDeliveryWorkerLogDeliveryToTerraform = exports.MskconnectConnectorLogDeliveryWorkerLogDeliveryS3OutputReference = exports.mskconnectConnectorLogDeliveryWorkerLogDeliveryS3ToTerraform = exports.MskconnectConnectorLogDeliveryWorkerLogDeliveryFirehoseOutputReference = exports.mskconnectConnectorLogDeliveryWorkerLogDeliveryFirehoseToTerraform = exports.MskconnectConnectorLogDeliveryWorkerLogDeliveryCloudwatchLogsOutputReference = exports.mskconnectConnectorLogDeliveryWorkerLogDeliveryCloudwatchLogsToTerraform = exports.MskconnectConnectorKafkaClusterEncryptionInTransitOutputReference = exports.mskconnectConnectorKafkaClusterEncryptionInTransitToTerraform = exports.MskconnectConnectorKafkaClusterClientAuthenticationOutputReference = exports.mskconnectConnectorKafkaClusterClientAuthenticationToTerraform = exports.MskconnectConnectorKafkaClusterOutputReference = exports.mskconnectConnectorKafkaClusterToTerraform = exports.MskconnectConnectorKafkaClusterApacheKafkaClusterOutputReference = exports.mskconnectConnectorKafkaClusterApacheKafkaClusterToTerraform = exports.MskconnectConnectorKafkaClusterApacheKafkaClusterVpcOutputReference = exports.mskconnectConnectorKafkaClusterApacheKafkaClusterVpcToTerraform = exports.MskconnectConnectorCapacityOutputReference = exports.mskconnectConnectorCapacityToTerraform = exports.MskconnectConnectorCapacityProvisionedCapacityOutputReference = exports.mskconnectConnectorCapacityProvisionedCapacityToTerraform = exports.MskconnectConnectorCapacityAutoscalingOutputReference = exports.mskconnectConnectorCapacityAutoscalingToTerraform = exports.MskconnectConnectorCapacityAutoscalingScaleOutPolicyOutputReference = exports.mskconnectConnectorCapacityAutoscalingScaleOutPolicyToTerraform = exports.MskconnectConnectorCapacityAutoscalingScaleInPolicyOutputReference = exports.mskconnectConnectorCapacityAutoscalingScaleInPolicyToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function mskconnectConnectorCapacityAutoscalingScaleInPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cpu_utilization_percentage: cdktf.numberToTerraform(struct.cpuUtilizationPercentage),
    };
}
exports.mskconnectConnectorCapacityAutoscalingScaleInPolicyToTerraform = mskconnectConnectorCapacityAutoscalingScaleInPolicyToTerraform;
class MskconnectConnectorCapacityAutoscalingScaleInPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cpuUtilizationPercentage !== undefined) {
            hasAnyValues = true;
            internalValueResult.cpuUtilizationPercentage = this._cpuUtilizationPercentage;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cpuUtilizationPercentage = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cpuUtilizationPercentage = value.cpuUtilizationPercentage;
        }
    }
    get cpuUtilizationPercentage() {
        return this.getNumberAttribute('cpu_utilization_percentage');
    }
    set cpuUtilizationPercentage(value) {
        this._cpuUtilizationPercentage = value;
    }
    resetCpuUtilizationPercentage() {
        this._cpuUtilizationPercentage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cpuUtilizationPercentageInput() {
        return this._cpuUtilizationPercentage;
    }
}
exports.MskconnectConnectorCapacityAutoscalingScaleInPolicyOutputReference = MskconnectConnectorCapacityAutoscalingScaleInPolicyOutputReference;
_a = JSII_RTTI_SYMBOL_1;
MskconnectConnectorCapacityAutoscalingScaleInPolicyOutputReference[_a] = { fqn: "@cdktf/provider-aws.mskconnectConnector.MskconnectConnectorCapacityAutoscalingScaleInPolicyOutputReference", version: "10.0.12" };
function mskconnectConnectorCapacityAutoscalingScaleOutPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cpu_utilization_percentage: cdktf.numberToTerraform(struct.cpuUtilizationPercentage),
    };
}
exports.mskconnectConnectorCapacityAutoscalingScaleOutPolicyToTerraform = mskconnectConnectorCapacityAutoscalingScaleOutPolicyToTerraform;
class MskconnectConnectorCapacityAutoscalingScaleOutPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cpuUtilizationPercentage !== undefined) {
            hasAnyValues = true;
            internalValueResult.cpuUtilizationPercentage = this._cpuUtilizationPercentage;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cpuUtilizationPercentage = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cpuUtilizationPercentage = value.cpuUtilizationPercentage;
        }
    }
    get cpuUtilizationPercentage() {
        return this.getNumberAttribute('cpu_utilization_percentage');
    }
    set cpuUtilizationPercentage(value) {
        this._cpuUtilizationPercentage = value;
    }
    resetCpuUtilizationPercentage() {
        this._cpuUtilizationPercentage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cpuUtilizationPercentageInput() {
        return this._cpuUtilizationPercentage;
    }
}
exports.MskconnectConnectorCapacityAutoscalingScaleOutPolicyOutputReference = MskconnectConnectorCapacityAutoscalingScaleOutPolicyOutputReference;
_b = JSII_RTTI_SYMBOL_1;
MskconnectConnectorCapacityAutoscalingScaleOutPolicyOutputReference[_b] = { fqn: "@cdktf/provider-aws.mskconnectConnector.MskconnectConnectorCapacityAutoscalingScaleOutPolicyOutputReference", version: "10.0.12" };
function mskconnectConnectorCapacityAutoscalingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        max_worker_count: cdktf.numberToTerraform(struct.maxWorkerCount),
        mcu_count: cdktf.numberToTerraform(struct.mcuCount),
        min_worker_count: cdktf.numberToTerraform(struct.minWorkerCount),
        scale_in_policy: mskconnectConnectorCapacityAutoscalingScaleInPolicyToTerraform(struct.scaleInPolicy),
        scale_out_policy: mskconnectConnectorCapacityAutoscalingScaleOutPolicyToTerraform(struct.scaleOutPolicy),
    };
}
exports.mskconnectConnectorCapacityAutoscalingToTerraform = mskconnectConnectorCapacityAutoscalingToTerraform;
class MskconnectConnectorCapacityAutoscalingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // scale_in_policy - computed: false, optional: true, required: false
        this._scaleInPolicy = new MskconnectConnectorCapacityAutoscalingScaleInPolicyOutputReference(this, "scale_in_policy");
        // scale_out_policy - computed: false, optional: true, required: false
        this._scaleOutPolicy = new MskconnectConnectorCapacityAutoscalingScaleOutPolicyOutputReference(this, "scale_out_policy");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maxWorkerCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxWorkerCount = this._maxWorkerCount;
        }
        if (this._mcuCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.mcuCount = this._mcuCount;
        }
        if (this._minWorkerCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.minWorkerCount = this._minWorkerCount;
        }
        if (this._scaleInPolicy?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.scaleInPolicy = this._scaleInPolicy?.internalValue;
        }
        if (this._scaleOutPolicy?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.scaleOutPolicy = this._scaleOutPolicy?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maxWorkerCount = undefined;
            this._mcuCount = undefined;
            this._minWorkerCount = undefined;
            this._scaleInPolicy.internalValue = undefined;
            this._scaleOutPolicy.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maxWorkerCount = value.maxWorkerCount;
            this._mcuCount = value.mcuCount;
            this._minWorkerCount = value.minWorkerCount;
            this._scaleInPolicy.internalValue = value.scaleInPolicy;
            this._scaleOutPolicy.internalValue = value.scaleOutPolicy;
        }
    }
    get maxWorkerCount() {
        return this.getNumberAttribute('max_worker_count');
    }
    set maxWorkerCount(value) {
        this._maxWorkerCount = value;
    }
    // Temporarily expose input value. Use with caution.
    get maxWorkerCountInput() {
        return this._maxWorkerCount;
    }
    get mcuCount() {
        return this.getNumberAttribute('mcu_count');
    }
    set mcuCount(value) {
        this._mcuCount = value;
    }
    resetMcuCount() {
        this._mcuCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get mcuCountInput() {
        return this._mcuCount;
    }
    get minWorkerCount() {
        return this.getNumberAttribute('min_worker_count');
    }
    set minWorkerCount(value) {
        this._minWorkerCount = value;
    }
    // Temporarily expose input value. Use with caution.
    get minWorkerCountInput() {
        return this._minWorkerCount;
    }
    get scaleInPolicy() {
        return this._scaleInPolicy;
    }
    putScaleInPolicy(value) {
        this._scaleInPolicy.internalValue = value;
    }
    resetScaleInPolicy() {
        this._scaleInPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scaleInPolicyInput() {
        return this._scaleInPolicy.internalValue;
    }
    get scaleOutPolicy() {
        return this._scaleOutPolicy;
    }
    putScaleOutPolicy(value) {
        this._scaleOutPolicy.internalValue = value;
    }
    resetScaleOutPolicy() {
        this._scaleOutPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scaleOutPolicyInput() {
        return this._scaleOutPolicy.internalValue;
    }
}
exports.MskconnectConnectorCapacityAutoscalingOutputReference = MskconnectConnectorCapacityAutoscalingOutputReference;
_c = JSII_RTTI_SYMBOL_1;
MskconnectConnectorCapacityAutoscalingOutputReference[_c] = { fqn: "@cdktf/provider-aws.mskconnectConnector.MskconnectConnectorCapacityAutoscalingOutputReference", version: "10.0.12" };
function mskconnectConnectorCapacityProvisionedCapacityToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        mcu_count: cdktf.numberToTerraform(struct.mcuCount),
        worker_count: cdktf.numberToTerraform(struct.workerCount),
    };
}
exports.mskconnectConnectorCapacityProvisionedCapacityToTerraform = mskconnectConnectorCapacityProvisionedCapacityToTerraform;
class MskconnectConnectorCapacityProvisionedCapacityOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._mcuCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.mcuCount = this._mcuCount;
        }
        if (this._workerCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.workerCount = this._workerCount;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._mcuCount = undefined;
            this._workerCount = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._mcuCount = value.mcuCount;
            this._workerCount = value.workerCount;
        }
    }
    get mcuCount() {
        return this.getNumberAttribute('mcu_count');
    }
    set mcuCount(value) {
        this._mcuCount = value;
    }
    resetMcuCount() {
        this._mcuCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get mcuCountInput() {
        return this._mcuCount;
    }
    get workerCount() {
        return this.getNumberAttribute('worker_count');
    }
    set workerCount(value) {
        this._workerCount = value;
    }
    // Temporarily expose input value. Use with caution.
    get workerCountInput() {
        return this._workerCount;
    }
}
exports.MskconnectConnectorCapacityProvisionedCapacityOutputReference = MskconnectConnectorCapacityProvisionedCapacityOutputReference;
_d = JSII_RTTI_SYMBOL_1;
MskconnectConnectorCapacityProvisionedCapacityOutputReference[_d] = { fqn: "@cdktf/provider-aws.mskconnectConnector.MskconnectConnectorCapacityProvisionedCapacityOutputReference", version: "10.0.12" };
function mskconnectConnectorCapacityToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        autoscaling: mskconnectConnectorCapacityAutoscalingToTerraform(struct.autoscaling),
        provisioned_capacity: mskconnectConnectorCapacityProvisionedCapacityToTerraform(struct.provisionedCapacity),
    };
}
exports.mskconnectConnectorCapacityToTerraform = mskconnectConnectorCapacityToTerraform;
class MskconnectConnectorCapacityOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // autoscaling - computed: false, optional: true, required: false
        this._autoscaling = new MskconnectConnectorCapacityAutoscalingOutputReference(this, "autoscaling");
        // provisioned_capacity - computed: false, optional: true, required: false
        this._provisionedCapacity = new MskconnectConnectorCapacityProvisionedCapacityOutputReference(this, "provisioned_capacity");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._autoscaling?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.autoscaling = this._autoscaling?.internalValue;
        }
        if (this._provisionedCapacity?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.provisionedCapacity = this._provisionedCapacity?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._autoscaling.internalValue = undefined;
            this._provisionedCapacity.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._autoscaling.internalValue = value.autoscaling;
            this._provisionedCapacity.internalValue = value.provisionedCapacity;
        }
    }
    get autoscaling() {
        return this._autoscaling;
    }
    putAutoscaling(value) {
        this._autoscaling.internalValue = value;
    }
    resetAutoscaling() {
        this._autoscaling.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoscalingInput() {
        return this._autoscaling.internalValue;
    }
    get provisionedCapacity() {
        return this._provisionedCapacity;
    }
    putProvisionedCapacity(value) {
        this._provisionedCapacity.internalValue = value;
    }
    resetProvisionedCapacity() {
        this._provisionedCapacity.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get provisionedCapacityInput() {
        return this._provisionedCapacity.internalValue;
    }
}
exports.MskconnectConnectorCapacityOutputReference = MskconnectConnectorCapacityOutputReference;
_e = JSII_RTTI_SYMBOL_1;
MskconnectConnectorCapacityOutputReference[_e] = { fqn: "@cdktf/provider-aws.mskconnectConnector.MskconnectConnectorCapacityOutputReference", version: "10.0.12" };
function mskconnectConnectorKafkaClusterApacheKafkaClusterVpcToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        security_groups: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.securityGroups),
        subnets: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.subnets),
    };
}
exports.mskconnectConnectorKafkaClusterApacheKafkaClusterVpcToTerraform = mskconnectConnectorKafkaClusterApacheKafkaClusterVpcToTerraform;
class MskconnectConnectorKafkaClusterApacheKafkaClusterVpcOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._securityGroups !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroups = this._securityGroups;
        }
        if (this._subnets !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnets = this._subnets;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._securityGroups = undefined;
            this._subnets = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._securityGroups = value.securityGroups;
            this._subnets = value.subnets;
        }
    }
    get securityGroups() {
        return cdktf.Fn.tolist(this.getListAttribute('security_groups'));
    }
    set securityGroups(value) {
        this._securityGroups = value;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupsInput() {
        return this._securityGroups;
    }
    get subnets() {
        return cdktf.Fn.tolist(this.getListAttribute('subnets'));
    }
    set subnets(value) {
        this._subnets = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetsInput() {
        return this._subnets;
    }
}
exports.MskconnectConnectorKafkaClusterApacheKafkaClusterVpcOutputReference = MskconnectConnectorKafkaClusterApacheKafkaClusterVpcOutputReference;
_f = JSII_RTTI_SYMBOL_1;
MskconnectConnectorKafkaClusterApacheKafkaClusterVpcOutputReference[_f] = { fqn: "@cdktf/provider-aws.mskconnectConnector.MskconnectConnectorKafkaClusterApacheKafkaClusterVpcOutputReference", version: "10.0.12" };
function mskconnectConnectorKafkaClusterApacheKafkaClusterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bootstrap_servers: cdktf.stringToTerraform(struct.bootstrapServers),
        vpc: mskconnectConnectorKafkaClusterApacheKafkaClusterVpcToTerraform(struct.vpc),
    };
}
exports.mskconnectConnectorKafkaClusterApacheKafkaClusterToTerraform = mskconnectConnectorKafkaClusterApacheKafkaClusterToTerraform;
class MskconnectConnectorKafkaClusterApacheKafkaClusterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // vpc - computed: false, optional: false, required: true
        this._vpc = new MskconnectConnectorKafkaClusterApacheKafkaClusterVpcOutputReference(this, "vpc");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bootstrapServers !== undefined) {
            hasAnyValues = true;
            internalValueResult.bootstrapServers = this._bootstrapServers;
        }
        if (this._vpc?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.vpc = this._vpc?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bootstrapServers = undefined;
            this._vpc.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bootstrapServers = value.bootstrapServers;
            this._vpc.internalValue = value.vpc;
        }
    }
    get bootstrapServers() {
        return this.getStringAttribute('bootstrap_servers');
    }
    set bootstrapServers(value) {
        this._bootstrapServers = value;
    }
    // Temporarily expose input value. Use with caution.
    get bootstrapServersInput() {
        return this._bootstrapServers;
    }
    get vpc() {
        return this._vpc;
    }
    putVpc(value) {
        this._vpc.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get vpcInput() {
        return this._vpc.internalValue;
    }
}
exports.MskconnectConnectorKafkaClusterApacheKafkaClusterOutputReference = MskconnectConnectorKafkaClusterApacheKafkaClusterOutputReference;
_g = JSII_RTTI_SYMBOL_1;
MskconnectConnectorKafkaClusterApacheKafkaClusterOutputReference[_g] = { fqn: "@cdktf/provider-aws.mskconnectConnector.MskconnectConnectorKafkaClusterApacheKafkaClusterOutputReference", version: "10.0.12" };
function mskconnectConnectorKafkaClusterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        apache_kafka_cluster: mskconnectConnectorKafkaClusterApacheKafkaClusterToTerraform(struct.apacheKafkaCluster),
    };
}
exports.mskconnectConnectorKafkaClusterToTerraform = mskconnectConnectorKafkaClusterToTerraform;
class MskconnectConnectorKafkaClusterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // apache_kafka_cluster - computed: false, optional: false, required: true
        this._apacheKafkaCluster = new MskconnectConnectorKafkaClusterApacheKafkaClusterOutputReference(this, "apache_kafka_cluster");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._apacheKafkaCluster?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.apacheKafkaCluster = this._apacheKafkaCluster?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._apacheKafkaCluster.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._apacheKafkaCluster.internalValue = value.apacheKafkaCluster;
        }
    }
    get apacheKafkaCluster() {
        return this._apacheKafkaCluster;
    }
    putApacheKafkaCluster(value) {
        this._apacheKafkaCluster.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get apacheKafkaClusterInput() {
        return this._apacheKafkaCluster.internalValue;
    }
}
exports.MskconnectConnectorKafkaClusterOutputReference = MskconnectConnectorKafkaClusterOutputReference;
_h = JSII_RTTI_SYMBOL_1;
MskconnectConnectorKafkaClusterOutputReference[_h] = { fqn: "@cdktf/provider-aws.mskconnectConnector.MskconnectConnectorKafkaClusterOutputReference", version: "10.0.12" };
function mskconnectConnectorKafkaClusterClientAuthenticationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        authentication_type: cdktf.stringToTerraform(struct.authenticationType),
    };
}
exports.mskconnectConnectorKafkaClusterClientAuthenticationToTerraform = mskconnectConnectorKafkaClusterClientAuthenticationToTerraform;
class MskconnectConnectorKafkaClusterClientAuthenticationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._authenticationType !== undefined) {
            hasAnyValues = true;
            internalValueResult.authenticationType = this._authenticationType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._authenticationType = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._authenticationType = value.authenticationType;
        }
    }
    get authenticationType() {
        return this.getStringAttribute('authentication_type');
    }
    set authenticationType(value) {
        this._authenticationType = value;
    }
    resetAuthenticationType() {
        this._authenticationType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authenticationTypeInput() {
        return this._authenticationType;
    }
}
exports.MskconnectConnectorKafkaClusterClientAuthenticationOutputReference = MskconnectConnectorKafkaClusterClientAuthenticationOutputReference;
_j = JSII_RTTI_SYMBOL_1;
MskconnectConnectorKafkaClusterClientAuthenticationOutputReference[_j] = { fqn: "@cdktf/provider-aws.mskconnectConnector.MskconnectConnectorKafkaClusterClientAuthenticationOutputReference", version: "10.0.12" };
function mskconnectConnectorKafkaClusterEncryptionInTransitToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        encryption_type: cdktf.stringToTerraform(struct.encryptionType),
    };
}
exports.mskconnectConnectorKafkaClusterEncryptionInTransitToTerraform = mskconnectConnectorKafkaClusterEncryptionInTransitToTerraform;
class MskconnectConnectorKafkaClusterEncryptionInTransitOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._encryptionType !== undefined) {
            hasAnyValues = true;
            internalValueResult.encryptionType = this._encryptionType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._encryptionType = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._encryptionType = value.encryptionType;
        }
    }
    get encryptionType() {
        return this.getStringAttribute('encryption_type');
    }
    set encryptionType(value) {
        this._encryptionType = value;
    }
    resetEncryptionType() {
        this._encryptionType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encryptionTypeInput() {
        return this._encryptionType;
    }
}
exports.MskconnectConnectorKafkaClusterEncryptionInTransitOutputReference = MskconnectConnectorKafkaClusterEncryptionInTransitOutputReference;
_k = JSII_RTTI_SYMBOL_1;
MskconnectConnectorKafkaClusterEncryptionInTransitOutputReference[_k] = { fqn: "@cdktf/provider-aws.mskconnectConnector.MskconnectConnectorKafkaClusterEncryptionInTransitOutputReference", version: "10.0.12" };
function mskconnectConnectorLogDeliveryWorkerLogDeliveryCloudwatchLogsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        log_group: cdktf.stringToTerraform(struct.logGroup),
    };
}
exports.mskconnectConnectorLogDeliveryWorkerLogDeliveryCloudwatchLogsToTerraform = mskconnectConnectorLogDeliveryWorkerLogDeliveryCloudwatchLogsToTerraform;
class MskconnectConnectorLogDeliveryWorkerLogDeliveryCloudwatchLogsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._logGroup !== undefined) {
            hasAnyValues = true;
            internalValueResult.logGroup = this._logGroup;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._logGroup = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._logGroup = value.logGroup;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get logGroup() {
        return this.getStringAttribute('log_group');
    }
    set logGroup(value) {
        this._logGroup = value;
    }
    resetLogGroup() {
        this._logGroup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get logGroupInput() {
        return this._logGroup;
    }
}
exports.MskconnectConnectorLogDeliveryWorkerLogDeliveryCloudwatchLogsOutputReference = MskconnectConnectorLogDeliveryWorkerLogDeliveryCloudwatchLogsOutputReference;
_l = JSII_RTTI_SYMBOL_1;
MskconnectConnectorLogDeliveryWorkerLogDeliveryCloudwatchLogsOutputReference[_l] = { fqn: "@cdktf/provider-aws.mskconnectConnector.MskconnectConnectorLogDeliveryWorkerLogDeliveryCloudwatchLogsOutputReference", version: "10.0.12" };
function mskconnectConnectorLogDeliveryWorkerLogDeliveryFirehoseToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        delivery_stream: cdktf.stringToTerraform(struct.deliveryStream),
        enabled: cdktf.booleanToTerraform(struct.enabled),
    };
}
exports.mskconnectConnectorLogDeliveryWorkerLogDeliveryFirehoseToTerraform = mskconnectConnectorLogDeliveryWorkerLogDeliveryFirehoseToTerraform;
class MskconnectConnectorLogDeliveryWorkerLogDeliveryFirehoseOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._deliveryStream !== undefined) {
            hasAnyValues = true;
            internalValueResult.deliveryStream = this._deliveryStream;
        }
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._deliveryStream = undefined;
            this._enabled = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._deliveryStream = value.deliveryStream;
            this._enabled = value.enabled;
        }
    }
    get deliveryStream() {
        return this.getStringAttribute('delivery_stream');
    }
    set deliveryStream(value) {
        this._deliveryStream = value;
    }
    resetDeliveryStream() {
        this._deliveryStream = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deliveryStreamInput() {
        return this._deliveryStream;
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
}
exports.MskconnectConnectorLogDeliveryWorkerLogDeliveryFirehoseOutputReference = MskconnectConnectorLogDeliveryWorkerLogDeliveryFirehoseOutputReference;
_m = JSII_RTTI_SYMBOL_1;
MskconnectConnectorLogDeliveryWorkerLogDeliveryFirehoseOutputReference[_m] = { fqn: "@cdktf/provider-aws.mskconnectConnector.MskconnectConnectorLogDeliveryWorkerLogDeliveryFirehoseOutputReference", version: "10.0.12" };
function mskconnectConnectorLogDeliveryWorkerLogDeliveryS3ToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket: cdktf.stringToTerraform(struct.bucket),
        enabled: cdktf.booleanToTerraform(struct.enabled),
        prefix: cdktf.stringToTerraform(struct.prefix),
    };
}
exports.mskconnectConnectorLogDeliveryWorkerLogDeliveryS3ToTerraform = mskconnectConnectorLogDeliveryWorkerLogDeliveryS3ToTerraform;
class MskconnectConnectorLogDeliveryWorkerLogDeliveryS3OutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucket !== undefined) {
            hasAnyValues = true;
            internalValueResult.bucket = this._bucket;
        }
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._prefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucket = undefined;
            this._enabled = undefined;
            this._prefix = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucket = value.bucket;
            this._enabled = value.enabled;
            this._prefix = value.prefix;
        }
    }
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    set bucket(value) {
        this._bucket = value;
    }
    resetBucket() {
        this._bucket = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bucketInput() {
        return this._bucket;
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
}
exports.MskconnectConnectorLogDeliveryWorkerLogDeliveryS3OutputReference = MskconnectConnectorLogDeliveryWorkerLogDeliveryS3OutputReference;
_o = JSII_RTTI_SYMBOL_1;
MskconnectConnectorLogDeliveryWorkerLogDeliveryS3OutputReference[_o] = { fqn: "@cdktf/provider-aws.mskconnectConnector.MskconnectConnectorLogDeliveryWorkerLogDeliveryS3OutputReference", version: "10.0.12" };
function mskconnectConnectorLogDeliveryWorkerLogDeliveryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cloudwatch_logs: mskconnectConnectorLogDeliveryWorkerLogDeliveryCloudwatchLogsToTerraform(struct.cloudwatchLogs),
        firehose: mskconnectConnectorLogDeliveryWorkerLogDeliveryFirehoseToTerraform(struct.firehose),
        s3: mskconnectConnectorLogDeliveryWorkerLogDeliveryS3ToTerraform(struct.s3),
    };
}
exports.mskconnectConnectorLogDeliveryWorkerLogDeliveryToTerraform = mskconnectConnectorLogDeliveryWorkerLogDeliveryToTerraform;
class MskconnectConnectorLogDeliveryWorkerLogDeliveryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // cloudwatch_logs - computed: false, optional: true, required: false
        this._cloudwatchLogs = new MskconnectConnectorLogDeliveryWorkerLogDeliveryCloudwatchLogsOutputReference(this, "cloudwatch_logs");
        // firehose - computed: false, optional: true, required: false
        this._firehose = new MskconnectConnectorLogDeliveryWorkerLogDeliveryFirehoseOutputReference(this, "firehose");
        // s3 - computed: false, optional: true, required: false
        this._s3 = new MskconnectConnectorLogDeliveryWorkerLogDeliveryS3OutputReference(this, "s3");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cloudwatchLogs?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.cloudwatchLogs = this._cloudwatchLogs?.internalValue;
        }
        if (this._firehose?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.firehose = this._firehose?.internalValue;
        }
        if (this._s3?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3 = this._s3?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cloudwatchLogs.internalValue = undefined;
            this._firehose.internalValue = undefined;
            this._s3.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cloudwatchLogs.internalValue = value.cloudwatchLogs;
            this._firehose.internalValue = value.firehose;
            this._s3.internalValue = value.s3;
        }
    }
    get cloudwatchLogs() {
        return this._cloudwatchLogs;
    }
    putCloudwatchLogs(value) {
        this._cloudwatchLogs.internalValue = value;
    }
    resetCloudwatchLogs() {
        this._cloudwatchLogs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudwatchLogsInput() {
        return this._cloudwatchLogs.internalValue;
    }
    get firehose() {
        return this._firehose;
    }
    putFirehose(value) {
        this._firehose.internalValue = value;
    }
    resetFirehose() {
        this._firehose.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get firehoseInput() {
        return this._firehose.internalValue;
    }
    get s3() {
        return this._s3;
    }
    putS3(value) {
        this._s3.internalValue = value;
    }
    resetS3() {
        this._s3.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get s3Input() {
        return this._s3.internalValue;
    }
}
exports.MskconnectConnectorLogDeliveryWorkerLogDeliveryOutputReference = MskconnectConnectorLogDeliveryWorkerLogDeliveryOutputReference;
_p = JSII_RTTI_SYMBOL_1;
MskconnectConnectorLogDeliveryWorkerLogDeliveryOutputReference[_p] = { fqn: "@cdktf/provider-aws.mskconnectConnector.MskconnectConnectorLogDeliveryWorkerLogDeliveryOutputReference", version: "10.0.12" };
function mskconnectConnectorLogDeliveryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        worker_log_delivery: mskconnectConnectorLogDeliveryWorkerLogDeliveryToTerraform(struct.workerLogDelivery),
    };
}
exports.mskconnectConnectorLogDeliveryToTerraform = mskconnectConnectorLogDeliveryToTerraform;
class MskconnectConnectorLogDeliveryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // worker_log_delivery - computed: false, optional: false, required: true
        this._workerLogDelivery = new MskconnectConnectorLogDeliveryWorkerLogDeliveryOutputReference(this, "worker_log_delivery");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._workerLogDelivery?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.workerLogDelivery = this._workerLogDelivery?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._workerLogDelivery.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._workerLogDelivery.internalValue = value.workerLogDelivery;
        }
    }
    get workerLogDelivery() {
        return this._workerLogDelivery;
    }
    putWorkerLogDelivery(value) {
        this._workerLogDelivery.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get workerLogDeliveryInput() {
        return this._workerLogDelivery.internalValue;
    }
}
exports.MskconnectConnectorLogDeliveryOutputReference = MskconnectConnectorLogDeliveryOutputReference;
_q = JSII_RTTI_SYMBOL_1;
MskconnectConnectorLogDeliveryOutputReference[_q] = { fqn: "@cdktf/provider-aws.mskconnectConnector.MskconnectConnectorLogDeliveryOutputReference", version: "10.0.12" };
function mskconnectConnectorPluginCustomPluginToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        arn: cdktf.stringToTerraform(struct.arn),
        revision: cdktf.numberToTerraform(struct.revision),
    };
}
exports.mskconnectConnectorPluginCustomPluginToTerraform = mskconnectConnectorPluginCustomPluginToTerraform;
class MskconnectConnectorPluginCustomPluginOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._arn !== undefined) {
            hasAnyValues = true;
            internalValueResult.arn = this._arn;
        }
        if (this._revision !== undefined) {
            hasAnyValues = true;
            internalValueResult.revision = this._revision;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._arn = undefined;
            this._revision = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._arn = value.arn;
            this._revision = value.revision;
        }
    }
    get arn() {
        return this.getStringAttribute('arn');
    }
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    get arnInput() {
        return this._arn;
    }
    get revision() {
        return this.getNumberAttribute('revision');
    }
    set revision(value) {
        this._revision = value;
    }
    // Temporarily expose input value. Use with caution.
    get revisionInput() {
        return this._revision;
    }
}
exports.MskconnectConnectorPluginCustomPluginOutputReference = MskconnectConnectorPluginCustomPluginOutputReference;
_r = JSII_RTTI_SYMBOL_1;
MskconnectConnectorPluginCustomPluginOutputReference[_r] = { fqn: "@cdktf/provider-aws.mskconnectConnector.MskconnectConnectorPluginCustomPluginOutputReference", version: "10.0.12" };
function mskconnectConnectorPluginToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        custom_plugin: mskconnectConnectorPluginCustomPluginToTerraform(struct.customPlugin),
    };
}
exports.mskconnectConnectorPluginToTerraform = mskconnectConnectorPluginToTerraform;
class MskconnectConnectorPluginOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // custom_plugin - computed: false, optional: false, required: true
        this._customPlugin = new MskconnectConnectorPluginCustomPluginOutputReference(this, "custom_plugin");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._customPlugin?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.customPlugin = this._customPlugin?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._customPlugin.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._customPlugin.internalValue = value.customPlugin;
        }
    }
    get customPlugin() {
        return this._customPlugin;
    }
    putCustomPlugin(value) {
        this._customPlugin.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get customPluginInput() {
        return this._customPlugin.internalValue;
    }
}
exports.MskconnectConnectorPluginOutputReference = MskconnectConnectorPluginOutputReference;
_s = JSII_RTTI_SYMBOL_1;
MskconnectConnectorPluginOutputReference[_s] = { fqn: "@cdktf/provider-aws.mskconnectConnector.MskconnectConnectorPluginOutputReference", version: "10.0.12" };
class MskconnectConnectorPluginList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new MskconnectConnectorPluginOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.MskconnectConnectorPluginList = MskconnectConnectorPluginList;
_t = JSII_RTTI_SYMBOL_1;
MskconnectConnectorPluginList[_t] = { fqn: "@cdktf/provider-aws.mskconnectConnector.MskconnectConnectorPluginList", version: "10.0.12" };
function mskconnectConnectorTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.mskconnectConnectorTimeoutsToTerraform = mskconnectConnectorTimeoutsToTerraform;
class MskconnectConnectorTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.MskconnectConnectorTimeoutsOutputReference = MskconnectConnectorTimeoutsOutputReference;
_u = JSII_RTTI_SYMBOL_1;
MskconnectConnectorTimeoutsOutputReference[_u] = { fqn: "@cdktf/provider-aws.mskconnectConnector.MskconnectConnectorTimeoutsOutputReference", version: "10.0.12" };
function mskconnectConnectorWorkerConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        arn: cdktf.stringToTerraform(struct.arn),
        revision: cdktf.numberToTerraform(struct.revision),
    };
}
exports.mskconnectConnectorWorkerConfigurationToTerraform = mskconnectConnectorWorkerConfigurationToTerraform;
class MskconnectConnectorWorkerConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._arn !== undefined) {
            hasAnyValues = true;
            internalValueResult.arn = this._arn;
        }
        if (this._revision !== undefined) {
            hasAnyValues = true;
            internalValueResult.revision = this._revision;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._arn = undefined;
            this._revision = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._arn = value.arn;
            this._revision = value.revision;
        }
    }
    get arn() {
        return this.getStringAttribute('arn');
    }
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    get arnInput() {
        return this._arn;
    }
    get revision() {
        return this.getNumberAttribute('revision');
    }
    set revision(value) {
        this._revision = value;
    }
    // Temporarily expose input value. Use with caution.
    get revisionInput() {
        return this._revision;
    }
}
exports.MskconnectConnectorWorkerConfigurationOutputReference = MskconnectConnectorWorkerConfigurationOutputReference;
_v = JSII_RTTI_SYMBOL_1;
MskconnectConnectorWorkerConfigurationOutputReference[_v] = { fqn: "@cdktf/provider-aws.mskconnectConnector.MskconnectConnectorWorkerConfigurationOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/mskconnect_connector aws_mskconnect_connector}
*/
class MskconnectConnector extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/mskconnect_connector aws_mskconnect_connector} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options MskconnectConnectorConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_mskconnect_connector',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // capacity - computed: false, optional: false, required: true
        this._capacity = new MskconnectConnectorCapacityOutputReference(this, "capacity");
        // kafka_cluster - computed: false, optional: false, required: true
        this._kafkaCluster = new MskconnectConnectorKafkaClusterOutputReference(this, "kafka_cluster");
        // kafka_cluster_client_authentication - computed: false, optional: false, required: true
        this._kafkaClusterClientAuthentication = new MskconnectConnectorKafkaClusterClientAuthenticationOutputReference(this, "kafka_cluster_client_authentication");
        // kafka_cluster_encryption_in_transit - computed: false, optional: false, required: true
        this._kafkaClusterEncryptionInTransit = new MskconnectConnectorKafkaClusterEncryptionInTransitOutputReference(this, "kafka_cluster_encryption_in_transit");
        // log_delivery - computed: false, optional: true, required: false
        this._logDelivery = new MskconnectConnectorLogDeliveryOutputReference(this, "log_delivery");
        // plugin - computed: false, optional: false, required: true
        this._plugin = new MskconnectConnectorPluginList(this, "plugin", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new MskconnectConnectorTimeoutsOutputReference(this, "timeouts");
        // worker_configuration - computed: false, optional: true, required: false
        this._workerConfiguration = new MskconnectConnectorWorkerConfigurationOutputReference(this, "worker_configuration");
        this._connectorConfiguration = config.connectorConfiguration;
        this._description = config.description;
        this._id = config.id;
        this._kafkaconnectVersion = config.kafkaconnectVersion;
        this._name = config.name;
        this._serviceExecutionRoleArn = config.serviceExecutionRoleArn;
        this._capacity.internalValue = config.capacity;
        this._kafkaCluster.internalValue = config.kafkaCluster;
        this._kafkaClusterClientAuthentication.internalValue = config.kafkaClusterClientAuthentication;
        this._kafkaClusterEncryptionInTransit.internalValue = config.kafkaClusterEncryptionInTransit;
        this._logDelivery.internalValue = config.logDelivery;
        this._plugin.internalValue = config.plugin;
        this._timeouts.internalValue = config.timeouts;
        this._workerConfiguration.internalValue = config.workerConfiguration;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get connectorConfiguration() {
        return this.getStringMapAttribute('connector_configuration');
    }
    set connectorConfiguration(value) {
        this._connectorConfiguration = value;
    }
    // Temporarily expose input value. Use with caution.
    get connectorConfigurationInput() {
        return this._connectorConfiguration;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get kafkaconnectVersion() {
        return this.getStringAttribute('kafkaconnect_version');
    }
    set kafkaconnectVersion(value) {
        this._kafkaconnectVersion = value;
    }
    // Temporarily expose input value. Use with caution.
    get kafkaconnectVersionInput() {
        return this._kafkaconnectVersion;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get serviceExecutionRoleArn() {
        return this.getStringAttribute('service_execution_role_arn');
    }
    set serviceExecutionRoleArn(value) {
        this._serviceExecutionRoleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get serviceExecutionRoleArnInput() {
        return this._serviceExecutionRoleArn;
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
    get capacity() {
        return this._capacity;
    }
    putCapacity(value) {
        this._capacity.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get capacityInput() {
        return this._capacity.internalValue;
    }
    get kafkaCluster() {
        return this._kafkaCluster;
    }
    putKafkaCluster(value) {
        this._kafkaCluster.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get kafkaClusterInput() {
        return this._kafkaCluster.internalValue;
    }
    get kafkaClusterClientAuthentication() {
        return this._kafkaClusterClientAuthentication;
    }
    putKafkaClusterClientAuthentication(value) {
        this._kafkaClusterClientAuthentication.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get kafkaClusterClientAuthenticationInput() {
        return this._kafkaClusterClientAuthentication.internalValue;
    }
    get kafkaClusterEncryptionInTransit() {
        return this._kafkaClusterEncryptionInTransit;
    }
    putKafkaClusterEncryptionInTransit(value) {
        this._kafkaClusterEncryptionInTransit.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get kafkaClusterEncryptionInTransitInput() {
        return this._kafkaClusterEncryptionInTransit.internalValue;
    }
    get logDelivery() {
        return this._logDelivery;
    }
    putLogDelivery(value) {
        this._logDelivery.internalValue = value;
    }
    resetLogDelivery() {
        this._logDelivery.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get logDeliveryInput() {
        return this._logDelivery.internalValue;
    }
    get plugin() {
        return this._plugin;
    }
    putPlugin(value) {
        this._plugin.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get pluginInput() {
        return this._plugin.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get workerConfiguration() {
        return this._workerConfiguration;
    }
    putWorkerConfiguration(value) {
        this._workerConfiguration.internalValue = value;
    }
    resetWorkerConfiguration() {
        this._workerConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get workerConfigurationInput() {
        return this._workerConfiguration.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            connector_configuration: cdktf.hashMapper(cdktf.stringToTerraform)(this._connectorConfiguration),
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            kafkaconnect_version: cdktf.stringToTerraform(this._kafkaconnectVersion),
            name: cdktf.stringToTerraform(this._name),
            service_execution_role_arn: cdktf.stringToTerraform(this._serviceExecutionRoleArn),
            capacity: mskconnectConnectorCapacityToTerraform(this._capacity.internalValue),
            kafka_cluster: mskconnectConnectorKafkaClusterToTerraform(this._kafkaCluster.internalValue),
            kafka_cluster_client_authentication: mskconnectConnectorKafkaClusterClientAuthenticationToTerraform(this._kafkaClusterClientAuthentication.internalValue),
            kafka_cluster_encryption_in_transit: mskconnectConnectorKafkaClusterEncryptionInTransitToTerraform(this._kafkaClusterEncryptionInTransit.internalValue),
            log_delivery: mskconnectConnectorLogDeliveryToTerraform(this._logDelivery.internalValue),
            plugin: cdktf.listMapper(mskconnectConnectorPluginToTerraform, true)(this._plugin.internalValue),
            timeouts: mskconnectConnectorTimeoutsToTerraform(this._timeouts.internalValue),
            worker_configuration: mskconnectConnectorWorkerConfigurationToTerraform(this._workerConfiguration.internalValue),
        };
    }
}
exports.MskconnectConnector = MskconnectConnector;
_w = JSII_RTTI_SYMBOL_1;
MskconnectConnector[_w] = { fqn: "@cdktf/provider-aws.mskconnectConnector.MskconnectConnector", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
MskconnectConnector.tfResourceType = "aws_mskconnect_connector";
