// @cdktf/provider-aws/lib/vpc-endpoint-service/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VpcEndpointService = exports.VpcEndpointServiceTimeoutsOutputReference = exports.vpcEndpointServiceTimeoutsToTerraform = exports.VpcEndpointServicePrivateDnsNameConfigurationList = exports.VpcEndpointServicePrivateDnsNameConfigurationOutputReference = exports.vpcEndpointServicePrivateDnsNameConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function vpcEndpointServicePrivateDnsNameConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.vpcEndpointServicePrivateDnsNameConfigurationToTerraform = vpcEndpointServicePrivateDnsNameConfigurationToTerraform;
class VpcEndpointServicePrivateDnsNameConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
    // value - computed: true, optional: false, required: false
    get value() {
        return this.getStringAttribute('value');
    }
}
exports.VpcEndpointServicePrivateDnsNameConfigurationOutputReference = VpcEndpointServicePrivateDnsNameConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
VpcEndpointServicePrivateDnsNameConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.vpcEndpointService.VpcEndpointServicePrivateDnsNameConfigurationOutputReference", version: "10.0.12" };
class VpcEndpointServicePrivateDnsNameConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new VpcEndpointServicePrivateDnsNameConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.VpcEndpointServicePrivateDnsNameConfigurationList = VpcEndpointServicePrivateDnsNameConfigurationList;
_b = JSII_RTTI_SYMBOL_1;
VpcEndpointServicePrivateDnsNameConfigurationList[_b] = { fqn: "@cdktf/provider-aws.vpcEndpointService.VpcEndpointServicePrivateDnsNameConfigurationList", version: "10.0.12" };
function vpcEndpointServiceTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.vpcEndpointServiceTimeoutsToTerraform = vpcEndpointServiceTimeoutsToTerraform;
class VpcEndpointServiceTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.VpcEndpointServiceTimeoutsOutputReference = VpcEndpointServiceTimeoutsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
VpcEndpointServiceTimeoutsOutputReference[_c] = { fqn: "@cdktf/provider-aws.vpcEndpointService.VpcEndpointServiceTimeoutsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/vpc_endpoint_service aws_vpc_endpoint_service}
*/
class VpcEndpointService extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/vpc_endpoint_service aws_vpc_endpoint_service} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options VpcEndpointServiceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_vpc_endpoint_service',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // private_dns_name_configuration - computed: true, optional: false, required: false
        this._privateDnsNameConfiguration = new VpcEndpointServicePrivateDnsNameConfigurationList(this, "private_dns_name_configuration", false);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new VpcEndpointServiceTimeoutsOutputReference(this, "timeouts");
        this._acceptanceRequired = config.acceptanceRequired;
        this._allowedPrincipals = config.allowedPrincipals;
        this._gatewayLoadBalancerArns = config.gatewayLoadBalancerArns;
        this._id = config.id;
        this._networkLoadBalancerArns = config.networkLoadBalancerArns;
        this._privateDnsName = config.privateDnsName;
        this._supportedIpAddressTypes = config.supportedIpAddressTypes;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._timeouts.internalValue = config.timeouts;
    }
    get acceptanceRequired() {
        return this.getBooleanAttribute('acceptance_required');
    }
    set acceptanceRequired(value) {
        this._acceptanceRequired = value;
    }
    // Temporarily expose input value. Use with caution.
    get acceptanceRequiredInput() {
        return this._acceptanceRequired;
    }
    get allowedPrincipals() {
        return cdktf.Fn.tolist(this.getListAttribute('allowed_principals'));
    }
    set allowedPrincipals(value) {
        this._allowedPrincipals = value;
    }
    resetAllowedPrincipals() {
        this._allowedPrincipals = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allowedPrincipalsInput() {
        return this._allowedPrincipals;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // availability_zones - computed: true, optional: false, required: false
    get availabilityZones() {
        return cdktf.Fn.tolist(this.getListAttribute('availability_zones'));
    }
    // base_endpoint_dns_names - computed: true, optional: false, required: false
    get baseEndpointDnsNames() {
        return cdktf.Fn.tolist(this.getListAttribute('base_endpoint_dns_names'));
    }
    get gatewayLoadBalancerArns() {
        return cdktf.Fn.tolist(this.getListAttribute('gateway_load_balancer_arns'));
    }
    set gatewayLoadBalancerArns(value) {
        this._gatewayLoadBalancerArns = value;
    }
    resetGatewayLoadBalancerArns() {
        this._gatewayLoadBalancerArns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get gatewayLoadBalancerArnsInput() {
        return this._gatewayLoadBalancerArns;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // manages_vpc_endpoints - computed: true, optional: false, required: false
    get managesVpcEndpoints() {
        return this.getBooleanAttribute('manages_vpc_endpoints');
    }
    get networkLoadBalancerArns() {
        return cdktf.Fn.tolist(this.getListAttribute('network_load_balancer_arns'));
    }
    set networkLoadBalancerArns(value) {
        this._networkLoadBalancerArns = value;
    }
    resetNetworkLoadBalancerArns() {
        this._networkLoadBalancerArns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkLoadBalancerArnsInput() {
        return this._networkLoadBalancerArns;
    }
    get privateDnsName() {
        return this.getStringAttribute('private_dns_name');
    }
    set privateDnsName(value) {
        this._privateDnsName = value;
    }
    resetPrivateDnsName() {
        this._privateDnsName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get privateDnsNameInput() {
        return this._privateDnsName;
    }
    get privateDnsNameConfiguration() {
        return this._privateDnsNameConfiguration;
    }
    // service_name - computed: true, optional: false, required: false
    get serviceName() {
        return this.getStringAttribute('service_name');
    }
    // service_type - computed: true, optional: false, required: false
    get serviceType() {
        return this.getStringAttribute('service_type');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get supportedIpAddressTypes() {
        return cdktf.Fn.tolist(this.getListAttribute('supported_ip_address_types'));
    }
    set supportedIpAddressTypes(value) {
        this._supportedIpAddressTypes = value;
    }
    resetSupportedIpAddressTypes() {
        this._supportedIpAddressTypes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get supportedIpAddressTypesInput() {
        return this._supportedIpAddressTypes;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            acceptance_required: cdktf.booleanToTerraform(this._acceptanceRequired),
            allowed_principals: cdktf.listMapper(cdktf.stringToTerraform, false)(this._allowedPrincipals),
            gateway_load_balancer_arns: cdktf.listMapper(cdktf.stringToTerraform, false)(this._gatewayLoadBalancerArns),
            id: cdktf.stringToTerraform(this._id),
            network_load_balancer_arns: cdktf.listMapper(cdktf.stringToTerraform, false)(this._networkLoadBalancerArns),
            private_dns_name: cdktf.stringToTerraform(this._privateDnsName),
            supported_ip_address_types: cdktf.listMapper(cdktf.stringToTerraform, false)(this._supportedIpAddressTypes),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            timeouts: vpcEndpointServiceTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.VpcEndpointService = VpcEndpointService;
_d = JSII_RTTI_SYMBOL_1;
VpcEndpointService[_d] = { fqn: "@cdktf/provider-aws.vpcEndpointService.VpcEndpointService", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
VpcEndpointService.tfResourceType = "aws_vpc_endpoint_service";
