// @cdktf/provider-aws/lib/elasticache-user/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElasticacheUser = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/elasticache_user aws_elasticache_user}
*/
class ElasticacheUser extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/elasticache_user aws_elasticache_user} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ElasticacheUserConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_elasticache_user',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._accessString = config.accessString;
        this._arn = config.arn;
        this._engine = config.engine;
        this._id = config.id;
        this._noPasswordRequired = config.noPasswordRequired;
        this._passwords = config.passwords;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._userId = config.userId;
        this._userName = config.userName;
    }
    get accessString() {
        return this.getStringAttribute('access_string');
    }
    set accessString(value) {
        this._accessString = value;
    }
    // Temporarily expose input value. Use with caution.
    get accessStringInput() {
        return this._accessString;
    }
    get arn() {
        return this.getStringAttribute('arn');
    }
    set arn(value) {
        this._arn = value;
    }
    resetArn() {
        this._arn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get arnInput() {
        return this._arn;
    }
    get engine() {
        return this.getStringAttribute('engine');
    }
    set engine(value) {
        this._engine = value;
    }
    // Temporarily expose input value. Use with caution.
    get engineInput() {
        return this._engine;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get noPasswordRequired() {
        return this.getBooleanAttribute('no_password_required');
    }
    set noPasswordRequired(value) {
        this._noPasswordRequired = value;
    }
    resetNoPasswordRequired() {
        this._noPasswordRequired = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get noPasswordRequiredInput() {
        return this._noPasswordRequired;
    }
    get passwords() {
        return cdktf.Fn.tolist(this.getListAttribute('passwords'));
    }
    set passwords(value) {
        this._passwords = value;
    }
    resetPasswords() {
        this._passwords = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get passwordsInput() {
        return this._passwords;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get userId() {
        return this.getStringAttribute('user_id');
    }
    set userId(value) {
        this._userId = value;
    }
    // Temporarily expose input value. Use with caution.
    get userIdInput() {
        return this._userId;
    }
    get userName() {
        return this.getStringAttribute('user_name');
    }
    set userName(value) {
        this._userName = value;
    }
    // Temporarily expose input value. Use with caution.
    get userNameInput() {
        return this._userName;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            access_string: cdktf.stringToTerraform(this._accessString),
            arn: cdktf.stringToTerraform(this._arn),
            engine: cdktf.stringToTerraform(this._engine),
            id: cdktf.stringToTerraform(this._id),
            no_password_required: cdktf.booleanToTerraform(this._noPasswordRequired),
            passwords: cdktf.listMapper(cdktf.stringToTerraform, false)(this._passwords),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            user_id: cdktf.stringToTerraform(this._userId),
            user_name: cdktf.stringToTerraform(this._userName),
        };
    }
}
exports.ElasticacheUser = ElasticacheUser;
_a = JSII_RTTI_SYMBOL_1;
ElasticacheUser[_a] = { fqn: "@cdktf/provider-aws.elasticacheUser.ElasticacheUser", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
ElasticacheUser.tfResourceType = "aws_elasticache_user";
