// @cdktf/provider-aws/lib/data-aws-location-map/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsLocationMap = exports.DataAwsLocationMapConfigurationList = exports.DataAwsLocationMapConfigurationOutputReference = exports.dataAwsLocationMapConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsLocationMapConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLocationMapConfigurationToTerraform = dataAwsLocationMapConfigurationToTerraform;
class DataAwsLocationMapConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // style - computed: true, optional: false, required: false
    get style() {
        return this.getStringAttribute('style');
    }
}
exports.DataAwsLocationMapConfigurationOutputReference = DataAwsLocationMapConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsLocationMapConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsLocationMap.DataAwsLocationMapConfigurationOutputReference", version: "10.0.12" };
class DataAwsLocationMapConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLocationMapConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLocationMapConfigurationList = DataAwsLocationMapConfigurationList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsLocationMapConfigurationList[_b] = { fqn: "@cdktf/provider-aws.dataAwsLocationMap.DataAwsLocationMapConfigurationList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/location_map aws_location_map}
*/
class DataAwsLocationMap extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/location_map aws_location_map} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsLocationMapConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_location_map',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // configuration - computed: true, optional: false, required: false
        this._configuration = new DataAwsLocationMapConfigurationList(this, "configuration", false);
        this._id = config.id;
        this._mapName = config.mapName;
        this._tags = config.tags;
    }
    get configuration() {
        return this._configuration;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // map_arn - computed: true, optional: false, required: false
    get mapArn() {
        return this.getStringAttribute('map_arn');
    }
    get mapName() {
        return this.getStringAttribute('map_name');
    }
    set mapName(value) {
        this._mapName = value;
    }
    // Temporarily expose input value. Use with caution.
    get mapNameInput() {
        return this._mapName;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            map_name: cdktf.stringToTerraform(this._mapName),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
}
exports.DataAwsLocationMap = DataAwsLocationMap;
_c = JSII_RTTI_SYMBOL_1;
DataAwsLocationMap[_c] = { fqn: "@cdktf/provider-aws.dataAwsLocationMap.DataAwsLocationMap", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsLocationMap.tfResourceType = "aws_location_map";
