// @cdktf/provider-aws/lib/data-aws-autoscaling-group/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsAutoscalingGroup = exports.DataAwsAutoscalingGroupLaunchTemplateList = exports.DataAwsAutoscalingGroupLaunchTemplateOutputReference = exports.dataAwsAutoscalingGroupLaunchTemplateToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsAutoscalingGroupLaunchTemplateToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsAutoscalingGroupLaunchTemplateToTerraform = dataAwsAutoscalingGroupLaunchTemplateToTerraform;
class DataAwsAutoscalingGroupLaunchTemplateOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
}
exports.DataAwsAutoscalingGroupLaunchTemplateOutputReference = DataAwsAutoscalingGroupLaunchTemplateOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsAutoscalingGroupLaunchTemplateOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsAutoscalingGroup.DataAwsAutoscalingGroupLaunchTemplateOutputReference", version: "10.0.12" };
class DataAwsAutoscalingGroupLaunchTemplateList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsAutoscalingGroupLaunchTemplateOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsAutoscalingGroupLaunchTemplateList = DataAwsAutoscalingGroupLaunchTemplateList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsAutoscalingGroupLaunchTemplateList[_b] = { fqn: "@cdktf/provider-aws.dataAwsAutoscalingGroup.DataAwsAutoscalingGroupLaunchTemplateList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/autoscaling_group aws_autoscaling_group}
*/
class DataAwsAutoscalingGroup extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/autoscaling_group aws_autoscaling_group} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsAutoscalingGroupConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_autoscaling_group',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // launch_template - computed: true, optional: false, required: false
        this._launchTemplate = new DataAwsAutoscalingGroupLaunchTemplateList(this, "launch_template", false);
        this._id = config.id;
        this._name = config.name;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // availability_zones - computed: true, optional: false, required: false
    get availabilityZones() {
        return cdktf.Fn.tolist(this.getListAttribute('availability_zones'));
    }
    // default_cooldown - computed: true, optional: false, required: false
    get defaultCooldown() {
        return this.getNumberAttribute('default_cooldown');
    }
    // desired_capacity - computed: true, optional: false, required: false
    get desiredCapacity() {
        return this.getNumberAttribute('desired_capacity');
    }
    // enabled_metrics - computed: true, optional: false, required: false
    get enabledMetrics() {
        return cdktf.Fn.tolist(this.getListAttribute('enabled_metrics'));
    }
    // health_check_grace_period - computed: true, optional: false, required: false
    get healthCheckGracePeriod() {
        return this.getNumberAttribute('health_check_grace_period');
    }
    // health_check_type - computed: true, optional: false, required: false
    get healthCheckType() {
        return this.getStringAttribute('health_check_type');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // launch_configuration - computed: true, optional: false, required: false
    get launchConfiguration() {
        return this.getStringAttribute('launch_configuration');
    }
    get launchTemplate() {
        return this._launchTemplate;
    }
    // load_balancers - computed: true, optional: false, required: false
    get loadBalancers() {
        return cdktf.Fn.tolist(this.getListAttribute('load_balancers'));
    }
    // max_size - computed: true, optional: false, required: false
    get maxSize() {
        return this.getNumberAttribute('max_size');
    }
    // min_size - computed: true, optional: false, required: false
    get minSize() {
        return this.getNumberAttribute('min_size');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // new_instances_protected_from_scale_in - computed: true, optional: false, required: false
    get newInstancesProtectedFromScaleIn() {
        return this.getBooleanAttribute('new_instances_protected_from_scale_in');
    }
    // placement_group - computed: true, optional: false, required: false
    get placementGroup() {
        return this.getStringAttribute('placement_group');
    }
    // service_linked_role_arn - computed: true, optional: false, required: false
    get serviceLinkedRoleArn() {
        return this.getStringAttribute('service_linked_role_arn');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    // target_group_arns - computed: true, optional: false, required: false
    get targetGroupArns() {
        return cdktf.Fn.tolist(this.getListAttribute('target_group_arns'));
    }
    // termination_policies - computed: true, optional: false, required: false
    get terminationPolicies() {
        return cdktf.Fn.tolist(this.getListAttribute('termination_policies'));
    }
    // vpc_zone_identifier - computed: true, optional: false, required: false
    get vpcZoneIdentifier() {
        return this.getStringAttribute('vpc_zone_identifier');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
        };
    }
}
exports.DataAwsAutoscalingGroup = DataAwsAutoscalingGroup;
_c = JSII_RTTI_SYMBOL_1;
DataAwsAutoscalingGroup[_c] = { fqn: "@cdktf/provider-aws.dataAwsAutoscalingGroup.DataAwsAutoscalingGroup", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsAutoscalingGroup.tfResourceType = "aws_autoscaling_group";
