// @cdktf/provider-aws/lib/emr-studio/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmrStudio = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/emr_studio aws_emr_studio}
*/
class EmrStudio extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/emr_studio aws_emr_studio} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options EmrStudioConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_emr_studio',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._authMode = config.authMode;
        this._defaultS3Location = config.defaultS3Location;
        this._description = config.description;
        this._engineSecurityGroupId = config.engineSecurityGroupId;
        this._id = config.id;
        this._idpAuthUrl = config.idpAuthUrl;
        this._idpRelayStateParameterName = config.idpRelayStateParameterName;
        this._name = config.name;
        this._serviceRole = config.serviceRole;
        this._subnetIds = config.subnetIds;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._userRole = config.userRole;
        this._vpcId = config.vpcId;
        this._workspaceSecurityGroupId = config.workspaceSecurityGroupId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get authMode() {
        return this.getStringAttribute('auth_mode');
    }
    set authMode(value) {
        this._authMode = value;
    }
    // Temporarily expose input value. Use with caution.
    get authModeInput() {
        return this._authMode;
    }
    get defaultS3Location() {
        return this.getStringAttribute('default_s3_location');
    }
    set defaultS3Location(value) {
        this._defaultS3Location = value;
    }
    // Temporarily expose input value. Use with caution.
    get defaultS3LocationInput() {
        return this._defaultS3Location;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get engineSecurityGroupId() {
        return this.getStringAttribute('engine_security_group_id');
    }
    set engineSecurityGroupId(value) {
        this._engineSecurityGroupId = value;
    }
    // Temporarily expose input value. Use with caution.
    get engineSecurityGroupIdInput() {
        return this._engineSecurityGroupId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get idpAuthUrl() {
        return this.getStringAttribute('idp_auth_url');
    }
    set idpAuthUrl(value) {
        this._idpAuthUrl = value;
    }
    resetIdpAuthUrl() {
        this._idpAuthUrl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idpAuthUrlInput() {
        return this._idpAuthUrl;
    }
    get idpRelayStateParameterName() {
        return this.getStringAttribute('idp_relay_state_parameter_name');
    }
    set idpRelayStateParameterName(value) {
        this._idpRelayStateParameterName = value;
    }
    resetIdpRelayStateParameterName() {
        this._idpRelayStateParameterName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idpRelayStateParameterNameInput() {
        return this._idpRelayStateParameterName;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get serviceRole() {
        return this.getStringAttribute('service_role');
    }
    set serviceRole(value) {
        this._serviceRole = value;
    }
    // Temporarily expose input value. Use with caution.
    get serviceRoleInput() {
        return this._serviceRole;
    }
    get subnetIds() {
        return cdktf.Fn.tolist(this.getListAttribute('subnet_ids'));
    }
    set subnetIds(value) {
        this._subnetIds = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdsInput() {
        return this._subnetIds;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // url - computed: true, optional: false, required: false
    get url() {
        return this.getStringAttribute('url');
    }
    get userRole() {
        return this.getStringAttribute('user_role');
    }
    set userRole(value) {
        this._userRole = value;
    }
    resetUserRole() {
        this._userRole = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get userRoleInput() {
        return this._userRole;
    }
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    set vpcId(value) {
        this._vpcId = value;
    }
    // Temporarily expose input value. Use with caution.
    get vpcIdInput() {
        return this._vpcId;
    }
    get workspaceSecurityGroupId() {
        return this.getStringAttribute('workspace_security_group_id');
    }
    set workspaceSecurityGroupId(value) {
        this._workspaceSecurityGroupId = value;
    }
    // Temporarily expose input value. Use with caution.
    get workspaceSecurityGroupIdInput() {
        return this._workspaceSecurityGroupId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            auth_mode: cdktf.stringToTerraform(this._authMode),
            default_s3_location: cdktf.stringToTerraform(this._defaultS3Location),
            description: cdktf.stringToTerraform(this._description),
            engine_security_group_id: cdktf.stringToTerraform(this._engineSecurityGroupId),
            id: cdktf.stringToTerraform(this._id),
            idp_auth_url: cdktf.stringToTerraform(this._idpAuthUrl),
            idp_relay_state_parameter_name: cdktf.stringToTerraform(this._idpRelayStateParameterName),
            name: cdktf.stringToTerraform(this._name),
            service_role: cdktf.stringToTerraform(this._serviceRole),
            subnet_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(this._subnetIds),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            user_role: cdktf.stringToTerraform(this._userRole),
            vpc_id: cdktf.stringToTerraform(this._vpcId),
            workspace_security_group_id: cdktf.stringToTerraform(this._workspaceSecurityGroupId),
        };
    }
}
exports.EmrStudio = EmrStudio;
_a = JSII_RTTI_SYMBOL_1;
EmrStudio[_a] = { fqn: "@cdktf/provider-aws.emrStudio.EmrStudio", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
EmrStudio.tfResourceType = "aws_emr_studio";
