// @cdktf/provider-aws/lib/route53-resolver-query-log-config-association/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Route53ResolverQueryLogConfigAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/route53_resolver_query_log_config_association aws_route53_resolver_query_log_config_association}
*/
class Route53ResolverQueryLogConfigAssociation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/route53_resolver_query_log_config_association aws_route53_resolver_query_log_config_association} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options Route53ResolverQueryLogConfigAssociationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_route53_resolver_query_log_config_association',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._resolverQueryLogConfigId = config.resolverQueryLogConfigId;
        this._resourceId = config.resourceId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get resolverQueryLogConfigId() {
        return this.getStringAttribute('resolver_query_log_config_id');
    }
    set resolverQueryLogConfigId(value) {
        this._resolverQueryLogConfigId = value;
    }
    // Temporarily expose input value. Use with caution.
    get resolverQueryLogConfigIdInput() {
        return this._resolverQueryLogConfigId;
    }
    get resourceId() {
        return this.getStringAttribute('resource_id');
    }
    set resourceId(value) {
        this._resourceId = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourceIdInput() {
        return this._resourceId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            resolver_query_log_config_id: cdktf.stringToTerraform(this._resolverQueryLogConfigId),
            resource_id: cdktf.stringToTerraform(this._resourceId),
        };
    }
}
exports.Route53ResolverQueryLogConfigAssociation = Route53ResolverQueryLogConfigAssociation;
_a = JSII_RTTI_SYMBOL_1;
Route53ResolverQueryLogConfigAssociation[_a] = { fqn: "@cdktf/provider-aws.route53ResolverQueryLogConfigAssociation.Route53ResolverQueryLogConfigAssociation", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
Route53ResolverQueryLogConfigAssociation.tfResourceType = "aws_route53_resolver_query_log_config_association";
