// @cdktf/provider-aws/lib/data-aws-cloudfront-realtime-log-config/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsCloudfrontRealtimeLogConfig = exports.DataAwsCloudfrontRealtimeLogConfigEndpointList = exports.DataAwsCloudfrontRealtimeLogConfigEndpointOutputReference = exports.dataAwsCloudfrontRealtimeLogConfigEndpointToTerraform = exports.DataAwsCloudfrontRealtimeLogConfigEndpointKinesisStreamConfigList = exports.DataAwsCloudfrontRealtimeLogConfigEndpointKinesisStreamConfigOutputReference = exports.dataAwsCloudfrontRealtimeLogConfigEndpointKinesisStreamConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsCloudfrontRealtimeLogConfigEndpointKinesisStreamConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCloudfrontRealtimeLogConfigEndpointKinesisStreamConfigToTerraform = dataAwsCloudfrontRealtimeLogConfigEndpointKinesisStreamConfigToTerraform;
class DataAwsCloudfrontRealtimeLogConfigEndpointKinesisStreamConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // role_arn - computed: true, optional: false, required: false
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    // stream_arn - computed: true, optional: false, required: false
    get streamArn() {
        return this.getStringAttribute('stream_arn');
    }
}
exports.DataAwsCloudfrontRealtimeLogConfigEndpointKinesisStreamConfigOutputReference = DataAwsCloudfrontRealtimeLogConfigEndpointKinesisStreamConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsCloudfrontRealtimeLogConfigEndpointKinesisStreamConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsCloudfrontRealtimeLogConfig.DataAwsCloudfrontRealtimeLogConfigEndpointKinesisStreamConfigOutputReference", version: "10.0.12" };
class DataAwsCloudfrontRealtimeLogConfigEndpointKinesisStreamConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCloudfrontRealtimeLogConfigEndpointKinesisStreamConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCloudfrontRealtimeLogConfigEndpointKinesisStreamConfigList = DataAwsCloudfrontRealtimeLogConfigEndpointKinesisStreamConfigList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsCloudfrontRealtimeLogConfigEndpointKinesisStreamConfigList[_b] = { fqn: "@cdktf/provider-aws.dataAwsCloudfrontRealtimeLogConfig.DataAwsCloudfrontRealtimeLogConfigEndpointKinesisStreamConfigList", version: "10.0.12" };
function dataAwsCloudfrontRealtimeLogConfigEndpointToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCloudfrontRealtimeLogConfigEndpointToTerraform = dataAwsCloudfrontRealtimeLogConfigEndpointToTerraform;
class DataAwsCloudfrontRealtimeLogConfigEndpointOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // kinesis_stream_config - computed: true, optional: false, required: false
        this._kinesisStreamConfig = new DataAwsCloudfrontRealtimeLogConfigEndpointKinesisStreamConfigList(this, "kinesis_stream_config", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get kinesisStreamConfig() {
        return this._kinesisStreamConfig;
    }
    // stream_type - computed: true, optional: false, required: false
    get streamType() {
        return this.getStringAttribute('stream_type');
    }
}
exports.DataAwsCloudfrontRealtimeLogConfigEndpointOutputReference = DataAwsCloudfrontRealtimeLogConfigEndpointOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsCloudfrontRealtimeLogConfigEndpointOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsCloudfrontRealtimeLogConfig.DataAwsCloudfrontRealtimeLogConfigEndpointOutputReference", version: "10.0.12" };
class DataAwsCloudfrontRealtimeLogConfigEndpointList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCloudfrontRealtimeLogConfigEndpointOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCloudfrontRealtimeLogConfigEndpointList = DataAwsCloudfrontRealtimeLogConfigEndpointList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsCloudfrontRealtimeLogConfigEndpointList[_d] = { fqn: "@cdktf/provider-aws.dataAwsCloudfrontRealtimeLogConfig.DataAwsCloudfrontRealtimeLogConfigEndpointList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/cloudfront_realtime_log_config aws_cloudfront_realtime_log_config}
*/
class DataAwsCloudfrontRealtimeLogConfig extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/cloudfront_realtime_log_config aws_cloudfront_realtime_log_config} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsCloudfrontRealtimeLogConfigConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cloudfront_realtime_log_config',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // endpoint - computed: true, optional: false, required: false
        this._endpoint = new DataAwsCloudfrontRealtimeLogConfigEndpointList(this, "endpoint", false);
        this._id = config.id;
        this._name = config.name;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get endpoint() {
        return this._endpoint;
    }
    // fields - computed: true, optional: false, required: false
    get fields() {
        return cdktf.Fn.tolist(this.getListAttribute('fields'));
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // sampling_rate - computed: true, optional: false, required: false
    get samplingRate() {
        return this.getNumberAttribute('sampling_rate');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
        };
    }
}
exports.DataAwsCloudfrontRealtimeLogConfig = DataAwsCloudfrontRealtimeLogConfig;
_e = JSII_RTTI_SYMBOL_1;
DataAwsCloudfrontRealtimeLogConfig[_e] = { fqn: "@cdktf/provider-aws.dataAwsCloudfrontRealtimeLogConfig.DataAwsCloudfrontRealtimeLogConfig", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsCloudfrontRealtimeLogConfig.tfResourceType = "aws_cloudfront_realtime_log_config";
