// @cdktf/provider-aws/lib/dx-gateway-association/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DxGatewayAssociation = exports.DxGatewayAssociationTimeoutsOutputReference = exports.dxGatewayAssociationTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dxGatewayAssociationTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.dxGatewayAssociationTimeoutsToTerraform = dxGatewayAssociationTimeoutsToTerraform;
class DxGatewayAssociationTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.DxGatewayAssociationTimeoutsOutputReference = DxGatewayAssociationTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DxGatewayAssociationTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.dxGatewayAssociation.DxGatewayAssociationTimeoutsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/dx_gateway_association aws_dx_gateway_association}
*/
class DxGatewayAssociation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/dx_gateway_association aws_dx_gateway_association} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DxGatewayAssociationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_dx_gateway_association',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new DxGatewayAssociationTimeoutsOutputReference(this, "timeouts");
        this._allowedPrefixes = config.allowedPrefixes;
        this._associatedGatewayId = config.associatedGatewayId;
        this._associatedGatewayOwnerAccountId = config.associatedGatewayOwnerAccountId;
        this._dxGatewayId = config.dxGatewayId;
        this._id = config.id;
        this._proposalId = config.proposalId;
        this._vpnGatewayId = config.vpnGatewayId;
        this._timeouts.internalValue = config.timeouts;
    }
    get allowedPrefixes() {
        return cdktf.Fn.tolist(this.getListAttribute('allowed_prefixes'));
    }
    set allowedPrefixes(value) {
        this._allowedPrefixes = value;
    }
    resetAllowedPrefixes() {
        this._allowedPrefixes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allowedPrefixesInput() {
        return this._allowedPrefixes;
    }
    get associatedGatewayId() {
        return this.getStringAttribute('associated_gateway_id');
    }
    set associatedGatewayId(value) {
        this._associatedGatewayId = value;
    }
    resetAssociatedGatewayId() {
        this._associatedGatewayId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get associatedGatewayIdInput() {
        return this._associatedGatewayId;
    }
    get associatedGatewayOwnerAccountId() {
        return this.getStringAttribute('associated_gateway_owner_account_id');
    }
    set associatedGatewayOwnerAccountId(value) {
        this._associatedGatewayOwnerAccountId = value;
    }
    resetAssociatedGatewayOwnerAccountId() {
        this._associatedGatewayOwnerAccountId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get associatedGatewayOwnerAccountIdInput() {
        return this._associatedGatewayOwnerAccountId;
    }
    // associated_gateway_type - computed: true, optional: false, required: false
    get associatedGatewayType() {
        return this.getStringAttribute('associated_gateway_type');
    }
    // dx_gateway_association_id - computed: true, optional: false, required: false
    get dxGatewayAssociationId() {
        return this.getStringAttribute('dx_gateway_association_id');
    }
    get dxGatewayId() {
        return this.getStringAttribute('dx_gateway_id');
    }
    set dxGatewayId(value) {
        this._dxGatewayId = value;
    }
    // Temporarily expose input value. Use with caution.
    get dxGatewayIdInput() {
        return this._dxGatewayId;
    }
    // dx_gateway_owner_account_id - computed: true, optional: false, required: false
    get dxGatewayOwnerAccountId() {
        return this.getStringAttribute('dx_gateway_owner_account_id');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get proposalId() {
        return this.getStringAttribute('proposal_id');
    }
    set proposalId(value) {
        this._proposalId = value;
    }
    resetProposalId() {
        this._proposalId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get proposalIdInput() {
        return this._proposalId;
    }
    get vpnGatewayId() {
        return this.getStringAttribute('vpn_gateway_id');
    }
    set vpnGatewayId(value) {
        this._vpnGatewayId = value;
    }
    resetVpnGatewayId() {
        this._vpnGatewayId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpnGatewayIdInput() {
        return this._vpnGatewayId;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            allowed_prefixes: cdktf.listMapper(cdktf.stringToTerraform, false)(this._allowedPrefixes),
            associated_gateway_id: cdktf.stringToTerraform(this._associatedGatewayId),
            associated_gateway_owner_account_id: cdktf.stringToTerraform(this._associatedGatewayOwnerAccountId),
            dx_gateway_id: cdktf.stringToTerraform(this._dxGatewayId),
            id: cdktf.stringToTerraform(this._id),
            proposal_id: cdktf.stringToTerraform(this._proposalId),
            vpn_gateway_id: cdktf.stringToTerraform(this._vpnGatewayId),
            timeouts: dxGatewayAssociationTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.DxGatewayAssociation = DxGatewayAssociation;
_b = JSII_RTTI_SYMBOL_1;
DxGatewayAssociation[_b] = { fqn: "@cdktf/provider-aws.dxGatewayAssociation.DxGatewayAssociation", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DxGatewayAssociation.tfResourceType = "aws_dx_gateway_association";
