// @cdktf/provider-aws/lib/macie-member-account-association/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MacieMemberAccountAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/macie_member_account_association aws_macie_member_account_association}
*/
class MacieMemberAccountAssociation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/macie_member_account_association aws_macie_member_account_association} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options MacieMemberAccountAssociationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_macie_member_account_association',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._memberAccountId = config.memberAccountId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get memberAccountId() {
        return this.getStringAttribute('member_account_id');
    }
    set memberAccountId(value) {
        this._memberAccountId = value;
    }
    // Temporarily expose input value. Use with caution.
    get memberAccountIdInput() {
        return this._memberAccountId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            member_account_id: cdktf.stringToTerraform(this._memberAccountId),
        };
    }
}
exports.MacieMemberAccountAssociation = MacieMemberAccountAssociation;
_a = JSII_RTTI_SYMBOL_1;
MacieMemberAccountAssociation[_a] = { fqn: "@cdktf/provider-aws.macieMemberAccountAssociation.MacieMemberAccountAssociation", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
MacieMemberAccountAssociation.tfResourceType = "aws_macie_member_account_association";
