// @cdktf/provider-aws/lib/glue-dev-endpoint/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlueDevEndpoint = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/glue_dev_endpoint aws_glue_dev_endpoint}
*/
class GlueDevEndpoint extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/glue_dev_endpoint aws_glue_dev_endpoint} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options GlueDevEndpointConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_glue_dev_endpoint',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._arguments = config.arguments;
        this._extraJarsS3Path = config.extraJarsS3Path;
        this._extraPythonLibsS3Path = config.extraPythonLibsS3Path;
        this._glueVersion = config.glueVersion;
        this._id = config.id;
        this._name = config.name;
        this._numberOfNodes = config.numberOfNodes;
        this._numberOfWorkers = config.numberOfWorkers;
        this._publicKey = config.publicKey;
        this._publicKeys = config.publicKeys;
        this._roleArn = config.roleArn;
        this._securityConfiguration = config.securityConfiguration;
        this._securityGroupIds = config.securityGroupIds;
        this._subnetId = config.subnetId;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._workerType = config.workerType;
    }
    get arguments() {
        return this.getStringMapAttribute('arguments');
    }
    set arguments(value) {
        this._arguments = value;
    }
    resetArguments() {
        this._arguments = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get argumentsInput() {
        return this._arguments;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // availability_zone - computed: true, optional: false, required: false
    get availabilityZone() {
        return this.getStringAttribute('availability_zone');
    }
    get extraJarsS3Path() {
        return this.getStringAttribute('extra_jars_s3_path');
    }
    set extraJarsS3Path(value) {
        this._extraJarsS3Path = value;
    }
    resetExtraJarsS3Path() {
        this._extraJarsS3Path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get extraJarsS3PathInput() {
        return this._extraJarsS3Path;
    }
    get extraPythonLibsS3Path() {
        return this.getStringAttribute('extra_python_libs_s3_path');
    }
    set extraPythonLibsS3Path(value) {
        this._extraPythonLibsS3Path = value;
    }
    resetExtraPythonLibsS3Path() {
        this._extraPythonLibsS3Path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get extraPythonLibsS3PathInput() {
        return this._extraPythonLibsS3Path;
    }
    // failure_reason - computed: true, optional: false, required: false
    get failureReason() {
        return this.getStringAttribute('failure_reason');
    }
    get glueVersion() {
        return this.getStringAttribute('glue_version');
    }
    set glueVersion(value) {
        this._glueVersion = value;
    }
    resetGlueVersion() {
        this._glueVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get glueVersionInput() {
        return this._glueVersion;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get numberOfNodes() {
        return this.getNumberAttribute('number_of_nodes');
    }
    set numberOfNodes(value) {
        this._numberOfNodes = value;
    }
    resetNumberOfNodes() {
        this._numberOfNodes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get numberOfNodesInput() {
        return this._numberOfNodes;
    }
    get numberOfWorkers() {
        return this.getNumberAttribute('number_of_workers');
    }
    set numberOfWorkers(value) {
        this._numberOfWorkers = value;
    }
    resetNumberOfWorkers() {
        this._numberOfWorkers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get numberOfWorkersInput() {
        return this._numberOfWorkers;
    }
    // private_address - computed: true, optional: false, required: false
    get privateAddress() {
        return this.getStringAttribute('private_address');
    }
    // public_address - computed: true, optional: false, required: false
    get publicAddress() {
        return this.getStringAttribute('public_address');
    }
    get publicKey() {
        return this.getStringAttribute('public_key');
    }
    set publicKey(value) {
        this._publicKey = value;
    }
    resetPublicKey() {
        this._publicKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get publicKeyInput() {
        return this._publicKey;
    }
    get publicKeys() {
        return cdktf.Fn.tolist(this.getListAttribute('public_keys'));
    }
    set publicKeys(value) {
        this._publicKeys = value;
    }
    resetPublicKeys() {
        this._publicKeys = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get publicKeysInput() {
        return this._publicKeys;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get securityConfiguration() {
        return this.getStringAttribute('security_configuration');
    }
    set securityConfiguration(value) {
        this._securityConfiguration = value;
    }
    resetSecurityConfiguration() {
        this._securityConfiguration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityConfigurationInput() {
        return this._securityConfiguration;
    }
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    resetSecurityGroupIds() {
        this._securityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    set subnetId(value) {
        this._subnetId = value;
    }
    resetSubnetId() {
        this._subnetId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdInput() {
        return this._subnetId;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // vpc_id - computed: true, optional: false, required: false
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    get workerType() {
        return this.getStringAttribute('worker_type');
    }
    set workerType(value) {
        this._workerType = value;
    }
    resetWorkerType() {
        this._workerType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get workerTypeInput() {
        return this._workerType;
    }
    // yarn_endpoint_address - computed: true, optional: false, required: false
    get yarnEndpointAddress() {
        return this.getStringAttribute('yarn_endpoint_address');
    }
    // zeppelin_remote_spark_interpreter_port - computed: true, optional: false, required: false
    get zeppelinRemoteSparkInterpreterPort() {
        return this.getNumberAttribute('zeppelin_remote_spark_interpreter_port');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            arguments: cdktf.hashMapper(cdktf.stringToTerraform)(this._arguments),
            extra_jars_s3_path: cdktf.stringToTerraform(this._extraJarsS3Path),
            extra_python_libs_s3_path: cdktf.stringToTerraform(this._extraPythonLibsS3Path),
            glue_version: cdktf.stringToTerraform(this._glueVersion),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            number_of_nodes: cdktf.numberToTerraform(this._numberOfNodes),
            number_of_workers: cdktf.numberToTerraform(this._numberOfWorkers),
            public_key: cdktf.stringToTerraform(this._publicKey),
            public_keys: cdktf.listMapper(cdktf.stringToTerraform, false)(this._publicKeys),
            role_arn: cdktf.stringToTerraform(this._roleArn),
            security_configuration: cdktf.stringToTerraform(this._securityConfiguration),
            security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(this._securityGroupIds),
            subnet_id: cdktf.stringToTerraform(this._subnetId),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            worker_type: cdktf.stringToTerraform(this._workerType),
        };
    }
}
exports.GlueDevEndpoint = GlueDevEndpoint;
_a = JSII_RTTI_SYMBOL_1;
GlueDevEndpoint[_a] = { fqn: "@cdktf/provider-aws.glueDevEndpoint.GlueDevEndpoint", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
GlueDevEndpoint.tfResourceType = "aws_glue_dev_endpoint";
