// @cdktf/provider-aws/lib/config-organization-custom-rule/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigOrganizationCustomRule = exports.ConfigOrganizationCustomRuleTimeoutsOutputReference = exports.configOrganizationCustomRuleTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function configOrganizationCustomRuleTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.configOrganizationCustomRuleTimeoutsToTerraform = configOrganizationCustomRuleTimeoutsToTerraform;
class ConfigOrganizationCustomRuleTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.ConfigOrganizationCustomRuleTimeoutsOutputReference = ConfigOrganizationCustomRuleTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ConfigOrganizationCustomRuleTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.configOrganizationCustomRule.ConfigOrganizationCustomRuleTimeoutsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/config_organization_custom_rule aws_config_organization_custom_rule}
*/
class ConfigOrganizationCustomRule extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/config_organization_custom_rule aws_config_organization_custom_rule} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ConfigOrganizationCustomRuleConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_config_organization_custom_rule',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new ConfigOrganizationCustomRuleTimeoutsOutputReference(this, "timeouts");
        this._description = config.description;
        this._excludedAccounts = config.excludedAccounts;
        this._id = config.id;
        this._inputParameters = config.inputParameters;
        this._lambdaFunctionArn = config.lambdaFunctionArn;
        this._maximumExecutionFrequency = config.maximumExecutionFrequency;
        this._name = config.name;
        this._resourceIdScope = config.resourceIdScope;
        this._resourceTypesScope = config.resourceTypesScope;
        this._tagKeyScope = config.tagKeyScope;
        this._tagValueScope = config.tagValueScope;
        this._triggerTypes = config.triggerTypes;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get excludedAccounts() {
        return cdktf.Fn.tolist(this.getListAttribute('excluded_accounts'));
    }
    set excludedAccounts(value) {
        this._excludedAccounts = value;
    }
    resetExcludedAccounts() {
        this._excludedAccounts = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get excludedAccountsInput() {
        return this._excludedAccounts;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get inputParameters() {
        return this.getStringAttribute('input_parameters');
    }
    set inputParameters(value) {
        this._inputParameters = value;
    }
    resetInputParameters() {
        this._inputParameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get inputParametersInput() {
        return this._inputParameters;
    }
    get lambdaFunctionArn() {
        return this.getStringAttribute('lambda_function_arn');
    }
    set lambdaFunctionArn(value) {
        this._lambdaFunctionArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get lambdaFunctionArnInput() {
        return this._lambdaFunctionArn;
    }
    get maximumExecutionFrequency() {
        return this.getStringAttribute('maximum_execution_frequency');
    }
    set maximumExecutionFrequency(value) {
        this._maximumExecutionFrequency = value;
    }
    resetMaximumExecutionFrequency() {
        this._maximumExecutionFrequency = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maximumExecutionFrequencyInput() {
        return this._maximumExecutionFrequency;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get resourceIdScope() {
        return this.getStringAttribute('resource_id_scope');
    }
    set resourceIdScope(value) {
        this._resourceIdScope = value;
    }
    resetResourceIdScope() {
        this._resourceIdScope = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourceIdScopeInput() {
        return this._resourceIdScope;
    }
    get resourceTypesScope() {
        return cdktf.Fn.tolist(this.getListAttribute('resource_types_scope'));
    }
    set resourceTypesScope(value) {
        this._resourceTypesScope = value;
    }
    resetResourceTypesScope() {
        this._resourceTypesScope = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourceTypesScopeInput() {
        return this._resourceTypesScope;
    }
    get tagKeyScope() {
        return this.getStringAttribute('tag_key_scope');
    }
    set tagKeyScope(value) {
        this._tagKeyScope = value;
    }
    resetTagKeyScope() {
        this._tagKeyScope = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagKeyScopeInput() {
        return this._tagKeyScope;
    }
    get tagValueScope() {
        return this.getStringAttribute('tag_value_scope');
    }
    set tagValueScope(value) {
        this._tagValueScope = value;
    }
    resetTagValueScope() {
        this._tagValueScope = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagValueScopeInput() {
        return this._tagValueScope;
    }
    get triggerTypes() {
        return cdktf.Fn.tolist(this.getListAttribute('trigger_types'));
    }
    set triggerTypes(value) {
        this._triggerTypes = value;
    }
    // Temporarily expose input value. Use with caution.
    get triggerTypesInput() {
        return this._triggerTypes;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            excluded_accounts: cdktf.listMapper(cdktf.stringToTerraform, false)(this._excludedAccounts),
            id: cdktf.stringToTerraform(this._id),
            input_parameters: cdktf.stringToTerraform(this._inputParameters),
            lambda_function_arn: cdktf.stringToTerraform(this._lambdaFunctionArn),
            maximum_execution_frequency: cdktf.stringToTerraform(this._maximumExecutionFrequency),
            name: cdktf.stringToTerraform(this._name),
            resource_id_scope: cdktf.stringToTerraform(this._resourceIdScope),
            resource_types_scope: cdktf.listMapper(cdktf.stringToTerraform, false)(this._resourceTypesScope),
            tag_key_scope: cdktf.stringToTerraform(this._tagKeyScope),
            tag_value_scope: cdktf.stringToTerraform(this._tagValueScope),
            trigger_types: cdktf.listMapper(cdktf.stringToTerraform, false)(this._triggerTypes),
            timeouts: configOrganizationCustomRuleTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.ConfigOrganizationCustomRule = ConfigOrganizationCustomRule;
_b = JSII_RTTI_SYMBOL_1;
ConfigOrganizationCustomRule[_b] = { fqn: "@cdktf/provider-aws.configOrganizationCustomRule.ConfigOrganizationCustomRule", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
ConfigOrganizationCustomRule.tfResourceType = "aws_config_organization_custom_rule";
