// @cdktf/provider-aws/lib/data-aws-service-discovery-service/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsServiceDiscoveryService = exports.DataAwsServiceDiscoveryServiceHealthCheckCustomConfigList = exports.DataAwsServiceDiscoveryServiceHealthCheckCustomConfigOutputReference = exports.dataAwsServiceDiscoveryServiceHealthCheckCustomConfigToTerraform = exports.DataAwsServiceDiscoveryServiceHealthCheckConfigList = exports.DataAwsServiceDiscoveryServiceHealthCheckConfigOutputReference = exports.dataAwsServiceDiscoveryServiceHealthCheckConfigToTerraform = exports.DataAwsServiceDiscoveryServiceDnsConfigList = exports.DataAwsServiceDiscoveryServiceDnsConfigOutputReference = exports.dataAwsServiceDiscoveryServiceDnsConfigToTerraform = exports.DataAwsServiceDiscoveryServiceDnsConfigDnsRecordsList = exports.DataAwsServiceDiscoveryServiceDnsConfigDnsRecordsOutputReference = exports.dataAwsServiceDiscoveryServiceDnsConfigDnsRecordsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsServiceDiscoveryServiceDnsConfigDnsRecordsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsServiceDiscoveryServiceDnsConfigDnsRecordsToTerraform = dataAwsServiceDiscoveryServiceDnsConfigDnsRecordsToTerraform;
class DataAwsServiceDiscoveryServiceDnsConfigDnsRecordsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ttl - computed: true, optional: false, required: false
    get ttl() {
        return this.getNumberAttribute('ttl');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataAwsServiceDiscoveryServiceDnsConfigDnsRecordsOutputReference = DataAwsServiceDiscoveryServiceDnsConfigDnsRecordsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsServiceDiscoveryServiceDnsConfigDnsRecordsOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsServiceDiscoveryService.DataAwsServiceDiscoveryServiceDnsConfigDnsRecordsOutputReference", version: "10.0.12" };
class DataAwsServiceDiscoveryServiceDnsConfigDnsRecordsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsServiceDiscoveryServiceDnsConfigDnsRecordsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsServiceDiscoveryServiceDnsConfigDnsRecordsList = DataAwsServiceDiscoveryServiceDnsConfigDnsRecordsList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsServiceDiscoveryServiceDnsConfigDnsRecordsList[_b] = { fqn: "@cdktf/provider-aws.dataAwsServiceDiscoveryService.DataAwsServiceDiscoveryServiceDnsConfigDnsRecordsList", version: "10.0.12" };
function dataAwsServiceDiscoveryServiceDnsConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsServiceDiscoveryServiceDnsConfigToTerraform = dataAwsServiceDiscoveryServiceDnsConfigToTerraform;
class DataAwsServiceDiscoveryServiceDnsConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // dns_records - computed: true, optional: false, required: false
        this._dnsRecords = new DataAwsServiceDiscoveryServiceDnsConfigDnsRecordsList(this, "dns_records", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get dnsRecords() {
        return this._dnsRecords;
    }
    // namespace_id - computed: true, optional: false, required: false
    get namespaceId() {
        return this.getStringAttribute('namespace_id');
    }
    // routing_policy - computed: true, optional: false, required: false
    get routingPolicy() {
        return this.getStringAttribute('routing_policy');
    }
}
exports.DataAwsServiceDiscoveryServiceDnsConfigOutputReference = DataAwsServiceDiscoveryServiceDnsConfigOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsServiceDiscoveryServiceDnsConfigOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsServiceDiscoveryService.DataAwsServiceDiscoveryServiceDnsConfigOutputReference", version: "10.0.12" };
class DataAwsServiceDiscoveryServiceDnsConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsServiceDiscoveryServiceDnsConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsServiceDiscoveryServiceDnsConfigList = DataAwsServiceDiscoveryServiceDnsConfigList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsServiceDiscoveryServiceDnsConfigList[_d] = { fqn: "@cdktf/provider-aws.dataAwsServiceDiscoveryService.DataAwsServiceDiscoveryServiceDnsConfigList", version: "10.0.12" };
function dataAwsServiceDiscoveryServiceHealthCheckConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsServiceDiscoveryServiceHealthCheckConfigToTerraform = dataAwsServiceDiscoveryServiceHealthCheckConfigToTerraform;
class DataAwsServiceDiscoveryServiceHealthCheckConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // failure_threshold - computed: true, optional: false, required: false
    get failureThreshold() {
        return this.getNumberAttribute('failure_threshold');
    }
    // resource_path - computed: true, optional: false, required: false
    get resourcePath() {
        return this.getStringAttribute('resource_path');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataAwsServiceDiscoveryServiceHealthCheckConfigOutputReference = DataAwsServiceDiscoveryServiceHealthCheckConfigOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsServiceDiscoveryServiceHealthCheckConfigOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsServiceDiscoveryService.DataAwsServiceDiscoveryServiceHealthCheckConfigOutputReference", version: "10.0.12" };
class DataAwsServiceDiscoveryServiceHealthCheckConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsServiceDiscoveryServiceHealthCheckConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsServiceDiscoveryServiceHealthCheckConfigList = DataAwsServiceDiscoveryServiceHealthCheckConfigList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsServiceDiscoveryServiceHealthCheckConfigList[_f] = { fqn: "@cdktf/provider-aws.dataAwsServiceDiscoveryService.DataAwsServiceDiscoveryServiceHealthCheckConfigList", version: "10.0.12" };
function dataAwsServiceDiscoveryServiceHealthCheckCustomConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsServiceDiscoveryServiceHealthCheckCustomConfigToTerraform = dataAwsServiceDiscoveryServiceHealthCheckCustomConfigToTerraform;
class DataAwsServiceDiscoveryServiceHealthCheckCustomConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // failure_threshold - computed: true, optional: false, required: false
    get failureThreshold() {
        return this.getNumberAttribute('failure_threshold');
    }
}
exports.DataAwsServiceDiscoveryServiceHealthCheckCustomConfigOutputReference = DataAwsServiceDiscoveryServiceHealthCheckCustomConfigOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsServiceDiscoveryServiceHealthCheckCustomConfigOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsServiceDiscoveryService.DataAwsServiceDiscoveryServiceHealthCheckCustomConfigOutputReference", version: "10.0.12" };
class DataAwsServiceDiscoveryServiceHealthCheckCustomConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsServiceDiscoveryServiceHealthCheckCustomConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsServiceDiscoveryServiceHealthCheckCustomConfigList = DataAwsServiceDiscoveryServiceHealthCheckCustomConfigList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsServiceDiscoveryServiceHealthCheckCustomConfigList[_h] = { fqn: "@cdktf/provider-aws.dataAwsServiceDiscoveryService.DataAwsServiceDiscoveryServiceHealthCheckCustomConfigList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/service_discovery_service aws_service_discovery_service}
*/
class DataAwsServiceDiscoveryService extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/service_discovery_service aws_service_discovery_service} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsServiceDiscoveryServiceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_service_discovery_service',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // dns_config - computed: true, optional: false, required: false
        this._dnsConfig = new DataAwsServiceDiscoveryServiceDnsConfigList(this, "dns_config", false);
        // health_check_config - computed: true, optional: false, required: false
        this._healthCheckConfig = new DataAwsServiceDiscoveryServiceHealthCheckConfigList(this, "health_check_config", false);
        // health_check_custom_config - computed: true, optional: false, required: false
        this._healthCheckCustomConfig = new DataAwsServiceDiscoveryServiceHealthCheckCustomConfigList(this, "health_check_custom_config", false);
        this._id = config.id;
        this._name = config.name;
        this._namespaceId = config.namespaceId;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get dnsConfig() {
        return this._dnsConfig;
    }
    get healthCheckConfig() {
        return this._healthCheckConfig;
    }
    get healthCheckCustomConfig() {
        return this._healthCheckCustomConfig;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get namespaceId() {
        return this.getStringAttribute('namespace_id');
    }
    set namespaceId(value) {
        this._namespaceId = value;
    }
    // Temporarily expose input value. Use with caution.
    get namespaceIdInput() {
        return this._namespaceId;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            namespace_id: cdktf.stringToTerraform(this._namespaceId),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
        };
    }
}
exports.DataAwsServiceDiscoveryService = DataAwsServiceDiscoveryService;
_j = JSII_RTTI_SYMBOL_1;
DataAwsServiceDiscoveryService[_j] = { fqn: "@cdktf/provider-aws.dataAwsServiceDiscoveryService.DataAwsServiceDiscoveryService", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsServiceDiscoveryService.tfResourceType = "aws_service_discovery_service";
