// @cdktf/provider-aws/lib/connect-user-hierarchy-structure/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectUserHierarchyStructure = exports.ConnectUserHierarchyStructureHierarchyStructureOutputReference = exports.connectUserHierarchyStructureHierarchyStructureToTerraform = exports.ConnectUserHierarchyStructureHierarchyStructureLevelTwoOutputReference = exports.connectUserHierarchyStructureHierarchyStructureLevelTwoToTerraform = exports.ConnectUserHierarchyStructureHierarchyStructureLevelThreeOutputReference = exports.connectUserHierarchyStructureHierarchyStructureLevelThreeToTerraform = exports.ConnectUserHierarchyStructureHierarchyStructureLevelOneOutputReference = exports.connectUserHierarchyStructureHierarchyStructureLevelOneToTerraform = exports.ConnectUserHierarchyStructureHierarchyStructureLevelFourOutputReference = exports.connectUserHierarchyStructureHierarchyStructureLevelFourToTerraform = exports.ConnectUserHierarchyStructureHierarchyStructureLevelFiveOutputReference = exports.connectUserHierarchyStructureHierarchyStructureLevelFiveToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function connectUserHierarchyStructureHierarchyStructureLevelFiveToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.connectUserHierarchyStructureHierarchyStructureLevelFiveToTerraform = connectUserHierarchyStructureHierarchyStructureLevelFiveToTerraform;
class ConnectUserHierarchyStructureHierarchyStructureLevelFiveOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._name = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._name = value.name;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
}
exports.ConnectUserHierarchyStructureHierarchyStructureLevelFiveOutputReference = ConnectUserHierarchyStructureHierarchyStructureLevelFiveOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ConnectUserHierarchyStructureHierarchyStructureLevelFiveOutputReference[_a] = { fqn: "@cdktf/provider-aws.connectUserHierarchyStructure.ConnectUserHierarchyStructureHierarchyStructureLevelFiveOutputReference", version: "10.0.12" };
function connectUserHierarchyStructureHierarchyStructureLevelFourToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.connectUserHierarchyStructureHierarchyStructureLevelFourToTerraform = connectUserHierarchyStructureHierarchyStructureLevelFourToTerraform;
class ConnectUserHierarchyStructureHierarchyStructureLevelFourOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._name = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._name = value.name;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
}
exports.ConnectUserHierarchyStructureHierarchyStructureLevelFourOutputReference = ConnectUserHierarchyStructureHierarchyStructureLevelFourOutputReference;
_b = JSII_RTTI_SYMBOL_1;
ConnectUserHierarchyStructureHierarchyStructureLevelFourOutputReference[_b] = { fqn: "@cdktf/provider-aws.connectUserHierarchyStructure.ConnectUserHierarchyStructureHierarchyStructureLevelFourOutputReference", version: "10.0.12" };
function connectUserHierarchyStructureHierarchyStructureLevelOneToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.connectUserHierarchyStructureHierarchyStructureLevelOneToTerraform = connectUserHierarchyStructureHierarchyStructureLevelOneToTerraform;
class ConnectUserHierarchyStructureHierarchyStructureLevelOneOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._name = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._name = value.name;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
}
exports.ConnectUserHierarchyStructureHierarchyStructureLevelOneOutputReference = ConnectUserHierarchyStructureHierarchyStructureLevelOneOutputReference;
_c = JSII_RTTI_SYMBOL_1;
ConnectUserHierarchyStructureHierarchyStructureLevelOneOutputReference[_c] = { fqn: "@cdktf/provider-aws.connectUserHierarchyStructure.ConnectUserHierarchyStructureHierarchyStructureLevelOneOutputReference", version: "10.0.12" };
function connectUserHierarchyStructureHierarchyStructureLevelThreeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.connectUserHierarchyStructureHierarchyStructureLevelThreeToTerraform = connectUserHierarchyStructureHierarchyStructureLevelThreeToTerraform;
class ConnectUserHierarchyStructureHierarchyStructureLevelThreeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._name = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._name = value.name;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
}
exports.ConnectUserHierarchyStructureHierarchyStructureLevelThreeOutputReference = ConnectUserHierarchyStructureHierarchyStructureLevelThreeOutputReference;
_d = JSII_RTTI_SYMBOL_1;
ConnectUserHierarchyStructureHierarchyStructureLevelThreeOutputReference[_d] = { fqn: "@cdktf/provider-aws.connectUserHierarchyStructure.ConnectUserHierarchyStructureHierarchyStructureLevelThreeOutputReference", version: "10.0.12" };
function connectUserHierarchyStructureHierarchyStructureLevelTwoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.connectUserHierarchyStructureHierarchyStructureLevelTwoToTerraform = connectUserHierarchyStructureHierarchyStructureLevelTwoToTerraform;
class ConnectUserHierarchyStructureHierarchyStructureLevelTwoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._name = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._name = value.name;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
}
exports.ConnectUserHierarchyStructureHierarchyStructureLevelTwoOutputReference = ConnectUserHierarchyStructureHierarchyStructureLevelTwoOutputReference;
_e = JSII_RTTI_SYMBOL_1;
ConnectUserHierarchyStructureHierarchyStructureLevelTwoOutputReference[_e] = { fqn: "@cdktf/provider-aws.connectUserHierarchyStructure.ConnectUserHierarchyStructureHierarchyStructureLevelTwoOutputReference", version: "10.0.12" };
function connectUserHierarchyStructureHierarchyStructureToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        level_five: connectUserHierarchyStructureHierarchyStructureLevelFiveToTerraform(struct.levelFive),
        level_four: connectUserHierarchyStructureHierarchyStructureLevelFourToTerraform(struct.levelFour),
        level_one: connectUserHierarchyStructureHierarchyStructureLevelOneToTerraform(struct.levelOne),
        level_three: connectUserHierarchyStructureHierarchyStructureLevelThreeToTerraform(struct.levelThree),
        level_two: connectUserHierarchyStructureHierarchyStructureLevelTwoToTerraform(struct.levelTwo),
    };
}
exports.connectUserHierarchyStructureHierarchyStructureToTerraform = connectUserHierarchyStructureHierarchyStructureToTerraform;
class ConnectUserHierarchyStructureHierarchyStructureOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // level_five - computed: false, optional: true, required: false
        this._levelFive = new ConnectUserHierarchyStructureHierarchyStructureLevelFiveOutputReference(this, "level_five");
        // level_four - computed: false, optional: true, required: false
        this._levelFour = new ConnectUserHierarchyStructureHierarchyStructureLevelFourOutputReference(this, "level_four");
        // level_one - computed: false, optional: true, required: false
        this._levelOne = new ConnectUserHierarchyStructureHierarchyStructureLevelOneOutputReference(this, "level_one");
        // level_three - computed: false, optional: true, required: false
        this._levelThree = new ConnectUserHierarchyStructureHierarchyStructureLevelThreeOutputReference(this, "level_three");
        // level_two - computed: false, optional: true, required: false
        this._levelTwo = new ConnectUserHierarchyStructureHierarchyStructureLevelTwoOutputReference(this, "level_two");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._levelFive?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.levelFive = this._levelFive?.internalValue;
        }
        if (this._levelFour?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.levelFour = this._levelFour?.internalValue;
        }
        if (this._levelOne?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.levelOne = this._levelOne?.internalValue;
        }
        if (this._levelThree?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.levelThree = this._levelThree?.internalValue;
        }
        if (this._levelTwo?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.levelTwo = this._levelTwo?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._levelFive.internalValue = undefined;
            this._levelFour.internalValue = undefined;
            this._levelOne.internalValue = undefined;
            this._levelThree.internalValue = undefined;
            this._levelTwo.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._levelFive.internalValue = value.levelFive;
            this._levelFour.internalValue = value.levelFour;
            this._levelOne.internalValue = value.levelOne;
            this._levelThree.internalValue = value.levelThree;
            this._levelTwo.internalValue = value.levelTwo;
        }
    }
    get levelFive() {
        return this._levelFive;
    }
    putLevelFive(value) {
        this._levelFive.internalValue = value;
    }
    resetLevelFive() {
        this._levelFive.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get levelFiveInput() {
        return this._levelFive.internalValue;
    }
    get levelFour() {
        return this._levelFour;
    }
    putLevelFour(value) {
        this._levelFour.internalValue = value;
    }
    resetLevelFour() {
        this._levelFour.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get levelFourInput() {
        return this._levelFour.internalValue;
    }
    get levelOne() {
        return this._levelOne;
    }
    putLevelOne(value) {
        this._levelOne.internalValue = value;
    }
    resetLevelOne() {
        this._levelOne.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get levelOneInput() {
        return this._levelOne.internalValue;
    }
    get levelThree() {
        return this._levelThree;
    }
    putLevelThree(value) {
        this._levelThree.internalValue = value;
    }
    resetLevelThree() {
        this._levelThree.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get levelThreeInput() {
        return this._levelThree.internalValue;
    }
    get levelTwo() {
        return this._levelTwo;
    }
    putLevelTwo(value) {
        this._levelTwo.internalValue = value;
    }
    resetLevelTwo() {
        this._levelTwo.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get levelTwoInput() {
        return this._levelTwo.internalValue;
    }
}
exports.ConnectUserHierarchyStructureHierarchyStructureOutputReference = ConnectUserHierarchyStructureHierarchyStructureOutputReference;
_f = JSII_RTTI_SYMBOL_1;
ConnectUserHierarchyStructureHierarchyStructureOutputReference[_f] = { fqn: "@cdktf/provider-aws.connectUserHierarchyStructure.ConnectUserHierarchyStructureHierarchyStructureOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/connect_user_hierarchy_structure aws_connect_user_hierarchy_structure}
*/
class ConnectUserHierarchyStructure extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/connect_user_hierarchy_structure aws_connect_user_hierarchy_structure} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ConnectUserHierarchyStructureConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_connect_user_hierarchy_structure',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // hierarchy_structure - computed: false, optional: false, required: true
        this._hierarchyStructure = new ConnectUserHierarchyStructureHierarchyStructureOutputReference(this, "hierarchy_structure");
        this._id = config.id;
        this._instanceId = config.instanceId;
        this._hierarchyStructure.internalValue = config.hierarchyStructure;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instanceId() {
        return this.getStringAttribute('instance_id');
    }
    set instanceId(value) {
        this._instanceId = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceIdInput() {
        return this._instanceId;
    }
    get hierarchyStructure() {
        return this._hierarchyStructure;
    }
    putHierarchyStructure(value) {
        this._hierarchyStructure.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get hierarchyStructureInput() {
        return this._hierarchyStructure.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            instance_id: cdktf.stringToTerraform(this._instanceId),
            hierarchy_structure: connectUserHierarchyStructureHierarchyStructureToTerraform(this._hierarchyStructure.internalValue),
        };
    }
}
exports.ConnectUserHierarchyStructure = ConnectUserHierarchyStructure;
_g = JSII_RTTI_SYMBOL_1;
ConnectUserHierarchyStructure[_g] = { fqn: "@cdktf/provider-aws.connectUserHierarchyStructure.ConnectUserHierarchyStructure", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
ConnectUserHierarchyStructure.tfResourceType = "aws_connect_user_hierarchy_structure";
