// @cdktf/provider-aws/lib/emrcontainers-virtual-cluster/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmrcontainersVirtualCluster = exports.EmrcontainersVirtualClusterTimeoutsOutputReference = exports.emrcontainersVirtualClusterTimeoutsToTerraform = exports.EmrcontainersVirtualClusterContainerProviderOutputReference = exports.emrcontainersVirtualClusterContainerProviderToTerraform = exports.EmrcontainersVirtualClusterContainerProviderInfoOutputReference = exports.emrcontainersVirtualClusterContainerProviderInfoToTerraform = exports.EmrcontainersVirtualClusterContainerProviderInfoEksInfoOutputReference = exports.emrcontainersVirtualClusterContainerProviderInfoEksInfoToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function emrcontainersVirtualClusterContainerProviderInfoEksInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        namespace: cdktf.stringToTerraform(struct.namespace),
    };
}
exports.emrcontainersVirtualClusterContainerProviderInfoEksInfoToTerraform = emrcontainersVirtualClusterContainerProviderInfoEksInfoToTerraform;
class EmrcontainersVirtualClusterContainerProviderInfoEksInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._namespace !== undefined) {
            hasAnyValues = true;
            internalValueResult.namespace = this._namespace;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._namespace = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._namespace = value.namespace;
        }
    }
    get namespace() {
        return this.getStringAttribute('namespace');
    }
    set namespace(value) {
        this._namespace = value;
    }
    resetNamespace() {
        this._namespace = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get namespaceInput() {
        return this._namespace;
    }
}
exports.EmrcontainersVirtualClusterContainerProviderInfoEksInfoOutputReference = EmrcontainersVirtualClusterContainerProviderInfoEksInfoOutputReference;
_a = JSII_RTTI_SYMBOL_1;
EmrcontainersVirtualClusterContainerProviderInfoEksInfoOutputReference[_a] = { fqn: "@cdktf/provider-aws.emrcontainersVirtualCluster.EmrcontainersVirtualClusterContainerProviderInfoEksInfoOutputReference", version: "10.0.12" };
function emrcontainersVirtualClusterContainerProviderInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        eks_info: emrcontainersVirtualClusterContainerProviderInfoEksInfoToTerraform(struct.eksInfo),
    };
}
exports.emrcontainersVirtualClusterContainerProviderInfoToTerraform = emrcontainersVirtualClusterContainerProviderInfoToTerraform;
class EmrcontainersVirtualClusterContainerProviderInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // eks_info - computed: false, optional: false, required: true
        this._eksInfo = new EmrcontainersVirtualClusterContainerProviderInfoEksInfoOutputReference(this, "eks_info");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._eksInfo?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.eksInfo = this._eksInfo?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._eksInfo.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._eksInfo.internalValue = value.eksInfo;
        }
    }
    get eksInfo() {
        return this._eksInfo;
    }
    putEksInfo(value) {
        this._eksInfo.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get eksInfoInput() {
        return this._eksInfo.internalValue;
    }
}
exports.EmrcontainersVirtualClusterContainerProviderInfoOutputReference = EmrcontainersVirtualClusterContainerProviderInfoOutputReference;
_b = JSII_RTTI_SYMBOL_1;
EmrcontainersVirtualClusterContainerProviderInfoOutputReference[_b] = { fqn: "@cdktf/provider-aws.emrcontainersVirtualCluster.EmrcontainersVirtualClusterContainerProviderInfoOutputReference", version: "10.0.12" };
function emrcontainersVirtualClusterContainerProviderToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        id: cdktf.stringToTerraform(struct.id),
        type: cdktf.stringToTerraform(struct.type),
        info: emrcontainersVirtualClusterContainerProviderInfoToTerraform(struct.info),
    };
}
exports.emrcontainersVirtualClusterContainerProviderToTerraform = emrcontainersVirtualClusterContainerProviderToTerraform;
class EmrcontainersVirtualClusterContainerProviderOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // info - computed: false, optional: false, required: true
        this._info = new EmrcontainersVirtualClusterContainerProviderInfoOutputReference(this, "info");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._id !== undefined) {
            hasAnyValues = true;
            internalValueResult.id = this._id;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._info?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.info = this._info?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._id = undefined;
            this._type = undefined;
            this._info.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._id = value.id;
            this._type = value.type;
            this._info.internalValue = value.info;
        }
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get info() {
        return this._info;
    }
    putInfo(value) {
        this._info.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get infoInput() {
        return this._info.internalValue;
    }
}
exports.EmrcontainersVirtualClusterContainerProviderOutputReference = EmrcontainersVirtualClusterContainerProviderOutputReference;
_c = JSII_RTTI_SYMBOL_1;
EmrcontainersVirtualClusterContainerProviderOutputReference[_c] = { fqn: "@cdktf/provider-aws.emrcontainersVirtualCluster.EmrcontainersVirtualClusterContainerProviderOutputReference", version: "10.0.12" };
function emrcontainersVirtualClusterTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.emrcontainersVirtualClusterTimeoutsToTerraform = emrcontainersVirtualClusterTimeoutsToTerraform;
class EmrcontainersVirtualClusterTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._delete = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._delete = value.delete;
        }
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
}
exports.EmrcontainersVirtualClusterTimeoutsOutputReference = EmrcontainersVirtualClusterTimeoutsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
EmrcontainersVirtualClusterTimeoutsOutputReference[_d] = { fqn: "@cdktf/provider-aws.emrcontainersVirtualCluster.EmrcontainersVirtualClusterTimeoutsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/emrcontainers_virtual_cluster aws_emrcontainers_virtual_cluster}
*/
class EmrcontainersVirtualCluster extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/emrcontainers_virtual_cluster aws_emrcontainers_virtual_cluster} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options EmrcontainersVirtualClusterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_emrcontainers_virtual_cluster',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // container_provider - computed: false, optional: false, required: true
        this._containerProvider = new EmrcontainersVirtualClusterContainerProviderOutputReference(this, "container_provider");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new EmrcontainersVirtualClusterTimeoutsOutputReference(this, "timeouts");
        this._id = config.id;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._containerProvider.internalValue = config.containerProvider;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get containerProvider() {
        return this._containerProvider;
    }
    putContainerProvider(value) {
        this._containerProvider.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get containerProviderInput() {
        return this._containerProvider.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            container_provider: emrcontainersVirtualClusterContainerProviderToTerraform(this._containerProvider.internalValue),
            timeouts: emrcontainersVirtualClusterTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.EmrcontainersVirtualCluster = EmrcontainersVirtualCluster;
_e = JSII_RTTI_SYMBOL_1;
EmrcontainersVirtualCluster[_e] = { fqn: "@cdktf/provider-aws.emrcontainersVirtualCluster.EmrcontainersVirtualCluster", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
EmrcontainersVirtualCluster.tfResourceType = "aws_emrcontainers_virtual_cluster";
