// @cdktf/provider-aws/lib/data-aws-secretsmanager-secret-version/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsSecretsmanagerSecretVersion = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/secretsmanager_secret_version aws_secretsmanager_secret_version}
*/
class DataAwsSecretsmanagerSecretVersion extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/secretsmanager_secret_version aws_secretsmanager_secret_version} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsSecretsmanagerSecretVersionConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_secretsmanager_secret_version',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._secretId = config.secretId;
        this._versionId = config.versionId;
        this._versionStage = config.versionStage;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // secret_binary - computed: true, optional: false, required: false
    get secretBinary() {
        return this.getStringAttribute('secret_binary');
    }
    get secretId() {
        return this.getStringAttribute('secret_id');
    }
    set secretId(value) {
        this._secretId = value;
    }
    // Temporarily expose input value. Use with caution.
    get secretIdInput() {
        return this._secretId;
    }
    // secret_string - computed: true, optional: false, required: false
    get secretString() {
        return this.getStringAttribute('secret_string');
    }
    get versionId() {
        return this.getStringAttribute('version_id');
    }
    set versionId(value) {
        this._versionId = value;
    }
    resetVersionId() {
        this._versionId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionIdInput() {
        return this._versionId;
    }
    get versionStage() {
        return this.getStringAttribute('version_stage');
    }
    set versionStage(value) {
        this._versionStage = value;
    }
    resetVersionStage() {
        this._versionStage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionStageInput() {
        return this._versionStage;
    }
    // version_stages - computed: true, optional: false, required: false
    get versionStages() {
        return cdktf.Fn.tolist(this.getListAttribute('version_stages'));
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            secret_id: cdktf.stringToTerraform(this._secretId),
            version_id: cdktf.stringToTerraform(this._versionId),
            version_stage: cdktf.stringToTerraform(this._versionStage),
        };
    }
}
exports.DataAwsSecretsmanagerSecretVersion = DataAwsSecretsmanagerSecretVersion;
_a = JSII_RTTI_SYMBOL_1;
DataAwsSecretsmanagerSecretVersion[_a] = { fqn: "@cdktf/provider-aws.dataAwsSecretsmanagerSecretVersion.DataAwsSecretsmanagerSecretVersion", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsSecretsmanagerSecretVersion.tfResourceType = "aws_secretsmanager_secret_version";
