// @cdktf/provider-aws/lib/data-aws-cur-report-definition/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsCurReportDefinition = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/cur_report_definition aws_cur_report_definition}
*/
class DataAwsCurReportDefinition extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/cur_report_definition aws_cur_report_definition} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsCurReportDefinitionConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cur_report_definition',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._reportName = config.reportName;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // additional_artifacts - computed: true, optional: false, required: false
    get additionalArtifacts() {
        return cdktf.Fn.tolist(this.getListAttribute('additional_artifacts'));
    }
    // additional_schema_elements - computed: true, optional: false, required: false
    get additionalSchemaElements() {
        return cdktf.Fn.tolist(this.getListAttribute('additional_schema_elements'));
    }
    // compression - computed: true, optional: false, required: false
    get compression() {
        return this.getStringAttribute('compression');
    }
    // format - computed: true, optional: false, required: false
    get format() {
        return this.getStringAttribute('format');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // refresh_closed_reports - computed: true, optional: false, required: false
    get refreshClosedReports() {
        return this.getBooleanAttribute('refresh_closed_reports');
    }
    get reportName() {
        return this.getStringAttribute('report_name');
    }
    set reportName(value) {
        this._reportName = value;
    }
    // Temporarily expose input value. Use with caution.
    get reportNameInput() {
        return this._reportName;
    }
    // report_versioning - computed: true, optional: false, required: false
    get reportVersioning() {
        return this.getStringAttribute('report_versioning');
    }
    // s3_bucket - computed: true, optional: false, required: false
    get s3Bucket() {
        return this.getStringAttribute('s3_bucket');
    }
    // s3_prefix - computed: true, optional: false, required: false
    get s3Prefix() {
        return this.getStringAttribute('s3_prefix');
    }
    // s3_region - computed: true, optional: false, required: false
    get s3Region() {
        return this.getStringAttribute('s3_region');
    }
    // time_unit - computed: true, optional: false, required: false
    get timeUnit() {
        return this.getStringAttribute('time_unit');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            report_name: cdktf.stringToTerraform(this._reportName),
        };
    }
}
exports.DataAwsCurReportDefinition = DataAwsCurReportDefinition;
_a = JSII_RTTI_SYMBOL_1;
DataAwsCurReportDefinition[_a] = { fqn: "@cdktf/provider-aws.dataAwsCurReportDefinition.DataAwsCurReportDefinition", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsCurReportDefinition.tfResourceType = "aws_cur_report_definition";
