// @cdktf/provider-aws/lib/networkmanager-attachment-accepter/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkmanagerAttachmentAccepter = exports.NetworkmanagerAttachmentAccepterTimeoutsOutputReference = exports.networkmanagerAttachmentAccepterTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function networkmanagerAttachmentAccepterTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
    };
}
exports.networkmanagerAttachmentAccepterTimeoutsToTerraform = networkmanagerAttachmentAccepterTimeoutsToTerraform;
class NetworkmanagerAttachmentAccepterTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
}
exports.NetworkmanagerAttachmentAccepterTimeoutsOutputReference = NetworkmanagerAttachmentAccepterTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
NetworkmanagerAttachmentAccepterTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.networkmanagerAttachmentAccepter.NetworkmanagerAttachmentAccepterTimeoutsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/networkmanager_attachment_accepter aws_networkmanager_attachment_accepter}
*/
class NetworkmanagerAttachmentAccepter extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/networkmanager_attachment_accepter aws_networkmanager_attachment_accepter} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options NetworkmanagerAttachmentAccepterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_networkmanager_attachment_accepter',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new NetworkmanagerAttachmentAccepterTimeoutsOutputReference(this, "timeouts");
        this._attachmentId = config.attachmentId;
        this._attachmentType = config.attachmentType;
        this._id = config.id;
        this._timeouts.internalValue = config.timeouts;
    }
    get attachmentId() {
        return this.getStringAttribute('attachment_id');
    }
    set attachmentId(value) {
        this._attachmentId = value;
    }
    // Temporarily expose input value. Use with caution.
    get attachmentIdInput() {
        return this._attachmentId;
    }
    // attachment_policy_rule_number - computed: true, optional: false, required: false
    get attachmentPolicyRuleNumber() {
        return this.getNumberAttribute('attachment_policy_rule_number');
    }
    get attachmentType() {
        return this.getStringAttribute('attachment_type');
    }
    set attachmentType(value) {
        this._attachmentType = value;
    }
    // Temporarily expose input value. Use with caution.
    get attachmentTypeInput() {
        return this._attachmentType;
    }
    // core_network_arn - computed: true, optional: false, required: false
    get coreNetworkArn() {
        return this.getStringAttribute('core_network_arn');
    }
    // core_network_id - computed: true, optional: false, required: false
    get coreNetworkId() {
        return this.getStringAttribute('core_network_id');
    }
    // edge_location - computed: true, optional: false, required: false
    get edgeLocation() {
        return this.getStringAttribute('edge_location');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // owner_account_id - computed: true, optional: false, required: false
    get ownerAccountId() {
        return this.getStringAttribute('owner_account_id');
    }
    // resource_arn - computed: true, optional: false, required: false
    get resourceArn() {
        return this.getStringAttribute('resource_arn');
    }
    // segment_name - computed: true, optional: false, required: false
    get segmentName() {
        return this.getStringAttribute('segment_name');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            attachment_id: cdktf.stringToTerraform(this._attachmentId),
            attachment_type: cdktf.stringToTerraform(this._attachmentType),
            id: cdktf.stringToTerraform(this._id),
            timeouts: networkmanagerAttachmentAccepterTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.NetworkmanagerAttachmentAccepter = NetworkmanagerAttachmentAccepter;
_b = JSII_RTTI_SYMBOL_1;
NetworkmanagerAttachmentAccepter[_b] = { fqn: "@cdktf/provider-aws.networkmanagerAttachmentAccepter.NetworkmanagerAttachmentAccepter", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
NetworkmanagerAttachmentAccepter.tfResourceType = "aws_networkmanager_attachment_accepter";
