// @cdktf/provider-aws/lib/redshiftserverless-usage-limit/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedshiftserverlessUsageLimit = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/redshiftserverless_usage_limit aws_redshiftserverless_usage_limit}
*/
class RedshiftserverlessUsageLimit extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/redshiftserverless_usage_limit aws_redshiftserverless_usage_limit} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options RedshiftserverlessUsageLimitConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_redshiftserverless_usage_limit',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._amount = config.amount;
        this._breachAction = config.breachAction;
        this._id = config.id;
        this._period = config.period;
        this._resourceArn = config.resourceArn;
        this._usageType = config.usageType;
    }
    get amount() {
        return this.getNumberAttribute('amount');
    }
    set amount(value) {
        this._amount = value;
    }
    // Temporarily expose input value. Use with caution.
    get amountInput() {
        return this._amount;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get breachAction() {
        return this.getStringAttribute('breach_action');
    }
    set breachAction(value) {
        this._breachAction = value;
    }
    resetBreachAction() {
        this._breachAction = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get breachActionInput() {
        return this._breachAction;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get period() {
        return this.getStringAttribute('period');
    }
    set period(value) {
        this._period = value;
    }
    resetPeriod() {
        this._period = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get periodInput() {
        return this._period;
    }
    get resourceArn() {
        return this.getStringAttribute('resource_arn');
    }
    set resourceArn(value) {
        this._resourceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourceArnInput() {
        return this._resourceArn;
    }
    get usageType() {
        return this.getStringAttribute('usage_type');
    }
    set usageType(value) {
        this._usageType = value;
    }
    // Temporarily expose input value. Use with caution.
    get usageTypeInput() {
        return this._usageType;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            amount: cdktf.numberToTerraform(this._amount),
            breach_action: cdktf.stringToTerraform(this._breachAction),
            id: cdktf.stringToTerraform(this._id),
            period: cdktf.stringToTerraform(this._period),
            resource_arn: cdktf.stringToTerraform(this._resourceArn),
            usage_type: cdktf.stringToTerraform(this._usageType),
        };
    }
}
exports.RedshiftserverlessUsageLimit = RedshiftserverlessUsageLimit;
_a = JSII_RTTI_SYMBOL_1;
RedshiftserverlessUsageLimit[_a] = { fqn: "@cdktf/provider-aws.redshiftserverlessUsageLimit.RedshiftserverlessUsageLimit", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
RedshiftserverlessUsageLimit.tfResourceType = "aws_redshiftserverless_usage_limit";
