// @cdktf/provider-aws/lib/backup-report-plan/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BackupReportPlan = exports.BackupReportPlanReportSettingOutputReference = exports.backupReportPlanReportSettingToTerraform = exports.BackupReportPlanReportDeliveryChannelOutputReference = exports.backupReportPlanReportDeliveryChannelToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function backupReportPlanReportDeliveryChannelToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        formats: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.formats),
        s3_bucket_name: cdktf.stringToTerraform(struct.s3BucketName),
        s3_key_prefix: cdktf.stringToTerraform(struct.s3KeyPrefix),
    };
}
exports.backupReportPlanReportDeliveryChannelToTerraform = backupReportPlanReportDeliveryChannelToTerraform;
class BackupReportPlanReportDeliveryChannelOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._formats !== undefined) {
            hasAnyValues = true;
            internalValueResult.formats = this._formats;
        }
        if (this._s3BucketName !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3BucketName = this._s3BucketName;
        }
        if (this._s3KeyPrefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3KeyPrefix = this._s3KeyPrefix;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._formats = undefined;
            this._s3BucketName = undefined;
            this._s3KeyPrefix = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._formats = value.formats;
            this._s3BucketName = value.s3BucketName;
            this._s3KeyPrefix = value.s3KeyPrefix;
        }
    }
    get formats() {
        return cdktf.Fn.tolist(this.getListAttribute('formats'));
    }
    set formats(value) {
        this._formats = value;
    }
    resetFormats() {
        this._formats = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get formatsInput() {
        return this._formats;
    }
    get s3BucketName() {
        return this.getStringAttribute('s3_bucket_name');
    }
    set s3BucketName(value) {
        this._s3BucketName = value;
    }
    // Temporarily expose input value. Use with caution.
    get s3BucketNameInput() {
        return this._s3BucketName;
    }
    get s3KeyPrefix() {
        return this.getStringAttribute('s3_key_prefix');
    }
    set s3KeyPrefix(value) {
        this._s3KeyPrefix = value;
    }
    resetS3KeyPrefix() {
        this._s3KeyPrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get s3KeyPrefixInput() {
        return this._s3KeyPrefix;
    }
}
exports.BackupReportPlanReportDeliveryChannelOutputReference = BackupReportPlanReportDeliveryChannelOutputReference;
_a = JSII_RTTI_SYMBOL_1;
BackupReportPlanReportDeliveryChannelOutputReference[_a] = { fqn: "@cdktf/provider-aws.backupReportPlan.BackupReportPlanReportDeliveryChannelOutputReference", version: "10.0.12" };
function backupReportPlanReportSettingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        framework_arns: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.frameworkArns),
        number_of_frameworks: cdktf.numberToTerraform(struct.numberOfFrameworks),
        report_template: cdktf.stringToTerraform(struct.reportTemplate),
    };
}
exports.backupReportPlanReportSettingToTerraform = backupReportPlanReportSettingToTerraform;
class BackupReportPlanReportSettingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._frameworkArns !== undefined) {
            hasAnyValues = true;
            internalValueResult.frameworkArns = this._frameworkArns;
        }
        if (this._numberOfFrameworks !== undefined) {
            hasAnyValues = true;
            internalValueResult.numberOfFrameworks = this._numberOfFrameworks;
        }
        if (this._reportTemplate !== undefined) {
            hasAnyValues = true;
            internalValueResult.reportTemplate = this._reportTemplate;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._frameworkArns = undefined;
            this._numberOfFrameworks = undefined;
            this._reportTemplate = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._frameworkArns = value.frameworkArns;
            this._numberOfFrameworks = value.numberOfFrameworks;
            this._reportTemplate = value.reportTemplate;
        }
    }
    get frameworkArns() {
        return cdktf.Fn.tolist(this.getListAttribute('framework_arns'));
    }
    set frameworkArns(value) {
        this._frameworkArns = value;
    }
    resetFrameworkArns() {
        this._frameworkArns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get frameworkArnsInput() {
        return this._frameworkArns;
    }
    get numberOfFrameworks() {
        return this.getNumberAttribute('number_of_frameworks');
    }
    set numberOfFrameworks(value) {
        this._numberOfFrameworks = value;
    }
    resetNumberOfFrameworks() {
        this._numberOfFrameworks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get numberOfFrameworksInput() {
        return this._numberOfFrameworks;
    }
    get reportTemplate() {
        return this.getStringAttribute('report_template');
    }
    set reportTemplate(value) {
        this._reportTemplate = value;
    }
    // Temporarily expose input value. Use with caution.
    get reportTemplateInput() {
        return this._reportTemplate;
    }
}
exports.BackupReportPlanReportSettingOutputReference = BackupReportPlanReportSettingOutputReference;
_b = JSII_RTTI_SYMBOL_1;
BackupReportPlanReportSettingOutputReference[_b] = { fqn: "@cdktf/provider-aws.backupReportPlan.BackupReportPlanReportSettingOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/backup_report_plan aws_backup_report_plan}
*/
class BackupReportPlan extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/backup_report_plan aws_backup_report_plan} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options BackupReportPlanConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_backup_report_plan',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // report_delivery_channel - computed: false, optional: false, required: true
        this._reportDeliveryChannel = new BackupReportPlanReportDeliveryChannelOutputReference(this, "report_delivery_channel");
        // report_setting - computed: false, optional: false, required: true
        this._reportSetting = new BackupReportPlanReportSettingOutputReference(this, "report_setting");
        this._description = config.description;
        this._id = config.id;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._reportDeliveryChannel.internalValue = config.reportDeliveryChannel;
        this._reportSetting.internalValue = config.reportSetting;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // creation_time - computed: true, optional: false, required: false
    get creationTime() {
        return this.getStringAttribute('creation_time');
    }
    // deployment_status - computed: true, optional: false, required: false
    get deploymentStatus() {
        return this.getStringAttribute('deployment_status');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get reportDeliveryChannel() {
        return this._reportDeliveryChannel;
    }
    putReportDeliveryChannel(value) {
        this._reportDeliveryChannel.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get reportDeliveryChannelInput() {
        return this._reportDeliveryChannel.internalValue;
    }
    get reportSetting() {
        return this._reportSetting;
    }
    putReportSetting(value) {
        this._reportSetting.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get reportSettingInput() {
        return this._reportSetting.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            report_delivery_channel: backupReportPlanReportDeliveryChannelToTerraform(this._reportDeliveryChannel.internalValue),
            report_setting: backupReportPlanReportSettingToTerraform(this._reportSetting.internalValue),
        };
    }
}
exports.BackupReportPlan = BackupReportPlan;
_c = JSII_RTTI_SYMBOL_1;
BackupReportPlan[_c] = { fqn: "@cdktf/provider-aws.backupReportPlan.BackupReportPlan", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
BackupReportPlan.tfResourceType = "aws_backup_report_plan";
