// @cdktf/provider-aws/lib/route53domains-registered-domain/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Route53DomainsRegisteredDomain = exports.Route53DomainsRegisteredDomainTimeoutsOutputReference = exports.route53DomainsRegisteredDomainTimeoutsToTerraform = exports.Route53DomainsRegisteredDomainTechContactOutputReference = exports.route53DomainsRegisteredDomainTechContactToTerraform = exports.Route53DomainsRegisteredDomainRegistrantContactOutputReference = exports.route53DomainsRegisteredDomainRegistrantContactToTerraform = exports.Route53DomainsRegisteredDomainNameServerList = exports.Route53DomainsRegisteredDomainNameServerOutputReference = exports.route53DomainsRegisteredDomainNameServerToTerraform = exports.Route53DomainsRegisteredDomainAdminContactOutputReference = exports.route53DomainsRegisteredDomainAdminContactToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function route53DomainsRegisteredDomainAdminContactToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        address_line_1: cdktf.stringToTerraform(struct.addressLine1),
        address_line_2: cdktf.stringToTerraform(struct.addressLine2),
        city: cdktf.stringToTerraform(struct.city),
        contact_type: cdktf.stringToTerraform(struct.contactType),
        country_code: cdktf.stringToTerraform(struct.countryCode),
        email: cdktf.stringToTerraform(struct.email),
        extra_params: cdktf.hashMapper(cdktf.stringToTerraform)(struct.extraParams),
        fax: cdktf.stringToTerraform(struct.fax),
        first_name: cdktf.stringToTerraform(struct.firstName),
        last_name: cdktf.stringToTerraform(struct.lastName),
        organization_name: cdktf.stringToTerraform(struct.organizationName),
        phone_number: cdktf.stringToTerraform(struct.phoneNumber),
        state: cdktf.stringToTerraform(struct.state),
        zip_code: cdktf.stringToTerraform(struct.zipCode),
    };
}
exports.route53DomainsRegisteredDomainAdminContactToTerraform = route53DomainsRegisteredDomainAdminContactToTerraform;
class Route53DomainsRegisteredDomainAdminContactOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._addressLine1 !== undefined) {
            hasAnyValues = true;
            internalValueResult.addressLine1 = this._addressLine1;
        }
        if (this._addressLine2 !== undefined) {
            hasAnyValues = true;
            internalValueResult.addressLine2 = this._addressLine2;
        }
        if (this._city !== undefined) {
            hasAnyValues = true;
            internalValueResult.city = this._city;
        }
        if (this._contactType !== undefined) {
            hasAnyValues = true;
            internalValueResult.contactType = this._contactType;
        }
        if (this._countryCode !== undefined) {
            hasAnyValues = true;
            internalValueResult.countryCode = this._countryCode;
        }
        if (this._email !== undefined) {
            hasAnyValues = true;
            internalValueResult.email = this._email;
        }
        if (this._extraParams !== undefined) {
            hasAnyValues = true;
            internalValueResult.extraParams = this._extraParams;
        }
        if (this._fax !== undefined) {
            hasAnyValues = true;
            internalValueResult.fax = this._fax;
        }
        if (this._firstName !== undefined) {
            hasAnyValues = true;
            internalValueResult.firstName = this._firstName;
        }
        if (this._lastName !== undefined) {
            hasAnyValues = true;
            internalValueResult.lastName = this._lastName;
        }
        if (this._organizationName !== undefined) {
            hasAnyValues = true;
            internalValueResult.organizationName = this._organizationName;
        }
        if (this._phoneNumber !== undefined) {
            hasAnyValues = true;
            internalValueResult.phoneNumber = this._phoneNumber;
        }
        if (this._state !== undefined) {
            hasAnyValues = true;
            internalValueResult.state = this._state;
        }
        if (this._zipCode !== undefined) {
            hasAnyValues = true;
            internalValueResult.zipCode = this._zipCode;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._addressLine1 = undefined;
            this._addressLine2 = undefined;
            this._city = undefined;
            this._contactType = undefined;
            this._countryCode = undefined;
            this._email = undefined;
            this._extraParams = undefined;
            this._fax = undefined;
            this._firstName = undefined;
            this._lastName = undefined;
            this._organizationName = undefined;
            this._phoneNumber = undefined;
            this._state = undefined;
            this._zipCode = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._addressLine1 = value.addressLine1;
            this._addressLine2 = value.addressLine2;
            this._city = value.city;
            this._contactType = value.contactType;
            this._countryCode = value.countryCode;
            this._email = value.email;
            this._extraParams = value.extraParams;
            this._fax = value.fax;
            this._firstName = value.firstName;
            this._lastName = value.lastName;
            this._organizationName = value.organizationName;
            this._phoneNumber = value.phoneNumber;
            this._state = value.state;
            this._zipCode = value.zipCode;
        }
    }
    get addressLine1() {
        return this.getStringAttribute('address_line_1');
    }
    set addressLine1(value) {
        this._addressLine1 = value;
    }
    resetAddressLine1() {
        this._addressLine1 = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get addressLine1Input() {
        return this._addressLine1;
    }
    get addressLine2() {
        return this.getStringAttribute('address_line_2');
    }
    set addressLine2(value) {
        this._addressLine2 = value;
    }
    resetAddressLine2() {
        this._addressLine2 = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get addressLine2Input() {
        return this._addressLine2;
    }
    get city() {
        return this.getStringAttribute('city');
    }
    set city(value) {
        this._city = value;
    }
    resetCity() {
        this._city = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cityInput() {
        return this._city;
    }
    get contactType() {
        return this.getStringAttribute('contact_type');
    }
    set contactType(value) {
        this._contactType = value;
    }
    resetContactType() {
        this._contactType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get contactTypeInput() {
        return this._contactType;
    }
    get countryCode() {
        return this.getStringAttribute('country_code');
    }
    set countryCode(value) {
        this._countryCode = value;
    }
    resetCountryCode() {
        this._countryCode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get countryCodeInput() {
        return this._countryCode;
    }
    get email() {
        return this.getStringAttribute('email');
    }
    set email(value) {
        this._email = value;
    }
    resetEmail() {
        this._email = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get emailInput() {
        return this._email;
    }
    get extraParams() {
        return this.getStringMapAttribute('extra_params');
    }
    set extraParams(value) {
        this._extraParams = value;
    }
    resetExtraParams() {
        this._extraParams = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get extraParamsInput() {
        return this._extraParams;
    }
    get fax() {
        return this.getStringAttribute('fax');
    }
    set fax(value) {
        this._fax = value;
    }
    resetFax() {
        this._fax = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get faxInput() {
        return this._fax;
    }
    get firstName() {
        return this.getStringAttribute('first_name');
    }
    set firstName(value) {
        this._firstName = value;
    }
    resetFirstName() {
        this._firstName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get firstNameInput() {
        return this._firstName;
    }
    get lastName() {
        return this.getStringAttribute('last_name');
    }
    set lastName(value) {
        this._lastName = value;
    }
    resetLastName() {
        this._lastName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lastNameInput() {
        return this._lastName;
    }
    get organizationName() {
        return this.getStringAttribute('organization_name');
    }
    set organizationName(value) {
        this._organizationName = value;
    }
    resetOrganizationName() {
        this._organizationName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get organizationNameInput() {
        return this._organizationName;
    }
    get phoneNumber() {
        return this.getStringAttribute('phone_number');
    }
    set phoneNumber(value) {
        this._phoneNumber = value;
    }
    resetPhoneNumber() {
        this._phoneNumber = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get phoneNumberInput() {
        return this._phoneNumber;
    }
    get state() {
        return this.getStringAttribute('state');
    }
    set state(value) {
        this._state = value;
    }
    resetState() {
        this._state = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get stateInput() {
        return this._state;
    }
    get zipCode() {
        return this.getStringAttribute('zip_code');
    }
    set zipCode(value) {
        this._zipCode = value;
    }
    resetZipCode() {
        this._zipCode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zipCodeInput() {
        return this._zipCode;
    }
}
exports.Route53DomainsRegisteredDomainAdminContactOutputReference = Route53DomainsRegisteredDomainAdminContactOutputReference;
_a = JSII_RTTI_SYMBOL_1;
Route53DomainsRegisteredDomainAdminContactOutputReference[_a] = { fqn: "@cdktf/provider-aws.route53DomainsRegisteredDomain.Route53DomainsRegisteredDomainAdminContactOutputReference", version: "10.0.12" };
function route53DomainsRegisteredDomainNameServerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        glue_ips: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.glueIps),
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.route53DomainsRegisteredDomainNameServerToTerraform = route53DomainsRegisteredDomainNameServerToTerraform;
class Route53DomainsRegisteredDomainNameServerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._glueIps !== undefined) {
            hasAnyValues = true;
            internalValueResult.glueIps = this._glueIps;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._glueIps = undefined;
            this._name = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._glueIps = value.glueIps;
            this._name = value.name;
        }
    }
    get glueIps() {
        return cdktf.Fn.tolist(this.getListAttribute('glue_ips'));
    }
    set glueIps(value) {
        this._glueIps = value;
    }
    resetGlueIps() {
        this._glueIps = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get glueIpsInput() {
        return this._glueIps;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
}
exports.Route53DomainsRegisteredDomainNameServerOutputReference = Route53DomainsRegisteredDomainNameServerOutputReference;
_b = JSII_RTTI_SYMBOL_1;
Route53DomainsRegisteredDomainNameServerOutputReference[_b] = { fqn: "@cdktf/provider-aws.route53DomainsRegisteredDomain.Route53DomainsRegisteredDomainNameServerOutputReference", version: "10.0.12" };
class Route53DomainsRegisteredDomainNameServerList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new Route53DomainsRegisteredDomainNameServerOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.Route53DomainsRegisteredDomainNameServerList = Route53DomainsRegisteredDomainNameServerList;
_c = JSII_RTTI_SYMBOL_1;
Route53DomainsRegisteredDomainNameServerList[_c] = { fqn: "@cdktf/provider-aws.route53DomainsRegisteredDomain.Route53DomainsRegisteredDomainNameServerList", version: "10.0.12" };
function route53DomainsRegisteredDomainRegistrantContactToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        address_line_1: cdktf.stringToTerraform(struct.addressLine1),
        address_line_2: cdktf.stringToTerraform(struct.addressLine2),
        city: cdktf.stringToTerraform(struct.city),
        contact_type: cdktf.stringToTerraform(struct.contactType),
        country_code: cdktf.stringToTerraform(struct.countryCode),
        email: cdktf.stringToTerraform(struct.email),
        extra_params: cdktf.hashMapper(cdktf.stringToTerraform)(struct.extraParams),
        fax: cdktf.stringToTerraform(struct.fax),
        first_name: cdktf.stringToTerraform(struct.firstName),
        last_name: cdktf.stringToTerraform(struct.lastName),
        organization_name: cdktf.stringToTerraform(struct.organizationName),
        phone_number: cdktf.stringToTerraform(struct.phoneNumber),
        state: cdktf.stringToTerraform(struct.state),
        zip_code: cdktf.stringToTerraform(struct.zipCode),
    };
}
exports.route53DomainsRegisteredDomainRegistrantContactToTerraform = route53DomainsRegisteredDomainRegistrantContactToTerraform;
class Route53DomainsRegisteredDomainRegistrantContactOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._addressLine1 !== undefined) {
            hasAnyValues = true;
            internalValueResult.addressLine1 = this._addressLine1;
        }
        if (this._addressLine2 !== undefined) {
            hasAnyValues = true;
            internalValueResult.addressLine2 = this._addressLine2;
        }
        if (this._city !== undefined) {
            hasAnyValues = true;
            internalValueResult.city = this._city;
        }
        if (this._contactType !== undefined) {
            hasAnyValues = true;
            internalValueResult.contactType = this._contactType;
        }
        if (this._countryCode !== undefined) {
            hasAnyValues = true;
            internalValueResult.countryCode = this._countryCode;
        }
        if (this._email !== undefined) {
            hasAnyValues = true;
            internalValueResult.email = this._email;
        }
        if (this._extraParams !== undefined) {
            hasAnyValues = true;
            internalValueResult.extraParams = this._extraParams;
        }
        if (this._fax !== undefined) {
            hasAnyValues = true;
            internalValueResult.fax = this._fax;
        }
        if (this._firstName !== undefined) {
            hasAnyValues = true;
            internalValueResult.firstName = this._firstName;
        }
        if (this._lastName !== undefined) {
            hasAnyValues = true;
            internalValueResult.lastName = this._lastName;
        }
        if (this._organizationName !== undefined) {
            hasAnyValues = true;
            internalValueResult.organizationName = this._organizationName;
        }
        if (this._phoneNumber !== undefined) {
            hasAnyValues = true;
            internalValueResult.phoneNumber = this._phoneNumber;
        }
        if (this._state !== undefined) {
            hasAnyValues = true;
            internalValueResult.state = this._state;
        }
        if (this._zipCode !== undefined) {
            hasAnyValues = true;
            internalValueResult.zipCode = this._zipCode;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._addressLine1 = undefined;
            this._addressLine2 = undefined;
            this._city = undefined;
            this._contactType = undefined;
            this._countryCode = undefined;
            this._email = undefined;
            this._extraParams = undefined;
            this._fax = undefined;
            this._firstName = undefined;
            this._lastName = undefined;
            this._organizationName = undefined;
            this._phoneNumber = undefined;
            this._state = undefined;
            this._zipCode = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._addressLine1 = value.addressLine1;
            this._addressLine2 = value.addressLine2;
            this._city = value.city;
            this._contactType = value.contactType;
            this._countryCode = value.countryCode;
            this._email = value.email;
            this._extraParams = value.extraParams;
            this._fax = value.fax;
            this._firstName = value.firstName;
            this._lastName = value.lastName;
            this._organizationName = value.organizationName;
            this._phoneNumber = value.phoneNumber;
            this._state = value.state;
            this._zipCode = value.zipCode;
        }
    }
    get addressLine1() {
        return this.getStringAttribute('address_line_1');
    }
    set addressLine1(value) {
        this._addressLine1 = value;
    }
    resetAddressLine1() {
        this._addressLine1 = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get addressLine1Input() {
        return this._addressLine1;
    }
    get addressLine2() {
        return this.getStringAttribute('address_line_2');
    }
    set addressLine2(value) {
        this._addressLine2 = value;
    }
    resetAddressLine2() {
        this._addressLine2 = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get addressLine2Input() {
        return this._addressLine2;
    }
    get city() {
        return this.getStringAttribute('city');
    }
    set city(value) {
        this._city = value;
    }
    resetCity() {
        this._city = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cityInput() {
        return this._city;
    }
    get contactType() {
        return this.getStringAttribute('contact_type');
    }
    set contactType(value) {
        this._contactType = value;
    }
    resetContactType() {
        this._contactType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get contactTypeInput() {
        return this._contactType;
    }
    get countryCode() {
        return this.getStringAttribute('country_code');
    }
    set countryCode(value) {
        this._countryCode = value;
    }
    resetCountryCode() {
        this._countryCode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get countryCodeInput() {
        return this._countryCode;
    }
    get email() {
        return this.getStringAttribute('email');
    }
    set email(value) {
        this._email = value;
    }
    resetEmail() {
        this._email = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get emailInput() {
        return this._email;
    }
    get extraParams() {
        return this.getStringMapAttribute('extra_params');
    }
    set extraParams(value) {
        this._extraParams = value;
    }
    resetExtraParams() {
        this._extraParams = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get extraParamsInput() {
        return this._extraParams;
    }
    get fax() {
        return this.getStringAttribute('fax');
    }
    set fax(value) {
        this._fax = value;
    }
    resetFax() {
        this._fax = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get faxInput() {
        return this._fax;
    }
    get firstName() {
        return this.getStringAttribute('first_name');
    }
    set firstName(value) {
        this._firstName = value;
    }
    resetFirstName() {
        this._firstName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get firstNameInput() {
        return this._firstName;
    }
    get lastName() {
        return this.getStringAttribute('last_name');
    }
    set lastName(value) {
        this._lastName = value;
    }
    resetLastName() {
        this._lastName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lastNameInput() {
        return this._lastName;
    }
    get organizationName() {
        return this.getStringAttribute('organization_name');
    }
    set organizationName(value) {
        this._organizationName = value;
    }
    resetOrganizationName() {
        this._organizationName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get organizationNameInput() {
        return this._organizationName;
    }
    get phoneNumber() {
        return this.getStringAttribute('phone_number');
    }
    set phoneNumber(value) {
        this._phoneNumber = value;
    }
    resetPhoneNumber() {
        this._phoneNumber = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get phoneNumberInput() {
        return this._phoneNumber;
    }
    get state() {
        return this.getStringAttribute('state');
    }
    set state(value) {
        this._state = value;
    }
    resetState() {
        this._state = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get stateInput() {
        return this._state;
    }
    get zipCode() {
        return this.getStringAttribute('zip_code');
    }
    set zipCode(value) {
        this._zipCode = value;
    }
    resetZipCode() {
        this._zipCode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zipCodeInput() {
        return this._zipCode;
    }
}
exports.Route53DomainsRegisteredDomainRegistrantContactOutputReference = Route53DomainsRegisteredDomainRegistrantContactOutputReference;
_d = JSII_RTTI_SYMBOL_1;
Route53DomainsRegisteredDomainRegistrantContactOutputReference[_d] = { fqn: "@cdktf/provider-aws.route53DomainsRegisteredDomain.Route53DomainsRegisteredDomainRegistrantContactOutputReference", version: "10.0.12" };
function route53DomainsRegisteredDomainTechContactToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        address_line_1: cdktf.stringToTerraform(struct.addressLine1),
        address_line_2: cdktf.stringToTerraform(struct.addressLine2),
        city: cdktf.stringToTerraform(struct.city),
        contact_type: cdktf.stringToTerraform(struct.contactType),
        country_code: cdktf.stringToTerraform(struct.countryCode),
        email: cdktf.stringToTerraform(struct.email),
        extra_params: cdktf.hashMapper(cdktf.stringToTerraform)(struct.extraParams),
        fax: cdktf.stringToTerraform(struct.fax),
        first_name: cdktf.stringToTerraform(struct.firstName),
        last_name: cdktf.stringToTerraform(struct.lastName),
        organization_name: cdktf.stringToTerraform(struct.organizationName),
        phone_number: cdktf.stringToTerraform(struct.phoneNumber),
        state: cdktf.stringToTerraform(struct.state),
        zip_code: cdktf.stringToTerraform(struct.zipCode),
    };
}
exports.route53DomainsRegisteredDomainTechContactToTerraform = route53DomainsRegisteredDomainTechContactToTerraform;
class Route53DomainsRegisteredDomainTechContactOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._addressLine1 !== undefined) {
            hasAnyValues = true;
            internalValueResult.addressLine1 = this._addressLine1;
        }
        if (this._addressLine2 !== undefined) {
            hasAnyValues = true;
            internalValueResult.addressLine2 = this._addressLine2;
        }
        if (this._city !== undefined) {
            hasAnyValues = true;
            internalValueResult.city = this._city;
        }
        if (this._contactType !== undefined) {
            hasAnyValues = true;
            internalValueResult.contactType = this._contactType;
        }
        if (this._countryCode !== undefined) {
            hasAnyValues = true;
            internalValueResult.countryCode = this._countryCode;
        }
        if (this._email !== undefined) {
            hasAnyValues = true;
            internalValueResult.email = this._email;
        }
        if (this._extraParams !== undefined) {
            hasAnyValues = true;
            internalValueResult.extraParams = this._extraParams;
        }
        if (this._fax !== undefined) {
            hasAnyValues = true;
            internalValueResult.fax = this._fax;
        }
        if (this._firstName !== undefined) {
            hasAnyValues = true;
            internalValueResult.firstName = this._firstName;
        }
        if (this._lastName !== undefined) {
            hasAnyValues = true;
            internalValueResult.lastName = this._lastName;
        }
        if (this._organizationName !== undefined) {
            hasAnyValues = true;
            internalValueResult.organizationName = this._organizationName;
        }
        if (this._phoneNumber !== undefined) {
            hasAnyValues = true;
            internalValueResult.phoneNumber = this._phoneNumber;
        }
        if (this._state !== undefined) {
            hasAnyValues = true;
            internalValueResult.state = this._state;
        }
        if (this._zipCode !== undefined) {
            hasAnyValues = true;
            internalValueResult.zipCode = this._zipCode;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._addressLine1 = undefined;
            this._addressLine2 = undefined;
            this._city = undefined;
            this._contactType = undefined;
            this._countryCode = undefined;
            this._email = undefined;
            this._extraParams = undefined;
            this._fax = undefined;
            this._firstName = undefined;
            this._lastName = undefined;
            this._organizationName = undefined;
            this._phoneNumber = undefined;
            this._state = undefined;
            this._zipCode = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._addressLine1 = value.addressLine1;
            this._addressLine2 = value.addressLine2;
            this._city = value.city;
            this._contactType = value.contactType;
            this._countryCode = value.countryCode;
            this._email = value.email;
            this._extraParams = value.extraParams;
            this._fax = value.fax;
            this._firstName = value.firstName;
            this._lastName = value.lastName;
            this._organizationName = value.organizationName;
            this._phoneNumber = value.phoneNumber;
            this._state = value.state;
            this._zipCode = value.zipCode;
        }
    }
    get addressLine1() {
        return this.getStringAttribute('address_line_1');
    }
    set addressLine1(value) {
        this._addressLine1 = value;
    }
    resetAddressLine1() {
        this._addressLine1 = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get addressLine1Input() {
        return this._addressLine1;
    }
    get addressLine2() {
        return this.getStringAttribute('address_line_2');
    }
    set addressLine2(value) {
        this._addressLine2 = value;
    }
    resetAddressLine2() {
        this._addressLine2 = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get addressLine2Input() {
        return this._addressLine2;
    }
    get city() {
        return this.getStringAttribute('city');
    }
    set city(value) {
        this._city = value;
    }
    resetCity() {
        this._city = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cityInput() {
        return this._city;
    }
    get contactType() {
        return this.getStringAttribute('contact_type');
    }
    set contactType(value) {
        this._contactType = value;
    }
    resetContactType() {
        this._contactType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get contactTypeInput() {
        return this._contactType;
    }
    get countryCode() {
        return this.getStringAttribute('country_code');
    }
    set countryCode(value) {
        this._countryCode = value;
    }
    resetCountryCode() {
        this._countryCode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get countryCodeInput() {
        return this._countryCode;
    }
    get email() {
        return this.getStringAttribute('email');
    }
    set email(value) {
        this._email = value;
    }
    resetEmail() {
        this._email = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get emailInput() {
        return this._email;
    }
    get extraParams() {
        return this.getStringMapAttribute('extra_params');
    }
    set extraParams(value) {
        this._extraParams = value;
    }
    resetExtraParams() {
        this._extraParams = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get extraParamsInput() {
        return this._extraParams;
    }
    get fax() {
        return this.getStringAttribute('fax');
    }
    set fax(value) {
        this._fax = value;
    }
    resetFax() {
        this._fax = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get faxInput() {
        return this._fax;
    }
    get firstName() {
        return this.getStringAttribute('first_name');
    }
    set firstName(value) {
        this._firstName = value;
    }
    resetFirstName() {
        this._firstName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get firstNameInput() {
        return this._firstName;
    }
    get lastName() {
        return this.getStringAttribute('last_name');
    }
    set lastName(value) {
        this._lastName = value;
    }
    resetLastName() {
        this._lastName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lastNameInput() {
        return this._lastName;
    }
    get organizationName() {
        return this.getStringAttribute('organization_name');
    }
    set organizationName(value) {
        this._organizationName = value;
    }
    resetOrganizationName() {
        this._organizationName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get organizationNameInput() {
        return this._organizationName;
    }
    get phoneNumber() {
        return this.getStringAttribute('phone_number');
    }
    set phoneNumber(value) {
        this._phoneNumber = value;
    }
    resetPhoneNumber() {
        this._phoneNumber = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get phoneNumberInput() {
        return this._phoneNumber;
    }
    get state() {
        return this.getStringAttribute('state');
    }
    set state(value) {
        this._state = value;
    }
    resetState() {
        this._state = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get stateInput() {
        return this._state;
    }
    get zipCode() {
        return this.getStringAttribute('zip_code');
    }
    set zipCode(value) {
        this._zipCode = value;
    }
    resetZipCode() {
        this._zipCode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zipCodeInput() {
        return this._zipCode;
    }
}
exports.Route53DomainsRegisteredDomainTechContactOutputReference = Route53DomainsRegisteredDomainTechContactOutputReference;
_e = JSII_RTTI_SYMBOL_1;
Route53DomainsRegisteredDomainTechContactOutputReference[_e] = { fqn: "@cdktf/provider-aws.route53DomainsRegisteredDomain.Route53DomainsRegisteredDomainTechContactOutputReference", version: "10.0.12" };
function route53DomainsRegisteredDomainTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.route53DomainsRegisteredDomainTimeoutsToTerraform = route53DomainsRegisteredDomainTimeoutsToTerraform;
class Route53DomainsRegisteredDomainTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.Route53DomainsRegisteredDomainTimeoutsOutputReference = Route53DomainsRegisteredDomainTimeoutsOutputReference;
_f = JSII_RTTI_SYMBOL_1;
Route53DomainsRegisteredDomainTimeoutsOutputReference[_f] = { fqn: "@cdktf/provider-aws.route53DomainsRegisteredDomain.Route53DomainsRegisteredDomainTimeoutsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/route53domains_registered_domain aws_route53domains_registered_domain}
*/
class Route53DomainsRegisteredDomain extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/route53domains_registered_domain aws_route53domains_registered_domain} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options Route53DomainsRegisteredDomainConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_route53domains_registered_domain',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // admin_contact - computed: false, optional: true, required: false
        this._adminContact = new Route53DomainsRegisteredDomainAdminContactOutputReference(this, "admin_contact");
        // name_server - computed: false, optional: true, required: false
        this._nameServer = new Route53DomainsRegisteredDomainNameServerList(this, "name_server", false);
        // registrant_contact - computed: false, optional: true, required: false
        this._registrantContact = new Route53DomainsRegisteredDomainRegistrantContactOutputReference(this, "registrant_contact");
        // tech_contact - computed: false, optional: true, required: false
        this._techContact = new Route53DomainsRegisteredDomainTechContactOutputReference(this, "tech_contact");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new Route53DomainsRegisteredDomainTimeoutsOutputReference(this, "timeouts");
        this._adminPrivacy = config.adminPrivacy;
        this._autoRenew = config.autoRenew;
        this._domainName = config.domainName;
        this._id = config.id;
        this._registrantPrivacy = config.registrantPrivacy;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._techPrivacy = config.techPrivacy;
        this._transferLock = config.transferLock;
        this._adminContact.internalValue = config.adminContact;
        this._nameServer.internalValue = config.nameServer;
        this._registrantContact.internalValue = config.registrantContact;
        this._techContact.internalValue = config.techContact;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // abuse_contact_email - computed: true, optional: false, required: false
    get abuseContactEmail() {
        return this.getStringAttribute('abuse_contact_email');
    }
    // abuse_contact_phone - computed: true, optional: false, required: false
    get abuseContactPhone() {
        return this.getStringAttribute('abuse_contact_phone');
    }
    get adminPrivacy() {
        return this.getBooleanAttribute('admin_privacy');
    }
    set adminPrivacy(value) {
        this._adminPrivacy = value;
    }
    resetAdminPrivacy() {
        this._adminPrivacy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get adminPrivacyInput() {
        return this._adminPrivacy;
    }
    get autoRenew() {
        return this.getBooleanAttribute('auto_renew');
    }
    set autoRenew(value) {
        this._autoRenew = value;
    }
    resetAutoRenew() {
        this._autoRenew = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoRenewInput() {
        return this._autoRenew;
    }
    // creation_date - computed: true, optional: false, required: false
    get creationDate() {
        return this.getStringAttribute('creation_date');
    }
    get domainName() {
        return this.getStringAttribute('domain_name');
    }
    set domainName(value) {
        this._domainName = value;
    }
    // Temporarily expose input value. Use with caution.
    get domainNameInput() {
        return this._domainName;
    }
    // expiration_date - computed: true, optional: false, required: false
    get expirationDate() {
        return this.getStringAttribute('expiration_date');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get registrantPrivacy() {
        return this.getBooleanAttribute('registrant_privacy');
    }
    set registrantPrivacy(value) {
        this._registrantPrivacy = value;
    }
    resetRegistrantPrivacy() {
        this._registrantPrivacy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get registrantPrivacyInput() {
        return this._registrantPrivacy;
    }
    // registrar_name - computed: true, optional: false, required: false
    get registrarName() {
        return this.getStringAttribute('registrar_name');
    }
    // registrar_url - computed: true, optional: false, required: false
    get registrarUrl() {
        return this.getStringAttribute('registrar_url');
    }
    // reseller - computed: true, optional: false, required: false
    get reseller() {
        return this.getStringAttribute('reseller');
    }
    // status_list - computed: true, optional: false, required: false
    get statusList() {
        return this.getListAttribute('status_list');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get techPrivacy() {
        return this.getBooleanAttribute('tech_privacy');
    }
    set techPrivacy(value) {
        this._techPrivacy = value;
    }
    resetTechPrivacy() {
        this._techPrivacy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get techPrivacyInput() {
        return this._techPrivacy;
    }
    get transferLock() {
        return this.getBooleanAttribute('transfer_lock');
    }
    set transferLock(value) {
        this._transferLock = value;
    }
    resetTransferLock() {
        this._transferLock = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get transferLockInput() {
        return this._transferLock;
    }
    // updated_date - computed: true, optional: false, required: false
    get updatedDate() {
        return this.getStringAttribute('updated_date');
    }
    // whois_server - computed: true, optional: false, required: false
    get whoisServer() {
        return this.getStringAttribute('whois_server');
    }
    get adminContact() {
        return this._adminContact;
    }
    putAdminContact(value) {
        this._adminContact.internalValue = value;
    }
    resetAdminContact() {
        this._adminContact.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get adminContactInput() {
        return this._adminContact.internalValue;
    }
    get nameServer() {
        return this._nameServer;
    }
    putNameServer(value) {
        this._nameServer.internalValue = value;
    }
    resetNameServer() {
        this._nameServer.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameServerInput() {
        return this._nameServer.internalValue;
    }
    get registrantContact() {
        return this._registrantContact;
    }
    putRegistrantContact(value) {
        this._registrantContact.internalValue = value;
    }
    resetRegistrantContact() {
        this._registrantContact.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get registrantContactInput() {
        return this._registrantContact.internalValue;
    }
    get techContact() {
        return this._techContact;
    }
    putTechContact(value) {
        this._techContact.internalValue = value;
    }
    resetTechContact() {
        this._techContact.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get techContactInput() {
        return this._techContact.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            admin_privacy: cdktf.booleanToTerraform(this._adminPrivacy),
            auto_renew: cdktf.booleanToTerraform(this._autoRenew),
            domain_name: cdktf.stringToTerraform(this._domainName),
            id: cdktf.stringToTerraform(this._id),
            registrant_privacy: cdktf.booleanToTerraform(this._registrantPrivacy),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            tech_privacy: cdktf.booleanToTerraform(this._techPrivacy),
            transfer_lock: cdktf.booleanToTerraform(this._transferLock),
            admin_contact: route53DomainsRegisteredDomainAdminContactToTerraform(this._adminContact.internalValue),
            name_server: cdktf.listMapper(route53DomainsRegisteredDomainNameServerToTerraform, true)(this._nameServer.internalValue),
            registrant_contact: route53DomainsRegisteredDomainRegistrantContactToTerraform(this._registrantContact.internalValue),
            tech_contact: route53DomainsRegisteredDomainTechContactToTerraform(this._techContact.internalValue),
            timeouts: route53DomainsRegisteredDomainTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.Route53DomainsRegisteredDomain = Route53DomainsRegisteredDomain;
_g = JSII_RTTI_SYMBOL_1;
Route53DomainsRegisteredDomain[_g] = { fqn: "@cdktf/provider-aws.route53DomainsRegisteredDomain.Route53DomainsRegisteredDomain", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
Route53DomainsRegisteredDomain.tfResourceType = "aws_route53domains_registered_domain";
