// @cdktf/provider-aws/lib/inspector2-organization-configuration/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Inspector2OrganizationConfiguration = exports.Inspector2OrganizationConfigurationTimeoutsOutputReference = exports.inspector2OrganizationConfigurationTimeoutsToTerraform = exports.Inspector2OrganizationConfigurationAutoEnableOutputReference = exports.inspector2OrganizationConfigurationAutoEnableToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function inspector2OrganizationConfigurationAutoEnableToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ec2: cdktf.booleanToTerraform(struct.ec2),
        ecr: cdktf.booleanToTerraform(struct.ecr),
    };
}
exports.inspector2OrganizationConfigurationAutoEnableToTerraform = inspector2OrganizationConfigurationAutoEnableToTerraform;
class Inspector2OrganizationConfigurationAutoEnableOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._ec2 !== undefined) {
            hasAnyValues = true;
            internalValueResult.ec2 = this._ec2;
        }
        if (this._ecr !== undefined) {
            hasAnyValues = true;
            internalValueResult.ecr = this._ecr;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._ec2 = undefined;
            this._ecr = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._ec2 = value.ec2;
            this._ecr = value.ecr;
        }
    }
    get ec2() {
        return this.getBooleanAttribute('ec2');
    }
    set ec2(value) {
        this._ec2 = value;
    }
    // Temporarily expose input value. Use with caution.
    get ec2Input() {
        return this._ec2;
    }
    get ecr() {
        return this.getBooleanAttribute('ecr');
    }
    set ecr(value) {
        this._ecr = value;
    }
    // Temporarily expose input value. Use with caution.
    get ecrInput() {
        return this._ecr;
    }
}
exports.Inspector2OrganizationConfigurationAutoEnableOutputReference = Inspector2OrganizationConfigurationAutoEnableOutputReference;
_a = JSII_RTTI_SYMBOL_1;
Inspector2OrganizationConfigurationAutoEnableOutputReference[_a] = { fqn: "@cdktf/provider-aws.inspector2OrganizationConfiguration.Inspector2OrganizationConfigurationAutoEnableOutputReference", version: "10.0.12" };
function inspector2OrganizationConfigurationTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.inspector2OrganizationConfigurationTimeoutsToTerraform = inspector2OrganizationConfigurationTimeoutsToTerraform;
class Inspector2OrganizationConfigurationTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.Inspector2OrganizationConfigurationTimeoutsOutputReference = Inspector2OrganizationConfigurationTimeoutsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
Inspector2OrganizationConfigurationTimeoutsOutputReference[_b] = { fqn: "@cdktf/provider-aws.inspector2OrganizationConfiguration.Inspector2OrganizationConfigurationTimeoutsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/inspector2_organization_configuration aws_inspector2_organization_configuration}
*/
class Inspector2OrganizationConfiguration extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/inspector2_organization_configuration aws_inspector2_organization_configuration} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options Inspector2OrganizationConfigurationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_inspector2_organization_configuration',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // auto_enable - computed: false, optional: false, required: true
        this._autoEnable = new Inspector2OrganizationConfigurationAutoEnableOutputReference(this, "auto_enable");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new Inspector2OrganizationConfigurationTimeoutsOutputReference(this, "timeouts");
        this._id = config.id;
        this._autoEnable.internalValue = config.autoEnable;
        this._timeouts.internalValue = config.timeouts;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // max_account_limit_reached - computed: true, optional: false, required: false
    get maxAccountLimitReached() {
        return this.getBooleanAttribute('max_account_limit_reached');
    }
    get autoEnable() {
        return this._autoEnable;
    }
    putAutoEnable(value) {
        this._autoEnable.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get autoEnableInput() {
        return this._autoEnable.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            auto_enable: inspector2OrganizationConfigurationAutoEnableToTerraform(this._autoEnable.internalValue),
            timeouts: inspector2OrganizationConfigurationTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.Inspector2OrganizationConfiguration = Inspector2OrganizationConfiguration;
_c = JSII_RTTI_SYMBOL_1;
Inspector2OrganizationConfiguration[_c] = { fqn: "@cdktf/provider-aws.inspector2OrganizationConfiguration.Inspector2OrganizationConfiguration", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
Inspector2OrganizationConfiguration.tfResourceType = "aws_inspector2_organization_configuration";
