// @cdktf/provider-aws/lib/data-aws-dynamodb-table/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsDynamodbTable = exports.DataAwsDynamodbTableServerSideEncryptionOutputReference = exports.dataAwsDynamodbTableServerSideEncryptionToTerraform = exports.DataAwsDynamodbTableTtlList = exports.DataAwsDynamodbTableTtlOutputReference = exports.dataAwsDynamodbTableTtlToTerraform = exports.DataAwsDynamodbTableReplicaList = exports.DataAwsDynamodbTableReplicaOutputReference = exports.dataAwsDynamodbTableReplicaToTerraform = exports.DataAwsDynamodbTablePointInTimeRecoveryList = exports.DataAwsDynamodbTablePointInTimeRecoveryOutputReference = exports.dataAwsDynamodbTablePointInTimeRecoveryToTerraform = exports.DataAwsDynamodbTableLocalSecondaryIndexList = exports.DataAwsDynamodbTableLocalSecondaryIndexOutputReference = exports.dataAwsDynamodbTableLocalSecondaryIndexToTerraform = exports.DataAwsDynamodbTableGlobalSecondaryIndexList = exports.DataAwsDynamodbTableGlobalSecondaryIndexOutputReference = exports.dataAwsDynamodbTableGlobalSecondaryIndexToTerraform = exports.DataAwsDynamodbTableAttributeList = exports.DataAwsDynamodbTableAttributeOutputReference = exports.dataAwsDynamodbTableAttributeToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsDynamodbTableAttributeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsDynamodbTableAttributeToTerraform = dataAwsDynamodbTableAttributeToTerraform;
class DataAwsDynamodbTableAttributeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataAwsDynamodbTableAttributeOutputReference = DataAwsDynamodbTableAttributeOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsDynamodbTableAttributeOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsDynamodbTable.DataAwsDynamodbTableAttributeOutputReference", version: "10.0.12" };
class DataAwsDynamodbTableAttributeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsDynamodbTableAttributeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsDynamodbTableAttributeList = DataAwsDynamodbTableAttributeList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsDynamodbTableAttributeList[_b] = { fqn: "@cdktf/provider-aws.dataAwsDynamodbTable.DataAwsDynamodbTableAttributeList", version: "10.0.12" };
function dataAwsDynamodbTableGlobalSecondaryIndexToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsDynamodbTableGlobalSecondaryIndexToTerraform = dataAwsDynamodbTableGlobalSecondaryIndexToTerraform;
class DataAwsDynamodbTableGlobalSecondaryIndexOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // hash_key - computed: true, optional: false, required: false
    get hashKey() {
        return this.getStringAttribute('hash_key');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // non_key_attributes - computed: true, optional: false, required: false
    get nonKeyAttributes() {
        return this.getListAttribute('non_key_attributes');
    }
    // projection_type - computed: true, optional: false, required: false
    get projectionType() {
        return this.getStringAttribute('projection_type');
    }
    // range_key - computed: true, optional: false, required: false
    get rangeKey() {
        return this.getStringAttribute('range_key');
    }
    // read_capacity - computed: true, optional: false, required: false
    get readCapacity() {
        return this.getNumberAttribute('read_capacity');
    }
    // write_capacity - computed: true, optional: false, required: false
    get writeCapacity() {
        return this.getNumberAttribute('write_capacity');
    }
}
exports.DataAwsDynamodbTableGlobalSecondaryIndexOutputReference = DataAwsDynamodbTableGlobalSecondaryIndexOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsDynamodbTableGlobalSecondaryIndexOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsDynamodbTable.DataAwsDynamodbTableGlobalSecondaryIndexOutputReference", version: "10.0.12" };
class DataAwsDynamodbTableGlobalSecondaryIndexList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsDynamodbTableGlobalSecondaryIndexOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsDynamodbTableGlobalSecondaryIndexList = DataAwsDynamodbTableGlobalSecondaryIndexList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsDynamodbTableGlobalSecondaryIndexList[_d] = { fqn: "@cdktf/provider-aws.dataAwsDynamodbTable.DataAwsDynamodbTableGlobalSecondaryIndexList", version: "10.0.12" };
function dataAwsDynamodbTableLocalSecondaryIndexToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsDynamodbTableLocalSecondaryIndexToTerraform = dataAwsDynamodbTableLocalSecondaryIndexToTerraform;
class DataAwsDynamodbTableLocalSecondaryIndexOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // non_key_attributes - computed: true, optional: false, required: false
    get nonKeyAttributes() {
        return this.getListAttribute('non_key_attributes');
    }
    // projection_type - computed: true, optional: false, required: false
    get projectionType() {
        return this.getStringAttribute('projection_type');
    }
    // range_key - computed: true, optional: false, required: false
    get rangeKey() {
        return this.getStringAttribute('range_key');
    }
}
exports.DataAwsDynamodbTableLocalSecondaryIndexOutputReference = DataAwsDynamodbTableLocalSecondaryIndexOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsDynamodbTableLocalSecondaryIndexOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsDynamodbTable.DataAwsDynamodbTableLocalSecondaryIndexOutputReference", version: "10.0.12" };
class DataAwsDynamodbTableLocalSecondaryIndexList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsDynamodbTableLocalSecondaryIndexOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsDynamodbTableLocalSecondaryIndexList = DataAwsDynamodbTableLocalSecondaryIndexList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsDynamodbTableLocalSecondaryIndexList[_f] = { fqn: "@cdktf/provider-aws.dataAwsDynamodbTable.DataAwsDynamodbTableLocalSecondaryIndexList", version: "10.0.12" };
function dataAwsDynamodbTablePointInTimeRecoveryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsDynamodbTablePointInTimeRecoveryToTerraform = dataAwsDynamodbTablePointInTimeRecoveryToTerraform;
class DataAwsDynamodbTablePointInTimeRecoveryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
}
exports.DataAwsDynamodbTablePointInTimeRecoveryOutputReference = DataAwsDynamodbTablePointInTimeRecoveryOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsDynamodbTablePointInTimeRecoveryOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsDynamodbTable.DataAwsDynamodbTablePointInTimeRecoveryOutputReference", version: "10.0.12" };
class DataAwsDynamodbTablePointInTimeRecoveryList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsDynamodbTablePointInTimeRecoveryOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsDynamodbTablePointInTimeRecoveryList = DataAwsDynamodbTablePointInTimeRecoveryList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsDynamodbTablePointInTimeRecoveryList[_h] = { fqn: "@cdktf/provider-aws.dataAwsDynamodbTable.DataAwsDynamodbTablePointInTimeRecoveryList", version: "10.0.12" };
function dataAwsDynamodbTableReplicaToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsDynamodbTableReplicaToTerraform = dataAwsDynamodbTableReplicaToTerraform;
class DataAwsDynamodbTableReplicaOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // kms_key_arn - computed: true, optional: false, required: false
    get kmsKeyArn() {
        return this.getStringAttribute('kms_key_arn');
    }
    // region_name - computed: true, optional: false, required: false
    get regionName() {
        return this.getStringAttribute('region_name');
    }
}
exports.DataAwsDynamodbTableReplicaOutputReference = DataAwsDynamodbTableReplicaOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsDynamodbTableReplicaOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsDynamodbTable.DataAwsDynamodbTableReplicaOutputReference", version: "10.0.12" };
class DataAwsDynamodbTableReplicaList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsDynamodbTableReplicaOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsDynamodbTableReplicaList = DataAwsDynamodbTableReplicaList;
_k = JSII_RTTI_SYMBOL_1;
DataAwsDynamodbTableReplicaList[_k] = { fqn: "@cdktf/provider-aws.dataAwsDynamodbTable.DataAwsDynamodbTableReplicaList", version: "10.0.12" };
function dataAwsDynamodbTableTtlToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsDynamodbTableTtlToTerraform = dataAwsDynamodbTableTtlToTerraform;
class DataAwsDynamodbTableTtlOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // attribute_name - computed: true, optional: false, required: false
    get attributeName() {
        return this.getStringAttribute('attribute_name');
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
}
exports.DataAwsDynamodbTableTtlOutputReference = DataAwsDynamodbTableTtlOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataAwsDynamodbTableTtlOutputReference[_l] = { fqn: "@cdktf/provider-aws.dataAwsDynamodbTable.DataAwsDynamodbTableTtlOutputReference", version: "10.0.12" };
class DataAwsDynamodbTableTtlList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsDynamodbTableTtlOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsDynamodbTableTtlList = DataAwsDynamodbTableTtlList;
_m = JSII_RTTI_SYMBOL_1;
DataAwsDynamodbTableTtlList[_m] = { fqn: "@cdktf/provider-aws.dataAwsDynamodbTable.DataAwsDynamodbTableTtlList", version: "10.0.12" };
function dataAwsDynamodbTableServerSideEncryptionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsDynamodbTableServerSideEncryptionToTerraform = dataAwsDynamodbTableServerSideEncryptionToTerraform;
class DataAwsDynamodbTableServerSideEncryptionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    // kms_key_arn - computed: true, optional: false, required: false
    get kmsKeyArn() {
        return this.getStringAttribute('kms_key_arn');
    }
}
exports.DataAwsDynamodbTableServerSideEncryptionOutputReference = DataAwsDynamodbTableServerSideEncryptionOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataAwsDynamodbTableServerSideEncryptionOutputReference[_o] = { fqn: "@cdktf/provider-aws.dataAwsDynamodbTable.DataAwsDynamodbTableServerSideEncryptionOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/dynamodb_table aws_dynamodb_table}
*/
class DataAwsDynamodbTable extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/dynamodb_table aws_dynamodb_table} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsDynamodbTableConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_dynamodb_table',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // attribute - computed: true, optional: false, required: false
        this._attribute = new DataAwsDynamodbTableAttributeList(this, "attribute", true);
        // global_secondary_index - computed: true, optional: false, required: false
        this._globalSecondaryIndex = new DataAwsDynamodbTableGlobalSecondaryIndexList(this, "global_secondary_index", true);
        // local_secondary_index - computed: true, optional: false, required: false
        this._localSecondaryIndex = new DataAwsDynamodbTableLocalSecondaryIndexList(this, "local_secondary_index", true);
        // point_in_time_recovery - computed: true, optional: false, required: false
        this._pointInTimeRecovery = new DataAwsDynamodbTablePointInTimeRecoveryList(this, "point_in_time_recovery", false);
        // replica - computed: true, optional: false, required: false
        this._replica = new DataAwsDynamodbTableReplicaList(this, "replica", true);
        // ttl - computed: true, optional: false, required: false
        this._ttl = new DataAwsDynamodbTableTtlList(this, "ttl", true);
        // server_side_encryption - computed: false, optional: true, required: false
        this._serverSideEncryption = new DataAwsDynamodbTableServerSideEncryptionOutputReference(this, "server_side_encryption");
        this._id = config.id;
        this._name = config.name;
        this._tags = config.tags;
        this._serverSideEncryption.internalValue = config.serverSideEncryption;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get attribute() {
        return this._attribute;
    }
    // billing_mode - computed: true, optional: false, required: false
    get billingMode() {
        return this.getStringAttribute('billing_mode');
    }
    get globalSecondaryIndex() {
        return this._globalSecondaryIndex;
    }
    // hash_key - computed: true, optional: false, required: false
    get hashKey() {
        return this.getStringAttribute('hash_key');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get localSecondaryIndex() {
        return this._localSecondaryIndex;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get pointInTimeRecovery() {
        return this._pointInTimeRecovery;
    }
    // range_key - computed: true, optional: false, required: false
    get rangeKey() {
        return this.getStringAttribute('range_key');
    }
    // read_capacity - computed: true, optional: false, required: false
    get readCapacity() {
        return this.getNumberAttribute('read_capacity');
    }
    get replica() {
        return this._replica;
    }
    // stream_arn - computed: true, optional: false, required: false
    get streamArn() {
        return this.getStringAttribute('stream_arn');
    }
    // stream_enabled - computed: true, optional: false, required: false
    get streamEnabled() {
        return this.getBooleanAttribute('stream_enabled');
    }
    // stream_label - computed: true, optional: false, required: false
    get streamLabel() {
        return this.getStringAttribute('stream_label');
    }
    // stream_view_type - computed: true, optional: false, required: false
    get streamViewType() {
        return this.getStringAttribute('stream_view_type');
    }
    // table_class - computed: true, optional: false, required: false
    get tableClass() {
        return this.getStringAttribute('table_class');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get ttl() {
        return this._ttl;
    }
    // write_capacity - computed: true, optional: false, required: false
    get writeCapacity() {
        return this.getNumberAttribute('write_capacity');
    }
    get serverSideEncryption() {
        return this._serverSideEncryption;
    }
    putServerSideEncryption(value) {
        this._serverSideEncryption.internalValue = value;
    }
    resetServerSideEncryption() {
        this._serverSideEncryption.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serverSideEncryptionInput() {
        return this._serverSideEncryption.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            server_side_encryption: dataAwsDynamodbTableServerSideEncryptionToTerraform(this._serverSideEncryption.internalValue),
        };
    }
}
exports.DataAwsDynamodbTable = DataAwsDynamodbTable;
_p = JSII_RTTI_SYMBOL_1;
DataAwsDynamodbTable[_p] = { fqn: "@cdktf/provider-aws.dataAwsDynamodbTable.DataAwsDynamodbTable", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsDynamodbTable.tfResourceType = "aws_dynamodb_table";
