// @cdktf/provider-aws/lib/securityhub-finding-aggregator/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecurityhubFindingAggregator = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/securityhub_finding_aggregator aws_securityhub_finding_aggregator}
*/
class SecurityhubFindingAggregator extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/securityhub_finding_aggregator aws_securityhub_finding_aggregator} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SecurityhubFindingAggregatorConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_securityhub_finding_aggregator',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._linkingMode = config.linkingMode;
        this._specifiedRegions = config.specifiedRegions;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get linkingMode() {
        return this.getStringAttribute('linking_mode');
    }
    set linkingMode(value) {
        this._linkingMode = value;
    }
    // Temporarily expose input value. Use with caution.
    get linkingModeInput() {
        return this._linkingMode;
    }
    get specifiedRegions() {
        return cdktf.Fn.tolist(this.getListAttribute('specified_regions'));
    }
    set specifiedRegions(value) {
        this._specifiedRegions = value;
    }
    resetSpecifiedRegions() {
        this._specifiedRegions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get specifiedRegionsInput() {
        return this._specifiedRegions;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            linking_mode: cdktf.stringToTerraform(this._linkingMode),
            specified_regions: cdktf.listMapper(cdktf.stringToTerraform, false)(this._specifiedRegions),
        };
    }
}
exports.SecurityhubFindingAggregator = SecurityhubFindingAggregator;
_a = JSII_RTTI_SYMBOL_1;
SecurityhubFindingAggregator[_a] = { fqn: "@cdktf/provider-aws.securityhubFindingAggregator.SecurityhubFindingAggregator", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
SecurityhubFindingAggregator.tfResourceType = "aws_securityhub_finding_aggregator";
