// @cdktf/provider-aws/lib/wafv2-web-acl/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Wafv2WebAcl = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
// generated from terraform resource schema
const index_structs_1 = require("./index-structs");
__exportStar(require("./index-structs"), exports);
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/wafv2_web_acl aws_wafv2_web_acl}
*/
class Wafv2WebAcl extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/wafv2_web_acl aws_wafv2_web_acl} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options Wafv2WebAclConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_wafv2_web_acl',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // custom_response_body - computed: false, optional: true, required: false
        this._customResponseBody = new index_structs_1.Wafv2WebAclCustomResponseBodyList(this, "custom_response_body", true);
        // default_action - computed: false, optional: false, required: true
        this._defaultAction = new index_structs_1.Wafv2WebAclDefaultActionOutputReference(this, "default_action");
        // rule - computed: false, optional: true, required: false
        this._rule = new index_structs_1.Wafv2WebAclRuleList(this, "rule", true);
        // visibility_config - computed: false, optional: false, required: true
        this._visibilityConfig = new index_structs_1.Wafv2WebAclVisibilityConfigOutputReference(this, "visibility_config");
        this._description = config.description;
        this._id = config.id;
        this._name = config.name;
        this._scope = config.scope;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._customResponseBody.internalValue = config.customResponseBody;
        this._defaultAction.internalValue = config.defaultAction;
        this._rule.internalValue = config.rule;
        this._visibilityConfig.internalValue = config.visibilityConfig;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // capacity - computed: true, optional: false, required: false
    get capacity() {
        return this.getNumberAttribute('capacity');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // lock_token - computed: true, optional: false, required: false
    get lockToken() {
        return this.getStringAttribute('lock_token');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get scope() {
        return this.getStringAttribute('scope');
    }
    set scope(value) {
        this._scope = value;
    }
    // Temporarily expose input value. Use with caution.
    get scopeInput() {
        return this._scope;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get customResponseBody() {
        return this._customResponseBody;
    }
    putCustomResponseBody(value) {
        this._customResponseBody.internalValue = value;
    }
    resetCustomResponseBody() {
        this._customResponseBody.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get customResponseBodyInput() {
        return this._customResponseBody.internalValue;
    }
    get defaultAction() {
        return this._defaultAction;
    }
    putDefaultAction(value) {
        this._defaultAction.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get defaultActionInput() {
        return this._defaultAction.internalValue;
    }
    get rule() {
        return this._rule;
    }
    putRule(value) {
        this._rule.internalValue = value;
    }
    resetRule() {
        this._rule.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ruleInput() {
        return this._rule.internalValue;
    }
    get visibilityConfig() {
        return this._visibilityConfig;
    }
    putVisibilityConfig(value) {
        this._visibilityConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get visibilityConfigInput() {
        return this._visibilityConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            scope: cdktf.stringToTerraform(this._scope),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            custom_response_body: cdktf.listMapper(index_structs_1.wafv2WebAclCustomResponseBodyToTerraform, true)(this._customResponseBody.internalValue),
            default_action: index_structs_1.wafv2WebAclDefaultActionToTerraform(this._defaultAction.internalValue),
            rule: cdktf.listMapper(index_structs_1.wafv2WebAclRuleToTerraform, true)(this._rule.internalValue),
            visibility_config: index_structs_1.wafv2WebAclVisibilityConfigToTerraform(this._visibilityConfig.internalValue),
        };
    }
}
exports.Wafv2WebAcl = Wafv2WebAcl;
_a = JSII_RTTI_SYMBOL_1;
Wafv2WebAcl[_a] = { fqn: "@cdktf/provider-aws.wafv2WebAcl.Wafv2WebAcl", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
Wafv2WebAcl.tfResourceType = "aws_wafv2_web_acl";
