// @cdktf/provider-aws/lib/elasticache-user-group-association/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElasticacheUserGroupAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/elasticache_user_group_association aws_elasticache_user_group_association}
*/
class ElasticacheUserGroupAssociation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/elasticache_user_group_association aws_elasticache_user_group_association} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ElasticacheUserGroupAssociationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_elasticache_user_group_association',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._userGroupId = config.userGroupId;
        this._userId = config.userId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get userGroupId() {
        return this.getStringAttribute('user_group_id');
    }
    set userGroupId(value) {
        this._userGroupId = value;
    }
    // Temporarily expose input value. Use with caution.
    get userGroupIdInput() {
        return this._userGroupId;
    }
    get userId() {
        return this.getStringAttribute('user_id');
    }
    set userId(value) {
        this._userId = value;
    }
    // Temporarily expose input value. Use with caution.
    get userIdInput() {
        return this._userId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            user_group_id: cdktf.stringToTerraform(this._userGroupId),
            user_id: cdktf.stringToTerraform(this._userId),
        };
    }
}
exports.ElasticacheUserGroupAssociation = ElasticacheUserGroupAssociation;
_a = JSII_RTTI_SYMBOL_1;
ElasticacheUserGroupAssociation[_a] = { fqn: "@cdktf/provider-aws.elasticacheUserGroupAssociation.ElasticacheUserGroupAssociation", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
ElasticacheUserGroupAssociation.tfResourceType = "aws_elasticache_user_group_association";
