// @cdktf/provider-aws/lib/ec2-traffic-mirror-session/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ec2TrafficMirrorSession = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/ec2_traffic_mirror_session aws_ec2_traffic_mirror_session}
*/
class Ec2TrafficMirrorSession extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ec2_traffic_mirror_session aws_ec2_traffic_mirror_session} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options Ec2TrafficMirrorSessionConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ec2_traffic_mirror_session',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._description = config.description;
        this._id = config.id;
        this._networkInterfaceId = config.networkInterfaceId;
        this._packetLength = config.packetLength;
        this._sessionNumber = config.sessionNumber;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._trafficMirrorFilterId = config.trafficMirrorFilterId;
        this._trafficMirrorTargetId = config.trafficMirrorTargetId;
        this._virtualNetworkId = config.virtualNetworkId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get networkInterfaceId() {
        return this.getStringAttribute('network_interface_id');
    }
    set networkInterfaceId(value) {
        this._networkInterfaceId = value;
    }
    // Temporarily expose input value. Use with caution.
    get networkInterfaceIdInput() {
        return this._networkInterfaceId;
    }
    // owner_id - computed: true, optional: false, required: false
    get ownerId() {
        return this.getStringAttribute('owner_id');
    }
    get packetLength() {
        return this.getNumberAttribute('packet_length');
    }
    set packetLength(value) {
        this._packetLength = value;
    }
    resetPacketLength() {
        this._packetLength = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get packetLengthInput() {
        return this._packetLength;
    }
    get sessionNumber() {
        return this.getNumberAttribute('session_number');
    }
    set sessionNumber(value) {
        this._sessionNumber = value;
    }
    // Temporarily expose input value. Use with caution.
    get sessionNumberInput() {
        return this._sessionNumber;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get trafficMirrorFilterId() {
        return this.getStringAttribute('traffic_mirror_filter_id');
    }
    set trafficMirrorFilterId(value) {
        this._trafficMirrorFilterId = value;
    }
    // Temporarily expose input value. Use with caution.
    get trafficMirrorFilterIdInput() {
        return this._trafficMirrorFilterId;
    }
    get trafficMirrorTargetId() {
        return this.getStringAttribute('traffic_mirror_target_id');
    }
    set trafficMirrorTargetId(value) {
        this._trafficMirrorTargetId = value;
    }
    // Temporarily expose input value. Use with caution.
    get trafficMirrorTargetIdInput() {
        return this._trafficMirrorTargetId;
    }
    get virtualNetworkId() {
        return this.getNumberAttribute('virtual_network_id');
    }
    set virtualNetworkId(value) {
        this._virtualNetworkId = value;
    }
    resetVirtualNetworkId() {
        this._virtualNetworkId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get virtualNetworkIdInput() {
        return this._virtualNetworkId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            network_interface_id: cdktf.stringToTerraform(this._networkInterfaceId),
            packet_length: cdktf.numberToTerraform(this._packetLength),
            session_number: cdktf.numberToTerraform(this._sessionNumber),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            traffic_mirror_filter_id: cdktf.stringToTerraform(this._trafficMirrorFilterId),
            traffic_mirror_target_id: cdktf.stringToTerraform(this._trafficMirrorTargetId),
            virtual_network_id: cdktf.numberToTerraform(this._virtualNetworkId),
        };
    }
}
exports.Ec2TrafficMirrorSession = Ec2TrafficMirrorSession;
_a = JSII_RTTI_SYMBOL_1;
Ec2TrafficMirrorSession[_a] = { fqn: "@cdktf/provider-aws.ec2TrafficMirrorSession.Ec2TrafficMirrorSession", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
Ec2TrafficMirrorSession.tfResourceType = "aws_ec2_traffic_mirror_session";
