// @cdktf/provider-aws/lib/cloudwatch-log-destination-policy/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudwatchLogDestinationPolicy = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/cloudwatch_log_destination_policy aws_cloudwatch_log_destination_policy}
*/
class CloudwatchLogDestinationPolicy extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/cloudwatch_log_destination_policy aws_cloudwatch_log_destination_policy} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CloudwatchLogDestinationPolicyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cloudwatch_log_destination_policy',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._accessPolicy = config.accessPolicy;
        this._destinationName = config.destinationName;
        this._forceUpdate = config.forceUpdate;
        this._id = config.id;
    }
    get accessPolicy() {
        return this.getStringAttribute('access_policy');
    }
    set accessPolicy(value) {
        this._accessPolicy = value;
    }
    // Temporarily expose input value. Use with caution.
    get accessPolicyInput() {
        return this._accessPolicy;
    }
    get destinationName() {
        return this.getStringAttribute('destination_name');
    }
    set destinationName(value) {
        this._destinationName = value;
    }
    // Temporarily expose input value. Use with caution.
    get destinationNameInput() {
        return this._destinationName;
    }
    get forceUpdate() {
        return this.getBooleanAttribute('force_update');
    }
    set forceUpdate(value) {
        this._forceUpdate = value;
    }
    resetForceUpdate() {
        this._forceUpdate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get forceUpdateInput() {
        return this._forceUpdate;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            access_policy: cdktf.stringToTerraform(this._accessPolicy),
            destination_name: cdktf.stringToTerraform(this._destinationName),
            force_update: cdktf.booleanToTerraform(this._forceUpdate),
            id: cdktf.stringToTerraform(this._id),
        };
    }
}
exports.CloudwatchLogDestinationPolicy = CloudwatchLogDestinationPolicy;
_a = JSII_RTTI_SYMBOL_1;
CloudwatchLogDestinationPolicy[_a] = { fqn: "@cdktf/provider-aws.cloudwatchLogDestinationPolicy.CloudwatchLogDestinationPolicy", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
CloudwatchLogDestinationPolicy.tfResourceType = "aws_cloudwatch_log_destination_policy";
