// @cdktf/provider-aws/lib/efs-access-point/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EfsAccessPoint = exports.EfsAccessPointRootDirectoryOutputReference = exports.efsAccessPointRootDirectoryToTerraform = exports.EfsAccessPointRootDirectoryCreationInfoOutputReference = exports.efsAccessPointRootDirectoryCreationInfoToTerraform = exports.EfsAccessPointPosixUserOutputReference = exports.efsAccessPointPosixUserToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function efsAccessPointPosixUserToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        gid: cdktf.numberToTerraform(struct.gid),
        secondary_gids: cdktf.listMapper(cdktf.numberToTerraform, false)(struct.secondaryGids),
        uid: cdktf.numberToTerraform(struct.uid),
    };
}
exports.efsAccessPointPosixUserToTerraform = efsAccessPointPosixUserToTerraform;
class EfsAccessPointPosixUserOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._gid !== undefined) {
            hasAnyValues = true;
            internalValueResult.gid = this._gid;
        }
        if (this._secondaryGids !== undefined) {
            hasAnyValues = true;
            internalValueResult.secondaryGids = this._secondaryGids;
        }
        if (this._uid !== undefined) {
            hasAnyValues = true;
            internalValueResult.uid = this._uid;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._gid = undefined;
            this._secondaryGids = undefined;
            this._uid = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._gid = value.gid;
            this._secondaryGids = value.secondaryGids;
            this._uid = value.uid;
        }
    }
    get gid() {
        return this.getNumberAttribute('gid');
    }
    set gid(value) {
        this._gid = value;
    }
    // Temporarily expose input value. Use with caution.
    get gidInput() {
        return this._gid;
    }
    get secondaryGids() {
        return cdktf.Token.asNumberList(cdktf.Fn.tolist(this.getNumberListAttribute('secondary_gids')));
    }
    set secondaryGids(value) {
        this._secondaryGids = value;
    }
    resetSecondaryGids() {
        this._secondaryGids = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secondaryGidsInput() {
        return this._secondaryGids;
    }
    get uid() {
        return this.getNumberAttribute('uid');
    }
    set uid(value) {
        this._uid = value;
    }
    // Temporarily expose input value. Use with caution.
    get uidInput() {
        return this._uid;
    }
}
exports.EfsAccessPointPosixUserOutputReference = EfsAccessPointPosixUserOutputReference;
_a = JSII_RTTI_SYMBOL_1;
EfsAccessPointPosixUserOutputReference[_a] = { fqn: "@cdktf/provider-aws.efsAccessPoint.EfsAccessPointPosixUserOutputReference", version: "10.0.12" };
function efsAccessPointRootDirectoryCreationInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        owner_gid: cdktf.numberToTerraform(struct.ownerGid),
        owner_uid: cdktf.numberToTerraform(struct.ownerUid),
        permissions: cdktf.stringToTerraform(struct.permissions),
    };
}
exports.efsAccessPointRootDirectoryCreationInfoToTerraform = efsAccessPointRootDirectoryCreationInfoToTerraform;
class EfsAccessPointRootDirectoryCreationInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._ownerGid !== undefined) {
            hasAnyValues = true;
            internalValueResult.ownerGid = this._ownerGid;
        }
        if (this._ownerUid !== undefined) {
            hasAnyValues = true;
            internalValueResult.ownerUid = this._ownerUid;
        }
        if (this._permissions !== undefined) {
            hasAnyValues = true;
            internalValueResult.permissions = this._permissions;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._ownerGid = undefined;
            this._ownerUid = undefined;
            this._permissions = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._ownerGid = value.ownerGid;
            this._ownerUid = value.ownerUid;
            this._permissions = value.permissions;
        }
    }
    get ownerGid() {
        return this.getNumberAttribute('owner_gid');
    }
    set ownerGid(value) {
        this._ownerGid = value;
    }
    // Temporarily expose input value. Use with caution.
    get ownerGidInput() {
        return this._ownerGid;
    }
    get ownerUid() {
        return this.getNumberAttribute('owner_uid');
    }
    set ownerUid(value) {
        this._ownerUid = value;
    }
    // Temporarily expose input value. Use with caution.
    get ownerUidInput() {
        return this._ownerUid;
    }
    get permissions() {
        return this.getStringAttribute('permissions');
    }
    set permissions(value) {
        this._permissions = value;
    }
    // Temporarily expose input value. Use with caution.
    get permissionsInput() {
        return this._permissions;
    }
}
exports.EfsAccessPointRootDirectoryCreationInfoOutputReference = EfsAccessPointRootDirectoryCreationInfoOutputReference;
_b = JSII_RTTI_SYMBOL_1;
EfsAccessPointRootDirectoryCreationInfoOutputReference[_b] = { fqn: "@cdktf/provider-aws.efsAccessPoint.EfsAccessPointRootDirectoryCreationInfoOutputReference", version: "10.0.12" };
function efsAccessPointRootDirectoryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        path: cdktf.stringToTerraform(struct.path),
        creation_info: efsAccessPointRootDirectoryCreationInfoToTerraform(struct.creationInfo),
    };
}
exports.efsAccessPointRootDirectoryToTerraform = efsAccessPointRootDirectoryToTerraform;
class EfsAccessPointRootDirectoryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // creation_info - computed: false, optional: true, required: false
        this._creationInfo = new EfsAccessPointRootDirectoryCreationInfoOutputReference(this, "creation_info");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        if (this._creationInfo?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.creationInfo = this._creationInfo?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._path = undefined;
            this._creationInfo.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._path = value.path;
            this._creationInfo.internalValue = value.creationInfo;
        }
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    resetPath() {
        this._path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
    get creationInfo() {
        return this._creationInfo;
    }
    putCreationInfo(value) {
        this._creationInfo.internalValue = value;
    }
    resetCreationInfo() {
        this._creationInfo.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get creationInfoInput() {
        return this._creationInfo.internalValue;
    }
}
exports.EfsAccessPointRootDirectoryOutputReference = EfsAccessPointRootDirectoryOutputReference;
_c = JSII_RTTI_SYMBOL_1;
EfsAccessPointRootDirectoryOutputReference[_c] = { fqn: "@cdktf/provider-aws.efsAccessPoint.EfsAccessPointRootDirectoryOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/efs_access_point aws_efs_access_point}
*/
class EfsAccessPoint extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/efs_access_point aws_efs_access_point} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options EfsAccessPointConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_efs_access_point',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // posix_user - computed: false, optional: true, required: false
        this._posixUser = new EfsAccessPointPosixUserOutputReference(this, "posix_user");
        // root_directory - computed: false, optional: true, required: false
        this._rootDirectory = new EfsAccessPointRootDirectoryOutputReference(this, "root_directory");
        this._fileSystemId = config.fileSystemId;
        this._id = config.id;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._posixUser.internalValue = config.posixUser;
        this._rootDirectory.internalValue = config.rootDirectory;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // file_system_arn - computed: true, optional: false, required: false
    get fileSystemArn() {
        return this.getStringAttribute('file_system_arn');
    }
    get fileSystemId() {
        return this.getStringAttribute('file_system_id');
    }
    set fileSystemId(value) {
        this._fileSystemId = value;
    }
    // Temporarily expose input value. Use with caution.
    get fileSystemIdInput() {
        return this._fileSystemId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // owner_id - computed: true, optional: false, required: false
    get ownerId() {
        return this.getStringAttribute('owner_id');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get posixUser() {
        return this._posixUser;
    }
    putPosixUser(value) {
        this._posixUser.internalValue = value;
    }
    resetPosixUser() {
        this._posixUser.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get posixUserInput() {
        return this._posixUser.internalValue;
    }
    get rootDirectory() {
        return this._rootDirectory;
    }
    putRootDirectory(value) {
        this._rootDirectory.internalValue = value;
    }
    resetRootDirectory() {
        this._rootDirectory.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rootDirectoryInput() {
        return this._rootDirectory.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            file_system_id: cdktf.stringToTerraform(this._fileSystemId),
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            posix_user: efsAccessPointPosixUserToTerraform(this._posixUser.internalValue),
            root_directory: efsAccessPointRootDirectoryToTerraform(this._rootDirectory.internalValue),
        };
    }
}
exports.EfsAccessPoint = EfsAccessPoint;
_d = JSII_RTTI_SYMBOL_1;
EfsAccessPoint[_d] = { fqn: "@cdktf/provider-aws.efsAccessPoint.EfsAccessPoint", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
EfsAccessPoint.tfResourceType = "aws_efs_access_point";
