// @cdktf/provider-aws/lib/sesv2-email-identity-feedback-attributes/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sesv2EmailIdentityFeedbackAttributes = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/sesv2_email_identity_feedback_attributes aws_sesv2_email_identity_feedback_attributes}
*/
class Sesv2EmailIdentityFeedbackAttributes extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/sesv2_email_identity_feedback_attributes aws_sesv2_email_identity_feedback_attributes} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options Sesv2EmailIdentityFeedbackAttributesConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_sesv2_email_identity_feedback_attributes',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._emailForwardingEnabled = config.emailForwardingEnabled;
        this._emailIdentity = config.emailIdentity;
        this._id = config.id;
    }
    get emailForwardingEnabled() {
        return this.getBooleanAttribute('email_forwarding_enabled');
    }
    set emailForwardingEnabled(value) {
        this._emailForwardingEnabled = value;
    }
    resetEmailForwardingEnabled() {
        this._emailForwardingEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get emailForwardingEnabledInput() {
        return this._emailForwardingEnabled;
    }
    get emailIdentity() {
        return this.getStringAttribute('email_identity');
    }
    set emailIdentity(value) {
        this._emailIdentity = value;
    }
    // Temporarily expose input value. Use with caution.
    get emailIdentityInput() {
        return this._emailIdentity;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            email_forwarding_enabled: cdktf.booleanToTerraform(this._emailForwardingEnabled),
            email_identity: cdktf.stringToTerraform(this._emailIdentity),
            id: cdktf.stringToTerraform(this._id),
        };
    }
}
exports.Sesv2EmailIdentityFeedbackAttributes = Sesv2EmailIdentityFeedbackAttributes;
_a = JSII_RTTI_SYMBOL_1;
Sesv2EmailIdentityFeedbackAttributes[_a] = { fqn: "@cdktf/provider-aws.sesv2EmailIdentityFeedbackAttributes.Sesv2EmailIdentityFeedbackAttributes", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
Sesv2EmailIdentityFeedbackAttributes.tfResourceType = "aws_sesv2_email_identity_feedback_attributes";
