// @cdktf/provider-aws/lib/ses-domain-mail-from/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SesDomainMailFrom = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/ses_domain_mail_from aws_ses_domain_mail_from}
*/
class SesDomainMailFrom extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ses_domain_mail_from aws_ses_domain_mail_from} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SesDomainMailFromConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ses_domain_mail_from',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._behaviorOnMxFailure = config.behaviorOnMxFailure;
        this._domain = config.domain;
        this._id = config.id;
        this._mailFromDomain = config.mailFromDomain;
    }
    get behaviorOnMxFailure() {
        return this.getStringAttribute('behavior_on_mx_failure');
    }
    set behaviorOnMxFailure(value) {
        this._behaviorOnMxFailure = value;
    }
    resetBehaviorOnMxFailure() {
        this._behaviorOnMxFailure = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get behaviorOnMxFailureInput() {
        return this._behaviorOnMxFailure;
    }
    get domain() {
        return this.getStringAttribute('domain');
    }
    set domain(value) {
        this._domain = value;
    }
    // Temporarily expose input value. Use with caution.
    get domainInput() {
        return this._domain;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get mailFromDomain() {
        return this.getStringAttribute('mail_from_domain');
    }
    set mailFromDomain(value) {
        this._mailFromDomain = value;
    }
    // Temporarily expose input value. Use with caution.
    get mailFromDomainInput() {
        return this._mailFromDomain;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            behavior_on_mx_failure: cdktf.stringToTerraform(this._behaviorOnMxFailure),
            domain: cdktf.stringToTerraform(this._domain),
            id: cdktf.stringToTerraform(this._id),
            mail_from_domain: cdktf.stringToTerraform(this._mailFromDomain),
        };
    }
}
exports.SesDomainMailFrom = SesDomainMailFrom;
_a = JSII_RTTI_SYMBOL_1;
SesDomainMailFrom[_a] = { fqn: "@cdktf/provider-aws.sesDomainMailFrom.SesDomainMailFrom", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
SesDomainMailFrom.tfResourceType = "aws_ses_domain_mail_from";
