// @cdktf/provider-aws/lib/networkfirewall-firewall/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkfirewallFirewall = exports.NetworkfirewallFirewallSubnetMappingList = exports.NetworkfirewallFirewallSubnetMappingOutputReference = exports.networkfirewallFirewallSubnetMappingToTerraform = exports.NetworkfirewallFirewallFirewallStatusList = exports.NetworkfirewallFirewallFirewallStatusOutputReference = exports.networkfirewallFirewallFirewallStatusToTerraform = exports.NetworkfirewallFirewallFirewallStatusSyncStatesList = exports.NetworkfirewallFirewallFirewallStatusSyncStatesOutputReference = exports.networkfirewallFirewallFirewallStatusSyncStatesToTerraform = exports.NetworkfirewallFirewallFirewallStatusSyncStatesAttachmentList = exports.NetworkfirewallFirewallFirewallStatusSyncStatesAttachmentOutputReference = exports.networkfirewallFirewallFirewallStatusSyncStatesAttachmentToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function networkfirewallFirewallFirewallStatusSyncStatesAttachmentToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.networkfirewallFirewallFirewallStatusSyncStatesAttachmentToTerraform = networkfirewallFirewallFirewallStatusSyncStatesAttachmentToTerraform;
class NetworkfirewallFirewallFirewallStatusSyncStatesAttachmentOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // endpoint_id - computed: true, optional: false, required: false
    get endpointId() {
        return this.getStringAttribute('endpoint_id');
    }
    // subnet_id - computed: true, optional: false, required: false
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
}
exports.NetworkfirewallFirewallFirewallStatusSyncStatesAttachmentOutputReference = NetworkfirewallFirewallFirewallStatusSyncStatesAttachmentOutputReference;
_a = JSII_RTTI_SYMBOL_1;
NetworkfirewallFirewallFirewallStatusSyncStatesAttachmentOutputReference[_a] = { fqn: "@cdktf/provider-aws.networkfirewallFirewall.NetworkfirewallFirewallFirewallStatusSyncStatesAttachmentOutputReference", version: "10.0.12" };
class NetworkfirewallFirewallFirewallStatusSyncStatesAttachmentList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkfirewallFirewallFirewallStatusSyncStatesAttachmentOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkfirewallFirewallFirewallStatusSyncStatesAttachmentList = NetworkfirewallFirewallFirewallStatusSyncStatesAttachmentList;
_b = JSII_RTTI_SYMBOL_1;
NetworkfirewallFirewallFirewallStatusSyncStatesAttachmentList[_b] = { fqn: "@cdktf/provider-aws.networkfirewallFirewall.NetworkfirewallFirewallFirewallStatusSyncStatesAttachmentList", version: "10.0.12" };
function networkfirewallFirewallFirewallStatusSyncStatesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.networkfirewallFirewallFirewallStatusSyncStatesToTerraform = networkfirewallFirewallFirewallStatusSyncStatesToTerraform;
class NetworkfirewallFirewallFirewallStatusSyncStatesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // attachment - computed: true, optional: false, required: false
        this._attachment = new NetworkfirewallFirewallFirewallStatusSyncStatesAttachmentList(this, "attachment", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get attachment() {
        return this._attachment;
    }
    // availability_zone - computed: true, optional: false, required: false
    get availabilityZone() {
        return this.getStringAttribute('availability_zone');
    }
}
exports.NetworkfirewallFirewallFirewallStatusSyncStatesOutputReference = NetworkfirewallFirewallFirewallStatusSyncStatesOutputReference;
_c = JSII_RTTI_SYMBOL_1;
NetworkfirewallFirewallFirewallStatusSyncStatesOutputReference[_c] = { fqn: "@cdktf/provider-aws.networkfirewallFirewall.NetworkfirewallFirewallFirewallStatusSyncStatesOutputReference", version: "10.0.12" };
class NetworkfirewallFirewallFirewallStatusSyncStatesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkfirewallFirewallFirewallStatusSyncStatesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkfirewallFirewallFirewallStatusSyncStatesList = NetworkfirewallFirewallFirewallStatusSyncStatesList;
_d = JSII_RTTI_SYMBOL_1;
NetworkfirewallFirewallFirewallStatusSyncStatesList[_d] = { fqn: "@cdktf/provider-aws.networkfirewallFirewall.NetworkfirewallFirewallFirewallStatusSyncStatesList", version: "10.0.12" };
function networkfirewallFirewallFirewallStatusToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.networkfirewallFirewallFirewallStatusToTerraform = networkfirewallFirewallFirewallStatusToTerraform;
class NetworkfirewallFirewallFirewallStatusOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // sync_states - computed: true, optional: false, required: false
        this._syncStates = new NetworkfirewallFirewallFirewallStatusSyncStatesList(this, "sync_states", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get syncStates() {
        return this._syncStates;
    }
}
exports.NetworkfirewallFirewallFirewallStatusOutputReference = NetworkfirewallFirewallFirewallStatusOutputReference;
_e = JSII_RTTI_SYMBOL_1;
NetworkfirewallFirewallFirewallStatusOutputReference[_e] = { fqn: "@cdktf/provider-aws.networkfirewallFirewall.NetworkfirewallFirewallFirewallStatusOutputReference", version: "10.0.12" };
class NetworkfirewallFirewallFirewallStatusList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkfirewallFirewallFirewallStatusOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkfirewallFirewallFirewallStatusList = NetworkfirewallFirewallFirewallStatusList;
_f = JSII_RTTI_SYMBOL_1;
NetworkfirewallFirewallFirewallStatusList[_f] = { fqn: "@cdktf/provider-aws.networkfirewallFirewall.NetworkfirewallFirewallFirewallStatusList", version: "10.0.12" };
function networkfirewallFirewallSubnetMappingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        subnet_id: cdktf.stringToTerraform(struct.subnetId),
    };
}
exports.networkfirewallFirewallSubnetMappingToTerraform = networkfirewallFirewallSubnetMappingToTerraform;
class NetworkfirewallFirewallSubnetMappingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._subnetId !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetId = this._subnetId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._subnetId = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._subnetId = value.subnetId;
        }
    }
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    set subnetId(value) {
        this._subnetId = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdInput() {
        return this._subnetId;
    }
}
exports.NetworkfirewallFirewallSubnetMappingOutputReference = NetworkfirewallFirewallSubnetMappingOutputReference;
_g = JSII_RTTI_SYMBOL_1;
NetworkfirewallFirewallSubnetMappingOutputReference[_g] = { fqn: "@cdktf/provider-aws.networkfirewallFirewall.NetworkfirewallFirewallSubnetMappingOutputReference", version: "10.0.12" };
class NetworkfirewallFirewallSubnetMappingList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkfirewallFirewallSubnetMappingOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkfirewallFirewallSubnetMappingList = NetworkfirewallFirewallSubnetMappingList;
_h = JSII_RTTI_SYMBOL_1;
NetworkfirewallFirewallSubnetMappingList[_h] = { fqn: "@cdktf/provider-aws.networkfirewallFirewall.NetworkfirewallFirewallSubnetMappingList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/networkfirewall_firewall aws_networkfirewall_firewall}
*/
class NetworkfirewallFirewall extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/networkfirewall_firewall aws_networkfirewall_firewall} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options NetworkfirewallFirewallConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_networkfirewall_firewall',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // firewall_status - computed: true, optional: false, required: false
        this._firewallStatus = new NetworkfirewallFirewallFirewallStatusList(this, "firewall_status", false);
        // subnet_mapping - computed: false, optional: false, required: true
        this._subnetMapping = new NetworkfirewallFirewallSubnetMappingList(this, "subnet_mapping", true);
        this._deleteProtection = config.deleteProtection;
        this._description = config.description;
        this._firewallPolicyArn = config.firewallPolicyArn;
        this._firewallPolicyChangeProtection = config.firewallPolicyChangeProtection;
        this._id = config.id;
        this._name = config.name;
        this._subnetChangeProtection = config.subnetChangeProtection;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._vpcId = config.vpcId;
        this._subnetMapping.internalValue = config.subnetMapping;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get deleteProtection() {
        return this.getBooleanAttribute('delete_protection');
    }
    set deleteProtection(value) {
        this._deleteProtection = value;
    }
    resetDeleteProtection() {
        this._deleteProtection = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteProtectionInput() {
        return this._deleteProtection;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get firewallPolicyArn() {
        return this.getStringAttribute('firewall_policy_arn');
    }
    set firewallPolicyArn(value) {
        this._firewallPolicyArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get firewallPolicyArnInput() {
        return this._firewallPolicyArn;
    }
    get firewallPolicyChangeProtection() {
        return this.getBooleanAttribute('firewall_policy_change_protection');
    }
    set firewallPolicyChangeProtection(value) {
        this._firewallPolicyChangeProtection = value;
    }
    resetFirewallPolicyChangeProtection() {
        this._firewallPolicyChangeProtection = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get firewallPolicyChangeProtectionInput() {
        return this._firewallPolicyChangeProtection;
    }
    get firewallStatus() {
        return this._firewallStatus;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get subnetChangeProtection() {
        return this.getBooleanAttribute('subnet_change_protection');
    }
    set subnetChangeProtection(value) {
        this._subnetChangeProtection = value;
    }
    resetSubnetChangeProtection() {
        this._subnetChangeProtection = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetChangeProtectionInput() {
        return this._subnetChangeProtection;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // update_token - computed: true, optional: false, required: false
    get updateToken() {
        return this.getStringAttribute('update_token');
    }
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    set vpcId(value) {
        this._vpcId = value;
    }
    // Temporarily expose input value. Use with caution.
    get vpcIdInput() {
        return this._vpcId;
    }
    get subnetMapping() {
        return this._subnetMapping;
    }
    putSubnetMapping(value) {
        this._subnetMapping.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetMappingInput() {
        return this._subnetMapping.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            delete_protection: cdktf.booleanToTerraform(this._deleteProtection),
            description: cdktf.stringToTerraform(this._description),
            firewall_policy_arn: cdktf.stringToTerraform(this._firewallPolicyArn),
            firewall_policy_change_protection: cdktf.booleanToTerraform(this._firewallPolicyChangeProtection),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            subnet_change_protection: cdktf.booleanToTerraform(this._subnetChangeProtection),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            vpc_id: cdktf.stringToTerraform(this._vpcId),
            subnet_mapping: cdktf.listMapper(networkfirewallFirewallSubnetMappingToTerraform, true)(this._subnetMapping.internalValue),
        };
    }
}
exports.NetworkfirewallFirewall = NetworkfirewallFirewall;
_j = JSII_RTTI_SYMBOL_1;
NetworkfirewallFirewall[_j] = { fqn: "@cdktf/provider-aws.networkfirewallFirewall.NetworkfirewallFirewall", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
NetworkfirewallFirewall.tfResourceType = "aws_networkfirewall_firewall";
