// @cdktf/provider-aws/lib/data-aws-connect-user-hierarchy-group/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsConnectUserHierarchyGroup = exports.DataAwsConnectUserHierarchyGroupHierarchyPathList = exports.DataAwsConnectUserHierarchyGroupHierarchyPathOutputReference = exports.dataAwsConnectUserHierarchyGroupHierarchyPathToTerraform = exports.DataAwsConnectUserHierarchyGroupHierarchyPathLevelTwoList = exports.DataAwsConnectUserHierarchyGroupHierarchyPathLevelTwoOutputReference = exports.dataAwsConnectUserHierarchyGroupHierarchyPathLevelTwoToTerraform = exports.DataAwsConnectUserHierarchyGroupHierarchyPathLevelThreeList = exports.DataAwsConnectUserHierarchyGroupHierarchyPathLevelThreeOutputReference = exports.dataAwsConnectUserHierarchyGroupHierarchyPathLevelThreeToTerraform = exports.DataAwsConnectUserHierarchyGroupHierarchyPathLevelOneList = exports.DataAwsConnectUserHierarchyGroupHierarchyPathLevelOneOutputReference = exports.dataAwsConnectUserHierarchyGroupHierarchyPathLevelOneToTerraform = exports.DataAwsConnectUserHierarchyGroupHierarchyPathLevelFourList = exports.DataAwsConnectUserHierarchyGroupHierarchyPathLevelFourOutputReference = exports.dataAwsConnectUserHierarchyGroupHierarchyPathLevelFourToTerraform = exports.DataAwsConnectUserHierarchyGroupHierarchyPathLevelFiveList = exports.DataAwsConnectUserHierarchyGroupHierarchyPathLevelFiveOutputReference = exports.dataAwsConnectUserHierarchyGroupHierarchyPathLevelFiveToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsConnectUserHierarchyGroupHierarchyPathLevelFiveToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsConnectUserHierarchyGroupHierarchyPathLevelFiveToTerraform = dataAwsConnectUserHierarchyGroupHierarchyPathLevelFiveToTerraform;
class DataAwsConnectUserHierarchyGroupHierarchyPathLevelFiveOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataAwsConnectUserHierarchyGroupHierarchyPathLevelFiveOutputReference = DataAwsConnectUserHierarchyGroupHierarchyPathLevelFiveOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyGroupHierarchyPathLevelFiveOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyGroup.DataAwsConnectUserHierarchyGroupHierarchyPathLevelFiveOutputReference", version: "10.0.12" };
class DataAwsConnectUserHierarchyGroupHierarchyPathLevelFiveList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsConnectUserHierarchyGroupHierarchyPathLevelFiveOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsConnectUserHierarchyGroupHierarchyPathLevelFiveList = DataAwsConnectUserHierarchyGroupHierarchyPathLevelFiveList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyGroupHierarchyPathLevelFiveList[_b] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyGroup.DataAwsConnectUserHierarchyGroupHierarchyPathLevelFiveList", version: "10.0.12" };
function dataAwsConnectUserHierarchyGroupHierarchyPathLevelFourToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsConnectUserHierarchyGroupHierarchyPathLevelFourToTerraform = dataAwsConnectUserHierarchyGroupHierarchyPathLevelFourToTerraform;
class DataAwsConnectUserHierarchyGroupHierarchyPathLevelFourOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataAwsConnectUserHierarchyGroupHierarchyPathLevelFourOutputReference = DataAwsConnectUserHierarchyGroupHierarchyPathLevelFourOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyGroupHierarchyPathLevelFourOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyGroup.DataAwsConnectUserHierarchyGroupHierarchyPathLevelFourOutputReference", version: "10.0.12" };
class DataAwsConnectUserHierarchyGroupHierarchyPathLevelFourList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsConnectUserHierarchyGroupHierarchyPathLevelFourOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsConnectUserHierarchyGroupHierarchyPathLevelFourList = DataAwsConnectUserHierarchyGroupHierarchyPathLevelFourList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyGroupHierarchyPathLevelFourList[_d] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyGroup.DataAwsConnectUserHierarchyGroupHierarchyPathLevelFourList", version: "10.0.12" };
function dataAwsConnectUserHierarchyGroupHierarchyPathLevelOneToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsConnectUserHierarchyGroupHierarchyPathLevelOneToTerraform = dataAwsConnectUserHierarchyGroupHierarchyPathLevelOneToTerraform;
class DataAwsConnectUserHierarchyGroupHierarchyPathLevelOneOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataAwsConnectUserHierarchyGroupHierarchyPathLevelOneOutputReference = DataAwsConnectUserHierarchyGroupHierarchyPathLevelOneOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyGroupHierarchyPathLevelOneOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyGroup.DataAwsConnectUserHierarchyGroupHierarchyPathLevelOneOutputReference", version: "10.0.12" };
class DataAwsConnectUserHierarchyGroupHierarchyPathLevelOneList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsConnectUserHierarchyGroupHierarchyPathLevelOneOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsConnectUserHierarchyGroupHierarchyPathLevelOneList = DataAwsConnectUserHierarchyGroupHierarchyPathLevelOneList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyGroupHierarchyPathLevelOneList[_f] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyGroup.DataAwsConnectUserHierarchyGroupHierarchyPathLevelOneList", version: "10.0.12" };
function dataAwsConnectUserHierarchyGroupHierarchyPathLevelThreeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsConnectUserHierarchyGroupHierarchyPathLevelThreeToTerraform = dataAwsConnectUserHierarchyGroupHierarchyPathLevelThreeToTerraform;
class DataAwsConnectUserHierarchyGroupHierarchyPathLevelThreeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataAwsConnectUserHierarchyGroupHierarchyPathLevelThreeOutputReference = DataAwsConnectUserHierarchyGroupHierarchyPathLevelThreeOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyGroupHierarchyPathLevelThreeOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyGroup.DataAwsConnectUserHierarchyGroupHierarchyPathLevelThreeOutputReference", version: "10.0.12" };
class DataAwsConnectUserHierarchyGroupHierarchyPathLevelThreeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsConnectUserHierarchyGroupHierarchyPathLevelThreeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsConnectUserHierarchyGroupHierarchyPathLevelThreeList = DataAwsConnectUserHierarchyGroupHierarchyPathLevelThreeList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyGroupHierarchyPathLevelThreeList[_h] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyGroup.DataAwsConnectUserHierarchyGroupHierarchyPathLevelThreeList", version: "10.0.12" };
function dataAwsConnectUserHierarchyGroupHierarchyPathLevelTwoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsConnectUserHierarchyGroupHierarchyPathLevelTwoToTerraform = dataAwsConnectUserHierarchyGroupHierarchyPathLevelTwoToTerraform;
class DataAwsConnectUserHierarchyGroupHierarchyPathLevelTwoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataAwsConnectUserHierarchyGroupHierarchyPathLevelTwoOutputReference = DataAwsConnectUserHierarchyGroupHierarchyPathLevelTwoOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyGroupHierarchyPathLevelTwoOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyGroup.DataAwsConnectUserHierarchyGroupHierarchyPathLevelTwoOutputReference", version: "10.0.12" };
class DataAwsConnectUserHierarchyGroupHierarchyPathLevelTwoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsConnectUserHierarchyGroupHierarchyPathLevelTwoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsConnectUserHierarchyGroupHierarchyPathLevelTwoList = DataAwsConnectUserHierarchyGroupHierarchyPathLevelTwoList;
_k = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyGroupHierarchyPathLevelTwoList[_k] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyGroup.DataAwsConnectUserHierarchyGroupHierarchyPathLevelTwoList", version: "10.0.12" };
function dataAwsConnectUserHierarchyGroupHierarchyPathToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsConnectUserHierarchyGroupHierarchyPathToTerraform = dataAwsConnectUserHierarchyGroupHierarchyPathToTerraform;
class DataAwsConnectUserHierarchyGroupHierarchyPathOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // level_five - computed: true, optional: false, required: false
        this._levelFive = new DataAwsConnectUserHierarchyGroupHierarchyPathLevelFiveList(this, "level_five", false);
        // level_four - computed: true, optional: false, required: false
        this._levelFour = new DataAwsConnectUserHierarchyGroupHierarchyPathLevelFourList(this, "level_four", false);
        // level_one - computed: true, optional: false, required: false
        this._levelOne = new DataAwsConnectUserHierarchyGroupHierarchyPathLevelOneList(this, "level_one", false);
        // level_three - computed: true, optional: false, required: false
        this._levelThree = new DataAwsConnectUserHierarchyGroupHierarchyPathLevelThreeList(this, "level_three", false);
        // level_two - computed: true, optional: false, required: false
        this._levelTwo = new DataAwsConnectUserHierarchyGroupHierarchyPathLevelTwoList(this, "level_two", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get levelFive() {
        return this._levelFive;
    }
    get levelFour() {
        return this._levelFour;
    }
    get levelOne() {
        return this._levelOne;
    }
    get levelThree() {
        return this._levelThree;
    }
    get levelTwo() {
        return this._levelTwo;
    }
}
exports.DataAwsConnectUserHierarchyGroupHierarchyPathOutputReference = DataAwsConnectUserHierarchyGroupHierarchyPathOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyGroupHierarchyPathOutputReference[_l] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyGroup.DataAwsConnectUserHierarchyGroupHierarchyPathOutputReference", version: "10.0.12" };
class DataAwsConnectUserHierarchyGroupHierarchyPathList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsConnectUserHierarchyGroupHierarchyPathOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsConnectUserHierarchyGroupHierarchyPathList = DataAwsConnectUserHierarchyGroupHierarchyPathList;
_m = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyGroupHierarchyPathList[_m] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyGroup.DataAwsConnectUserHierarchyGroupHierarchyPathList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/connect_user_hierarchy_group aws_connect_user_hierarchy_group}
*/
class DataAwsConnectUserHierarchyGroup extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/connect_user_hierarchy_group aws_connect_user_hierarchy_group} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsConnectUserHierarchyGroupConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_connect_user_hierarchy_group',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // hierarchy_path - computed: true, optional: false, required: false
        this._hierarchyPath = new DataAwsConnectUserHierarchyGroupHierarchyPathList(this, "hierarchy_path", false);
        this._hierarchyGroupId = config.hierarchyGroupId;
        this._id = config.id;
        this._instanceId = config.instanceId;
        this._name = config.name;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get hierarchyGroupId() {
        return this.getStringAttribute('hierarchy_group_id');
    }
    set hierarchyGroupId(value) {
        this._hierarchyGroupId = value;
    }
    resetHierarchyGroupId() {
        this._hierarchyGroupId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hierarchyGroupIdInput() {
        return this._hierarchyGroupId;
    }
    get hierarchyPath() {
        return this._hierarchyPath;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instanceId() {
        return this.getStringAttribute('instance_id');
    }
    set instanceId(value) {
        this._instanceId = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceIdInput() {
        return this._instanceId;
    }
    // level_id - computed: true, optional: false, required: false
    get levelId() {
        return this.getStringAttribute('level_id');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            hierarchy_group_id: cdktf.stringToTerraform(this._hierarchyGroupId),
            id: cdktf.stringToTerraform(this._id),
            instance_id: cdktf.stringToTerraform(this._instanceId),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
}
exports.DataAwsConnectUserHierarchyGroup = DataAwsConnectUserHierarchyGroup;
_o = JSII_RTTI_SYMBOL_1;
DataAwsConnectUserHierarchyGroup[_o] = { fqn: "@cdktf/provider-aws.dataAwsConnectUserHierarchyGroup.DataAwsConnectUserHierarchyGroup", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsConnectUserHierarchyGroup.tfResourceType = "aws_connect_user_hierarchy_group";
