// @cdktf/provider-aws/lib/msk-scram-secret-association/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MskScramSecretAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/msk_scram_secret_association aws_msk_scram_secret_association}
*/
class MskScramSecretAssociation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/msk_scram_secret_association aws_msk_scram_secret_association} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options MskScramSecretAssociationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_msk_scram_secret_association',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._clusterArn = config.clusterArn;
        this._id = config.id;
        this._secretArnList = config.secretArnList;
    }
    get clusterArn() {
        return this.getStringAttribute('cluster_arn');
    }
    set clusterArn(value) {
        this._clusterArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterArnInput() {
        return this._clusterArn;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get secretArnList() {
        return cdktf.Fn.tolist(this.getListAttribute('secret_arn_list'));
    }
    set secretArnList(value) {
        this._secretArnList = value;
    }
    // Temporarily expose input value. Use with caution.
    get secretArnListInput() {
        return this._secretArnList;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_arn: cdktf.stringToTerraform(this._clusterArn),
            id: cdktf.stringToTerraform(this._id),
            secret_arn_list: cdktf.listMapper(cdktf.stringToTerraform, false)(this._secretArnList),
        };
    }
}
exports.MskScramSecretAssociation = MskScramSecretAssociation;
_a = JSII_RTTI_SYMBOL_1;
MskScramSecretAssociation[_a] = { fqn: "@cdktf/provider-aws.mskScramSecretAssociation.MskScramSecretAssociation", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
MskScramSecretAssociation.tfResourceType = "aws_msk_scram_secret_association";
