// @cdktf/provider-aws/lib/data-aws-globalaccelerator-accelerator/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsGlobalacceleratorAccelerator = exports.DataAwsGlobalacceleratorAcceleratorIpSetsList = exports.DataAwsGlobalacceleratorAcceleratorIpSetsOutputReference = exports.dataAwsGlobalacceleratorAcceleratorIpSetsToTerraform = exports.DataAwsGlobalacceleratorAcceleratorAttributesList = exports.DataAwsGlobalacceleratorAcceleratorAttributesOutputReference = exports.dataAwsGlobalacceleratorAcceleratorAttributesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsGlobalacceleratorAcceleratorAttributesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsGlobalacceleratorAcceleratorAttributesToTerraform = dataAwsGlobalacceleratorAcceleratorAttributesToTerraform;
class DataAwsGlobalacceleratorAcceleratorAttributesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // flow_logs_enabled - computed: true, optional: false, required: false
    get flowLogsEnabled() {
        return this.getBooleanAttribute('flow_logs_enabled');
    }
    // flow_logs_s3_bucket - computed: true, optional: false, required: false
    get flowLogsS3Bucket() {
        return this.getStringAttribute('flow_logs_s3_bucket');
    }
    // flow_logs_s3_prefix - computed: true, optional: false, required: false
    get flowLogsS3Prefix() {
        return this.getStringAttribute('flow_logs_s3_prefix');
    }
}
exports.DataAwsGlobalacceleratorAcceleratorAttributesOutputReference = DataAwsGlobalacceleratorAcceleratorAttributesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsGlobalacceleratorAcceleratorAttributesOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsGlobalacceleratorAccelerator.DataAwsGlobalacceleratorAcceleratorAttributesOutputReference", version: "10.0.12" };
class DataAwsGlobalacceleratorAcceleratorAttributesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsGlobalacceleratorAcceleratorAttributesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsGlobalacceleratorAcceleratorAttributesList = DataAwsGlobalacceleratorAcceleratorAttributesList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsGlobalacceleratorAcceleratorAttributesList[_b] = { fqn: "@cdktf/provider-aws.dataAwsGlobalacceleratorAccelerator.DataAwsGlobalacceleratorAcceleratorAttributesList", version: "10.0.12" };
function dataAwsGlobalacceleratorAcceleratorIpSetsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsGlobalacceleratorAcceleratorIpSetsToTerraform = dataAwsGlobalacceleratorAcceleratorIpSetsToTerraform;
class DataAwsGlobalacceleratorAcceleratorIpSetsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ip_addresses - computed: true, optional: false, required: false
    get ipAddresses() {
        return this.getListAttribute('ip_addresses');
    }
    // ip_family - computed: true, optional: false, required: false
    get ipFamily() {
        return this.getStringAttribute('ip_family');
    }
}
exports.DataAwsGlobalacceleratorAcceleratorIpSetsOutputReference = DataAwsGlobalacceleratorAcceleratorIpSetsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsGlobalacceleratorAcceleratorIpSetsOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsGlobalacceleratorAccelerator.DataAwsGlobalacceleratorAcceleratorIpSetsOutputReference", version: "10.0.12" };
class DataAwsGlobalacceleratorAcceleratorIpSetsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsGlobalacceleratorAcceleratorIpSetsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsGlobalacceleratorAcceleratorIpSetsList = DataAwsGlobalacceleratorAcceleratorIpSetsList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsGlobalacceleratorAcceleratorIpSetsList[_d] = { fqn: "@cdktf/provider-aws.dataAwsGlobalacceleratorAccelerator.DataAwsGlobalacceleratorAcceleratorIpSetsList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/globalaccelerator_accelerator aws_globalaccelerator_accelerator}
*/
class DataAwsGlobalacceleratorAccelerator extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/globalaccelerator_accelerator aws_globalaccelerator_accelerator} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsGlobalacceleratorAcceleratorConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_globalaccelerator_accelerator',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // attributes - computed: true, optional: false, required: false
        this._attributes = new DataAwsGlobalacceleratorAcceleratorAttributesList(this, "attributes", false);
        // ip_sets - computed: true, optional: false, required: false
        this._ipSets = new DataAwsGlobalacceleratorAcceleratorIpSetsList(this, "ip_sets", false);
        this._arn = config.arn;
        this._id = config.id;
        this._name = config.name;
        this._tags = config.tags;
    }
    get arn() {
        return this.getStringAttribute('arn');
    }
    set arn(value) {
        this._arn = value;
    }
    resetArn() {
        this._arn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get arnInput() {
        return this._arn;
    }
    get attributes() {
        return this._attributes;
    }
    // dns_name - computed: true, optional: false, required: false
    get dnsName() {
        return this.getStringAttribute('dns_name');
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    // hosted_zone_id - computed: true, optional: false, required: false
    get hostedZoneId() {
        return this.getStringAttribute('hosted_zone_id');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // ip_address_type - computed: true, optional: false, required: false
    get ipAddressType() {
        return this.getStringAttribute('ip_address_type');
    }
    get ipSets() {
        return this._ipSets;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            arn: cdktf.stringToTerraform(this._arn),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
}
exports.DataAwsGlobalacceleratorAccelerator = DataAwsGlobalacceleratorAccelerator;
_e = JSII_RTTI_SYMBOL_1;
DataAwsGlobalacceleratorAccelerator[_e] = { fqn: "@cdktf/provider-aws.dataAwsGlobalacceleratorAccelerator.DataAwsGlobalacceleratorAccelerator", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsGlobalacceleratorAccelerator.tfResourceType = "aws_globalaccelerator_accelerator";
