// @cdktf/provider-aws/lib/data-aws-cloudwatch-log-groups/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsCloudwatchLogGroups = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/cloudwatch_log_groups aws_cloudwatch_log_groups}
*/
class DataAwsCloudwatchLogGroups extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/cloudwatch_log_groups aws_cloudwatch_log_groups} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsCloudwatchLogGroupsConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_cloudwatch_log_groups',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._logGroupNamePrefix = config.logGroupNamePrefix;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arns - computed: true, optional: false, required: false
    get arns() {
        return cdktf.Fn.tolist(this.getListAttribute('arns'));
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get logGroupNamePrefix() {
        return this.getStringAttribute('log_group_name_prefix');
    }
    set logGroupNamePrefix(value) {
        this._logGroupNamePrefix = value;
    }
    resetLogGroupNamePrefix() {
        this._logGroupNamePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get logGroupNamePrefixInput() {
        return this._logGroupNamePrefix;
    }
    // log_group_names - computed: true, optional: false, required: false
    get logGroupNames() {
        return cdktf.Fn.tolist(this.getListAttribute('log_group_names'));
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            log_group_name_prefix: cdktf.stringToTerraform(this._logGroupNamePrefix),
        };
    }
}
exports.DataAwsCloudwatchLogGroups = DataAwsCloudwatchLogGroups;
_a = JSII_RTTI_SYMBOL_1;
DataAwsCloudwatchLogGroups[_a] = { fqn: "@cdktf/provider-aws.dataAwsCloudwatchLogGroups.DataAwsCloudwatchLogGroups", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsCloudwatchLogGroups.tfResourceType = "aws_cloudwatch_log_groups";
