// @cdktf/provider-aws/lib/sagemaker-feature-group/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SagemakerFeatureGroup = exports.SagemakerFeatureGroupOnlineStoreConfigOutputReference = exports.sagemakerFeatureGroupOnlineStoreConfigToTerraform = exports.SagemakerFeatureGroupOnlineStoreConfigSecurityConfigOutputReference = exports.sagemakerFeatureGroupOnlineStoreConfigSecurityConfigToTerraform = exports.SagemakerFeatureGroupOfflineStoreConfigOutputReference = exports.sagemakerFeatureGroupOfflineStoreConfigToTerraform = exports.SagemakerFeatureGroupOfflineStoreConfigS3StorageConfigOutputReference = exports.sagemakerFeatureGroupOfflineStoreConfigS3StorageConfigToTerraform = exports.SagemakerFeatureGroupOfflineStoreConfigDataCatalogConfigOutputReference = exports.sagemakerFeatureGroupOfflineStoreConfigDataCatalogConfigToTerraform = exports.SagemakerFeatureGroupFeatureDefinitionList = exports.SagemakerFeatureGroupFeatureDefinitionOutputReference = exports.sagemakerFeatureGroupFeatureDefinitionToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function sagemakerFeatureGroupFeatureDefinitionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        feature_name: cdktf.stringToTerraform(struct.featureName),
        feature_type: cdktf.stringToTerraform(struct.featureType),
    };
}
exports.sagemakerFeatureGroupFeatureDefinitionToTerraform = sagemakerFeatureGroupFeatureDefinitionToTerraform;
class SagemakerFeatureGroupFeatureDefinitionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._featureName !== undefined) {
            hasAnyValues = true;
            internalValueResult.featureName = this._featureName;
        }
        if (this._featureType !== undefined) {
            hasAnyValues = true;
            internalValueResult.featureType = this._featureType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._featureName = undefined;
            this._featureType = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._featureName = value.featureName;
            this._featureType = value.featureType;
        }
    }
    get featureName() {
        return this.getStringAttribute('feature_name');
    }
    set featureName(value) {
        this._featureName = value;
    }
    resetFeatureName() {
        this._featureName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get featureNameInput() {
        return this._featureName;
    }
    get featureType() {
        return this.getStringAttribute('feature_type');
    }
    set featureType(value) {
        this._featureType = value;
    }
    resetFeatureType() {
        this._featureType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get featureTypeInput() {
        return this._featureType;
    }
}
exports.SagemakerFeatureGroupFeatureDefinitionOutputReference = SagemakerFeatureGroupFeatureDefinitionOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SagemakerFeatureGroupFeatureDefinitionOutputReference[_a] = { fqn: "@cdktf/provider-aws.sagemakerFeatureGroup.SagemakerFeatureGroupFeatureDefinitionOutputReference", version: "10.0.12" };
class SagemakerFeatureGroupFeatureDefinitionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SagemakerFeatureGroupFeatureDefinitionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SagemakerFeatureGroupFeatureDefinitionList = SagemakerFeatureGroupFeatureDefinitionList;
_b = JSII_RTTI_SYMBOL_1;
SagemakerFeatureGroupFeatureDefinitionList[_b] = { fqn: "@cdktf/provider-aws.sagemakerFeatureGroup.SagemakerFeatureGroupFeatureDefinitionList", version: "10.0.12" };
function sagemakerFeatureGroupOfflineStoreConfigDataCatalogConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        catalog: cdktf.stringToTerraform(struct.catalog),
        database: cdktf.stringToTerraform(struct.database),
        table_name: cdktf.stringToTerraform(struct.tableName),
    };
}
exports.sagemakerFeatureGroupOfflineStoreConfigDataCatalogConfigToTerraform = sagemakerFeatureGroupOfflineStoreConfigDataCatalogConfigToTerraform;
class SagemakerFeatureGroupOfflineStoreConfigDataCatalogConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._catalog !== undefined) {
            hasAnyValues = true;
            internalValueResult.catalog = this._catalog;
        }
        if (this._database !== undefined) {
            hasAnyValues = true;
            internalValueResult.database = this._database;
        }
        if (this._tableName !== undefined) {
            hasAnyValues = true;
            internalValueResult.tableName = this._tableName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._catalog = undefined;
            this._database = undefined;
            this._tableName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._catalog = value.catalog;
            this._database = value.database;
            this._tableName = value.tableName;
        }
    }
    get catalog() {
        return this.getStringAttribute('catalog');
    }
    set catalog(value) {
        this._catalog = value;
    }
    resetCatalog() {
        this._catalog = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get catalogInput() {
        return this._catalog;
    }
    get database() {
        return this.getStringAttribute('database');
    }
    set database(value) {
        this._database = value;
    }
    resetDatabase() {
        this._database = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get databaseInput() {
        return this._database;
    }
    get tableName() {
        return this.getStringAttribute('table_name');
    }
    set tableName(value) {
        this._tableName = value;
    }
    resetTableName() {
        this._tableName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tableNameInput() {
        return this._tableName;
    }
}
exports.SagemakerFeatureGroupOfflineStoreConfigDataCatalogConfigOutputReference = SagemakerFeatureGroupOfflineStoreConfigDataCatalogConfigOutputReference;
_c = JSII_RTTI_SYMBOL_1;
SagemakerFeatureGroupOfflineStoreConfigDataCatalogConfigOutputReference[_c] = { fqn: "@cdktf/provider-aws.sagemakerFeatureGroup.SagemakerFeatureGroupOfflineStoreConfigDataCatalogConfigOutputReference", version: "10.0.12" };
function sagemakerFeatureGroupOfflineStoreConfigS3StorageConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        kms_key_id: cdktf.stringToTerraform(struct.kmsKeyId),
        s3_uri: cdktf.stringToTerraform(struct.s3Uri),
    };
}
exports.sagemakerFeatureGroupOfflineStoreConfigS3StorageConfigToTerraform = sagemakerFeatureGroupOfflineStoreConfigS3StorageConfigToTerraform;
class SagemakerFeatureGroupOfflineStoreConfigS3StorageConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._kmsKeyId !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKeyId = this._kmsKeyId;
        }
        if (this._s3Uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3Uri = this._s3Uri;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._kmsKeyId = undefined;
            this._s3Uri = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._kmsKeyId = value.kmsKeyId;
            this._s3Uri = value.s3Uri;
        }
    }
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    resetKmsKeyId() {
        this._kmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
    get s3Uri() {
        return this.getStringAttribute('s3_uri');
    }
    set s3Uri(value) {
        this._s3Uri = value;
    }
    // Temporarily expose input value. Use with caution.
    get s3UriInput() {
        return this._s3Uri;
    }
}
exports.SagemakerFeatureGroupOfflineStoreConfigS3StorageConfigOutputReference = SagemakerFeatureGroupOfflineStoreConfigS3StorageConfigOutputReference;
_d = JSII_RTTI_SYMBOL_1;
SagemakerFeatureGroupOfflineStoreConfigS3StorageConfigOutputReference[_d] = { fqn: "@cdktf/provider-aws.sagemakerFeatureGroup.SagemakerFeatureGroupOfflineStoreConfigS3StorageConfigOutputReference", version: "10.0.12" };
function sagemakerFeatureGroupOfflineStoreConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        disable_glue_table_creation: cdktf.booleanToTerraform(struct.disableGlueTableCreation),
        data_catalog_config: sagemakerFeatureGroupOfflineStoreConfigDataCatalogConfigToTerraform(struct.dataCatalogConfig),
        s3_storage_config: sagemakerFeatureGroupOfflineStoreConfigS3StorageConfigToTerraform(struct.s3StorageConfig),
    };
}
exports.sagemakerFeatureGroupOfflineStoreConfigToTerraform = sagemakerFeatureGroupOfflineStoreConfigToTerraform;
class SagemakerFeatureGroupOfflineStoreConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // data_catalog_config - computed: false, optional: true, required: false
        this._dataCatalogConfig = new SagemakerFeatureGroupOfflineStoreConfigDataCatalogConfigOutputReference(this, "data_catalog_config");
        // s3_storage_config - computed: false, optional: false, required: true
        this._s3StorageConfig = new SagemakerFeatureGroupOfflineStoreConfigS3StorageConfigOutputReference(this, "s3_storage_config");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._disableGlueTableCreation !== undefined) {
            hasAnyValues = true;
            internalValueResult.disableGlueTableCreation = this._disableGlueTableCreation;
        }
        if (this._dataCatalogConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.dataCatalogConfig = this._dataCatalogConfig?.internalValue;
        }
        if (this._s3StorageConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3StorageConfig = this._s3StorageConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._disableGlueTableCreation = undefined;
            this._dataCatalogConfig.internalValue = undefined;
            this._s3StorageConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._disableGlueTableCreation = value.disableGlueTableCreation;
            this._dataCatalogConfig.internalValue = value.dataCatalogConfig;
            this._s3StorageConfig.internalValue = value.s3StorageConfig;
        }
    }
    get disableGlueTableCreation() {
        return this.getBooleanAttribute('disable_glue_table_creation');
    }
    set disableGlueTableCreation(value) {
        this._disableGlueTableCreation = value;
    }
    resetDisableGlueTableCreation() {
        this._disableGlueTableCreation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get disableGlueTableCreationInput() {
        return this._disableGlueTableCreation;
    }
    get dataCatalogConfig() {
        return this._dataCatalogConfig;
    }
    putDataCatalogConfig(value) {
        this._dataCatalogConfig.internalValue = value;
    }
    resetDataCatalogConfig() {
        this._dataCatalogConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataCatalogConfigInput() {
        return this._dataCatalogConfig.internalValue;
    }
    get s3StorageConfig() {
        return this._s3StorageConfig;
    }
    putS3StorageConfig(value) {
        this._s3StorageConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get s3StorageConfigInput() {
        return this._s3StorageConfig.internalValue;
    }
}
exports.SagemakerFeatureGroupOfflineStoreConfigOutputReference = SagemakerFeatureGroupOfflineStoreConfigOutputReference;
_e = JSII_RTTI_SYMBOL_1;
SagemakerFeatureGroupOfflineStoreConfigOutputReference[_e] = { fqn: "@cdktf/provider-aws.sagemakerFeatureGroup.SagemakerFeatureGroupOfflineStoreConfigOutputReference", version: "10.0.12" };
function sagemakerFeatureGroupOnlineStoreConfigSecurityConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        kms_key_id: cdktf.stringToTerraform(struct.kmsKeyId),
    };
}
exports.sagemakerFeatureGroupOnlineStoreConfigSecurityConfigToTerraform = sagemakerFeatureGroupOnlineStoreConfigSecurityConfigToTerraform;
class SagemakerFeatureGroupOnlineStoreConfigSecurityConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._kmsKeyId !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKeyId = this._kmsKeyId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._kmsKeyId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._kmsKeyId = value.kmsKeyId;
        }
    }
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    resetKmsKeyId() {
        this._kmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
}
exports.SagemakerFeatureGroupOnlineStoreConfigSecurityConfigOutputReference = SagemakerFeatureGroupOnlineStoreConfigSecurityConfigOutputReference;
_f = JSII_RTTI_SYMBOL_1;
SagemakerFeatureGroupOnlineStoreConfigSecurityConfigOutputReference[_f] = { fqn: "@cdktf/provider-aws.sagemakerFeatureGroup.SagemakerFeatureGroupOnlineStoreConfigSecurityConfigOutputReference", version: "10.0.12" };
function sagemakerFeatureGroupOnlineStoreConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enable_online_store: cdktf.booleanToTerraform(struct.enableOnlineStore),
        security_config: sagemakerFeatureGroupOnlineStoreConfigSecurityConfigToTerraform(struct.securityConfig),
    };
}
exports.sagemakerFeatureGroupOnlineStoreConfigToTerraform = sagemakerFeatureGroupOnlineStoreConfigToTerraform;
class SagemakerFeatureGroupOnlineStoreConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // security_config - computed: false, optional: true, required: false
        this._securityConfig = new SagemakerFeatureGroupOnlineStoreConfigSecurityConfigOutputReference(this, "security_config");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enableOnlineStore !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableOnlineStore = this._enableOnlineStore;
        }
        if (this._securityConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityConfig = this._securityConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enableOnlineStore = undefined;
            this._securityConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enableOnlineStore = value.enableOnlineStore;
            this._securityConfig.internalValue = value.securityConfig;
        }
    }
    get enableOnlineStore() {
        return this.getBooleanAttribute('enable_online_store');
    }
    set enableOnlineStore(value) {
        this._enableOnlineStore = value;
    }
    resetEnableOnlineStore() {
        this._enableOnlineStore = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableOnlineStoreInput() {
        return this._enableOnlineStore;
    }
    get securityConfig() {
        return this._securityConfig;
    }
    putSecurityConfig(value) {
        this._securityConfig.internalValue = value;
    }
    resetSecurityConfig() {
        this._securityConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityConfigInput() {
        return this._securityConfig.internalValue;
    }
}
exports.SagemakerFeatureGroupOnlineStoreConfigOutputReference = SagemakerFeatureGroupOnlineStoreConfigOutputReference;
_g = JSII_RTTI_SYMBOL_1;
SagemakerFeatureGroupOnlineStoreConfigOutputReference[_g] = { fqn: "@cdktf/provider-aws.sagemakerFeatureGroup.SagemakerFeatureGroupOnlineStoreConfigOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/sagemaker_feature_group aws_sagemaker_feature_group}
*/
class SagemakerFeatureGroup extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/sagemaker_feature_group aws_sagemaker_feature_group} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SagemakerFeatureGroupConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_sagemaker_feature_group',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // feature_definition - computed: false, optional: false, required: true
        this._featureDefinition = new SagemakerFeatureGroupFeatureDefinitionList(this, "feature_definition", false);
        // offline_store_config - computed: false, optional: true, required: false
        this._offlineStoreConfig = new SagemakerFeatureGroupOfflineStoreConfigOutputReference(this, "offline_store_config");
        // online_store_config - computed: false, optional: true, required: false
        this._onlineStoreConfig = new SagemakerFeatureGroupOnlineStoreConfigOutputReference(this, "online_store_config");
        this._description = config.description;
        this._eventTimeFeatureName = config.eventTimeFeatureName;
        this._featureGroupName = config.featureGroupName;
        this._id = config.id;
        this._recordIdentifierFeatureName = config.recordIdentifierFeatureName;
        this._roleArn = config.roleArn;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._featureDefinition.internalValue = config.featureDefinition;
        this._offlineStoreConfig.internalValue = config.offlineStoreConfig;
        this._onlineStoreConfig.internalValue = config.onlineStoreConfig;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get eventTimeFeatureName() {
        return this.getStringAttribute('event_time_feature_name');
    }
    set eventTimeFeatureName(value) {
        this._eventTimeFeatureName = value;
    }
    // Temporarily expose input value. Use with caution.
    get eventTimeFeatureNameInput() {
        return this._eventTimeFeatureName;
    }
    get featureGroupName() {
        return this.getStringAttribute('feature_group_name');
    }
    set featureGroupName(value) {
        this._featureGroupName = value;
    }
    // Temporarily expose input value. Use with caution.
    get featureGroupNameInput() {
        return this._featureGroupName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get recordIdentifierFeatureName() {
        return this.getStringAttribute('record_identifier_feature_name');
    }
    set recordIdentifierFeatureName(value) {
        this._recordIdentifierFeatureName = value;
    }
    // Temporarily expose input value. Use with caution.
    get recordIdentifierFeatureNameInput() {
        return this._recordIdentifierFeatureName;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get featureDefinition() {
        return this._featureDefinition;
    }
    putFeatureDefinition(value) {
        this._featureDefinition.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get featureDefinitionInput() {
        return this._featureDefinition.internalValue;
    }
    get offlineStoreConfig() {
        return this._offlineStoreConfig;
    }
    putOfflineStoreConfig(value) {
        this._offlineStoreConfig.internalValue = value;
    }
    resetOfflineStoreConfig() {
        this._offlineStoreConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get offlineStoreConfigInput() {
        return this._offlineStoreConfig.internalValue;
    }
    get onlineStoreConfig() {
        return this._onlineStoreConfig;
    }
    putOnlineStoreConfig(value) {
        this._onlineStoreConfig.internalValue = value;
    }
    resetOnlineStoreConfig() {
        this._onlineStoreConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get onlineStoreConfigInput() {
        return this._onlineStoreConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            event_time_feature_name: cdktf.stringToTerraform(this._eventTimeFeatureName),
            feature_group_name: cdktf.stringToTerraform(this._featureGroupName),
            id: cdktf.stringToTerraform(this._id),
            record_identifier_feature_name: cdktf.stringToTerraform(this._recordIdentifierFeatureName),
            role_arn: cdktf.stringToTerraform(this._roleArn),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            feature_definition: cdktf.listMapper(sagemakerFeatureGroupFeatureDefinitionToTerraform, true)(this._featureDefinition.internalValue),
            offline_store_config: sagemakerFeatureGroupOfflineStoreConfigToTerraform(this._offlineStoreConfig.internalValue),
            online_store_config: sagemakerFeatureGroupOnlineStoreConfigToTerraform(this._onlineStoreConfig.internalValue),
        };
    }
}
exports.SagemakerFeatureGroup = SagemakerFeatureGroup;
_h = JSII_RTTI_SYMBOL_1;
SagemakerFeatureGroup[_h] = { fqn: "@cdktf/provider-aws.sagemakerFeatureGroup.SagemakerFeatureGroup", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
SagemakerFeatureGroup.tfResourceType = "aws_sagemaker_feature_group";
