// @cdktf/provider-aws/lib/datapipeline-pipeline-definition/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatapipelinePipelineDefinition = exports.DatapipelinePipelineDefinitionPipelineObjectList = exports.DatapipelinePipelineDefinitionPipelineObjectOutputReference = exports.datapipelinePipelineDefinitionPipelineObjectToTerraform = exports.DatapipelinePipelineDefinitionPipelineObjectFieldList = exports.DatapipelinePipelineDefinitionPipelineObjectFieldOutputReference = exports.datapipelinePipelineDefinitionPipelineObjectFieldToTerraform = exports.DatapipelinePipelineDefinitionParameterValueList = exports.DatapipelinePipelineDefinitionParameterValueOutputReference = exports.datapipelinePipelineDefinitionParameterValueToTerraform = exports.DatapipelinePipelineDefinitionParameterObjectList = exports.DatapipelinePipelineDefinitionParameterObjectOutputReference = exports.datapipelinePipelineDefinitionParameterObjectToTerraform = exports.DatapipelinePipelineDefinitionParameterObjectAttributeList = exports.DatapipelinePipelineDefinitionParameterObjectAttributeOutputReference = exports.datapipelinePipelineDefinitionParameterObjectAttributeToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function datapipelinePipelineDefinitionParameterObjectAttributeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        string_value: cdktf.stringToTerraform(struct.stringValue),
    };
}
exports.datapipelinePipelineDefinitionParameterObjectAttributeToTerraform = datapipelinePipelineDefinitionParameterObjectAttributeToTerraform;
class DatapipelinePipelineDefinitionParameterObjectAttributeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._stringValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.stringValue = this._stringValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._key = undefined;
            this._stringValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._key = value.key;
            this._stringValue = value.stringValue;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get stringValue() {
        return this.getStringAttribute('string_value');
    }
    set stringValue(value) {
        this._stringValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get stringValueInput() {
        return this._stringValue;
    }
}
exports.DatapipelinePipelineDefinitionParameterObjectAttributeOutputReference = DatapipelinePipelineDefinitionParameterObjectAttributeOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DatapipelinePipelineDefinitionParameterObjectAttributeOutputReference[_a] = { fqn: "@cdktf/provider-aws.datapipelinePipelineDefinition.DatapipelinePipelineDefinitionParameterObjectAttributeOutputReference", version: "10.0.12" };
class DatapipelinePipelineDefinitionParameterObjectAttributeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DatapipelinePipelineDefinitionParameterObjectAttributeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DatapipelinePipelineDefinitionParameterObjectAttributeList = DatapipelinePipelineDefinitionParameterObjectAttributeList;
_b = JSII_RTTI_SYMBOL_1;
DatapipelinePipelineDefinitionParameterObjectAttributeList[_b] = { fqn: "@cdktf/provider-aws.datapipelinePipelineDefinition.DatapipelinePipelineDefinitionParameterObjectAttributeList", version: "10.0.12" };
function datapipelinePipelineDefinitionParameterObjectToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        id: cdktf.stringToTerraform(struct.id),
        attribute: cdktf.listMapper(datapipelinePipelineDefinitionParameterObjectAttributeToTerraform, true)(struct.attribute),
    };
}
exports.datapipelinePipelineDefinitionParameterObjectToTerraform = datapipelinePipelineDefinitionParameterObjectToTerraform;
class DatapipelinePipelineDefinitionParameterObjectOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // attribute - computed: false, optional: true, required: false
        this._attribute = new DatapipelinePipelineDefinitionParameterObjectAttributeList(this, "attribute", true);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._id !== undefined) {
            hasAnyValues = true;
            internalValueResult.id = this._id;
        }
        if (this._attribute?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.attribute = this._attribute?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._id = undefined;
            this._attribute.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._id = value.id;
            this._attribute.internalValue = value.attribute;
        }
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get attribute() {
        return this._attribute;
    }
    putAttribute(value) {
        this._attribute.internalValue = value;
    }
    resetAttribute() {
        this._attribute.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get attributeInput() {
        return this._attribute.internalValue;
    }
}
exports.DatapipelinePipelineDefinitionParameterObjectOutputReference = DatapipelinePipelineDefinitionParameterObjectOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DatapipelinePipelineDefinitionParameterObjectOutputReference[_c] = { fqn: "@cdktf/provider-aws.datapipelinePipelineDefinition.DatapipelinePipelineDefinitionParameterObjectOutputReference", version: "10.0.12" };
class DatapipelinePipelineDefinitionParameterObjectList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DatapipelinePipelineDefinitionParameterObjectOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DatapipelinePipelineDefinitionParameterObjectList = DatapipelinePipelineDefinitionParameterObjectList;
_d = JSII_RTTI_SYMBOL_1;
DatapipelinePipelineDefinitionParameterObjectList[_d] = { fqn: "@cdktf/provider-aws.datapipelinePipelineDefinition.DatapipelinePipelineDefinitionParameterObjectList", version: "10.0.12" };
function datapipelinePipelineDefinitionParameterValueToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        id: cdktf.stringToTerraform(struct.id),
        string_value: cdktf.stringToTerraform(struct.stringValue),
    };
}
exports.datapipelinePipelineDefinitionParameterValueToTerraform = datapipelinePipelineDefinitionParameterValueToTerraform;
class DatapipelinePipelineDefinitionParameterValueOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._id !== undefined) {
            hasAnyValues = true;
            internalValueResult.id = this._id;
        }
        if (this._stringValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.stringValue = this._stringValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._id = undefined;
            this._stringValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._id = value.id;
            this._stringValue = value.stringValue;
        }
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get stringValue() {
        return this.getStringAttribute('string_value');
    }
    set stringValue(value) {
        this._stringValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get stringValueInput() {
        return this._stringValue;
    }
}
exports.DatapipelinePipelineDefinitionParameterValueOutputReference = DatapipelinePipelineDefinitionParameterValueOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DatapipelinePipelineDefinitionParameterValueOutputReference[_e] = { fqn: "@cdktf/provider-aws.datapipelinePipelineDefinition.DatapipelinePipelineDefinitionParameterValueOutputReference", version: "10.0.12" };
class DatapipelinePipelineDefinitionParameterValueList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DatapipelinePipelineDefinitionParameterValueOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DatapipelinePipelineDefinitionParameterValueList = DatapipelinePipelineDefinitionParameterValueList;
_f = JSII_RTTI_SYMBOL_1;
DatapipelinePipelineDefinitionParameterValueList[_f] = { fqn: "@cdktf/provider-aws.datapipelinePipelineDefinition.DatapipelinePipelineDefinitionParameterValueList", version: "10.0.12" };
function datapipelinePipelineDefinitionPipelineObjectFieldToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        ref_value: cdktf.stringToTerraform(struct.refValue),
        string_value: cdktf.stringToTerraform(struct.stringValue),
    };
}
exports.datapipelinePipelineDefinitionPipelineObjectFieldToTerraform = datapipelinePipelineDefinitionPipelineObjectFieldToTerraform;
class DatapipelinePipelineDefinitionPipelineObjectFieldOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._refValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.refValue = this._refValue;
        }
        if (this._stringValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.stringValue = this._stringValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._key = undefined;
            this._refValue = undefined;
            this._stringValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._key = value.key;
            this._refValue = value.refValue;
            this._stringValue = value.stringValue;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get refValue() {
        return this.getStringAttribute('ref_value');
    }
    set refValue(value) {
        this._refValue = value;
    }
    resetRefValue() {
        this._refValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get refValueInput() {
        return this._refValue;
    }
    get stringValue() {
        return this.getStringAttribute('string_value');
    }
    set stringValue(value) {
        this._stringValue = value;
    }
    resetStringValue() {
        this._stringValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get stringValueInput() {
        return this._stringValue;
    }
}
exports.DatapipelinePipelineDefinitionPipelineObjectFieldOutputReference = DatapipelinePipelineDefinitionPipelineObjectFieldOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DatapipelinePipelineDefinitionPipelineObjectFieldOutputReference[_g] = { fqn: "@cdktf/provider-aws.datapipelinePipelineDefinition.DatapipelinePipelineDefinitionPipelineObjectFieldOutputReference", version: "10.0.12" };
class DatapipelinePipelineDefinitionPipelineObjectFieldList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DatapipelinePipelineDefinitionPipelineObjectFieldOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DatapipelinePipelineDefinitionPipelineObjectFieldList = DatapipelinePipelineDefinitionPipelineObjectFieldList;
_h = JSII_RTTI_SYMBOL_1;
DatapipelinePipelineDefinitionPipelineObjectFieldList[_h] = { fqn: "@cdktf/provider-aws.datapipelinePipelineDefinition.DatapipelinePipelineDefinitionPipelineObjectFieldList", version: "10.0.12" };
function datapipelinePipelineDefinitionPipelineObjectToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        id: cdktf.stringToTerraform(struct.id),
        name: cdktf.stringToTerraform(struct.name),
        field: cdktf.listMapper(datapipelinePipelineDefinitionPipelineObjectFieldToTerraform, true)(struct.field),
    };
}
exports.datapipelinePipelineDefinitionPipelineObjectToTerraform = datapipelinePipelineDefinitionPipelineObjectToTerraform;
class DatapipelinePipelineDefinitionPipelineObjectOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // field - computed: false, optional: true, required: false
        this._field = new DatapipelinePipelineDefinitionPipelineObjectFieldList(this, "field", true);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._id !== undefined) {
            hasAnyValues = true;
            internalValueResult.id = this._id;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._field?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.field = this._field?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._id = undefined;
            this._name = undefined;
            this._field.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._id = value.id;
            this._name = value.name;
            this._field.internalValue = value.field;
        }
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get field() {
        return this._field;
    }
    putField(value) {
        this._field.internalValue = value;
    }
    resetField() {
        this._field.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fieldInput() {
        return this._field.internalValue;
    }
}
exports.DatapipelinePipelineDefinitionPipelineObjectOutputReference = DatapipelinePipelineDefinitionPipelineObjectOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DatapipelinePipelineDefinitionPipelineObjectOutputReference[_j] = { fqn: "@cdktf/provider-aws.datapipelinePipelineDefinition.DatapipelinePipelineDefinitionPipelineObjectOutputReference", version: "10.0.12" };
class DatapipelinePipelineDefinitionPipelineObjectList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DatapipelinePipelineDefinitionPipelineObjectOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DatapipelinePipelineDefinitionPipelineObjectList = DatapipelinePipelineDefinitionPipelineObjectList;
_k = JSII_RTTI_SYMBOL_1;
DatapipelinePipelineDefinitionPipelineObjectList[_k] = { fqn: "@cdktf/provider-aws.datapipelinePipelineDefinition.DatapipelinePipelineDefinitionPipelineObjectList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/datapipeline_pipeline_definition aws_datapipeline_pipeline_definition}
*/
class DatapipelinePipelineDefinition extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/datapipeline_pipeline_definition aws_datapipeline_pipeline_definition} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DatapipelinePipelineDefinitionConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_datapipeline_pipeline_definition',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // parameter_object - computed: false, optional: true, required: false
        this._parameterObject = new DatapipelinePipelineDefinitionParameterObjectList(this, "parameter_object", true);
        // parameter_value - computed: false, optional: true, required: false
        this._parameterValue = new DatapipelinePipelineDefinitionParameterValueList(this, "parameter_value", true);
        // pipeline_object - computed: false, optional: false, required: true
        this._pipelineObject = new DatapipelinePipelineDefinitionPipelineObjectList(this, "pipeline_object", true);
        this._id = config.id;
        this._pipelineId = config.pipelineId;
        this._parameterObject.internalValue = config.parameterObject;
        this._parameterValue.internalValue = config.parameterValue;
        this._pipelineObject.internalValue = config.pipelineObject;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get pipelineId() {
        return this.getStringAttribute('pipeline_id');
    }
    set pipelineId(value) {
        this._pipelineId = value;
    }
    // Temporarily expose input value. Use with caution.
    get pipelineIdInput() {
        return this._pipelineId;
    }
    get parameterObject() {
        return this._parameterObject;
    }
    putParameterObject(value) {
        this._parameterObject.internalValue = value;
    }
    resetParameterObject() {
        this._parameterObject.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parameterObjectInput() {
        return this._parameterObject.internalValue;
    }
    get parameterValue() {
        return this._parameterValue;
    }
    putParameterValue(value) {
        this._parameterValue.internalValue = value;
    }
    resetParameterValue() {
        this._parameterValue.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parameterValueInput() {
        return this._parameterValue.internalValue;
    }
    get pipelineObject() {
        return this._pipelineObject;
    }
    putPipelineObject(value) {
        this._pipelineObject.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get pipelineObjectInput() {
        return this._pipelineObject.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            pipeline_id: cdktf.stringToTerraform(this._pipelineId),
            parameter_object: cdktf.listMapper(datapipelinePipelineDefinitionParameterObjectToTerraform, true)(this._parameterObject.internalValue),
            parameter_value: cdktf.listMapper(datapipelinePipelineDefinitionParameterValueToTerraform, true)(this._parameterValue.internalValue),
            pipeline_object: cdktf.listMapper(datapipelinePipelineDefinitionPipelineObjectToTerraform, true)(this._pipelineObject.internalValue),
        };
    }
}
exports.DatapipelinePipelineDefinition = DatapipelinePipelineDefinition;
_l = JSII_RTTI_SYMBOL_1;
DatapipelinePipelineDefinition[_l] = { fqn: "@cdktf/provider-aws.datapipelinePipelineDefinition.DatapipelinePipelineDefinition", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DatapipelinePipelineDefinition.tfResourceType = "aws_datapipeline_pipeline_definition";
