// @cdktf/provider-aws/lib/data-aws-kinesis-stream/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsKinesisStream = exports.DataAwsKinesisStreamStreamModeDetailsList = exports.DataAwsKinesisStreamStreamModeDetailsOutputReference = exports.dataAwsKinesisStreamStreamModeDetailsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsKinesisStreamStreamModeDetailsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsKinesisStreamStreamModeDetailsToTerraform = dataAwsKinesisStreamStreamModeDetailsToTerraform;
class DataAwsKinesisStreamStreamModeDetailsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // stream_mode - computed: true, optional: false, required: false
    get streamMode() {
        return this.getStringAttribute('stream_mode');
    }
}
exports.DataAwsKinesisStreamStreamModeDetailsOutputReference = DataAwsKinesisStreamStreamModeDetailsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsKinesisStreamStreamModeDetailsOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsKinesisStream.DataAwsKinesisStreamStreamModeDetailsOutputReference", version: "10.0.12" };
class DataAwsKinesisStreamStreamModeDetailsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsKinesisStreamStreamModeDetailsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsKinesisStreamStreamModeDetailsList = DataAwsKinesisStreamStreamModeDetailsList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsKinesisStreamStreamModeDetailsList[_b] = { fqn: "@cdktf/provider-aws.dataAwsKinesisStream.DataAwsKinesisStreamStreamModeDetailsList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/kinesis_stream aws_kinesis_stream}
*/
class DataAwsKinesisStream extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/kinesis_stream aws_kinesis_stream} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsKinesisStreamConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_kinesis_stream',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // stream_mode_details - computed: true, optional: false, required: false
        this._streamModeDetails = new DataAwsKinesisStreamStreamModeDetailsList(this, "stream_mode_details", false);
        this._id = config.id;
        this._name = config.name;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // closed_shards - computed: true, optional: false, required: false
    get closedShards() {
        return cdktf.Fn.tolist(this.getListAttribute('closed_shards'));
    }
    // creation_timestamp - computed: true, optional: false, required: false
    get creationTimestamp() {
        return this.getNumberAttribute('creation_timestamp');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // open_shards - computed: true, optional: false, required: false
    get openShards() {
        return cdktf.Fn.tolist(this.getListAttribute('open_shards'));
    }
    // retention_period - computed: true, optional: false, required: false
    get retentionPeriod() {
        return this.getNumberAttribute('retention_period');
    }
    // shard_level_metrics - computed: true, optional: false, required: false
    get shardLevelMetrics() {
        return cdktf.Fn.tolist(this.getListAttribute('shard_level_metrics'));
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get streamModeDetails() {
        return this._streamModeDetails;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
}
exports.DataAwsKinesisStream = DataAwsKinesisStream;
_c = JSII_RTTI_SYMBOL_1;
DataAwsKinesisStream[_c] = { fqn: "@cdktf/provider-aws.dataAwsKinesisStream.DataAwsKinesisStream", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsKinesisStream.tfResourceType = "aws_kinesis_stream";
