// @cdktf/provider-aws/lib/data-aws-kms-secret/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsKmsSecret = exports.DataAwsKmsSecretSecretList = exports.DataAwsKmsSecretSecretOutputReference = exports.dataAwsKmsSecretSecretToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsKmsSecretSecretToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        context: cdktf.hashMapper(cdktf.stringToTerraform)(struct.context),
        grant_tokens: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.grantTokens),
        name: cdktf.stringToTerraform(struct.name),
        payload: cdktf.stringToTerraform(struct.payload),
    };
}
exports.dataAwsKmsSecretSecretToTerraform = dataAwsKmsSecretSecretToTerraform;
class DataAwsKmsSecretSecretOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._context !== undefined) {
            hasAnyValues = true;
            internalValueResult.context = this._context;
        }
        if (this._grantTokens !== undefined) {
            hasAnyValues = true;
            internalValueResult.grantTokens = this._grantTokens;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._payload !== undefined) {
            hasAnyValues = true;
            internalValueResult.payload = this._payload;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._context = undefined;
            this._grantTokens = undefined;
            this._name = undefined;
            this._payload = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._context = value.context;
            this._grantTokens = value.grantTokens;
            this._name = value.name;
            this._payload = value.payload;
        }
    }
    get context() {
        return this.getStringMapAttribute('context');
    }
    set context(value) {
        this._context = value;
    }
    resetContext() {
        this._context = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get contextInput() {
        return this._context;
    }
    get grantTokens() {
        return this.getListAttribute('grant_tokens');
    }
    set grantTokens(value) {
        this._grantTokens = value;
    }
    resetGrantTokens() {
        this._grantTokens = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get grantTokensInput() {
        return this._grantTokens;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get payload() {
        return this.getStringAttribute('payload');
    }
    set payload(value) {
        this._payload = value;
    }
    // Temporarily expose input value. Use with caution.
    get payloadInput() {
        return this._payload;
    }
}
exports.DataAwsKmsSecretSecretOutputReference = DataAwsKmsSecretSecretOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsKmsSecretSecretOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsKmsSecret.DataAwsKmsSecretSecretOutputReference", version: "10.0.12" };
class DataAwsKmsSecretSecretList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsKmsSecretSecretOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsKmsSecretSecretList = DataAwsKmsSecretSecretList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsKmsSecretSecretList[_b] = { fqn: "@cdktf/provider-aws.dataAwsKmsSecret.DataAwsKmsSecretSecretList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/kms_secret aws_kms_secret}
*/
class DataAwsKmsSecret extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/kms_secret aws_kms_secret} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsKmsSecretConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_kms_secret',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // secret - computed: false, optional: false, required: true
        this._secret = new DataAwsKmsSecretSecretList(this, "secret", true);
        this._id = config.id;
        this._secret.internalValue = config.secret;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get secret() {
        return this._secret;
    }
    putSecret(value) {
        this._secret.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get secretInput() {
        return this._secret.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            secret: cdktf.listMapper(dataAwsKmsSecretSecretToTerraform, true)(this._secret.internalValue),
        };
    }
}
exports.DataAwsKmsSecret = DataAwsKmsSecret;
_c = JSII_RTTI_SYMBOL_1;
DataAwsKmsSecret[_c] = { fqn: "@cdktf/provider-aws.dataAwsKmsSecret.DataAwsKmsSecret", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsKmsSecret.tfResourceType = "aws_kms_secret";
