// @cdktf/provider-aws/lib/kms-alias/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.KmsAlias = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/kms_alias aws_kms_alias}
*/
class KmsAlias extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/kms_alias aws_kms_alias} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options KmsAliasConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_kms_alias',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._name = config.name;
        this._namePrefix = config.namePrefix;
        this._targetKeyId = config.targetKeyId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get namePrefix() {
        return this.getStringAttribute('name_prefix');
    }
    set namePrefix(value) {
        this._namePrefix = value;
    }
    resetNamePrefix() {
        this._namePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get namePrefixInput() {
        return this._namePrefix;
    }
    // target_key_arn - computed: true, optional: false, required: false
    get targetKeyArn() {
        return this.getStringAttribute('target_key_arn');
    }
    get targetKeyId() {
        return this.getStringAttribute('target_key_id');
    }
    set targetKeyId(value) {
        this._targetKeyId = value;
    }
    // Temporarily expose input value. Use with caution.
    get targetKeyIdInput() {
        return this._targetKeyId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            name_prefix: cdktf.stringToTerraform(this._namePrefix),
            target_key_id: cdktf.stringToTerraform(this._targetKeyId),
        };
    }
}
exports.KmsAlias = KmsAlias;
_a = JSII_RTTI_SYMBOL_1;
KmsAlias[_a] = { fqn: "@cdktf/provider-aws.kmsAlias.KmsAlias", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
KmsAlias.tfResourceType = "aws_kms_alias";
