// @cdktf/provider-aws/lib/ses-identity-notification-topic/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SesIdentityNotificationTopic = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/ses_identity_notification_topic aws_ses_identity_notification_topic}
*/
class SesIdentityNotificationTopic extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ses_identity_notification_topic aws_ses_identity_notification_topic} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SesIdentityNotificationTopicConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ses_identity_notification_topic',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._identity = config.identity;
        this._includeOriginalHeaders = config.includeOriginalHeaders;
        this._notificationType = config.notificationType;
        this._topicArn = config.topicArn;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get identity() {
        return this.getStringAttribute('identity');
    }
    set identity(value) {
        this._identity = value;
    }
    // Temporarily expose input value. Use with caution.
    get identityInput() {
        return this._identity;
    }
    get includeOriginalHeaders() {
        return this.getBooleanAttribute('include_original_headers');
    }
    set includeOriginalHeaders(value) {
        this._includeOriginalHeaders = value;
    }
    resetIncludeOriginalHeaders() {
        this._includeOriginalHeaders = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includeOriginalHeadersInput() {
        return this._includeOriginalHeaders;
    }
    get notificationType() {
        return this.getStringAttribute('notification_type');
    }
    set notificationType(value) {
        this._notificationType = value;
    }
    // Temporarily expose input value. Use with caution.
    get notificationTypeInput() {
        return this._notificationType;
    }
    get topicArn() {
        return this.getStringAttribute('topic_arn');
    }
    set topicArn(value) {
        this._topicArn = value;
    }
    resetTopicArn() {
        this._topicArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get topicArnInput() {
        return this._topicArn;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            identity: cdktf.stringToTerraform(this._identity),
            include_original_headers: cdktf.booleanToTerraform(this._includeOriginalHeaders),
            notification_type: cdktf.stringToTerraform(this._notificationType),
            topic_arn: cdktf.stringToTerraform(this._topicArn),
        };
    }
}
exports.SesIdentityNotificationTopic = SesIdentityNotificationTopic;
_a = JSII_RTTI_SYMBOL_1;
SesIdentityNotificationTopic[_a] = { fqn: "@cdktf/provider-aws.sesIdentityNotificationTopic.SesIdentityNotificationTopic", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
SesIdentityNotificationTopic.tfResourceType = "aws_ses_identity_notification_topic";
