// @cdktf/provider-aws/lib/connect-quick-connect/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectQuickConnect = exports.ConnectQuickConnectQuickConnectConfigOutputReference = exports.connectQuickConnectQuickConnectConfigToTerraform = exports.ConnectQuickConnectQuickConnectConfigUserConfigList = exports.ConnectQuickConnectQuickConnectConfigUserConfigOutputReference = exports.connectQuickConnectQuickConnectConfigUserConfigToTerraform = exports.ConnectQuickConnectQuickConnectConfigQueueConfigList = exports.ConnectQuickConnectQuickConnectConfigQueueConfigOutputReference = exports.connectQuickConnectQuickConnectConfigQueueConfigToTerraform = exports.ConnectQuickConnectQuickConnectConfigPhoneConfigList = exports.ConnectQuickConnectQuickConnectConfigPhoneConfigOutputReference = exports.connectQuickConnectQuickConnectConfigPhoneConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function connectQuickConnectQuickConnectConfigPhoneConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        phone_number: cdktf.stringToTerraform(struct.phoneNumber),
    };
}
exports.connectQuickConnectQuickConnectConfigPhoneConfigToTerraform = connectQuickConnectQuickConnectConfigPhoneConfigToTerraform;
class ConnectQuickConnectQuickConnectConfigPhoneConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._phoneNumber !== undefined) {
            hasAnyValues = true;
            internalValueResult.phoneNumber = this._phoneNumber;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._phoneNumber = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._phoneNumber = value.phoneNumber;
        }
    }
    get phoneNumber() {
        return this.getStringAttribute('phone_number');
    }
    set phoneNumber(value) {
        this._phoneNumber = value;
    }
    // Temporarily expose input value. Use with caution.
    get phoneNumberInput() {
        return this._phoneNumber;
    }
}
exports.ConnectQuickConnectQuickConnectConfigPhoneConfigOutputReference = ConnectQuickConnectQuickConnectConfigPhoneConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ConnectQuickConnectQuickConnectConfigPhoneConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.connectQuickConnect.ConnectQuickConnectQuickConnectConfigPhoneConfigOutputReference", version: "10.0.12" };
class ConnectQuickConnectQuickConnectConfigPhoneConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ConnectQuickConnectQuickConnectConfigPhoneConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ConnectQuickConnectQuickConnectConfigPhoneConfigList = ConnectQuickConnectQuickConnectConfigPhoneConfigList;
_b = JSII_RTTI_SYMBOL_1;
ConnectQuickConnectQuickConnectConfigPhoneConfigList[_b] = { fqn: "@cdktf/provider-aws.connectQuickConnect.ConnectQuickConnectQuickConnectConfigPhoneConfigList", version: "10.0.12" };
function connectQuickConnectQuickConnectConfigQueueConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        contact_flow_id: cdktf.stringToTerraform(struct.contactFlowId),
        queue_id: cdktf.stringToTerraform(struct.queueId),
    };
}
exports.connectQuickConnectQuickConnectConfigQueueConfigToTerraform = connectQuickConnectQuickConnectConfigQueueConfigToTerraform;
class ConnectQuickConnectQuickConnectConfigQueueConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._contactFlowId !== undefined) {
            hasAnyValues = true;
            internalValueResult.contactFlowId = this._contactFlowId;
        }
        if (this._queueId !== undefined) {
            hasAnyValues = true;
            internalValueResult.queueId = this._queueId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._contactFlowId = undefined;
            this._queueId = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._contactFlowId = value.contactFlowId;
            this._queueId = value.queueId;
        }
    }
    get contactFlowId() {
        return this.getStringAttribute('contact_flow_id');
    }
    set contactFlowId(value) {
        this._contactFlowId = value;
    }
    // Temporarily expose input value. Use with caution.
    get contactFlowIdInput() {
        return this._contactFlowId;
    }
    get queueId() {
        return this.getStringAttribute('queue_id');
    }
    set queueId(value) {
        this._queueId = value;
    }
    // Temporarily expose input value. Use with caution.
    get queueIdInput() {
        return this._queueId;
    }
}
exports.ConnectQuickConnectQuickConnectConfigQueueConfigOutputReference = ConnectQuickConnectQuickConnectConfigQueueConfigOutputReference;
_c = JSII_RTTI_SYMBOL_1;
ConnectQuickConnectQuickConnectConfigQueueConfigOutputReference[_c] = { fqn: "@cdktf/provider-aws.connectQuickConnect.ConnectQuickConnectQuickConnectConfigQueueConfigOutputReference", version: "10.0.12" };
class ConnectQuickConnectQuickConnectConfigQueueConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ConnectQuickConnectQuickConnectConfigQueueConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ConnectQuickConnectQuickConnectConfigQueueConfigList = ConnectQuickConnectQuickConnectConfigQueueConfigList;
_d = JSII_RTTI_SYMBOL_1;
ConnectQuickConnectQuickConnectConfigQueueConfigList[_d] = { fqn: "@cdktf/provider-aws.connectQuickConnect.ConnectQuickConnectQuickConnectConfigQueueConfigList", version: "10.0.12" };
function connectQuickConnectQuickConnectConfigUserConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        contact_flow_id: cdktf.stringToTerraform(struct.contactFlowId),
        user_id: cdktf.stringToTerraform(struct.userId),
    };
}
exports.connectQuickConnectQuickConnectConfigUserConfigToTerraform = connectQuickConnectQuickConnectConfigUserConfigToTerraform;
class ConnectQuickConnectQuickConnectConfigUserConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._contactFlowId !== undefined) {
            hasAnyValues = true;
            internalValueResult.contactFlowId = this._contactFlowId;
        }
        if (this._userId !== undefined) {
            hasAnyValues = true;
            internalValueResult.userId = this._userId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._contactFlowId = undefined;
            this._userId = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._contactFlowId = value.contactFlowId;
            this._userId = value.userId;
        }
    }
    get contactFlowId() {
        return this.getStringAttribute('contact_flow_id');
    }
    set contactFlowId(value) {
        this._contactFlowId = value;
    }
    // Temporarily expose input value. Use with caution.
    get contactFlowIdInput() {
        return this._contactFlowId;
    }
    get userId() {
        return this.getStringAttribute('user_id');
    }
    set userId(value) {
        this._userId = value;
    }
    // Temporarily expose input value. Use with caution.
    get userIdInput() {
        return this._userId;
    }
}
exports.ConnectQuickConnectQuickConnectConfigUserConfigOutputReference = ConnectQuickConnectQuickConnectConfigUserConfigOutputReference;
_e = JSII_RTTI_SYMBOL_1;
ConnectQuickConnectQuickConnectConfigUserConfigOutputReference[_e] = { fqn: "@cdktf/provider-aws.connectQuickConnect.ConnectQuickConnectQuickConnectConfigUserConfigOutputReference", version: "10.0.12" };
class ConnectQuickConnectQuickConnectConfigUserConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ConnectQuickConnectQuickConnectConfigUserConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ConnectQuickConnectQuickConnectConfigUserConfigList = ConnectQuickConnectQuickConnectConfigUserConfigList;
_f = JSII_RTTI_SYMBOL_1;
ConnectQuickConnectQuickConnectConfigUserConfigList[_f] = { fqn: "@cdktf/provider-aws.connectQuickConnect.ConnectQuickConnectQuickConnectConfigUserConfigList", version: "10.0.12" };
function connectQuickConnectQuickConnectConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        quick_connect_type: cdktf.stringToTerraform(struct.quickConnectType),
        phone_config: cdktf.listMapper(connectQuickConnectQuickConnectConfigPhoneConfigToTerraform, true)(struct.phoneConfig),
        queue_config: cdktf.listMapper(connectQuickConnectQuickConnectConfigQueueConfigToTerraform, true)(struct.queueConfig),
        user_config: cdktf.listMapper(connectQuickConnectQuickConnectConfigUserConfigToTerraform, true)(struct.userConfig),
    };
}
exports.connectQuickConnectQuickConnectConfigToTerraform = connectQuickConnectQuickConnectConfigToTerraform;
class ConnectQuickConnectQuickConnectConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // phone_config - computed: false, optional: true, required: false
        this._phoneConfig = new ConnectQuickConnectQuickConnectConfigPhoneConfigList(this, "phone_config", false);
        // queue_config - computed: false, optional: true, required: false
        this._queueConfig = new ConnectQuickConnectQuickConnectConfigQueueConfigList(this, "queue_config", false);
        // user_config - computed: false, optional: true, required: false
        this._userConfig = new ConnectQuickConnectQuickConnectConfigUserConfigList(this, "user_config", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._quickConnectType !== undefined) {
            hasAnyValues = true;
            internalValueResult.quickConnectType = this._quickConnectType;
        }
        if (this._phoneConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.phoneConfig = this._phoneConfig?.internalValue;
        }
        if (this._queueConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.queueConfig = this._queueConfig?.internalValue;
        }
        if (this._userConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.userConfig = this._userConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._quickConnectType = undefined;
            this._phoneConfig.internalValue = undefined;
            this._queueConfig.internalValue = undefined;
            this._userConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._quickConnectType = value.quickConnectType;
            this._phoneConfig.internalValue = value.phoneConfig;
            this._queueConfig.internalValue = value.queueConfig;
            this._userConfig.internalValue = value.userConfig;
        }
    }
    get quickConnectType() {
        return this.getStringAttribute('quick_connect_type');
    }
    set quickConnectType(value) {
        this._quickConnectType = value;
    }
    // Temporarily expose input value. Use with caution.
    get quickConnectTypeInput() {
        return this._quickConnectType;
    }
    get phoneConfig() {
        return this._phoneConfig;
    }
    putPhoneConfig(value) {
        this._phoneConfig.internalValue = value;
    }
    resetPhoneConfig() {
        this._phoneConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get phoneConfigInput() {
        return this._phoneConfig.internalValue;
    }
    get queueConfig() {
        return this._queueConfig;
    }
    putQueueConfig(value) {
        this._queueConfig.internalValue = value;
    }
    resetQueueConfig() {
        this._queueConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get queueConfigInput() {
        return this._queueConfig.internalValue;
    }
    get userConfig() {
        return this._userConfig;
    }
    putUserConfig(value) {
        this._userConfig.internalValue = value;
    }
    resetUserConfig() {
        this._userConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get userConfigInput() {
        return this._userConfig.internalValue;
    }
}
exports.ConnectQuickConnectQuickConnectConfigOutputReference = ConnectQuickConnectQuickConnectConfigOutputReference;
_g = JSII_RTTI_SYMBOL_1;
ConnectQuickConnectQuickConnectConfigOutputReference[_g] = { fqn: "@cdktf/provider-aws.connectQuickConnect.ConnectQuickConnectQuickConnectConfigOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/connect_quick_connect aws_connect_quick_connect}
*/
class ConnectQuickConnect extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/connect_quick_connect aws_connect_quick_connect} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ConnectQuickConnectConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_connect_quick_connect',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // quick_connect_config - computed: false, optional: false, required: true
        this._quickConnectConfig = new ConnectQuickConnectQuickConnectConfigOutputReference(this, "quick_connect_config");
        this._description = config.description;
        this._id = config.id;
        this._instanceId = config.instanceId;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._quickConnectConfig.internalValue = config.quickConnectConfig;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instanceId() {
        return this.getStringAttribute('instance_id');
    }
    set instanceId(value) {
        this._instanceId = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceIdInput() {
        return this._instanceId;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // quick_connect_id - computed: true, optional: false, required: false
    get quickConnectId() {
        return this.getStringAttribute('quick_connect_id');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get quickConnectConfig() {
        return this._quickConnectConfig;
    }
    putQuickConnectConfig(value) {
        this._quickConnectConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get quickConnectConfigInput() {
        return this._quickConnectConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            instance_id: cdktf.stringToTerraform(this._instanceId),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            quick_connect_config: connectQuickConnectQuickConnectConfigToTerraform(this._quickConnectConfig.internalValue),
        };
    }
}
exports.ConnectQuickConnect = ConnectQuickConnect;
_h = JSII_RTTI_SYMBOL_1;
ConnectQuickConnect[_h] = { fqn: "@cdktf/provider-aws.connectQuickConnect.ConnectQuickConnect", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
ConnectQuickConnect.tfResourceType = "aws_connect_quick_connect";
