// @cdktf/provider-aws/lib/codepipeline-webhook/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodepipelineWebhook = exports.CodepipelineWebhookFilterList = exports.CodepipelineWebhookFilterOutputReference = exports.codepipelineWebhookFilterToTerraform = exports.CodepipelineWebhookAuthenticationConfigurationOutputReference = exports.codepipelineWebhookAuthenticationConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function codepipelineWebhookAuthenticationConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        allowed_ip_range: cdktf.stringToTerraform(struct.allowedIpRange),
        secret_token: cdktf.stringToTerraform(struct.secretToken),
    };
}
exports.codepipelineWebhookAuthenticationConfigurationToTerraform = codepipelineWebhookAuthenticationConfigurationToTerraform;
class CodepipelineWebhookAuthenticationConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._allowedIpRange !== undefined) {
            hasAnyValues = true;
            internalValueResult.allowedIpRange = this._allowedIpRange;
        }
        if (this._secretToken !== undefined) {
            hasAnyValues = true;
            internalValueResult.secretToken = this._secretToken;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._allowedIpRange = undefined;
            this._secretToken = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._allowedIpRange = value.allowedIpRange;
            this._secretToken = value.secretToken;
        }
    }
    get allowedIpRange() {
        return this.getStringAttribute('allowed_ip_range');
    }
    set allowedIpRange(value) {
        this._allowedIpRange = value;
    }
    resetAllowedIpRange() {
        this._allowedIpRange = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allowedIpRangeInput() {
        return this._allowedIpRange;
    }
    get secretToken() {
        return this.getStringAttribute('secret_token');
    }
    set secretToken(value) {
        this._secretToken = value;
    }
    resetSecretToken() {
        this._secretToken = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secretTokenInput() {
        return this._secretToken;
    }
}
exports.CodepipelineWebhookAuthenticationConfigurationOutputReference = CodepipelineWebhookAuthenticationConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CodepipelineWebhookAuthenticationConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.codepipelineWebhook.CodepipelineWebhookAuthenticationConfigurationOutputReference", version: "10.0.12" };
function codepipelineWebhookFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        json_path: cdktf.stringToTerraform(struct.jsonPath),
        match_equals: cdktf.stringToTerraform(struct.matchEquals),
    };
}
exports.codepipelineWebhookFilterToTerraform = codepipelineWebhookFilterToTerraform;
class CodepipelineWebhookFilterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._jsonPath !== undefined) {
            hasAnyValues = true;
            internalValueResult.jsonPath = this._jsonPath;
        }
        if (this._matchEquals !== undefined) {
            hasAnyValues = true;
            internalValueResult.matchEquals = this._matchEquals;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._jsonPath = undefined;
            this._matchEquals = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._jsonPath = value.jsonPath;
            this._matchEquals = value.matchEquals;
        }
    }
    get jsonPath() {
        return this.getStringAttribute('json_path');
    }
    set jsonPath(value) {
        this._jsonPath = value;
    }
    // Temporarily expose input value. Use with caution.
    get jsonPathInput() {
        return this._jsonPath;
    }
    get matchEquals() {
        return this.getStringAttribute('match_equals');
    }
    set matchEquals(value) {
        this._matchEquals = value;
    }
    // Temporarily expose input value. Use with caution.
    get matchEqualsInput() {
        return this._matchEquals;
    }
}
exports.CodepipelineWebhookFilterOutputReference = CodepipelineWebhookFilterOutputReference;
_b = JSII_RTTI_SYMBOL_1;
CodepipelineWebhookFilterOutputReference[_b] = { fqn: "@cdktf/provider-aws.codepipelineWebhook.CodepipelineWebhookFilterOutputReference", version: "10.0.12" };
class CodepipelineWebhookFilterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CodepipelineWebhookFilterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CodepipelineWebhookFilterList = CodepipelineWebhookFilterList;
_c = JSII_RTTI_SYMBOL_1;
CodepipelineWebhookFilterList[_c] = { fqn: "@cdktf/provider-aws.codepipelineWebhook.CodepipelineWebhookFilterList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/codepipeline_webhook aws_codepipeline_webhook}
*/
class CodepipelineWebhook extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/codepipeline_webhook aws_codepipeline_webhook} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CodepipelineWebhookConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_codepipeline_webhook',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // authentication_configuration - computed: false, optional: true, required: false
        this._authenticationConfiguration = new CodepipelineWebhookAuthenticationConfigurationOutputReference(this, "authentication_configuration");
        // filter - computed: false, optional: false, required: true
        this._filter = new CodepipelineWebhookFilterList(this, "filter", true);
        this._authentication = config.authentication;
        this._id = config.id;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._targetAction = config.targetAction;
        this._targetPipeline = config.targetPipeline;
        this._authenticationConfiguration.internalValue = config.authenticationConfiguration;
        this._filter.internalValue = config.filter;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get authentication() {
        return this.getStringAttribute('authentication');
    }
    set authentication(value) {
        this._authentication = value;
    }
    // Temporarily expose input value. Use with caution.
    get authenticationInput() {
        return this._authentication;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get targetAction() {
        return this.getStringAttribute('target_action');
    }
    set targetAction(value) {
        this._targetAction = value;
    }
    // Temporarily expose input value. Use with caution.
    get targetActionInput() {
        return this._targetAction;
    }
    get targetPipeline() {
        return this.getStringAttribute('target_pipeline');
    }
    set targetPipeline(value) {
        this._targetPipeline = value;
    }
    // Temporarily expose input value. Use with caution.
    get targetPipelineInput() {
        return this._targetPipeline;
    }
    // url - computed: true, optional: false, required: false
    get url() {
        return this.getStringAttribute('url');
    }
    get authenticationConfiguration() {
        return this._authenticationConfiguration;
    }
    putAuthenticationConfiguration(value) {
        this._authenticationConfiguration.internalValue = value;
    }
    resetAuthenticationConfiguration() {
        this._authenticationConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authenticationConfigurationInput() {
        return this._authenticationConfiguration.internalValue;
    }
    get filter() {
        return this._filter;
    }
    putFilter(value) {
        this._filter.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            authentication: cdktf.stringToTerraform(this._authentication),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            target_action: cdktf.stringToTerraform(this._targetAction),
            target_pipeline: cdktf.stringToTerraform(this._targetPipeline),
            authentication_configuration: codepipelineWebhookAuthenticationConfigurationToTerraform(this._authenticationConfiguration.internalValue),
            filter: cdktf.listMapper(codepipelineWebhookFilterToTerraform, true)(this._filter.internalValue),
        };
    }
}
exports.CodepipelineWebhook = CodepipelineWebhook;
_d = JSII_RTTI_SYMBOL_1;
CodepipelineWebhook[_d] = { fqn: "@cdktf/provider-aws.codepipelineWebhook.CodepipelineWebhook", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
CodepipelineWebhook.tfResourceType = "aws_codepipeline_webhook";
