// @cdktf/provider-aws/lib/apigatewayv2-authorizer/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Apigatewayv2Authorizer = exports.Apigatewayv2AuthorizerJwtConfigurationOutputReference = exports.apigatewayv2AuthorizerJwtConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function apigatewayv2AuthorizerJwtConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        audience: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.audience),
        issuer: cdktf.stringToTerraform(struct.issuer),
    };
}
exports.apigatewayv2AuthorizerJwtConfigurationToTerraform = apigatewayv2AuthorizerJwtConfigurationToTerraform;
class Apigatewayv2AuthorizerJwtConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._audience !== undefined) {
            hasAnyValues = true;
            internalValueResult.audience = this._audience;
        }
        if (this._issuer !== undefined) {
            hasAnyValues = true;
            internalValueResult.issuer = this._issuer;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._audience = undefined;
            this._issuer = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._audience = value.audience;
            this._issuer = value.issuer;
        }
    }
    get audience() {
        return cdktf.Fn.tolist(this.getListAttribute('audience'));
    }
    set audience(value) {
        this._audience = value;
    }
    resetAudience() {
        this._audience = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get audienceInput() {
        return this._audience;
    }
    get issuer() {
        return this.getStringAttribute('issuer');
    }
    set issuer(value) {
        this._issuer = value;
    }
    resetIssuer() {
        this._issuer = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get issuerInput() {
        return this._issuer;
    }
}
exports.Apigatewayv2AuthorizerJwtConfigurationOutputReference = Apigatewayv2AuthorizerJwtConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
Apigatewayv2AuthorizerJwtConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.apigatewayv2Authorizer.Apigatewayv2AuthorizerJwtConfigurationOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/apigatewayv2_authorizer aws_apigatewayv2_authorizer}
*/
class Apigatewayv2Authorizer extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/apigatewayv2_authorizer aws_apigatewayv2_authorizer} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options Apigatewayv2AuthorizerConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_apigatewayv2_authorizer',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // jwt_configuration - computed: false, optional: true, required: false
        this._jwtConfiguration = new Apigatewayv2AuthorizerJwtConfigurationOutputReference(this, "jwt_configuration");
        this._apiId = config.apiId;
        this._authorizerCredentialsArn = config.authorizerCredentialsArn;
        this._authorizerPayloadFormatVersion = config.authorizerPayloadFormatVersion;
        this._authorizerResultTtlInSeconds = config.authorizerResultTtlInSeconds;
        this._authorizerType = config.authorizerType;
        this._authorizerUri = config.authorizerUri;
        this._enableSimpleResponses = config.enableSimpleResponses;
        this._id = config.id;
        this._identitySources = config.identitySources;
        this._name = config.name;
        this._jwtConfiguration.internalValue = config.jwtConfiguration;
    }
    get apiId() {
        return this.getStringAttribute('api_id');
    }
    set apiId(value) {
        this._apiId = value;
    }
    // Temporarily expose input value. Use with caution.
    get apiIdInput() {
        return this._apiId;
    }
    get authorizerCredentialsArn() {
        return this.getStringAttribute('authorizer_credentials_arn');
    }
    set authorizerCredentialsArn(value) {
        this._authorizerCredentialsArn = value;
    }
    resetAuthorizerCredentialsArn() {
        this._authorizerCredentialsArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authorizerCredentialsArnInput() {
        return this._authorizerCredentialsArn;
    }
    get authorizerPayloadFormatVersion() {
        return this.getStringAttribute('authorizer_payload_format_version');
    }
    set authorizerPayloadFormatVersion(value) {
        this._authorizerPayloadFormatVersion = value;
    }
    resetAuthorizerPayloadFormatVersion() {
        this._authorizerPayloadFormatVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authorizerPayloadFormatVersionInput() {
        return this._authorizerPayloadFormatVersion;
    }
    get authorizerResultTtlInSeconds() {
        return this.getNumberAttribute('authorizer_result_ttl_in_seconds');
    }
    set authorizerResultTtlInSeconds(value) {
        this._authorizerResultTtlInSeconds = value;
    }
    resetAuthorizerResultTtlInSeconds() {
        this._authorizerResultTtlInSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authorizerResultTtlInSecondsInput() {
        return this._authorizerResultTtlInSeconds;
    }
    get authorizerType() {
        return this.getStringAttribute('authorizer_type');
    }
    set authorizerType(value) {
        this._authorizerType = value;
    }
    // Temporarily expose input value. Use with caution.
    get authorizerTypeInput() {
        return this._authorizerType;
    }
    get authorizerUri() {
        return this.getStringAttribute('authorizer_uri');
    }
    set authorizerUri(value) {
        this._authorizerUri = value;
    }
    resetAuthorizerUri() {
        this._authorizerUri = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authorizerUriInput() {
        return this._authorizerUri;
    }
    get enableSimpleResponses() {
        return this.getBooleanAttribute('enable_simple_responses');
    }
    set enableSimpleResponses(value) {
        this._enableSimpleResponses = value;
    }
    resetEnableSimpleResponses() {
        this._enableSimpleResponses = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableSimpleResponsesInput() {
        return this._enableSimpleResponses;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get identitySources() {
        return cdktf.Fn.tolist(this.getListAttribute('identity_sources'));
    }
    set identitySources(value) {
        this._identitySources = value;
    }
    resetIdentitySources() {
        this._identitySources = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get identitySourcesInput() {
        return this._identitySources;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get jwtConfiguration() {
        return this._jwtConfiguration;
    }
    putJwtConfiguration(value) {
        this._jwtConfiguration.internalValue = value;
    }
    resetJwtConfiguration() {
        this._jwtConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get jwtConfigurationInput() {
        return this._jwtConfiguration.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            api_id: cdktf.stringToTerraform(this._apiId),
            authorizer_credentials_arn: cdktf.stringToTerraform(this._authorizerCredentialsArn),
            authorizer_payload_format_version: cdktf.stringToTerraform(this._authorizerPayloadFormatVersion),
            authorizer_result_ttl_in_seconds: cdktf.numberToTerraform(this._authorizerResultTtlInSeconds),
            authorizer_type: cdktf.stringToTerraform(this._authorizerType),
            authorizer_uri: cdktf.stringToTerraform(this._authorizerUri),
            enable_simple_responses: cdktf.booleanToTerraform(this._enableSimpleResponses),
            id: cdktf.stringToTerraform(this._id),
            identity_sources: cdktf.listMapper(cdktf.stringToTerraform, false)(this._identitySources),
            name: cdktf.stringToTerraform(this._name),
            jwt_configuration: apigatewayv2AuthorizerJwtConfigurationToTerraform(this._jwtConfiguration.internalValue),
        };
    }
}
exports.Apigatewayv2Authorizer = Apigatewayv2Authorizer;
_b = JSII_RTTI_SYMBOL_1;
Apigatewayv2Authorizer[_b] = { fqn: "@cdktf/provider-aws.apigatewayv2Authorizer.Apigatewayv2Authorizer", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
Apigatewayv2Authorizer.tfResourceType = "aws_apigatewayv2_authorizer";
