// @cdktf/provider-aws/lib/route53-traffic-policy-instance/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Route53TrafficPolicyInstance = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/route53_traffic_policy_instance aws_route53_traffic_policy_instance}
*/
class Route53TrafficPolicyInstance extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/route53_traffic_policy_instance aws_route53_traffic_policy_instance} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options Route53TrafficPolicyInstanceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_route53_traffic_policy_instance',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._hostedZoneId = config.hostedZoneId;
        this._id = config.id;
        this._name = config.name;
        this._trafficPolicyId = config.trafficPolicyId;
        this._trafficPolicyVersion = config.trafficPolicyVersion;
        this._ttl = config.ttl;
    }
    get hostedZoneId() {
        return this.getStringAttribute('hosted_zone_id');
    }
    set hostedZoneId(value) {
        this._hostedZoneId = value;
    }
    // Temporarily expose input value. Use with caution.
    get hostedZoneIdInput() {
        return this._hostedZoneId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get trafficPolicyId() {
        return this.getStringAttribute('traffic_policy_id');
    }
    set trafficPolicyId(value) {
        this._trafficPolicyId = value;
    }
    // Temporarily expose input value. Use with caution.
    get trafficPolicyIdInput() {
        return this._trafficPolicyId;
    }
    get trafficPolicyVersion() {
        return this.getNumberAttribute('traffic_policy_version');
    }
    set trafficPolicyVersion(value) {
        this._trafficPolicyVersion = value;
    }
    // Temporarily expose input value. Use with caution.
    get trafficPolicyVersionInput() {
        return this._trafficPolicyVersion;
    }
    get ttl() {
        return this.getNumberAttribute('ttl');
    }
    set ttl(value) {
        this._ttl = value;
    }
    // Temporarily expose input value. Use with caution.
    get ttlInput() {
        return this._ttl;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            hosted_zone_id: cdktf.stringToTerraform(this._hostedZoneId),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            traffic_policy_id: cdktf.stringToTerraform(this._trafficPolicyId),
            traffic_policy_version: cdktf.numberToTerraform(this._trafficPolicyVersion),
            ttl: cdktf.numberToTerraform(this._ttl),
        };
    }
}
exports.Route53TrafficPolicyInstance = Route53TrafficPolicyInstance;
_a = JSII_RTTI_SYMBOL_1;
Route53TrafficPolicyInstance[_a] = { fqn: "@cdktf/provider-aws.route53TrafficPolicyInstance.Route53TrafficPolicyInstance", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
Route53TrafficPolicyInstance.tfResourceType = "aws_route53_traffic_policy_instance";
