// @cdktf/provider-aws/lib/glue-connection/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlueConnection = exports.GlueConnectionPhysicalConnectionRequirementsOutputReference = exports.glueConnectionPhysicalConnectionRequirementsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function glueConnectionPhysicalConnectionRequirementsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        availability_zone: cdktf.stringToTerraform(struct.availabilityZone),
        security_group_id_list: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.securityGroupIdList),
        subnet_id: cdktf.stringToTerraform(struct.subnetId),
    };
}
exports.glueConnectionPhysicalConnectionRequirementsToTerraform = glueConnectionPhysicalConnectionRequirementsToTerraform;
class GlueConnectionPhysicalConnectionRequirementsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._availabilityZone !== undefined) {
            hasAnyValues = true;
            internalValueResult.availabilityZone = this._availabilityZone;
        }
        if (this._securityGroupIdList !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroupIdList = this._securityGroupIdList;
        }
        if (this._subnetId !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetId = this._subnetId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._availabilityZone = undefined;
            this._securityGroupIdList = undefined;
            this._subnetId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._availabilityZone = value.availabilityZone;
            this._securityGroupIdList = value.securityGroupIdList;
            this._subnetId = value.subnetId;
        }
    }
    get availabilityZone() {
        return this.getStringAttribute('availability_zone');
    }
    set availabilityZone(value) {
        this._availabilityZone = value;
    }
    resetAvailabilityZone() {
        this._availabilityZone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get availabilityZoneInput() {
        return this._availabilityZone;
    }
    get securityGroupIdList() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_id_list'));
    }
    set securityGroupIdList(value) {
        this._securityGroupIdList = value;
    }
    resetSecurityGroupIdList() {
        this._securityGroupIdList = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdListInput() {
        return this._securityGroupIdList;
    }
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    set subnetId(value) {
        this._subnetId = value;
    }
    resetSubnetId() {
        this._subnetId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdInput() {
        return this._subnetId;
    }
}
exports.GlueConnectionPhysicalConnectionRequirementsOutputReference = GlueConnectionPhysicalConnectionRequirementsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
GlueConnectionPhysicalConnectionRequirementsOutputReference[_a] = { fqn: "@cdktf/provider-aws.glueConnection.GlueConnectionPhysicalConnectionRequirementsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/glue_connection aws_glue_connection}
*/
class GlueConnection extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/glue_connection aws_glue_connection} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options GlueConnectionConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_glue_connection',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // physical_connection_requirements - computed: false, optional: true, required: false
        this._physicalConnectionRequirements = new GlueConnectionPhysicalConnectionRequirementsOutputReference(this, "physical_connection_requirements");
        this._catalogId = config.catalogId;
        this._connectionProperties = config.connectionProperties;
        this._connectionType = config.connectionType;
        this._description = config.description;
        this._id = config.id;
        this._matchCriteria = config.matchCriteria;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._physicalConnectionRequirements.internalValue = config.physicalConnectionRequirements;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get catalogId() {
        return this.getStringAttribute('catalog_id');
    }
    set catalogId(value) {
        this._catalogId = value;
    }
    resetCatalogId() {
        this._catalogId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get catalogIdInput() {
        return this._catalogId;
    }
    get connectionProperties() {
        return this.getStringMapAttribute('connection_properties');
    }
    set connectionProperties(value) {
        this._connectionProperties = value;
    }
    resetConnectionProperties() {
        this._connectionProperties = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get connectionPropertiesInput() {
        return this._connectionProperties;
    }
    get connectionType() {
        return this.getStringAttribute('connection_type');
    }
    set connectionType(value) {
        this._connectionType = value;
    }
    resetConnectionType() {
        this._connectionType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get connectionTypeInput() {
        return this._connectionType;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get matchCriteria() {
        return this.getListAttribute('match_criteria');
    }
    set matchCriteria(value) {
        this._matchCriteria = value;
    }
    resetMatchCriteria() {
        this._matchCriteria = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get matchCriteriaInput() {
        return this._matchCriteria;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get physicalConnectionRequirements() {
        return this._physicalConnectionRequirements;
    }
    putPhysicalConnectionRequirements(value) {
        this._physicalConnectionRequirements.internalValue = value;
    }
    resetPhysicalConnectionRequirements() {
        this._physicalConnectionRequirements.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get physicalConnectionRequirementsInput() {
        return this._physicalConnectionRequirements.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            catalog_id: cdktf.stringToTerraform(this._catalogId),
            connection_properties: cdktf.hashMapper(cdktf.stringToTerraform)(this._connectionProperties),
            connection_type: cdktf.stringToTerraform(this._connectionType),
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            match_criteria: cdktf.listMapper(cdktf.stringToTerraform, false)(this._matchCriteria),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            physical_connection_requirements: glueConnectionPhysicalConnectionRequirementsToTerraform(this._physicalConnectionRequirements.internalValue),
        };
    }
}
exports.GlueConnection = GlueConnection;
_b = JSII_RTTI_SYMBOL_1;
GlueConnection[_b] = { fqn: "@cdktf/provider-aws.glueConnection.GlueConnection", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
GlueConnection.tfResourceType = "aws_glue_connection";
