// @cdktf/provider-aws/lib/data-aws-workspaces-directory/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsWorkspacesDirectory = exports.DataAwsWorkspacesDirectoryWorkspaceCreationPropertiesList = exports.DataAwsWorkspacesDirectoryWorkspaceCreationPropertiesOutputReference = exports.dataAwsWorkspacesDirectoryWorkspaceCreationPropertiesToTerraform = exports.DataAwsWorkspacesDirectoryWorkspaceAccessPropertiesList = exports.DataAwsWorkspacesDirectoryWorkspaceAccessPropertiesOutputReference = exports.dataAwsWorkspacesDirectoryWorkspaceAccessPropertiesToTerraform = exports.DataAwsWorkspacesDirectorySelfServicePermissionsList = exports.DataAwsWorkspacesDirectorySelfServicePermissionsOutputReference = exports.dataAwsWorkspacesDirectorySelfServicePermissionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsWorkspacesDirectorySelfServicePermissionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsWorkspacesDirectorySelfServicePermissionsToTerraform = dataAwsWorkspacesDirectorySelfServicePermissionsToTerraform;
class DataAwsWorkspacesDirectorySelfServicePermissionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // change_compute_type - computed: true, optional: false, required: false
    get changeComputeType() {
        return this.getBooleanAttribute('change_compute_type');
    }
    // increase_volume_size - computed: true, optional: false, required: false
    get increaseVolumeSize() {
        return this.getBooleanAttribute('increase_volume_size');
    }
    // rebuild_workspace - computed: true, optional: false, required: false
    get rebuildWorkspace() {
        return this.getBooleanAttribute('rebuild_workspace');
    }
    // restart_workspace - computed: true, optional: false, required: false
    get restartWorkspace() {
        return this.getBooleanAttribute('restart_workspace');
    }
    // switch_running_mode - computed: true, optional: false, required: false
    get switchRunningMode() {
        return this.getBooleanAttribute('switch_running_mode');
    }
}
exports.DataAwsWorkspacesDirectorySelfServicePermissionsOutputReference = DataAwsWorkspacesDirectorySelfServicePermissionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsWorkspacesDirectorySelfServicePermissionsOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsWorkspacesDirectory.DataAwsWorkspacesDirectorySelfServicePermissionsOutputReference", version: "10.0.12" };
class DataAwsWorkspacesDirectorySelfServicePermissionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsWorkspacesDirectorySelfServicePermissionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsWorkspacesDirectorySelfServicePermissionsList = DataAwsWorkspacesDirectorySelfServicePermissionsList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsWorkspacesDirectorySelfServicePermissionsList[_b] = { fqn: "@cdktf/provider-aws.dataAwsWorkspacesDirectory.DataAwsWorkspacesDirectorySelfServicePermissionsList", version: "10.0.12" };
function dataAwsWorkspacesDirectoryWorkspaceAccessPropertiesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsWorkspacesDirectoryWorkspaceAccessPropertiesToTerraform = dataAwsWorkspacesDirectoryWorkspaceAccessPropertiesToTerraform;
class DataAwsWorkspacesDirectoryWorkspaceAccessPropertiesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // device_type_android - computed: true, optional: false, required: false
    get deviceTypeAndroid() {
        return this.getStringAttribute('device_type_android');
    }
    // device_type_chromeos - computed: true, optional: false, required: false
    get deviceTypeChromeos() {
        return this.getStringAttribute('device_type_chromeos');
    }
    // device_type_ios - computed: true, optional: false, required: false
    get deviceTypeIos() {
        return this.getStringAttribute('device_type_ios');
    }
    // device_type_linux - computed: true, optional: false, required: false
    get deviceTypeLinux() {
        return this.getStringAttribute('device_type_linux');
    }
    // device_type_osx - computed: true, optional: false, required: false
    get deviceTypeOsx() {
        return this.getStringAttribute('device_type_osx');
    }
    // device_type_web - computed: true, optional: false, required: false
    get deviceTypeWeb() {
        return this.getStringAttribute('device_type_web');
    }
    // device_type_windows - computed: true, optional: false, required: false
    get deviceTypeWindows() {
        return this.getStringAttribute('device_type_windows');
    }
    // device_type_zeroclient - computed: true, optional: false, required: false
    get deviceTypeZeroclient() {
        return this.getStringAttribute('device_type_zeroclient');
    }
}
exports.DataAwsWorkspacesDirectoryWorkspaceAccessPropertiesOutputReference = DataAwsWorkspacesDirectoryWorkspaceAccessPropertiesOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsWorkspacesDirectoryWorkspaceAccessPropertiesOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsWorkspacesDirectory.DataAwsWorkspacesDirectoryWorkspaceAccessPropertiesOutputReference", version: "10.0.12" };
class DataAwsWorkspacesDirectoryWorkspaceAccessPropertiesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsWorkspacesDirectoryWorkspaceAccessPropertiesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsWorkspacesDirectoryWorkspaceAccessPropertiesList = DataAwsWorkspacesDirectoryWorkspaceAccessPropertiesList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsWorkspacesDirectoryWorkspaceAccessPropertiesList[_d] = { fqn: "@cdktf/provider-aws.dataAwsWorkspacesDirectory.DataAwsWorkspacesDirectoryWorkspaceAccessPropertiesList", version: "10.0.12" };
function dataAwsWorkspacesDirectoryWorkspaceCreationPropertiesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsWorkspacesDirectoryWorkspaceCreationPropertiesToTerraform = dataAwsWorkspacesDirectoryWorkspaceCreationPropertiesToTerraform;
class DataAwsWorkspacesDirectoryWorkspaceCreationPropertiesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // custom_security_group_id - computed: true, optional: false, required: false
    get customSecurityGroupId() {
        return this.getStringAttribute('custom_security_group_id');
    }
    // default_ou - computed: true, optional: false, required: false
    get defaultOu() {
        return this.getStringAttribute('default_ou');
    }
    // enable_internet_access - computed: true, optional: false, required: false
    get enableInternetAccess() {
        return this.getBooleanAttribute('enable_internet_access');
    }
    // enable_maintenance_mode - computed: true, optional: false, required: false
    get enableMaintenanceMode() {
        return this.getBooleanAttribute('enable_maintenance_mode');
    }
    // user_enabled_as_local_administrator - computed: true, optional: false, required: false
    get userEnabledAsLocalAdministrator() {
        return this.getBooleanAttribute('user_enabled_as_local_administrator');
    }
}
exports.DataAwsWorkspacesDirectoryWorkspaceCreationPropertiesOutputReference = DataAwsWorkspacesDirectoryWorkspaceCreationPropertiesOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsWorkspacesDirectoryWorkspaceCreationPropertiesOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsWorkspacesDirectory.DataAwsWorkspacesDirectoryWorkspaceCreationPropertiesOutputReference", version: "10.0.12" };
class DataAwsWorkspacesDirectoryWorkspaceCreationPropertiesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsWorkspacesDirectoryWorkspaceCreationPropertiesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsWorkspacesDirectoryWorkspaceCreationPropertiesList = DataAwsWorkspacesDirectoryWorkspaceCreationPropertiesList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsWorkspacesDirectoryWorkspaceCreationPropertiesList[_f] = { fqn: "@cdktf/provider-aws.dataAwsWorkspacesDirectory.DataAwsWorkspacesDirectoryWorkspaceCreationPropertiesList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/workspaces_directory aws_workspaces_directory}
*/
class DataAwsWorkspacesDirectory extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/workspaces_directory aws_workspaces_directory} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsWorkspacesDirectoryConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_workspaces_directory',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // self_service_permissions - computed: true, optional: false, required: false
        this._selfServicePermissions = new DataAwsWorkspacesDirectorySelfServicePermissionsList(this, "self_service_permissions", false);
        // workspace_access_properties - computed: true, optional: false, required: false
        this._workspaceAccessProperties = new DataAwsWorkspacesDirectoryWorkspaceAccessPropertiesList(this, "workspace_access_properties", false);
        // workspace_creation_properties - computed: true, optional: false, required: false
        this._workspaceCreationProperties = new DataAwsWorkspacesDirectoryWorkspaceCreationPropertiesList(this, "workspace_creation_properties", false);
        this._directoryId = config.directoryId;
        this._id = config.id;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // alias - computed: true, optional: false, required: false
    get alias() {
        return this.getStringAttribute('alias');
    }
    // customer_user_name - computed: true, optional: false, required: false
    get customerUserName() {
        return this.getStringAttribute('customer_user_name');
    }
    get directoryId() {
        return this.getStringAttribute('directory_id');
    }
    set directoryId(value) {
        this._directoryId = value;
    }
    // Temporarily expose input value. Use with caution.
    get directoryIdInput() {
        return this._directoryId;
    }
    // directory_name - computed: true, optional: false, required: false
    get directoryName() {
        return this.getStringAttribute('directory_name');
    }
    // directory_type - computed: true, optional: false, required: false
    get directoryType() {
        return this.getStringAttribute('directory_type');
    }
    // dns_ip_addresses - computed: true, optional: false, required: false
    get dnsIpAddresses() {
        return cdktf.Fn.tolist(this.getListAttribute('dns_ip_addresses'));
    }
    // iam_role_id - computed: true, optional: false, required: false
    get iamRoleId() {
        return this.getStringAttribute('iam_role_id');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // ip_group_ids - computed: true, optional: false, required: false
    get ipGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('ip_group_ids'));
    }
    // registration_code - computed: true, optional: false, required: false
    get registrationCode() {
        return this.getStringAttribute('registration_code');
    }
    get selfServicePermissions() {
        return this._selfServicePermissions;
    }
    // subnet_ids - computed: true, optional: false, required: false
    get subnetIds() {
        return cdktf.Fn.tolist(this.getListAttribute('subnet_ids'));
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get workspaceAccessProperties() {
        return this._workspaceAccessProperties;
    }
    get workspaceCreationProperties() {
        return this._workspaceCreationProperties;
    }
    // workspace_security_group_id - computed: true, optional: false, required: false
    get workspaceSecurityGroupId() {
        return this.getStringAttribute('workspace_security_group_id');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            directory_id: cdktf.stringToTerraform(this._directoryId),
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
}
exports.DataAwsWorkspacesDirectory = DataAwsWorkspacesDirectory;
_g = JSII_RTTI_SYMBOL_1;
DataAwsWorkspacesDirectory[_g] = { fqn: "@cdktf/provider-aws.dataAwsWorkspacesDirectory.DataAwsWorkspacesDirectory", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsWorkspacesDirectory.tfResourceType = "aws_workspaces_directory";
