// @cdktf/provider-aws/lib/cloudformation-stack-set-instance/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudformationStackSetInstance = exports.CloudformationStackSetInstanceTimeoutsOutputReference = exports.cloudformationStackSetInstanceTimeoutsToTerraform = exports.CloudformationStackSetInstanceOperationPreferencesOutputReference = exports.cloudformationStackSetInstanceOperationPreferencesToTerraform = exports.CloudformationStackSetInstanceDeploymentTargetsOutputReference = exports.cloudformationStackSetInstanceDeploymentTargetsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function cloudformationStackSetInstanceDeploymentTargetsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        organizational_unit_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.organizationalUnitIds),
    };
}
exports.cloudformationStackSetInstanceDeploymentTargetsToTerraform = cloudformationStackSetInstanceDeploymentTargetsToTerraform;
class CloudformationStackSetInstanceDeploymentTargetsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._organizationalUnitIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.organizationalUnitIds = this._organizationalUnitIds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._organizationalUnitIds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._organizationalUnitIds = value.organizationalUnitIds;
        }
    }
    get organizationalUnitIds() {
        return cdktf.Fn.tolist(this.getListAttribute('organizational_unit_ids'));
    }
    set organizationalUnitIds(value) {
        this._organizationalUnitIds = value;
    }
    resetOrganizationalUnitIds() {
        this._organizationalUnitIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get organizationalUnitIdsInput() {
        return this._organizationalUnitIds;
    }
}
exports.CloudformationStackSetInstanceDeploymentTargetsOutputReference = CloudformationStackSetInstanceDeploymentTargetsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CloudformationStackSetInstanceDeploymentTargetsOutputReference[_a] = { fqn: "@cdktf/provider-aws.cloudformationStackSetInstance.CloudformationStackSetInstanceDeploymentTargetsOutputReference", version: "10.0.12" };
function cloudformationStackSetInstanceOperationPreferencesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        failure_tolerance_count: cdktf.numberToTerraform(struct.failureToleranceCount),
        failure_tolerance_percentage: cdktf.numberToTerraform(struct.failureTolerancePercentage),
        max_concurrent_count: cdktf.numberToTerraform(struct.maxConcurrentCount),
        max_concurrent_percentage: cdktf.numberToTerraform(struct.maxConcurrentPercentage),
        region_concurrency_type: cdktf.stringToTerraform(struct.regionConcurrencyType),
        region_order: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.regionOrder),
    };
}
exports.cloudformationStackSetInstanceOperationPreferencesToTerraform = cloudformationStackSetInstanceOperationPreferencesToTerraform;
class CloudformationStackSetInstanceOperationPreferencesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._failureToleranceCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.failureToleranceCount = this._failureToleranceCount;
        }
        if (this._failureTolerancePercentage !== undefined) {
            hasAnyValues = true;
            internalValueResult.failureTolerancePercentage = this._failureTolerancePercentage;
        }
        if (this._maxConcurrentCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxConcurrentCount = this._maxConcurrentCount;
        }
        if (this._maxConcurrentPercentage !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxConcurrentPercentage = this._maxConcurrentPercentage;
        }
        if (this._regionConcurrencyType !== undefined) {
            hasAnyValues = true;
            internalValueResult.regionConcurrencyType = this._regionConcurrencyType;
        }
        if (this._regionOrder !== undefined) {
            hasAnyValues = true;
            internalValueResult.regionOrder = this._regionOrder;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._failureToleranceCount = undefined;
            this._failureTolerancePercentage = undefined;
            this._maxConcurrentCount = undefined;
            this._maxConcurrentPercentage = undefined;
            this._regionConcurrencyType = undefined;
            this._regionOrder = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._failureToleranceCount = value.failureToleranceCount;
            this._failureTolerancePercentage = value.failureTolerancePercentage;
            this._maxConcurrentCount = value.maxConcurrentCount;
            this._maxConcurrentPercentage = value.maxConcurrentPercentage;
            this._regionConcurrencyType = value.regionConcurrencyType;
            this._regionOrder = value.regionOrder;
        }
    }
    get failureToleranceCount() {
        return this.getNumberAttribute('failure_tolerance_count');
    }
    set failureToleranceCount(value) {
        this._failureToleranceCount = value;
    }
    resetFailureToleranceCount() {
        this._failureToleranceCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get failureToleranceCountInput() {
        return this._failureToleranceCount;
    }
    get failureTolerancePercentage() {
        return this.getNumberAttribute('failure_tolerance_percentage');
    }
    set failureTolerancePercentage(value) {
        this._failureTolerancePercentage = value;
    }
    resetFailureTolerancePercentage() {
        this._failureTolerancePercentage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get failureTolerancePercentageInput() {
        return this._failureTolerancePercentage;
    }
    get maxConcurrentCount() {
        return this.getNumberAttribute('max_concurrent_count');
    }
    set maxConcurrentCount(value) {
        this._maxConcurrentCount = value;
    }
    resetMaxConcurrentCount() {
        this._maxConcurrentCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxConcurrentCountInput() {
        return this._maxConcurrentCount;
    }
    get maxConcurrentPercentage() {
        return this.getNumberAttribute('max_concurrent_percentage');
    }
    set maxConcurrentPercentage(value) {
        this._maxConcurrentPercentage = value;
    }
    resetMaxConcurrentPercentage() {
        this._maxConcurrentPercentage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxConcurrentPercentageInput() {
        return this._maxConcurrentPercentage;
    }
    get regionConcurrencyType() {
        return this.getStringAttribute('region_concurrency_type');
    }
    set regionConcurrencyType(value) {
        this._regionConcurrencyType = value;
    }
    resetRegionConcurrencyType() {
        this._regionConcurrencyType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionConcurrencyTypeInput() {
        return this._regionConcurrencyType;
    }
    get regionOrder() {
        return this.getListAttribute('region_order');
    }
    set regionOrder(value) {
        this._regionOrder = value;
    }
    resetRegionOrder() {
        this._regionOrder = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionOrderInput() {
        return this._regionOrder;
    }
}
exports.CloudformationStackSetInstanceOperationPreferencesOutputReference = CloudformationStackSetInstanceOperationPreferencesOutputReference;
_b = JSII_RTTI_SYMBOL_1;
CloudformationStackSetInstanceOperationPreferencesOutputReference[_b] = { fqn: "@cdktf/provider-aws.cloudformationStackSetInstance.CloudformationStackSetInstanceOperationPreferencesOutputReference", version: "10.0.12" };
function cloudformationStackSetInstanceTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.cloudformationStackSetInstanceTimeoutsToTerraform = cloudformationStackSetInstanceTimeoutsToTerraform;
class CloudformationStackSetInstanceTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.CloudformationStackSetInstanceTimeoutsOutputReference = CloudformationStackSetInstanceTimeoutsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
CloudformationStackSetInstanceTimeoutsOutputReference[_c] = { fqn: "@cdktf/provider-aws.cloudformationStackSetInstance.CloudformationStackSetInstanceTimeoutsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/cloudformation_stack_set_instance aws_cloudformation_stack_set_instance}
*/
class CloudformationStackSetInstance extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/cloudformation_stack_set_instance aws_cloudformation_stack_set_instance} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CloudformationStackSetInstanceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cloudformation_stack_set_instance',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // deployment_targets - computed: false, optional: true, required: false
        this._deploymentTargets = new CloudformationStackSetInstanceDeploymentTargetsOutputReference(this, "deployment_targets");
        // operation_preferences - computed: false, optional: true, required: false
        this._operationPreferences = new CloudformationStackSetInstanceOperationPreferencesOutputReference(this, "operation_preferences");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new CloudformationStackSetInstanceTimeoutsOutputReference(this, "timeouts");
        this._accountId = config.accountId;
        this._callAs = config.callAs;
        this._id = config.id;
        this._parameterOverrides = config.parameterOverrides;
        this._region = config.region;
        this._retainStack = config.retainStack;
        this._stackSetName = config.stackSetName;
        this._deploymentTargets.internalValue = config.deploymentTargets;
        this._operationPreferences.internalValue = config.operationPreferences;
        this._timeouts.internalValue = config.timeouts;
    }
    get accountId() {
        return this.getStringAttribute('account_id');
    }
    set accountId(value) {
        this._accountId = value;
    }
    resetAccountId() {
        this._accountId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accountIdInput() {
        return this._accountId;
    }
    get callAs() {
        return this.getStringAttribute('call_as');
    }
    set callAs(value) {
        this._callAs = value;
    }
    resetCallAs() {
        this._callAs = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get callAsInput() {
        return this._callAs;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // organizational_unit_id - computed: true, optional: false, required: false
    get organizationalUnitId() {
        return this.getStringAttribute('organizational_unit_id');
    }
    get parameterOverrides() {
        return this.getStringMapAttribute('parameter_overrides');
    }
    set parameterOverrides(value) {
        this._parameterOverrides = value;
    }
    resetParameterOverrides() {
        this._parameterOverrides = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parameterOverridesInput() {
        return this._parameterOverrides;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    get retainStack() {
        return this.getBooleanAttribute('retain_stack');
    }
    set retainStack(value) {
        this._retainStack = value;
    }
    resetRetainStack() {
        this._retainStack = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get retainStackInput() {
        return this._retainStack;
    }
    // stack_id - computed: true, optional: false, required: false
    get stackId() {
        return this.getStringAttribute('stack_id');
    }
    get stackSetName() {
        return this.getStringAttribute('stack_set_name');
    }
    set stackSetName(value) {
        this._stackSetName = value;
    }
    // Temporarily expose input value. Use with caution.
    get stackSetNameInput() {
        return this._stackSetName;
    }
    get deploymentTargets() {
        return this._deploymentTargets;
    }
    putDeploymentTargets(value) {
        this._deploymentTargets.internalValue = value;
    }
    resetDeploymentTargets() {
        this._deploymentTargets.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deploymentTargetsInput() {
        return this._deploymentTargets.internalValue;
    }
    get operationPreferences() {
        return this._operationPreferences;
    }
    putOperationPreferences(value) {
        this._operationPreferences.internalValue = value;
    }
    resetOperationPreferences() {
        this._operationPreferences.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get operationPreferencesInput() {
        return this._operationPreferences.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            account_id: cdktf.stringToTerraform(this._accountId),
            call_as: cdktf.stringToTerraform(this._callAs),
            id: cdktf.stringToTerraform(this._id),
            parameter_overrides: cdktf.hashMapper(cdktf.stringToTerraform)(this._parameterOverrides),
            region: cdktf.stringToTerraform(this._region),
            retain_stack: cdktf.booleanToTerraform(this._retainStack),
            stack_set_name: cdktf.stringToTerraform(this._stackSetName),
            deployment_targets: cloudformationStackSetInstanceDeploymentTargetsToTerraform(this._deploymentTargets.internalValue),
            operation_preferences: cloudformationStackSetInstanceOperationPreferencesToTerraform(this._operationPreferences.internalValue),
            timeouts: cloudformationStackSetInstanceTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.CloudformationStackSetInstance = CloudformationStackSetInstance;
_d = JSII_RTTI_SYMBOL_1;
CloudformationStackSetInstance[_d] = { fqn: "@cdktf/provider-aws.cloudformationStackSetInstance.CloudformationStackSetInstance", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
CloudformationStackSetInstance.tfResourceType = "aws_cloudformation_stack_set_instance";
