// @cdktf/provider-aws/lib/data-aws-imagebuilder-image/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsImagebuilderImage = exports.DataAwsImagebuilderImageOutputResourcesList = exports.DataAwsImagebuilderImageOutputResourcesOutputReference = exports.dataAwsImagebuilderImageOutputResourcesToTerraform = exports.DataAwsImagebuilderImageOutputResourcesAmisList = exports.DataAwsImagebuilderImageOutputResourcesAmisOutputReference = exports.dataAwsImagebuilderImageOutputResourcesAmisToTerraform = exports.DataAwsImagebuilderImageImageTestsConfigurationList = exports.DataAwsImagebuilderImageImageTestsConfigurationOutputReference = exports.dataAwsImagebuilderImageImageTestsConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsImagebuilderImageImageTestsConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsImagebuilderImageImageTestsConfigurationToTerraform = dataAwsImagebuilderImageImageTestsConfigurationToTerraform;
class DataAwsImagebuilderImageImageTestsConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // image_tests_enabled - computed: true, optional: false, required: false
    get imageTestsEnabled() {
        return this.getBooleanAttribute('image_tests_enabled');
    }
    // timeout_minutes - computed: true, optional: false, required: false
    get timeoutMinutes() {
        return this.getNumberAttribute('timeout_minutes');
    }
}
exports.DataAwsImagebuilderImageImageTestsConfigurationOutputReference = DataAwsImagebuilderImageImageTestsConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderImageImageTestsConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsImagebuilderImage.DataAwsImagebuilderImageImageTestsConfigurationOutputReference", version: "10.0.12" };
class DataAwsImagebuilderImageImageTestsConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsImagebuilderImageImageTestsConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsImagebuilderImageImageTestsConfigurationList = DataAwsImagebuilderImageImageTestsConfigurationList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderImageImageTestsConfigurationList[_b] = { fqn: "@cdktf/provider-aws.dataAwsImagebuilderImage.DataAwsImagebuilderImageImageTestsConfigurationList", version: "10.0.12" };
function dataAwsImagebuilderImageOutputResourcesAmisToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsImagebuilderImageOutputResourcesAmisToTerraform = dataAwsImagebuilderImageOutputResourcesAmisToTerraform;
class DataAwsImagebuilderImageOutputResourcesAmisOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // account_id - computed: true, optional: false, required: false
    get accountId() {
        return this.getStringAttribute('account_id');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // image - computed: true, optional: false, required: false
    get image() {
        return this.getStringAttribute('image');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // region - computed: true, optional: false, required: false
    get region() {
        return this.getStringAttribute('region');
    }
}
exports.DataAwsImagebuilderImageOutputResourcesAmisOutputReference = DataAwsImagebuilderImageOutputResourcesAmisOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderImageOutputResourcesAmisOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsImagebuilderImage.DataAwsImagebuilderImageOutputResourcesAmisOutputReference", version: "10.0.12" };
class DataAwsImagebuilderImageOutputResourcesAmisList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsImagebuilderImageOutputResourcesAmisOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsImagebuilderImageOutputResourcesAmisList = DataAwsImagebuilderImageOutputResourcesAmisList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderImageOutputResourcesAmisList[_d] = { fqn: "@cdktf/provider-aws.dataAwsImagebuilderImage.DataAwsImagebuilderImageOutputResourcesAmisList", version: "10.0.12" };
function dataAwsImagebuilderImageOutputResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsImagebuilderImageOutputResourcesToTerraform = dataAwsImagebuilderImageOutputResourcesToTerraform;
class DataAwsImagebuilderImageOutputResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // amis - computed: true, optional: false, required: false
        this._amis = new DataAwsImagebuilderImageOutputResourcesAmisList(this, "amis", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get amis() {
        return this._amis;
    }
}
exports.DataAwsImagebuilderImageOutputResourcesOutputReference = DataAwsImagebuilderImageOutputResourcesOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderImageOutputResourcesOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsImagebuilderImage.DataAwsImagebuilderImageOutputResourcesOutputReference", version: "10.0.12" };
class DataAwsImagebuilderImageOutputResourcesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsImagebuilderImageOutputResourcesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsImagebuilderImageOutputResourcesList = DataAwsImagebuilderImageOutputResourcesList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderImageOutputResourcesList[_f] = { fqn: "@cdktf/provider-aws.dataAwsImagebuilderImage.DataAwsImagebuilderImageOutputResourcesList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/imagebuilder_image aws_imagebuilder_image}
*/
class DataAwsImagebuilderImage extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/imagebuilder_image aws_imagebuilder_image} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsImagebuilderImageConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_imagebuilder_image',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // image_tests_configuration - computed: true, optional: false, required: false
        this._imageTestsConfiguration = new DataAwsImagebuilderImageImageTestsConfigurationList(this, "image_tests_configuration", false);
        // output_resources - computed: true, optional: false, required: false
        this._outputResources = new DataAwsImagebuilderImageOutputResourcesList(this, "output_resources", false);
        this._arn = config.arn;
        this._id = config.id;
        this._tags = config.tags;
    }
    get arn() {
        return this.getStringAttribute('arn');
    }
    set arn(value) {
        this._arn = value;
    }
    // Temporarily expose input value. Use with caution.
    get arnInput() {
        return this._arn;
    }
    // build_version_arn - computed: true, optional: false, required: false
    get buildVersionArn() {
        return this.getStringAttribute('build_version_arn');
    }
    // container_recipe_arn - computed: true, optional: false, required: false
    get containerRecipeArn() {
        return this.getStringAttribute('container_recipe_arn');
    }
    // date_created - computed: true, optional: false, required: false
    get dateCreated() {
        return this.getStringAttribute('date_created');
    }
    // distribution_configuration_arn - computed: true, optional: false, required: false
    get distributionConfigurationArn() {
        return this.getStringAttribute('distribution_configuration_arn');
    }
    // enhanced_image_metadata_enabled - computed: true, optional: false, required: false
    get enhancedImageMetadataEnabled() {
        return this.getBooleanAttribute('enhanced_image_metadata_enabled');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // image_recipe_arn - computed: true, optional: false, required: false
    get imageRecipeArn() {
        return this.getStringAttribute('image_recipe_arn');
    }
    get imageTestsConfiguration() {
        return this._imageTestsConfiguration;
    }
    // infrastructure_configuration_arn - computed: true, optional: false, required: false
    get infrastructureConfigurationArn() {
        return this.getStringAttribute('infrastructure_configuration_arn');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // os_version - computed: true, optional: false, required: false
    get osVersion() {
        return this.getStringAttribute('os_version');
    }
    get outputResources() {
        return this._outputResources;
    }
    // platform - computed: true, optional: false, required: false
    get platform() {
        return this.getStringAttribute('platform');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            arn: cdktf.stringToTerraform(this._arn),
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
}
exports.DataAwsImagebuilderImage = DataAwsImagebuilderImage;
_g = JSII_RTTI_SYMBOL_1;
DataAwsImagebuilderImage[_g] = { fqn: "@cdktf/provider-aws.dataAwsImagebuilderImage.DataAwsImagebuilderImage", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsImagebuilderImage.tfResourceType = "aws_imagebuilder_image";
