// @cdktf/provider-aws/lib/data-aws-route53-resolver-firewall-rule-group-association/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsRoute53ResolverFirewallRuleGroupAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/route53_resolver_firewall_rule_group_association aws_route53_resolver_firewall_rule_group_association}
*/
class DataAwsRoute53ResolverFirewallRuleGroupAssociation extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/route53_resolver_firewall_rule_group_association aws_route53_resolver_firewall_rule_group_association} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsRoute53ResolverFirewallRuleGroupAssociationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_route53_resolver_firewall_rule_group_association',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._firewallRuleGroupAssociationId = config.firewallRuleGroupAssociationId;
        this._id = config.id;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // creation_time - computed: true, optional: false, required: false
    get creationTime() {
        return this.getStringAttribute('creation_time');
    }
    // creator_request_id - computed: true, optional: false, required: false
    get creatorRequestId() {
        return this.getStringAttribute('creator_request_id');
    }
    get firewallRuleGroupAssociationId() {
        return this.getStringAttribute('firewall_rule_group_association_id');
    }
    set firewallRuleGroupAssociationId(value) {
        this._firewallRuleGroupAssociationId = value;
    }
    // Temporarily expose input value. Use with caution.
    get firewallRuleGroupAssociationIdInput() {
        return this._firewallRuleGroupAssociationId;
    }
    // firewall_rule_group_id - computed: true, optional: false, required: false
    get firewallRuleGroupId() {
        return this.getStringAttribute('firewall_rule_group_id');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // managed_owner_name - computed: true, optional: false, required: false
    get managedOwnerName() {
        return this.getStringAttribute('managed_owner_name');
    }
    // modification_time - computed: true, optional: false, required: false
    get modificationTime() {
        return this.getStringAttribute('modification_time');
    }
    // mutation_protection - computed: true, optional: false, required: false
    get mutationProtection() {
        return this.getStringAttribute('mutation_protection');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // priority - computed: true, optional: false, required: false
    get priority() {
        return this.getNumberAttribute('priority');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    // status_message - computed: true, optional: false, required: false
    get statusMessage() {
        return this.getStringAttribute('status_message');
    }
    // vpc_id - computed: true, optional: false, required: false
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            firewall_rule_group_association_id: cdktf.stringToTerraform(this._firewallRuleGroupAssociationId),
            id: cdktf.stringToTerraform(this._id),
        };
    }
}
exports.DataAwsRoute53ResolverFirewallRuleGroupAssociation = DataAwsRoute53ResolverFirewallRuleGroupAssociation;
_a = JSII_RTTI_SYMBOL_1;
DataAwsRoute53ResolverFirewallRuleGroupAssociation[_a] = { fqn: "@cdktf/provider-aws.dataAwsRoute53ResolverFirewallRuleGroupAssociation.DataAwsRoute53ResolverFirewallRuleGroupAssociation", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsRoute53ResolverFirewallRuleGroupAssociation.tfResourceType = "aws_route53_resolver_firewall_rule_group_association";
