// @cdktf/provider-aws/lib/data-aws-eks-cluster/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsEksCluster = exports.DataAwsEksClusterVpcConfigList = exports.DataAwsEksClusterVpcConfigOutputReference = exports.dataAwsEksClusterVpcConfigToTerraform = exports.DataAwsEksClusterOutpostConfigList = exports.DataAwsEksClusterOutpostConfigOutputReference = exports.dataAwsEksClusterOutpostConfigToTerraform = exports.DataAwsEksClusterKubernetesNetworkConfigList = exports.DataAwsEksClusterKubernetesNetworkConfigOutputReference = exports.dataAwsEksClusterKubernetesNetworkConfigToTerraform = exports.DataAwsEksClusterIdentityList = exports.DataAwsEksClusterIdentityOutputReference = exports.dataAwsEksClusterIdentityToTerraform = exports.DataAwsEksClusterIdentityOidcList = exports.DataAwsEksClusterIdentityOidcOutputReference = exports.dataAwsEksClusterIdentityOidcToTerraform = exports.DataAwsEksClusterCertificateAuthorityList = exports.DataAwsEksClusterCertificateAuthorityOutputReference = exports.dataAwsEksClusterCertificateAuthorityToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsEksClusterCertificateAuthorityToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEksClusterCertificateAuthorityToTerraform = dataAwsEksClusterCertificateAuthorityToTerraform;
class DataAwsEksClusterCertificateAuthorityOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // data - computed: true, optional: false, required: false
    get data() {
        return this.getStringAttribute('data');
    }
}
exports.DataAwsEksClusterCertificateAuthorityOutputReference = DataAwsEksClusterCertificateAuthorityOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterCertificateAuthorityOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterCertificateAuthorityOutputReference", version: "10.0.12" };
class DataAwsEksClusterCertificateAuthorityList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEksClusterCertificateAuthorityOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEksClusterCertificateAuthorityList = DataAwsEksClusterCertificateAuthorityList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterCertificateAuthorityList[_b] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterCertificateAuthorityList", version: "10.0.12" };
function dataAwsEksClusterIdentityOidcToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEksClusterIdentityOidcToTerraform = dataAwsEksClusterIdentityOidcToTerraform;
class DataAwsEksClusterIdentityOidcOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // issuer - computed: true, optional: false, required: false
    get issuer() {
        return this.getStringAttribute('issuer');
    }
}
exports.DataAwsEksClusterIdentityOidcOutputReference = DataAwsEksClusterIdentityOidcOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterIdentityOidcOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterIdentityOidcOutputReference", version: "10.0.12" };
class DataAwsEksClusterIdentityOidcList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEksClusterIdentityOidcOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEksClusterIdentityOidcList = DataAwsEksClusterIdentityOidcList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterIdentityOidcList[_d] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterIdentityOidcList", version: "10.0.12" };
function dataAwsEksClusterIdentityToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEksClusterIdentityToTerraform = dataAwsEksClusterIdentityToTerraform;
class DataAwsEksClusterIdentityOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // oidc - computed: true, optional: false, required: false
        this._oidc = new DataAwsEksClusterIdentityOidcList(this, "oidc", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get oidc() {
        return this._oidc;
    }
}
exports.DataAwsEksClusterIdentityOutputReference = DataAwsEksClusterIdentityOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterIdentityOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterIdentityOutputReference", version: "10.0.12" };
class DataAwsEksClusterIdentityList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEksClusterIdentityOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEksClusterIdentityList = DataAwsEksClusterIdentityList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterIdentityList[_f] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterIdentityList", version: "10.0.12" };
function dataAwsEksClusterKubernetesNetworkConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEksClusterKubernetesNetworkConfigToTerraform = dataAwsEksClusterKubernetesNetworkConfigToTerraform;
class DataAwsEksClusterKubernetesNetworkConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ip_family - computed: true, optional: false, required: false
    get ipFamily() {
        return this.getStringAttribute('ip_family');
    }
    // service_ipv4_cidr - computed: true, optional: false, required: false
    get serviceIpv4Cidr() {
        return this.getStringAttribute('service_ipv4_cidr');
    }
    // service_ipv6_cidr - computed: true, optional: false, required: false
    get serviceIpv6Cidr() {
        return this.getStringAttribute('service_ipv6_cidr');
    }
}
exports.DataAwsEksClusterKubernetesNetworkConfigOutputReference = DataAwsEksClusterKubernetesNetworkConfigOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterKubernetesNetworkConfigOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterKubernetesNetworkConfigOutputReference", version: "10.0.12" };
class DataAwsEksClusterKubernetesNetworkConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEksClusterKubernetesNetworkConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEksClusterKubernetesNetworkConfigList = DataAwsEksClusterKubernetesNetworkConfigList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterKubernetesNetworkConfigList[_h] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterKubernetesNetworkConfigList", version: "10.0.12" };
function dataAwsEksClusterOutpostConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEksClusterOutpostConfigToTerraform = dataAwsEksClusterOutpostConfigToTerraform;
class DataAwsEksClusterOutpostConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // control_plane_instance_type - computed: true, optional: false, required: false
    get controlPlaneInstanceType() {
        return this.getStringAttribute('control_plane_instance_type');
    }
    // outpost_arns - computed: true, optional: false, required: false
    get outpostArns() {
        return cdktf.Fn.tolist(this.getListAttribute('outpost_arns'));
    }
}
exports.DataAwsEksClusterOutpostConfigOutputReference = DataAwsEksClusterOutpostConfigOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterOutpostConfigOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterOutpostConfigOutputReference", version: "10.0.12" };
class DataAwsEksClusterOutpostConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEksClusterOutpostConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEksClusterOutpostConfigList = DataAwsEksClusterOutpostConfigList;
_k = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterOutpostConfigList[_k] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterOutpostConfigList", version: "10.0.12" };
function dataAwsEksClusterVpcConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEksClusterVpcConfigToTerraform = dataAwsEksClusterVpcConfigToTerraform;
class DataAwsEksClusterVpcConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // cluster_security_group_id - computed: true, optional: false, required: false
    get clusterSecurityGroupId() {
        return this.getStringAttribute('cluster_security_group_id');
    }
    // endpoint_private_access - computed: true, optional: false, required: false
    get endpointPrivateAccess() {
        return this.getBooleanAttribute('endpoint_private_access');
    }
    // endpoint_public_access - computed: true, optional: false, required: false
    get endpointPublicAccess() {
        return this.getBooleanAttribute('endpoint_public_access');
    }
    // public_access_cidrs - computed: true, optional: false, required: false
    get publicAccessCidrs() {
        return cdktf.Fn.tolist(this.getListAttribute('public_access_cidrs'));
    }
    // security_group_ids - computed: true, optional: false, required: false
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    // subnet_ids - computed: true, optional: false, required: false
    get subnetIds() {
        return cdktf.Fn.tolist(this.getListAttribute('subnet_ids'));
    }
    // vpc_id - computed: true, optional: false, required: false
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
}
exports.DataAwsEksClusterVpcConfigOutputReference = DataAwsEksClusterVpcConfigOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterVpcConfigOutputReference[_l] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterVpcConfigOutputReference", version: "10.0.12" };
class DataAwsEksClusterVpcConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEksClusterVpcConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEksClusterVpcConfigList = DataAwsEksClusterVpcConfigList;
_m = JSII_RTTI_SYMBOL_1;
DataAwsEksClusterVpcConfigList[_m] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksClusterVpcConfigList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/eks_cluster aws_eks_cluster}
*/
class DataAwsEksCluster extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/eks_cluster aws_eks_cluster} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsEksClusterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_eks_cluster',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // certificate_authority - computed: true, optional: false, required: false
        this._certificateAuthority = new DataAwsEksClusterCertificateAuthorityList(this, "certificate_authority", false);
        // identity - computed: true, optional: false, required: false
        this._identity = new DataAwsEksClusterIdentityList(this, "identity", false);
        // kubernetes_network_config - computed: true, optional: false, required: false
        this._kubernetesNetworkConfig = new DataAwsEksClusterKubernetesNetworkConfigList(this, "kubernetes_network_config", false);
        // outpost_config - computed: true, optional: false, required: false
        this._outpostConfig = new DataAwsEksClusterOutpostConfigList(this, "outpost_config", false);
        // vpc_config - computed: true, optional: false, required: false
        this._vpcConfig = new DataAwsEksClusterVpcConfigList(this, "vpc_config", false);
        this._id = config.id;
        this._name = config.name;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get certificateAuthority() {
        return this._certificateAuthority;
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    // enabled_cluster_log_types - computed: true, optional: false, required: false
    get enabledClusterLogTypes() {
        return cdktf.Fn.tolist(this.getListAttribute('enabled_cluster_log_types'));
    }
    // endpoint - computed: true, optional: false, required: false
    get endpoint() {
        return this.getStringAttribute('endpoint');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get identity() {
        return this._identity;
    }
    get kubernetesNetworkConfig() {
        return this._kubernetesNetworkConfig;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get outpostConfig() {
        return this._outpostConfig;
    }
    // platform_version - computed: true, optional: false, required: false
    get platformVersion() {
        return this.getStringAttribute('platform_version');
    }
    // role_arn - computed: true, optional: false, required: false
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
    get vpcConfig() {
        return this._vpcConfig;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
}
exports.DataAwsEksCluster = DataAwsEksCluster;
_o = JSII_RTTI_SYMBOL_1;
DataAwsEksCluster[_o] = { fqn: "@cdktf/provider-aws.dataAwsEksCluster.DataAwsEksCluster", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsEksCluster.tfResourceType = "aws_eks_cluster";
