// @cdktf/provider-aws/lib/efs-replication-configuration/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EfsReplicationConfiguration = exports.EfsReplicationConfigurationTimeoutsOutputReference = exports.efsReplicationConfigurationTimeoutsToTerraform = exports.EfsReplicationConfigurationDestinationOutputReference = exports.efsReplicationConfigurationDestinationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function efsReplicationConfigurationDestinationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        availability_zone_name: cdktf.stringToTerraform(struct.availabilityZoneName),
        kms_key_id: cdktf.stringToTerraform(struct.kmsKeyId),
        region: cdktf.stringToTerraform(struct.region),
    };
}
exports.efsReplicationConfigurationDestinationToTerraform = efsReplicationConfigurationDestinationToTerraform;
class EfsReplicationConfigurationDestinationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._availabilityZoneName !== undefined) {
            hasAnyValues = true;
            internalValueResult.availabilityZoneName = this._availabilityZoneName;
        }
        if (this._kmsKeyId !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKeyId = this._kmsKeyId;
        }
        if (this._region !== undefined) {
            hasAnyValues = true;
            internalValueResult.region = this._region;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._availabilityZoneName = undefined;
            this._kmsKeyId = undefined;
            this._region = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._availabilityZoneName = value.availabilityZoneName;
            this._kmsKeyId = value.kmsKeyId;
            this._region = value.region;
        }
    }
    get availabilityZoneName() {
        return this.getStringAttribute('availability_zone_name');
    }
    set availabilityZoneName(value) {
        this._availabilityZoneName = value;
    }
    resetAvailabilityZoneName() {
        this._availabilityZoneName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get availabilityZoneNameInput() {
        return this._availabilityZoneName;
    }
    // file_system_id - computed: true, optional: false, required: false
    get fileSystemId() {
        return this.getStringAttribute('file_system_id');
    }
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    resetKmsKeyId() {
        this._kmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
}
exports.EfsReplicationConfigurationDestinationOutputReference = EfsReplicationConfigurationDestinationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
EfsReplicationConfigurationDestinationOutputReference[_a] = { fqn: "@cdktf/provider-aws.efsReplicationConfiguration.EfsReplicationConfigurationDestinationOutputReference", version: "10.0.12" };
function efsReplicationConfigurationTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.efsReplicationConfigurationTimeoutsToTerraform = efsReplicationConfigurationTimeoutsToTerraform;
class EfsReplicationConfigurationTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
}
exports.EfsReplicationConfigurationTimeoutsOutputReference = EfsReplicationConfigurationTimeoutsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
EfsReplicationConfigurationTimeoutsOutputReference[_b] = { fqn: "@cdktf/provider-aws.efsReplicationConfiguration.EfsReplicationConfigurationTimeoutsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/efs_replication_configuration aws_efs_replication_configuration}
*/
class EfsReplicationConfiguration extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/efs_replication_configuration aws_efs_replication_configuration} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options EfsReplicationConfigurationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_efs_replication_configuration',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // destination - computed: false, optional: false, required: true
        this._destination = new EfsReplicationConfigurationDestinationOutputReference(this, "destination");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new EfsReplicationConfigurationTimeoutsOutputReference(this, "timeouts");
        this._id = config.id;
        this._sourceFileSystemId = config.sourceFileSystemId;
        this._destination.internalValue = config.destination;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // creation_time - computed: true, optional: false, required: false
    get creationTime() {
        return this.getStringAttribute('creation_time');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // original_source_file_system_arn - computed: true, optional: false, required: false
    get originalSourceFileSystemArn() {
        return this.getStringAttribute('original_source_file_system_arn');
    }
    // source_file_system_arn - computed: true, optional: false, required: false
    get sourceFileSystemArn() {
        return this.getStringAttribute('source_file_system_arn');
    }
    get sourceFileSystemId() {
        return this.getStringAttribute('source_file_system_id');
    }
    set sourceFileSystemId(value) {
        this._sourceFileSystemId = value;
    }
    // Temporarily expose input value. Use with caution.
    get sourceFileSystemIdInput() {
        return this._sourceFileSystemId;
    }
    // source_file_system_region - computed: true, optional: false, required: false
    get sourceFileSystemRegion() {
        return this.getStringAttribute('source_file_system_region');
    }
    get destination() {
        return this._destination;
    }
    putDestination(value) {
        this._destination.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get destinationInput() {
        return this._destination.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            source_file_system_id: cdktf.stringToTerraform(this._sourceFileSystemId),
            destination: efsReplicationConfigurationDestinationToTerraform(this._destination.internalValue),
            timeouts: efsReplicationConfigurationTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.EfsReplicationConfiguration = EfsReplicationConfiguration;
_c = JSII_RTTI_SYMBOL_1;
EfsReplicationConfiguration[_c] = { fqn: "@cdktf/provider-aws.efsReplicationConfiguration.EfsReplicationConfiguration", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
EfsReplicationConfiguration.tfResourceType = "aws_efs_replication_configuration";
