// @cdktf/provider-aws/lib/fsx-openzfs-volume/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h;
Object.defineProperty(exports, "__esModule", { value: true });
exports.FsxOpenzfsVolume = exports.FsxOpenzfsVolumeUserAndGroupQuotasList = exports.FsxOpenzfsVolumeUserAndGroupQuotasOutputReference = exports.fsxOpenzfsVolumeUserAndGroupQuotasToTerraform = exports.FsxOpenzfsVolumeTimeoutsOutputReference = exports.fsxOpenzfsVolumeTimeoutsToTerraform = exports.FsxOpenzfsVolumeOriginSnapshotOutputReference = exports.fsxOpenzfsVolumeOriginSnapshotToTerraform = exports.FsxOpenzfsVolumeNfsExportsOutputReference = exports.fsxOpenzfsVolumeNfsExportsToTerraform = exports.FsxOpenzfsVolumeNfsExportsClientConfigurationsList = exports.FsxOpenzfsVolumeNfsExportsClientConfigurationsOutputReference = exports.fsxOpenzfsVolumeNfsExportsClientConfigurationsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function fsxOpenzfsVolumeNfsExportsClientConfigurationsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        clients: cdktf.stringToTerraform(struct.clients),
        options: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.options),
    };
}
exports.fsxOpenzfsVolumeNfsExportsClientConfigurationsToTerraform = fsxOpenzfsVolumeNfsExportsClientConfigurationsToTerraform;
class FsxOpenzfsVolumeNfsExportsClientConfigurationsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._clients !== undefined) {
            hasAnyValues = true;
            internalValueResult.clients = this._clients;
        }
        if (this._options !== undefined) {
            hasAnyValues = true;
            internalValueResult.options = this._options;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._clients = undefined;
            this._options = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._clients = value.clients;
            this._options = value.options;
        }
    }
    get clients() {
        return this.getStringAttribute('clients');
    }
    set clients(value) {
        this._clients = value;
    }
    // Temporarily expose input value. Use with caution.
    get clientsInput() {
        return this._clients;
    }
    get options() {
        return this.getListAttribute('options');
    }
    set options(value) {
        this._options = value;
    }
    // Temporarily expose input value. Use with caution.
    get optionsInput() {
        return this._options;
    }
}
exports.FsxOpenzfsVolumeNfsExportsClientConfigurationsOutputReference = FsxOpenzfsVolumeNfsExportsClientConfigurationsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
FsxOpenzfsVolumeNfsExportsClientConfigurationsOutputReference[_a] = { fqn: "@cdktf/provider-aws.fsxOpenzfsVolume.FsxOpenzfsVolumeNfsExportsClientConfigurationsOutputReference", version: "10.0.12" };
class FsxOpenzfsVolumeNfsExportsClientConfigurationsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new FsxOpenzfsVolumeNfsExportsClientConfigurationsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.FsxOpenzfsVolumeNfsExportsClientConfigurationsList = FsxOpenzfsVolumeNfsExportsClientConfigurationsList;
_b = JSII_RTTI_SYMBOL_1;
FsxOpenzfsVolumeNfsExportsClientConfigurationsList[_b] = { fqn: "@cdktf/provider-aws.fsxOpenzfsVolume.FsxOpenzfsVolumeNfsExportsClientConfigurationsList", version: "10.0.12" };
function fsxOpenzfsVolumeNfsExportsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        client_configurations: cdktf.listMapper(fsxOpenzfsVolumeNfsExportsClientConfigurationsToTerraform, true)(struct.clientConfigurations),
    };
}
exports.fsxOpenzfsVolumeNfsExportsToTerraform = fsxOpenzfsVolumeNfsExportsToTerraform;
class FsxOpenzfsVolumeNfsExportsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // client_configurations - computed: false, optional: false, required: true
        this._clientConfigurations = new FsxOpenzfsVolumeNfsExportsClientConfigurationsList(this, "client_configurations", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._clientConfigurations?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.clientConfigurations = this._clientConfigurations?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._clientConfigurations.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._clientConfigurations.internalValue = value.clientConfigurations;
        }
    }
    get clientConfigurations() {
        return this._clientConfigurations;
    }
    putClientConfigurations(value) {
        this._clientConfigurations.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get clientConfigurationsInput() {
        return this._clientConfigurations.internalValue;
    }
}
exports.FsxOpenzfsVolumeNfsExportsOutputReference = FsxOpenzfsVolumeNfsExportsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
FsxOpenzfsVolumeNfsExportsOutputReference[_c] = { fqn: "@cdktf/provider-aws.fsxOpenzfsVolume.FsxOpenzfsVolumeNfsExportsOutputReference", version: "10.0.12" };
function fsxOpenzfsVolumeOriginSnapshotToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        copy_strategy: cdktf.stringToTerraform(struct.copyStrategy),
        snapshot_arn: cdktf.stringToTerraform(struct.snapshotArn),
    };
}
exports.fsxOpenzfsVolumeOriginSnapshotToTerraform = fsxOpenzfsVolumeOriginSnapshotToTerraform;
class FsxOpenzfsVolumeOriginSnapshotOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._copyStrategy !== undefined) {
            hasAnyValues = true;
            internalValueResult.copyStrategy = this._copyStrategy;
        }
        if (this._snapshotArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.snapshotArn = this._snapshotArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._copyStrategy = undefined;
            this._snapshotArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._copyStrategy = value.copyStrategy;
            this._snapshotArn = value.snapshotArn;
        }
    }
    get copyStrategy() {
        return this.getStringAttribute('copy_strategy');
    }
    set copyStrategy(value) {
        this._copyStrategy = value;
    }
    // Temporarily expose input value. Use with caution.
    get copyStrategyInput() {
        return this._copyStrategy;
    }
    get snapshotArn() {
        return this.getStringAttribute('snapshot_arn');
    }
    set snapshotArn(value) {
        this._snapshotArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get snapshotArnInput() {
        return this._snapshotArn;
    }
}
exports.FsxOpenzfsVolumeOriginSnapshotOutputReference = FsxOpenzfsVolumeOriginSnapshotOutputReference;
_d = JSII_RTTI_SYMBOL_1;
FsxOpenzfsVolumeOriginSnapshotOutputReference[_d] = { fqn: "@cdktf/provider-aws.fsxOpenzfsVolume.FsxOpenzfsVolumeOriginSnapshotOutputReference", version: "10.0.12" };
function fsxOpenzfsVolumeTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.fsxOpenzfsVolumeTimeoutsToTerraform = fsxOpenzfsVolumeTimeoutsToTerraform;
class FsxOpenzfsVolumeTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.FsxOpenzfsVolumeTimeoutsOutputReference = FsxOpenzfsVolumeTimeoutsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
FsxOpenzfsVolumeTimeoutsOutputReference[_e] = { fqn: "@cdktf/provider-aws.fsxOpenzfsVolume.FsxOpenzfsVolumeTimeoutsOutputReference", version: "10.0.12" };
function fsxOpenzfsVolumeUserAndGroupQuotasToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        id: cdktf.numberToTerraform(struct.id),
        storage_capacity_quota_gib: cdktf.numberToTerraform(struct.storageCapacityQuotaGib),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.fsxOpenzfsVolumeUserAndGroupQuotasToTerraform = fsxOpenzfsVolumeUserAndGroupQuotasToTerraform;
class FsxOpenzfsVolumeUserAndGroupQuotasOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._id !== undefined) {
            hasAnyValues = true;
            internalValueResult.id = this._id;
        }
        if (this._storageCapacityQuotaGib !== undefined) {
            hasAnyValues = true;
            internalValueResult.storageCapacityQuotaGib = this._storageCapacityQuotaGib;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._id = undefined;
            this._storageCapacityQuotaGib = undefined;
            this._type = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._id = value.id;
            this._storageCapacityQuotaGib = value.storageCapacityQuotaGib;
            this._type = value.type;
        }
    }
    get id() {
        return this.getNumberAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get storageCapacityQuotaGib() {
        return this.getNumberAttribute('storage_capacity_quota_gib');
    }
    set storageCapacityQuotaGib(value) {
        this._storageCapacityQuotaGib = value;
    }
    // Temporarily expose input value. Use with caution.
    get storageCapacityQuotaGibInput() {
        return this._storageCapacityQuotaGib;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.FsxOpenzfsVolumeUserAndGroupQuotasOutputReference = FsxOpenzfsVolumeUserAndGroupQuotasOutputReference;
_f = JSII_RTTI_SYMBOL_1;
FsxOpenzfsVolumeUserAndGroupQuotasOutputReference[_f] = { fqn: "@cdktf/provider-aws.fsxOpenzfsVolume.FsxOpenzfsVolumeUserAndGroupQuotasOutputReference", version: "10.0.12" };
class FsxOpenzfsVolumeUserAndGroupQuotasList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new FsxOpenzfsVolumeUserAndGroupQuotasOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.FsxOpenzfsVolumeUserAndGroupQuotasList = FsxOpenzfsVolumeUserAndGroupQuotasList;
_g = JSII_RTTI_SYMBOL_1;
FsxOpenzfsVolumeUserAndGroupQuotasList[_g] = { fqn: "@cdktf/provider-aws.fsxOpenzfsVolume.FsxOpenzfsVolumeUserAndGroupQuotasList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/fsx_openzfs_volume aws_fsx_openzfs_volume}
*/
class FsxOpenzfsVolume extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/fsx_openzfs_volume aws_fsx_openzfs_volume} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options FsxOpenzfsVolumeConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_fsx_openzfs_volume',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // nfs_exports - computed: false, optional: true, required: false
        this._nfsExports = new FsxOpenzfsVolumeNfsExportsOutputReference(this, "nfs_exports");
        // origin_snapshot - computed: false, optional: true, required: false
        this._originSnapshot = new FsxOpenzfsVolumeOriginSnapshotOutputReference(this, "origin_snapshot");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new FsxOpenzfsVolumeTimeoutsOutputReference(this, "timeouts");
        // user_and_group_quotas - computed: false, optional: true, required: false
        this._userAndGroupQuotas = new FsxOpenzfsVolumeUserAndGroupQuotasList(this, "user_and_group_quotas", true);
        this._copyTagsToSnapshots = config.copyTagsToSnapshots;
        this._dataCompressionType = config.dataCompressionType;
        this._id = config.id;
        this._name = config.name;
        this._parentVolumeId = config.parentVolumeId;
        this._readOnly = config.readOnly;
        this._recordSizeKib = config.recordSizeKib;
        this._storageCapacityQuotaGib = config.storageCapacityQuotaGib;
        this._storageCapacityReservationGib = config.storageCapacityReservationGib;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._volumeType = config.volumeType;
        this._nfsExports.internalValue = config.nfsExports;
        this._originSnapshot.internalValue = config.originSnapshot;
        this._timeouts.internalValue = config.timeouts;
        this._userAndGroupQuotas.internalValue = config.userAndGroupQuotas;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get copyTagsToSnapshots() {
        return this.getBooleanAttribute('copy_tags_to_snapshots');
    }
    set copyTagsToSnapshots(value) {
        this._copyTagsToSnapshots = value;
    }
    resetCopyTagsToSnapshots() {
        this._copyTagsToSnapshots = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get copyTagsToSnapshotsInput() {
        return this._copyTagsToSnapshots;
    }
    get dataCompressionType() {
        return this.getStringAttribute('data_compression_type');
    }
    set dataCompressionType(value) {
        this._dataCompressionType = value;
    }
    resetDataCompressionType() {
        this._dataCompressionType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataCompressionTypeInput() {
        return this._dataCompressionType;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get parentVolumeId() {
        return this.getStringAttribute('parent_volume_id');
    }
    set parentVolumeId(value) {
        this._parentVolumeId = value;
    }
    // Temporarily expose input value. Use with caution.
    get parentVolumeIdInput() {
        return this._parentVolumeId;
    }
    get readOnly() {
        return this.getBooleanAttribute('read_only');
    }
    set readOnly(value) {
        this._readOnly = value;
    }
    resetReadOnly() {
        this._readOnly = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get readOnlyInput() {
        return this._readOnly;
    }
    get recordSizeKib() {
        return this.getNumberAttribute('record_size_kib');
    }
    set recordSizeKib(value) {
        this._recordSizeKib = value;
    }
    resetRecordSizeKib() {
        this._recordSizeKib = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get recordSizeKibInput() {
        return this._recordSizeKib;
    }
    get storageCapacityQuotaGib() {
        return this.getNumberAttribute('storage_capacity_quota_gib');
    }
    set storageCapacityQuotaGib(value) {
        this._storageCapacityQuotaGib = value;
    }
    resetStorageCapacityQuotaGib() {
        this._storageCapacityQuotaGib = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get storageCapacityQuotaGibInput() {
        return this._storageCapacityQuotaGib;
    }
    get storageCapacityReservationGib() {
        return this.getNumberAttribute('storage_capacity_reservation_gib');
    }
    set storageCapacityReservationGib(value) {
        this._storageCapacityReservationGib = value;
    }
    resetStorageCapacityReservationGib() {
        this._storageCapacityReservationGib = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get storageCapacityReservationGibInput() {
        return this._storageCapacityReservationGib;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get volumeType() {
        return this.getStringAttribute('volume_type');
    }
    set volumeType(value) {
        this._volumeType = value;
    }
    resetVolumeType() {
        this._volumeType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumeTypeInput() {
        return this._volumeType;
    }
    get nfsExports() {
        return this._nfsExports;
    }
    putNfsExports(value) {
        this._nfsExports.internalValue = value;
    }
    resetNfsExports() {
        this._nfsExports.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nfsExportsInput() {
        return this._nfsExports.internalValue;
    }
    get originSnapshot() {
        return this._originSnapshot;
    }
    putOriginSnapshot(value) {
        this._originSnapshot.internalValue = value;
    }
    resetOriginSnapshot() {
        this._originSnapshot.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get originSnapshotInput() {
        return this._originSnapshot.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get userAndGroupQuotas() {
        return this._userAndGroupQuotas;
    }
    putUserAndGroupQuotas(value) {
        this._userAndGroupQuotas.internalValue = value;
    }
    resetUserAndGroupQuotas() {
        this._userAndGroupQuotas.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get userAndGroupQuotasInput() {
        return this._userAndGroupQuotas.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            copy_tags_to_snapshots: cdktf.booleanToTerraform(this._copyTagsToSnapshots),
            data_compression_type: cdktf.stringToTerraform(this._dataCompressionType),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            parent_volume_id: cdktf.stringToTerraform(this._parentVolumeId),
            read_only: cdktf.booleanToTerraform(this._readOnly),
            record_size_kib: cdktf.numberToTerraform(this._recordSizeKib),
            storage_capacity_quota_gib: cdktf.numberToTerraform(this._storageCapacityQuotaGib),
            storage_capacity_reservation_gib: cdktf.numberToTerraform(this._storageCapacityReservationGib),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            volume_type: cdktf.stringToTerraform(this._volumeType),
            nfs_exports: fsxOpenzfsVolumeNfsExportsToTerraform(this._nfsExports.internalValue),
            origin_snapshot: fsxOpenzfsVolumeOriginSnapshotToTerraform(this._originSnapshot.internalValue),
            timeouts: fsxOpenzfsVolumeTimeoutsToTerraform(this._timeouts.internalValue),
            user_and_group_quotas: cdktf.listMapper(fsxOpenzfsVolumeUserAndGroupQuotasToTerraform, true)(this._userAndGroupQuotas.internalValue),
        };
    }
}
exports.FsxOpenzfsVolume = FsxOpenzfsVolume;
_h = JSII_RTTI_SYMBOL_1;
FsxOpenzfsVolume[_h] = { fqn: "@cdktf/provider-aws.fsxOpenzfsVolume.FsxOpenzfsVolume", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
FsxOpenzfsVolume.tfResourceType = "aws_fsx_openzfs_volume";
