// @cdktf/provider-aws/lib/iot-thing-group-membership/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IotThingGroupMembership = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/iot_thing_group_membership aws_iot_thing_group_membership}
*/
class IotThingGroupMembership extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/iot_thing_group_membership aws_iot_thing_group_membership} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options IotThingGroupMembershipConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iot_thing_group_membership',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._overrideDynamicGroup = config.overrideDynamicGroup;
        this._thingGroupName = config.thingGroupName;
        this._thingName = config.thingName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get overrideDynamicGroup() {
        return this.getBooleanAttribute('override_dynamic_group');
    }
    set overrideDynamicGroup(value) {
        this._overrideDynamicGroup = value;
    }
    resetOverrideDynamicGroup() {
        this._overrideDynamicGroup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get overrideDynamicGroupInput() {
        return this._overrideDynamicGroup;
    }
    get thingGroupName() {
        return this.getStringAttribute('thing_group_name');
    }
    set thingGroupName(value) {
        this._thingGroupName = value;
    }
    // Temporarily expose input value. Use with caution.
    get thingGroupNameInput() {
        return this._thingGroupName;
    }
    get thingName() {
        return this.getStringAttribute('thing_name');
    }
    set thingName(value) {
        this._thingName = value;
    }
    // Temporarily expose input value. Use with caution.
    get thingNameInput() {
        return this._thingName;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            override_dynamic_group: cdktf.booleanToTerraform(this._overrideDynamicGroup),
            thing_group_name: cdktf.stringToTerraform(this._thingGroupName),
            thing_name: cdktf.stringToTerraform(this._thingName),
        };
    }
}
exports.IotThingGroupMembership = IotThingGroupMembership;
_a = JSII_RTTI_SYMBOL_1;
IotThingGroupMembership[_a] = { fqn: "@cdktf/provider-aws.iotThingGroupMembership.IotThingGroupMembership", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
IotThingGroupMembership.tfResourceType = "aws_iot_thing_group_membership";
