// @cdktf/provider-aws/lib/data-aws-organizations-delegated-services/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsOrganizationsDelegatedServices = exports.DataAwsOrganizationsDelegatedServicesDelegatedServicesList = exports.DataAwsOrganizationsDelegatedServicesDelegatedServicesOutputReference = exports.dataAwsOrganizationsDelegatedServicesDelegatedServicesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsOrganizationsDelegatedServicesDelegatedServicesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsOrganizationsDelegatedServicesDelegatedServicesToTerraform = dataAwsOrganizationsDelegatedServicesDelegatedServicesToTerraform;
class DataAwsOrganizationsDelegatedServicesDelegatedServicesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // delegation_enabled_date - computed: true, optional: false, required: false
    get delegationEnabledDate() {
        return this.getStringAttribute('delegation_enabled_date');
    }
    // service_principal - computed: true, optional: false, required: false
    get servicePrincipal() {
        return this.getStringAttribute('service_principal');
    }
}
exports.DataAwsOrganizationsDelegatedServicesDelegatedServicesOutputReference = DataAwsOrganizationsDelegatedServicesDelegatedServicesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsOrganizationsDelegatedServicesDelegatedServicesOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsOrganizationsDelegatedServices.DataAwsOrganizationsDelegatedServicesDelegatedServicesOutputReference", version: "10.0.12" };
class DataAwsOrganizationsDelegatedServicesDelegatedServicesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsOrganizationsDelegatedServicesDelegatedServicesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsOrganizationsDelegatedServicesDelegatedServicesList = DataAwsOrganizationsDelegatedServicesDelegatedServicesList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsOrganizationsDelegatedServicesDelegatedServicesList[_b] = { fqn: "@cdktf/provider-aws.dataAwsOrganizationsDelegatedServices.DataAwsOrganizationsDelegatedServicesDelegatedServicesList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/organizations_delegated_services aws_organizations_delegated_services}
*/
class DataAwsOrganizationsDelegatedServices extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/organizations_delegated_services aws_organizations_delegated_services} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsOrganizationsDelegatedServicesConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_organizations_delegated_services',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // delegated_services - computed: true, optional: false, required: false
        this._delegatedServices = new DataAwsOrganizationsDelegatedServicesDelegatedServicesList(this, "delegated_services", true);
        this._accountId = config.accountId;
        this._id = config.id;
    }
    get accountId() {
        return this.getStringAttribute('account_id');
    }
    set accountId(value) {
        this._accountId = value;
    }
    // Temporarily expose input value. Use with caution.
    get accountIdInput() {
        return this._accountId;
    }
    get delegatedServices() {
        return this._delegatedServices;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            account_id: cdktf.stringToTerraform(this._accountId),
            id: cdktf.stringToTerraform(this._id),
        };
    }
}
exports.DataAwsOrganizationsDelegatedServices = DataAwsOrganizationsDelegatedServices;
_c = JSII_RTTI_SYMBOL_1;
DataAwsOrganizationsDelegatedServices[_c] = { fqn: "@cdktf/provider-aws.dataAwsOrganizationsDelegatedServices.DataAwsOrganizationsDelegatedServices", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsOrganizationsDelegatedServices.tfResourceType = "aws_organizations_delegated_services";
