// @cdktf/provider-aws/lib/ivs-recording-configuration/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IvsRecordingConfiguration = exports.IvsRecordingConfigurationTimeoutsOutputReference = exports.ivsRecordingConfigurationTimeoutsToTerraform = exports.IvsRecordingConfigurationThumbnailConfigurationOutputReference = exports.ivsRecordingConfigurationThumbnailConfigurationToTerraform = exports.IvsRecordingConfigurationDestinationConfigurationOutputReference = exports.ivsRecordingConfigurationDestinationConfigurationToTerraform = exports.IvsRecordingConfigurationDestinationConfigurationS3OutputReference = exports.ivsRecordingConfigurationDestinationConfigurationS3ToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function ivsRecordingConfigurationDestinationConfigurationS3ToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket_name: cdktf.stringToTerraform(struct.bucketName),
    };
}
exports.ivsRecordingConfigurationDestinationConfigurationS3ToTerraform = ivsRecordingConfigurationDestinationConfigurationS3ToTerraform;
class IvsRecordingConfigurationDestinationConfigurationS3OutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucketName !== undefined) {
            hasAnyValues = true;
            internalValueResult.bucketName = this._bucketName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucketName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucketName = value.bucketName;
        }
    }
    get bucketName() {
        return this.getStringAttribute('bucket_name');
    }
    set bucketName(value) {
        this._bucketName = value;
    }
    // Temporarily expose input value. Use with caution.
    get bucketNameInput() {
        return this._bucketName;
    }
}
exports.IvsRecordingConfigurationDestinationConfigurationS3OutputReference = IvsRecordingConfigurationDestinationConfigurationS3OutputReference;
_a = JSII_RTTI_SYMBOL_1;
IvsRecordingConfigurationDestinationConfigurationS3OutputReference[_a] = { fqn: "@cdktf/provider-aws.ivsRecordingConfiguration.IvsRecordingConfigurationDestinationConfigurationS3OutputReference", version: "10.0.12" };
function ivsRecordingConfigurationDestinationConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        s3: ivsRecordingConfigurationDestinationConfigurationS3ToTerraform(struct.s3),
    };
}
exports.ivsRecordingConfigurationDestinationConfigurationToTerraform = ivsRecordingConfigurationDestinationConfigurationToTerraform;
class IvsRecordingConfigurationDestinationConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // s3 - computed: false, optional: false, required: true
        this._s3 = new IvsRecordingConfigurationDestinationConfigurationS3OutputReference(this, "s3");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._s3?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3 = this._s3?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._s3.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._s3.internalValue = value.s3;
        }
    }
    get s3() {
        return this._s3;
    }
    putS3(value) {
        this._s3.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get s3Input() {
        return this._s3.internalValue;
    }
}
exports.IvsRecordingConfigurationDestinationConfigurationOutputReference = IvsRecordingConfigurationDestinationConfigurationOutputReference;
_b = JSII_RTTI_SYMBOL_1;
IvsRecordingConfigurationDestinationConfigurationOutputReference[_b] = { fqn: "@cdktf/provider-aws.ivsRecordingConfiguration.IvsRecordingConfigurationDestinationConfigurationOutputReference", version: "10.0.12" };
function ivsRecordingConfigurationThumbnailConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        recording_mode: cdktf.stringToTerraform(struct.recordingMode),
        target_interval_seconds: cdktf.numberToTerraform(struct.targetIntervalSeconds),
    };
}
exports.ivsRecordingConfigurationThumbnailConfigurationToTerraform = ivsRecordingConfigurationThumbnailConfigurationToTerraform;
class IvsRecordingConfigurationThumbnailConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._recordingMode !== undefined) {
            hasAnyValues = true;
            internalValueResult.recordingMode = this._recordingMode;
        }
        if (this._targetIntervalSeconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetIntervalSeconds = this._targetIntervalSeconds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._recordingMode = undefined;
            this._targetIntervalSeconds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._recordingMode = value.recordingMode;
            this._targetIntervalSeconds = value.targetIntervalSeconds;
        }
    }
    get recordingMode() {
        return this.getStringAttribute('recording_mode');
    }
    set recordingMode(value) {
        this._recordingMode = value;
    }
    resetRecordingMode() {
        this._recordingMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get recordingModeInput() {
        return this._recordingMode;
    }
    get targetIntervalSeconds() {
        return this.getNumberAttribute('target_interval_seconds');
    }
    set targetIntervalSeconds(value) {
        this._targetIntervalSeconds = value;
    }
    resetTargetIntervalSeconds() {
        this._targetIntervalSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetIntervalSecondsInput() {
        return this._targetIntervalSeconds;
    }
}
exports.IvsRecordingConfigurationThumbnailConfigurationOutputReference = IvsRecordingConfigurationThumbnailConfigurationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
IvsRecordingConfigurationThumbnailConfigurationOutputReference[_c] = { fqn: "@cdktf/provider-aws.ivsRecordingConfiguration.IvsRecordingConfigurationThumbnailConfigurationOutputReference", version: "10.0.12" };
function ivsRecordingConfigurationTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.ivsRecordingConfigurationTimeoutsToTerraform = ivsRecordingConfigurationTimeoutsToTerraform;
class IvsRecordingConfigurationTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
}
exports.IvsRecordingConfigurationTimeoutsOutputReference = IvsRecordingConfigurationTimeoutsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
IvsRecordingConfigurationTimeoutsOutputReference[_d] = { fqn: "@cdktf/provider-aws.ivsRecordingConfiguration.IvsRecordingConfigurationTimeoutsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/ivs_recording_configuration aws_ivs_recording_configuration}
*/
class IvsRecordingConfiguration extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ivs_recording_configuration aws_ivs_recording_configuration} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options IvsRecordingConfigurationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ivs_recording_configuration',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // destination_configuration - computed: false, optional: false, required: true
        this._destinationConfiguration = new IvsRecordingConfigurationDestinationConfigurationOutputReference(this, "destination_configuration");
        // thumbnail_configuration - computed: false, optional: true, required: false
        this._thumbnailConfiguration = new IvsRecordingConfigurationThumbnailConfigurationOutputReference(this, "thumbnail_configuration");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new IvsRecordingConfigurationTimeoutsOutputReference(this, "timeouts");
        this._id = config.id;
        this._name = config.name;
        this._recordingReconnectWindowSeconds = config.recordingReconnectWindowSeconds;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._destinationConfiguration.internalValue = config.destinationConfiguration;
        this._thumbnailConfiguration.internalValue = config.thumbnailConfiguration;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get recordingReconnectWindowSeconds() {
        return this.getNumberAttribute('recording_reconnect_window_seconds');
    }
    set recordingReconnectWindowSeconds(value) {
        this._recordingReconnectWindowSeconds = value;
    }
    resetRecordingReconnectWindowSeconds() {
        this._recordingReconnectWindowSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get recordingReconnectWindowSecondsInput() {
        return this._recordingReconnectWindowSeconds;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get destinationConfiguration() {
        return this._destinationConfiguration;
    }
    putDestinationConfiguration(value) {
        this._destinationConfiguration.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get destinationConfigurationInput() {
        return this._destinationConfiguration.internalValue;
    }
    get thumbnailConfiguration() {
        return this._thumbnailConfiguration;
    }
    putThumbnailConfiguration(value) {
        this._thumbnailConfiguration.internalValue = value;
    }
    resetThumbnailConfiguration() {
        this._thumbnailConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get thumbnailConfigurationInput() {
        return this._thumbnailConfiguration.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            recording_reconnect_window_seconds: cdktf.numberToTerraform(this._recordingReconnectWindowSeconds),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            destination_configuration: ivsRecordingConfigurationDestinationConfigurationToTerraform(this._destinationConfiguration.internalValue),
            thumbnail_configuration: ivsRecordingConfigurationThumbnailConfigurationToTerraform(this._thumbnailConfiguration.internalValue),
            timeouts: ivsRecordingConfigurationTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.IvsRecordingConfiguration = IvsRecordingConfiguration;
_e = JSII_RTTI_SYMBOL_1;
IvsRecordingConfiguration[_e] = { fqn: "@cdktf/provider-aws.ivsRecordingConfiguration.IvsRecordingConfiguration", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
IvsRecordingConfiguration.tfResourceType = "aws_ivs_recording_configuration";
