// @cdktf/provider-aws/lib/guardduty-invite-accepter/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GuarddutyInviteAccepter = exports.GuarddutyInviteAccepterTimeoutsOutputReference = exports.guarddutyInviteAccepterTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function guarddutyInviteAccepterTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
    };
}
exports.guarddutyInviteAccepterTimeoutsToTerraform = guarddutyInviteAccepterTimeoutsToTerraform;
class GuarddutyInviteAccepterTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
}
exports.GuarddutyInviteAccepterTimeoutsOutputReference = GuarddutyInviteAccepterTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
GuarddutyInviteAccepterTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.guarddutyInviteAccepter.GuarddutyInviteAccepterTimeoutsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/guardduty_invite_accepter aws_guardduty_invite_accepter}
*/
class GuarddutyInviteAccepter extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/guardduty_invite_accepter aws_guardduty_invite_accepter} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options GuarddutyInviteAccepterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_guardduty_invite_accepter',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new GuarddutyInviteAccepterTimeoutsOutputReference(this, "timeouts");
        this._detectorId = config.detectorId;
        this._id = config.id;
        this._masterAccountId = config.masterAccountId;
        this._timeouts.internalValue = config.timeouts;
    }
    get detectorId() {
        return this.getStringAttribute('detector_id');
    }
    set detectorId(value) {
        this._detectorId = value;
    }
    // Temporarily expose input value. Use with caution.
    get detectorIdInput() {
        return this._detectorId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get masterAccountId() {
        return this.getStringAttribute('master_account_id');
    }
    set masterAccountId(value) {
        this._masterAccountId = value;
    }
    // Temporarily expose input value. Use with caution.
    get masterAccountIdInput() {
        return this._masterAccountId;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            detector_id: cdktf.stringToTerraform(this._detectorId),
            id: cdktf.stringToTerraform(this._id),
            master_account_id: cdktf.stringToTerraform(this._masterAccountId),
            timeouts: guarddutyInviteAccepterTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.GuarddutyInviteAccepter = GuarddutyInviteAccepter;
_b = JSII_RTTI_SYMBOL_1;
GuarddutyInviteAccepter[_b] = { fqn: "@cdktf/provider-aws.guarddutyInviteAccepter.GuarddutyInviteAccepter", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
GuarddutyInviteAccepter.tfResourceType = "aws_guardduty_invite_accepter";
