// @cdktf/provider-aws/lib/glacier-vault-lock/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlacierVaultLock = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/glacier_vault_lock aws_glacier_vault_lock}
*/
class GlacierVaultLock extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/glacier_vault_lock aws_glacier_vault_lock} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options GlacierVaultLockConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_glacier_vault_lock',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._completeLock = config.completeLock;
        this._id = config.id;
        this._ignoreDeletionError = config.ignoreDeletionError;
        this._policy = config.policy;
        this._vaultName = config.vaultName;
    }
    get completeLock() {
        return this.getBooleanAttribute('complete_lock');
    }
    set completeLock(value) {
        this._completeLock = value;
    }
    // Temporarily expose input value. Use with caution.
    get completeLockInput() {
        return this._completeLock;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get ignoreDeletionError() {
        return this.getBooleanAttribute('ignore_deletion_error');
    }
    set ignoreDeletionError(value) {
        this._ignoreDeletionError = value;
    }
    resetIgnoreDeletionError() {
        this._ignoreDeletionError = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ignoreDeletionErrorInput() {
        return this._ignoreDeletionError;
    }
    get policy() {
        return this.getStringAttribute('policy');
    }
    set policy(value) {
        this._policy = value;
    }
    // Temporarily expose input value. Use with caution.
    get policyInput() {
        return this._policy;
    }
    get vaultName() {
        return this.getStringAttribute('vault_name');
    }
    set vaultName(value) {
        this._vaultName = value;
    }
    // Temporarily expose input value. Use with caution.
    get vaultNameInput() {
        return this._vaultName;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            complete_lock: cdktf.booleanToTerraform(this._completeLock),
            id: cdktf.stringToTerraform(this._id),
            ignore_deletion_error: cdktf.booleanToTerraform(this._ignoreDeletionError),
            policy: cdktf.stringToTerraform(this._policy),
            vault_name: cdktf.stringToTerraform(this._vaultName),
        };
    }
}
exports.GlacierVaultLock = GlacierVaultLock;
_a = JSII_RTTI_SYMBOL_1;
GlacierVaultLock[_a] = { fqn: "@cdktf/provider-aws.glacierVaultLock.GlacierVaultLock", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
GlacierVaultLock.tfResourceType = "aws_glacier_vault_lock";
