// @cdktf/provider-aws/lib/ssm-patch-group/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SsmPatchGroup = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/ssm_patch_group aws_ssm_patch_group}
*/
class SsmPatchGroup extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ssm_patch_group aws_ssm_patch_group} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SsmPatchGroupConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ssm_patch_group',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._baselineId = config.baselineId;
        this._id = config.id;
        this._patchGroup = config.patchGroup;
    }
    get baselineId() {
        return this.getStringAttribute('baseline_id');
    }
    set baselineId(value) {
        this._baselineId = value;
    }
    // Temporarily expose input value. Use with caution.
    get baselineIdInput() {
        return this._baselineId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get patchGroup() {
        return this.getStringAttribute('patch_group');
    }
    set patchGroup(value) {
        this._patchGroup = value;
    }
    // Temporarily expose input value. Use with caution.
    get patchGroupInput() {
        return this._patchGroup;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            baseline_id: cdktf.stringToTerraform(this._baselineId),
            id: cdktf.stringToTerraform(this._id),
            patch_group: cdktf.stringToTerraform(this._patchGroup),
        };
    }
}
exports.SsmPatchGroup = SsmPatchGroup;
_a = JSII_RTTI_SYMBOL_1;
SsmPatchGroup[_a] = { fqn: "@cdktf/provider-aws.ssmPatchGroup.SsmPatchGroup", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
SsmPatchGroup.tfResourceType = "aws_ssm_patch_group";
