// @cdktf/provider-aws/lib/sagemaker-model/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SagemakerModel = exports.SagemakerModelVpcConfigOutputReference = exports.sagemakerModelVpcConfigToTerraform = exports.SagemakerModelPrimaryContainerOutputReference = exports.sagemakerModelPrimaryContainerToTerraform = exports.SagemakerModelPrimaryContainerImageConfigOutputReference = exports.sagemakerModelPrimaryContainerImageConfigToTerraform = exports.SagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigOutputReference = exports.sagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigToTerraform = exports.SagemakerModelInferenceExecutionConfigOutputReference = exports.sagemakerModelInferenceExecutionConfigToTerraform = exports.SagemakerModelContainerList = exports.SagemakerModelContainerOutputReference = exports.sagemakerModelContainerToTerraform = exports.SagemakerModelContainerImageConfigOutputReference = exports.sagemakerModelContainerImageConfigToTerraform = exports.SagemakerModelContainerImageConfigRepositoryAuthConfigOutputReference = exports.sagemakerModelContainerImageConfigRepositoryAuthConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function sagemakerModelContainerImageConfigRepositoryAuthConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        repository_credentials_provider_arn: cdktf.stringToTerraform(struct.repositoryCredentialsProviderArn),
    };
}
exports.sagemakerModelContainerImageConfigRepositoryAuthConfigToTerraform = sagemakerModelContainerImageConfigRepositoryAuthConfigToTerraform;
class SagemakerModelContainerImageConfigRepositoryAuthConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._repositoryCredentialsProviderArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.repositoryCredentialsProviderArn = this._repositoryCredentialsProviderArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._repositoryCredentialsProviderArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._repositoryCredentialsProviderArn = value.repositoryCredentialsProviderArn;
        }
    }
    get repositoryCredentialsProviderArn() {
        return this.getStringAttribute('repository_credentials_provider_arn');
    }
    set repositoryCredentialsProviderArn(value) {
        this._repositoryCredentialsProviderArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryCredentialsProviderArnInput() {
        return this._repositoryCredentialsProviderArn;
    }
}
exports.SagemakerModelContainerImageConfigRepositoryAuthConfigOutputReference = SagemakerModelContainerImageConfigRepositoryAuthConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SagemakerModelContainerImageConfigRepositoryAuthConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModelContainerImageConfigRepositoryAuthConfigOutputReference", version: "10.0.12" };
function sagemakerModelContainerImageConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        repository_access_mode: cdktf.stringToTerraform(struct.repositoryAccessMode),
        repository_auth_config: sagemakerModelContainerImageConfigRepositoryAuthConfigToTerraform(struct.repositoryAuthConfig),
    };
}
exports.sagemakerModelContainerImageConfigToTerraform = sagemakerModelContainerImageConfigToTerraform;
class SagemakerModelContainerImageConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // repository_auth_config - computed: false, optional: true, required: false
        this._repositoryAuthConfig = new SagemakerModelContainerImageConfigRepositoryAuthConfigOutputReference(this, "repository_auth_config");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._repositoryAccessMode !== undefined) {
            hasAnyValues = true;
            internalValueResult.repositoryAccessMode = this._repositoryAccessMode;
        }
        if (this._repositoryAuthConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.repositoryAuthConfig = this._repositoryAuthConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._repositoryAccessMode = undefined;
            this._repositoryAuthConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._repositoryAccessMode = value.repositoryAccessMode;
            this._repositoryAuthConfig.internalValue = value.repositoryAuthConfig;
        }
    }
    get repositoryAccessMode() {
        return this.getStringAttribute('repository_access_mode');
    }
    set repositoryAccessMode(value) {
        this._repositoryAccessMode = value;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryAccessModeInput() {
        return this._repositoryAccessMode;
    }
    get repositoryAuthConfig() {
        return this._repositoryAuthConfig;
    }
    putRepositoryAuthConfig(value) {
        this._repositoryAuthConfig.internalValue = value;
    }
    resetRepositoryAuthConfig() {
        this._repositoryAuthConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryAuthConfigInput() {
        return this._repositoryAuthConfig.internalValue;
    }
}
exports.SagemakerModelContainerImageConfigOutputReference = SagemakerModelContainerImageConfigOutputReference;
_b = JSII_RTTI_SYMBOL_1;
SagemakerModelContainerImageConfigOutputReference[_b] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModelContainerImageConfigOutputReference", version: "10.0.12" };
function sagemakerModelContainerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        container_hostname: cdktf.stringToTerraform(struct.containerHostname),
        environment: cdktf.hashMapper(cdktf.stringToTerraform)(struct.environment),
        image: cdktf.stringToTerraform(struct.image),
        mode: cdktf.stringToTerraform(struct.mode),
        model_data_url: cdktf.stringToTerraform(struct.modelDataUrl),
        image_config: sagemakerModelContainerImageConfigToTerraform(struct.imageConfig),
    };
}
exports.sagemakerModelContainerToTerraform = sagemakerModelContainerToTerraform;
class SagemakerModelContainerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // image_config - computed: false, optional: true, required: false
        this._imageConfig = new SagemakerModelContainerImageConfigOutputReference(this, "image_config");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._containerHostname !== undefined) {
            hasAnyValues = true;
            internalValueResult.containerHostname = this._containerHostname;
        }
        if (this._environment !== undefined) {
            hasAnyValues = true;
            internalValueResult.environment = this._environment;
        }
        if (this._image !== undefined) {
            hasAnyValues = true;
            internalValueResult.image = this._image;
        }
        if (this._mode !== undefined) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        if (this._modelDataUrl !== undefined) {
            hasAnyValues = true;
            internalValueResult.modelDataUrl = this._modelDataUrl;
        }
        if (this._imageConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.imageConfig = this._imageConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._containerHostname = undefined;
            this._environment = undefined;
            this._image = undefined;
            this._mode = undefined;
            this._modelDataUrl = undefined;
            this._imageConfig.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._containerHostname = value.containerHostname;
            this._environment = value.environment;
            this._image = value.image;
            this._mode = value.mode;
            this._modelDataUrl = value.modelDataUrl;
            this._imageConfig.internalValue = value.imageConfig;
        }
    }
    get containerHostname() {
        return this.getStringAttribute('container_hostname');
    }
    set containerHostname(value) {
        this._containerHostname = value;
    }
    resetContainerHostname() {
        this._containerHostname = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerHostnameInput() {
        return this._containerHostname;
    }
    get environment() {
        return this.getStringMapAttribute('environment');
    }
    set environment(value) {
        this._environment = value;
    }
    resetEnvironment() {
        this._environment = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get environmentInput() {
        return this._environment;
    }
    get image() {
        return this.getStringAttribute('image');
    }
    set image(value) {
        this._image = value;
    }
    // Temporarily expose input value. Use with caution.
    get imageInput() {
        return this._image;
    }
    get mode() {
        return this.getStringAttribute('mode');
    }
    set mode(value) {
        this._mode = value;
    }
    resetMode() {
        this._mode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modeInput() {
        return this._mode;
    }
    get modelDataUrl() {
        return this.getStringAttribute('model_data_url');
    }
    set modelDataUrl(value) {
        this._modelDataUrl = value;
    }
    resetModelDataUrl() {
        this._modelDataUrl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modelDataUrlInput() {
        return this._modelDataUrl;
    }
    get imageConfig() {
        return this._imageConfig;
    }
    putImageConfig(value) {
        this._imageConfig.internalValue = value;
    }
    resetImageConfig() {
        this._imageConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get imageConfigInput() {
        return this._imageConfig.internalValue;
    }
}
exports.SagemakerModelContainerOutputReference = SagemakerModelContainerOutputReference;
_c = JSII_RTTI_SYMBOL_1;
SagemakerModelContainerOutputReference[_c] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModelContainerOutputReference", version: "10.0.12" };
class SagemakerModelContainerList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SagemakerModelContainerOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SagemakerModelContainerList = SagemakerModelContainerList;
_d = JSII_RTTI_SYMBOL_1;
SagemakerModelContainerList[_d] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModelContainerList", version: "10.0.12" };
function sagemakerModelInferenceExecutionConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        mode: cdktf.stringToTerraform(struct.mode),
    };
}
exports.sagemakerModelInferenceExecutionConfigToTerraform = sagemakerModelInferenceExecutionConfigToTerraform;
class SagemakerModelInferenceExecutionConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._mode !== undefined) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._mode = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._mode = value.mode;
        }
    }
    get mode() {
        return this.getStringAttribute('mode');
    }
    set mode(value) {
        this._mode = value;
    }
    // Temporarily expose input value. Use with caution.
    get modeInput() {
        return this._mode;
    }
}
exports.SagemakerModelInferenceExecutionConfigOutputReference = SagemakerModelInferenceExecutionConfigOutputReference;
_e = JSII_RTTI_SYMBOL_1;
SagemakerModelInferenceExecutionConfigOutputReference[_e] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModelInferenceExecutionConfigOutputReference", version: "10.0.12" };
function sagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        repository_credentials_provider_arn: cdktf.stringToTerraform(struct.repositoryCredentialsProviderArn),
    };
}
exports.sagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigToTerraform = sagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigToTerraform;
class SagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._repositoryCredentialsProviderArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.repositoryCredentialsProviderArn = this._repositoryCredentialsProviderArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._repositoryCredentialsProviderArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._repositoryCredentialsProviderArn = value.repositoryCredentialsProviderArn;
        }
    }
    get repositoryCredentialsProviderArn() {
        return this.getStringAttribute('repository_credentials_provider_arn');
    }
    set repositoryCredentialsProviderArn(value) {
        this._repositoryCredentialsProviderArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryCredentialsProviderArnInput() {
        return this._repositoryCredentialsProviderArn;
    }
}
exports.SagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigOutputReference = SagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigOutputReference;
_f = JSII_RTTI_SYMBOL_1;
SagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigOutputReference[_f] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigOutputReference", version: "10.0.12" };
function sagemakerModelPrimaryContainerImageConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        repository_access_mode: cdktf.stringToTerraform(struct.repositoryAccessMode),
        repository_auth_config: sagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigToTerraform(struct.repositoryAuthConfig),
    };
}
exports.sagemakerModelPrimaryContainerImageConfigToTerraform = sagemakerModelPrimaryContainerImageConfigToTerraform;
class SagemakerModelPrimaryContainerImageConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // repository_auth_config - computed: false, optional: true, required: false
        this._repositoryAuthConfig = new SagemakerModelPrimaryContainerImageConfigRepositoryAuthConfigOutputReference(this, "repository_auth_config");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._repositoryAccessMode !== undefined) {
            hasAnyValues = true;
            internalValueResult.repositoryAccessMode = this._repositoryAccessMode;
        }
        if (this._repositoryAuthConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.repositoryAuthConfig = this._repositoryAuthConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._repositoryAccessMode = undefined;
            this._repositoryAuthConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._repositoryAccessMode = value.repositoryAccessMode;
            this._repositoryAuthConfig.internalValue = value.repositoryAuthConfig;
        }
    }
    get repositoryAccessMode() {
        return this.getStringAttribute('repository_access_mode');
    }
    set repositoryAccessMode(value) {
        this._repositoryAccessMode = value;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryAccessModeInput() {
        return this._repositoryAccessMode;
    }
    get repositoryAuthConfig() {
        return this._repositoryAuthConfig;
    }
    putRepositoryAuthConfig(value) {
        this._repositoryAuthConfig.internalValue = value;
    }
    resetRepositoryAuthConfig() {
        this._repositoryAuthConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryAuthConfigInput() {
        return this._repositoryAuthConfig.internalValue;
    }
}
exports.SagemakerModelPrimaryContainerImageConfigOutputReference = SagemakerModelPrimaryContainerImageConfigOutputReference;
_g = JSII_RTTI_SYMBOL_1;
SagemakerModelPrimaryContainerImageConfigOutputReference[_g] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModelPrimaryContainerImageConfigOutputReference", version: "10.0.12" };
function sagemakerModelPrimaryContainerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        container_hostname: cdktf.stringToTerraform(struct.containerHostname),
        environment: cdktf.hashMapper(cdktf.stringToTerraform)(struct.environment),
        image: cdktf.stringToTerraform(struct.image),
        mode: cdktf.stringToTerraform(struct.mode),
        model_data_url: cdktf.stringToTerraform(struct.modelDataUrl),
        image_config: sagemakerModelPrimaryContainerImageConfigToTerraform(struct.imageConfig),
    };
}
exports.sagemakerModelPrimaryContainerToTerraform = sagemakerModelPrimaryContainerToTerraform;
class SagemakerModelPrimaryContainerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // image_config - computed: false, optional: true, required: false
        this._imageConfig = new SagemakerModelPrimaryContainerImageConfigOutputReference(this, "image_config");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._containerHostname !== undefined) {
            hasAnyValues = true;
            internalValueResult.containerHostname = this._containerHostname;
        }
        if (this._environment !== undefined) {
            hasAnyValues = true;
            internalValueResult.environment = this._environment;
        }
        if (this._image !== undefined) {
            hasAnyValues = true;
            internalValueResult.image = this._image;
        }
        if (this._mode !== undefined) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        if (this._modelDataUrl !== undefined) {
            hasAnyValues = true;
            internalValueResult.modelDataUrl = this._modelDataUrl;
        }
        if (this._imageConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.imageConfig = this._imageConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._containerHostname = undefined;
            this._environment = undefined;
            this._image = undefined;
            this._mode = undefined;
            this._modelDataUrl = undefined;
            this._imageConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._containerHostname = value.containerHostname;
            this._environment = value.environment;
            this._image = value.image;
            this._mode = value.mode;
            this._modelDataUrl = value.modelDataUrl;
            this._imageConfig.internalValue = value.imageConfig;
        }
    }
    get containerHostname() {
        return this.getStringAttribute('container_hostname');
    }
    set containerHostname(value) {
        this._containerHostname = value;
    }
    resetContainerHostname() {
        this._containerHostname = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerHostnameInput() {
        return this._containerHostname;
    }
    get environment() {
        return this.getStringMapAttribute('environment');
    }
    set environment(value) {
        this._environment = value;
    }
    resetEnvironment() {
        this._environment = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get environmentInput() {
        return this._environment;
    }
    get image() {
        return this.getStringAttribute('image');
    }
    set image(value) {
        this._image = value;
    }
    // Temporarily expose input value. Use with caution.
    get imageInput() {
        return this._image;
    }
    get mode() {
        return this.getStringAttribute('mode');
    }
    set mode(value) {
        this._mode = value;
    }
    resetMode() {
        this._mode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modeInput() {
        return this._mode;
    }
    get modelDataUrl() {
        return this.getStringAttribute('model_data_url');
    }
    set modelDataUrl(value) {
        this._modelDataUrl = value;
    }
    resetModelDataUrl() {
        this._modelDataUrl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modelDataUrlInput() {
        return this._modelDataUrl;
    }
    get imageConfig() {
        return this._imageConfig;
    }
    putImageConfig(value) {
        this._imageConfig.internalValue = value;
    }
    resetImageConfig() {
        this._imageConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get imageConfigInput() {
        return this._imageConfig.internalValue;
    }
}
exports.SagemakerModelPrimaryContainerOutputReference = SagemakerModelPrimaryContainerOutputReference;
_h = JSII_RTTI_SYMBOL_1;
SagemakerModelPrimaryContainerOutputReference[_h] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModelPrimaryContainerOutputReference", version: "10.0.12" };
function sagemakerModelVpcConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.securityGroupIds),
        subnets: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.subnets),
    };
}
exports.sagemakerModelVpcConfigToTerraform = sagemakerModelVpcConfigToTerraform;
class SagemakerModelVpcConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._securityGroupIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroupIds = this._securityGroupIds;
        }
        if (this._subnets !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnets = this._subnets;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._securityGroupIds = undefined;
            this._subnets = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._securityGroupIds = value.securityGroupIds;
            this._subnets = value.subnets;
        }
    }
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    get subnets() {
        return cdktf.Fn.tolist(this.getListAttribute('subnets'));
    }
    set subnets(value) {
        this._subnets = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetsInput() {
        return this._subnets;
    }
}
exports.SagemakerModelVpcConfigOutputReference = SagemakerModelVpcConfigOutputReference;
_j = JSII_RTTI_SYMBOL_1;
SagemakerModelVpcConfigOutputReference[_j] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModelVpcConfigOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/sagemaker_model aws_sagemaker_model}
*/
class SagemakerModel extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/sagemaker_model aws_sagemaker_model} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SagemakerModelConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_sagemaker_model',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // container - computed: false, optional: true, required: false
        this._container = new SagemakerModelContainerList(this, "container", false);
        // inference_execution_config - computed: false, optional: true, required: false
        this._inferenceExecutionConfig = new SagemakerModelInferenceExecutionConfigOutputReference(this, "inference_execution_config");
        // primary_container - computed: false, optional: true, required: false
        this._primaryContainer = new SagemakerModelPrimaryContainerOutputReference(this, "primary_container");
        // vpc_config - computed: false, optional: true, required: false
        this._vpcConfig = new SagemakerModelVpcConfigOutputReference(this, "vpc_config");
        this._enableNetworkIsolation = config.enableNetworkIsolation;
        this._executionRoleArn = config.executionRoleArn;
        this._id = config.id;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._container.internalValue = config.container;
        this._inferenceExecutionConfig.internalValue = config.inferenceExecutionConfig;
        this._primaryContainer.internalValue = config.primaryContainer;
        this._vpcConfig.internalValue = config.vpcConfig;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get enableNetworkIsolation() {
        return this.getBooleanAttribute('enable_network_isolation');
    }
    set enableNetworkIsolation(value) {
        this._enableNetworkIsolation = value;
    }
    resetEnableNetworkIsolation() {
        this._enableNetworkIsolation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableNetworkIsolationInput() {
        return this._enableNetworkIsolation;
    }
    get executionRoleArn() {
        return this.getStringAttribute('execution_role_arn');
    }
    set executionRoleArn(value) {
        this._executionRoleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get executionRoleArnInput() {
        return this._executionRoleArn;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get container() {
        return this._container;
    }
    putContainer(value) {
        this._container.internalValue = value;
    }
    resetContainer() {
        this._container.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerInput() {
        return this._container.internalValue;
    }
    get inferenceExecutionConfig() {
        return this._inferenceExecutionConfig;
    }
    putInferenceExecutionConfig(value) {
        this._inferenceExecutionConfig.internalValue = value;
    }
    resetInferenceExecutionConfig() {
        this._inferenceExecutionConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get inferenceExecutionConfigInput() {
        return this._inferenceExecutionConfig.internalValue;
    }
    get primaryContainer() {
        return this._primaryContainer;
    }
    putPrimaryContainer(value) {
        this._primaryContainer.internalValue = value;
    }
    resetPrimaryContainer() {
        this._primaryContainer.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get primaryContainerInput() {
        return this._primaryContainer.internalValue;
    }
    get vpcConfig() {
        return this._vpcConfig;
    }
    putVpcConfig(value) {
        this._vpcConfig.internalValue = value;
    }
    resetVpcConfig() {
        this._vpcConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcConfigInput() {
        return this._vpcConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            enable_network_isolation: cdktf.booleanToTerraform(this._enableNetworkIsolation),
            execution_role_arn: cdktf.stringToTerraform(this._executionRoleArn),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            container: cdktf.listMapper(sagemakerModelContainerToTerraform, true)(this._container.internalValue),
            inference_execution_config: sagemakerModelInferenceExecutionConfigToTerraform(this._inferenceExecutionConfig.internalValue),
            primary_container: sagemakerModelPrimaryContainerToTerraform(this._primaryContainer.internalValue),
            vpc_config: sagemakerModelVpcConfigToTerraform(this._vpcConfig.internalValue),
        };
    }
}
exports.SagemakerModel = SagemakerModel;
_k = JSII_RTTI_SYMBOL_1;
SagemakerModel[_k] = { fqn: "@cdktf/provider-aws.sagemakerModel.SagemakerModel", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
SagemakerModel.tfResourceType = "aws_sagemaker_model";
