// @cdktf/provider-aws/lib/batch-job-queue/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BatchJobQueue = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/batch_job_queue aws_batch_job_queue}
*/
class BatchJobQueue extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/batch_job_queue aws_batch_job_queue} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options BatchJobQueueConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_batch_job_queue',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._computeEnvironments = config.computeEnvironments;
        this._id = config.id;
        this._name = config.name;
        this._priority = config.priority;
        this._schedulingPolicyArn = config.schedulingPolicyArn;
        this._state = config.state;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get computeEnvironments() {
        return this.getListAttribute('compute_environments');
    }
    set computeEnvironments(value) {
        this._computeEnvironments = value;
    }
    // Temporarily expose input value. Use with caution.
    get computeEnvironmentsInput() {
        return this._computeEnvironments;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get priority() {
        return this.getNumberAttribute('priority');
    }
    set priority(value) {
        this._priority = value;
    }
    // Temporarily expose input value. Use with caution.
    get priorityInput() {
        return this._priority;
    }
    get schedulingPolicyArn() {
        return this.getStringAttribute('scheduling_policy_arn');
    }
    set schedulingPolicyArn(value) {
        this._schedulingPolicyArn = value;
    }
    resetSchedulingPolicyArn() {
        this._schedulingPolicyArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get schedulingPolicyArnInput() {
        return this._schedulingPolicyArn;
    }
    get state() {
        return this.getStringAttribute('state');
    }
    set state(value) {
        this._state = value;
    }
    // Temporarily expose input value. Use with caution.
    get stateInput() {
        return this._state;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            compute_environments: cdktf.listMapper(cdktf.stringToTerraform, false)(this._computeEnvironments),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            priority: cdktf.numberToTerraform(this._priority),
            scheduling_policy_arn: cdktf.stringToTerraform(this._schedulingPolicyArn),
            state: cdktf.stringToTerraform(this._state),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
        };
    }
}
exports.BatchJobQueue = BatchJobQueue;
_a = JSII_RTTI_SYMBOL_1;
BatchJobQueue[_a] = { fqn: "@cdktf/provider-aws.batchJobQueue.BatchJobQueue", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
BatchJobQueue.tfResourceType = "aws_batch_job_queue";
