// @cdktf/provider-aws/lib/organizations-organization/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrganizationsOrganization = exports.OrganizationsOrganizationRootsList = exports.OrganizationsOrganizationRootsOutputReference = exports.organizationsOrganizationRootsToTerraform = exports.OrganizationsOrganizationRootsPolicyTypesList = exports.OrganizationsOrganizationRootsPolicyTypesOutputReference = exports.organizationsOrganizationRootsPolicyTypesToTerraform = exports.OrganizationsOrganizationNonMasterAccountsList = exports.OrganizationsOrganizationNonMasterAccountsOutputReference = exports.organizationsOrganizationNonMasterAccountsToTerraform = exports.OrganizationsOrganizationAccountsList = exports.OrganizationsOrganizationAccountsOutputReference = exports.organizationsOrganizationAccountsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function organizationsOrganizationAccountsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.organizationsOrganizationAccountsToTerraform = organizationsOrganizationAccountsToTerraform;
class OrganizationsOrganizationAccountsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // email - computed: true, optional: false, required: false
    get email() {
        return this.getStringAttribute('email');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
}
exports.OrganizationsOrganizationAccountsOutputReference = OrganizationsOrganizationAccountsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
OrganizationsOrganizationAccountsOutputReference[_a] = { fqn: "@cdktf/provider-aws.organizationsOrganization.OrganizationsOrganizationAccountsOutputReference", version: "10.0.12" };
class OrganizationsOrganizationAccountsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new OrganizationsOrganizationAccountsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.OrganizationsOrganizationAccountsList = OrganizationsOrganizationAccountsList;
_b = JSII_RTTI_SYMBOL_1;
OrganizationsOrganizationAccountsList[_b] = { fqn: "@cdktf/provider-aws.organizationsOrganization.OrganizationsOrganizationAccountsList", version: "10.0.12" };
function organizationsOrganizationNonMasterAccountsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.organizationsOrganizationNonMasterAccountsToTerraform = organizationsOrganizationNonMasterAccountsToTerraform;
class OrganizationsOrganizationNonMasterAccountsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // email - computed: true, optional: false, required: false
    get email() {
        return this.getStringAttribute('email');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
}
exports.OrganizationsOrganizationNonMasterAccountsOutputReference = OrganizationsOrganizationNonMasterAccountsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
OrganizationsOrganizationNonMasterAccountsOutputReference[_c] = { fqn: "@cdktf/provider-aws.organizationsOrganization.OrganizationsOrganizationNonMasterAccountsOutputReference", version: "10.0.12" };
class OrganizationsOrganizationNonMasterAccountsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new OrganizationsOrganizationNonMasterAccountsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.OrganizationsOrganizationNonMasterAccountsList = OrganizationsOrganizationNonMasterAccountsList;
_d = JSII_RTTI_SYMBOL_1;
OrganizationsOrganizationNonMasterAccountsList[_d] = { fqn: "@cdktf/provider-aws.organizationsOrganization.OrganizationsOrganizationNonMasterAccountsList", version: "10.0.12" };
function organizationsOrganizationRootsPolicyTypesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.organizationsOrganizationRootsPolicyTypesToTerraform = organizationsOrganizationRootsPolicyTypesToTerraform;
class OrganizationsOrganizationRootsPolicyTypesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.OrganizationsOrganizationRootsPolicyTypesOutputReference = OrganizationsOrganizationRootsPolicyTypesOutputReference;
_e = JSII_RTTI_SYMBOL_1;
OrganizationsOrganizationRootsPolicyTypesOutputReference[_e] = { fqn: "@cdktf/provider-aws.organizationsOrganization.OrganizationsOrganizationRootsPolicyTypesOutputReference", version: "10.0.12" };
class OrganizationsOrganizationRootsPolicyTypesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new OrganizationsOrganizationRootsPolicyTypesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.OrganizationsOrganizationRootsPolicyTypesList = OrganizationsOrganizationRootsPolicyTypesList;
_f = JSII_RTTI_SYMBOL_1;
OrganizationsOrganizationRootsPolicyTypesList[_f] = { fqn: "@cdktf/provider-aws.organizationsOrganization.OrganizationsOrganizationRootsPolicyTypesList", version: "10.0.12" };
function organizationsOrganizationRootsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.organizationsOrganizationRootsToTerraform = organizationsOrganizationRootsToTerraform;
class OrganizationsOrganizationRootsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // policy_types - computed: true, optional: false, required: false
        this._policyTypes = new OrganizationsOrganizationRootsPolicyTypesList(this, "policy_types", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get policyTypes() {
        return this._policyTypes;
    }
}
exports.OrganizationsOrganizationRootsOutputReference = OrganizationsOrganizationRootsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
OrganizationsOrganizationRootsOutputReference[_g] = { fqn: "@cdktf/provider-aws.organizationsOrganization.OrganizationsOrganizationRootsOutputReference", version: "10.0.12" };
class OrganizationsOrganizationRootsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new OrganizationsOrganizationRootsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.OrganizationsOrganizationRootsList = OrganizationsOrganizationRootsList;
_h = JSII_RTTI_SYMBOL_1;
OrganizationsOrganizationRootsList[_h] = { fqn: "@cdktf/provider-aws.organizationsOrganization.OrganizationsOrganizationRootsList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/organizations_organization aws_organizations_organization}
*/
class OrganizationsOrganization extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/organizations_organization aws_organizations_organization} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options OrganizationsOrganizationConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_organizations_organization',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // accounts - computed: true, optional: false, required: false
        this._accounts = new OrganizationsOrganizationAccountsList(this, "accounts", false);
        // non_master_accounts - computed: true, optional: false, required: false
        this._nonMasterAccounts = new OrganizationsOrganizationNonMasterAccountsList(this, "non_master_accounts", false);
        // roots - computed: true, optional: false, required: false
        this._roots = new OrganizationsOrganizationRootsList(this, "roots", false);
        this._awsServiceAccessPrincipals = config.awsServiceAccessPrincipals;
        this._enabledPolicyTypes = config.enabledPolicyTypes;
        this._featureSet = config.featureSet;
        this._id = config.id;
    }
    get accounts() {
        return this._accounts;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get awsServiceAccessPrincipals() {
        return cdktf.Fn.tolist(this.getListAttribute('aws_service_access_principals'));
    }
    set awsServiceAccessPrincipals(value) {
        this._awsServiceAccessPrincipals = value;
    }
    resetAwsServiceAccessPrincipals() {
        this._awsServiceAccessPrincipals = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get awsServiceAccessPrincipalsInput() {
        return this._awsServiceAccessPrincipals;
    }
    get enabledPolicyTypes() {
        return cdktf.Fn.tolist(this.getListAttribute('enabled_policy_types'));
    }
    set enabledPolicyTypes(value) {
        this._enabledPolicyTypes = value;
    }
    resetEnabledPolicyTypes() {
        this._enabledPolicyTypes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledPolicyTypesInput() {
        return this._enabledPolicyTypes;
    }
    get featureSet() {
        return this.getStringAttribute('feature_set');
    }
    set featureSet(value) {
        this._featureSet = value;
    }
    resetFeatureSet() {
        this._featureSet = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get featureSetInput() {
        return this._featureSet;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // master_account_arn - computed: true, optional: false, required: false
    get masterAccountArn() {
        return this.getStringAttribute('master_account_arn');
    }
    // master_account_email - computed: true, optional: false, required: false
    get masterAccountEmail() {
        return this.getStringAttribute('master_account_email');
    }
    // master_account_id - computed: true, optional: false, required: false
    get masterAccountId() {
        return this.getStringAttribute('master_account_id');
    }
    get nonMasterAccounts() {
        return this._nonMasterAccounts;
    }
    get roots() {
        return this._roots;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            aws_service_access_principals: cdktf.listMapper(cdktf.stringToTerraform, false)(this._awsServiceAccessPrincipals),
            enabled_policy_types: cdktf.listMapper(cdktf.stringToTerraform, false)(this._enabledPolicyTypes),
            feature_set: cdktf.stringToTerraform(this._featureSet),
            id: cdktf.stringToTerraform(this._id),
        };
    }
}
exports.OrganizationsOrganization = OrganizationsOrganization;
_j = JSII_RTTI_SYMBOL_1;
OrganizationsOrganization[_j] = { fqn: "@cdktf/provider-aws.organizationsOrganization.OrganizationsOrganization", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
OrganizationsOrganization.tfResourceType = "aws_organizations_organization";
