// @cdktf/provider-aws/lib/transcribe-vocabulary/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TranscribeVocabulary = exports.TranscribeVocabularyTimeoutsOutputReference = exports.transcribeVocabularyTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function transcribeVocabularyTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.transcribeVocabularyTimeoutsToTerraform = transcribeVocabularyTimeoutsToTerraform;
class TranscribeVocabularyTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.TranscribeVocabularyTimeoutsOutputReference = TranscribeVocabularyTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
TranscribeVocabularyTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.transcribeVocabulary.TranscribeVocabularyTimeoutsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/transcribe_vocabulary aws_transcribe_vocabulary}
*/
class TranscribeVocabulary extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/transcribe_vocabulary aws_transcribe_vocabulary} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options TranscribeVocabularyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_transcribe_vocabulary',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new TranscribeVocabularyTimeoutsOutputReference(this, "timeouts");
        this._id = config.id;
        this._languageCode = config.languageCode;
        this._phrases = config.phrases;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._vocabularyFileUri = config.vocabularyFileUri;
        this._vocabularyName = config.vocabularyName;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // download_uri - computed: true, optional: false, required: false
    get downloadUri() {
        return this.getStringAttribute('download_uri');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get languageCode() {
        return this.getStringAttribute('language_code');
    }
    set languageCode(value) {
        this._languageCode = value;
    }
    // Temporarily expose input value. Use with caution.
    get languageCodeInput() {
        return this._languageCode;
    }
    get phrases() {
        return this.getListAttribute('phrases');
    }
    set phrases(value) {
        this._phrases = value;
    }
    resetPhrases() {
        this._phrases = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get phrasesInput() {
        return this._phrases;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get vocabularyFileUri() {
        return this.getStringAttribute('vocabulary_file_uri');
    }
    set vocabularyFileUri(value) {
        this._vocabularyFileUri = value;
    }
    resetVocabularyFileUri() {
        this._vocabularyFileUri = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vocabularyFileUriInput() {
        return this._vocabularyFileUri;
    }
    get vocabularyName() {
        return this.getStringAttribute('vocabulary_name');
    }
    set vocabularyName(value) {
        this._vocabularyName = value;
    }
    // Temporarily expose input value. Use with caution.
    get vocabularyNameInput() {
        return this._vocabularyName;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            language_code: cdktf.stringToTerraform(this._languageCode),
            phrases: cdktf.listMapper(cdktf.stringToTerraform, false)(this._phrases),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            vocabulary_file_uri: cdktf.stringToTerraform(this._vocabularyFileUri),
            vocabulary_name: cdktf.stringToTerraform(this._vocabularyName),
            timeouts: transcribeVocabularyTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.TranscribeVocabulary = TranscribeVocabulary;
_b = JSII_RTTI_SYMBOL_1;
TranscribeVocabulary[_b] = { fqn: "@cdktf/provider-aws.transcribeVocabulary.TranscribeVocabulary", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
TranscribeVocabulary.tfResourceType = "aws_transcribe_vocabulary";
