// @cdktf/provider-aws/lib/iot-thing-group/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IotThingGroup = exports.IotThingGroupPropertiesOutputReference = exports.iotThingGroupPropertiesToTerraform = exports.IotThingGroupPropertiesAttributePayloadOutputReference = exports.iotThingGroupPropertiesAttributePayloadToTerraform = exports.IotThingGroupMetadataList = exports.IotThingGroupMetadataOutputReference = exports.iotThingGroupMetadataToTerraform = exports.IotThingGroupMetadataRootToParentGroupsList = exports.IotThingGroupMetadataRootToParentGroupsOutputReference = exports.iotThingGroupMetadataRootToParentGroupsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function iotThingGroupMetadataRootToParentGroupsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.iotThingGroupMetadataRootToParentGroupsToTerraform = iotThingGroupMetadataRootToParentGroupsToTerraform;
class IotThingGroupMetadataRootToParentGroupsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // group_arn - computed: true, optional: false, required: false
    get groupArn() {
        return this.getStringAttribute('group_arn');
    }
    // group_name - computed: true, optional: false, required: false
    get groupName() {
        return this.getStringAttribute('group_name');
    }
}
exports.IotThingGroupMetadataRootToParentGroupsOutputReference = IotThingGroupMetadataRootToParentGroupsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
IotThingGroupMetadataRootToParentGroupsOutputReference[_a] = { fqn: "@cdktf/provider-aws.iotThingGroup.IotThingGroupMetadataRootToParentGroupsOutputReference", version: "10.0.12" };
class IotThingGroupMetadataRootToParentGroupsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new IotThingGroupMetadataRootToParentGroupsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.IotThingGroupMetadataRootToParentGroupsList = IotThingGroupMetadataRootToParentGroupsList;
_b = JSII_RTTI_SYMBOL_1;
IotThingGroupMetadataRootToParentGroupsList[_b] = { fqn: "@cdktf/provider-aws.iotThingGroup.IotThingGroupMetadataRootToParentGroupsList", version: "10.0.12" };
function iotThingGroupMetadataToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.iotThingGroupMetadataToTerraform = iotThingGroupMetadataToTerraform;
class IotThingGroupMetadataOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // root_to_parent_groups - computed: true, optional: false, required: false
        this._rootToParentGroups = new IotThingGroupMetadataRootToParentGroupsList(this, "root_to_parent_groups", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // creation_date - computed: true, optional: false, required: false
    get creationDate() {
        return this.getStringAttribute('creation_date');
    }
    // parent_group_name - computed: true, optional: false, required: false
    get parentGroupName() {
        return this.getStringAttribute('parent_group_name');
    }
    get rootToParentGroups() {
        return this._rootToParentGroups;
    }
}
exports.IotThingGroupMetadataOutputReference = IotThingGroupMetadataOutputReference;
_c = JSII_RTTI_SYMBOL_1;
IotThingGroupMetadataOutputReference[_c] = { fqn: "@cdktf/provider-aws.iotThingGroup.IotThingGroupMetadataOutputReference", version: "10.0.12" };
class IotThingGroupMetadataList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new IotThingGroupMetadataOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.IotThingGroupMetadataList = IotThingGroupMetadataList;
_d = JSII_RTTI_SYMBOL_1;
IotThingGroupMetadataList[_d] = { fqn: "@cdktf/provider-aws.iotThingGroup.IotThingGroupMetadataList", version: "10.0.12" };
function iotThingGroupPropertiesAttributePayloadToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        attributes: cdktf.hashMapper(cdktf.stringToTerraform)(struct.attributes),
    };
}
exports.iotThingGroupPropertiesAttributePayloadToTerraform = iotThingGroupPropertiesAttributePayloadToTerraform;
class IotThingGroupPropertiesAttributePayloadOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._attributes !== undefined) {
            hasAnyValues = true;
            internalValueResult.attributes = this._attributes;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._attributes = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._attributes = value.attributes;
        }
    }
    get attributes() {
        return this.getStringMapAttribute('attributes');
    }
    set attributes(value) {
        this._attributes = value;
    }
    resetAttributes() {
        this._attributes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get attributesInput() {
        return this._attributes;
    }
}
exports.IotThingGroupPropertiesAttributePayloadOutputReference = IotThingGroupPropertiesAttributePayloadOutputReference;
_e = JSII_RTTI_SYMBOL_1;
IotThingGroupPropertiesAttributePayloadOutputReference[_e] = { fqn: "@cdktf/provider-aws.iotThingGroup.IotThingGroupPropertiesAttributePayloadOutputReference", version: "10.0.12" };
function iotThingGroupPropertiesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        description: cdktf.stringToTerraform(struct.description),
        attribute_payload: iotThingGroupPropertiesAttributePayloadToTerraform(struct.attributePayload),
    };
}
exports.iotThingGroupPropertiesToTerraform = iotThingGroupPropertiesToTerraform;
class IotThingGroupPropertiesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // attribute_payload - computed: false, optional: true, required: false
        this._attributePayload = new IotThingGroupPropertiesAttributePayloadOutputReference(this, "attribute_payload");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._attributePayload?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.attributePayload = this._attributePayload?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._description = undefined;
            this._attributePayload.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._description = value.description;
            this._attributePayload.internalValue = value.attributePayload;
        }
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get attributePayload() {
        return this._attributePayload;
    }
    putAttributePayload(value) {
        this._attributePayload.internalValue = value;
    }
    resetAttributePayload() {
        this._attributePayload.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get attributePayloadInput() {
        return this._attributePayload.internalValue;
    }
}
exports.IotThingGroupPropertiesOutputReference = IotThingGroupPropertiesOutputReference;
_f = JSII_RTTI_SYMBOL_1;
IotThingGroupPropertiesOutputReference[_f] = { fqn: "@cdktf/provider-aws.iotThingGroup.IotThingGroupPropertiesOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/iot_thing_group aws_iot_thing_group}
*/
class IotThingGroup extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/iot_thing_group aws_iot_thing_group} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options IotThingGroupConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iot_thing_group',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // metadata - computed: true, optional: false, required: false
        this._metadata = new IotThingGroupMetadataList(this, "metadata", false);
        // properties - computed: false, optional: true, required: false
        this._properties = new IotThingGroupPropertiesOutputReference(this, "properties");
        this._id = config.id;
        this._name = config.name;
        this._parentGroupName = config.parentGroupName;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._properties.internalValue = config.properties;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get metadata() {
        return this._metadata;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get parentGroupName() {
        return this.getStringAttribute('parent_group_name');
    }
    set parentGroupName(value) {
        this._parentGroupName = value;
    }
    resetParentGroupName() {
        this._parentGroupName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parentGroupNameInput() {
        return this._parentGroupName;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getNumberAttribute('version');
    }
    get properties() {
        return this._properties;
    }
    putProperties(value) {
        this._properties.internalValue = value;
    }
    resetProperties() {
        this._properties.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get propertiesInput() {
        return this._properties.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            parent_group_name: cdktf.stringToTerraform(this._parentGroupName),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            properties: iotThingGroupPropertiesToTerraform(this._properties.internalValue),
        };
    }
}
exports.IotThingGroup = IotThingGroup;
_g = JSII_RTTI_SYMBOL_1;
IotThingGroup[_g] = { fqn: "@cdktf/provider-aws.iotThingGroup.IotThingGroup", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
IotThingGroup.tfResourceType = "aws_iot_thing_group";
