// @cdktf/provider-aws/lib/msk-serverless-cluster/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MskServerlessCluster = exports.MskServerlessClusterVpcConfigList = exports.MskServerlessClusterVpcConfigOutputReference = exports.mskServerlessClusterVpcConfigToTerraform = exports.MskServerlessClusterTimeoutsOutputReference = exports.mskServerlessClusterTimeoutsToTerraform = exports.MskServerlessClusterClientAuthenticationOutputReference = exports.mskServerlessClusterClientAuthenticationToTerraform = exports.MskServerlessClusterClientAuthenticationSaslOutputReference = exports.mskServerlessClusterClientAuthenticationSaslToTerraform = exports.MskServerlessClusterClientAuthenticationSaslIamOutputReference = exports.mskServerlessClusterClientAuthenticationSaslIamToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function mskServerlessClusterClientAuthenticationSaslIamToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
    };
}
exports.mskServerlessClusterClientAuthenticationSaslIamToTerraform = mskServerlessClusterClientAuthenticationSaslIamToTerraform;
class MskServerlessClusterClientAuthenticationSaslIamOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
}
exports.MskServerlessClusterClientAuthenticationSaslIamOutputReference = MskServerlessClusterClientAuthenticationSaslIamOutputReference;
_a = JSII_RTTI_SYMBOL_1;
MskServerlessClusterClientAuthenticationSaslIamOutputReference[_a] = { fqn: "@cdktf/provider-aws.mskServerlessCluster.MskServerlessClusterClientAuthenticationSaslIamOutputReference", version: "10.0.12" };
function mskServerlessClusterClientAuthenticationSaslToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        iam: mskServerlessClusterClientAuthenticationSaslIamToTerraform(struct.iam),
    };
}
exports.mskServerlessClusterClientAuthenticationSaslToTerraform = mskServerlessClusterClientAuthenticationSaslToTerraform;
class MskServerlessClusterClientAuthenticationSaslOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // iam - computed: false, optional: false, required: true
        this._iam = new MskServerlessClusterClientAuthenticationSaslIamOutputReference(this, "iam");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._iam?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.iam = this._iam?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._iam.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._iam.internalValue = value.iam;
        }
    }
    get iam() {
        return this._iam;
    }
    putIam(value) {
        this._iam.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get iamInput() {
        return this._iam.internalValue;
    }
}
exports.MskServerlessClusterClientAuthenticationSaslOutputReference = MskServerlessClusterClientAuthenticationSaslOutputReference;
_b = JSII_RTTI_SYMBOL_1;
MskServerlessClusterClientAuthenticationSaslOutputReference[_b] = { fqn: "@cdktf/provider-aws.mskServerlessCluster.MskServerlessClusterClientAuthenticationSaslOutputReference", version: "10.0.12" };
function mskServerlessClusterClientAuthenticationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        sasl: mskServerlessClusterClientAuthenticationSaslToTerraform(struct.sasl),
    };
}
exports.mskServerlessClusterClientAuthenticationToTerraform = mskServerlessClusterClientAuthenticationToTerraform;
class MskServerlessClusterClientAuthenticationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // sasl - computed: false, optional: false, required: true
        this._sasl = new MskServerlessClusterClientAuthenticationSaslOutputReference(this, "sasl");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._sasl?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.sasl = this._sasl?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._sasl.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._sasl.internalValue = value.sasl;
        }
    }
    get sasl() {
        return this._sasl;
    }
    putSasl(value) {
        this._sasl.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get saslInput() {
        return this._sasl.internalValue;
    }
}
exports.MskServerlessClusterClientAuthenticationOutputReference = MskServerlessClusterClientAuthenticationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
MskServerlessClusterClientAuthenticationOutputReference[_c] = { fqn: "@cdktf/provider-aws.mskServerlessCluster.MskServerlessClusterClientAuthenticationOutputReference", version: "10.0.12" };
function mskServerlessClusterTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.mskServerlessClusterTimeoutsToTerraform = mskServerlessClusterTimeoutsToTerraform;
class MskServerlessClusterTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
}
exports.MskServerlessClusterTimeoutsOutputReference = MskServerlessClusterTimeoutsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
MskServerlessClusterTimeoutsOutputReference[_d] = { fqn: "@cdktf/provider-aws.mskServerlessCluster.MskServerlessClusterTimeoutsOutputReference", version: "10.0.12" };
function mskServerlessClusterVpcConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.securityGroupIds),
        subnet_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.subnetIds),
    };
}
exports.mskServerlessClusterVpcConfigToTerraform = mskServerlessClusterVpcConfigToTerraform;
class MskServerlessClusterVpcConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._securityGroupIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroupIds = this._securityGroupIds;
        }
        if (this._subnetIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetIds = this._subnetIds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._securityGroupIds = undefined;
            this._subnetIds = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._securityGroupIds = value.securityGroupIds;
            this._subnetIds = value.subnetIds;
        }
    }
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    resetSecurityGroupIds() {
        this._securityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    get subnetIds() {
        return cdktf.Fn.tolist(this.getListAttribute('subnet_ids'));
    }
    set subnetIds(value) {
        this._subnetIds = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdsInput() {
        return this._subnetIds;
    }
}
exports.MskServerlessClusterVpcConfigOutputReference = MskServerlessClusterVpcConfigOutputReference;
_e = JSII_RTTI_SYMBOL_1;
MskServerlessClusterVpcConfigOutputReference[_e] = { fqn: "@cdktf/provider-aws.mskServerlessCluster.MskServerlessClusterVpcConfigOutputReference", version: "10.0.12" };
class MskServerlessClusterVpcConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new MskServerlessClusterVpcConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.MskServerlessClusterVpcConfigList = MskServerlessClusterVpcConfigList;
_f = JSII_RTTI_SYMBOL_1;
MskServerlessClusterVpcConfigList[_f] = { fqn: "@cdktf/provider-aws.mskServerlessCluster.MskServerlessClusterVpcConfigList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/msk_serverless_cluster aws_msk_serverless_cluster}
*/
class MskServerlessCluster extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/msk_serverless_cluster aws_msk_serverless_cluster} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options MskServerlessClusterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_msk_serverless_cluster',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // client_authentication - computed: false, optional: false, required: true
        this._clientAuthentication = new MskServerlessClusterClientAuthenticationOutputReference(this, "client_authentication");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new MskServerlessClusterTimeoutsOutputReference(this, "timeouts");
        // vpc_config - computed: false, optional: false, required: true
        this._vpcConfig = new MskServerlessClusterVpcConfigList(this, "vpc_config", false);
        this._clusterName = config.clusterName;
        this._id = config.id;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._clientAuthentication.internalValue = config.clientAuthentication;
        this._timeouts.internalValue = config.timeouts;
        this._vpcConfig.internalValue = config.vpcConfig;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get clusterName() {
        return this.getStringAttribute('cluster_name');
    }
    set clusterName(value) {
        this._clusterName = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterNameInput() {
        return this._clusterName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get clientAuthentication() {
        return this._clientAuthentication;
    }
    putClientAuthentication(value) {
        this._clientAuthentication.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get clientAuthenticationInput() {
        return this._clientAuthentication.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get vpcConfig() {
        return this._vpcConfig;
    }
    putVpcConfig(value) {
        this._vpcConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get vpcConfigInput() {
        return this._vpcConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_name: cdktf.stringToTerraform(this._clusterName),
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            client_authentication: mskServerlessClusterClientAuthenticationToTerraform(this._clientAuthentication.internalValue),
            timeouts: mskServerlessClusterTimeoutsToTerraform(this._timeouts.internalValue),
            vpc_config: cdktf.listMapper(mskServerlessClusterVpcConfigToTerraform, true)(this._vpcConfig.internalValue),
        };
    }
}
exports.MskServerlessCluster = MskServerlessCluster;
_g = JSII_RTTI_SYMBOL_1;
MskServerlessCluster[_g] = { fqn: "@cdktf/provider-aws.mskServerlessCluster.MskServerlessCluster", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
MskServerlessCluster.tfResourceType = "aws_msk_serverless_cluster";
