// @cdktf/provider-aws/lib/kms-key/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.KmsKey = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/kms_key aws_kms_key}
*/
class KmsKey extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/kms_key aws_kms_key} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options KmsKeyConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_kms_key',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._bypassPolicyLockoutSafetyCheck = config.bypassPolicyLockoutSafetyCheck;
        this._customKeyStoreId = config.customKeyStoreId;
        this._customerMasterKeySpec = config.customerMasterKeySpec;
        this._deletionWindowInDays = config.deletionWindowInDays;
        this._description = config.description;
        this._enableKeyRotation = config.enableKeyRotation;
        this._id = config.id;
        this._isEnabled = config.isEnabled;
        this._keyUsage = config.keyUsage;
        this._multiRegion = config.multiRegion;
        this._policy = config.policy;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get bypassPolicyLockoutSafetyCheck() {
        return this.getBooleanAttribute('bypass_policy_lockout_safety_check');
    }
    set bypassPolicyLockoutSafetyCheck(value) {
        this._bypassPolicyLockoutSafetyCheck = value;
    }
    resetBypassPolicyLockoutSafetyCheck() {
        this._bypassPolicyLockoutSafetyCheck = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bypassPolicyLockoutSafetyCheckInput() {
        return this._bypassPolicyLockoutSafetyCheck;
    }
    get customKeyStoreId() {
        return this.getStringAttribute('custom_key_store_id');
    }
    set customKeyStoreId(value) {
        this._customKeyStoreId = value;
    }
    resetCustomKeyStoreId() {
        this._customKeyStoreId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get customKeyStoreIdInput() {
        return this._customKeyStoreId;
    }
    get customerMasterKeySpec() {
        return this.getStringAttribute('customer_master_key_spec');
    }
    set customerMasterKeySpec(value) {
        this._customerMasterKeySpec = value;
    }
    resetCustomerMasterKeySpec() {
        this._customerMasterKeySpec = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get customerMasterKeySpecInput() {
        return this._customerMasterKeySpec;
    }
    get deletionWindowInDays() {
        return this.getNumberAttribute('deletion_window_in_days');
    }
    set deletionWindowInDays(value) {
        this._deletionWindowInDays = value;
    }
    resetDeletionWindowInDays() {
        this._deletionWindowInDays = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deletionWindowInDaysInput() {
        return this._deletionWindowInDays;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get enableKeyRotation() {
        return this.getBooleanAttribute('enable_key_rotation');
    }
    set enableKeyRotation(value) {
        this._enableKeyRotation = value;
    }
    resetEnableKeyRotation() {
        this._enableKeyRotation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableKeyRotationInput() {
        return this._enableKeyRotation;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get isEnabled() {
        return this.getBooleanAttribute('is_enabled');
    }
    set isEnabled(value) {
        this._isEnabled = value;
    }
    resetIsEnabled() {
        this._isEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get isEnabledInput() {
        return this._isEnabled;
    }
    // key_id - computed: true, optional: false, required: false
    get keyId() {
        return this.getStringAttribute('key_id');
    }
    get keyUsage() {
        return this.getStringAttribute('key_usage');
    }
    set keyUsage(value) {
        this._keyUsage = value;
    }
    resetKeyUsage() {
        this._keyUsage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyUsageInput() {
        return this._keyUsage;
    }
    get multiRegion() {
        return this.getBooleanAttribute('multi_region');
    }
    set multiRegion(value) {
        this._multiRegion = value;
    }
    resetMultiRegion() {
        this._multiRegion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get multiRegionInput() {
        return this._multiRegion;
    }
    get policy() {
        return this.getStringAttribute('policy');
    }
    set policy(value) {
        this._policy = value;
    }
    resetPolicy() {
        this._policy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get policyInput() {
        return this._policy;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            bypass_policy_lockout_safety_check: cdktf.booleanToTerraform(this._bypassPolicyLockoutSafetyCheck),
            custom_key_store_id: cdktf.stringToTerraform(this._customKeyStoreId),
            customer_master_key_spec: cdktf.stringToTerraform(this._customerMasterKeySpec),
            deletion_window_in_days: cdktf.numberToTerraform(this._deletionWindowInDays),
            description: cdktf.stringToTerraform(this._description),
            enable_key_rotation: cdktf.booleanToTerraform(this._enableKeyRotation),
            id: cdktf.stringToTerraform(this._id),
            is_enabled: cdktf.booleanToTerraform(this._isEnabled),
            key_usage: cdktf.stringToTerraform(this._keyUsage),
            multi_region: cdktf.booleanToTerraform(this._multiRegion),
            policy: cdktf.stringToTerraform(this._policy),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
        };
    }
}
exports.KmsKey = KmsKey;
_a = JSII_RTTI_SYMBOL_1;
KmsKey[_a] = { fqn: "@cdktf/provider-aws.kmsKey.KmsKey", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
KmsKey.tfResourceType = "aws_kms_key";
