// @cdktf/provider-aws/lib/data-aws-appmesh-virtual-service/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsAppmeshVirtualService = exports.DataAwsAppmeshVirtualServiceSpecList = exports.DataAwsAppmeshVirtualServiceSpecOutputReference = exports.dataAwsAppmeshVirtualServiceSpecToTerraform = exports.DataAwsAppmeshVirtualServiceSpecProviderList = exports.DataAwsAppmeshVirtualServiceSpecProviderOutputReference = exports.dataAwsAppmeshVirtualServiceSpecProviderToTerraform = exports.DataAwsAppmeshVirtualServiceSpecProviderVirtualRouterList = exports.DataAwsAppmeshVirtualServiceSpecProviderVirtualRouterOutputReference = exports.dataAwsAppmeshVirtualServiceSpecProviderVirtualRouterToTerraform = exports.DataAwsAppmeshVirtualServiceSpecProviderVirtualNodeList = exports.DataAwsAppmeshVirtualServiceSpecProviderVirtualNodeOutputReference = exports.dataAwsAppmeshVirtualServiceSpecProviderVirtualNodeToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsAppmeshVirtualServiceSpecProviderVirtualNodeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsAppmeshVirtualServiceSpecProviderVirtualNodeToTerraform = dataAwsAppmeshVirtualServiceSpecProviderVirtualNodeToTerraform;
class DataAwsAppmeshVirtualServiceSpecProviderVirtualNodeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // virtual_node_name - computed: true, optional: false, required: false
    get virtualNodeName() {
        return this.getStringAttribute('virtual_node_name');
    }
}
exports.DataAwsAppmeshVirtualServiceSpecProviderVirtualNodeOutputReference = DataAwsAppmeshVirtualServiceSpecProviderVirtualNodeOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsAppmeshVirtualServiceSpecProviderVirtualNodeOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsAppmeshVirtualService.DataAwsAppmeshVirtualServiceSpecProviderVirtualNodeOutputReference", version: "10.0.12" };
class DataAwsAppmeshVirtualServiceSpecProviderVirtualNodeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsAppmeshVirtualServiceSpecProviderVirtualNodeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsAppmeshVirtualServiceSpecProviderVirtualNodeList = DataAwsAppmeshVirtualServiceSpecProviderVirtualNodeList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsAppmeshVirtualServiceSpecProviderVirtualNodeList[_b] = { fqn: "@cdktf/provider-aws.dataAwsAppmeshVirtualService.DataAwsAppmeshVirtualServiceSpecProviderVirtualNodeList", version: "10.0.12" };
function dataAwsAppmeshVirtualServiceSpecProviderVirtualRouterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsAppmeshVirtualServiceSpecProviderVirtualRouterToTerraform = dataAwsAppmeshVirtualServiceSpecProviderVirtualRouterToTerraform;
class DataAwsAppmeshVirtualServiceSpecProviderVirtualRouterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // virtual_router_name - computed: true, optional: false, required: false
    get virtualRouterName() {
        return this.getStringAttribute('virtual_router_name');
    }
}
exports.DataAwsAppmeshVirtualServiceSpecProviderVirtualRouterOutputReference = DataAwsAppmeshVirtualServiceSpecProviderVirtualRouterOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsAppmeshVirtualServiceSpecProviderVirtualRouterOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsAppmeshVirtualService.DataAwsAppmeshVirtualServiceSpecProviderVirtualRouterOutputReference", version: "10.0.12" };
class DataAwsAppmeshVirtualServiceSpecProviderVirtualRouterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsAppmeshVirtualServiceSpecProviderVirtualRouterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsAppmeshVirtualServiceSpecProviderVirtualRouterList = DataAwsAppmeshVirtualServiceSpecProviderVirtualRouterList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsAppmeshVirtualServiceSpecProviderVirtualRouterList[_d] = { fqn: "@cdktf/provider-aws.dataAwsAppmeshVirtualService.DataAwsAppmeshVirtualServiceSpecProviderVirtualRouterList", version: "10.0.12" };
function dataAwsAppmeshVirtualServiceSpecProviderToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsAppmeshVirtualServiceSpecProviderToTerraform = dataAwsAppmeshVirtualServiceSpecProviderToTerraform;
class DataAwsAppmeshVirtualServiceSpecProviderOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // virtual_node - computed: true, optional: false, required: false
        this._virtualNode = new DataAwsAppmeshVirtualServiceSpecProviderVirtualNodeList(this, "virtual_node", false);
        // virtual_router - computed: true, optional: false, required: false
        this._virtualRouter = new DataAwsAppmeshVirtualServiceSpecProviderVirtualRouterList(this, "virtual_router", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get virtualNode() {
        return this._virtualNode;
    }
    get virtualRouter() {
        return this._virtualRouter;
    }
}
exports.DataAwsAppmeshVirtualServiceSpecProviderOutputReference = DataAwsAppmeshVirtualServiceSpecProviderOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsAppmeshVirtualServiceSpecProviderOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsAppmeshVirtualService.DataAwsAppmeshVirtualServiceSpecProviderOutputReference", version: "10.0.12" };
class DataAwsAppmeshVirtualServiceSpecProviderList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsAppmeshVirtualServiceSpecProviderOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsAppmeshVirtualServiceSpecProviderList = DataAwsAppmeshVirtualServiceSpecProviderList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsAppmeshVirtualServiceSpecProviderList[_f] = { fqn: "@cdktf/provider-aws.dataAwsAppmeshVirtualService.DataAwsAppmeshVirtualServiceSpecProviderList", version: "10.0.12" };
function dataAwsAppmeshVirtualServiceSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsAppmeshVirtualServiceSpecToTerraform = dataAwsAppmeshVirtualServiceSpecToTerraform;
class DataAwsAppmeshVirtualServiceSpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // provider - computed: true, optional: false, required: false
        this._provider = new DataAwsAppmeshVirtualServiceSpecProviderList(this, "provider", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get provider() {
        return this._provider;
    }
}
exports.DataAwsAppmeshVirtualServiceSpecOutputReference = DataAwsAppmeshVirtualServiceSpecOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsAppmeshVirtualServiceSpecOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsAppmeshVirtualService.DataAwsAppmeshVirtualServiceSpecOutputReference", version: "10.0.12" };
class DataAwsAppmeshVirtualServiceSpecList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsAppmeshVirtualServiceSpecOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsAppmeshVirtualServiceSpecList = DataAwsAppmeshVirtualServiceSpecList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsAppmeshVirtualServiceSpecList[_h] = { fqn: "@cdktf/provider-aws.dataAwsAppmeshVirtualService.DataAwsAppmeshVirtualServiceSpecList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/appmesh_virtual_service aws_appmesh_virtual_service}
*/
class DataAwsAppmeshVirtualService extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/appmesh_virtual_service aws_appmesh_virtual_service} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsAppmeshVirtualServiceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_appmesh_virtual_service',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // spec - computed: true, optional: false, required: false
        this._spec = new DataAwsAppmeshVirtualServiceSpecList(this, "spec", false);
        this._id = config.id;
        this._meshName = config.meshName;
        this._meshOwner = config.meshOwner;
        this._name = config.name;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // created_date - computed: true, optional: false, required: false
    get createdDate() {
        return this.getStringAttribute('created_date');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // last_updated_date - computed: true, optional: false, required: false
    get lastUpdatedDate() {
        return this.getStringAttribute('last_updated_date');
    }
    get meshName() {
        return this.getStringAttribute('mesh_name');
    }
    set meshName(value) {
        this._meshName = value;
    }
    // Temporarily expose input value. Use with caution.
    get meshNameInput() {
        return this._meshName;
    }
    get meshOwner() {
        return this.getStringAttribute('mesh_owner');
    }
    set meshOwner(value) {
        this._meshOwner = value;
    }
    resetMeshOwner() {
        this._meshOwner = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get meshOwnerInput() {
        return this._meshOwner;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // resource_owner - computed: true, optional: false, required: false
    get resourceOwner() {
        return this.getStringAttribute('resource_owner');
    }
    get spec() {
        return this._spec;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            mesh_name: cdktf.stringToTerraform(this._meshName),
            mesh_owner: cdktf.stringToTerraform(this._meshOwner),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
}
exports.DataAwsAppmeshVirtualService = DataAwsAppmeshVirtualService;
_j = JSII_RTTI_SYMBOL_1;
DataAwsAppmeshVirtualService[_j] = { fqn: "@cdktf/provider-aws.dataAwsAppmeshVirtualService.DataAwsAppmeshVirtualService", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsAppmeshVirtualService.tfResourceType = "aws_appmesh_virtual_service";
