// @cdktf/provider-aws/lib/appstream-user-stack-association/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppstreamUserStackAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/appstream_user_stack_association aws_appstream_user_stack_association}
*/
class AppstreamUserStackAssociation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/appstream_user_stack_association aws_appstream_user_stack_association} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options AppstreamUserStackAssociationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_appstream_user_stack_association',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._authenticationType = config.authenticationType;
        this._id = config.id;
        this._sendEmailNotification = config.sendEmailNotification;
        this._stackName = config.stackName;
        this._userName = config.userName;
    }
    get authenticationType() {
        return this.getStringAttribute('authentication_type');
    }
    set authenticationType(value) {
        this._authenticationType = value;
    }
    // Temporarily expose input value. Use with caution.
    get authenticationTypeInput() {
        return this._authenticationType;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get sendEmailNotification() {
        return this.getBooleanAttribute('send_email_notification');
    }
    set sendEmailNotification(value) {
        this._sendEmailNotification = value;
    }
    resetSendEmailNotification() {
        this._sendEmailNotification = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sendEmailNotificationInput() {
        return this._sendEmailNotification;
    }
    get stackName() {
        return this.getStringAttribute('stack_name');
    }
    set stackName(value) {
        this._stackName = value;
    }
    // Temporarily expose input value. Use with caution.
    get stackNameInput() {
        return this._stackName;
    }
    get userName() {
        return this.getStringAttribute('user_name');
    }
    set userName(value) {
        this._userName = value;
    }
    // Temporarily expose input value. Use with caution.
    get userNameInput() {
        return this._userName;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            authentication_type: cdktf.stringToTerraform(this._authenticationType),
            id: cdktf.stringToTerraform(this._id),
            send_email_notification: cdktf.booleanToTerraform(this._sendEmailNotification),
            stack_name: cdktf.stringToTerraform(this._stackName),
            user_name: cdktf.stringToTerraform(this._userName),
        };
    }
}
exports.AppstreamUserStackAssociation = AppstreamUserStackAssociation;
_a = JSII_RTTI_SYMBOL_1;
AppstreamUserStackAssociation[_a] = { fqn: "@cdktf/provider-aws.appstreamUserStackAssociation.AppstreamUserStackAssociation", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
AppstreamUserStackAssociation.tfResourceType = "aws_appstream_user_stack_association";
