// @cdktf/provider-aws/lib/globalaccelerator-listener/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlobalacceleratorListener = exports.GlobalacceleratorListenerTimeoutsOutputReference = exports.globalacceleratorListenerTimeoutsToTerraform = exports.GlobalacceleratorListenerPortRangeList = exports.GlobalacceleratorListenerPortRangeOutputReference = exports.globalacceleratorListenerPortRangeToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function globalacceleratorListenerPortRangeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        from_port: cdktf.numberToTerraform(struct.fromPort),
        to_port: cdktf.numberToTerraform(struct.toPort),
    };
}
exports.globalacceleratorListenerPortRangeToTerraform = globalacceleratorListenerPortRangeToTerraform;
class GlobalacceleratorListenerPortRangeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._fromPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.fromPort = this._fromPort;
        }
        if (this._toPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.toPort = this._toPort;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._fromPort = undefined;
            this._toPort = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._fromPort = value.fromPort;
            this._toPort = value.toPort;
        }
    }
    get fromPort() {
        return this.getNumberAttribute('from_port');
    }
    set fromPort(value) {
        this._fromPort = value;
    }
    resetFromPort() {
        this._fromPort = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fromPortInput() {
        return this._fromPort;
    }
    get toPort() {
        return this.getNumberAttribute('to_port');
    }
    set toPort(value) {
        this._toPort = value;
    }
    resetToPort() {
        this._toPort = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get toPortInput() {
        return this._toPort;
    }
}
exports.GlobalacceleratorListenerPortRangeOutputReference = GlobalacceleratorListenerPortRangeOutputReference;
_a = JSII_RTTI_SYMBOL_1;
GlobalacceleratorListenerPortRangeOutputReference[_a] = { fqn: "@cdktf/provider-aws.globalacceleratorListener.GlobalacceleratorListenerPortRangeOutputReference", version: "10.0.12" };
class GlobalacceleratorListenerPortRangeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new GlobalacceleratorListenerPortRangeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.GlobalacceleratorListenerPortRangeList = GlobalacceleratorListenerPortRangeList;
_b = JSII_RTTI_SYMBOL_1;
GlobalacceleratorListenerPortRangeList[_b] = { fqn: "@cdktf/provider-aws.globalacceleratorListener.GlobalacceleratorListenerPortRangeList", version: "10.0.12" };
function globalacceleratorListenerTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.globalacceleratorListenerTimeoutsToTerraform = globalacceleratorListenerTimeoutsToTerraform;
class GlobalacceleratorListenerTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.GlobalacceleratorListenerTimeoutsOutputReference = GlobalacceleratorListenerTimeoutsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
GlobalacceleratorListenerTimeoutsOutputReference[_c] = { fqn: "@cdktf/provider-aws.globalacceleratorListener.GlobalacceleratorListenerTimeoutsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/globalaccelerator_listener aws_globalaccelerator_listener}
*/
class GlobalacceleratorListener extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/globalaccelerator_listener aws_globalaccelerator_listener} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options GlobalacceleratorListenerConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_globalaccelerator_listener',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // port_range - computed: false, optional: false, required: true
        this._portRange = new GlobalacceleratorListenerPortRangeList(this, "port_range", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new GlobalacceleratorListenerTimeoutsOutputReference(this, "timeouts");
        this._acceleratorArn = config.acceleratorArn;
        this._clientAffinity = config.clientAffinity;
        this._id = config.id;
        this._protocol = config.protocol;
        this._portRange.internalValue = config.portRange;
        this._timeouts.internalValue = config.timeouts;
    }
    get acceleratorArn() {
        return this.getStringAttribute('accelerator_arn');
    }
    set acceleratorArn(value) {
        this._acceleratorArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get acceleratorArnInput() {
        return this._acceleratorArn;
    }
    get clientAffinity() {
        return this.getStringAttribute('client_affinity');
    }
    set clientAffinity(value) {
        this._clientAffinity = value;
    }
    resetClientAffinity() {
        this._clientAffinity = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clientAffinityInput() {
        return this._clientAffinity;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    set protocol(value) {
        this._protocol = value;
    }
    // Temporarily expose input value. Use with caution.
    get protocolInput() {
        return this._protocol;
    }
    get portRange() {
        return this._portRange;
    }
    putPortRange(value) {
        this._portRange.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get portRangeInput() {
        return this._portRange.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            accelerator_arn: cdktf.stringToTerraform(this._acceleratorArn),
            client_affinity: cdktf.stringToTerraform(this._clientAffinity),
            id: cdktf.stringToTerraform(this._id),
            protocol: cdktf.stringToTerraform(this._protocol),
            port_range: cdktf.listMapper(globalacceleratorListenerPortRangeToTerraform, true)(this._portRange.internalValue),
            timeouts: globalacceleratorListenerTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.GlobalacceleratorListener = GlobalacceleratorListener;
_d = JSII_RTTI_SYMBOL_1;
GlobalacceleratorListener[_d] = { fqn: "@cdktf/provider-aws.globalacceleratorListener.GlobalacceleratorListener", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
GlobalacceleratorListener.tfResourceType = "aws_globalaccelerator_listener";
