// @cdktf/provider-aws/lib/data-aws-lakeformation-data-lake-settings/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsLakeformationDataLakeSettings = exports.DataAwsLakeformationDataLakeSettingsCreateTableDefaultPermissionsList = exports.DataAwsLakeformationDataLakeSettingsCreateTableDefaultPermissionsOutputReference = exports.dataAwsLakeformationDataLakeSettingsCreateTableDefaultPermissionsToTerraform = exports.DataAwsLakeformationDataLakeSettingsCreateDatabaseDefaultPermissionsList = exports.DataAwsLakeformationDataLakeSettingsCreateDatabaseDefaultPermissionsOutputReference = exports.dataAwsLakeformationDataLakeSettingsCreateDatabaseDefaultPermissionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsLakeformationDataLakeSettingsCreateDatabaseDefaultPermissionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLakeformationDataLakeSettingsCreateDatabaseDefaultPermissionsToTerraform = dataAwsLakeformationDataLakeSettingsCreateDatabaseDefaultPermissionsToTerraform;
class DataAwsLakeformationDataLakeSettingsCreateDatabaseDefaultPermissionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // permissions - computed: true, optional: false, required: false
    get permissions() {
        return cdktf.Fn.tolist(this.getListAttribute('permissions'));
    }
    // principal - computed: true, optional: false, required: false
    get principal() {
        return this.getStringAttribute('principal');
    }
}
exports.DataAwsLakeformationDataLakeSettingsCreateDatabaseDefaultPermissionsOutputReference = DataAwsLakeformationDataLakeSettingsCreateDatabaseDefaultPermissionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsLakeformationDataLakeSettingsCreateDatabaseDefaultPermissionsOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsLakeformationDataLakeSettings.DataAwsLakeformationDataLakeSettingsCreateDatabaseDefaultPermissionsOutputReference", version: "10.0.12" };
class DataAwsLakeformationDataLakeSettingsCreateDatabaseDefaultPermissionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLakeformationDataLakeSettingsCreateDatabaseDefaultPermissionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLakeformationDataLakeSettingsCreateDatabaseDefaultPermissionsList = DataAwsLakeformationDataLakeSettingsCreateDatabaseDefaultPermissionsList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsLakeformationDataLakeSettingsCreateDatabaseDefaultPermissionsList[_b] = { fqn: "@cdktf/provider-aws.dataAwsLakeformationDataLakeSettings.DataAwsLakeformationDataLakeSettingsCreateDatabaseDefaultPermissionsList", version: "10.0.12" };
function dataAwsLakeformationDataLakeSettingsCreateTableDefaultPermissionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLakeformationDataLakeSettingsCreateTableDefaultPermissionsToTerraform = dataAwsLakeformationDataLakeSettingsCreateTableDefaultPermissionsToTerraform;
class DataAwsLakeformationDataLakeSettingsCreateTableDefaultPermissionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // permissions - computed: true, optional: false, required: false
    get permissions() {
        return cdktf.Fn.tolist(this.getListAttribute('permissions'));
    }
    // principal - computed: true, optional: false, required: false
    get principal() {
        return this.getStringAttribute('principal');
    }
}
exports.DataAwsLakeformationDataLakeSettingsCreateTableDefaultPermissionsOutputReference = DataAwsLakeformationDataLakeSettingsCreateTableDefaultPermissionsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsLakeformationDataLakeSettingsCreateTableDefaultPermissionsOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsLakeformationDataLakeSettings.DataAwsLakeformationDataLakeSettingsCreateTableDefaultPermissionsOutputReference", version: "10.0.12" };
class DataAwsLakeformationDataLakeSettingsCreateTableDefaultPermissionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLakeformationDataLakeSettingsCreateTableDefaultPermissionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLakeformationDataLakeSettingsCreateTableDefaultPermissionsList = DataAwsLakeformationDataLakeSettingsCreateTableDefaultPermissionsList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsLakeformationDataLakeSettingsCreateTableDefaultPermissionsList[_d] = { fqn: "@cdktf/provider-aws.dataAwsLakeformationDataLakeSettings.DataAwsLakeformationDataLakeSettingsCreateTableDefaultPermissionsList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/lakeformation_data_lake_settings aws_lakeformation_data_lake_settings}
*/
class DataAwsLakeformationDataLakeSettings extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/lakeformation_data_lake_settings aws_lakeformation_data_lake_settings} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsLakeformationDataLakeSettingsConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_lakeformation_data_lake_settings',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // create_database_default_permissions - computed: true, optional: false, required: false
        this._createDatabaseDefaultPermissions = new DataAwsLakeformationDataLakeSettingsCreateDatabaseDefaultPermissionsList(this, "create_database_default_permissions", false);
        // create_table_default_permissions - computed: true, optional: false, required: false
        this._createTableDefaultPermissions = new DataAwsLakeformationDataLakeSettingsCreateTableDefaultPermissionsList(this, "create_table_default_permissions", false);
        this._catalogId = config.catalogId;
        this._id = config.id;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // admins - computed: true, optional: false, required: false
    get admins() {
        return cdktf.Fn.tolist(this.getListAttribute('admins'));
    }
    get catalogId() {
        return this.getStringAttribute('catalog_id');
    }
    set catalogId(value) {
        this._catalogId = value;
    }
    resetCatalogId() {
        this._catalogId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get catalogIdInput() {
        return this._catalogId;
    }
    get createDatabaseDefaultPermissions() {
        return this._createDatabaseDefaultPermissions;
    }
    get createTableDefaultPermissions() {
        return this._createTableDefaultPermissions;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // trusted_resource_owners - computed: true, optional: false, required: false
    get trustedResourceOwners() {
        return this.getListAttribute('trusted_resource_owners');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            catalog_id: cdktf.stringToTerraform(this._catalogId),
            id: cdktf.stringToTerraform(this._id),
        };
    }
}
exports.DataAwsLakeformationDataLakeSettings = DataAwsLakeformationDataLakeSettings;
_e = JSII_RTTI_SYMBOL_1;
DataAwsLakeformationDataLakeSettings[_e] = { fqn: "@cdktf/provider-aws.dataAwsLakeformationDataLakeSettings.DataAwsLakeformationDataLakeSettings", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsLakeformationDataLakeSettings.tfResourceType = "aws_lakeformation_data_lake_settings";
