// @cdktf/provider-aws/lib/redshiftdata-statement/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedshiftdataStatement = exports.RedshiftdataStatementTimeoutsOutputReference = exports.redshiftdataStatementTimeoutsToTerraform = exports.RedshiftdataStatementParametersList = exports.RedshiftdataStatementParametersOutputReference = exports.redshiftdataStatementParametersToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function redshiftdataStatementParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.redshiftdataStatementParametersToTerraform = redshiftdataStatementParametersToTerraform;
class RedshiftdataStatementParametersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._value = value.value;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.RedshiftdataStatementParametersOutputReference = RedshiftdataStatementParametersOutputReference;
_a = JSII_RTTI_SYMBOL_1;
RedshiftdataStatementParametersOutputReference[_a] = { fqn: "@cdktf/provider-aws.redshiftdataStatement.RedshiftdataStatementParametersOutputReference", version: "10.0.12" };
class RedshiftdataStatementParametersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new RedshiftdataStatementParametersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.RedshiftdataStatementParametersList = RedshiftdataStatementParametersList;
_b = JSII_RTTI_SYMBOL_1;
RedshiftdataStatementParametersList[_b] = { fqn: "@cdktf/provider-aws.redshiftdataStatement.RedshiftdataStatementParametersList", version: "10.0.12" };
function redshiftdataStatementTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
    };
}
exports.redshiftdataStatementTimeoutsToTerraform = redshiftdataStatementTimeoutsToTerraform;
class RedshiftdataStatementTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
}
exports.RedshiftdataStatementTimeoutsOutputReference = RedshiftdataStatementTimeoutsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
RedshiftdataStatementTimeoutsOutputReference[_c] = { fqn: "@cdktf/provider-aws.redshiftdataStatement.RedshiftdataStatementTimeoutsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/redshiftdata_statement aws_redshiftdata_statement}
*/
class RedshiftdataStatement extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/redshiftdata_statement aws_redshiftdata_statement} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options RedshiftdataStatementConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_redshiftdata_statement',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // parameters - computed: false, optional: true, required: false
        this._parameters = new RedshiftdataStatementParametersList(this, "parameters", false);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new RedshiftdataStatementTimeoutsOutputReference(this, "timeouts");
        this._clusterIdentifier = config.clusterIdentifier;
        this._database = config.database;
        this._dbUser = config.dbUser;
        this._id = config.id;
        this._secretArn = config.secretArn;
        this._sql = config.sql;
        this._statementName = config.statementName;
        this._withEvent = config.withEvent;
        this._parameters.internalValue = config.parameters;
        this._timeouts.internalValue = config.timeouts;
    }
    get clusterIdentifier() {
        return this.getStringAttribute('cluster_identifier');
    }
    set clusterIdentifier(value) {
        this._clusterIdentifier = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterIdentifierInput() {
        return this._clusterIdentifier;
    }
    get database() {
        return this.getStringAttribute('database');
    }
    set database(value) {
        this._database = value;
    }
    // Temporarily expose input value. Use with caution.
    get databaseInput() {
        return this._database;
    }
    get dbUser() {
        return this.getStringAttribute('db_user');
    }
    set dbUser(value) {
        this._dbUser = value;
    }
    resetDbUser() {
        this._dbUser = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dbUserInput() {
        return this._dbUser;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get secretArn() {
        return this.getStringAttribute('secret_arn');
    }
    set secretArn(value) {
        this._secretArn = value;
    }
    resetSecretArn() {
        this._secretArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secretArnInput() {
        return this._secretArn;
    }
    get sql() {
        return this.getStringAttribute('sql');
    }
    set sql(value) {
        this._sql = value;
    }
    // Temporarily expose input value. Use with caution.
    get sqlInput() {
        return this._sql;
    }
    get statementName() {
        return this.getStringAttribute('statement_name');
    }
    set statementName(value) {
        this._statementName = value;
    }
    resetStatementName() {
        this._statementName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get statementNameInput() {
        return this._statementName;
    }
    get withEvent() {
        return this.getBooleanAttribute('with_event');
    }
    set withEvent(value) {
        this._withEvent = value;
    }
    resetWithEvent() {
        this._withEvent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get withEventInput() {
        return this._withEvent;
    }
    get parameters() {
        return this._parameters;
    }
    putParameters(value) {
        this._parameters.internalValue = value;
    }
    resetParameters() {
        this._parameters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parametersInput() {
        return this._parameters.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_identifier: cdktf.stringToTerraform(this._clusterIdentifier),
            database: cdktf.stringToTerraform(this._database),
            db_user: cdktf.stringToTerraform(this._dbUser),
            id: cdktf.stringToTerraform(this._id),
            secret_arn: cdktf.stringToTerraform(this._secretArn),
            sql: cdktf.stringToTerraform(this._sql),
            statement_name: cdktf.stringToTerraform(this._statementName),
            with_event: cdktf.booleanToTerraform(this._withEvent),
            parameters: cdktf.listMapper(redshiftdataStatementParametersToTerraform, true)(this._parameters.internalValue),
            timeouts: redshiftdataStatementTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.RedshiftdataStatement = RedshiftdataStatement;
_d = JSII_RTTI_SYMBOL_1;
RedshiftdataStatement[_d] = { fqn: "@cdktf/provider-aws.redshiftdataStatement.RedshiftdataStatement", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
RedshiftdataStatement.tfResourceType = "aws_redshiftdata_statement";
