// @cdktf/provider-aws/lib/data-aws-cloudfront-origin-access-identities/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsCloudfrontOriginAccessIdentities = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/cloudfront_origin_access_identities aws_cloudfront_origin_access_identities}
*/
class DataAwsCloudfrontOriginAccessIdentities extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/cloudfront_origin_access_identities aws_cloudfront_origin_access_identities} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsCloudfrontOriginAccessIdentitiesConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_cloudfront_origin_access_identities',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._comments = config.comments;
        this._id = config.id;
    }
    get comments() {
        return cdktf.Fn.tolist(this.getListAttribute('comments'));
    }
    set comments(value) {
        this._comments = value;
    }
    resetComments() {
        this._comments = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get commentsInput() {
        return this._comments;
    }
    // iam_arns - computed: true, optional: false, required: false
    get iamArns() {
        return cdktf.Fn.tolist(this.getListAttribute('iam_arns'));
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // ids - computed: true, optional: false, required: false
    get ids() {
        return cdktf.Fn.tolist(this.getListAttribute('ids'));
    }
    // s3_canonical_user_ids - computed: true, optional: false, required: false
    get s3CanonicalUserIds() {
        return cdktf.Fn.tolist(this.getListAttribute('s3_canonical_user_ids'));
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            comments: cdktf.listMapper(cdktf.stringToTerraform, false)(this._comments),
            id: cdktf.stringToTerraform(this._id),
        };
    }
}
exports.DataAwsCloudfrontOriginAccessIdentities = DataAwsCloudfrontOriginAccessIdentities;
_a = JSII_RTTI_SYMBOL_1;
DataAwsCloudfrontOriginAccessIdentities[_a] = { fqn: "@cdktf/provider-aws.dataAwsCloudfrontOriginAccessIdentities.DataAwsCloudfrontOriginAccessIdentities", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsCloudfrontOriginAccessIdentities.tfResourceType = "aws_cloudfront_origin_access_identities";
