// @cdktf/provider-aws/lib/data-aws-route53-traffic-policy-document/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsRoute53TrafficPolicyDocument = exports.DataAwsRoute53TrafficPolicyDocumentRuleList = exports.DataAwsRoute53TrafficPolicyDocumentRuleOutputReference = exports.dataAwsRoute53TrafficPolicyDocumentRuleToTerraform = exports.DataAwsRoute53TrafficPolicyDocumentRuleSecondaryOutputReference = exports.dataAwsRoute53TrafficPolicyDocumentRuleSecondaryToTerraform = exports.DataAwsRoute53TrafficPolicyDocumentRuleRegionList = exports.DataAwsRoute53TrafficPolicyDocumentRuleRegionOutputReference = exports.dataAwsRoute53TrafficPolicyDocumentRuleRegionToTerraform = exports.DataAwsRoute53TrafficPolicyDocumentRulePrimaryOutputReference = exports.dataAwsRoute53TrafficPolicyDocumentRulePrimaryToTerraform = exports.DataAwsRoute53TrafficPolicyDocumentRuleLocationList = exports.DataAwsRoute53TrafficPolicyDocumentRuleLocationOutputReference = exports.dataAwsRoute53TrafficPolicyDocumentRuleLocationToTerraform = exports.DataAwsRoute53TrafficPolicyDocumentRuleItemsList = exports.DataAwsRoute53TrafficPolicyDocumentRuleItemsOutputReference = exports.dataAwsRoute53TrafficPolicyDocumentRuleItemsToTerraform = exports.DataAwsRoute53TrafficPolicyDocumentRuleGeoProximityLocationList = exports.DataAwsRoute53TrafficPolicyDocumentRuleGeoProximityLocationOutputReference = exports.dataAwsRoute53TrafficPolicyDocumentRuleGeoProximityLocationToTerraform = exports.DataAwsRoute53TrafficPolicyDocumentEndpointList = exports.DataAwsRoute53TrafficPolicyDocumentEndpointOutputReference = exports.dataAwsRoute53TrafficPolicyDocumentEndpointToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsRoute53TrafficPolicyDocumentEndpointToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        id: cdktf.stringToTerraform(struct.id),
        region: cdktf.stringToTerraform(struct.region),
        type: cdktf.stringToTerraform(struct.type),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.dataAwsRoute53TrafficPolicyDocumentEndpointToTerraform = dataAwsRoute53TrafficPolicyDocumentEndpointToTerraform;
class DataAwsRoute53TrafficPolicyDocumentEndpointOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._id !== undefined) {
            hasAnyValues = true;
            internalValueResult.id = this._id;
        }
        if (this._region !== undefined) {
            hasAnyValues = true;
            internalValueResult.region = this._region;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._id = undefined;
            this._region = undefined;
            this._type = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._id = value.id;
            this._region = value.region;
            this._type = value.type;
            this._value = value.value;
        }
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.DataAwsRoute53TrafficPolicyDocumentEndpointOutputReference = DataAwsRoute53TrafficPolicyDocumentEndpointOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsRoute53TrafficPolicyDocumentEndpointOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsRoute53TrafficPolicyDocument.DataAwsRoute53TrafficPolicyDocumentEndpointOutputReference", version: "10.0.12" };
class DataAwsRoute53TrafficPolicyDocumentEndpointList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsRoute53TrafficPolicyDocumentEndpointOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsRoute53TrafficPolicyDocumentEndpointList = DataAwsRoute53TrafficPolicyDocumentEndpointList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsRoute53TrafficPolicyDocumentEndpointList[_b] = { fqn: "@cdktf/provider-aws.dataAwsRoute53TrafficPolicyDocument.DataAwsRoute53TrafficPolicyDocumentEndpointList", version: "10.0.12" };
function dataAwsRoute53TrafficPolicyDocumentRuleGeoProximityLocationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bias: cdktf.stringToTerraform(struct.bias),
        endpoint_reference: cdktf.stringToTerraform(struct.endpointReference),
        evaluate_target_health: cdktf.booleanToTerraform(struct.evaluateTargetHealth),
        health_check: cdktf.stringToTerraform(struct.healthCheck),
        latitude: cdktf.stringToTerraform(struct.latitude),
        longitude: cdktf.stringToTerraform(struct.longitude),
        region: cdktf.stringToTerraform(struct.region),
        rule_reference: cdktf.stringToTerraform(struct.ruleReference),
    };
}
exports.dataAwsRoute53TrafficPolicyDocumentRuleGeoProximityLocationToTerraform = dataAwsRoute53TrafficPolicyDocumentRuleGeoProximityLocationToTerraform;
class DataAwsRoute53TrafficPolicyDocumentRuleGeoProximityLocationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bias !== undefined) {
            hasAnyValues = true;
            internalValueResult.bias = this._bias;
        }
        if (this._endpointReference !== undefined) {
            hasAnyValues = true;
            internalValueResult.endpointReference = this._endpointReference;
        }
        if (this._evaluateTargetHealth !== undefined) {
            hasAnyValues = true;
            internalValueResult.evaluateTargetHealth = this._evaluateTargetHealth;
        }
        if (this._healthCheck !== undefined) {
            hasAnyValues = true;
            internalValueResult.healthCheck = this._healthCheck;
        }
        if (this._latitude !== undefined) {
            hasAnyValues = true;
            internalValueResult.latitude = this._latitude;
        }
        if (this._longitude !== undefined) {
            hasAnyValues = true;
            internalValueResult.longitude = this._longitude;
        }
        if (this._region !== undefined) {
            hasAnyValues = true;
            internalValueResult.region = this._region;
        }
        if (this._ruleReference !== undefined) {
            hasAnyValues = true;
            internalValueResult.ruleReference = this._ruleReference;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._bias = undefined;
            this._endpointReference = undefined;
            this._evaluateTargetHealth = undefined;
            this._healthCheck = undefined;
            this._latitude = undefined;
            this._longitude = undefined;
            this._region = undefined;
            this._ruleReference = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._bias = value.bias;
            this._endpointReference = value.endpointReference;
            this._evaluateTargetHealth = value.evaluateTargetHealth;
            this._healthCheck = value.healthCheck;
            this._latitude = value.latitude;
            this._longitude = value.longitude;
            this._region = value.region;
            this._ruleReference = value.ruleReference;
        }
    }
    get bias() {
        return this.getStringAttribute('bias');
    }
    set bias(value) {
        this._bias = value;
    }
    resetBias() {
        this._bias = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get biasInput() {
        return this._bias;
    }
    get endpointReference() {
        return this.getStringAttribute('endpoint_reference');
    }
    set endpointReference(value) {
        this._endpointReference = value;
    }
    resetEndpointReference() {
        this._endpointReference = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endpointReferenceInput() {
        return this._endpointReference;
    }
    get evaluateTargetHealth() {
        return this.getBooleanAttribute('evaluate_target_health');
    }
    set evaluateTargetHealth(value) {
        this._evaluateTargetHealth = value;
    }
    resetEvaluateTargetHealth() {
        this._evaluateTargetHealth = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get evaluateTargetHealthInput() {
        return this._evaluateTargetHealth;
    }
    get healthCheck() {
        return this.getStringAttribute('health_check');
    }
    set healthCheck(value) {
        this._healthCheck = value;
    }
    resetHealthCheck() {
        this._healthCheck = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get healthCheckInput() {
        return this._healthCheck;
    }
    get latitude() {
        return this.getStringAttribute('latitude');
    }
    set latitude(value) {
        this._latitude = value;
    }
    resetLatitude() {
        this._latitude = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get latitudeInput() {
        return this._latitude;
    }
    get longitude() {
        return this.getStringAttribute('longitude');
    }
    set longitude(value) {
        this._longitude = value;
    }
    resetLongitude() {
        this._longitude = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get longitudeInput() {
        return this._longitude;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    get ruleReference() {
        return this.getStringAttribute('rule_reference');
    }
    set ruleReference(value) {
        this._ruleReference = value;
    }
    resetRuleReference() {
        this._ruleReference = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ruleReferenceInput() {
        return this._ruleReference;
    }
}
exports.DataAwsRoute53TrafficPolicyDocumentRuleGeoProximityLocationOutputReference = DataAwsRoute53TrafficPolicyDocumentRuleGeoProximityLocationOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsRoute53TrafficPolicyDocumentRuleGeoProximityLocationOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsRoute53TrafficPolicyDocument.DataAwsRoute53TrafficPolicyDocumentRuleGeoProximityLocationOutputReference", version: "10.0.12" };
class DataAwsRoute53TrafficPolicyDocumentRuleGeoProximityLocationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsRoute53TrafficPolicyDocumentRuleGeoProximityLocationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsRoute53TrafficPolicyDocumentRuleGeoProximityLocationList = DataAwsRoute53TrafficPolicyDocumentRuleGeoProximityLocationList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsRoute53TrafficPolicyDocumentRuleGeoProximityLocationList[_d] = { fqn: "@cdktf/provider-aws.dataAwsRoute53TrafficPolicyDocument.DataAwsRoute53TrafficPolicyDocumentRuleGeoProximityLocationList", version: "10.0.12" };
function dataAwsRoute53TrafficPolicyDocumentRuleItemsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        endpoint_reference: cdktf.stringToTerraform(struct.endpointReference),
        health_check: cdktf.stringToTerraform(struct.healthCheck),
    };
}
exports.dataAwsRoute53TrafficPolicyDocumentRuleItemsToTerraform = dataAwsRoute53TrafficPolicyDocumentRuleItemsToTerraform;
class DataAwsRoute53TrafficPolicyDocumentRuleItemsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._endpointReference !== undefined) {
            hasAnyValues = true;
            internalValueResult.endpointReference = this._endpointReference;
        }
        if (this._healthCheck !== undefined) {
            hasAnyValues = true;
            internalValueResult.healthCheck = this._healthCheck;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._endpointReference = undefined;
            this._healthCheck = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._endpointReference = value.endpointReference;
            this._healthCheck = value.healthCheck;
        }
    }
    get endpointReference() {
        return this.getStringAttribute('endpoint_reference');
    }
    set endpointReference(value) {
        this._endpointReference = value;
    }
    resetEndpointReference() {
        this._endpointReference = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endpointReferenceInput() {
        return this._endpointReference;
    }
    get healthCheck() {
        return this.getStringAttribute('health_check');
    }
    set healthCheck(value) {
        this._healthCheck = value;
    }
    resetHealthCheck() {
        this._healthCheck = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get healthCheckInput() {
        return this._healthCheck;
    }
}
exports.DataAwsRoute53TrafficPolicyDocumentRuleItemsOutputReference = DataAwsRoute53TrafficPolicyDocumentRuleItemsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsRoute53TrafficPolicyDocumentRuleItemsOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsRoute53TrafficPolicyDocument.DataAwsRoute53TrafficPolicyDocumentRuleItemsOutputReference", version: "10.0.12" };
class DataAwsRoute53TrafficPolicyDocumentRuleItemsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsRoute53TrafficPolicyDocumentRuleItemsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsRoute53TrafficPolicyDocumentRuleItemsList = DataAwsRoute53TrafficPolicyDocumentRuleItemsList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsRoute53TrafficPolicyDocumentRuleItemsList[_f] = { fqn: "@cdktf/provider-aws.dataAwsRoute53TrafficPolicyDocument.DataAwsRoute53TrafficPolicyDocumentRuleItemsList", version: "10.0.12" };
function dataAwsRoute53TrafficPolicyDocumentRuleLocationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        continent: cdktf.stringToTerraform(struct.continent),
        country: cdktf.stringToTerraform(struct.country),
        endpoint_reference: cdktf.stringToTerraform(struct.endpointReference),
        evaluate_target_health: cdktf.booleanToTerraform(struct.evaluateTargetHealth),
        health_check: cdktf.stringToTerraform(struct.healthCheck),
        is_default: cdktf.booleanToTerraform(struct.isDefault),
        rule_reference: cdktf.stringToTerraform(struct.ruleReference),
        subdivision: cdktf.stringToTerraform(struct.subdivision),
    };
}
exports.dataAwsRoute53TrafficPolicyDocumentRuleLocationToTerraform = dataAwsRoute53TrafficPolicyDocumentRuleLocationToTerraform;
class DataAwsRoute53TrafficPolicyDocumentRuleLocationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._continent !== undefined) {
            hasAnyValues = true;
            internalValueResult.continent = this._continent;
        }
        if (this._country !== undefined) {
            hasAnyValues = true;
            internalValueResult.country = this._country;
        }
        if (this._endpointReference !== undefined) {
            hasAnyValues = true;
            internalValueResult.endpointReference = this._endpointReference;
        }
        if (this._evaluateTargetHealth !== undefined) {
            hasAnyValues = true;
            internalValueResult.evaluateTargetHealth = this._evaluateTargetHealth;
        }
        if (this._healthCheck !== undefined) {
            hasAnyValues = true;
            internalValueResult.healthCheck = this._healthCheck;
        }
        if (this._isDefault !== undefined) {
            hasAnyValues = true;
            internalValueResult.isDefault = this._isDefault;
        }
        if (this._ruleReference !== undefined) {
            hasAnyValues = true;
            internalValueResult.ruleReference = this._ruleReference;
        }
        if (this._subdivision !== undefined) {
            hasAnyValues = true;
            internalValueResult.subdivision = this._subdivision;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._continent = undefined;
            this._country = undefined;
            this._endpointReference = undefined;
            this._evaluateTargetHealth = undefined;
            this._healthCheck = undefined;
            this._isDefault = undefined;
            this._ruleReference = undefined;
            this._subdivision = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._continent = value.continent;
            this._country = value.country;
            this._endpointReference = value.endpointReference;
            this._evaluateTargetHealth = value.evaluateTargetHealth;
            this._healthCheck = value.healthCheck;
            this._isDefault = value.isDefault;
            this._ruleReference = value.ruleReference;
            this._subdivision = value.subdivision;
        }
    }
    get continent() {
        return this.getStringAttribute('continent');
    }
    set continent(value) {
        this._continent = value;
    }
    resetContinent() {
        this._continent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get continentInput() {
        return this._continent;
    }
    get country() {
        return this.getStringAttribute('country');
    }
    set country(value) {
        this._country = value;
    }
    resetCountry() {
        this._country = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get countryInput() {
        return this._country;
    }
    get endpointReference() {
        return this.getStringAttribute('endpoint_reference');
    }
    set endpointReference(value) {
        this._endpointReference = value;
    }
    resetEndpointReference() {
        this._endpointReference = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endpointReferenceInput() {
        return this._endpointReference;
    }
    get evaluateTargetHealth() {
        return this.getBooleanAttribute('evaluate_target_health');
    }
    set evaluateTargetHealth(value) {
        this._evaluateTargetHealth = value;
    }
    resetEvaluateTargetHealth() {
        this._evaluateTargetHealth = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get evaluateTargetHealthInput() {
        return this._evaluateTargetHealth;
    }
    get healthCheck() {
        return this.getStringAttribute('health_check');
    }
    set healthCheck(value) {
        this._healthCheck = value;
    }
    resetHealthCheck() {
        this._healthCheck = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get healthCheckInput() {
        return this._healthCheck;
    }
    get isDefault() {
        return this.getBooleanAttribute('is_default');
    }
    set isDefault(value) {
        this._isDefault = value;
    }
    resetIsDefault() {
        this._isDefault = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get isDefaultInput() {
        return this._isDefault;
    }
    get ruleReference() {
        return this.getStringAttribute('rule_reference');
    }
    set ruleReference(value) {
        this._ruleReference = value;
    }
    resetRuleReference() {
        this._ruleReference = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ruleReferenceInput() {
        return this._ruleReference;
    }
    get subdivision() {
        return this.getStringAttribute('subdivision');
    }
    set subdivision(value) {
        this._subdivision = value;
    }
    resetSubdivision() {
        this._subdivision = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subdivisionInput() {
        return this._subdivision;
    }
}
exports.DataAwsRoute53TrafficPolicyDocumentRuleLocationOutputReference = DataAwsRoute53TrafficPolicyDocumentRuleLocationOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsRoute53TrafficPolicyDocumentRuleLocationOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsRoute53TrafficPolicyDocument.DataAwsRoute53TrafficPolicyDocumentRuleLocationOutputReference", version: "10.0.12" };
class DataAwsRoute53TrafficPolicyDocumentRuleLocationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsRoute53TrafficPolicyDocumentRuleLocationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsRoute53TrafficPolicyDocumentRuleLocationList = DataAwsRoute53TrafficPolicyDocumentRuleLocationList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsRoute53TrafficPolicyDocumentRuleLocationList[_h] = { fqn: "@cdktf/provider-aws.dataAwsRoute53TrafficPolicyDocument.DataAwsRoute53TrafficPolicyDocumentRuleLocationList", version: "10.0.12" };
function dataAwsRoute53TrafficPolicyDocumentRulePrimaryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        endpoint_reference: cdktf.stringToTerraform(struct.endpointReference),
        evaluate_target_health: cdktf.booleanToTerraform(struct.evaluateTargetHealth),
        health_check: cdktf.stringToTerraform(struct.healthCheck),
        rule_reference: cdktf.stringToTerraform(struct.ruleReference),
    };
}
exports.dataAwsRoute53TrafficPolicyDocumentRulePrimaryToTerraform = dataAwsRoute53TrafficPolicyDocumentRulePrimaryToTerraform;
class DataAwsRoute53TrafficPolicyDocumentRulePrimaryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._endpointReference !== undefined) {
            hasAnyValues = true;
            internalValueResult.endpointReference = this._endpointReference;
        }
        if (this._evaluateTargetHealth !== undefined) {
            hasAnyValues = true;
            internalValueResult.evaluateTargetHealth = this._evaluateTargetHealth;
        }
        if (this._healthCheck !== undefined) {
            hasAnyValues = true;
            internalValueResult.healthCheck = this._healthCheck;
        }
        if (this._ruleReference !== undefined) {
            hasAnyValues = true;
            internalValueResult.ruleReference = this._ruleReference;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._endpointReference = undefined;
            this._evaluateTargetHealth = undefined;
            this._healthCheck = undefined;
            this._ruleReference = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._endpointReference = value.endpointReference;
            this._evaluateTargetHealth = value.evaluateTargetHealth;
            this._healthCheck = value.healthCheck;
            this._ruleReference = value.ruleReference;
        }
    }
    get endpointReference() {
        return this.getStringAttribute('endpoint_reference');
    }
    set endpointReference(value) {
        this._endpointReference = value;
    }
    resetEndpointReference() {
        this._endpointReference = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endpointReferenceInput() {
        return this._endpointReference;
    }
    get evaluateTargetHealth() {
        return this.getBooleanAttribute('evaluate_target_health');
    }
    set evaluateTargetHealth(value) {
        this._evaluateTargetHealth = value;
    }
    resetEvaluateTargetHealth() {
        this._evaluateTargetHealth = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get evaluateTargetHealthInput() {
        return this._evaluateTargetHealth;
    }
    get healthCheck() {
        return this.getStringAttribute('health_check');
    }
    set healthCheck(value) {
        this._healthCheck = value;
    }
    resetHealthCheck() {
        this._healthCheck = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get healthCheckInput() {
        return this._healthCheck;
    }
    get ruleReference() {
        return this.getStringAttribute('rule_reference');
    }
    set ruleReference(value) {
        this._ruleReference = value;
    }
    resetRuleReference() {
        this._ruleReference = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ruleReferenceInput() {
        return this._ruleReference;
    }
}
exports.DataAwsRoute53TrafficPolicyDocumentRulePrimaryOutputReference = DataAwsRoute53TrafficPolicyDocumentRulePrimaryOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsRoute53TrafficPolicyDocumentRulePrimaryOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsRoute53TrafficPolicyDocument.DataAwsRoute53TrafficPolicyDocumentRulePrimaryOutputReference", version: "10.0.12" };
function dataAwsRoute53TrafficPolicyDocumentRuleRegionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        endpoint_reference: cdktf.stringToTerraform(struct.endpointReference),
        evaluate_target_health: cdktf.booleanToTerraform(struct.evaluateTargetHealth),
        health_check: cdktf.stringToTerraform(struct.healthCheck),
        region: cdktf.stringToTerraform(struct.region),
        rule_reference: cdktf.stringToTerraform(struct.ruleReference),
    };
}
exports.dataAwsRoute53TrafficPolicyDocumentRuleRegionToTerraform = dataAwsRoute53TrafficPolicyDocumentRuleRegionToTerraform;
class DataAwsRoute53TrafficPolicyDocumentRuleRegionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._endpointReference !== undefined) {
            hasAnyValues = true;
            internalValueResult.endpointReference = this._endpointReference;
        }
        if (this._evaluateTargetHealth !== undefined) {
            hasAnyValues = true;
            internalValueResult.evaluateTargetHealth = this._evaluateTargetHealth;
        }
        if (this._healthCheck !== undefined) {
            hasAnyValues = true;
            internalValueResult.healthCheck = this._healthCheck;
        }
        if (this._region !== undefined) {
            hasAnyValues = true;
            internalValueResult.region = this._region;
        }
        if (this._ruleReference !== undefined) {
            hasAnyValues = true;
            internalValueResult.ruleReference = this._ruleReference;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._endpointReference = undefined;
            this._evaluateTargetHealth = undefined;
            this._healthCheck = undefined;
            this._region = undefined;
            this._ruleReference = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._endpointReference = value.endpointReference;
            this._evaluateTargetHealth = value.evaluateTargetHealth;
            this._healthCheck = value.healthCheck;
            this._region = value.region;
            this._ruleReference = value.ruleReference;
        }
    }
    get endpointReference() {
        return this.getStringAttribute('endpoint_reference');
    }
    set endpointReference(value) {
        this._endpointReference = value;
    }
    resetEndpointReference() {
        this._endpointReference = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endpointReferenceInput() {
        return this._endpointReference;
    }
    get evaluateTargetHealth() {
        return this.getBooleanAttribute('evaluate_target_health');
    }
    set evaluateTargetHealth(value) {
        this._evaluateTargetHealth = value;
    }
    resetEvaluateTargetHealth() {
        this._evaluateTargetHealth = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get evaluateTargetHealthInput() {
        return this._evaluateTargetHealth;
    }
    get healthCheck() {
        return this.getStringAttribute('health_check');
    }
    set healthCheck(value) {
        this._healthCheck = value;
    }
    resetHealthCheck() {
        this._healthCheck = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get healthCheckInput() {
        return this._healthCheck;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    get ruleReference() {
        return this.getStringAttribute('rule_reference');
    }
    set ruleReference(value) {
        this._ruleReference = value;
    }
    resetRuleReference() {
        this._ruleReference = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ruleReferenceInput() {
        return this._ruleReference;
    }
}
exports.DataAwsRoute53TrafficPolicyDocumentRuleRegionOutputReference = DataAwsRoute53TrafficPolicyDocumentRuleRegionOutputReference;
_k = JSII_RTTI_SYMBOL_1;
DataAwsRoute53TrafficPolicyDocumentRuleRegionOutputReference[_k] = { fqn: "@cdktf/provider-aws.dataAwsRoute53TrafficPolicyDocument.DataAwsRoute53TrafficPolicyDocumentRuleRegionOutputReference", version: "10.0.12" };
class DataAwsRoute53TrafficPolicyDocumentRuleRegionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsRoute53TrafficPolicyDocumentRuleRegionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsRoute53TrafficPolicyDocumentRuleRegionList = DataAwsRoute53TrafficPolicyDocumentRuleRegionList;
_l = JSII_RTTI_SYMBOL_1;
DataAwsRoute53TrafficPolicyDocumentRuleRegionList[_l] = { fqn: "@cdktf/provider-aws.dataAwsRoute53TrafficPolicyDocument.DataAwsRoute53TrafficPolicyDocumentRuleRegionList", version: "10.0.12" };
function dataAwsRoute53TrafficPolicyDocumentRuleSecondaryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        endpoint_reference: cdktf.stringToTerraform(struct.endpointReference),
        evaluate_target_health: cdktf.booleanToTerraform(struct.evaluateTargetHealth),
        health_check: cdktf.stringToTerraform(struct.healthCheck),
        rule_reference: cdktf.stringToTerraform(struct.ruleReference),
    };
}
exports.dataAwsRoute53TrafficPolicyDocumentRuleSecondaryToTerraform = dataAwsRoute53TrafficPolicyDocumentRuleSecondaryToTerraform;
class DataAwsRoute53TrafficPolicyDocumentRuleSecondaryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._endpointReference !== undefined) {
            hasAnyValues = true;
            internalValueResult.endpointReference = this._endpointReference;
        }
        if (this._evaluateTargetHealth !== undefined) {
            hasAnyValues = true;
            internalValueResult.evaluateTargetHealth = this._evaluateTargetHealth;
        }
        if (this._healthCheck !== undefined) {
            hasAnyValues = true;
            internalValueResult.healthCheck = this._healthCheck;
        }
        if (this._ruleReference !== undefined) {
            hasAnyValues = true;
            internalValueResult.ruleReference = this._ruleReference;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._endpointReference = undefined;
            this._evaluateTargetHealth = undefined;
            this._healthCheck = undefined;
            this._ruleReference = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._endpointReference = value.endpointReference;
            this._evaluateTargetHealth = value.evaluateTargetHealth;
            this._healthCheck = value.healthCheck;
            this._ruleReference = value.ruleReference;
        }
    }
    get endpointReference() {
        return this.getStringAttribute('endpoint_reference');
    }
    set endpointReference(value) {
        this._endpointReference = value;
    }
    resetEndpointReference() {
        this._endpointReference = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endpointReferenceInput() {
        return this._endpointReference;
    }
    get evaluateTargetHealth() {
        return this.getBooleanAttribute('evaluate_target_health');
    }
    set evaluateTargetHealth(value) {
        this._evaluateTargetHealth = value;
    }
    resetEvaluateTargetHealth() {
        this._evaluateTargetHealth = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get evaluateTargetHealthInput() {
        return this._evaluateTargetHealth;
    }
    get healthCheck() {
        return this.getStringAttribute('health_check');
    }
    set healthCheck(value) {
        this._healthCheck = value;
    }
    resetHealthCheck() {
        this._healthCheck = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get healthCheckInput() {
        return this._healthCheck;
    }
    get ruleReference() {
        return this.getStringAttribute('rule_reference');
    }
    set ruleReference(value) {
        this._ruleReference = value;
    }
    resetRuleReference() {
        this._ruleReference = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ruleReferenceInput() {
        return this._ruleReference;
    }
}
exports.DataAwsRoute53TrafficPolicyDocumentRuleSecondaryOutputReference = DataAwsRoute53TrafficPolicyDocumentRuleSecondaryOutputReference;
_m = JSII_RTTI_SYMBOL_1;
DataAwsRoute53TrafficPolicyDocumentRuleSecondaryOutputReference[_m] = { fqn: "@cdktf/provider-aws.dataAwsRoute53TrafficPolicyDocument.DataAwsRoute53TrafficPolicyDocumentRuleSecondaryOutputReference", version: "10.0.12" };
function dataAwsRoute53TrafficPolicyDocumentRuleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        id: cdktf.stringToTerraform(struct.id),
        type: cdktf.stringToTerraform(struct.type),
        geo_proximity_location: cdktf.listMapper(dataAwsRoute53TrafficPolicyDocumentRuleGeoProximityLocationToTerraform, true)(struct.geoProximityLocation),
        items: cdktf.listMapper(dataAwsRoute53TrafficPolicyDocumentRuleItemsToTerraform, true)(struct.items),
        location: cdktf.listMapper(dataAwsRoute53TrafficPolicyDocumentRuleLocationToTerraform, true)(struct.location),
        primary: dataAwsRoute53TrafficPolicyDocumentRulePrimaryToTerraform(struct.primary),
        region: cdktf.listMapper(dataAwsRoute53TrafficPolicyDocumentRuleRegionToTerraform, true)(struct.region),
        secondary: dataAwsRoute53TrafficPolicyDocumentRuleSecondaryToTerraform(struct.secondary),
    };
}
exports.dataAwsRoute53TrafficPolicyDocumentRuleToTerraform = dataAwsRoute53TrafficPolicyDocumentRuleToTerraform;
class DataAwsRoute53TrafficPolicyDocumentRuleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // geo_proximity_location - computed: false, optional: true, required: false
        this._geoProximityLocation = new DataAwsRoute53TrafficPolicyDocumentRuleGeoProximityLocationList(this, "geo_proximity_location", true);
        // items - computed: false, optional: true, required: false
        this._items = new DataAwsRoute53TrafficPolicyDocumentRuleItemsList(this, "items", true);
        // location - computed: false, optional: true, required: false
        this._location = new DataAwsRoute53TrafficPolicyDocumentRuleLocationList(this, "location", true);
        // primary - computed: false, optional: true, required: false
        this._primary = new DataAwsRoute53TrafficPolicyDocumentRulePrimaryOutputReference(this, "primary");
        // region - computed: false, optional: true, required: false
        this._region = new DataAwsRoute53TrafficPolicyDocumentRuleRegionList(this, "region", true);
        // secondary - computed: false, optional: true, required: false
        this._secondary = new DataAwsRoute53TrafficPolicyDocumentRuleSecondaryOutputReference(this, "secondary");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._id !== undefined) {
            hasAnyValues = true;
            internalValueResult.id = this._id;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._geoProximityLocation?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.geoProximityLocation = this._geoProximityLocation?.internalValue;
        }
        if (this._items?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.items = this._items?.internalValue;
        }
        if (this._location?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.location = this._location?.internalValue;
        }
        if (this._primary?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.primary = this._primary?.internalValue;
        }
        if (this._region?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.region = this._region?.internalValue;
        }
        if (this._secondary?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.secondary = this._secondary?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._id = undefined;
            this._type = undefined;
            this._geoProximityLocation.internalValue = undefined;
            this._items.internalValue = undefined;
            this._location.internalValue = undefined;
            this._primary.internalValue = undefined;
            this._region.internalValue = undefined;
            this._secondary.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._id = value.id;
            this._type = value.type;
            this._geoProximityLocation.internalValue = value.geoProximityLocation;
            this._items.internalValue = value.items;
            this._location.internalValue = value.location;
            this._primary.internalValue = value.primary;
            this._region.internalValue = value.region;
            this._secondary.internalValue = value.secondary;
        }
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get geoProximityLocation() {
        return this._geoProximityLocation;
    }
    putGeoProximityLocation(value) {
        this._geoProximityLocation.internalValue = value;
    }
    resetGeoProximityLocation() {
        this._geoProximityLocation.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get geoProximityLocationInput() {
        return this._geoProximityLocation.internalValue;
    }
    get items() {
        return this._items;
    }
    putItems(value) {
        this._items.internalValue = value;
    }
    resetItems() {
        this._items.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get itemsInput() {
        return this._items.internalValue;
    }
    get location() {
        return this._location;
    }
    putLocation(value) {
        this._location.internalValue = value;
    }
    resetLocation() {
        this._location.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location.internalValue;
    }
    get primary() {
        return this._primary;
    }
    putPrimary(value) {
        this._primary.internalValue = value;
    }
    resetPrimary() {
        this._primary.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get primaryInput() {
        return this._primary.internalValue;
    }
    get region() {
        return this._region;
    }
    putRegion(value) {
        this._region.internalValue = value;
    }
    resetRegion() {
        this._region.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region.internalValue;
    }
    get secondary() {
        return this._secondary;
    }
    putSecondary(value) {
        this._secondary.internalValue = value;
    }
    resetSecondary() {
        this._secondary.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secondaryInput() {
        return this._secondary.internalValue;
    }
}
exports.DataAwsRoute53TrafficPolicyDocumentRuleOutputReference = DataAwsRoute53TrafficPolicyDocumentRuleOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataAwsRoute53TrafficPolicyDocumentRuleOutputReference[_o] = { fqn: "@cdktf/provider-aws.dataAwsRoute53TrafficPolicyDocument.DataAwsRoute53TrafficPolicyDocumentRuleOutputReference", version: "10.0.12" };
class DataAwsRoute53TrafficPolicyDocumentRuleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsRoute53TrafficPolicyDocumentRuleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsRoute53TrafficPolicyDocumentRuleList = DataAwsRoute53TrafficPolicyDocumentRuleList;
_p = JSII_RTTI_SYMBOL_1;
DataAwsRoute53TrafficPolicyDocumentRuleList[_p] = { fqn: "@cdktf/provider-aws.dataAwsRoute53TrafficPolicyDocument.DataAwsRoute53TrafficPolicyDocumentRuleList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/route53_traffic_policy_document aws_route53_traffic_policy_document}
*/
class DataAwsRoute53TrafficPolicyDocument extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/route53_traffic_policy_document aws_route53_traffic_policy_document} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsRoute53TrafficPolicyDocumentConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_route53_traffic_policy_document',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // endpoint - computed: false, optional: true, required: false
        this._endpoint = new DataAwsRoute53TrafficPolicyDocumentEndpointList(this, "endpoint", true);
        // rule - computed: false, optional: true, required: false
        this._rule = new DataAwsRoute53TrafficPolicyDocumentRuleList(this, "rule", true);
        this._id = config.id;
        this._recordType = config.recordType;
        this._startEndpoint = config.startEndpoint;
        this._startRule = config.startRule;
        this._version = config.version;
        this._endpoint.internalValue = config.endpoint;
        this._rule.internalValue = config.rule;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // json - computed: true, optional: false, required: false
    get json() {
        return this.getStringAttribute('json');
    }
    get recordType() {
        return this.getStringAttribute('record_type');
    }
    set recordType(value) {
        this._recordType = value;
    }
    resetRecordType() {
        this._recordType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get recordTypeInput() {
        return this._recordType;
    }
    get startEndpoint() {
        return this.getStringAttribute('start_endpoint');
    }
    set startEndpoint(value) {
        this._startEndpoint = value;
    }
    resetStartEndpoint() {
        this._startEndpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startEndpointInput() {
        return this._startEndpoint;
    }
    get startRule() {
        return this.getStringAttribute('start_rule');
    }
    set startRule(value) {
        this._startRule = value;
    }
    resetStartRule() {
        this._startRule = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startRuleInput() {
        return this._startRule;
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    resetVersion() {
        this._version = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
    get endpoint() {
        return this._endpoint;
    }
    putEndpoint(value) {
        this._endpoint.internalValue = value;
    }
    resetEndpoint() {
        this._endpoint.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endpointInput() {
        return this._endpoint.internalValue;
    }
    get rule() {
        return this._rule;
    }
    putRule(value) {
        this._rule.internalValue = value;
    }
    resetRule() {
        this._rule.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ruleInput() {
        return this._rule.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            record_type: cdktf.stringToTerraform(this._recordType),
            start_endpoint: cdktf.stringToTerraform(this._startEndpoint),
            start_rule: cdktf.stringToTerraform(this._startRule),
            version: cdktf.stringToTerraform(this._version),
            endpoint: cdktf.listMapper(dataAwsRoute53TrafficPolicyDocumentEndpointToTerraform, true)(this._endpoint.internalValue),
            rule: cdktf.listMapper(dataAwsRoute53TrafficPolicyDocumentRuleToTerraform, true)(this._rule.internalValue),
        };
    }
}
exports.DataAwsRoute53TrafficPolicyDocument = DataAwsRoute53TrafficPolicyDocument;
_q = JSII_RTTI_SYMBOL_1;
DataAwsRoute53TrafficPolicyDocument[_q] = { fqn: "@cdktf/provider-aws.dataAwsRoute53TrafficPolicyDocument.DataAwsRoute53TrafficPolicyDocument", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsRoute53TrafficPolicyDocument.tfResourceType = "aws_route53_traffic_policy_document";
