// @cdktf/provider-aws/lib/network-interface/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkInterface = exports.NetworkInterfaceAttachmentList = exports.NetworkInterfaceAttachmentOutputReference = exports.networkInterfaceAttachmentToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function networkInterfaceAttachmentToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        device_index: cdktf.numberToTerraform(struct.deviceIndex),
        instance: cdktf.stringToTerraform(struct.instance),
    };
}
exports.networkInterfaceAttachmentToTerraform = networkInterfaceAttachmentToTerraform;
class NetworkInterfaceAttachmentOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._deviceIndex !== undefined) {
            hasAnyValues = true;
            internalValueResult.deviceIndex = this._deviceIndex;
        }
        if (this._instance !== undefined) {
            hasAnyValues = true;
            internalValueResult.instance = this._instance;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._deviceIndex = undefined;
            this._instance = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._deviceIndex = value.deviceIndex;
            this._instance = value.instance;
        }
    }
    // attachment_id - computed: true, optional: false, required: false
    get attachmentId() {
        return this.getStringAttribute('attachment_id');
    }
    get deviceIndex() {
        return this.getNumberAttribute('device_index');
    }
    set deviceIndex(value) {
        this._deviceIndex = value;
    }
    // Temporarily expose input value. Use with caution.
    get deviceIndexInput() {
        return this._deviceIndex;
    }
    get instance() {
        return this.getStringAttribute('instance');
    }
    set instance(value) {
        this._instance = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceInput() {
        return this._instance;
    }
}
exports.NetworkInterfaceAttachmentOutputReference = NetworkInterfaceAttachmentOutputReference;
_a = JSII_RTTI_SYMBOL_1;
NetworkInterfaceAttachmentOutputReference[_a] = { fqn: "@cdktf/provider-aws.networkInterface.NetworkInterfaceAttachmentOutputReference", version: "10.0.12" };
class NetworkInterfaceAttachmentList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkInterfaceAttachmentOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkInterfaceAttachmentList = NetworkInterfaceAttachmentList;
_b = JSII_RTTI_SYMBOL_1;
NetworkInterfaceAttachmentList[_b] = { fqn: "@cdktf/provider-aws.networkInterface.NetworkInterfaceAttachmentList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/network_interface aws_network_interface}
*/
class NetworkInterface extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/network_interface aws_network_interface} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options NetworkInterfaceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_network_interface',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // attachment - computed: false, optional: true, required: false
        this._attachment = new NetworkInterfaceAttachmentList(this, "attachment", true);
        this._description = config.description;
        this._id = config.id;
        this._interfaceType = config.interfaceType;
        this._ipv4PrefixCount = config.ipv4PrefixCount;
        this._ipv4Prefixes = config.ipv4Prefixes;
        this._ipv6AddressCount = config.ipv6AddressCount;
        this._ipv6AddressList = config.ipv6AddressList;
        this._ipv6AddressListEnabled = config.ipv6AddressListEnabled;
        this._ipv6Addresses = config.ipv6Addresses;
        this._ipv6PrefixCount = config.ipv6PrefixCount;
        this._ipv6Prefixes = config.ipv6Prefixes;
        this._privateIp = config.privateIp;
        this._privateIpList = config.privateIpList;
        this._privateIpListEnabled = config.privateIpListEnabled;
        this._privateIps = config.privateIps;
        this._privateIpsCount = config.privateIpsCount;
        this._securityGroups = config.securityGroups;
        this._sourceDestCheck = config.sourceDestCheck;
        this._subnetId = config.subnetId;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._attachment.internalValue = config.attachment;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get interfaceType() {
        return this.getStringAttribute('interface_type');
    }
    set interfaceType(value) {
        this._interfaceType = value;
    }
    resetInterfaceType() {
        this._interfaceType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get interfaceTypeInput() {
        return this._interfaceType;
    }
    get ipv4PrefixCount() {
        return this.getNumberAttribute('ipv4_prefix_count');
    }
    set ipv4PrefixCount(value) {
        this._ipv4PrefixCount = value;
    }
    resetIpv4PrefixCount() {
        this._ipv4PrefixCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipv4PrefixCountInput() {
        return this._ipv4PrefixCount;
    }
    get ipv4Prefixes() {
        return cdktf.Fn.tolist(this.getListAttribute('ipv4_prefixes'));
    }
    set ipv4Prefixes(value) {
        this._ipv4Prefixes = value;
    }
    resetIpv4Prefixes() {
        this._ipv4Prefixes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipv4PrefixesInput() {
        return this._ipv4Prefixes;
    }
    get ipv6AddressCount() {
        return this.getNumberAttribute('ipv6_address_count');
    }
    set ipv6AddressCount(value) {
        this._ipv6AddressCount = value;
    }
    resetIpv6AddressCount() {
        this._ipv6AddressCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipv6AddressCountInput() {
        return this._ipv6AddressCount;
    }
    get ipv6AddressList() {
        return this.getListAttribute('ipv6_address_list');
    }
    set ipv6AddressList(value) {
        this._ipv6AddressList = value;
    }
    resetIpv6AddressList() {
        this._ipv6AddressList = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipv6AddressListInput() {
        return this._ipv6AddressList;
    }
    get ipv6AddressListEnabled() {
        return this.getBooleanAttribute('ipv6_address_list_enabled');
    }
    set ipv6AddressListEnabled(value) {
        this._ipv6AddressListEnabled = value;
    }
    resetIpv6AddressListEnabled() {
        this._ipv6AddressListEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipv6AddressListEnabledInput() {
        return this._ipv6AddressListEnabled;
    }
    get ipv6Addresses() {
        return cdktf.Fn.tolist(this.getListAttribute('ipv6_addresses'));
    }
    set ipv6Addresses(value) {
        this._ipv6Addresses = value;
    }
    resetIpv6Addresses() {
        this._ipv6Addresses = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipv6AddressesInput() {
        return this._ipv6Addresses;
    }
    get ipv6PrefixCount() {
        return this.getNumberAttribute('ipv6_prefix_count');
    }
    set ipv6PrefixCount(value) {
        this._ipv6PrefixCount = value;
    }
    resetIpv6PrefixCount() {
        this._ipv6PrefixCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipv6PrefixCountInput() {
        return this._ipv6PrefixCount;
    }
    get ipv6Prefixes() {
        return cdktf.Fn.tolist(this.getListAttribute('ipv6_prefixes'));
    }
    set ipv6Prefixes(value) {
        this._ipv6Prefixes = value;
    }
    resetIpv6Prefixes() {
        this._ipv6Prefixes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipv6PrefixesInput() {
        return this._ipv6Prefixes;
    }
    // mac_address - computed: true, optional: false, required: false
    get macAddress() {
        return this.getStringAttribute('mac_address');
    }
    // outpost_arn - computed: true, optional: false, required: false
    get outpostArn() {
        return this.getStringAttribute('outpost_arn');
    }
    // owner_id - computed: true, optional: false, required: false
    get ownerId() {
        return this.getStringAttribute('owner_id');
    }
    // private_dns_name - computed: true, optional: false, required: false
    get privateDnsName() {
        return this.getStringAttribute('private_dns_name');
    }
    get privateIp() {
        return this.getStringAttribute('private_ip');
    }
    set privateIp(value) {
        this._privateIp = value;
    }
    resetPrivateIp() {
        this._privateIp = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get privateIpInput() {
        return this._privateIp;
    }
    get privateIpList() {
        return this.getListAttribute('private_ip_list');
    }
    set privateIpList(value) {
        this._privateIpList = value;
    }
    resetPrivateIpList() {
        this._privateIpList = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get privateIpListInput() {
        return this._privateIpList;
    }
    get privateIpListEnabled() {
        return this.getBooleanAttribute('private_ip_list_enabled');
    }
    set privateIpListEnabled(value) {
        this._privateIpListEnabled = value;
    }
    resetPrivateIpListEnabled() {
        this._privateIpListEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get privateIpListEnabledInput() {
        return this._privateIpListEnabled;
    }
    get privateIps() {
        return cdktf.Fn.tolist(this.getListAttribute('private_ips'));
    }
    set privateIps(value) {
        this._privateIps = value;
    }
    resetPrivateIps() {
        this._privateIps = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get privateIpsInput() {
        return this._privateIps;
    }
    get privateIpsCount() {
        return this.getNumberAttribute('private_ips_count');
    }
    set privateIpsCount(value) {
        this._privateIpsCount = value;
    }
    resetPrivateIpsCount() {
        this._privateIpsCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get privateIpsCountInput() {
        return this._privateIpsCount;
    }
    get securityGroups() {
        return cdktf.Fn.tolist(this.getListAttribute('security_groups'));
    }
    set securityGroups(value) {
        this._securityGroups = value;
    }
    resetSecurityGroups() {
        this._securityGroups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupsInput() {
        return this._securityGroups;
    }
    get sourceDestCheck() {
        return this.getBooleanAttribute('source_dest_check');
    }
    set sourceDestCheck(value) {
        this._sourceDestCheck = value;
    }
    resetSourceDestCheck() {
        this._sourceDestCheck = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceDestCheckInput() {
        return this._sourceDestCheck;
    }
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    set subnetId(value) {
        this._subnetId = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdInput() {
        return this._subnetId;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get attachment() {
        return this._attachment;
    }
    putAttachment(value) {
        this._attachment.internalValue = value;
    }
    resetAttachment() {
        this._attachment.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get attachmentInput() {
        return this._attachment.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            interface_type: cdktf.stringToTerraform(this._interfaceType),
            ipv4_prefix_count: cdktf.numberToTerraform(this._ipv4PrefixCount),
            ipv4_prefixes: cdktf.listMapper(cdktf.stringToTerraform, false)(this._ipv4Prefixes),
            ipv6_address_count: cdktf.numberToTerraform(this._ipv6AddressCount),
            ipv6_address_list: cdktf.listMapper(cdktf.stringToTerraform, false)(this._ipv6AddressList),
            ipv6_address_list_enabled: cdktf.booleanToTerraform(this._ipv6AddressListEnabled),
            ipv6_addresses: cdktf.listMapper(cdktf.stringToTerraform, false)(this._ipv6Addresses),
            ipv6_prefix_count: cdktf.numberToTerraform(this._ipv6PrefixCount),
            ipv6_prefixes: cdktf.listMapper(cdktf.stringToTerraform, false)(this._ipv6Prefixes),
            private_ip: cdktf.stringToTerraform(this._privateIp),
            private_ip_list: cdktf.listMapper(cdktf.stringToTerraform, false)(this._privateIpList),
            private_ip_list_enabled: cdktf.booleanToTerraform(this._privateIpListEnabled),
            private_ips: cdktf.listMapper(cdktf.stringToTerraform, false)(this._privateIps),
            private_ips_count: cdktf.numberToTerraform(this._privateIpsCount),
            security_groups: cdktf.listMapper(cdktf.stringToTerraform, false)(this._securityGroups),
            source_dest_check: cdktf.booleanToTerraform(this._sourceDestCheck),
            subnet_id: cdktf.stringToTerraform(this._subnetId),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            attachment: cdktf.listMapper(networkInterfaceAttachmentToTerraform, true)(this._attachment.internalValue),
        };
    }
}
exports.NetworkInterface = NetworkInterface;
_c = JSII_RTTI_SYMBOL_1;
NetworkInterface[_c] = { fqn: "@cdktf/provider-aws.networkInterface.NetworkInterface", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
NetworkInterface.tfResourceType = "aws_network_interface";
