// @cdktf/provider-aws/lib/rolesanywhere-profile/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RolesanywhereProfile = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/rolesanywhere_profile aws_rolesanywhere_profile}
*/
class RolesanywhereProfile extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/rolesanywhere_profile aws_rolesanywhere_profile} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options RolesanywhereProfileConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_rolesanywhere_profile',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._durationSeconds = config.durationSeconds;
        this._enabled = config.enabled;
        this._id = config.id;
        this._managedPolicyArns = config.managedPolicyArns;
        this._name = config.name;
        this._requireInstanceProperties = config.requireInstanceProperties;
        this._roleArns = config.roleArns;
        this._sessionPolicy = config.sessionPolicy;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get durationSeconds() {
        return this.getNumberAttribute('duration_seconds');
    }
    set durationSeconds(value) {
        this._durationSeconds = value;
    }
    resetDurationSeconds() {
        this._durationSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get durationSecondsInput() {
        return this._durationSeconds;
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get managedPolicyArns() {
        return cdktf.Fn.tolist(this.getListAttribute('managed_policy_arns'));
    }
    set managedPolicyArns(value) {
        this._managedPolicyArns = value;
    }
    resetManagedPolicyArns() {
        this._managedPolicyArns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get managedPolicyArnsInput() {
        return this._managedPolicyArns;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get requireInstanceProperties() {
        return this.getBooleanAttribute('require_instance_properties');
    }
    set requireInstanceProperties(value) {
        this._requireInstanceProperties = value;
    }
    resetRequireInstanceProperties() {
        this._requireInstanceProperties = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get requireInstancePropertiesInput() {
        return this._requireInstanceProperties;
    }
    get roleArns() {
        return cdktf.Fn.tolist(this.getListAttribute('role_arns'));
    }
    set roleArns(value) {
        this._roleArns = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnsInput() {
        return this._roleArns;
    }
    get sessionPolicy() {
        return this.getStringAttribute('session_policy');
    }
    set sessionPolicy(value) {
        this._sessionPolicy = value;
    }
    resetSessionPolicy() {
        this._sessionPolicy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sessionPolicyInput() {
        return this._sessionPolicy;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            duration_seconds: cdktf.numberToTerraform(this._durationSeconds),
            enabled: cdktf.booleanToTerraform(this._enabled),
            id: cdktf.stringToTerraform(this._id),
            managed_policy_arns: cdktf.listMapper(cdktf.stringToTerraform, false)(this._managedPolicyArns),
            name: cdktf.stringToTerraform(this._name),
            require_instance_properties: cdktf.booleanToTerraform(this._requireInstanceProperties),
            role_arns: cdktf.listMapper(cdktf.stringToTerraform, false)(this._roleArns),
            session_policy: cdktf.stringToTerraform(this._sessionPolicy),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
        };
    }
}
exports.RolesanywhereProfile = RolesanywhereProfile;
_a = JSII_RTTI_SYMBOL_1;
RolesanywhereProfile[_a] = { fqn: "@cdktf/provider-aws.rolesanywhereProfile.RolesanywhereProfile", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
RolesanywhereProfile.tfResourceType = "aws_rolesanywhere_profile";
