// @cdktf/provider-aws/lib/ec2-availability-zone-group/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ec2AvailabilityZoneGroup = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/ec2_availability_zone_group aws_ec2_availability_zone_group}
*/
class Ec2AvailabilityZoneGroup extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ec2_availability_zone_group aws_ec2_availability_zone_group} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options Ec2AvailabilityZoneGroupConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ec2_availability_zone_group',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._groupName = config.groupName;
        this._id = config.id;
        this._optInStatus = config.optInStatus;
    }
    get groupName() {
        return this.getStringAttribute('group_name');
    }
    set groupName(value) {
        this._groupName = value;
    }
    // Temporarily expose input value. Use with caution.
    get groupNameInput() {
        return this._groupName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get optInStatus() {
        return this.getStringAttribute('opt_in_status');
    }
    set optInStatus(value) {
        this._optInStatus = value;
    }
    // Temporarily expose input value. Use with caution.
    get optInStatusInput() {
        return this._optInStatus;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            group_name: cdktf.stringToTerraform(this._groupName),
            id: cdktf.stringToTerraform(this._id),
            opt_in_status: cdktf.stringToTerraform(this._optInStatus),
        };
    }
}
exports.Ec2AvailabilityZoneGroup = Ec2AvailabilityZoneGroup;
_a = JSII_RTTI_SYMBOL_1;
Ec2AvailabilityZoneGroup[_a] = { fqn: "@cdktf/provider-aws.ec2AvailabilityZoneGroup.Ec2AvailabilityZoneGroup", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
Ec2AvailabilityZoneGroup.tfResourceType = "aws_ec2_availability_zone_group";
