// @cdktf/provider-aws/lib/route53-resolver-config/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Route53ResolverConfig = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/route53_resolver_config aws_route53_resolver_config}
*/
class Route53ResolverConfig extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/route53_resolver_config aws_route53_resolver_config} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options Route53ResolverConfigConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_route53_resolver_config',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._autodefinedReverseFlag = config.autodefinedReverseFlag;
        this._id = config.id;
        this._resourceId = config.resourceId;
    }
    get autodefinedReverseFlag() {
        return this.getStringAttribute('autodefined_reverse_flag');
    }
    set autodefinedReverseFlag(value) {
        this._autodefinedReverseFlag = value;
    }
    // Temporarily expose input value. Use with caution.
    get autodefinedReverseFlagInput() {
        return this._autodefinedReverseFlag;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // owner_id - computed: true, optional: false, required: false
    get ownerId() {
        return this.getStringAttribute('owner_id');
    }
    get resourceId() {
        return this.getStringAttribute('resource_id');
    }
    set resourceId(value) {
        this._resourceId = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourceIdInput() {
        return this._resourceId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            autodefined_reverse_flag: cdktf.stringToTerraform(this._autodefinedReverseFlag),
            id: cdktf.stringToTerraform(this._id),
            resource_id: cdktf.stringToTerraform(this._resourceId),
        };
    }
}
exports.Route53ResolverConfig = Route53ResolverConfig;
_a = JSII_RTTI_SYMBOL_1;
Route53ResolverConfig[_a] = { fqn: "@cdktf/provider-aws.route53ResolverConfig.Route53ResolverConfig", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
Route53ResolverConfig.tfResourceType = "aws_route53_resolver_config";
