// @cdktf/provider-aws/lib/ecr-registry-scanning-configuration/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EcrRegistryScanningConfiguration = exports.EcrRegistryScanningConfigurationRuleList = exports.EcrRegistryScanningConfigurationRuleOutputReference = exports.ecrRegistryScanningConfigurationRuleToTerraform = exports.EcrRegistryScanningConfigurationRuleRepositoryFilterList = exports.EcrRegistryScanningConfigurationRuleRepositoryFilterOutputReference = exports.ecrRegistryScanningConfigurationRuleRepositoryFilterToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function ecrRegistryScanningConfigurationRuleRepositoryFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        filter: cdktf.stringToTerraform(struct.filter),
        filter_type: cdktf.stringToTerraform(struct.filterType),
    };
}
exports.ecrRegistryScanningConfigurationRuleRepositoryFilterToTerraform = ecrRegistryScanningConfigurationRuleRepositoryFilterToTerraform;
class EcrRegistryScanningConfigurationRuleRepositoryFilterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._filter !== undefined) {
            hasAnyValues = true;
            internalValueResult.filter = this._filter;
        }
        if (this._filterType !== undefined) {
            hasAnyValues = true;
            internalValueResult.filterType = this._filterType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._filter = undefined;
            this._filterType = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._filter = value.filter;
            this._filterType = value.filterType;
        }
    }
    get filter() {
        return this.getStringAttribute('filter');
    }
    set filter(value) {
        this._filter = value;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter;
    }
    get filterType() {
        return this.getStringAttribute('filter_type');
    }
    set filterType(value) {
        this._filterType = value;
    }
    // Temporarily expose input value. Use with caution.
    get filterTypeInput() {
        return this._filterType;
    }
}
exports.EcrRegistryScanningConfigurationRuleRepositoryFilterOutputReference = EcrRegistryScanningConfigurationRuleRepositoryFilterOutputReference;
_a = JSII_RTTI_SYMBOL_1;
EcrRegistryScanningConfigurationRuleRepositoryFilterOutputReference[_a] = { fqn: "@cdktf/provider-aws.ecrRegistryScanningConfiguration.EcrRegistryScanningConfigurationRuleRepositoryFilterOutputReference", version: "10.0.12" };
class EcrRegistryScanningConfigurationRuleRepositoryFilterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EcrRegistryScanningConfigurationRuleRepositoryFilterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EcrRegistryScanningConfigurationRuleRepositoryFilterList = EcrRegistryScanningConfigurationRuleRepositoryFilterList;
_b = JSII_RTTI_SYMBOL_1;
EcrRegistryScanningConfigurationRuleRepositoryFilterList[_b] = { fqn: "@cdktf/provider-aws.ecrRegistryScanningConfiguration.EcrRegistryScanningConfigurationRuleRepositoryFilterList", version: "10.0.12" };
function ecrRegistryScanningConfigurationRuleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        scan_frequency: cdktf.stringToTerraform(struct.scanFrequency),
        repository_filter: cdktf.listMapper(ecrRegistryScanningConfigurationRuleRepositoryFilterToTerraform, true)(struct.repositoryFilter),
    };
}
exports.ecrRegistryScanningConfigurationRuleToTerraform = ecrRegistryScanningConfigurationRuleToTerraform;
class EcrRegistryScanningConfigurationRuleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // repository_filter - computed: false, optional: false, required: true
        this._repositoryFilter = new EcrRegistryScanningConfigurationRuleRepositoryFilterList(this, "repository_filter", true);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._scanFrequency !== undefined) {
            hasAnyValues = true;
            internalValueResult.scanFrequency = this._scanFrequency;
        }
        if (this._repositoryFilter?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.repositoryFilter = this._repositoryFilter?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._scanFrequency = undefined;
            this._repositoryFilter.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._scanFrequency = value.scanFrequency;
            this._repositoryFilter.internalValue = value.repositoryFilter;
        }
    }
    get scanFrequency() {
        return this.getStringAttribute('scan_frequency');
    }
    set scanFrequency(value) {
        this._scanFrequency = value;
    }
    // Temporarily expose input value. Use with caution.
    get scanFrequencyInput() {
        return this._scanFrequency;
    }
    get repositoryFilter() {
        return this._repositoryFilter;
    }
    putRepositoryFilter(value) {
        this._repositoryFilter.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryFilterInput() {
        return this._repositoryFilter.internalValue;
    }
}
exports.EcrRegistryScanningConfigurationRuleOutputReference = EcrRegistryScanningConfigurationRuleOutputReference;
_c = JSII_RTTI_SYMBOL_1;
EcrRegistryScanningConfigurationRuleOutputReference[_c] = { fqn: "@cdktf/provider-aws.ecrRegistryScanningConfiguration.EcrRegistryScanningConfigurationRuleOutputReference", version: "10.0.12" };
class EcrRegistryScanningConfigurationRuleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EcrRegistryScanningConfigurationRuleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EcrRegistryScanningConfigurationRuleList = EcrRegistryScanningConfigurationRuleList;
_d = JSII_RTTI_SYMBOL_1;
EcrRegistryScanningConfigurationRuleList[_d] = { fqn: "@cdktf/provider-aws.ecrRegistryScanningConfiguration.EcrRegistryScanningConfigurationRuleList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/ecr_registry_scanning_configuration aws_ecr_registry_scanning_configuration}
*/
class EcrRegistryScanningConfiguration extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ecr_registry_scanning_configuration aws_ecr_registry_scanning_configuration} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options EcrRegistryScanningConfigurationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ecr_registry_scanning_configuration',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // rule - computed: false, optional: true, required: false
        this._rule = new EcrRegistryScanningConfigurationRuleList(this, "rule", true);
        this._id = config.id;
        this._scanType = config.scanType;
        this._rule.internalValue = config.rule;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // registry_id - computed: true, optional: false, required: false
    get registryId() {
        return this.getStringAttribute('registry_id');
    }
    get scanType() {
        return this.getStringAttribute('scan_type');
    }
    set scanType(value) {
        this._scanType = value;
    }
    // Temporarily expose input value. Use with caution.
    get scanTypeInput() {
        return this._scanType;
    }
    get rule() {
        return this._rule;
    }
    putRule(value) {
        this._rule.internalValue = value;
    }
    resetRule() {
        this._rule.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ruleInput() {
        return this._rule.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            scan_type: cdktf.stringToTerraform(this._scanType),
            rule: cdktf.listMapper(ecrRegistryScanningConfigurationRuleToTerraform, true)(this._rule.internalValue),
        };
    }
}
exports.EcrRegistryScanningConfiguration = EcrRegistryScanningConfiguration;
_e = JSII_RTTI_SYMBOL_1;
EcrRegistryScanningConfiguration[_e] = { fqn: "@cdktf/provider-aws.ecrRegistryScanningConfiguration.EcrRegistryScanningConfiguration", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
EcrRegistryScanningConfiguration.tfResourceType = "aws_ecr_registry_scanning_configuration";
