// @cdktf/provider-aws/lib/vpn-connection-route/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VpnConnectionRoute = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/vpn_connection_route aws_vpn_connection_route}
*/
class VpnConnectionRoute extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/vpn_connection_route aws_vpn_connection_route} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options VpnConnectionRouteConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_vpn_connection_route',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._destinationCidrBlock = config.destinationCidrBlock;
        this._id = config.id;
        this._vpnConnectionId = config.vpnConnectionId;
    }
    get destinationCidrBlock() {
        return this.getStringAttribute('destination_cidr_block');
    }
    set destinationCidrBlock(value) {
        this._destinationCidrBlock = value;
    }
    // Temporarily expose input value. Use with caution.
    get destinationCidrBlockInput() {
        return this._destinationCidrBlock;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get vpnConnectionId() {
        return this.getStringAttribute('vpn_connection_id');
    }
    set vpnConnectionId(value) {
        this._vpnConnectionId = value;
    }
    // Temporarily expose input value. Use with caution.
    get vpnConnectionIdInput() {
        return this._vpnConnectionId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            destination_cidr_block: cdktf.stringToTerraform(this._destinationCidrBlock),
            id: cdktf.stringToTerraform(this._id),
            vpn_connection_id: cdktf.stringToTerraform(this._vpnConnectionId),
        };
    }
}
exports.VpnConnectionRoute = VpnConnectionRoute;
_a = JSII_RTTI_SYMBOL_1;
VpnConnectionRoute[_a] = { fqn: "@cdktf/provider-aws.vpnConnectionRoute.VpnConnectionRoute", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
VpnConnectionRoute.tfResourceType = "aws_vpn_connection_route";
