// @cdktf/provider-aws/lib/ec2-traffic-mirror-target/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ec2TrafficMirrorTarget = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/ec2_traffic_mirror_target aws_ec2_traffic_mirror_target}
*/
class Ec2TrafficMirrorTarget extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ec2_traffic_mirror_target aws_ec2_traffic_mirror_target} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options Ec2TrafficMirrorTargetConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_ec2_traffic_mirror_target',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._description = config.description;
        this._gatewayLoadBalancerEndpointId = config.gatewayLoadBalancerEndpointId;
        this._id = config.id;
        this._networkInterfaceId = config.networkInterfaceId;
        this._networkLoadBalancerArn = config.networkLoadBalancerArn;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get gatewayLoadBalancerEndpointId() {
        return this.getStringAttribute('gateway_load_balancer_endpoint_id');
    }
    set gatewayLoadBalancerEndpointId(value) {
        this._gatewayLoadBalancerEndpointId = value;
    }
    resetGatewayLoadBalancerEndpointId() {
        this._gatewayLoadBalancerEndpointId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get gatewayLoadBalancerEndpointIdInput() {
        return this._gatewayLoadBalancerEndpointId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get networkInterfaceId() {
        return this.getStringAttribute('network_interface_id');
    }
    set networkInterfaceId(value) {
        this._networkInterfaceId = value;
    }
    resetNetworkInterfaceId() {
        this._networkInterfaceId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkInterfaceIdInput() {
        return this._networkInterfaceId;
    }
    get networkLoadBalancerArn() {
        return this.getStringAttribute('network_load_balancer_arn');
    }
    set networkLoadBalancerArn(value) {
        this._networkLoadBalancerArn = value;
    }
    resetNetworkLoadBalancerArn() {
        this._networkLoadBalancerArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkLoadBalancerArnInput() {
        return this._networkLoadBalancerArn;
    }
    // owner_id - computed: true, optional: false, required: false
    get ownerId() {
        return this.getStringAttribute('owner_id');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            gateway_load_balancer_endpoint_id: cdktf.stringToTerraform(this._gatewayLoadBalancerEndpointId),
            id: cdktf.stringToTerraform(this._id),
            network_interface_id: cdktf.stringToTerraform(this._networkInterfaceId),
            network_load_balancer_arn: cdktf.stringToTerraform(this._networkLoadBalancerArn),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
        };
    }
}
exports.Ec2TrafficMirrorTarget = Ec2TrafficMirrorTarget;
_a = JSII_RTTI_SYMBOL_1;
Ec2TrafficMirrorTarget[_a] = { fqn: "@cdktf/provider-aws.ec2TrafficMirrorTarget.Ec2TrafficMirrorTarget", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
Ec2TrafficMirrorTarget.tfResourceType = "aws_ec2_traffic_mirror_target";
