// @cdktf/provider-aws/lib/data-aws-rds-reserved-instance-offering/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsRdsReservedInstanceOffering = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/rds_reserved_instance_offering aws_rds_reserved_instance_offering}
*/
class DataAwsRdsReservedInstanceOffering extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/rds_reserved_instance_offering aws_rds_reserved_instance_offering} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsRdsReservedInstanceOfferingConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_rds_reserved_instance_offering',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._dbInstanceClass = config.dbInstanceClass;
        this._duration = config.duration;
        this._id = config.id;
        this._multiAz = config.multiAz;
        this._offeringType = config.offeringType;
        this._productDescription = config.productDescription;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // currency_code - computed: true, optional: false, required: false
    get currencyCode() {
        return this.getStringAttribute('currency_code');
    }
    get dbInstanceClass() {
        return this.getStringAttribute('db_instance_class');
    }
    set dbInstanceClass(value) {
        this._dbInstanceClass = value;
    }
    // Temporarily expose input value. Use with caution.
    get dbInstanceClassInput() {
        return this._dbInstanceClass;
    }
    get duration() {
        return this.getNumberAttribute('duration');
    }
    set duration(value) {
        this._duration = value;
    }
    // Temporarily expose input value. Use with caution.
    get durationInput() {
        return this._duration;
    }
    // fixed_price - computed: true, optional: false, required: false
    get fixedPrice() {
        return this.getNumberAttribute('fixed_price');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get multiAz() {
        return this.getBooleanAttribute('multi_az');
    }
    set multiAz(value) {
        this._multiAz = value;
    }
    // Temporarily expose input value. Use with caution.
    get multiAzInput() {
        return this._multiAz;
    }
    // offering_id - computed: true, optional: false, required: false
    get offeringId() {
        return this.getStringAttribute('offering_id');
    }
    get offeringType() {
        return this.getStringAttribute('offering_type');
    }
    set offeringType(value) {
        this._offeringType = value;
    }
    // Temporarily expose input value. Use with caution.
    get offeringTypeInput() {
        return this._offeringType;
    }
    get productDescription() {
        return this.getStringAttribute('product_description');
    }
    set productDescription(value) {
        this._productDescription = value;
    }
    // Temporarily expose input value. Use with caution.
    get productDescriptionInput() {
        return this._productDescription;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            db_instance_class: cdktf.stringToTerraform(this._dbInstanceClass),
            duration: cdktf.numberToTerraform(this._duration),
            id: cdktf.stringToTerraform(this._id),
            multi_az: cdktf.booleanToTerraform(this._multiAz),
            offering_type: cdktf.stringToTerraform(this._offeringType),
            product_description: cdktf.stringToTerraform(this._productDescription),
        };
    }
}
exports.DataAwsRdsReservedInstanceOffering = DataAwsRdsReservedInstanceOffering;
_a = JSII_RTTI_SYMBOL_1;
DataAwsRdsReservedInstanceOffering[_a] = { fqn: "@cdktf/provider-aws.dataAwsRdsReservedInstanceOffering.DataAwsRdsReservedInstanceOffering", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsRdsReservedInstanceOffering.tfResourceType = "aws_rds_reserved_instance_offering";
