// @cdktf/provider-aws/lib/data-aws-ecs-cluster/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsEcsCluster = exports.DataAwsEcsClusterSettingList = exports.DataAwsEcsClusterSettingOutputReference = exports.dataAwsEcsClusterSettingToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsEcsClusterSettingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEcsClusterSettingToTerraform = dataAwsEcsClusterSettingToTerraform;
class DataAwsEcsClusterSettingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // value - computed: true, optional: false, required: false
    get value() {
        return this.getStringAttribute('value');
    }
}
exports.DataAwsEcsClusterSettingOutputReference = DataAwsEcsClusterSettingOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsEcsClusterSettingOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsEcsCluster.DataAwsEcsClusterSettingOutputReference", version: "10.0.12" };
class DataAwsEcsClusterSettingList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEcsClusterSettingOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEcsClusterSettingList = DataAwsEcsClusterSettingList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsEcsClusterSettingList[_b] = { fqn: "@cdktf/provider-aws.dataAwsEcsCluster.DataAwsEcsClusterSettingList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/ecs_cluster aws_ecs_cluster}
*/
class DataAwsEcsCluster extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/ecs_cluster aws_ecs_cluster} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsEcsClusterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ecs_cluster',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // setting - computed: true, optional: false, required: false
        this._setting = new DataAwsEcsClusterSettingList(this, "setting", true);
        this._clusterName = config.clusterName;
        this._id = config.id;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get clusterName() {
        return this.getStringAttribute('cluster_name');
    }
    set clusterName(value) {
        this._clusterName = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterNameInput() {
        return this._clusterName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // pending_tasks_count - computed: true, optional: false, required: false
    get pendingTasksCount() {
        return this.getNumberAttribute('pending_tasks_count');
    }
    // registered_container_instances_count - computed: true, optional: false, required: false
    get registeredContainerInstancesCount() {
        return this.getNumberAttribute('registered_container_instances_count');
    }
    // running_tasks_count - computed: true, optional: false, required: false
    get runningTasksCount() {
        return this.getNumberAttribute('running_tasks_count');
    }
    get setting() {
        return this._setting;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_name: cdktf.stringToTerraform(this._clusterName),
            id: cdktf.stringToTerraform(this._id),
        };
    }
}
exports.DataAwsEcsCluster = DataAwsEcsCluster;
_c = JSII_RTTI_SYMBOL_1;
DataAwsEcsCluster[_c] = { fqn: "@cdktf/provider-aws.dataAwsEcsCluster.DataAwsEcsCluster", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsEcsCluster.tfResourceType = "aws_ecs_cluster";
