// @cdktf/provider-aws/lib/cognito-user-pool-ui-customization/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CognitoUserPoolUiCustomization = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/cognito_user_pool_ui_customization aws_cognito_user_pool_ui_customization}
*/
class CognitoUserPoolUiCustomization extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/cognito_user_pool_ui_customization aws_cognito_user_pool_ui_customization} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CognitoUserPoolUiCustomizationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cognito_user_pool_ui_customization',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._clientId = config.clientId;
        this._css = config.css;
        this._id = config.id;
        this._imageFile = config.imageFile;
        this._userPoolId = config.userPoolId;
    }
    get clientId() {
        return this.getStringAttribute('client_id');
    }
    set clientId(value) {
        this._clientId = value;
    }
    resetClientId() {
        this._clientId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clientIdInput() {
        return this._clientId;
    }
    // creation_date - computed: true, optional: false, required: false
    get creationDate() {
        return this.getStringAttribute('creation_date');
    }
    get css() {
        return this.getStringAttribute('css');
    }
    set css(value) {
        this._css = value;
    }
    resetCss() {
        this._css = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cssInput() {
        return this._css;
    }
    // css_version - computed: true, optional: false, required: false
    get cssVersion() {
        return this.getStringAttribute('css_version');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get imageFile() {
        return this.getStringAttribute('image_file');
    }
    set imageFile(value) {
        this._imageFile = value;
    }
    resetImageFile() {
        this._imageFile = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get imageFileInput() {
        return this._imageFile;
    }
    // image_url - computed: true, optional: false, required: false
    get imageUrl() {
        return this.getStringAttribute('image_url');
    }
    // last_modified_date - computed: true, optional: false, required: false
    get lastModifiedDate() {
        return this.getStringAttribute('last_modified_date');
    }
    get userPoolId() {
        return this.getStringAttribute('user_pool_id');
    }
    set userPoolId(value) {
        this._userPoolId = value;
    }
    // Temporarily expose input value. Use with caution.
    get userPoolIdInput() {
        return this._userPoolId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            client_id: cdktf.stringToTerraform(this._clientId),
            css: cdktf.stringToTerraform(this._css),
            id: cdktf.stringToTerraform(this._id),
            image_file: cdktf.stringToTerraform(this._imageFile),
            user_pool_id: cdktf.stringToTerraform(this._userPoolId),
        };
    }
}
exports.CognitoUserPoolUiCustomization = CognitoUserPoolUiCustomization;
_a = JSII_RTTI_SYMBOL_1;
CognitoUserPoolUiCustomization[_a] = { fqn: "@cdktf/provider-aws.cognitoUserPoolUiCustomization.CognitoUserPoolUiCustomization", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
CognitoUserPoolUiCustomization.tfResourceType = "aws_cognito_user_pool_ui_customization";
