// @cdktf/provider-aws/lib/eks-cluster/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EksCluster = exports.EksClusterVpcConfigOutputReference = exports.eksClusterVpcConfigToTerraform = exports.EksClusterTimeoutsOutputReference = exports.eksClusterTimeoutsToTerraform = exports.EksClusterOutpostConfigOutputReference = exports.eksClusterOutpostConfigToTerraform = exports.EksClusterKubernetesNetworkConfigOutputReference = exports.eksClusterKubernetesNetworkConfigToTerraform = exports.EksClusterEncryptionConfigOutputReference = exports.eksClusterEncryptionConfigToTerraform = exports.EksClusterEncryptionConfigProviderOutputReference = exports.eksClusterEncryptionConfigProviderToTerraform = exports.EksClusterIdentityList = exports.EksClusterIdentityOutputReference = exports.eksClusterIdentityToTerraform = exports.EksClusterIdentityOidcList = exports.EksClusterIdentityOidcOutputReference = exports.eksClusterIdentityOidcToTerraform = exports.EksClusterCertificateAuthorityList = exports.EksClusterCertificateAuthorityOutputReference = exports.eksClusterCertificateAuthorityToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function eksClusterCertificateAuthorityToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.eksClusterCertificateAuthorityToTerraform = eksClusterCertificateAuthorityToTerraform;
class EksClusterCertificateAuthorityOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // data - computed: true, optional: false, required: false
    get data() {
        return this.getStringAttribute('data');
    }
}
exports.EksClusterCertificateAuthorityOutputReference = EksClusterCertificateAuthorityOutputReference;
_a = JSII_RTTI_SYMBOL_1;
EksClusterCertificateAuthorityOutputReference[_a] = { fqn: "@cdktf/provider-aws.eksCluster.EksClusterCertificateAuthorityOutputReference", version: "10.0.12" };
class EksClusterCertificateAuthorityList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EksClusterCertificateAuthorityOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EksClusterCertificateAuthorityList = EksClusterCertificateAuthorityList;
_b = JSII_RTTI_SYMBOL_1;
EksClusterCertificateAuthorityList[_b] = { fqn: "@cdktf/provider-aws.eksCluster.EksClusterCertificateAuthorityList", version: "10.0.12" };
function eksClusterIdentityOidcToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.eksClusterIdentityOidcToTerraform = eksClusterIdentityOidcToTerraform;
class EksClusterIdentityOidcOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // issuer - computed: true, optional: false, required: false
    get issuer() {
        return this.getStringAttribute('issuer');
    }
}
exports.EksClusterIdentityOidcOutputReference = EksClusterIdentityOidcOutputReference;
_c = JSII_RTTI_SYMBOL_1;
EksClusterIdentityOidcOutputReference[_c] = { fqn: "@cdktf/provider-aws.eksCluster.EksClusterIdentityOidcOutputReference", version: "10.0.12" };
class EksClusterIdentityOidcList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EksClusterIdentityOidcOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EksClusterIdentityOidcList = EksClusterIdentityOidcList;
_d = JSII_RTTI_SYMBOL_1;
EksClusterIdentityOidcList[_d] = { fqn: "@cdktf/provider-aws.eksCluster.EksClusterIdentityOidcList", version: "10.0.12" };
function eksClusterIdentityToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.eksClusterIdentityToTerraform = eksClusterIdentityToTerraform;
class EksClusterIdentityOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // oidc - computed: true, optional: false, required: false
        this._oidc = new EksClusterIdentityOidcList(this, "oidc", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get oidc() {
        return this._oidc;
    }
}
exports.EksClusterIdentityOutputReference = EksClusterIdentityOutputReference;
_e = JSII_RTTI_SYMBOL_1;
EksClusterIdentityOutputReference[_e] = { fqn: "@cdktf/provider-aws.eksCluster.EksClusterIdentityOutputReference", version: "10.0.12" };
class EksClusterIdentityList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EksClusterIdentityOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EksClusterIdentityList = EksClusterIdentityList;
_f = JSII_RTTI_SYMBOL_1;
EksClusterIdentityList[_f] = { fqn: "@cdktf/provider-aws.eksCluster.EksClusterIdentityList", version: "10.0.12" };
function eksClusterEncryptionConfigProviderToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key_arn: cdktf.stringToTerraform(struct.keyArn),
    };
}
exports.eksClusterEncryptionConfigProviderToTerraform = eksClusterEncryptionConfigProviderToTerraform;
class EksClusterEncryptionConfigProviderOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._keyArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.keyArn = this._keyArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._keyArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._keyArn = value.keyArn;
        }
    }
    get keyArn() {
        return this.getStringAttribute('key_arn');
    }
    set keyArn(value) {
        this._keyArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyArnInput() {
        return this._keyArn;
    }
}
exports.EksClusterEncryptionConfigProviderOutputReference = EksClusterEncryptionConfigProviderOutputReference;
_g = JSII_RTTI_SYMBOL_1;
EksClusterEncryptionConfigProviderOutputReference[_g] = { fqn: "@cdktf/provider-aws.eksCluster.EksClusterEncryptionConfigProviderOutputReference", version: "10.0.12" };
function eksClusterEncryptionConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resources: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.resources),
        provider: eksClusterEncryptionConfigProviderToTerraform(struct.provider),
    };
}
exports.eksClusterEncryptionConfigToTerraform = eksClusterEncryptionConfigToTerraform;
class EksClusterEncryptionConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // provider - computed: false, optional: false, required: true
        this._provider = new EksClusterEncryptionConfigProviderOutputReference(this, "provider");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resources !== undefined) {
            hasAnyValues = true;
            internalValueResult.resources = this._resources;
        }
        if (this._provider?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.provider = this._provider?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._resources = undefined;
            this._provider.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._resources = value.resources;
            this._provider.internalValue = value.provider;
        }
    }
    get resources() {
        return cdktf.Fn.tolist(this.getListAttribute('resources'));
    }
    set resources(value) {
        this._resources = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourcesInput() {
        return this._resources;
    }
    get provider() {
        return this._provider;
    }
    putProvider(value) {
        this._provider.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get providerInput() {
        return this._provider.internalValue;
    }
}
exports.EksClusterEncryptionConfigOutputReference = EksClusterEncryptionConfigOutputReference;
_h = JSII_RTTI_SYMBOL_1;
EksClusterEncryptionConfigOutputReference[_h] = { fqn: "@cdktf/provider-aws.eksCluster.EksClusterEncryptionConfigOutputReference", version: "10.0.12" };
function eksClusterKubernetesNetworkConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ip_family: cdktf.stringToTerraform(struct.ipFamily),
        service_ipv4_cidr: cdktf.stringToTerraform(struct.serviceIpv4Cidr),
    };
}
exports.eksClusterKubernetesNetworkConfigToTerraform = eksClusterKubernetesNetworkConfigToTerraform;
class EksClusterKubernetesNetworkConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._ipFamily !== undefined) {
            hasAnyValues = true;
            internalValueResult.ipFamily = this._ipFamily;
        }
        if (this._serviceIpv4Cidr !== undefined) {
            hasAnyValues = true;
            internalValueResult.serviceIpv4Cidr = this._serviceIpv4Cidr;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._ipFamily = undefined;
            this._serviceIpv4Cidr = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._ipFamily = value.ipFamily;
            this._serviceIpv4Cidr = value.serviceIpv4Cidr;
        }
    }
    get ipFamily() {
        return this.getStringAttribute('ip_family');
    }
    set ipFamily(value) {
        this._ipFamily = value;
    }
    resetIpFamily() {
        this._ipFamily = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipFamilyInput() {
        return this._ipFamily;
    }
    get serviceIpv4Cidr() {
        return this.getStringAttribute('service_ipv4_cidr');
    }
    set serviceIpv4Cidr(value) {
        this._serviceIpv4Cidr = value;
    }
    resetServiceIpv4Cidr() {
        this._serviceIpv4Cidr = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceIpv4CidrInput() {
        return this._serviceIpv4Cidr;
    }
    // service_ipv6_cidr - computed: true, optional: false, required: false
    get serviceIpv6Cidr() {
        return this.getStringAttribute('service_ipv6_cidr');
    }
}
exports.EksClusterKubernetesNetworkConfigOutputReference = EksClusterKubernetesNetworkConfigOutputReference;
_j = JSII_RTTI_SYMBOL_1;
EksClusterKubernetesNetworkConfigOutputReference[_j] = { fqn: "@cdktf/provider-aws.eksCluster.EksClusterKubernetesNetworkConfigOutputReference", version: "10.0.12" };
function eksClusterOutpostConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        control_plane_instance_type: cdktf.stringToTerraform(struct.controlPlaneInstanceType),
        outpost_arns: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.outpostArns),
    };
}
exports.eksClusterOutpostConfigToTerraform = eksClusterOutpostConfigToTerraform;
class EksClusterOutpostConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._controlPlaneInstanceType !== undefined) {
            hasAnyValues = true;
            internalValueResult.controlPlaneInstanceType = this._controlPlaneInstanceType;
        }
        if (this._outpostArns !== undefined) {
            hasAnyValues = true;
            internalValueResult.outpostArns = this._outpostArns;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._controlPlaneInstanceType = undefined;
            this._outpostArns = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._controlPlaneInstanceType = value.controlPlaneInstanceType;
            this._outpostArns = value.outpostArns;
        }
    }
    get controlPlaneInstanceType() {
        return this.getStringAttribute('control_plane_instance_type');
    }
    set controlPlaneInstanceType(value) {
        this._controlPlaneInstanceType = value;
    }
    // Temporarily expose input value. Use with caution.
    get controlPlaneInstanceTypeInput() {
        return this._controlPlaneInstanceType;
    }
    get outpostArns() {
        return cdktf.Fn.tolist(this.getListAttribute('outpost_arns'));
    }
    set outpostArns(value) {
        this._outpostArns = value;
    }
    // Temporarily expose input value. Use with caution.
    get outpostArnsInput() {
        return this._outpostArns;
    }
}
exports.EksClusterOutpostConfigOutputReference = EksClusterOutpostConfigOutputReference;
_k = JSII_RTTI_SYMBOL_1;
EksClusterOutpostConfigOutputReference[_k] = { fqn: "@cdktf/provider-aws.eksCluster.EksClusterOutpostConfigOutputReference", version: "10.0.12" };
function eksClusterTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.eksClusterTimeoutsToTerraform = eksClusterTimeoutsToTerraform;
class EksClusterTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.EksClusterTimeoutsOutputReference = EksClusterTimeoutsOutputReference;
_l = JSII_RTTI_SYMBOL_1;
EksClusterTimeoutsOutputReference[_l] = { fqn: "@cdktf/provider-aws.eksCluster.EksClusterTimeoutsOutputReference", version: "10.0.12" };
function eksClusterVpcConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        endpoint_private_access: cdktf.booleanToTerraform(struct.endpointPrivateAccess),
        endpoint_public_access: cdktf.booleanToTerraform(struct.endpointPublicAccess),
        public_access_cidrs: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.publicAccessCidrs),
        security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.securityGroupIds),
        subnet_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.subnetIds),
    };
}
exports.eksClusterVpcConfigToTerraform = eksClusterVpcConfigToTerraform;
class EksClusterVpcConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._endpointPrivateAccess !== undefined) {
            hasAnyValues = true;
            internalValueResult.endpointPrivateAccess = this._endpointPrivateAccess;
        }
        if (this._endpointPublicAccess !== undefined) {
            hasAnyValues = true;
            internalValueResult.endpointPublicAccess = this._endpointPublicAccess;
        }
        if (this._publicAccessCidrs !== undefined) {
            hasAnyValues = true;
            internalValueResult.publicAccessCidrs = this._publicAccessCidrs;
        }
        if (this._securityGroupIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroupIds = this._securityGroupIds;
        }
        if (this._subnetIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetIds = this._subnetIds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._endpointPrivateAccess = undefined;
            this._endpointPublicAccess = undefined;
            this._publicAccessCidrs = undefined;
            this._securityGroupIds = undefined;
            this._subnetIds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._endpointPrivateAccess = value.endpointPrivateAccess;
            this._endpointPublicAccess = value.endpointPublicAccess;
            this._publicAccessCidrs = value.publicAccessCidrs;
            this._securityGroupIds = value.securityGroupIds;
            this._subnetIds = value.subnetIds;
        }
    }
    // cluster_security_group_id - computed: true, optional: false, required: false
    get clusterSecurityGroupId() {
        return this.getStringAttribute('cluster_security_group_id');
    }
    get endpointPrivateAccess() {
        return this.getBooleanAttribute('endpoint_private_access');
    }
    set endpointPrivateAccess(value) {
        this._endpointPrivateAccess = value;
    }
    resetEndpointPrivateAccess() {
        this._endpointPrivateAccess = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endpointPrivateAccessInput() {
        return this._endpointPrivateAccess;
    }
    get endpointPublicAccess() {
        return this.getBooleanAttribute('endpoint_public_access');
    }
    set endpointPublicAccess(value) {
        this._endpointPublicAccess = value;
    }
    resetEndpointPublicAccess() {
        this._endpointPublicAccess = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endpointPublicAccessInput() {
        return this._endpointPublicAccess;
    }
    get publicAccessCidrs() {
        return cdktf.Fn.tolist(this.getListAttribute('public_access_cidrs'));
    }
    set publicAccessCidrs(value) {
        this._publicAccessCidrs = value;
    }
    resetPublicAccessCidrs() {
        this._publicAccessCidrs = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get publicAccessCidrsInput() {
        return this._publicAccessCidrs;
    }
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    resetSecurityGroupIds() {
        this._securityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    get subnetIds() {
        return cdktf.Fn.tolist(this.getListAttribute('subnet_ids'));
    }
    set subnetIds(value) {
        this._subnetIds = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdsInput() {
        return this._subnetIds;
    }
    // vpc_id - computed: true, optional: false, required: false
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
}
exports.EksClusterVpcConfigOutputReference = EksClusterVpcConfigOutputReference;
_m = JSII_RTTI_SYMBOL_1;
EksClusterVpcConfigOutputReference[_m] = { fqn: "@cdktf/provider-aws.eksCluster.EksClusterVpcConfigOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/eks_cluster aws_eks_cluster}
*/
class EksCluster extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/eks_cluster aws_eks_cluster} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options EksClusterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_eks_cluster',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // certificate_authority - computed: true, optional: false, required: false
        this._certificateAuthority = new EksClusterCertificateAuthorityList(this, "certificate_authority", false);
        // identity - computed: true, optional: false, required: false
        this._identity = new EksClusterIdentityList(this, "identity", false);
        // encryption_config - computed: false, optional: true, required: false
        this._encryptionConfig = new EksClusterEncryptionConfigOutputReference(this, "encryption_config");
        // kubernetes_network_config - computed: false, optional: true, required: false
        this._kubernetesNetworkConfig = new EksClusterKubernetesNetworkConfigOutputReference(this, "kubernetes_network_config");
        // outpost_config - computed: false, optional: true, required: false
        this._outpostConfig = new EksClusterOutpostConfigOutputReference(this, "outpost_config");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new EksClusterTimeoutsOutputReference(this, "timeouts");
        // vpc_config - computed: false, optional: false, required: true
        this._vpcConfig = new EksClusterVpcConfigOutputReference(this, "vpc_config");
        this._enabledClusterLogTypes = config.enabledClusterLogTypes;
        this._id = config.id;
        this._name = config.name;
        this._roleArn = config.roleArn;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._version = config.version;
        this._encryptionConfig.internalValue = config.encryptionConfig;
        this._kubernetesNetworkConfig.internalValue = config.kubernetesNetworkConfig;
        this._outpostConfig.internalValue = config.outpostConfig;
        this._timeouts.internalValue = config.timeouts;
        this._vpcConfig.internalValue = config.vpcConfig;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get certificateAuthority() {
        return this._certificateAuthority;
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    get enabledClusterLogTypes() {
        return cdktf.Fn.tolist(this.getListAttribute('enabled_cluster_log_types'));
    }
    set enabledClusterLogTypes(value) {
        this._enabledClusterLogTypes = value;
    }
    resetEnabledClusterLogTypes() {
        this._enabledClusterLogTypes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledClusterLogTypesInput() {
        return this._enabledClusterLogTypes;
    }
    // endpoint - computed: true, optional: false, required: false
    get endpoint() {
        return this.getStringAttribute('endpoint');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get identity() {
        return this._identity;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // platform_version - computed: true, optional: false, required: false
    get platformVersion() {
        return this.getStringAttribute('platform_version');
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    resetVersion() {
        this._version = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
    get encryptionConfig() {
        return this._encryptionConfig;
    }
    putEncryptionConfig(value) {
        this._encryptionConfig.internalValue = value;
    }
    resetEncryptionConfig() {
        this._encryptionConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encryptionConfigInput() {
        return this._encryptionConfig.internalValue;
    }
    get kubernetesNetworkConfig() {
        return this._kubernetesNetworkConfig;
    }
    putKubernetesNetworkConfig(value) {
        this._kubernetesNetworkConfig.internalValue = value;
    }
    resetKubernetesNetworkConfig() {
        this._kubernetesNetworkConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kubernetesNetworkConfigInput() {
        return this._kubernetesNetworkConfig.internalValue;
    }
    get outpostConfig() {
        return this._outpostConfig;
    }
    putOutpostConfig(value) {
        this._outpostConfig.internalValue = value;
    }
    resetOutpostConfig() {
        this._outpostConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get outpostConfigInput() {
        return this._outpostConfig.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get vpcConfig() {
        return this._vpcConfig;
    }
    putVpcConfig(value) {
        this._vpcConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get vpcConfigInput() {
        return this._vpcConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            enabled_cluster_log_types: cdktf.listMapper(cdktf.stringToTerraform, false)(this._enabledClusterLogTypes),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            role_arn: cdktf.stringToTerraform(this._roleArn),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            version: cdktf.stringToTerraform(this._version),
            encryption_config: eksClusterEncryptionConfigToTerraform(this._encryptionConfig.internalValue),
            kubernetes_network_config: eksClusterKubernetesNetworkConfigToTerraform(this._kubernetesNetworkConfig.internalValue),
            outpost_config: eksClusterOutpostConfigToTerraform(this._outpostConfig.internalValue),
            timeouts: eksClusterTimeoutsToTerraform(this._timeouts.internalValue),
            vpc_config: eksClusterVpcConfigToTerraform(this._vpcConfig.internalValue),
        };
    }
}
exports.EksCluster = EksCluster;
_o = JSII_RTTI_SYMBOL_1;
EksCluster[_o] = { fqn: "@cdktf/provider-aws.eksCluster.EksCluster", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
EksCluster.tfResourceType = "aws_eks_cluster";
