// @cdktf/provider-aws/lib/shield-protection-health-check-association/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShieldProtectionHealthCheckAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/shield_protection_health_check_association aws_shield_protection_health_check_association}
*/
class ShieldProtectionHealthCheckAssociation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/shield_protection_health_check_association aws_shield_protection_health_check_association} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ShieldProtectionHealthCheckAssociationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_shield_protection_health_check_association',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._healthCheckArn = config.healthCheckArn;
        this._id = config.id;
        this._shieldProtectionId = config.shieldProtectionId;
    }
    get healthCheckArn() {
        return this.getStringAttribute('health_check_arn');
    }
    set healthCheckArn(value) {
        this._healthCheckArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get healthCheckArnInput() {
        return this._healthCheckArn;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get shieldProtectionId() {
        return this.getStringAttribute('shield_protection_id');
    }
    set shieldProtectionId(value) {
        this._shieldProtectionId = value;
    }
    // Temporarily expose input value. Use with caution.
    get shieldProtectionIdInput() {
        return this._shieldProtectionId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            health_check_arn: cdktf.stringToTerraform(this._healthCheckArn),
            id: cdktf.stringToTerraform(this._id),
            shield_protection_id: cdktf.stringToTerraform(this._shieldProtectionId),
        };
    }
}
exports.ShieldProtectionHealthCheckAssociation = ShieldProtectionHealthCheckAssociation;
_a = JSII_RTTI_SYMBOL_1;
ShieldProtectionHealthCheckAssociation[_a] = { fqn: "@cdktf/provider-aws.shieldProtectionHealthCheckAssociation.ShieldProtectionHealthCheckAssociation", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
ShieldProtectionHealthCheckAssociation.tfResourceType = "aws_shield_protection_health_check_association";
