// @cdktf/provider-aws/lib/data-aws-vpc-ipam-pools/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsVpcIpamPools = exports.DataAwsVpcIpamPoolsFilterList = exports.DataAwsVpcIpamPoolsFilterOutputReference = exports.dataAwsVpcIpamPoolsFilterToTerraform = exports.DataAwsVpcIpamPoolsIpamPoolsList = exports.DataAwsVpcIpamPoolsIpamPoolsOutputReference = exports.dataAwsVpcIpamPoolsIpamPoolsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsVpcIpamPoolsIpamPoolsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsVpcIpamPoolsIpamPoolsToTerraform = dataAwsVpcIpamPoolsIpamPoolsToTerraform;
class DataAwsVpcIpamPoolsIpamPoolsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // allocation_resource_tags - computed: true, optional: false, required: false
        this._allocationResourceTags = new cdktf.StringMap(this, "allocation_resource_tags");
        // tags - computed: true, optional: false, required: false
        this._tags = new cdktf.StringMap(this, "tags");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // address_family - computed: true, optional: false, required: false
    get addressFamily() {
        return this.getStringAttribute('address_family');
    }
    // allocation_default_netmask_length - computed: true, optional: false, required: false
    get allocationDefaultNetmaskLength() {
        return this.getNumberAttribute('allocation_default_netmask_length');
    }
    // allocation_max_netmask_length - computed: true, optional: false, required: false
    get allocationMaxNetmaskLength() {
        return this.getNumberAttribute('allocation_max_netmask_length');
    }
    // allocation_min_netmask_length - computed: true, optional: false, required: false
    get allocationMinNetmaskLength() {
        return this.getNumberAttribute('allocation_min_netmask_length');
    }
    get allocationResourceTags() {
        return this._allocationResourceTags;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // auto_import - computed: true, optional: false, required: false
    get autoImport() {
        return this.getBooleanAttribute('auto_import');
    }
    // aws_service - computed: true, optional: false, required: false
    get awsService() {
        return this.getStringAttribute('aws_service');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // ipam_pool_id - computed: true, optional: false, required: false
    get ipamPoolId() {
        return this.getStringAttribute('ipam_pool_id');
    }
    // ipam_scope_id - computed: true, optional: false, required: false
    get ipamScopeId() {
        return this.getStringAttribute('ipam_scope_id');
    }
    // ipam_scope_type - computed: true, optional: false, required: false
    get ipamScopeType() {
        return this.getStringAttribute('ipam_scope_type');
    }
    // locale - computed: true, optional: false, required: false
    get locale() {
        return this.getStringAttribute('locale');
    }
    // pool_depth - computed: true, optional: false, required: false
    get poolDepth() {
        return this.getNumberAttribute('pool_depth');
    }
    // publicly_advertisable - computed: true, optional: false, required: false
    get publiclyAdvertisable() {
        return this.getBooleanAttribute('publicly_advertisable');
    }
    // source_ipam_pool_id - computed: true, optional: false, required: false
    get sourceIpamPoolId() {
        return this.getStringAttribute('source_ipam_pool_id');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get tags() {
        return this._tags;
    }
}
exports.DataAwsVpcIpamPoolsIpamPoolsOutputReference = DataAwsVpcIpamPoolsIpamPoolsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsVpcIpamPoolsIpamPoolsOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsVpcIpamPools.DataAwsVpcIpamPoolsIpamPoolsOutputReference", version: "10.0.12" };
class DataAwsVpcIpamPoolsIpamPoolsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsVpcIpamPoolsIpamPoolsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsVpcIpamPoolsIpamPoolsList = DataAwsVpcIpamPoolsIpamPoolsList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsVpcIpamPoolsIpamPoolsList[_b] = { fqn: "@cdktf/provider-aws.dataAwsVpcIpamPools.DataAwsVpcIpamPoolsIpamPoolsList", version: "10.0.12" };
function dataAwsVpcIpamPoolsFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.values),
    };
}
exports.dataAwsVpcIpamPoolsFilterToTerraform = dataAwsVpcIpamPoolsFilterToTerraform;
class DataAwsVpcIpamPoolsFilterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._values !== undefined) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._values = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._values = value.values;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get values() {
        return this.getListAttribute('values');
    }
    set values(value) {
        this._values = value;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
}
exports.DataAwsVpcIpamPoolsFilterOutputReference = DataAwsVpcIpamPoolsFilterOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsVpcIpamPoolsFilterOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsVpcIpamPools.DataAwsVpcIpamPoolsFilterOutputReference", version: "10.0.12" };
class DataAwsVpcIpamPoolsFilterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsVpcIpamPoolsFilterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsVpcIpamPoolsFilterList = DataAwsVpcIpamPoolsFilterList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsVpcIpamPoolsFilterList[_d] = { fqn: "@cdktf/provider-aws.dataAwsVpcIpamPools.DataAwsVpcIpamPoolsFilterList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/vpc_ipam_pools aws_vpc_ipam_pools}
*/
class DataAwsVpcIpamPools extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/vpc_ipam_pools aws_vpc_ipam_pools} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsVpcIpamPoolsConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_vpc_ipam_pools',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ipam_pools - computed: true, optional: false, required: false
        this._ipamPools = new DataAwsVpcIpamPoolsIpamPoolsList(this, "ipam_pools", true);
        // filter - computed: false, optional: true, required: false
        this._filter = new DataAwsVpcIpamPoolsFilterList(this, "filter", true);
        this._id = config.id;
        this._filter.internalValue = config.filter;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get ipamPools() {
        return this._ipamPools;
    }
    get filter() {
        return this._filter;
    }
    putFilter(value) {
        this._filter.internalValue = value;
    }
    resetFilter() {
        this._filter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            filter: cdktf.listMapper(dataAwsVpcIpamPoolsFilterToTerraform, true)(this._filter.internalValue),
        };
    }
}
exports.DataAwsVpcIpamPools = DataAwsVpcIpamPools;
_e = JSII_RTTI_SYMBOL_1;
DataAwsVpcIpamPools[_e] = { fqn: "@cdktf/provider-aws.dataAwsVpcIpamPools.DataAwsVpcIpamPools", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsVpcIpamPools.tfResourceType = "aws_vpc_ipam_pools";
