// @cdktf/provider-aws/lib/data-aws-efs-access-point/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsEfsAccessPoint = exports.DataAwsEfsAccessPointRootDirectoryList = exports.DataAwsEfsAccessPointRootDirectoryOutputReference = exports.dataAwsEfsAccessPointRootDirectoryToTerraform = exports.DataAwsEfsAccessPointRootDirectoryCreationInfoList = exports.DataAwsEfsAccessPointRootDirectoryCreationInfoOutputReference = exports.dataAwsEfsAccessPointRootDirectoryCreationInfoToTerraform = exports.DataAwsEfsAccessPointPosixUserList = exports.DataAwsEfsAccessPointPosixUserOutputReference = exports.dataAwsEfsAccessPointPosixUserToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsEfsAccessPointPosixUserToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEfsAccessPointPosixUserToTerraform = dataAwsEfsAccessPointPosixUserToTerraform;
class DataAwsEfsAccessPointPosixUserOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // gid - computed: true, optional: false, required: false
    get gid() {
        return this.getNumberAttribute('gid');
    }
    // secondary_gids - computed: true, optional: false, required: false
    get secondaryGids() {
        return cdktf.Token.asNumberList(cdktf.Fn.tolist(this.getNumberListAttribute('secondary_gids')));
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getNumberAttribute('uid');
    }
}
exports.DataAwsEfsAccessPointPosixUserOutputReference = DataAwsEfsAccessPointPosixUserOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsEfsAccessPointPosixUserOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsEfsAccessPoint.DataAwsEfsAccessPointPosixUserOutputReference", version: "10.0.12" };
class DataAwsEfsAccessPointPosixUserList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEfsAccessPointPosixUserOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEfsAccessPointPosixUserList = DataAwsEfsAccessPointPosixUserList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsEfsAccessPointPosixUserList[_b] = { fqn: "@cdktf/provider-aws.dataAwsEfsAccessPoint.DataAwsEfsAccessPointPosixUserList", version: "10.0.12" };
function dataAwsEfsAccessPointRootDirectoryCreationInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEfsAccessPointRootDirectoryCreationInfoToTerraform = dataAwsEfsAccessPointRootDirectoryCreationInfoToTerraform;
class DataAwsEfsAccessPointRootDirectoryCreationInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // owner_gid - computed: true, optional: false, required: false
    get ownerGid() {
        return this.getNumberAttribute('owner_gid');
    }
    // owner_uid - computed: true, optional: false, required: false
    get ownerUid() {
        return this.getNumberAttribute('owner_uid');
    }
    // permissions - computed: true, optional: false, required: false
    get permissions() {
        return this.getStringAttribute('permissions');
    }
}
exports.DataAwsEfsAccessPointRootDirectoryCreationInfoOutputReference = DataAwsEfsAccessPointRootDirectoryCreationInfoOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsEfsAccessPointRootDirectoryCreationInfoOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsEfsAccessPoint.DataAwsEfsAccessPointRootDirectoryCreationInfoOutputReference", version: "10.0.12" };
class DataAwsEfsAccessPointRootDirectoryCreationInfoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEfsAccessPointRootDirectoryCreationInfoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEfsAccessPointRootDirectoryCreationInfoList = DataAwsEfsAccessPointRootDirectoryCreationInfoList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsEfsAccessPointRootDirectoryCreationInfoList[_d] = { fqn: "@cdktf/provider-aws.dataAwsEfsAccessPoint.DataAwsEfsAccessPointRootDirectoryCreationInfoList", version: "10.0.12" };
function dataAwsEfsAccessPointRootDirectoryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEfsAccessPointRootDirectoryToTerraform = dataAwsEfsAccessPointRootDirectoryToTerraform;
class DataAwsEfsAccessPointRootDirectoryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // creation_info - computed: true, optional: false, required: false
        this._creationInfo = new DataAwsEfsAccessPointRootDirectoryCreationInfoList(this, "creation_info", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get creationInfo() {
        return this._creationInfo;
    }
    // path - computed: true, optional: false, required: false
    get path() {
        return this.getStringAttribute('path');
    }
}
exports.DataAwsEfsAccessPointRootDirectoryOutputReference = DataAwsEfsAccessPointRootDirectoryOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsEfsAccessPointRootDirectoryOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsEfsAccessPoint.DataAwsEfsAccessPointRootDirectoryOutputReference", version: "10.0.12" };
class DataAwsEfsAccessPointRootDirectoryList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEfsAccessPointRootDirectoryOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEfsAccessPointRootDirectoryList = DataAwsEfsAccessPointRootDirectoryList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsEfsAccessPointRootDirectoryList[_f] = { fqn: "@cdktf/provider-aws.dataAwsEfsAccessPoint.DataAwsEfsAccessPointRootDirectoryList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/efs_access_point aws_efs_access_point}
*/
class DataAwsEfsAccessPoint extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/efs_access_point aws_efs_access_point} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsEfsAccessPointConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_efs_access_point',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // posix_user - computed: true, optional: false, required: false
        this._posixUser = new DataAwsEfsAccessPointPosixUserList(this, "posix_user", false);
        // root_directory - computed: true, optional: false, required: false
        this._rootDirectory = new DataAwsEfsAccessPointRootDirectoryList(this, "root_directory", false);
        this._accessPointId = config.accessPointId;
        this._id = config.id;
        this._tags = config.tags;
    }
    get accessPointId() {
        return this.getStringAttribute('access_point_id');
    }
    set accessPointId(value) {
        this._accessPointId = value;
    }
    // Temporarily expose input value. Use with caution.
    get accessPointIdInput() {
        return this._accessPointId;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // file_system_arn - computed: true, optional: false, required: false
    get fileSystemArn() {
        return this.getStringAttribute('file_system_arn');
    }
    // file_system_id - computed: true, optional: false, required: false
    get fileSystemId() {
        return this.getStringAttribute('file_system_id');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // owner_id - computed: true, optional: false, required: false
    get ownerId() {
        return this.getStringAttribute('owner_id');
    }
    get posixUser() {
        return this._posixUser;
    }
    get rootDirectory() {
        return this._rootDirectory;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            access_point_id: cdktf.stringToTerraform(this._accessPointId),
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
}
exports.DataAwsEfsAccessPoint = DataAwsEfsAccessPoint;
_g = JSII_RTTI_SYMBOL_1;
DataAwsEfsAccessPoint[_g] = { fqn: "@cdktf/provider-aws.dataAwsEfsAccessPoint.DataAwsEfsAccessPoint", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsEfsAccessPoint.tfResourceType = "aws_efs_access_point";
