// @cdktf/provider-aws/lib/data-aws-dynamodb-table-item/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsDynamodbTableItem = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/dynamodb_table_item aws_dynamodb_table_item}
*/
class DataAwsDynamodbTableItem extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/dynamodb_table_item aws_dynamodb_table_item} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsDynamodbTableItemConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_dynamodb_table_item',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._expressionAttributeNames = config.expressionAttributeNames;
        this._id = config.id;
        this._key = config.key;
        this._projectionExpression = config.projectionExpression;
        this._tableName = config.tableName;
    }
    get expressionAttributeNames() {
        return this.getStringMapAttribute('expression_attribute_names');
    }
    set expressionAttributeNames(value) {
        this._expressionAttributeNames = value;
    }
    resetExpressionAttributeNames() {
        this._expressionAttributeNames = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get expressionAttributeNamesInput() {
        return this._expressionAttributeNames;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // item - computed: true, optional: false, required: false
    get item() {
        return this.getStringAttribute('item');
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get projectionExpression() {
        return this.getStringAttribute('projection_expression');
    }
    set projectionExpression(value) {
        this._projectionExpression = value;
    }
    resetProjectionExpression() {
        this._projectionExpression = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectionExpressionInput() {
        return this._projectionExpression;
    }
    get tableName() {
        return this.getStringAttribute('table_name');
    }
    set tableName(value) {
        this._tableName = value;
    }
    // Temporarily expose input value. Use with caution.
    get tableNameInput() {
        return this._tableName;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            expression_attribute_names: cdktf.hashMapper(cdktf.stringToTerraform)(this._expressionAttributeNames),
            id: cdktf.stringToTerraform(this._id),
            key: cdktf.stringToTerraform(this._key),
            projection_expression: cdktf.stringToTerraform(this._projectionExpression),
            table_name: cdktf.stringToTerraform(this._tableName),
        };
    }
}
exports.DataAwsDynamodbTableItem = DataAwsDynamodbTableItem;
_a = JSII_RTTI_SYMBOL_1;
DataAwsDynamodbTableItem[_a] = { fqn: "@cdktf/provider-aws.dataAwsDynamodbTableItem.DataAwsDynamodbTableItem", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsDynamodbTableItem.tfResourceType = "aws_dynamodb_table_item";
