// @cdktf/provider-aws/lib/waf-ipset/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.WafIpset = exports.WafIpsetIpSetDescriptorsList = exports.WafIpsetIpSetDescriptorsOutputReference = exports.wafIpsetIpSetDescriptorsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function wafIpsetIpSetDescriptorsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.wafIpsetIpSetDescriptorsToTerraform = wafIpsetIpSetDescriptorsToTerraform;
class WafIpsetIpSetDescriptorsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._type = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._type = value.type;
            this._value = value.value;
        }
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.WafIpsetIpSetDescriptorsOutputReference = WafIpsetIpSetDescriptorsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
WafIpsetIpSetDescriptorsOutputReference[_a] = { fqn: "@cdktf/provider-aws.wafIpset.WafIpsetIpSetDescriptorsOutputReference", version: "10.0.12" };
class WafIpsetIpSetDescriptorsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new WafIpsetIpSetDescriptorsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.WafIpsetIpSetDescriptorsList = WafIpsetIpSetDescriptorsList;
_b = JSII_RTTI_SYMBOL_1;
WafIpsetIpSetDescriptorsList[_b] = { fqn: "@cdktf/provider-aws.wafIpset.WafIpsetIpSetDescriptorsList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/waf_ipset aws_waf_ipset}
*/
class WafIpset extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/waf_ipset aws_waf_ipset} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options WafIpsetConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_waf_ipset',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ip_set_descriptors - computed: false, optional: true, required: false
        this._ipSetDescriptors = new WafIpsetIpSetDescriptorsList(this, "ip_set_descriptors", true);
        this._id = config.id;
        this._name = config.name;
        this._ipSetDescriptors.internalValue = config.ipSetDescriptors;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get ipSetDescriptors() {
        return this._ipSetDescriptors;
    }
    putIpSetDescriptors(value) {
        this._ipSetDescriptors.internalValue = value;
    }
    resetIpSetDescriptors() {
        this._ipSetDescriptors.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipSetDescriptorsInput() {
        return this._ipSetDescriptors.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            ip_set_descriptors: cdktf.listMapper(wafIpsetIpSetDescriptorsToTerraform, true)(this._ipSetDescriptors.internalValue),
        };
    }
}
exports.WafIpset = WafIpset;
_c = JSII_RTTI_SYMBOL_1;
WafIpset[_c] = { fqn: "@cdktf/provider-aws.wafIpset.WafIpset", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
WafIpset.tfResourceType = "aws_waf_ipset";
