// @cdktf/provider-aws/lib/data-aws-api-gateway-domain-name/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsApiGatewayDomainName = exports.DataAwsApiGatewayDomainNameEndpointConfigurationList = exports.DataAwsApiGatewayDomainNameEndpointConfigurationOutputReference = exports.dataAwsApiGatewayDomainNameEndpointConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsApiGatewayDomainNameEndpointConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsApiGatewayDomainNameEndpointConfigurationToTerraform = dataAwsApiGatewayDomainNameEndpointConfigurationToTerraform;
class DataAwsApiGatewayDomainNameEndpointConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // types - computed: true, optional: false, required: false
    get types() {
        return this.getListAttribute('types');
    }
}
exports.DataAwsApiGatewayDomainNameEndpointConfigurationOutputReference = DataAwsApiGatewayDomainNameEndpointConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsApiGatewayDomainNameEndpointConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsApiGatewayDomainName.DataAwsApiGatewayDomainNameEndpointConfigurationOutputReference", version: "10.0.12" };
class DataAwsApiGatewayDomainNameEndpointConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsApiGatewayDomainNameEndpointConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsApiGatewayDomainNameEndpointConfigurationList = DataAwsApiGatewayDomainNameEndpointConfigurationList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsApiGatewayDomainNameEndpointConfigurationList[_b] = { fqn: "@cdktf/provider-aws.dataAwsApiGatewayDomainName.DataAwsApiGatewayDomainNameEndpointConfigurationList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/api_gateway_domain_name aws_api_gateway_domain_name}
*/
class DataAwsApiGatewayDomainName extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/api_gateway_domain_name aws_api_gateway_domain_name} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsApiGatewayDomainNameConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_api_gateway_domain_name',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // endpoint_configuration - computed: true, optional: false, required: false
        this._endpointConfiguration = new DataAwsApiGatewayDomainNameEndpointConfigurationList(this, "endpoint_configuration", false);
        this._domainName = config.domainName;
        this._id = config.id;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // certificate_arn - computed: true, optional: false, required: false
    get certificateArn() {
        return this.getStringAttribute('certificate_arn');
    }
    // certificate_name - computed: true, optional: false, required: false
    get certificateName() {
        return this.getStringAttribute('certificate_name');
    }
    // certificate_upload_date - computed: true, optional: false, required: false
    get certificateUploadDate() {
        return this.getStringAttribute('certificate_upload_date');
    }
    // cloudfront_domain_name - computed: true, optional: false, required: false
    get cloudfrontDomainName() {
        return this.getStringAttribute('cloudfront_domain_name');
    }
    // cloudfront_zone_id - computed: true, optional: false, required: false
    get cloudfrontZoneId() {
        return this.getStringAttribute('cloudfront_zone_id');
    }
    get domainName() {
        return this.getStringAttribute('domain_name');
    }
    set domainName(value) {
        this._domainName = value;
    }
    // Temporarily expose input value. Use with caution.
    get domainNameInput() {
        return this._domainName;
    }
    get endpointConfiguration() {
        return this._endpointConfiguration;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // regional_certificate_arn - computed: true, optional: false, required: false
    get regionalCertificateArn() {
        return this.getStringAttribute('regional_certificate_arn');
    }
    // regional_certificate_name - computed: true, optional: false, required: false
    get regionalCertificateName() {
        return this.getStringAttribute('regional_certificate_name');
    }
    // regional_domain_name - computed: true, optional: false, required: false
    get regionalDomainName() {
        return this.getStringAttribute('regional_domain_name');
    }
    // regional_zone_id - computed: true, optional: false, required: false
    get regionalZoneId() {
        return this.getStringAttribute('regional_zone_id');
    }
    // security_policy - computed: true, optional: false, required: false
    get securityPolicy() {
        return this.getStringAttribute('security_policy');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            domain_name: cdktf.stringToTerraform(this._domainName),
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
}
exports.DataAwsApiGatewayDomainName = DataAwsApiGatewayDomainName;
_c = JSII_RTTI_SYMBOL_1;
DataAwsApiGatewayDomainName[_c] = { fqn: "@cdktf/provider-aws.dataAwsApiGatewayDomainName.DataAwsApiGatewayDomainName", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsApiGatewayDomainName.tfResourceType = "aws_api_gateway_domain_name";
