// @cdktf/provider-aws/lib/ses-configuration-set/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SesConfigurationSet = exports.SesConfigurationSetTrackingOptionsOutputReference = exports.sesConfigurationSetTrackingOptionsToTerraform = exports.SesConfigurationSetDeliveryOptionsOutputReference = exports.sesConfigurationSetDeliveryOptionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function sesConfigurationSetDeliveryOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        tls_policy: cdktf.stringToTerraform(struct.tlsPolicy),
    };
}
exports.sesConfigurationSetDeliveryOptionsToTerraform = sesConfigurationSetDeliveryOptionsToTerraform;
class SesConfigurationSetDeliveryOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._tlsPolicy !== undefined) {
            hasAnyValues = true;
            internalValueResult.tlsPolicy = this._tlsPolicy;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._tlsPolicy = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._tlsPolicy = value.tlsPolicy;
        }
    }
    get tlsPolicy() {
        return this.getStringAttribute('tls_policy');
    }
    set tlsPolicy(value) {
        this._tlsPolicy = value;
    }
    resetTlsPolicy() {
        this._tlsPolicy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tlsPolicyInput() {
        return this._tlsPolicy;
    }
}
exports.SesConfigurationSetDeliveryOptionsOutputReference = SesConfigurationSetDeliveryOptionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SesConfigurationSetDeliveryOptionsOutputReference[_a] = { fqn: "@cdktf/provider-aws.sesConfigurationSet.SesConfigurationSetDeliveryOptionsOutputReference", version: "10.0.12" };
function sesConfigurationSetTrackingOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        custom_redirect_domain: cdktf.stringToTerraform(struct.customRedirectDomain),
    };
}
exports.sesConfigurationSetTrackingOptionsToTerraform = sesConfigurationSetTrackingOptionsToTerraform;
class SesConfigurationSetTrackingOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._customRedirectDomain !== undefined) {
            hasAnyValues = true;
            internalValueResult.customRedirectDomain = this._customRedirectDomain;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._customRedirectDomain = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._customRedirectDomain = value.customRedirectDomain;
        }
    }
    get customRedirectDomain() {
        return this.getStringAttribute('custom_redirect_domain');
    }
    set customRedirectDomain(value) {
        this._customRedirectDomain = value;
    }
    resetCustomRedirectDomain() {
        this._customRedirectDomain = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get customRedirectDomainInput() {
        return this._customRedirectDomain;
    }
}
exports.SesConfigurationSetTrackingOptionsOutputReference = SesConfigurationSetTrackingOptionsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
SesConfigurationSetTrackingOptionsOutputReference[_b] = { fqn: "@cdktf/provider-aws.sesConfigurationSet.SesConfigurationSetTrackingOptionsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/ses_configuration_set aws_ses_configuration_set}
*/
class SesConfigurationSet extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ses_configuration_set aws_ses_configuration_set} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SesConfigurationSetConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ses_configuration_set',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // delivery_options - computed: false, optional: true, required: false
        this._deliveryOptions = new SesConfigurationSetDeliveryOptionsOutputReference(this, "delivery_options");
        // tracking_options - computed: false, optional: true, required: false
        this._trackingOptions = new SesConfigurationSetTrackingOptionsOutputReference(this, "tracking_options");
        this._id = config.id;
        this._name = config.name;
        this._reputationMetricsEnabled = config.reputationMetricsEnabled;
        this._sendingEnabled = config.sendingEnabled;
        this._deliveryOptions.internalValue = config.deliveryOptions;
        this._trackingOptions.internalValue = config.trackingOptions;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // last_fresh_start - computed: true, optional: false, required: false
    get lastFreshStart() {
        return this.getStringAttribute('last_fresh_start');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get reputationMetricsEnabled() {
        return this.getBooleanAttribute('reputation_metrics_enabled');
    }
    set reputationMetricsEnabled(value) {
        this._reputationMetricsEnabled = value;
    }
    resetReputationMetricsEnabled() {
        this._reputationMetricsEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get reputationMetricsEnabledInput() {
        return this._reputationMetricsEnabled;
    }
    get sendingEnabled() {
        return this.getBooleanAttribute('sending_enabled');
    }
    set sendingEnabled(value) {
        this._sendingEnabled = value;
    }
    resetSendingEnabled() {
        this._sendingEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sendingEnabledInput() {
        return this._sendingEnabled;
    }
    get deliveryOptions() {
        return this._deliveryOptions;
    }
    putDeliveryOptions(value) {
        this._deliveryOptions.internalValue = value;
    }
    resetDeliveryOptions() {
        this._deliveryOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deliveryOptionsInput() {
        return this._deliveryOptions.internalValue;
    }
    get trackingOptions() {
        return this._trackingOptions;
    }
    putTrackingOptions(value) {
        this._trackingOptions.internalValue = value;
    }
    resetTrackingOptions() {
        this._trackingOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get trackingOptionsInput() {
        return this._trackingOptions.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            reputation_metrics_enabled: cdktf.booleanToTerraform(this._reputationMetricsEnabled),
            sending_enabled: cdktf.booleanToTerraform(this._sendingEnabled),
            delivery_options: sesConfigurationSetDeliveryOptionsToTerraform(this._deliveryOptions.internalValue),
            tracking_options: sesConfigurationSetTrackingOptionsToTerraform(this._trackingOptions.internalValue),
        };
    }
}
exports.SesConfigurationSet = SesConfigurationSet;
_c = JSII_RTTI_SYMBOL_1;
SesConfigurationSet[_c] = { fqn: "@cdktf/provider-aws.sesConfigurationSet.SesConfigurationSet", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
SesConfigurationSet.tfResourceType = "aws_ses_configuration_set";
