// @cdktf/provider-aws/lib/redshiftserverless-workgroup/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedshiftserverlessWorkgroup = exports.RedshiftserverlessWorkgroupConfigParameterList = exports.RedshiftserverlessWorkgroupConfigParameterOutputReference = exports.redshiftserverlessWorkgroupConfigParameterToTerraform = exports.RedshiftserverlessWorkgroupEndpointList = exports.RedshiftserverlessWorkgroupEndpointOutputReference = exports.redshiftserverlessWorkgroupEndpointToTerraform = exports.RedshiftserverlessWorkgroupEndpointVpcEndpointList = exports.RedshiftserverlessWorkgroupEndpointVpcEndpointOutputReference = exports.redshiftserverlessWorkgroupEndpointVpcEndpointToTerraform = exports.RedshiftserverlessWorkgroupEndpointVpcEndpointNetworkInterfaceList = exports.RedshiftserverlessWorkgroupEndpointVpcEndpointNetworkInterfaceOutputReference = exports.redshiftserverlessWorkgroupEndpointVpcEndpointNetworkInterfaceToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function redshiftserverlessWorkgroupEndpointVpcEndpointNetworkInterfaceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.redshiftserverlessWorkgroupEndpointVpcEndpointNetworkInterfaceToTerraform = redshiftserverlessWorkgroupEndpointVpcEndpointNetworkInterfaceToTerraform;
class RedshiftserverlessWorkgroupEndpointVpcEndpointNetworkInterfaceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // availability_zone - computed: true, optional: false, required: false
    get availabilityZone() {
        return this.getStringAttribute('availability_zone');
    }
    // network_interface_id - computed: true, optional: false, required: false
    get networkInterfaceId() {
        return this.getStringAttribute('network_interface_id');
    }
    // private_ip_address - computed: true, optional: false, required: false
    get privateIpAddress() {
        return this.getStringAttribute('private_ip_address');
    }
    // subnet_id - computed: true, optional: false, required: false
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
}
exports.RedshiftserverlessWorkgroupEndpointVpcEndpointNetworkInterfaceOutputReference = RedshiftserverlessWorkgroupEndpointVpcEndpointNetworkInterfaceOutputReference;
_a = JSII_RTTI_SYMBOL_1;
RedshiftserverlessWorkgroupEndpointVpcEndpointNetworkInterfaceOutputReference[_a] = { fqn: "@cdktf/provider-aws.redshiftserverlessWorkgroup.RedshiftserverlessWorkgroupEndpointVpcEndpointNetworkInterfaceOutputReference", version: "10.0.12" };
class RedshiftserverlessWorkgroupEndpointVpcEndpointNetworkInterfaceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new RedshiftserverlessWorkgroupEndpointVpcEndpointNetworkInterfaceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.RedshiftserverlessWorkgroupEndpointVpcEndpointNetworkInterfaceList = RedshiftserverlessWorkgroupEndpointVpcEndpointNetworkInterfaceList;
_b = JSII_RTTI_SYMBOL_1;
RedshiftserverlessWorkgroupEndpointVpcEndpointNetworkInterfaceList[_b] = { fqn: "@cdktf/provider-aws.redshiftserverlessWorkgroup.RedshiftserverlessWorkgroupEndpointVpcEndpointNetworkInterfaceList", version: "10.0.12" };
function redshiftserverlessWorkgroupEndpointVpcEndpointToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.redshiftserverlessWorkgroupEndpointVpcEndpointToTerraform = redshiftserverlessWorkgroupEndpointVpcEndpointToTerraform;
class RedshiftserverlessWorkgroupEndpointVpcEndpointOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // network_interface - computed: true, optional: false, required: false
        this._networkInterface = new RedshiftserverlessWorkgroupEndpointVpcEndpointNetworkInterfaceList(this, "network_interface", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get networkInterface() {
        return this._networkInterface;
    }
    // vpc_endpoint_id - computed: true, optional: false, required: false
    get vpcEndpointId() {
        return this.getStringAttribute('vpc_endpoint_id');
    }
    // vpc_id - computed: true, optional: false, required: false
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
}
exports.RedshiftserverlessWorkgroupEndpointVpcEndpointOutputReference = RedshiftserverlessWorkgroupEndpointVpcEndpointOutputReference;
_c = JSII_RTTI_SYMBOL_1;
RedshiftserverlessWorkgroupEndpointVpcEndpointOutputReference[_c] = { fqn: "@cdktf/provider-aws.redshiftserverlessWorkgroup.RedshiftserverlessWorkgroupEndpointVpcEndpointOutputReference", version: "10.0.12" };
class RedshiftserverlessWorkgroupEndpointVpcEndpointList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new RedshiftserverlessWorkgroupEndpointVpcEndpointOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.RedshiftserverlessWorkgroupEndpointVpcEndpointList = RedshiftserverlessWorkgroupEndpointVpcEndpointList;
_d = JSII_RTTI_SYMBOL_1;
RedshiftserverlessWorkgroupEndpointVpcEndpointList[_d] = { fqn: "@cdktf/provider-aws.redshiftserverlessWorkgroup.RedshiftserverlessWorkgroupEndpointVpcEndpointList", version: "10.0.12" };
function redshiftserverlessWorkgroupEndpointToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.redshiftserverlessWorkgroupEndpointToTerraform = redshiftserverlessWorkgroupEndpointToTerraform;
class RedshiftserverlessWorkgroupEndpointOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // vpc_endpoint - computed: true, optional: false, required: false
        this._vpcEndpoint = new RedshiftserverlessWorkgroupEndpointVpcEndpointList(this, "vpc_endpoint", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // address - computed: true, optional: false, required: false
    get address() {
        return this.getStringAttribute('address');
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getNumberAttribute('port');
    }
    get vpcEndpoint() {
        return this._vpcEndpoint;
    }
}
exports.RedshiftserverlessWorkgroupEndpointOutputReference = RedshiftserverlessWorkgroupEndpointOutputReference;
_e = JSII_RTTI_SYMBOL_1;
RedshiftserverlessWorkgroupEndpointOutputReference[_e] = { fqn: "@cdktf/provider-aws.redshiftserverlessWorkgroup.RedshiftserverlessWorkgroupEndpointOutputReference", version: "10.0.12" };
class RedshiftserverlessWorkgroupEndpointList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new RedshiftserverlessWorkgroupEndpointOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.RedshiftserverlessWorkgroupEndpointList = RedshiftserverlessWorkgroupEndpointList;
_f = JSII_RTTI_SYMBOL_1;
RedshiftserverlessWorkgroupEndpointList[_f] = { fqn: "@cdktf/provider-aws.redshiftserverlessWorkgroup.RedshiftserverlessWorkgroupEndpointList", version: "10.0.12" };
function redshiftserverlessWorkgroupConfigParameterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        parameter_key: cdktf.stringToTerraform(struct.parameterKey),
        parameter_value: cdktf.stringToTerraform(struct.parameterValue),
    };
}
exports.redshiftserverlessWorkgroupConfigParameterToTerraform = redshiftserverlessWorkgroupConfigParameterToTerraform;
class RedshiftserverlessWorkgroupConfigParameterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._parameterKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.parameterKey = this._parameterKey;
        }
        if (this._parameterValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.parameterValue = this._parameterValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._parameterKey = undefined;
            this._parameterValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._parameterKey = value.parameterKey;
            this._parameterValue = value.parameterValue;
        }
    }
    get parameterKey() {
        return this.getStringAttribute('parameter_key');
    }
    set parameterKey(value) {
        this._parameterKey = value;
    }
    // Temporarily expose input value. Use with caution.
    get parameterKeyInput() {
        return this._parameterKey;
    }
    get parameterValue() {
        return this.getStringAttribute('parameter_value');
    }
    set parameterValue(value) {
        this._parameterValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get parameterValueInput() {
        return this._parameterValue;
    }
}
exports.RedshiftserverlessWorkgroupConfigParameterOutputReference = RedshiftserverlessWorkgroupConfigParameterOutputReference;
_g = JSII_RTTI_SYMBOL_1;
RedshiftserverlessWorkgroupConfigParameterOutputReference[_g] = { fqn: "@cdktf/provider-aws.redshiftserverlessWorkgroup.RedshiftserverlessWorkgroupConfigParameterOutputReference", version: "10.0.12" };
class RedshiftserverlessWorkgroupConfigParameterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new RedshiftserverlessWorkgroupConfigParameterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.RedshiftserverlessWorkgroupConfigParameterList = RedshiftserverlessWorkgroupConfigParameterList;
_h = JSII_RTTI_SYMBOL_1;
RedshiftserverlessWorkgroupConfigParameterList[_h] = { fqn: "@cdktf/provider-aws.redshiftserverlessWorkgroup.RedshiftserverlessWorkgroupConfigParameterList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/redshiftserverless_workgroup aws_redshiftserverless_workgroup}
*/
class RedshiftserverlessWorkgroup extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/redshiftserverless_workgroup aws_redshiftserverless_workgroup} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options RedshiftserverlessWorkgroupConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_redshiftserverless_workgroup',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // endpoint - computed: true, optional: false, required: false
        this._endpoint = new RedshiftserverlessWorkgroupEndpointList(this, "endpoint", false);
        // config_parameter - computed: false, optional: true, required: false
        this._configParameter = new RedshiftserverlessWorkgroupConfigParameterList(this, "config_parameter", false);
        this._baseCapacity = config.baseCapacity;
        this._enhancedVpcRouting = config.enhancedVpcRouting;
        this._id = config.id;
        this._namespaceName = config.namespaceName;
        this._publiclyAccessible = config.publiclyAccessible;
        this._securityGroupIds = config.securityGroupIds;
        this._subnetIds = config.subnetIds;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._workgroupName = config.workgroupName;
        this._configParameter.internalValue = config.configParameter;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get baseCapacity() {
        return this.getNumberAttribute('base_capacity');
    }
    set baseCapacity(value) {
        this._baseCapacity = value;
    }
    resetBaseCapacity() {
        this._baseCapacity = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get baseCapacityInput() {
        return this._baseCapacity;
    }
    get endpoint() {
        return this._endpoint;
    }
    get enhancedVpcRouting() {
        return this.getBooleanAttribute('enhanced_vpc_routing');
    }
    set enhancedVpcRouting(value) {
        this._enhancedVpcRouting = value;
    }
    resetEnhancedVpcRouting() {
        this._enhancedVpcRouting = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enhancedVpcRoutingInput() {
        return this._enhancedVpcRouting;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get namespaceName() {
        return this.getStringAttribute('namespace_name');
    }
    set namespaceName(value) {
        this._namespaceName = value;
    }
    // Temporarily expose input value. Use with caution.
    get namespaceNameInput() {
        return this._namespaceName;
    }
    get publiclyAccessible() {
        return this.getBooleanAttribute('publicly_accessible');
    }
    set publiclyAccessible(value) {
        this._publiclyAccessible = value;
    }
    resetPubliclyAccessible() {
        this._publiclyAccessible = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get publiclyAccessibleInput() {
        return this._publiclyAccessible;
    }
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    resetSecurityGroupIds() {
        this._securityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    get subnetIds() {
        return cdktf.Fn.tolist(this.getListAttribute('subnet_ids'));
    }
    set subnetIds(value) {
        this._subnetIds = value;
    }
    resetSubnetIds() {
        this._subnetIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdsInput() {
        return this._subnetIds;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // workgroup_id - computed: true, optional: false, required: false
    get workgroupId() {
        return this.getStringAttribute('workgroup_id');
    }
    get workgroupName() {
        return this.getStringAttribute('workgroup_name');
    }
    set workgroupName(value) {
        this._workgroupName = value;
    }
    // Temporarily expose input value. Use with caution.
    get workgroupNameInput() {
        return this._workgroupName;
    }
    get configParameter() {
        return this._configParameter;
    }
    putConfigParameter(value) {
        this._configParameter.internalValue = value;
    }
    resetConfigParameter() {
        this._configParameter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get configParameterInput() {
        return this._configParameter.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            base_capacity: cdktf.numberToTerraform(this._baseCapacity),
            enhanced_vpc_routing: cdktf.booleanToTerraform(this._enhancedVpcRouting),
            id: cdktf.stringToTerraform(this._id),
            namespace_name: cdktf.stringToTerraform(this._namespaceName),
            publicly_accessible: cdktf.booleanToTerraform(this._publiclyAccessible),
            security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(this._securityGroupIds),
            subnet_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(this._subnetIds),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            workgroup_name: cdktf.stringToTerraform(this._workgroupName),
            config_parameter: cdktf.listMapper(redshiftserverlessWorkgroupConfigParameterToTerraform, true)(this._configParameter.internalValue),
        };
    }
}
exports.RedshiftserverlessWorkgroup = RedshiftserverlessWorkgroup;
_j = JSII_RTTI_SYMBOL_1;
RedshiftserverlessWorkgroup[_j] = { fqn: "@cdktf/provider-aws.redshiftserverlessWorkgroup.RedshiftserverlessWorkgroup", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
RedshiftserverlessWorkgroup.tfResourceType = "aws_redshiftserverless_workgroup";
