// @cdktf/provider-aws/lib/data-aws-organizations-organization/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsOrganizationsOrganization = exports.DataAwsOrganizationsOrganizationRootsList = exports.DataAwsOrganizationsOrganizationRootsOutputReference = exports.dataAwsOrganizationsOrganizationRootsToTerraform = exports.DataAwsOrganizationsOrganizationRootsPolicyTypesList = exports.DataAwsOrganizationsOrganizationRootsPolicyTypesOutputReference = exports.dataAwsOrganizationsOrganizationRootsPolicyTypesToTerraform = exports.DataAwsOrganizationsOrganizationNonMasterAccountsList = exports.DataAwsOrganizationsOrganizationNonMasterAccountsOutputReference = exports.dataAwsOrganizationsOrganizationNonMasterAccountsToTerraform = exports.DataAwsOrganizationsOrganizationAccountsList = exports.DataAwsOrganizationsOrganizationAccountsOutputReference = exports.dataAwsOrganizationsOrganizationAccountsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsOrganizationsOrganizationAccountsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsOrganizationsOrganizationAccountsToTerraform = dataAwsOrganizationsOrganizationAccountsToTerraform;
class DataAwsOrganizationsOrganizationAccountsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // email - computed: true, optional: false, required: false
    get email() {
        return this.getStringAttribute('email');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
}
exports.DataAwsOrganizationsOrganizationAccountsOutputReference = DataAwsOrganizationsOrganizationAccountsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsOrganizationsOrganizationAccountsOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsOrganizationsOrganization.DataAwsOrganizationsOrganizationAccountsOutputReference", version: "10.0.12" };
class DataAwsOrganizationsOrganizationAccountsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsOrganizationsOrganizationAccountsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsOrganizationsOrganizationAccountsList = DataAwsOrganizationsOrganizationAccountsList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsOrganizationsOrganizationAccountsList[_b] = { fqn: "@cdktf/provider-aws.dataAwsOrganizationsOrganization.DataAwsOrganizationsOrganizationAccountsList", version: "10.0.12" };
function dataAwsOrganizationsOrganizationNonMasterAccountsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsOrganizationsOrganizationNonMasterAccountsToTerraform = dataAwsOrganizationsOrganizationNonMasterAccountsToTerraform;
class DataAwsOrganizationsOrganizationNonMasterAccountsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // email - computed: true, optional: false, required: false
    get email() {
        return this.getStringAttribute('email');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
}
exports.DataAwsOrganizationsOrganizationNonMasterAccountsOutputReference = DataAwsOrganizationsOrganizationNonMasterAccountsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsOrganizationsOrganizationNonMasterAccountsOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsOrganizationsOrganization.DataAwsOrganizationsOrganizationNonMasterAccountsOutputReference", version: "10.0.12" };
class DataAwsOrganizationsOrganizationNonMasterAccountsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsOrganizationsOrganizationNonMasterAccountsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsOrganizationsOrganizationNonMasterAccountsList = DataAwsOrganizationsOrganizationNonMasterAccountsList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsOrganizationsOrganizationNonMasterAccountsList[_d] = { fqn: "@cdktf/provider-aws.dataAwsOrganizationsOrganization.DataAwsOrganizationsOrganizationNonMasterAccountsList", version: "10.0.12" };
function dataAwsOrganizationsOrganizationRootsPolicyTypesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsOrganizationsOrganizationRootsPolicyTypesToTerraform = dataAwsOrganizationsOrganizationRootsPolicyTypesToTerraform;
class DataAwsOrganizationsOrganizationRootsPolicyTypesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataAwsOrganizationsOrganizationRootsPolicyTypesOutputReference = DataAwsOrganizationsOrganizationRootsPolicyTypesOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsOrganizationsOrganizationRootsPolicyTypesOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsOrganizationsOrganization.DataAwsOrganizationsOrganizationRootsPolicyTypesOutputReference", version: "10.0.12" };
class DataAwsOrganizationsOrganizationRootsPolicyTypesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsOrganizationsOrganizationRootsPolicyTypesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsOrganizationsOrganizationRootsPolicyTypesList = DataAwsOrganizationsOrganizationRootsPolicyTypesList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsOrganizationsOrganizationRootsPolicyTypesList[_f] = { fqn: "@cdktf/provider-aws.dataAwsOrganizationsOrganization.DataAwsOrganizationsOrganizationRootsPolicyTypesList", version: "10.0.12" };
function dataAwsOrganizationsOrganizationRootsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsOrganizationsOrganizationRootsToTerraform = dataAwsOrganizationsOrganizationRootsToTerraform;
class DataAwsOrganizationsOrganizationRootsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // policy_types - computed: true, optional: false, required: false
        this._policyTypes = new DataAwsOrganizationsOrganizationRootsPolicyTypesList(this, "policy_types", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get policyTypes() {
        return this._policyTypes;
    }
}
exports.DataAwsOrganizationsOrganizationRootsOutputReference = DataAwsOrganizationsOrganizationRootsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsOrganizationsOrganizationRootsOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsOrganizationsOrganization.DataAwsOrganizationsOrganizationRootsOutputReference", version: "10.0.12" };
class DataAwsOrganizationsOrganizationRootsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsOrganizationsOrganizationRootsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsOrganizationsOrganizationRootsList = DataAwsOrganizationsOrganizationRootsList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsOrganizationsOrganizationRootsList[_h] = { fqn: "@cdktf/provider-aws.dataAwsOrganizationsOrganization.DataAwsOrganizationsOrganizationRootsList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/organizations_organization aws_organizations_organization}
*/
class DataAwsOrganizationsOrganization extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/organizations_organization aws_organizations_organization} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsOrganizationsOrganizationConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_organizations_organization',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // accounts - computed: true, optional: false, required: false
        this._accounts = new DataAwsOrganizationsOrganizationAccountsList(this, "accounts", false);
        // non_master_accounts - computed: true, optional: false, required: false
        this._nonMasterAccounts = new DataAwsOrganizationsOrganizationNonMasterAccountsList(this, "non_master_accounts", false);
        // roots - computed: true, optional: false, required: false
        this._roots = new DataAwsOrganizationsOrganizationRootsList(this, "roots", false);
        this._id = config.id;
    }
    get accounts() {
        return this._accounts;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // aws_service_access_principals - computed: true, optional: false, required: false
    get awsServiceAccessPrincipals() {
        return cdktf.Fn.tolist(this.getListAttribute('aws_service_access_principals'));
    }
    // enabled_policy_types - computed: true, optional: false, required: false
    get enabledPolicyTypes() {
        return cdktf.Fn.tolist(this.getListAttribute('enabled_policy_types'));
    }
    // feature_set - computed: true, optional: false, required: false
    get featureSet() {
        return this.getStringAttribute('feature_set');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // master_account_arn - computed: true, optional: false, required: false
    get masterAccountArn() {
        return this.getStringAttribute('master_account_arn');
    }
    // master_account_email - computed: true, optional: false, required: false
    get masterAccountEmail() {
        return this.getStringAttribute('master_account_email');
    }
    // master_account_id - computed: true, optional: false, required: false
    get masterAccountId() {
        return this.getStringAttribute('master_account_id');
    }
    get nonMasterAccounts() {
        return this._nonMasterAccounts;
    }
    get roots() {
        return this._roots;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
        };
    }
}
exports.DataAwsOrganizationsOrganization = DataAwsOrganizationsOrganization;
_j = JSII_RTTI_SYMBOL_1;
DataAwsOrganizationsOrganization[_j] = { fqn: "@cdktf/provider-aws.dataAwsOrganizationsOrganization.DataAwsOrganizationsOrganization", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsOrganizationsOrganization.tfResourceType = "aws_organizations_organization";
