// @cdktf/provider-aws/lib/quicksight-user/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.QuicksightUser = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/quicksight_user aws_quicksight_user}
*/
class QuicksightUser extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/quicksight_user aws_quicksight_user} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options QuicksightUserConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_quicksight_user',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._awsAccountId = config.awsAccountId;
        this._email = config.email;
        this._iamArn = config.iamArn;
        this._id = config.id;
        this._identityType = config.identityType;
        this._namespace = config.namespace;
        this._sessionName = config.sessionName;
        this._userName = config.userName;
        this._userRole = config.userRole;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get awsAccountId() {
        return this.getStringAttribute('aws_account_id');
    }
    set awsAccountId(value) {
        this._awsAccountId = value;
    }
    resetAwsAccountId() {
        this._awsAccountId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get awsAccountIdInput() {
        return this._awsAccountId;
    }
    get email() {
        return this.getStringAttribute('email');
    }
    set email(value) {
        this._email = value;
    }
    // Temporarily expose input value. Use with caution.
    get emailInput() {
        return this._email;
    }
    get iamArn() {
        return this.getStringAttribute('iam_arn');
    }
    set iamArn(value) {
        this._iamArn = value;
    }
    resetIamArn() {
        this._iamArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iamArnInput() {
        return this._iamArn;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get identityType() {
        return this.getStringAttribute('identity_type');
    }
    set identityType(value) {
        this._identityType = value;
    }
    // Temporarily expose input value. Use with caution.
    get identityTypeInput() {
        return this._identityType;
    }
    get namespace() {
        return this.getStringAttribute('namespace');
    }
    set namespace(value) {
        this._namespace = value;
    }
    resetNamespace() {
        this._namespace = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get namespaceInput() {
        return this._namespace;
    }
    get sessionName() {
        return this.getStringAttribute('session_name');
    }
    set sessionName(value) {
        this._sessionName = value;
    }
    resetSessionName() {
        this._sessionName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sessionNameInput() {
        return this._sessionName;
    }
    get userName() {
        return this.getStringAttribute('user_name');
    }
    set userName(value) {
        this._userName = value;
    }
    resetUserName() {
        this._userName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get userNameInput() {
        return this._userName;
    }
    get userRole() {
        return this.getStringAttribute('user_role');
    }
    set userRole(value) {
        this._userRole = value;
    }
    // Temporarily expose input value. Use with caution.
    get userRoleInput() {
        return this._userRole;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            aws_account_id: cdktf.stringToTerraform(this._awsAccountId),
            email: cdktf.stringToTerraform(this._email),
            iam_arn: cdktf.stringToTerraform(this._iamArn),
            id: cdktf.stringToTerraform(this._id),
            identity_type: cdktf.stringToTerraform(this._identityType),
            namespace: cdktf.stringToTerraform(this._namespace),
            session_name: cdktf.stringToTerraform(this._sessionName),
            user_name: cdktf.stringToTerraform(this._userName),
            user_role: cdktf.stringToTerraform(this._userRole),
        };
    }
}
exports.QuicksightUser = QuicksightUser;
_a = JSII_RTTI_SYMBOL_1;
QuicksightUser[_a] = { fqn: "@cdktf/provider-aws.quicksightUser.QuicksightUser", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
QuicksightUser.tfResourceType = "aws_quicksight_user";
