// @cdktf/provider-aws/lib/data-aws-signer-signing-profile/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsSignerSigningProfile = exports.DataAwsSignerSigningProfileSignatureValidityPeriodList = exports.DataAwsSignerSigningProfileSignatureValidityPeriodOutputReference = exports.dataAwsSignerSigningProfileSignatureValidityPeriodToTerraform = exports.DataAwsSignerSigningProfileRevocationRecordList = exports.DataAwsSignerSigningProfileRevocationRecordOutputReference = exports.dataAwsSignerSigningProfileRevocationRecordToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsSignerSigningProfileRevocationRecordToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsSignerSigningProfileRevocationRecordToTerraform = dataAwsSignerSigningProfileRevocationRecordToTerraform;
class DataAwsSignerSigningProfileRevocationRecordOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // revocation_effective_from - computed: true, optional: false, required: false
    get revocationEffectiveFrom() {
        return this.getStringAttribute('revocation_effective_from');
    }
    // revoked_at - computed: true, optional: false, required: false
    get revokedAt() {
        return this.getStringAttribute('revoked_at');
    }
    // revoked_by - computed: true, optional: false, required: false
    get revokedBy() {
        return this.getStringAttribute('revoked_by');
    }
}
exports.DataAwsSignerSigningProfileRevocationRecordOutputReference = DataAwsSignerSigningProfileRevocationRecordOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsSignerSigningProfileRevocationRecordOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsSignerSigningProfile.DataAwsSignerSigningProfileRevocationRecordOutputReference", version: "10.0.12" };
class DataAwsSignerSigningProfileRevocationRecordList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsSignerSigningProfileRevocationRecordOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsSignerSigningProfileRevocationRecordList = DataAwsSignerSigningProfileRevocationRecordList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsSignerSigningProfileRevocationRecordList[_b] = { fqn: "@cdktf/provider-aws.dataAwsSignerSigningProfile.DataAwsSignerSigningProfileRevocationRecordList", version: "10.0.12" };
function dataAwsSignerSigningProfileSignatureValidityPeriodToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsSignerSigningProfileSignatureValidityPeriodToTerraform = dataAwsSignerSigningProfileSignatureValidityPeriodToTerraform;
class DataAwsSignerSigningProfileSignatureValidityPeriodOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
    // value - computed: true, optional: false, required: false
    get value() {
        return this.getNumberAttribute('value');
    }
}
exports.DataAwsSignerSigningProfileSignatureValidityPeriodOutputReference = DataAwsSignerSigningProfileSignatureValidityPeriodOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsSignerSigningProfileSignatureValidityPeriodOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsSignerSigningProfile.DataAwsSignerSigningProfileSignatureValidityPeriodOutputReference", version: "10.0.12" };
class DataAwsSignerSigningProfileSignatureValidityPeriodList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsSignerSigningProfileSignatureValidityPeriodOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsSignerSigningProfileSignatureValidityPeriodList = DataAwsSignerSigningProfileSignatureValidityPeriodList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsSignerSigningProfileSignatureValidityPeriodList[_d] = { fqn: "@cdktf/provider-aws.dataAwsSignerSigningProfile.DataAwsSignerSigningProfileSignatureValidityPeriodList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/signer_signing_profile aws_signer_signing_profile}
*/
class DataAwsSignerSigningProfile extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/signer_signing_profile aws_signer_signing_profile} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsSignerSigningProfileConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_signer_signing_profile',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // revocation_record - computed: true, optional: false, required: false
        this._revocationRecord = new DataAwsSignerSigningProfileRevocationRecordList(this, "revocation_record", false);
        // signature_validity_period - computed: true, optional: false, required: false
        this._signatureValidityPeriod = new DataAwsSignerSigningProfileSignatureValidityPeriodList(this, "signature_validity_period", false);
        this._id = config.id;
        this._name = config.name;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // platform_display_name - computed: true, optional: false, required: false
    get platformDisplayName() {
        return this.getStringAttribute('platform_display_name');
    }
    // platform_id - computed: true, optional: false, required: false
    get platformId() {
        return this.getStringAttribute('platform_id');
    }
    get revocationRecord() {
        return this._revocationRecord;
    }
    get signatureValidityPeriod() {
        return this._signatureValidityPeriod;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
    // version_arn - computed: true, optional: false, required: false
    get versionArn() {
        return this.getStringAttribute('version_arn');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
}
exports.DataAwsSignerSigningProfile = DataAwsSignerSigningProfile;
_e = JSII_RTTI_SYMBOL_1;
DataAwsSignerSigningProfile[_e] = { fqn: "@cdktf/provider-aws.dataAwsSignerSigningProfile.DataAwsSignerSigningProfile", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsSignerSigningProfile.tfResourceType = "aws_signer_signing_profile";
