// @cdktf/provider-aws/lib/appmesh-route/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppmeshRoute = exports.AppmeshRouteSpecOutputReference = exports.appmeshRouteSpecToTerraform = exports.AppmeshRouteSpecTcpRouteOutputReference = exports.appmeshRouteSpecTcpRouteToTerraform = exports.AppmeshRouteSpecTcpRouteTimeoutOutputReference = exports.appmeshRouteSpecTcpRouteTimeoutToTerraform = exports.AppmeshRouteSpecTcpRouteTimeoutIdleOutputReference = exports.appmeshRouteSpecTcpRouteTimeoutIdleToTerraform = exports.AppmeshRouteSpecTcpRouteActionOutputReference = exports.appmeshRouteSpecTcpRouteActionToTerraform = exports.AppmeshRouteSpecTcpRouteActionWeightedTargetList = exports.AppmeshRouteSpecTcpRouteActionWeightedTargetOutputReference = exports.appmeshRouteSpecTcpRouteActionWeightedTargetToTerraform = exports.AppmeshRouteSpecHttpRouteOutputReference = exports.appmeshRouteSpecHttpRouteToTerraform = exports.AppmeshRouteSpecHttpRouteTimeoutOutputReference = exports.appmeshRouteSpecHttpRouteTimeoutToTerraform = exports.AppmeshRouteSpecHttpRouteTimeoutPerRequestOutputReference = exports.appmeshRouteSpecHttpRouteTimeoutPerRequestToTerraform = exports.AppmeshRouteSpecHttpRouteTimeoutIdleOutputReference = exports.appmeshRouteSpecHttpRouteTimeoutIdleToTerraform = exports.AppmeshRouteSpecHttpRouteRetryPolicyOutputReference = exports.appmeshRouteSpecHttpRouteRetryPolicyToTerraform = exports.AppmeshRouteSpecHttpRouteRetryPolicyPerRetryTimeoutOutputReference = exports.appmeshRouteSpecHttpRouteRetryPolicyPerRetryTimeoutToTerraform = exports.AppmeshRouteSpecHttpRouteMatchOutputReference = exports.appmeshRouteSpecHttpRouteMatchToTerraform = exports.AppmeshRouteSpecHttpRouteMatchHeaderList = exports.AppmeshRouteSpecHttpRouteMatchHeaderOutputReference = exports.appmeshRouteSpecHttpRouteMatchHeaderToTerraform = exports.AppmeshRouteSpecHttpRouteMatchHeaderMatchOutputReference = exports.appmeshRouteSpecHttpRouteMatchHeaderMatchToTerraform = exports.AppmeshRouteSpecHttpRouteMatchHeaderMatchRangeOutputReference = exports.appmeshRouteSpecHttpRouteMatchHeaderMatchRangeToTerraform = exports.AppmeshRouteSpecHttpRouteActionOutputReference = exports.appmeshRouteSpecHttpRouteActionToTerraform = exports.AppmeshRouteSpecHttpRouteActionWeightedTargetList = exports.AppmeshRouteSpecHttpRouteActionWeightedTargetOutputReference = exports.appmeshRouteSpecHttpRouteActionWeightedTargetToTerraform = exports.AppmeshRouteSpecHttp2RouteOutputReference = exports.appmeshRouteSpecHttp2RouteToTerraform = exports.AppmeshRouteSpecHttp2RouteTimeoutOutputReference = exports.appmeshRouteSpecHttp2RouteTimeoutToTerraform = exports.AppmeshRouteSpecHttp2RouteTimeoutPerRequestOutputReference = exports.appmeshRouteSpecHttp2RouteTimeoutPerRequestToTerraform = exports.AppmeshRouteSpecHttp2RouteTimeoutIdleOutputReference = exports.appmeshRouteSpecHttp2RouteTimeoutIdleToTerraform = exports.AppmeshRouteSpecHttp2RouteRetryPolicyOutputReference = exports.appmeshRouteSpecHttp2RouteRetryPolicyToTerraform = exports.AppmeshRouteSpecHttp2RouteRetryPolicyPerRetryTimeoutOutputReference = exports.appmeshRouteSpecHttp2RouteRetryPolicyPerRetryTimeoutToTerraform = exports.AppmeshRouteSpecHttp2RouteMatchOutputReference = exports.appmeshRouteSpecHttp2RouteMatchToTerraform = exports.AppmeshRouteSpecHttp2RouteMatchHeaderList = exports.AppmeshRouteSpecHttp2RouteMatchHeaderOutputReference = exports.appmeshRouteSpecHttp2RouteMatchHeaderToTerraform = exports.AppmeshRouteSpecHttp2RouteMatchHeaderMatchOutputReference = exports.appmeshRouteSpecHttp2RouteMatchHeaderMatchToTerraform = exports.AppmeshRouteSpecHttp2RouteMatchHeaderMatchRangeOutputReference = exports.appmeshRouteSpecHttp2RouteMatchHeaderMatchRangeToTerraform = exports.AppmeshRouteSpecHttp2RouteActionOutputReference = exports.appmeshRouteSpecHttp2RouteActionToTerraform = exports.AppmeshRouteSpecHttp2RouteActionWeightedTargetList = exports.AppmeshRouteSpecHttp2RouteActionWeightedTargetOutputReference = exports.appmeshRouteSpecHttp2RouteActionWeightedTargetToTerraform = exports.AppmeshRouteSpecGrpcRouteOutputReference = exports.appmeshRouteSpecGrpcRouteToTerraform = exports.AppmeshRouteSpecGrpcRouteTimeoutOutputReference = exports.appmeshRouteSpecGrpcRouteTimeoutToTerraform = exports.AppmeshRouteSpecGrpcRouteTimeoutPerRequestOutputReference = exports.appmeshRouteSpecGrpcRouteTimeoutPerRequestToTerraform = exports.AppmeshRouteSpecGrpcRouteTimeoutIdleOutputReference = exports.appmeshRouteSpecGrpcRouteTimeoutIdleToTerraform = exports.AppmeshRouteSpecGrpcRouteRetryPolicyOutputReference = exports.appmeshRouteSpecGrpcRouteRetryPolicyToTerraform = exports.AppmeshRouteSpecGrpcRouteRetryPolicyPerRetryTimeoutOutputReference = exports.appmeshRouteSpecGrpcRouteRetryPolicyPerRetryTimeoutToTerraform = exports.AppmeshRouteSpecGrpcRouteMatchOutputReference = exports.appmeshRouteSpecGrpcRouteMatchToTerraform = exports.AppmeshRouteSpecGrpcRouteMatchMetadataList = exports.AppmeshRouteSpecGrpcRouteMatchMetadataOutputReference = exports.appmeshRouteSpecGrpcRouteMatchMetadataToTerraform = exports.AppmeshRouteSpecGrpcRouteMatchMetadataMatchOutputReference = exports.appmeshRouteSpecGrpcRouteMatchMetadataMatchToTerraform = exports.AppmeshRouteSpecGrpcRouteMatchMetadataMatchRangeOutputReference = exports.appmeshRouteSpecGrpcRouteMatchMetadataMatchRangeToTerraform = exports.AppmeshRouteSpecGrpcRouteActionOutputReference = exports.appmeshRouteSpecGrpcRouteActionToTerraform = exports.AppmeshRouteSpecGrpcRouteActionWeightedTargetList = exports.AppmeshRouteSpecGrpcRouteActionWeightedTargetOutputReference = exports.appmeshRouteSpecGrpcRouteActionWeightedTargetToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function appmeshRouteSpecGrpcRouteActionWeightedTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        virtual_node: cdktf.stringToTerraform(struct.virtualNode),
        weight: cdktf.numberToTerraform(struct.weight),
    };
}
exports.appmeshRouteSpecGrpcRouteActionWeightedTargetToTerraform = appmeshRouteSpecGrpcRouteActionWeightedTargetToTerraform;
class AppmeshRouteSpecGrpcRouteActionWeightedTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._virtualNode !== undefined) {
            hasAnyValues = true;
            internalValueResult.virtualNode = this._virtualNode;
        }
        if (this._weight !== undefined) {
            hasAnyValues = true;
            internalValueResult.weight = this._weight;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._virtualNode = undefined;
            this._weight = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._virtualNode = value.virtualNode;
            this._weight = value.weight;
        }
    }
    get virtualNode() {
        return this.getStringAttribute('virtual_node');
    }
    set virtualNode(value) {
        this._virtualNode = value;
    }
    // Temporarily expose input value. Use with caution.
    get virtualNodeInput() {
        return this._virtualNode;
    }
    get weight() {
        return this.getNumberAttribute('weight');
    }
    set weight(value) {
        this._weight = value;
    }
    // Temporarily expose input value. Use with caution.
    get weightInput() {
        return this._weight;
    }
}
exports.AppmeshRouteSpecGrpcRouteActionWeightedTargetOutputReference = AppmeshRouteSpecGrpcRouteActionWeightedTargetOutputReference;
_a = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecGrpcRouteActionWeightedTargetOutputReference[_a] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecGrpcRouteActionWeightedTargetOutputReference", version: "10.0.12" };
class AppmeshRouteSpecGrpcRouteActionWeightedTargetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AppmeshRouteSpecGrpcRouteActionWeightedTargetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AppmeshRouteSpecGrpcRouteActionWeightedTargetList = AppmeshRouteSpecGrpcRouteActionWeightedTargetList;
_b = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecGrpcRouteActionWeightedTargetList[_b] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecGrpcRouteActionWeightedTargetList", version: "10.0.12" };
function appmeshRouteSpecGrpcRouteActionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        weighted_target: cdktf.listMapper(appmeshRouteSpecGrpcRouteActionWeightedTargetToTerraform, true)(struct.weightedTarget),
    };
}
exports.appmeshRouteSpecGrpcRouteActionToTerraform = appmeshRouteSpecGrpcRouteActionToTerraform;
class AppmeshRouteSpecGrpcRouteActionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // weighted_target - computed: false, optional: false, required: true
        this._weightedTarget = new AppmeshRouteSpecGrpcRouteActionWeightedTargetList(this, "weighted_target", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._weightedTarget?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.weightedTarget = this._weightedTarget?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._weightedTarget.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._weightedTarget.internalValue = value.weightedTarget;
        }
    }
    get weightedTarget() {
        return this._weightedTarget;
    }
    putWeightedTarget(value) {
        this._weightedTarget.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get weightedTargetInput() {
        return this._weightedTarget.internalValue;
    }
}
exports.AppmeshRouteSpecGrpcRouteActionOutputReference = AppmeshRouteSpecGrpcRouteActionOutputReference;
_c = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecGrpcRouteActionOutputReference[_c] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecGrpcRouteActionOutputReference", version: "10.0.12" };
function appmeshRouteSpecGrpcRouteMatchMetadataMatchRangeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        end: cdktf.numberToTerraform(struct.end),
        start: cdktf.numberToTerraform(struct.start),
    };
}
exports.appmeshRouteSpecGrpcRouteMatchMetadataMatchRangeToTerraform = appmeshRouteSpecGrpcRouteMatchMetadataMatchRangeToTerraform;
class AppmeshRouteSpecGrpcRouteMatchMetadataMatchRangeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._end !== undefined) {
            hasAnyValues = true;
            internalValueResult.end = this._end;
        }
        if (this._start !== undefined) {
            hasAnyValues = true;
            internalValueResult.start = this._start;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._end = undefined;
            this._start = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._end = value.end;
            this._start = value.start;
        }
    }
    get end() {
        return this.getNumberAttribute('end');
    }
    set end(value) {
        this._end = value;
    }
    // Temporarily expose input value. Use with caution.
    get endInput() {
        return this._end;
    }
    get start() {
        return this.getNumberAttribute('start');
    }
    set start(value) {
        this._start = value;
    }
    // Temporarily expose input value. Use with caution.
    get startInput() {
        return this._start;
    }
}
exports.AppmeshRouteSpecGrpcRouteMatchMetadataMatchRangeOutputReference = AppmeshRouteSpecGrpcRouteMatchMetadataMatchRangeOutputReference;
_d = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecGrpcRouteMatchMetadataMatchRangeOutputReference[_d] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecGrpcRouteMatchMetadataMatchRangeOutputReference", version: "10.0.12" };
function appmeshRouteSpecGrpcRouteMatchMetadataMatchToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        exact: cdktf.stringToTerraform(struct.exact),
        prefix: cdktf.stringToTerraform(struct.prefix),
        regex: cdktf.stringToTerraform(struct.regex),
        suffix: cdktf.stringToTerraform(struct.suffix),
        range: appmeshRouteSpecGrpcRouteMatchMetadataMatchRangeToTerraform(struct.range),
    };
}
exports.appmeshRouteSpecGrpcRouteMatchMetadataMatchToTerraform = appmeshRouteSpecGrpcRouteMatchMetadataMatchToTerraform;
class AppmeshRouteSpecGrpcRouteMatchMetadataMatchOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // range - computed: false, optional: true, required: false
        this._range = new AppmeshRouteSpecGrpcRouteMatchMetadataMatchRangeOutputReference(this, "range");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._exact !== undefined) {
            hasAnyValues = true;
            internalValueResult.exact = this._exact;
        }
        if (this._prefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        if (this._regex !== undefined) {
            hasAnyValues = true;
            internalValueResult.regex = this._regex;
        }
        if (this._suffix !== undefined) {
            hasAnyValues = true;
            internalValueResult.suffix = this._suffix;
        }
        if (this._range?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.range = this._range?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._exact = undefined;
            this._prefix = undefined;
            this._regex = undefined;
            this._suffix = undefined;
            this._range.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._exact = value.exact;
            this._prefix = value.prefix;
            this._regex = value.regex;
            this._suffix = value.suffix;
            this._range.internalValue = value.range;
        }
    }
    get exact() {
        return this.getStringAttribute('exact');
    }
    set exact(value) {
        this._exact = value;
    }
    resetExact() {
        this._exact = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exactInput() {
        return this._exact;
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
    get regex() {
        return this.getStringAttribute('regex');
    }
    set regex(value) {
        this._regex = value;
    }
    resetRegex() {
        this._regex = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regexInput() {
        return this._regex;
    }
    get suffix() {
        return this.getStringAttribute('suffix');
    }
    set suffix(value) {
        this._suffix = value;
    }
    resetSuffix() {
        this._suffix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get suffixInput() {
        return this._suffix;
    }
    get range() {
        return this._range;
    }
    putRange(value) {
        this._range.internalValue = value;
    }
    resetRange() {
        this._range.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rangeInput() {
        return this._range.internalValue;
    }
}
exports.AppmeshRouteSpecGrpcRouteMatchMetadataMatchOutputReference = AppmeshRouteSpecGrpcRouteMatchMetadataMatchOutputReference;
_e = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecGrpcRouteMatchMetadataMatchOutputReference[_e] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecGrpcRouteMatchMetadataMatchOutputReference", version: "10.0.12" };
function appmeshRouteSpecGrpcRouteMatchMetadataToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        invert: cdktf.booleanToTerraform(struct.invert),
        name: cdktf.stringToTerraform(struct.name),
        match: appmeshRouteSpecGrpcRouteMatchMetadataMatchToTerraform(struct.match),
    };
}
exports.appmeshRouteSpecGrpcRouteMatchMetadataToTerraform = appmeshRouteSpecGrpcRouteMatchMetadataToTerraform;
class AppmeshRouteSpecGrpcRouteMatchMetadataOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // match - computed: false, optional: true, required: false
        this._match = new AppmeshRouteSpecGrpcRouteMatchMetadataMatchOutputReference(this, "match");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._invert !== undefined) {
            hasAnyValues = true;
            internalValueResult.invert = this._invert;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._match?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.match = this._match?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._invert = undefined;
            this._name = undefined;
            this._match.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._invert = value.invert;
            this._name = value.name;
            this._match.internalValue = value.match;
        }
    }
    get invert() {
        return this.getBooleanAttribute('invert');
    }
    set invert(value) {
        this._invert = value;
    }
    resetInvert() {
        this._invert = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get invertInput() {
        return this._invert;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get match() {
        return this._match;
    }
    putMatch(value) {
        this._match.internalValue = value;
    }
    resetMatch() {
        this._match.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get matchInput() {
        return this._match.internalValue;
    }
}
exports.AppmeshRouteSpecGrpcRouteMatchMetadataOutputReference = AppmeshRouteSpecGrpcRouteMatchMetadataOutputReference;
_f = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecGrpcRouteMatchMetadataOutputReference[_f] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecGrpcRouteMatchMetadataOutputReference", version: "10.0.12" };
class AppmeshRouteSpecGrpcRouteMatchMetadataList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AppmeshRouteSpecGrpcRouteMatchMetadataOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AppmeshRouteSpecGrpcRouteMatchMetadataList = AppmeshRouteSpecGrpcRouteMatchMetadataList;
_g = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecGrpcRouteMatchMetadataList[_g] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecGrpcRouteMatchMetadataList", version: "10.0.12" };
function appmeshRouteSpecGrpcRouteMatchToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        method_name: cdktf.stringToTerraform(struct.methodName),
        prefix: cdktf.stringToTerraform(struct.prefix),
        service_name: cdktf.stringToTerraform(struct.serviceName),
        metadata: cdktf.listMapper(appmeshRouteSpecGrpcRouteMatchMetadataToTerraform, true)(struct.metadata),
    };
}
exports.appmeshRouteSpecGrpcRouteMatchToTerraform = appmeshRouteSpecGrpcRouteMatchToTerraform;
class AppmeshRouteSpecGrpcRouteMatchOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // metadata - computed: false, optional: true, required: false
        this._metadata = new AppmeshRouteSpecGrpcRouteMatchMetadataList(this, "metadata", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._methodName !== undefined) {
            hasAnyValues = true;
            internalValueResult.methodName = this._methodName;
        }
        if (this._prefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        if (this._serviceName !== undefined) {
            hasAnyValues = true;
            internalValueResult.serviceName = this._serviceName;
        }
        if (this._metadata?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.metadata = this._metadata?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._methodName = undefined;
            this._prefix = undefined;
            this._serviceName = undefined;
            this._metadata.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._methodName = value.methodName;
            this._prefix = value.prefix;
            this._serviceName = value.serviceName;
            this._metadata.internalValue = value.metadata;
        }
    }
    get methodName() {
        return this.getStringAttribute('method_name');
    }
    set methodName(value) {
        this._methodName = value;
    }
    resetMethodName() {
        this._methodName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get methodNameInput() {
        return this._methodName;
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
    get serviceName() {
        return this.getStringAttribute('service_name');
    }
    set serviceName(value) {
        this._serviceName = value;
    }
    resetServiceName() {
        this._serviceName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceNameInput() {
        return this._serviceName;
    }
    get metadata() {
        return this._metadata;
    }
    putMetadata(value) {
        this._metadata.internalValue = value;
    }
    resetMetadata() {
        this._metadata.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get metadataInput() {
        return this._metadata.internalValue;
    }
}
exports.AppmeshRouteSpecGrpcRouteMatchOutputReference = AppmeshRouteSpecGrpcRouteMatchOutputReference;
_h = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecGrpcRouteMatchOutputReference[_h] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecGrpcRouteMatchOutputReference", version: "10.0.12" };
function appmeshRouteSpecGrpcRouteRetryPolicyPerRetryTimeoutToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        unit: cdktf.stringToTerraform(struct.unit),
        value: cdktf.numberToTerraform(struct.value),
    };
}
exports.appmeshRouteSpecGrpcRouteRetryPolicyPerRetryTimeoutToTerraform = appmeshRouteSpecGrpcRouteRetryPolicyPerRetryTimeoutToTerraform;
class AppmeshRouteSpecGrpcRouteRetryPolicyPerRetryTimeoutOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._unit !== undefined) {
            hasAnyValues = true;
            internalValueResult.unit = this._unit;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._unit = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._unit = value.unit;
            this._value = value.value;
        }
    }
    get unit() {
        return this.getStringAttribute('unit');
    }
    set unit(value) {
        this._unit = value;
    }
    // Temporarily expose input value. Use with caution.
    get unitInput() {
        return this._unit;
    }
    get value() {
        return this.getNumberAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.AppmeshRouteSpecGrpcRouteRetryPolicyPerRetryTimeoutOutputReference = AppmeshRouteSpecGrpcRouteRetryPolicyPerRetryTimeoutOutputReference;
_j = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecGrpcRouteRetryPolicyPerRetryTimeoutOutputReference[_j] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecGrpcRouteRetryPolicyPerRetryTimeoutOutputReference", version: "10.0.12" };
function appmeshRouteSpecGrpcRouteRetryPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        grpc_retry_events: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.grpcRetryEvents),
        http_retry_events: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.httpRetryEvents),
        max_retries: cdktf.numberToTerraform(struct.maxRetries),
        tcp_retry_events: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.tcpRetryEvents),
        per_retry_timeout: appmeshRouteSpecGrpcRouteRetryPolicyPerRetryTimeoutToTerraform(struct.perRetryTimeout),
    };
}
exports.appmeshRouteSpecGrpcRouteRetryPolicyToTerraform = appmeshRouteSpecGrpcRouteRetryPolicyToTerraform;
class AppmeshRouteSpecGrpcRouteRetryPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // per_retry_timeout - computed: false, optional: false, required: true
        this._perRetryTimeout = new AppmeshRouteSpecGrpcRouteRetryPolicyPerRetryTimeoutOutputReference(this, "per_retry_timeout");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._grpcRetryEvents !== undefined) {
            hasAnyValues = true;
            internalValueResult.grpcRetryEvents = this._grpcRetryEvents;
        }
        if (this._httpRetryEvents !== undefined) {
            hasAnyValues = true;
            internalValueResult.httpRetryEvents = this._httpRetryEvents;
        }
        if (this._maxRetries !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxRetries = this._maxRetries;
        }
        if (this._tcpRetryEvents !== undefined) {
            hasAnyValues = true;
            internalValueResult.tcpRetryEvents = this._tcpRetryEvents;
        }
        if (this._perRetryTimeout?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.perRetryTimeout = this._perRetryTimeout?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._grpcRetryEvents = undefined;
            this._httpRetryEvents = undefined;
            this._maxRetries = undefined;
            this._tcpRetryEvents = undefined;
            this._perRetryTimeout.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._grpcRetryEvents = value.grpcRetryEvents;
            this._httpRetryEvents = value.httpRetryEvents;
            this._maxRetries = value.maxRetries;
            this._tcpRetryEvents = value.tcpRetryEvents;
            this._perRetryTimeout.internalValue = value.perRetryTimeout;
        }
    }
    get grpcRetryEvents() {
        return cdktf.Fn.tolist(this.getListAttribute('grpc_retry_events'));
    }
    set grpcRetryEvents(value) {
        this._grpcRetryEvents = value;
    }
    resetGrpcRetryEvents() {
        this._grpcRetryEvents = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get grpcRetryEventsInput() {
        return this._grpcRetryEvents;
    }
    get httpRetryEvents() {
        return cdktf.Fn.tolist(this.getListAttribute('http_retry_events'));
    }
    set httpRetryEvents(value) {
        this._httpRetryEvents = value;
    }
    resetHttpRetryEvents() {
        this._httpRetryEvents = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get httpRetryEventsInput() {
        return this._httpRetryEvents;
    }
    get maxRetries() {
        return this.getNumberAttribute('max_retries');
    }
    set maxRetries(value) {
        this._maxRetries = value;
    }
    // Temporarily expose input value. Use with caution.
    get maxRetriesInput() {
        return this._maxRetries;
    }
    get tcpRetryEvents() {
        return cdktf.Fn.tolist(this.getListAttribute('tcp_retry_events'));
    }
    set tcpRetryEvents(value) {
        this._tcpRetryEvents = value;
    }
    resetTcpRetryEvents() {
        this._tcpRetryEvents = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tcpRetryEventsInput() {
        return this._tcpRetryEvents;
    }
    get perRetryTimeout() {
        return this._perRetryTimeout;
    }
    putPerRetryTimeout(value) {
        this._perRetryTimeout.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get perRetryTimeoutInput() {
        return this._perRetryTimeout.internalValue;
    }
}
exports.AppmeshRouteSpecGrpcRouteRetryPolicyOutputReference = AppmeshRouteSpecGrpcRouteRetryPolicyOutputReference;
_k = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecGrpcRouteRetryPolicyOutputReference[_k] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecGrpcRouteRetryPolicyOutputReference", version: "10.0.12" };
function appmeshRouteSpecGrpcRouteTimeoutIdleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        unit: cdktf.stringToTerraform(struct.unit),
        value: cdktf.numberToTerraform(struct.value),
    };
}
exports.appmeshRouteSpecGrpcRouteTimeoutIdleToTerraform = appmeshRouteSpecGrpcRouteTimeoutIdleToTerraform;
class AppmeshRouteSpecGrpcRouteTimeoutIdleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._unit !== undefined) {
            hasAnyValues = true;
            internalValueResult.unit = this._unit;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._unit = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._unit = value.unit;
            this._value = value.value;
        }
    }
    get unit() {
        return this.getStringAttribute('unit');
    }
    set unit(value) {
        this._unit = value;
    }
    // Temporarily expose input value. Use with caution.
    get unitInput() {
        return this._unit;
    }
    get value() {
        return this.getNumberAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.AppmeshRouteSpecGrpcRouteTimeoutIdleOutputReference = AppmeshRouteSpecGrpcRouteTimeoutIdleOutputReference;
_l = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecGrpcRouteTimeoutIdleOutputReference[_l] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecGrpcRouteTimeoutIdleOutputReference", version: "10.0.12" };
function appmeshRouteSpecGrpcRouteTimeoutPerRequestToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        unit: cdktf.stringToTerraform(struct.unit),
        value: cdktf.numberToTerraform(struct.value),
    };
}
exports.appmeshRouteSpecGrpcRouteTimeoutPerRequestToTerraform = appmeshRouteSpecGrpcRouteTimeoutPerRequestToTerraform;
class AppmeshRouteSpecGrpcRouteTimeoutPerRequestOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._unit !== undefined) {
            hasAnyValues = true;
            internalValueResult.unit = this._unit;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._unit = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._unit = value.unit;
            this._value = value.value;
        }
    }
    get unit() {
        return this.getStringAttribute('unit');
    }
    set unit(value) {
        this._unit = value;
    }
    // Temporarily expose input value. Use with caution.
    get unitInput() {
        return this._unit;
    }
    get value() {
        return this.getNumberAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.AppmeshRouteSpecGrpcRouteTimeoutPerRequestOutputReference = AppmeshRouteSpecGrpcRouteTimeoutPerRequestOutputReference;
_m = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecGrpcRouteTimeoutPerRequestOutputReference[_m] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecGrpcRouteTimeoutPerRequestOutputReference", version: "10.0.12" };
function appmeshRouteSpecGrpcRouteTimeoutToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        idle: appmeshRouteSpecGrpcRouteTimeoutIdleToTerraform(struct.idle),
        per_request: appmeshRouteSpecGrpcRouteTimeoutPerRequestToTerraform(struct.perRequest),
    };
}
exports.appmeshRouteSpecGrpcRouteTimeoutToTerraform = appmeshRouteSpecGrpcRouteTimeoutToTerraform;
class AppmeshRouteSpecGrpcRouteTimeoutOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // idle - computed: false, optional: true, required: false
        this._idle = new AppmeshRouteSpecGrpcRouteTimeoutIdleOutputReference(this, "idle");
        // per_request - computed: false, optional: true, required: false
        this._perRequest = new AppmeshRouteSpecGrpcRouteTimeoutPerRequestOutputReference(this, "per_request");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._idle?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.idle = this._idle?.internalValue;
        }
        if (this._perRequest?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.perRequest = this._perRequest?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._idle.internalValue = undefined;
            this._perRequest.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._idle.internalValue = value.idle;
            this._perRequest.internalValue = value.perRequest;
        }
    }
    get idle() {
        return this._idle;
    }
    putIdle(value) {
        this._idle.internalValue = value;
    }
    resetIdle() {
        this._idle.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idleInput() {
        return this._idle.internalValue;
    }
    get perRequest() {
        return this._perRequest;
    }
    putPerRequest(value) {
        this._perRequest.internalValue = value;
    }
    resetPerRequest() {
        this._perRequest.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get perRequestInput() {
        return this._perRequest.internalValue;
    }
}
exports.AppmeshRouteSpecGrpcRouteTimeoutOutputReference = AppmeshRouteSpecGrpcRouteTimeoutOutputReference;
_o = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecGrpcRouteTimeoutOutputReference[_o] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecGrpcRouteTimeoutOutputReference", version: "10.0.12" };
function appmeshRouteSpecGrpcRouteToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        action: appmeshRouteSpecGrpcRouteActionToTerraform(struct.action),
        match: appmeshRouteSpecGrpcRouteMatchToTerraform(struct.match),
        retry_policy: appmeshRouteSpecGrpcRouteRetryPolicyToTerraform(struct.retryPolicy),
        timeout: appmeshRouteSpecGrpcRouteTimeoutToTerraform(struct.timeout),
    };
}
exports.appmeshRouteSpecGrpcRouteToTerraform = appmeshRouteSpecGrpcRouteToTerraform;
class AppmeshRouteSpecGrpcRouteOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // action - computed: false, optional: false, required: true
        this._action = new AppmeshRouteSpecGrpcRouteActionOutputReference(this, "action");
        // match - computed: false, optional: true, required: false
        this._match = new AppmeshRouteSpecGrpcRouteMatchOutputReference(this, "match");
        // retry_policy - computed: false, optional: true, required: false
        this._retryPolicy = new AppmeshRouteSpecGrpcRouteRetryPolicyOutputReference(this, "retry_policy");
        // timeout - computed: false, optional: true, required: false
        this._timeout = new AppmeshRouteSpecGrpcRouteTimeoutOutputReference(this, "timeout");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._action?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.action = this._action?.internalValue;
        }
        if (this._match?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.match = this._match?.internalValue;
        }
        if (this._retryPolicy?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.retryPolicy = this._retryPolicy?.internalValue;
        }
        if (this._timeout?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeout = this._timeout?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._action.internalValue = undefined;
            this._match.internalValue = undefined;
            this._retryPolicy.internalValue = undefined;
            this._timeout.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._action.internalValue = value.action;
            this._match.internalValue = value.match;
            this._retryPolicy.internalValue = value.retryPolicy;
            this._timeout.internalValue = value.timeout;
        }
    }
    get action() {
        return this._action;
    }
    putAction(value) {
        this._action.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get actionInput() {
        return this._action.internalValue;
    }
    get match() {
        return this._match;
    }
    putMatch(value) {
        this._match.internalValue = value;
    }
    resetMatch() {
        this._match.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get matchInput() {
        return this._match.internalValue;
    }
    get retryPolicy() {
        return this._retryPolicy;
    }
    putRetryPolicy(value) {
        this._retryPolicy.internalValue = value;
    }
    resetRetryPolicy() {
        this._retryPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get retryPolicyInput() {
        return this._retryPolicy.internalValue;
    }
    get timeout() {
        return this._timeout;
    }
    putTimeout(value) {
        this._timeout.internalValue = value;
    }
    resetTimeout() {
        this._timeout.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutInput() {
        return this._timeout.internalValue;
    }
}
exports.AppmeshRouteSpecGrpcRouteOutputReference = AppmeshRouteSpecGrpcRouteOutputReference;
_p = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecGrpcRouteOutputReference[_p] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecGrpcRouteOutputReference", version: "10.0.12" };
function appmeshRouteSpecHttp2RouteActionWeightedTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        virtual_node: cdktf.stringToTerraform(struct.virtualNode),
        weight: cdktf.numberToTerraform(struct.weight),
    };
}
exports.appmeshRouteSpecHttp2RouteActionWeightedTargetToTerraform = appmeshRouteSpecHttp2RouteActionWeightedTargetToTerraform;
class AppmeshRouteSpecHttp2RouteActionWeightedTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._virtualNode !== undefined) {
            hasAnyValues = true;
            internalValueResult.virtualNode = this._virtualNode;
        }
        if (this._weight !== undefined) {
            hasAnyValues = true;
            internalValueResult.weight = this._weight;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._virtualNode = undefined;
            this._weight = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._virtualNode = value.virtualNode;
            this._weight = value.weight;
        }
    }
    get virtualNode() {
        return this.getStringAttribute('virtual_node');
    }
    set virtualNode(value) {
        this._virtualNode = value;
    }
    // Temporarily expose input value. Use with caution.
    get virtualNodeInput() {
        return this._virtualNode;
    }
    get weight() {
        return this.getNumberAttribute('weight');
    }
    set weight(value) {
        this._weight = value;
    }
    // Temporarily expose input value. Use with caution.
    get weightInput() {
        return this._weight;
    }
}
exports.AppmeshRouteSpecHttp2RouteActionWeightedTargetOutputReference = AppmeshRouteSpecHttp2RouteActionWeightedTargetOutputReference;
_q = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttp2RouteActionWeightedTargetOutputReference[_q] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttp2RouteActionWeightedTargetOutputReference", version: "10.0.12" };
class AppmeshRouteSpecHttp2RouteActionWeightedTargetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AppmeshRouteSpecHttp2RouteActionWeightedTargetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AppmeshRouteSpecHttp2RouteActionWeightedTargetList = AppmeshRouteSpecHttp2RouteActionWeightedTargetList;
_r = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttp2RouteActionWeightedTargetList[_r] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttp2RouteActionWeightedTargetList", version: "10.0.12" };
function appmeshRouteSpecHttp2RouteActionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        weighted_target: cdktf.listMapper(appmeshRouteSpecHttp2RouteActionWeightedTargetToTerraform, true)(struct.weightedTarget),
    };
}
exports.appmeshRouteSpecHttp2RouteActionToTerraform = appmeshRouteSpecHttp2RouteActionToTerraform;
class AppmeshRouteSpecHttp2RouteActionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // weighted_target - computed: false, optional: false, required: true
        this._weightedTarget = new AppmeshRouteSpecHttp2RouteActionWeightedTargetList(this, "weighted_target", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._weightedTarget?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.weightedTarget = this._weightedTarget?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._weightedTarget.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._weightedTarget.internalValue = value.weightedTarget;
        }
    }
    get weightedTarget() {
        return this._weightedTarget;
    }
    putWeightedTarget(value) {
        this._weightedTarget.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get weightedTargetInput() {
        return this._weightedTarget.internalValue;
    }
}
exports.AppmeshRouteSpecHttp2RouteActionOutputReference = AppmeshRouteSpecHttp2RouteActionOutputReference;
_s = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttp2RouteActionOutputReference[_s] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttp2RouteActionOutputReference", version: "10.0.12" };
function appmeshRouteSpecHttp2RouteMatchHeaderMatchRangeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        end: cdktf.numberToTerraform(struct.end),
        start: cdktf.numberToTerraform(struct.start),
    };
}
exports.appmeshRouteSpecHttp2RouteMatchHeaderMatchRangeToTerraform = appmeshRouteSpecHttp2RouteMatchHeaderMatchRangeToTerraform;
class AppmeshRouteSpecHttp2RouteMatchHeaderMatchRangeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._end !== undefined) {
            hasAnyValues = true;
            internalValueResult.end = this._end;
        }
        if (this._start !== undefined) {
            hasAnyValues = true;
            internalValueResult.start = this._start;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._end = undefined;
            this._start = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._end = value.end;
            this._start = value.start;
        }
    }
    get end() {
        return this.getNumberAttribute('end');
    }
    set end(value) {
        this._end = value;
    }
    // Temporarily expose input value. Use with caution.
    get endInput() {
        return this._end;
    }
    get start() {
        return this.getNumberAttribute('start');
    }
    set start(value) {
        this._start = value;
    }
    // Temporarily expose input value. Use with caution.
    get startInput() {
        return this._start;
    }
}
exports.AppmeshRouteSpecHttp2RouteMatchHeaderMatchRangeOutputReference = AppmeshRouteSpecHttp2RouteMatchHeaderMatchRangeOutputReference;
_t = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttp2RouteMatchHeaderMatchRangeOutputReference[_t] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttp2RouteMatchHeaderMatchRangeOutputReference", version: "10.0.12" };
function appmeshRouteSpecHttp2RouteMatchHeaderMatchToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        exact: cdktf.stringToTerraform(struct.exact),
        prefix: cdktf.stringToTerraform(struct.prefix),
        regex: cdktf.stringToTerraform(struct.regex),
        suffix: cdktf.stringToTerraform(struct.suffix),
        range: appmeshRouteSpecHttp2RouteMatchHeaderMatchRangeToTerraform(struct.range),
    };
}
exports.appmeshRouteSpecHttp2RouteMatchHeaderMatchToTerraform = appmeshRouteSpecHttp2RouteMatchHeaderMatchToTerraform;
class AppmeshRouteSpecHttp2RouteMatchHeaderMatchOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // range - computed: false, optional: true, required: false
        this._range = new AppmeshRouteSpecHttp2RouteMatchHeaderMatchRangeOutputReference(this, "range");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._exact !== undefined) {
            hasAnyValues = true;
            internalValueResult.exact = this._exact;
        }
        if (this._prefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        if (this._regex !== undefined) {
            hasAnyValues = true;
            internalValueResult.regex = this._regex;
        }
        if (this._suffix !== undefined) {
            hasAnyValues = true;
            internalValueResult.suffix = this._suffix;
        }
        if (this._range?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.range = this._range?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._exact = undefined;
            this._prefix = undefined;
            this._regex = undefined;
            this._suffix = undefined;
            this._range.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._exact = value.exact;
            this._prefix = value.prefix;
            this._regex = value.regex;
            this._suffix = value.suffix;
            this._range.internalValue = value.range;
        }
    }
    get exact() {
        return this.getStringAttribute('exact');
    }
    set exact(value) {
        this._exact = value;
    }
    resetExact() {
        this._exact = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exactInput() {
        return this._exact;
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
    get regex() {
        return this.getStringAttribute('regex');
    }
    set regex(value) {
        this._regex = value;
    }
    resetRegex() {
        this._regex = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regexInput() {
        return this._regex;
    }
    get suffix() {
        return this.getStringAttribute('suffix');
    }
    set suffix(value) {
        this._suffix = value;
    }
    resetSuffix() {
        this._suffix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get suffixInput() {
        return this._suffix;
    }
    get range() {
        return this._range;
    }
    putRange(value) {
        this._range.internalValue = value;
    }
    resetRange() {
        this._range.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rangeInput() {
        return this._range.internalValue;
    }
}
exports.AppmeshRouteSpecHttp2RouteMatchHeaderMatchOutputReference = AppmeshRouteSpecHttp2RouteMatchHeaderMatchOutputReference;
_u = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttp2RouteMatchHeaderMatchOutputReference[_u] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttp2RouteMatchHeaderMatchOutputReference", version: "10.0.12" };
function appmeshRouteSpecHttp2RouteMatchHeaderToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        invert: cdktf.booleanToTerraform(struct.invert),
        name: cdktf.stringToTerraform(struct.name),
        match: appmeshRouteSpecHttp2RouteMatchHeaderMatchToTerraform(struct.match),
    };
}
exports.appmeshRouteSpecHttp2RouteMatchHeaderToTerraform = appmeshRouteSpecHttp2RouteMatchHeaderToTerraform;
class AppmeshRouteSpecHttp2RouteMatchHeaderOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // match - computed: false, optional: true, required: false
        this._match = new AppmeshRouteSpecHttp2RouteMatchHeaderMatchOutputReference(this, "match");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._invert !== undefined) {
            hasAnyValues = true;
            internalValueResult.invert = this._invert;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._match?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.match = this._match?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._invert = undefined;
            this._name = undefined;
            this._match.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._invert = value.invert;
            this._name = value.name;
            this._match.internalValue = value.match;
        }
    }
    get invert() {
        return this.getBooleanAttribute('invert');
    }
    set invert(value) {
        this._invert = value;
    }
    resetInvert() {
        this._invert = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get invertInput() {
        return this._invert;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get match() {
        return this._match;
    }
    putMatch(value) {
        this._match.internalValue = value;
    }
    resetMatch() {
        this._match.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get matchInput() {
        return this._match.internalValue;
    }
}
exports.AppmeshRouteSpecHttp2RouteMatchHeaderOutputReference = AppmeshRouteSpecHttp2RouteMatchHeaderOutputReference;
_v = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttp2RouteMatchHeaderOutputReference[_v] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttp2RouteMatchHeaderOutputReference", version: "10.0.12" };
class AppmeshRouteSpecHttp2RouteMatchHeaderList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AppmeshRouteSpecHttp2RouteMatchHeaderOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AppmeshRouteSpecHttp2RouteMatchHeaderList = AppmeshRouteSpecHttp2RouteMatchHeaderList;
_w = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttp2RouteMatchHeaderList[_w] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttp2RouteMatchHeaderList", version: "10.0.12" };
function appmeshRouteSpecHttp2RouteMatchToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        method: cdktf.stringToTerraform(struct.method),
        prefix: cdktf.stringToTerraform(struct.prefix),
        scheme: cdktf.stringToTerraform(struct.scheme),
        header: cdktf.listMapper(appmeshRouteSpecHttp2RouteMatchHeaderToTerraform, true)(struct.header),
    };
}
exports.appmeshRouteSpecHttp2RouteMatchToTerraform = appmeshRouteSpecHttp2RouteMatchToTerraform;
class AppmeshRouteSpecHttp2RouteMatchOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // header - computed: false, optional: true, required: false
        this._header = new AppmeshRouteSpecHttp2RouteMatchHeaderList(this, "header", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._method !== undefined) {
            hasAnyValues = true;
            internalValueResult.method = this._method;
        }
        if (this._prefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        if (this._scheme !== undefined) {
            hasAnyValues = true;
            internalValueResult.scheme = this._scheme;
        }
        if (this._header?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.header = this._header?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._method = undefined;
            this._prefix = undefined;
            this._scheme = undefined;
            this._header.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._method = value.method;
            this._prefix = value.prefix;
            this._scheme = value.scheme;
            this._header.internalValue = value.header;
        }
    }
    get method() {
        return this.getStringAttribute('method');
    }
    set method(value) {
        this._method = value;
    }
    resetMethod() {
        this._method = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get methodInput() {
        return this._method;
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
    get scheme() {
        return this.getStringAttribute('scheme');
    }
    set scheme(value) {
        this._scheme = value;
    }
    resetScheme() {
        this._scheme = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get schemeInput() {
        return this._scheme;
    }
    get header() {
        return this._header;
    }
    putHeader(value) {
        this._header.internalValue = value;
    }
    resetHeader() {
        this._header.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get headerInput() {
        return this._header.internalValue;
    }
}
exports.AppmeshRouteSpecHttp2RouteMatchOutputReference = AppmeshRouteSpecHttp2RouteMatchOutputReference;
_x = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttp2RouteMatchOutputReference[_x] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttp2RouteMatchOutputReference", version: "10.0.12" };
function appmeshRouteSpecHttp2RouteRetryPolicyPerRetryTimeoutToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        unit: cdktf.stringToTerraform(struct.unit),
        value: cdktf.numberToTerraform(struct.value),
    };
}
exports.appmeshRouteSpecHttp2RouteRetryPolicyPerRetryTimeoutToTerraform = appmeshRouteSpecHttp2RouteRetryPolicyPerRetryTimeoutToTerraform;
class AppmeshRouteSpecHttp2RouteRetryPolicyPerRetryTimeoutOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._unit !== undefined) {
            hasAnyValues = true;
            internalValueResult.unit = this._unit;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._unit = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._unit = value.unit;
            this._value = value.value;
        }
    }
    get unit() {
        return this.getStringAttribute('unit');
    }
    set unit(value) {
        this._unit = value;
    }
    // Temporarily expose input value. Use with caution.
    get unitInput() {
        return this._unit;
    }
    get value() {
        return this.getNumberAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.AppmeshRouteSpecHttp2RouteRetryPolicyPerRetryTimeoutOutputReference = AppmeshRouteSpecHttp2RouteRetryPolicyPerRetryTimeoutOutputReference;
_y = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttp2RouteRetryPolicyPerRetryTimeoutOutputReference[_y] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttp2RouteRetryPolicyPerRetryTimeoutOutputReference", version: "10.0.12" };
function appmeshRouteSpecHttp2RouteRetryPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        http_retry_events: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.httpRetryEvents),
        max_retries: cdktf.numberToTerraform(struct.maxRetries),
        tcp_retry_events: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.tcpRetryEvents),
        per_retry_timeout: appmeshRouteSpecHttp2RouteRetryPolicyPerRetryTimeoutToTerraform(struct.perRetryTimeout),
    };
}
exports.appmeshRouteSpecHttp2RouteRetryPolicyToTerraform = appmeshRouteSpecHttp2RouteRetryPolicyToTerraform;
class AppmeshRouteSpecHttp2RouteRetryPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // per_retry_timeout - computed: false, optional: false, required: true
        this._perRetryTimeout = new AppmeshRouteSpecHttp2RouteRetryPolicyPerRetryTimeoutOutputReference(this, "per_retry_timeout");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._httpRetryEvents !== undefined) {
            hasAnyValues = true;
            internalValueResult.httpRetryEvents = this._httpRetryEvents;
        }
        if (this._maxRetries !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxRetries = this._maxRetries;
        }
        if (this._tcpRetryEvents !== undefined) {
            hasAnyValues = true;
            internalValueResult.tcpRetryEvents = this._tcpRetryEvents;
        }
        if (this._perRetryTimeout?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.perRetryTimeout = this._perRetryTimeout?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._httpRetryEvents = undefined;
            this._maxRetries = undefined;
            this._tcpRetryEvents = undefined;
            this._perRetryTimeout.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._httpRetryEvents = value.httpRetryEvents;
            this._maxRetries = value.maxRetries;
            this._tcpRetryEvents = value.tcpRetryEvents;
            this._perRetryTimeout.internalValue = value.perRetryTimeout;
        }
    }
    get httpRetryEvents() {
        return cdktf.Fn.tolist(this.getListAttribute('http_retry_events'));
    }
    set httpRetryEvents(value) {
        this._httpRetryEvents = value;
    }
    resetHttpRetryEvents() {
        this._httpRetryEvents = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get httpRetryEventsInput() {
        return this._httpRetryEvents;
    }
    get maxRetries() {
        return this.getNumberAttribute('max_retries');
    }
    set maxRetries(value) {
        this._maxRetries = value;
    }
    // Temporarily expose input value. Use with caution.
    get maxRetriesInput() {
        return this._maxRetries;
    }
    get tcpRetryEvents() {
        return cdktf.Fn.tolist(this.getListAttribute('tcp_retry_events'));
    }
    set tcpRetryEvents(value) {
        this._tcpRetryEvents = value;
    }
    resetTcpRetryEvents() {
        this._tcpRetryEvents = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tcpRetryEventsInput() {
        return this._tcpRetryEvents;
    }
    get perRetryTimeout() {
        return this._perRetryTimeout;
    }
    putPerRetryTimeout(value) {
        this._perRetryTimeout.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get perRetryTimeoutInput() {
        return this._perRetryTimeout.internalValue;
    }
}
exports.AppmeshRouteSpecHttp2RouteRetryPolicyOutputReference = AppmeshRouteSpecHttp2RouteRetryPolicyOutputReference;
_z = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttp2RouteRetryPolicyOutputReference[_z] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttp2RouteRetryPolicyOutputReference", version: "10.0.12" };
function appmeshRouteSpecHttp2RouteTimeoutIdleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        unit: cdktf.stringToTerraform(struct.unit),
        value: cdktf.numberToTerraform(struct.value),
    };
}
exports.appmeshRouteSpecHttp2RouteTimeoutIdleToTerraform = appmeshRouteSpecHttp2RouteTimeoutIdleToTerraform;
class AppmeshRouteSpecHttp2RouteTimeoutIdleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._unit !== undefined) {
            hasAnyValues = true;
            internalValueResult.unit = this._unit;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._unit = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._unit = value.unit;
            this._value = value.value;
        }
    }
    get unit() {
        return this.getStringAttribute('unit');
    }
    set unit(value) {
        this._unit = value;
    }
    // Temporarily expose input value. Use with caution.
    get unitInput() {
        return this._unit;
    }
    get value() {
        return this.getNumberAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.AppmeshRouteSpecHttp2RouteTimeoutIdleOutputReference = AppmeshRouteSpecHttp2RouteTimeoutIdleOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttp2RouteTimeoutIdleOutputReference[_0] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttp2RouteTimeoutIdleOutputReference", version: "10.0.12" };
function appmeshRouteSpecHttp2RouteTimeoutPerRequestToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        unit: cdktf.stringToTerraform(struct.unit),
        value: cdktf.numberToTerraform(struct.value),
    };
}
exports.appmeshRouteSpecHttp2RouteTimeoutPerRequestToTerraform = appmeshRouteSpecHttp2RouteTimeoutPerRequestToTerraform;
class AppmeshRouteSpecHttp2RouteTimeoutPerRequestOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._unit !== undefined) {
            hasAnyValues = true;
            internalValueResult.unit = this._unit;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._unit = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._unit = value.unit;
            this._value = value.value;
        }
    }
    get unit() {
        return this.getStringAttribute('unit');
    }
    set unit(value) {
        this._unit = value;
    }
    // Temporarily expose input value. Use with caution.
    get unitInput() {
        return this._unit;
    }
    get value() {
        return this.getNumberAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.AppmeshRouteSpecHttp2RouteTimeoutPerRequestOutputReference = AppmeshRouteSpecHttp2RouteTimeoutPerRequestOutputReference;
_1 = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttp2RouteTimeoutPerRequestOutputReference[_1] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttp2RouteTimeoutPerRequestOutputReference", version: "10.0.12" };
function appmeshRouteSpecHttp2RouteTimeoutToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        idle: appmeshRouteSpecHttp2RouteTimeoutIdleToTerraform(struct.idle),
        per_request: appmeshRouteSpecHttp2RouteTimeoutPerRequestToTerraform(struct.perRequest),
    };
}
exports.appmeshRouteSpecHttp2RouteTimeoutToTerraform = appmeshRouteSpecHttp2RouteTimeoutToTerraform;
class AppmeshRouteSpecHttp2RouteTimeoutOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // idle - computed: false, optional: true, required: false
        this._idle = new AppmeshRouteSpecHttp2RouteTimeoutIdleOutputReference(this, "idle");
        // per_request - computed: false, optional: true, required: false
        this._perRequest = new AppmeshRouteSpecHttp2RouteTimeoutPerRequestOutputReference(this, "per_request");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._idle?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.idle = this._idle?.internalValue;
        }
        if (this._perRequest?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.perRequest = this._perRequest?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._idle.internalValue = undefined;
            this._perRequest.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._idle.internalValue = value.idle;
            this._perRequest.internalValue = value.perRequest;
        }
    }
    get idle() {
        return this._idle;
    }
    putIdle(value) {
        this._idle.internalValue = value;
    }
    resetIdle() {
        this._idle.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idleInput() {
        return this._idle.internalValue;
    }
    get perRequest() {
        return this._perRequest;
    }
    putPerRequest(value) {
        this._perRequest.internalValue = value;
    }
    resetPerRequest() {
        this._perRequest.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get perRequestInput() {
        return this._perRequest.internalValue;
    }
}
exports.AppmeshRouteSpecHttp2RouteTimeoutOutputReference = AppmeshRouteSpecHttp2RouteTimeoutOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttp2RouteTimeoutOutputReference[_2] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttp2RouteTimeoutOutputReference", version: "10.0.12" };
function appmeshRouteSpecHttp2RouteToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        action: appmeshRouteSpecHttp2RouteActionToTerraform(struct.action),
        match: appmeshRouteSpecHttp2RouteMatchToTerraform(struct.match),
        retry_policy: appmeshRouteSpecHttp2RouteRetryPolicyToTerraform(struct.retryPolicy),
        timeout: appmeshRouteSpecHttp2RouteTimeoutToTerraform(struct.timeout),
    };
}
exports.appmeshRouteSpecHttp2RouteToTerraform = appmeshRouteSpecHttp2RouteToTerraform;
class AppmeshRouteSpecHttp2RouteOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // action - computed: false, optional: false, required: true
        this._action = new AppmeshRouteSpecHttp2RouteActionOutputReference(this, "action");
        // match - computed: false, optional: false, required: true
        this._match = new AppmeshRouteSpecHttp2RouteMatchOutputReference(this, "match");
        // retry_policy - computed: false, optional: true, required: false
        this._retryPolicy = new AppmeshRouteSpecHttp2RouteRetryPolicyOutputReference(this, "retry_policy");
        // timeout - computed: false, optional: true, required: false
        this._timeout = new AppmeshRouteSpecHttp2RouteTimeoutOutputReference(this, "timeout");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._action?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.action = this._action?.internalValue;
        }
        if (this._match?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.match = this._match?.internalValue;
        }
        if (this._retryPolicy?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.retryPolicy = this._retryPolicy?.internalValue;
        }
        if (this._timeout?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeout = this._timeout?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._action.internalValue = undefined;
            this._match.internalValue = undefined;
            this._retryPolicy.internalValue = undefined;
            this._timeout.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._action.internalValue = value.action;
            this._match.internalValue = value.match;
            this._retryPolicy.internalValue = value.retryPolicy;
            this._timeout.internalValue = value.timeout;
        }
    }
    get action() {
        return this._action;
    }
    putAction(value) {
        this._action.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get actionInput() {
        return this._action.internalValue;
    }
    get match() {
        return this._match;
    }
    putMatch(value) {
        this._match.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get matchInput() {
        return this._match.internalValue;
    }
    get retryPolicy() {
        return this._retryPolicy;
    }
    putRetryPolicy(value) {
        this._retryPolicy.internalValue = value;
    }
    resetRetryPolicy() {
        this._retryPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get retryPolicyInput() {
        return this._retryPolicy.internalValue;
    }
    get timeout() {
        return this._timeout;
    }
    putTimeout(value) {
        this._timeout.internalValue = value;
    }
    resetTimeout() {
        this._timeout.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutInput() {
        return this._timeout.internalValue;
    }
}
exports.AppmeshRouteSpecHttp2RouteOutputReference = AppmeshRouteSpecHttp2RouteOutputReference;
_3 = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttp2RouteOutputReference[_3] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttp2RouteOutputReference", version: "10.0.12" };
function appmeshRouteSpecHttpRouteActionWeightedTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        virtual_node: cdktf.stringToTerraform(struct.virtualNode),
        weight: cdktf.numberToTerraform(struct.weight),
    };
}
exports.appmeshRouteSpecHttpRouteActionWeightedTargetToTerraform = appmeshRouteSpecHttpRouteActionWeightedTargetToTerraform;
class AppmeshRouteSpecHttpRouteActionWeightedTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._virtualNode !== undefined) {
            hasAnyValues = true;
            internalValueResult.virtualNode = this._virtualNode;
        }
        if (this._weight !== undefined) {
            hasAnyValues = true;
            internalValueResult.weight = this._weight;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._virtualNode = undefined;
            this._weight = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._virtualNode = value.virtualNode;
            this._weight = value.weight;
        }
    }
    get virtualNode() {
        return this.getStringAttribute('virtual_node');
    }
    set virtualNode(value) {
        this._virtualNode = value;
    }
    // Temporarily expose input value. Use with caution.
    get virtualNodeInput() {
        return this._virtualNode;
    }
    get weight() {
        return this.getNumberAttribute('weight');
    }
    set weight(value) {
        this._weight = value;
    }
    // Temporarily expose input value. Use with caution.
    get weightInput() {
        return this._weight;
    }
}
exports.AppmeshRouteSpecHttpRouteActionWeightedTargetOutputReference = AppmeshRouteSpecHttpRouteActionWeightedTargetOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttpRouteActionWeightedTargetOutputReference[_4] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttpRouteActionWeightedTargetOutputReference", version: "10.0.12" };
class AppmeshRouteSpecHttpRouteActionWeightedTargetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AppmeshRouteSpecHttpRouteActionWeightedTargetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AppmeshRouteSpecHttpRouteActionWeightedTargetList = AppmeshRouteSpecHttpRouteActionWeightedTargetList;
_5 = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttpRouteActionWeightedTargetList[_5] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttpRouteActionWeightedTargetList", version: "10.0.12" };
function appmeshRouteSpecHttpRouteActionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        weighted_target: cdktf.listMapper(appmeshRouteSpecHttpRouteActionWeightedTargetToTerraform, true)(struct.weightedTarget),
    };
}
exports.appmeshRouteSpecHttpRouteActionToTerraform = appmeshRouteSpecHttpRouteActionToTerraform;
class AppmeshRouteSpecHttpRouteActionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // weighted_target - computed: false, optional: false, required: true
        this._weightedTarget = new AppmeshRouteSpecHttpRouteActionWeightedTargetList(this, "weighted_target", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._weightedTarget?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.weightedTarget = this._weightedTarget?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._weightedTarget.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._weightedTarget.internalValue = value.weightedTarget;
        }
    }
    get weightedTarget() {
        return this._weightedTarget;
    }
    putWeightedTarget(value) {
        this._weightedTarget.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get weightedTargetInput() {
        return this._weightedTarget.internalValue;
    }
}
exports.AppmeshRouteSpecHttpRouteActionOutputReference = AppmeshRouteSpecHttpRouteActionOutputReference;
_6 = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttpRouteActionOutputReference[_6] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttpRouteActionOutputReference", version: "10.0.12" };
function appmeshRouteSpecHttpRouteMatchHeaderMatchRangeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        end: cdktf.numberToTerraform(struct.end),
        start: cdktf.numberToTerraform(struct.start),
    };
}
exports.appmeshRouteSpecHttpRouteMatchHeaderMatchRangeToTerraform = appmeshRouteSpecHttpRouteMatchHeaderMatchRangeToTerraform;
class AppmeshRouteSpecHttpRouteMatchHeaderMatchRangeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._end !== undefined) {
            hasAnyValues = true;
            internalValueResult.end = this._end;
        }
        if (this._start !== undefined) {
            hasAnyValues = true;
            internalValueResult.start = this._start;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._end = undefined;
            this._start = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._end = value.end;
            this._start = value.start;
        }
    }
    get end() {
        return this.getNumberAttribute('end');
    }
    set end(value) {
        this._end = value;
    }
    // Temporarily expose input value. Use with caution.
    get endInput() {
        return this._end;
    }
    get start() {
        return this.getNumberAttribute('start');
    }
    set start(value) {
        this._start = value;
    }
    // Temporarily expose input value. Use with caution.
    get startInput() {
        return this._start;
    }
}
exports.AppmeshRouteSpecHttpRouteMatchHeaderMatchRangeOutputReference = AppmeshRouteSpecHttpRouteMatchHeaderMatchRangeOutputReference;
_7 = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttpRouteMatchHeaderMatchRangeOutputReference[_7] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttpRouteMatchHeaderMatchRangeOutputReference", version: "10.0.12" };
function appmeshRouteSpecHttpRouteMatchHeaderMatchToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        exact: cdktf.stringToTerraform(struct.exact),
        prefix: cdktf.stringToTerraform(struct.prefix),
        regex: cdktf.stringToTerraform(struct.regex),
        suffix: cdktf.stringToTerraform(struct.suffix),
        range: appmeshRouteSpecHttpRouteMatchHeaderMatchRangeToTerraform(struct.range),
    };
}
exports.appmeshRouteSpecHttpRouteMatchHeaderMatchToTerraform = appmeshRouteSpecHttpRouteMatchHeaderMatchToTerraform;
class AppmeshRouteSpecHttpRouteMatchHeaderMatchOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // range - computed: false, optional: true, required: false
        this._range = new AppmeshRouteSpecHttpRouteMatchHeaderMatchRangeOutputReference(this, "range");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._exact !== undefined) {
            hasAnyValues = true;
            internalValueResult.exact = this._exact;
        }
        if (this._prefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        if (this._regex !== undefined) {
            hasAnyValues = true;
            internalValueResult.regex = this._regex;
        }
        if (this._suffix !== undefined) {
            hasAnyValues = true;
            internalValueResult.suffix = this._suffix;
        }
        if (this._range?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.range = this._range?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._exact = undefined;
            this._prefix = undefined;
            this._regex = undefined;
            this._suffix = undefined;
            this._range.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._exact = value.exact;
            this._prefix = value.prefix;
            this._regex = value.regex;
            this._suffix = value.suffix;
            this._range.internalValue = value.range;
        }
    }
    get exact() {
        return this.getStringAttribute('exact');
    }
    set exact(value) {
        this._exact = value;
    }
    resetExact() {
        this._exact = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exactInput() {
        return this._exact;
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
    get regex() {
        return this.getStringAttribute('regex');
    }
    set regex(value) {
        this._regex = value;
    }
    resetRegex() {
        this._regex = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regexInput() {
        return this._regex;
    }
    get suffix() {
        return this.getStringAttribute('suffix');
    }
    set suffix(value) {
        this._suffix = value;
    }
    resetSuffix() {
        this._suffix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get suffixInput() {
        return this._suffix;
    }
    get range() {
        return this._range;
    }
    putRange(value) {
        this._range.internalValue = value;
    }
    resetRange() {
        this._range.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rangeInput() {
        return this._range.internalValue;
    }
}
exports.AppmeshRouteSpecHttpRouteMatchHeaderMatchOutputReference = AppmeshRouteSpecHttpRouteMatchHeaderMatchOutputReference;
_8 = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttpRouteMatchHeaderMatchOutputReference[_8] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttpRouteMatchHeaderMatchOutputReference", version: "10.0.12" };
function appmeshRouteSpecHttpRouteMatchHeaderToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        invert: cdktf.booleanToTerraform(struct.invert),
        name: cdktf.stringToTerraform(struct.name),
        match: appmeshRouteSpecHttpRouteMatchHeaderMatchToTerraform(struct.match),
    };
}
exports.appmeshRouteSpecHttpRouteMatchHeaderToTerraform = appmeshRouteSpecHttpRouteMatchHeaderToTerraform;
class AppmeshRouteSpecHttpRouteMatchHeaderOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // match - computed: false, optional: true, required: false
        this._match = new AppmeshRouteSpecHttpRouteMatchHeaderMatchOutputReference(this, "match");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._invert !== undefined) {
            hasAnyValues = true;
            internalValueResult.invert = this._invert;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._match?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.match = this._match?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._invert = undefined;
            this._name = undefined;
            this._match.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._invert = value.invert;
            this._name = value.name;
            this._match.internalValue = value.match;
        }
    }
    get invert() {
        return this.getBooleanAttribute('invert');
    }
    set invert(value) {
        this._invert = value;
    }
    resetInvert() {
        this._invert = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get invertInput() {
        return this._invert;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get match() {
        return this._match;
    }
    putMatch(value) {
        this._match.internalValue = value;
    }
    resetMatch() {
        this._match.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get matchInput() {
        return this._match.internalValue;
    }
}
exports.AppmeshRouteSpecHttpRouteMatchHeaderOutputReference = AppmeshRouteSpecHttpRouteMatchHeaderOutputReference;
_9 = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttpRouteMatchHeaderOutputReference[_9] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttpRouteMatchHeaderOutputReference", version: "10.0.12" };
class AppmeshRouteSpecHttpRouteMatchHeaderList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AppmeshRouteSpecHttpRouteMatchHeaderOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AppmeshRouteSpecHttpRouteMatchHeaderList = AppmeshRouteSpecHttpRouteMatchHeaderList;
_10 = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttpRouteMatchHeaderList[_10] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttpRouteMatchHeaderList", version: "10.0.12" };
function appmeshRouteSpecHttpRouteMatchToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        method: cdktf.stringToTerraform(struct.method),
        prefix: cdktf.stringToTerraform(struct.prefix),
        scheme: cdktf.stringToTerraform(struct.scheme),
        header: cdktf.listMapper(appmeshRouteSpecHttpRouteMatchHeaderToTerraform, true)(struct.header),
    };
}
exports.appmeshRouteSpecHttpRouteMatchToTerraform = appmeshRouteSpecHttpRouteMatchToTerraform;
class AppmeshRouteSpecHttpRouteMatchOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // header - computed: false, optional: true, required: false
        this._header = new AppmeshRouteSpecHttpRouteMatchHeaderList(this, "header", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._method !== undefined) {
            hasAnyValues = true;
            internalValueResult.method = this._method;
        }
        if (this._prefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        if (this._scheme !== undefined) {
            hasAnyValues = true;
            internalValueResult.scheme = this._scheme;
        }
        if (this._header?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.header = this._header?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._method = undefined;
            this._prefix = undefined;
            this._scheme = undefined;
            this._header.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._method = value.method;
            this._prefix = value.prefix;
            this._scheme = value.scheme;
            this._header.internalValue = value.header;
        }
    }
    get method() {
        return this.getStringAttribute('method');
    }
    set method(value) {
        this._method = value;
    }
    resetMethod() {
        this._method = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get methodInput() {
        return this._method;
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
    get scheme() {
        return this.getStringAttribute('scheme');
    }
    set scheme(value) {
        this._scheme = value;
    }
    resetScheme() {
        this._scheme = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get schemeInput() {
        return this._scheme;
    }
    get header() {
        return this._header;
    }
    putHeader(value) {
        this._header.internalValue = value;
    }
    resetHeader() {
        this._header.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get headerInput() {
        return this._header.internalValue;
    }
}
exports.AppmeshRouteSpecHttpRouteMatchOutputReference = AppmeshRouteSpecHttpRouteMatchOutputReference;
_11 = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttpRouteMatchOutputReference[_11] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttpRouteMatchOutputReference", version: "10.0.12" };
function appmeshRouteSpecHttpRouteRetryPolicyPerRetryTimeoutToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        unit: cdktf.stringToTerraform(struct.unit),
        value: cdktf.numberToTerraform(struct.value),
    };
}
exports.appmeshRouteSpecHttpRouteRetryPolicyPerRetryTimeoutToTerraform = appmeshRouteSpecHttpRouteRetryPolicyPerRetryTimeoutToTerraform;
class AppmeshRouteSpecHttpRouteRetryPolicyPerRetryTimeoutOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._unit !== undefined) {
            hasAnyValues = true;
            internalValueResult.unit = this._unit;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._unit = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._unit = value.unit;
            this._value = value.value;
        }
    }
    get unit() {
        return this.getStringAttribute('unit');
    }
    set unit(value) {
        this._unit = value;
    }
    // Temporarily expose input value. Use with caution.
    get unitInput() {
        return this._unit;
    }
    get value() {
        return this.getNumberAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.AppmeshRouteSpecHttpRouteRetryPolicyPerRetryTimeoutOutputReference = AppmeshRouteSpecHttpRouteRetryPolicyPerRetryTimeoutOutputReference;
_12 = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttpRouteRetryPolicyPerRetryTimeoutOutputReference[_12] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttpRouteRetryPolicyPerRetryTimeoutOutputReference", version: "10.0.12" };
function appmeshRouteSpecHttpRouteRetryPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        http_retry_events: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.httpRetryEvents),
        max_retries: cdktf.numberToTerraform(struct.maxRetries),
        tcp_retry_events: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.tcpRetryEvents),
        per_retry_timeout: appmeshRouteSpecHttpRouteRetryPolicyPerRetryTimeoutToTerraform(struct.perRetryTimeout),
    };
}
exports.appmeshRouteSpecHttpRouteRetryPolicyToTerraform = appmeshRouteSpecHttpRouteRetryPolicyToTerraform;
class AppmeshRouteSpecHttpRouteRetryPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // per_retry_timeout - computed: false, optional: false, required: true
        this._perRetryTimeout = new AppmeshRouteSpecHttpRouteRetryPolicyPerRetryTimeoutOutputReference(this, "per_retry_timeout");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._httpRetryEvents !== undefined) {
            hasAnyValues = true;
            internalValueResult.httpRetryEvents = this._httpRetryEvents;
        }
        if (this._maxRetries !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxRetries = this._maxRetries;
        }
        if (this._tcpRetryEvents !== undefined) {
            hasAnyValues = true;
            internalValueResult.tcpRetryEvents = this._tcpRetryEvents;
        }
        if (this._perRetryTimeout?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.perRetryTimeout = this._perRetryTimeout?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._httpRetryEvents = undefined;
            this._maxRetries = undefined;
            this._tcpRetryEvents = undefined;
            this._perRetryTimeout.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._httpRetryEvents = value.httpRetryEvents;
            this._maxRetries = value.maxRetries;
            this._tcpRetryEvents = value.tcpRetryEvents;
            this._perRetryTimeout.internalValue = value.perRetryTimeout;
        }
    }
    get httpRetryEvents() {
        return cdktf.Fn.tolist(this.getListAttribute('http_retry_events'));
    }
    set httpRetryEvents(value) {
        this._httpRetryEvents = value;
    }
    resetHttpRetryEvents() {
        this._httpRetryEvents = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get httpRetryEventsInput() {
        return this._httpRetryEvents;
    }
    get maxRetries() {
        return this.getNumberAttribute('max_retries');
    }
    set maxRetries(value) {
        this._maxRetries = value;
    }
    // Temporarily expose input value. Use with caution.
    get maxRetriesInput() {
        return this._maxRetries;
    }
    get tcpRetryEvents() {
        return cdktf.Fn.tolist(this.getListAttribute('tcp_retry_events'));
    }
    set tcpRetryEvents(value) {
        this._tcpRetryEvents = value;
    }
    resetTcpRetryEvents() {
        this._tcpRetryEvents = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tcpRetryEventsInput() {
        return this._tcpRetryEvents;
    }
    get perRetryTimeout() {
        return this._perRetryTimeout;
    }
    putPerRetryTimeout(value) {
        this._perRetryTimeout.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get perRetryTimeoutInput() {
        return this._perRetryTimeout.internalValue;
    }
}
exports.AppmeshRouteSpecHttpRouteRetryPolicyOutputReference = AppmeshRouteSpecHttpRouteRetryPolicyOutputReference;
_13 = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttpRouteRetryPolicyOutputReference[_13] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttpRouteRetryPolicyOutputReference", version: "10.0.12" };
function appmeshRouteSpecHttpRouteTimeoutIdleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        unit: cdktf.stringToTerraform(struct.unit),
        value: cdktf.numberToTerraform(struct.value),
    };
}
exports.appmeshRouteSpecHttpRouteTimeoutIdleToTerraform = appmeshRouteSpecHttpRouteTimeoutIdleToTerraform;
class AppmeshRouteSpecHttpRouteTimeoutIdleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._unit !== undefined) {
            hasAnyValues = true;
            internalValueResult.unit = this._unit;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._unit = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._unit = value.unit;
            this._value = value.value;
        }
    }
    get unit() {
        return this.getStringAttribute('unit');
    }
    set unit(value) {
        this._unit = value;
    }
    // Temporarily expose input value. Use with caution.
    get unitInput() {
        return this._unit;
    }
    get value() {
        return this.getNumberAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.AppmeshRouteSpecHttpRouteTimeoutIdleOutputReference = AppmeshRouteSpecHttpRouteTimeoutIdleOutputReference;
_14 = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttpRouteTimeoutIdleOutputReference[_14] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttpRouteTimeoutIdleOutputReference", version: "10.0.12" };
function appmeshRouteSpecHttpRouteTimeoutPerRequestToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        unit: cdktf.stringToTerraform(struct.unit),
        value: cdktf.numberToTerraform(struct.value),
    };
}
exports.appmeshRouteSpecHttpRouteTimeoutPerRequestToTerraform = appmeshRouteSpecHttpRouteTimeoutPerRequestToTerraform;
class AppmeshRouteSpecHttpRouteTimeoutPerRequestOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._unit !== undefined) {
            hasAnyValues = true;
            internalValueResult.unit = this._unit;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._unit = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._unit = value.unit;
            this._value = value.value;
        }
    }
    get unit() {
        return this.getStringAttribute('unit');
    }
    set unit(value) {
        this._unit = value;
    }
    // Temporarily expose input value. Use with caution.
    get unitInput() {
        return this._unit;
    }
    get value() {
        return this.getNumberAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.AppmeshRouteSpecHttpRouteTimeoutPerRequestOutputReference = AppmeshRouteSpecHttpRouteTimeoutPerRequestOutputReference;
_15 = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttpRouteTimeoutPerRequestOutputReference[_15] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttpRouteTimeoutPerRequestOutputReference", version: "10.0.12" };
function appmeshRouteSpecHttpRouteTimeoutToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        idle: appmeshRouteSpecHttpRouteTimeoutIdleToTerraform(struct.idle),
        per_request: appmeshRouteSpecHttpRouteTimeoutPerRequestToTerraform(struct.perRequest),
    };
}
exports.appmeshRouteSpecHttpRouteTimeoutToTerraform = appmeshRouteSpecHttpRouteTimeoutToTerraform;
class AppmeshRouteSpecHttpRouteTimeoutOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // idle - computed: false, optional: true, required: false
        this._idle = new AppmeshRouteSpecHttpRouteTimeoutIdleOutputReference(this, "idle");
        // per_request - computed: false, optional: true, required: false
        this._perRequest = new AppmeshRouteSpecHttpRouteTimeoutPerRequestOutputReference(this, "per_request");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._idle?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.idle = this._idle?.internalValue;
        }
        if (this._perRequest?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.perRequest = this._perRequest?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._idle.internalValue = undefined;
            this._perRequest.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._idle.internalValue = value.idle;
            this._perRequest.internalValue = value.perRequest;
        }
    }
    get idle() {
        return this._idle;
    }
    putIdle(value) {
        this._idle.internalValue = value;
    }
    resetIdle() {
        this._idle.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idleInput() {
        return this._idle.internalValue;
    }
    get perRequest() {
        return this._perRequest;
    }
    putPerRequest(value) {
        this._perRequest.internalValue = value;
    }
    resetPerRequest() {
        this._perRequest.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get perRequestInput() {
        return this._perRequest.internalValue;
    }
}
exports.AppmeshRouteSpecHttpRouteTimeoutOutputReference = AppmeshRouteSpecHttpRouteTimeoutOutputReference;
_16 = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttpRouteTimeoutOutputReference[_16] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttpRouteTimeoutOutputReference", version: "10.0.12" };
function appmeshRouteSpecHttpRouteToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        action: appmeshRouteSpecHttpRouteActionToTerraform(struct.action),
        match: appmeshRouteSpecHttpRouteMatchToTerraform(struct.match),
        retry_policy: appmeshRouteSpecHttpRouteRetryPolicyToTerraform(struct.retryPolicy),
        timeout: appmeshRouteSpecHttpRouteTimeoutToTerraform(struct.timeout),
    };
}
exports.appmeshRouteSpecHttpRouteToTerraform = appmeshRouteSpecHttpRouteToTerraform;
class AppmeshRouteSpecHttpRouteOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // action - computed: false, optional: false, required: true
        this._action = new AppmeshRouteSpecHttpRouteActionOutputReference(this, "action");
        // match - computed: false, optional: false, required: true
        this._match = new AppmeshRouteSpecHttpRouteMatchOutputReference(this, "match");
        // retry_policy - computed: false, optional: true, required: false
        this._retryPolicy = new AppmeshRouteSpecHttpRouteRetryPolicyOutputReference(this, "retry_policy");
        // timeout - computed: false, optional: true, required: false
        this._timeout = new AppmeshRouteSpecHttpRouteTimeoutOutputReference(this, "timeout");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._action?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.action = this._action?.internalValue;
        }
        if (this._match?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.match = this._match?.internalValue;
        }
        if (this._retryPolicy?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.retryPolicy = this._retryPolicy?.internalValue;
        }
        if (this._timeout?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeout = this._timeout?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._action.internalValue = undefined;
            this._match.internalValue = undefined;
            this._retryPolicy.internalValue = undefined;
            this._timeout.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._action.internalValue = value.action;
            this._match.internalValue = value.match;
            this._retryPolicy.internalValue = value.retryPolicy;
            this._timeout.internalValue = value.timeout;
        }
    }
    get action() {
        return this._action;
    }
    putAction(value) {
        this._action.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get actionInput() {
        return this._action.internalValue;
    }
    get match() {
        return this._match;
    }
    putMatch(value) {
        this._match.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get matchInput() {
        return this._match.internalValue;
    }
    get retryPolicy() {
        return this._retryPolicy;
    }
    putRetryPolicy(value) {
        this._retryPolicy.internalValue = value;
    }
    resetRetryPolicy() {
        this._retryPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get retryPolicyInput() {
        return this._retryPolicy.internalValue;
    }
    get timeout() {
        return this._timeout;
    }
    putTimeout(value) {
        this._timeout.internalValue = value;
    }
    resetTimeout() {
        this._timeout.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutInput() {
        return this._timeout.internalValue;
    }
}
exports.AppmeshRouteSpecHttpRouteOutputReference = AppmeshRouteSpecHttpRouteOutputReference;
_17 = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecHttpRouteOutputReference[_17] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecHttpRouteOutputReference", version: "10.0.12" };
function appmeshRouteSpecTcpRouteActionWeightedTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        virtual_node: cdktf.stringToTerraform(struct.virtualNode),
        weight: cdktf.numberToTerraform(struct.weight),
    };
}
exports.appmeshRouteSpecTcpRouteActionWeightedTargetToTerraform = appmeshRouteSpecTcpRouteActionWeightedTargetToTerraform;
class AppmeshRouteSpecTcpRouteActionWeightedTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._virtualNode !== undefined) {
            hasAnyValues = true;
            internalValueResult.virtualNode = this._virtualNode;
        }
        if (this._weight !== undefined) {
            hasAnyValues = true;
            internalValueResult.weight = this._weight;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._virtualNode = undefined;
            this._weight = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._virtualNode = value.virtualNode;
            this._weight = value.weight;
        }
    }
    get virtualNode() {
        return this.getStringAttribute('virtual_node');
    }
    set virtualNode(value) {
        this._virtualNode = value;
    }
    // Temporarily expose input value. Use with caution.
    get virtualNodeInput() {
        return this._virtualNode;
    }
    get weight() {
        return this.getNumberAttribute('weight');
    }
    set weight(value) {
        this._weight = value;
    }
    // Temporarily expose input value. Use with caution.
    get weightInput() {
        return this._weight;
    }
}
exports.AppmeshRouteSpecTcpRouteActionWeightedTargetOutputReference = AppmeshRouteSpecTcpRouteActionWeightedTargetOutputReference;
_18 = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecTcpRouteActionWeightedTargetOutputReference[_18] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecTcpRouteActionWeightedTargetOutputReference", version: "10.0.12" };
class AppmeshRouteSpecTcpRouteActionWeightedTargetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AppmeshRouteSpecTcpRouteActionWeightedTargetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AppmeshRouteSpecTcpRouteActionWeightedTargetList = AppmeshRouteSpecTcpRouteActionWeightedTargetList;
_19 = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecTcpRouteActionWeightedTargetList[_19] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecTcpRouteActionWeightedTargetList", version: "10.0.12" };
function appmeshRouteSpecTcpRouteActionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        weighted_target: cdktf.listMapper(appmeshRouteSpecTcpRouteActionWeightedTargetToTerraform, true)(struct.weightedTarget),
    };
}
exports.appmeshRouteSpecTcpRouteActionToTerraform = appmeshRouteSpecTcpRouteActionToTerraform;
class AppmeshRouteSpecTcpRouteActionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // weighted_target - computed: false, optional: false, required: true
        this._weightedTarget = new AppmeshRouteSpecTcpRouteActionWeightedTargetList(this, "weighted_target", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._weightedTarget?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.weightedTarget = this._weightedTarget?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._weightedTarget.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._weightedTarget.internalValue = value.weightedTarget;
        }
    }
    get weightedTarget() {
        return this._weightedTarget;
    }
    putWeightedTarget(value) {
        this._weightedTarget.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get weightedTargetInput() {
        return this._weightedTarget.internalValue;
    }
}
exports.AppmeshRouteSpecTcpRouteActionOutputReference = AppmeshRouteSpecTcpRouteActionOutputReference;
_20 = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecTcpRouteActionOutputReference[_20] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecTcpRouteActionOutputReference", version: "10.0.12" };
function appmeshRouteSpecTcpRouteTimeoutIdleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        unit: cdktf.stringToTerraform(struct.unit),
        value: cdktf.numberToTerraform(struct.value),
    };
}
exports.appmeshRouteSpecTcpRouteTimeoutIdleToTerraform = appmeshRouteSpecTcpRouteTimeoutIdleToTerraform;
class AppmeshRouteSpecTcpRouteTimeoutIdleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._unit !== undefined) {
            hasAnyValues = true;
            internalValueResult.unit = this._unit;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._unit = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._unit = value.unit;
            this._value = value.value;
        }
    }
    get unit() {
        return this.getStringAttribute('unit');
    }
    set unit(value) {
        this._unit = value;
    }
    // Temporarily expose input value. Use with caution.
    get unitInput() {
        return this._unit;
    }
    get value() {
        return this.getNumberAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.AppmeshRouteSpecTcpRouteTimeoutIdleOutputReference = AppmeshRouteSpecTcpRouteTimeoutIdleOutputReference;
_21 = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecTcpRouteTimeoutIdleOutputReference[_21] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecTcpRouteTimeoutIdleOutputReference", version: "10.0.12" };
function appmeshRouteSpecTcpRouteTimeoutToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        idle: appmeshRouteSpecTcpRouteTimeoutIdleToTerraform(struct.idle),
    };
}
exports.appmeshRouteSpecTcpRouteTimeoutToTerraform = appmeshRouteSpecTcpRouteTimeoutToTerraform;
class AppmeshRouteSpecTcpRouteTimeoutOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // idle - computed: false, optional: true, required: false
        this._idle = new AppmeshRouteSpecTcpRouteTimeoutIdleOutputReference(this, "idle");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._idle?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.idle = this._idle?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._idle.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._idle.internalValue = value.idle;
        }
    }
    get idle() {
        return this._idle;
    }
    putIdle(value) {
        this._idle.internalValue = value;
    }
    resetIdle() {
        this._idle.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idleInput() {
        return this._idle.internalValue;
    }
}
exports.AppmeshRouteSpecTcpRouteTimeoutOutputReference = AppmeshRouteSpecTcpRouteTimeoutOutputReference;
_22 = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecTcpRouteTimeoutOutputReference[_22] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecTcpRouteTimeoutOutputReference", version: "10.0.12" };
function appmeshRouteSpecTcpRouteToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        action: appmeshRouteSpecTcpRouteActionToTerraform(struct.action),
        timeout: appmeshRouteSpecTcpRouteTimeoutToTerraform(struct.timeout),
    };
}
exports.appmeshRouteSpecTcpRouteToTerraform = appmeshRouteSpecTcpRouteToTerraform;
class AppmeshRouteSpecTcpRouteOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // action - computed: false, optional: false, required: true
        this._action = new AppmeshRouteSpecTcpRouteActionOutputReference(this, "action");
        // timeout - computed: false, optional: true, required: false
        this._timeout = new AppmeshRouteSpecTcpRouteTimeoutOutputReference(this, "timeout");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._action?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.action = this._action?.internalValue;
        }
        if (this._timeout?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeout = this._timeout?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._action.internalValue = undefined;
            this._timeout.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._action.internalValue = value.action;
            this._timeout.internalValue = value.timeout;
        }
    }
    get action() {
        return this._action;
    }
    putAction(value) {
        this._action.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get actionInput() {
        return this._action.internalValue;
    }
    get timeout() {
        return this._timeout;
    }
    putTimeout(value) {
        this._timeout.internalValue = value;
    }
    resetTimeout() {
        this._timeout.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutInput() {
        return this._timeout.internalValue;
    }
}
exports.AppmeshRouteSpecTcpRouteOutputReference = AppmeshRouteSpecTcpRouteOutputReference;
_23 = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecTcpRouteOutputReference[_23] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecTcpRouteOutputReference", version: "10.0.12" };
function appmeshRouteSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        priority: cdktf.numberToTerraform(struct.priority),
        grpc_route: appmeshRouteSpecGrpcRouteToTerraform(struct.grpcRoute),
        http2_route: appmeshRouteSpecHttp2RouteToTerraform(struct.http2Route),
        http_route: appmeshRouteSpecHttpRouteToTerraform(struct.httpRoute),
        tcp_route: appmeshRouteSpecTcpRouteToTerraform(struct.tcpRoute),
    };
}
exports.appmeshRouteSpecToTerraform = appmeshRouteSpecToTerraform;
class AppmeshRouteSpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // grpc_route - computed: false, optional: true, required: false
        this._grpcRoute = new AppmeshRouteSpecGrpcRouteOutputReference(this, "grpc_route");
        // http2_route - computed: false, optional: true, required: false
        this._http2Route = new AppmeshRouteSpecHttp2RouteOutputReference(this, "http2_route");
        // http_route - computed: false, optional: true, required: false
        this._httpRoute = new AppmeshRouteSpecHttpRouteOutputReference(this, "http_route");
        // tcp_route - computed: false, optional: true, required: false
        this._tcpRoute = new AppmeshRouteSpecTcpRouteOutputReference(this, "tcp_route");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._priority !== undefined) {
            hasAnyValues = true;
            internalValueResult.priority = this._priority;
        }
        if (this._grpcRoute?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.grpcRoute = this._grpcRoute?.internalValue;
        }
        if (this._http2Route?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.http2Route = this._http2Route?.internalValue;
        }
        if (this._httpRoute?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.httpRoute = this._httpRoute?.internalValue;
        }
        if (this._tcpRoute?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.tcpRoute = this._tcpRoute?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._priority = undefined;
            this._grpcRoute.internalValue = undefined;
            this._http2Route.internalValue = undefined;
            this._httpRoute.internalValue = undefined;
            this._tcpRoute.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._priority = value.priority;
            this._grpcRoute.internalValue = value.grpcRoute;
            this._http2Route.internalValue = value.http2Route;
            this._httpRoute.internalValue = value.httpRoute;
            this._tcpRoute.internalValue = value.tcpRoute;
        }
    }
    get priority() {
        return this.getNumberAttribute('priority');
    }
    set priority(value) {
        this._priority = value;
    }
    resetPriority() {
        this._priority = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get priorityInput() {
        return this._priority;
    }
    get grpcRoute() {
        return this._grpcRoute;
    }
    putGrpcRoute(value) {
        this._grpcRoute.internalValue = value;
    }
    resetGrpcRoute() {
        this._grpcRoute.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get grpcRouteInput() {
        return this._grpcRoute.internalValue;
    }
    get http2Route() {
        return this._http2Route;
    }
    putHttp2Route(value) {
        this._http2Route.internalValue = value;
    }
    resetHttp2Route() {
        this._http2Route.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get http2RouteInput() {
        return this._http2Route.internalValue;
    }
    get httpRoute() {
        return this._httpRoute;
    }
    putHttpRoute(value) {
        this._httpRoute.internalValue = value;
    }
    resetHttpRoute() {
        this._httpRoute.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get httpRouteInput() {
        return this._httpRoute.internalValue;
    }
    get tcpRoute() {
        return this._tcpRoute;
    }
    putTcpRoute(value) {
        this._tcpRoute.internalValue = value;
    }
    resetTcpRoute() {
        this._tcpRoute.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tcpRouteInput() {
        return this._tcpRoute.internalValue;
    }
}
exports.AppmeshRouteSpecOutputReference = AppmeshRouteSpecOutputReference;
_24 = JSII_RTTI_SYMBOL_1;
AppmeshRouteSpecOutputReference[_24] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRouteSpecOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/appmesh_route aws_appmesh_route}
*/
class AppmeshRoute extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/appmesh_route aws_appmesh_route} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options AppmeshRouteConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_appmesh_route',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // spec - computed: false, optional: false, required: true
        this._spec = new AppmeshRouteSpecOutputReference(this, "spec");
        this._id = config.id;
        this._meshName = config.meshName;
        this._meshOwner = config.meshOwner;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._virtualRouterName = config.virtualRouterName;
        this._spec.internalValue = config.spec;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // created_date - computed: true, optional: false, required: false
    get createdDate() {
        return this.getStringAttribute('created_date');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // last_updated_date - computed: true, optional: false, required: false
    get lastUpdatedDate() {
        return this.getStringAttribute('last_updated_date');
    }
    get meshName() {
        return this.getStringAttribute('mesh_name');
    }
    set meshName(value) {
        this._meshName = value;
    }
    // Temporarily expose input value. Use with caution.
    get meshNameInput() {
        return this._meshName;
    }
    get meshOwner() {
        return this.getStringAttribute('mesh_owner');
    }
    set meshOwner(value) {
        this._meshOwner = value;
    }
    resetMeshOwner() {
        this._meshOwner = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get meshOwnerInput() {
        return this._meshOwner;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // resource_owner - computed: true, optional: false, required: false
    get resourceOwner() {
        return this.getStringAttribute('resource_owner');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get virtualRouterName() {
        return this.getStringAttribute('virtual_router_name');
    }
    set virtualRouterName(value) {
        this._virtualRouterName = value;
    }
    // Temporarily expose input value. Use with caution.
    get virtualRouterNameInput() {
        return this._virtualRouterName;
    }
    get spec() {
        return this._spec;
    }
    putSpec(value) {
        this._spec.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get specInput() {
        return this._spec.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            mesh_name: cdktf.stringToTerraform(this._meshName),
            mesh_owner: cdktf.stringToTerraform(this._meshOwner),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            virtual_router_name: cdktf.stringToTerraform(this._virtualRouterName),
            spec: appmeshRouteSpecToTerraform(this._spec.internalValue),
        };
    }
}
exports.AppmeshRoute = AppmeshRoute;
_25 = JSII_RTTI_SYMBOL_1;
AppmeshRoute[_25] = { fqn: "@cdktf/provider-aws.appmeshRoute.AppmeshRoute", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
AppmeshRoute.tfResourceType = "aws_appmesh_route";
