// @cdktf/provider-aws/lib/s3-bucket-inventory/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h;
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3BucketInventory = exports.S3BucketInventoryScheduleOutputReference = exports.s3BucketInventoryScheduleToTerraform = exports.S3BucketInventoryFilterOutputReference = exports.s3BucketInventoryFilterToTerraform = exports.S3BucketInventoryDestinationOutputReference = exports.s3BucketInventoryDestinationToTerraform = exports.S3BucketInventoryDestinationBucketOutputReference = exports.s3BucketInventoryDestinationBucketToTerraform = exports.S3BucketInventoryDestinationBucketEncryptionOutputReference = exports.s3BucketInventoryDestinationBucketEncryptionToTerraform = exports.S3BucketInventoryDestinationBucketEncryptionSseS3OutputReference = exports.s3BucketInventoryDestinationBucketEncryptionSseS3ToTerraform = exports.S3BucketInventoryDestinationBucketEncryptionSseKmsOutputReference = exports.s3BucketInventoryDestinationBucketEncryptionSseKmsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function s3BucketInventoryDestinationBucketEncryptionSseKmsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key_id: cdktf.stringToTerraform(struct.keyId),
    };
}
exports.s3BucketInventoryDestinationBucketEncryptionSseKmsToTerraform = s3BucketInventoryDestinationBucketEncryptionSseKmsToTerraform;
class S3BucketInventoryDestinationBucketEncryptionSseKmsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._keyId !== undefined) {
            hasAnyValues = true;
            internalValueResult.keyId = this._keyId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._keyId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._keyId = value.keyId;
        }
    }
    get keyId() {
        return this.getStringAttribute('key_id');
    }
    set keyId(value) {
        this._keyId = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyIdInput() {
        return this._keyId;
    }
}
exports.S3BucketInventoryDestinationBucketEncryptionSseKmsOutputReference = S3BucketInventoryDestinationBucketEncryptionSseKmsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
S3BucketInventoryDestinationBucketEncryptionSseKmsOutputReference[_a] = { fqn: "@cdktf/provider-aws.s3BucketInventory.S3BucketInventoryDestinationBucketEncryptionSseKmsOutputReference", version: "10.0.12" };
function s3BucketInventoryDestinationBucketEncryptionSseS3ToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.s3BucketInventoryDestinationBucketEncryptionSseS3ToTerraform = s3BucketInventoryDestinationBucketEncryptionSseS3ToTerraform;
class S3BucketInventoryDestinationBucketEncryptionSseS3OutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.S3BucketInventoryDestinationBucketEncryptionSseS3OutputReference = S3BucketInventoryDestinationBucketEncryptionSseS3OutputReference;
_b = JSII_RTTI_SYMBOL_1;
S3BucketInventoryDestinationBucketEncryptionSseS3OutputReference[_b] = { fqn: "@cdktf/provider-aws.s3BucketInventory.S3BucketInventoryDestinationBucketEncryptionSseS3OutputReference", version: "10.0.12" };
function s3BucketInventoryDestinationBucketEncryptionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        sse_kms: s3BucketInventoryDestinationBucketEncryptionSseKmsToTerraform(struct.sseKms),
        sse_s3: s3BucketInventoryDestinationBucketEncryptionSseS3ToTerraform(struct.sseS3),
    };
}
exports.s3BucketInventoryDestinationBucketEncryptionToTerraform = s3BucketInventoryDestinationBucketEncryptionToTerraform;
class S3BucketInventoryDestinationBucketEncryptionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // sse_kms - computed: false, optional: true, required: false
        this._sseKms = new S3BucketInventoryDestinationBucketEncryptionSseKmsOutputReference(this, "sse_kms");
        // sse_s3 - computed: false, optional: true, required: false
        this._sseS3 = new S3BucketInventoryDestinationBucketEncryptionSseS3OutputReference(this, "sse_s3");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._sseKms?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.sseKms = this._sseKms?.internalValue;
        }
        if (this._sseS3?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.sseS3 = this._sseS3?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._sseKms.internalValue = undefined;
            this._sseS3.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._sseKms.internalValue = value.sseKms;
            this._sseS3.internalValue = value.sseS3;
        }
    }
    get sseKms() {
        return this._sseKms;
    }
    putSseKms(value) {
        this._sseKms.internalValue = value;
    }
    resetSseKms() {
        this._sseKms.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sseKmsInput() {
        return this._sseKms.internalValue;
    }
    get sseS3() {
        return this._sseS3;
    }
    putSseS3(value) {
        this._sseS3.internalValue = value;
    }
    resetSseS3() {
        this._sseS3.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sseS3Input() {
        return this._sseS3.internalValue;
    }
}
exports.S3BucketInventoryDestinationBucketEncryptionOutputReference = S3BucketInventoryDestinationBucketEncryptionOutputReference;
_c = JSII_RTTI_SYMBOL_1;
S3BucketInventoryDestinationBucketEncryptionOutputReference[_c] = { fqn: "@cdktf/provider-aws.s3BucketInventory.S3BucketInventoryDestinationBucketEncryptionOutputReference", version: "10.0.12" };
function s3BucketInventoryDestinationBucketToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        account_id: cdktf.stringToTerraform(struct.accountId),
        bucket_arn: cdktf.stringToTerraform(struct.bucketArn),
        format: cdktf.stringToTerraform(struct.format),
        prefix: cdktf.stringToTerraform(struct.prefix),
        encryption: s3BucketInventoryDestinationBucketEncryptionToTerraform(struct.encryption),
    };
}
exports.s3BucketInventoryDestinationBucketToTerraform = s3BucketInventoryDestinationBucketToTerraform;
class S3BucketInventoryDestinationBucketOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // encryption - computed: false, optional: true, required: false
        this._encryption = new S3BucketInventoryDestinationBucketEncryptionOutputReference(this, "encryption");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._accountId !== undefined) {
            hasAnyValues = true;
            internalValueResult.accountId = this._accountId;
        }
        if (this._bucketArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.bucketArn = this._bucketArn;
        }
        if (this._format !== undefined) {
            hasAnyValues = true;
            internalValueResult.format = this._format;
        }
        if (this._prefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        if (this._encryption?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.encryption = this._encryption?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._accountId = undefined;
            this._bucketArn = undefined;
            this._format = undefined;
            this._prefix = undefined;
            this._encryption.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._accountId = value.accountId;
            this._bucketArn = value.bucketArn;
            this._format = value.format;
            this._prefix = value.prefix;
            this._encryption.internalValue = value.encryption;
        }
    }
    get accountId() {
        return this.getStringAttribute('account_id');
    }
    set accountId(value) {
        this._accountId = value;
    }
    resetAccountId() {
        this._accountId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accountIdInput() {
        return this._accountId;
    }
    get bucketArn() {
        return this.getStringAttribute('bucket_arn');
    }
    set bucketArn(value) {
        this._bucketArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get bucketArnInput() {
        return this._bucketArn;
    }
    get format() {
        return this.getStringAttribute('format');
    }
    set format(value) {
        this._format = value;
    }
    // Temporarily expose input value. Use with caution.
    get formatInput() {
        return this._format;
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
    get encryption() {
        return this._encryption;
    }
    putEncryption(value) {
        this._encryption.internalValue = value;
    }
    resetEncryption() {
        this._encryption.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encryptionInput() {
        return this._encryption.internalValue;
    }
}
exports.S3BucketInventoryDestinationBucketOutputReference = S3BucketInventoryDestinationBucketOutputReference;
_d = JSII_RTTI_SYMBOL_1;
S3BucketInventoryDestinationBucketOutputReference[_d] = { fqn: "@cdktf/provider-aws.s3BucketInventory.S3BucketInventoryDestinationBucketOutputReference", version: "10.0.12" };
function s3BucketInventoryDestinationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket: s3BucketInventoryDestinationBucketToTerraform(struct.bucket),
    };
}
exports.s3BucketInventoryDestinationToTerraform = s3BucketInventoryDestinationToTerraform;
class S3BucketInventoryDestinationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // bucket - computed: false, optional: false, required: true
        this._bucket = new S3BucketInventoryDestinationBucketOutputReference(this, "bucket");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucket?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.bucket = this._bucket?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucket.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucket.internalValue = value.bucket;
        }
    }
    get bucket() {
        return this._bucket;
    }
    putBucket(value) {
        this._bucket.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get bucketInput() {
        return this._bucket.internalValue;
    }
}
exports.S3BucketInventoryDestinationOutputReference = S3BucketInventoryDestinationOutputReference;
_e = JSII_RTTI_SYMBOL_1;
S3BucketInventoryDestinationOutputReference[_e] = { fqn: "@cdktf/provider-aws.s3BucketInventory.S3BucketInventoryDestinationOutputReference", version: "10.0.12" };
function s3BucketInventoryFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        prefix: cdktf.stringToTerraform(struct.prefix),
    };
}
exports.s3BucketInventoryFilterToTerraform = s3BucketInventoryFilterToTerraform;
class S3BucketInventoryFilterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._prefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._prefix = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._prefix = value.prefix;
        }
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
}
exports.S3BucketInventoryFilterOutputReference = S3BucketInventoryFilterOutputReference;
_f = JSII_RTTI_SYMBOL_1;
S3BucketInventoryFilterOutputReference[_f] = { fqn: "@cdktf/provider-aws.s3BucketInventory.S3BucketInventoryFilterOutputReference", version: "10.0.12" };
function s3BucketInventoryScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        frequency: cdktf.stringToTerraform(struct.frequency),
    };
}
exports.s3BucketInventoryScheduleToTerraform = s3BucketInventoryScheduleToTerraform;
class S3BucketInventoryScheduleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._frequency !== undefined) {
            hasAnyValues = true;
            internalValueResult.frequency = this._frequency;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._frequency = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._frequency = value.frequency;
        }
    }
    get frequency() {
        return this.getStringAttribute('frequency');
    }
    set frequency(value) {
        this._frequency = value;
    }
    // Temporarily expose input value. Use with caution.
    get frequencyInput() {
        return this._frequency;
    }
}
exports.S3BucketInventoryScheduleOutputReference = S3BucketInventoryScheduleOutputReference;
_g = JSII_RTTI_SYMBOL_1;
S3BucketInventoryScheduleOutputReference[_g] = { fqn: "@cdktf/provider-aws.s3BucketInventory.S3BucketInventoryScheduleOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/s3_bucket_inventory aws_s3_bucket_inventory}
*/
class S3BucketInventory extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/s3_bucket_inventory aws_s3_bucket_inventory} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options S3BucketInventoryConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_s3_bucket_inventory',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // destination - computed: false, optional: false, required: true
        this._destination = new S3BucketInventoryDestinationOutputReference(this, "destination");
        // filter - computed: false, optional: true, required: false
        this._filter = new S3BucketInventoryFilterOutputReference(this, "filter");
        // schedule - computed: false, optional: false, required: true
        this._schedule = new S3BucketInventoryScheduleOutputReference(this, "schedule");
        this._bucket = config.bucket;
        this._enabled = config.enabled;
        this._id = config.id;
        this._includedObjectVersions = config.includedObjectVersions;
        this._name = config.name;
        this._optionalFields = config.optionalFields;
        this._destination.internalValue = config.destination;
        this._filter.internalValue = config.filter;
        this._schedule.internalValue = config.schedule;
    }
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    set bucket(value) {
        this._bucket = value;
    }
    // Temporarily expose input value. Use with caution.
    get bucketInput() {
        return this._bucket;
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get includedObjectVersions() {
        return this.getStringAttribute('included_object_versions');
    }
    set includedObjectVersions(value) {
        this._includedObjectVersions = value;
    }
    // Temporarily expose input value. Use with caution.
    get includedObjectVersionsInput() {
        return this._includedObjectVersions;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get optionalFields() {
        return cdktf.Fn.tolist(this.getListAttribute('optional_fields'));
    }
    set optionalFields(value) {
        this._optionalFields = value;
    }
    resetOptionalFields() {
        this._optionalFields = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get optionalFieldsInput() {
        return this._optionalFields;
    }
    get destination() {
        return this._destination;
    }
    putDestination(value) {
        this._destination.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get destinationInput() {
        return this._destination.internalValue;
    }
    get filter() {
        return this._filter;
    }
    putFilter(value) {
        this._filter.internalValue = value;
    }
    resetFilter() {
        this._filter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter.internalValue;
    }
    get schedule() {
        return this._schedule;
    }
    putSchedule(value) {
        this._schedule.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get scheduleInput() {
        return this._schedule.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            bucket: cdktf.stringToTerraform(this._bucket),
            enabled: cdktf.booleanToTerraform(this._enabled),
            id: cdktf.stringToTerraform(this._id),
            included_object_versions: cdktf.stringToTerraform(this._includedObjectVersions),
            name: cdktf.stringToTerraform(this._name),
            optional_fields: cdktf.listMapper(cdktf.stringToTerraform, false)(this._optionalFields),
            destination: s3BucketInventoryDestinationToTerraform(this._destination.internalValue),
            filter: s3BucketInventoryFilterToTerraform(this._filter.internalValue),
            schedule: s3BucketInventoryScheduleToTerraform(this._schedule.internalValue),
        };
    }
}
exports.S3BucketInventory = S3BucketInventory;
_h = JSII_RTTI_SYMBOL_1;
S3BucketInventory[_h] = { fqn: "@cdktf/provider-aws.s3BucketInventory.S3BucketInventory", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
S3BucketInventory.tfResourceType = "aws_s3_bucket_inventory";
