// @cdktf/provider-aws/lib/vpc-endpoint-service-allowed-principal/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VpcEndpointServiceAllowedPrincipal = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/vpc_endpoint_service_allowed_principal aws_vpc_endpoint_service_allowed_principal}
*/
class VpcEndpointServiceAllowedPrincipal extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/vpc_endpoint_service_allowed_principal aws_vpc_endpoint_service_allowed_principal} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options VpcEndpointServiceAllowedPrincipalConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_vpc_endpoint_service_allowed_principal',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._principalArn = config.principalArn;
        this._vpcEndpointServiceId = config.vpcEndpointServiceId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get principalArn() {
        return this.getStringAttribute('principal_arn');
    }
    set principalArn(value) {
        this._principalArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get principalArnInput() {
        return this._principalArn;
    }
    get vpcEndpointServiceId() {
        return this.getStringAttribute('vpc_endpoint_service_id');
    }
    set vpcEndpointServiceId(value) {
        this._vpcEndpointServiceId = value;
    }
    // Temporarily expose input value. Use with caution.
    get vpcEndpointServiceIdInput() {
        return this._vpcEndpointServiceId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            principal_arn: cdktf.stringToTerraform(this._principalArn),
            vpc_endpoint_service_id: cdktf.stringToTerraform(this._vpcEndpointServiceId),
        };
    }
}
exports.VpcEndpointServiceAllowedPrincipal = VpcEndpointServiceAllowedPrincipal;
_a = JSII_RTTI_SYMBOL_1;
VpcEndpointServiceAllowedPrincipal[_a] = { fqn: "@cdktf/provider-aws.vpcEndpointServiceAllowedPrincipal.VpcEndpointServiceAllowedPrincipal", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
VpcEndpointServiceAllowedPrincipal.tfResourceType = "aws_vpc_endpoint_service_allowed_principal";
