// @cdktf/provider-aws/lib/config-configuration-recorder/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigConfigurationRecorder = exports.ConfigConfigurationRecorderRecordingGroupOutputReference = exports.configConfigurationRecorderRecordingGroupToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function configConfigurationRecorderRecordingGroupToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        all_supported: cdktf.booleanToTerraform(struct.allSupported),
        include_global_resource_types: cdktf.booleanToTerraform(struct.includeGlobalResourceTypes),
        resource_types: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.resourceTypes),
    };
}
exports.configConfigurationRecorderRecordingGroupToTerraform = configConfigurationRecorderRecordingGroupToTerraform;
class ConfigConfigurationRecorderRecordingGroupOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._allSupported !== undefined) {
            hasAnyValues = true;
            internalValueResult.allSupported = this._allSupported;
        }
        if (this._includeGlobalResourceTypes !== undefined) {
            hasAnyValues = true;
            internalValueResult.includeGlobalResourceTypes = this._includeGlobalResourceTypes;
        }
        if (this._resourceTypes !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourceTypes = this._resourceTypes;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._allSupported = undefined;
            this._includeGlobalResourceTypes = undefined;
            this._resourceTypes = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._allSupported = value.allSupported;
            this._includeGlobalResourceTypes = value.includeGlobalResourceTypes;
            this._resourceTypes = value.resourceTypes;
        }
    }
    get allSupported() {
        return this.getBooleanAttribute('all_supported');
    }
    set allSupported(value) {
        this._allSupported = value;
    }
    resetAllSupported() {
        this._allSupported = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allSupportedInput() {
        return this._allSupported;
    }
    get includeGlobalResourceTypes() {
        return this.getBooleanAttribute('include_global_resource_types');
    }
    set includeGlobalResourceTypes(value) {
        this._includeGlobalResourceTypes = value;
    }
    resetIncludeGlobalResourceTypes() {
        this._includeGlobalResourceTypes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includeGlobalResourceTypesInput() {
        return this._includeGlobalResourceTypes;
    }
    get resourceTypes() {
        return cdktf.Fn.tolist(this.getListAttribute('resource_types'));
    }
    set resourceTypes(value) {
        this._resourceTypes = value;
    }
    resetResourceTypes() {
        this._resourceTypes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourceTypesInput() {
        return this._resourceTypes;
    }
}
exports.ConfigConfigurationRecorderRecordingGroupOutputReference = ConfigConfigurationRecorderRecordingGroupOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ConfigConfigurationRecorderRecordingGroupOutputReference[_a] = { fqn: "@cdktf/provider-aws.configConfigurationRecorder.ConfigConfigurationRecorderRecordingGroupOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/config_configuration_recorder aws_config_configuration_recorder}
*/
class ConfigConfigurationRecorder extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/config_configuration_recorder aws_config_configuration_recorder} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ConfigConfigurationRecorderConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_config_configuration_recorder',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // recording_group - computed: false, optional: true, required: false
        this._recordingGroup = new ConfigConfigurationRecorderRecordingGroupOutputReference(this, "recording_group");
        this._id = config.id;
        this._name = config.name;
        this._roleArn = config.roleArn;
        this._recordingGroup.internalValue = config.recordingGroup;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get recordingGroup() {
        return this._recordingGroup;
    }
    putRecordingGroup(value) {
        this._recordingGroup.internalValue = value;
    }
    resetRecordingGroup() {
        this._recordingGroup.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get recordingGroupInput() {
        return this._recordingGroup.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            role_arn: cdktf.stringToTerraform(this._roleArn),
            recording_group: configConfigurationRecorderRecordingGroupToTerraform(this._recordingGroup.internalValue),
        };
    }
}
exports.ConfigConfigurationRecorder = ConfigConfigurationRecorder;
_b = JSII_RTTI_SYMBOL_1;
ConfigConfigurationRecorder[_b] = { fqn: "@cdktf/provider-aws.configConfigurationRecorder.ConfigConfigurationRecorder", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
ConfigConfigurationRecorder.tfResourceType = "aws_config_configuration_recorder";
