// @cdktf/provider-aws/lib/data-aws-route53-zone/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsRoute53Zone = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/route53_zone aws_route53_zone}
*/
class DataAwsRoute53Zone extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/route53_zone aws_route53_zone} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsRoute53ZoneConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_route53_zone',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._name = config.name;
        this._privateZone = config.privateZone;
        this._resourceRecordSetCount = config.resourceRecordSetCount;
        this._tags = config.tags;
        this._vpcId = config.vpcId;
        this._zoneId = config.zoneId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // caller_reference - computed: true, optional: false, required: false
    get callerReference() {
        return this.getStringAttribute('caller_reference');
    }
    // comment - computed: true, optional: false, required: false
    get comment() {
        return this.getStringAttribute('comment');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // linked_service_description - computed: true, optional: false, required: false
    get linkedServiceDescription() {
        return this.getStringAttribute('linked_service_description');
    }
    // linked_service_principal - computed: true, optional: false, required: false
    get linkedServicePrincipal() {
        return this.getStringAttribute('linked_service_principal');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // name_servers - computed: true, optional: false, required: false
    get nameServers() {
        return this.getListAttribute('name_servers');
    }
    // primary_name_server - computed: true, optional: false, required: false
    get primaryNameServer() {
        return this.getStringAttribute('primary_name_server');
    }
    get privateZone() {
        return this.getBooleanAttribute('private_zone');
    }
    set privateZone(value) {
        this._privateZone = value;
    }
    resetPrivateZone() {
        this._privateZone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get privateZoneInput() {
        return this._privateZone;
    }
    get resourceRecordSetCount() {
        return this.getNumberAttribute('resource_record_set_count');
    }
    set resourceRecordSetCount(value) {
        this._resourceRecordSetCount = value;
    }
    resetResourceRecordSetCount() {
        this._resourceRecordSetCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourceRecordSetCountInput() {
        return this._resourceRecordSetCount;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    set vpcId(value) {
        this._vpcId = value;
    }
    resetVpcId() {
        this._vpcId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcIdInput() {
        return this._vpcId;
    }
    get zoneId() {
        return this.getStringAttribute('zone_id');
    }
    set zoneId(value) {
        this._zoneId = value;
    }
    resetZoneId() {
        this._zoneId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zoneIdInput() {
        return this._zoneId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            private_zone: cdktf.booleanToTerraform(this._privateZone),
            resource_record_set_count: cdktf.numberToTerraform(this._resourceRecordSetCount),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            vpc_id: cdktf.stringToTerraform(this._vpcId),
            zone_id: cdktf.stringToTerraform(this._zoneId),
        };
    }
}
exports.DataAwsRoute53Zone = DataAwsRoute53Zone;
_a = JSII_RTTI_SYMBOL_1;
DataAwsRoute53Zone[_a] = { fqn: "@cdktf/provider-aws.dataAwsRoute53Zone.DataAwsRoute53Zone", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsRoute53Zone.tfResourceType = "aws_route53_zone";
