// @nrwl/workspace/src/generators/new/generate-workspace-files.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateWorkspaceFiles = exports.DEFAULT_NRWL_PRETTIER_CONFIG = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nrwl/devkit");
const versions_1 = require("../../utils/versions");
const fs_1 = require("fs");
const path_1 = require("path");
const presets_1 = require("../utils/presets");
const default_base_1 = require("../../utilities/default-base");
exports.DEFAULT_NRWL_PRETTIER_CONFIG = {
    singleQuote: true,
};
function generateWorkspaceFiles(host, options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        if (!options.name) {
            throw new Error(`Invalid options, "name" is required.`);
        }
        options = normalizeOptions(options);
        createFiles(host, options);
        createNxJson(host, options);
        createPrettierrc(host, options);
        if (options.preset === presets_1.Preset.Angular) {
            decorateAngularClI(host, options);
        }
        const [packageMajor] = (0, devkit_1.getPackageManagerVersion)(options.packageManager).split('.');
        if (options.packageManager === 'pnpm' && +packageMajor >= 7) {
            createNpmrc(host, options);
        }
        else if (options.packageManager === 'yarn' && +packageMajor >= 2) {
            createYarnrcYml(host, options);
        }
        setPresetProperty(host, options);
        addNpmScripts(host, options);
        createAppsAndLibsFolders(host, options);
        setUpWorkspacesInPackageJson(host, options);
        yield (0, devkit_1.formatFiles)(host);
    });
}
exports.generateWorkspaceFiles = generateWorkspaceFiles;
function decorateAngularClI(host, options) {
    const decorateCli = (0, fs_1.readFileSync)((0, path_1.join)(__dirname, '..', 'utils', 'decorate-angular-cli.js__tmpl__')).toString();
    host.write((0, path_1.join)(options.directory, 'decorate-angular-cli.js'), decorateCli);
}
function setPresetProperty(tree, options) {
    (0, devkit_1.updateJson)(tree, (0, path_1.join)(options.directory, 'nx.json'), (json) => {
        if (options.preset === presets_1.Preset.Core || options.preset === presets_1.Preset.NPM) {
            addPropertyWithStableKeys(json, 'extends', 'nx/presets/npm.json');
            delete json.implicitDependencies;
            delete json.targetDefaults;
            delete json.targetDependencies;
            delete json.workspaceLayout;
            delete json.npmScope;
        }
        return json;
    });
}
function createAppsAndLibsFolders(host, options) {
    if (options.preset === presets_1.Preset.Core ||
        options.preset === presets_1.Preset.TS ||
        options.preset === presets_1.Preset.NPM) {
        host.write((0, path_1.join)(options.directory, 'packages/.gitkeep'), '');
    }
    else if (options.preset === presets_1.Preset.ReactExperimental) {
        // don't generate any folders
    }
    else {
        host.write((0, path_1.join)(options.directory, 'apps/.gitkeep'), '');
        host.write((0, path_1.join)(options.directory, 'libs/.gitkeep'), '');
    }
}
function createNxJson(host, { directory, npmScope, packageManager, defaultBase, preset }) {
    const nxJson = {
        $schema: './node_modules/nx/schemas/nx-schema.json',
        npmScope: npmScope,
        affected: {
            defaultBase,
        },
        tasksRunnerOptions: {
            default: {
                runner: 'nx/tasks-runners/default',
                options: {
                    cacheableOperations: ['build', 'lint', 'test', 'e2e'],
                },
            },
        },
    };
    nxJson.targetDefaults = {
        build: {
            dependsOn: ['^build'],
        },
    };
    if (defaultBase === 'main') {
        delete nxJson.affected;
    }
    if (preset !== presets_1.Preset.Core &&
        preset !== presets_1.Preset.NPM &&
        preset !== presets_1.Preset.Empty) {
        nxJson.namedInputs = {
            default: ['{projectRoot}/**/*', 'sharedGlobals'],
            production: ['default'],
            sharedGlobals: [],
        };
        nxJson.targetDefaults.build.inputs = ['production', '^production'];
    }
    (0, devkit_1.writeJson)(host, (0, path_1.join)(directory, 'nx.json'), nxJson);
}
function createFiles(host, options) {
    const formattedNames = (0, devkit_1.names)(options.name);
    const filesDirName = options.preset === presets_1.Preset.ReactExperimental
        ? './files-root-app'
        : options.preset === presets_1.Preset.NPM || options.preset === presets_1.Preset.Core
            ? './files-package-based-repo'
            : './files-integrated-repo';
    (0, devkit_1.generateFiles)(host, (0, path_1.join)(__dirname, filesDirName), options.directory, Object.assign(Object.assign({ formattedNames, dot: '.', tmpl: '', workspaceFile: options.preset === presets_1.Preset.Angular ? 'angular' : 'workspace', cliCommand: options.preset === presets_1.Preset.Angular ? 'ng' : 'nx', nxCli: false, typescriptVersion: versions_1.typescriptVersion,
        prettierVersion: versions_1.prettierVersion,
        // angular cli is used only when workspace schematics is added to angular cli
        angularCliVersion: versions_1.angularCliVersion }, options), { nxVersion: versions_1.nxVersion, packageManager: options.packageManager }));
}
function createPrettierrc(host, options) {
    (0, devkit_1.writeJson)(host, (0, path_1.join)(options.directory, '.prettierrc'), exports.DEFAULT_NRWL_PRETTIER_CONFIG);
}
// ensure that pnpm install add all the missing peer deps
function createNpmrc(host, options) {
    host.write((0, path_1.join)(options.directory, '.npmrc'), 'strict-peer-dependencies=false\nauto-install-peers=true\n');
}
// ensure that yarn (berry) install uses classic node linker
function createYarnrcYml(host, options) {
    host.write((0, path_1.join)(options.directory, '.yarnrc.yml'), 'nodeLinker: node-modules\n');
}
function addNpmScripts(host, options) {
    if (options.preset === presets_1.Preset.Angular) {
        (0, devkit_1.updateJson)(host, (0, path_1.join)(options.directory, 'package.json'), (json) => {
            Object.assign(json.scripts, {
                ng: 'nx',
                postinstall: 'node ./decorate-angular-cli.js',
            });
            return json;
        });
    }
    if (options.preset !== presets_1.Preset.TS &&
        options.preset !== presets_1.Preset.Core &&
        options.preset !== presets_1.Preset.NPM) {
        (0, devkit_1.updateJson)(host, (0, path_1.join)(options.directory, 'package.json'), (json) => {
            Object.assign(json.scripts, {
                start: 'nx serve',
                build: 'nx build',
                test: 'nx test',
            });
            return json;
        });
    }
}
function addPropertyWithStableKeys(obj, key, value) {
    const copy = Object.assign({}, obj);
    Object.keys(obj).forEach((k) => {
        delete obj[k];
    });
    obj[key] = value;
    Object.keys(copy).forEach((k) => {
        obj[k] = copy[k];
    });
}
function normalizeOptions(options) {
    let defaultBase = options.defaultBase || (0, default_base_1.deduceDefaultBase)();
    return Object.assign(Object.assign({ npmScope: options.name }, options), { defaultBase });
}
function setUpWorkspacesInPackageJson(tree, options) {
    if (options.preset === presets_1.Preset.NPM || options.preset === presets_1.Preset.Core) {
        if (options.packageManager === 'pnpm') {
            tree.write((0, path_1.join)(options.directory, 'pnpm-workspace.yaml'), `packages:
  - 'packages/*'
`);
        }
        else {
            (0, devkit_1.updateJson)(tree, (0, path_1.join)(options.directory, 'package.json'), (json) => {
                json.workspaces = ['packages/*'];
                return json;
            });
        }
    }
}
