// @cdktf/provider-aws/lib/redshiftserverless-namespace/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedshiftserverlessNamespace = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/redshiftserverless_namespace aws_redshiftserverless_namespace}
*/
class RedshiftserverlessNamespace extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/redshiftserverless_namespace aws_redshiftserverless_namespace} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options RedshiftserverlessNamespaceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_redshiftserverless_namespace',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._adminUserPassword = config.adminUserPassword;
        this._adminUsername = config.adminUsername;
        this._dbName = config.dbName;
        this._defaultIamRoleArn = config.defaultIamRoleArn;
        this._iamRoles = config.iamRoles;
        this._id = config.id;
        this._kmsKeyId = config.kmsKeyId;
        this._logExports = config.logExports;
        this._namespaceName = config.namespaceName;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    get adminUserPassword() {
        return this.getStringAttribute('admin_user_password');
    }
    set adminUserPassword(value) {
        this._adminUserPassword = value;
    }
    resetAdminUserPassword() {
        this._adminUserPassword = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get adminUserPasswordInput() {
        return this._adminUserPassword;
    }
    get adminUsername() {
        return this.getStringAttribute('admin_username');
    }
    set adminUsername(value) {
        this._adminUsername = value;
    }
    resetAdminUsername() {
        this._adminUsername = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get adminUsernameInput() {
        return this._adminUsername;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get dbName() {
        return this.getStringAttribute('db_name');
    }
    set dbName(value) {
        this._dbName = value;
    }
    resetDbName() {
        this._dbName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dbNameInput() {
        return this._dbName;
    }
    get defaultIamRoleArn() {
        return this.getStringAttribute('default_iam_role_arn');
    }
    set defaultIamRoleArn(value) {
        this._defaultIamRoleArn = value;
    }
    resetDefaultIamRoleArn() {
        this._defaultIamRoleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultIamRoleArnInput() {
        return this._defaultIamRoleArn;
    }
    get iamRoles() {
        return cdktf.Fn.tolist(this.getListAttribute('iam_roles'));
    }
    set iamRoles(value) {
        this._iamRoles = value;
    }
    resetIamRoles() {
        this._iamRoles = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iamRolesInput() {
        return this._iamRoles;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    resetKmsKeyId() {
        this._kmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
    get logExports() {
        return cdktf.Fn.tolist(this.getListAttribute('log_exports'));
    }
    set logExports(value) {
        this._logExports = value;
    }
    resetLogExports() {
        this._logExports = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get logExportsInput() {
        return this._logExports;
    }
    // namespace_id - computed: true, optional: false, required: false
    get namespaceId() {
        return this.getStringAttribute('namespace_id');
    }
    get namespaceName() {
        return this.getStringAttribute('namespace_name');
    }
    set namespaceName(value) {
        this._namespaceName = value;
    }
    // Temporarily expose input value. Use with caution.
    get namespaceNameInput() {
        return this._namespaceName;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            admin_user_password: cdktf.stringToTerraform(this._adminUserPassword),
            admin_username: cdktf.stringToTerraform(this._adminUsername),
            db_name: cdktf.stringToTerraform(this._dbName),
            default_iam_role_arn: cdktf.stringToTerraform(this._defaultIamRoleArn),
            iam_roles: cdktf.listMapper(cdktf.stringToTerraform, false)(this._iamRoles),
            id: cdktf.stringToTerraform(this._id),
            kms_key_id: cdktf.stringToTerraform(this._kmsKeyId),
            log_exports: cdktf.listMapper(cdktf.stringToTerraform, false)(this._logExports),
            namespace_name: cdktf.stringToTerraform(this._namespaceName),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
        };
    }
}
exports.RedshiftserverlessNamespace = RedshiftserverlessNamespace;
_a = JSII_RTTI_SYMBOL_1;
RedshiftserverlessNamespace[_a] = { fqn: "@cdktf/provider-aws.redshiftserverlessNamespace.RedshiftserverlessNamespace", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
RedshiftserverlessNamespace.tfResourceType = "aws_redshiftserverless_namespace";
