// @cdktf/provider-aws/lib/data-aws-location-place-index/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsLocationPlaceIndex = exports.DataAwsLocationPlaceIndexDataSourceConfigurationList = exports.DataAwsLocationPlaceIndexDataSourceConfigurationOutputReference = exports.dataAwsLocationPlaceIndexDataSourceConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsLocationPlaceIndexDataSourceConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsLocationPlaceIndexDataSourceConfigurationToTerraform = dataAwsLocationPlaceIndexDataSourceConfigurationToTerraform;
class DataAwsLocationPlaceIndexDataSourceConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // intended_use - computed: true, optional: false, required: false
    get intendedUse() {
        return this.getStringAttribute('intended_use');
    }
}
exports.DataAwsLocationPlaceIndexDataSourceConfigurationOutputReference = DataAwsLocationPlaceIndexDataSourceConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsLocationPlaceIndexDataSourceConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsLocationPlaceIndex.DataAwsLocationPlaceIndexDataSourceConfigurationOutputReference", version: "10.0.12" };
class DataAwsLocationPlaceIndexDataSourceConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsLocationPlaceIndexDataSourceConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsLocationPlaceIndexDataSourceConfigurationList = DataAwsLocationPlaceIndexDataSourceConfigurationList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsLocationPlaceIndexDataSourceConfigurationList[_b] = { fqn: "@cdktf/provider-aws.dataAwsLocationPlaceIndex.DataAwsLocationPlaceIndexDataSourceConfigurationList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/location_place_index aws_location_place_index}
*/
class DataAwsLocationPlaceIndex extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/location_place_index aws_location_place_index} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsLocationPlaceIndexConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_location_place_index',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // data_source_configuration - computed: true, optional: false, required: false
        this._dataSourceConfiguration = new DataAwsLocationPlaceIndexDataSourceConfigurationList(this, "data_source_configuration", false);
        this._id = config.id;
        this._indexName = config.indexName;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    // data_source - computed: true, optional: false, required: false
    get dataSource() {
        return this.getStringAttribute('data_source');
    }
    get dataSourceConfiguration() {
        return this._dataSourceConfiguration;
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // index_arn - computed: true, optional: false, required: false
    get indexArn() {
        return this.getStringAttribute('index_arn');
    }
    get indexName() {
        return this.getStringAttribute('index_name');
    }
    set indexName(value) {
        this._indexName = value;
    }
    // Temporarily expose input value. Use with caution.
    get indexNameInput() {
        return this._indexName;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            index_name: cdktf.stringToTerraform(this._indexName),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
}
exports.DataAwsLocationPlaceIndex = DataAwsLocationPlaceIndex;
_c = JSII_RTTI_SYMBOL_1;
DataAwsLocationPlaceIndex[_c] = { fqn: "@cdktf/provider-aws.dataAwsLocationPlaceIndex.DataAwsLocationPlaceIndex", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsLocationPlaceIndex.tfResourceType = "aws_location_place_index";
