// @cdktf/provider-aws/lib/elastictranscoder-pipeline/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElastictranscoderPipeline = exports.ElastictranscoderPipelineThumbnailConfigPermissionsList = exports.ElastictranscoderPipelineThumbnailConfigPermissionsOutputReference = exports.elastictranscoderPipelineThumbnailConfigPermissionsToTerraform = exports.ElastictranscoderPipelineThumbnailConfigOutputReference = exports.elastictranscoderPipelineThumbnailConfigToTerraform = exports.ElastictranscoderPipelineNotificationsOutputReference = exports.elastictranscoderPipelineNotificationsToTerraform = exports.ElastictranscoderPipelineContentConfigPermissionsList = exports.ElastictranscoderPipelineContentConfigPermissionsOutputReference = exports.elastictranscoderPipelineContentConfigPermissionsToTerraform = exports.ElastictranscoderPipelineContentConfigOutputReference = exports.elastictranscoderPipelineContentConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function elastictranscoderPipelineContentConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket: cdktf.stringToTerraform(struct.bucket),
        storage_class: cdktf.stringToTerraform(struct.storageClass),
    };
}
exports.elastictranscoderPipelineContentConfigToTerraform = elastictranscoderPipelineContentConfigToTerraform;
class ElastictranscoderPipelineContentConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucket !== undefined) {
            hasAnyValues = true;
            internalValueResult.bucket = this._bucket;
        }
        if (this._storageClass !== undefined) {
            hasAnyValues = true;
            internalValueResult.storageClass = this._storageClass;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucket = undefined;
            this._storageClass = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucket = value.bucket;
            this._storageClass = value.storageClass;
        }
    }
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    set bucket(value) {
        this._bucket = value;
    }
    resetBucket() {
        this._bucket = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bucketInput() {
        return this._bucket;
    }
    get storageClass() {
        return this.getStringAttribute('storage_class');
    }
    set storageClass(value) {
        this._storageClass = value;
    }
    resetStorageClass() {
        this._storageClass = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get storageClassInput() {
        return this._storageClass;
    }
}
exports.ElastictranscoderPipelineContentConfigOutputReference = ElastictranscoderPipelineContentConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ElastictranscoderPipelineContentConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.elastictranscoderPipeline.ElastictranscoderPipelineContentConfigOutputReference", version: "10.0.12" };
function elastictranscoderPipelineContentConfigPermissionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        access: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.access),
        grantee: cdktf.stringToTerraform(struct.grantee),
        grantee_type: cdktf.stringToTerraform(struct.granteeType),
    };
}
exports.elastictranscoderPipelineContentConfigPermissionsToTerraform = elastictranscoderPipelineContentConfigPermissionsToTerraform;
class ElastictranscoderPipelineContentConfigPermissionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._access !== undefined) {
            hasAnyValues = true;
            internalValueResult.access = this._access;
        }
        if (this._grantee !== undefined) {
            hasAnyValues = true;
            internalValueResult.grantee = this._grantee;
        }
        if (this._granteeType !== undefined) {
            hasAnyValues = true;
            internalValueResult.granteeType = this._granteeType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._access = undefined;
            this._grantee = undefined;
            this._granteeType = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._access = value.access;
            this._grantee = value.grantee;
            this._granteeType = value.granteeType;
        }
    }
    get access() {
        return this.getListAttribute('access');
    }
    set access(value) {
        this._access = value;
    }
    resetAccess() {
        this._access = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessInput() {
        return this._access;
    }
    get grantee() {
        return this.getStringAttribute('grantee');
    }
    set grantee(value) {
        this._grantee = value;
    }
    resetGrantee() {
        this._grantee = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get granteeInput() {
        return this._grantee;
    }
    get granteeType() {
        return this.getStringAttribute('grantee_type');
    }
    set granteeType(value) {
        this._granteeType = value;
    }
    resetGranteeType() {
        this._granteeType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get granteeTypeInput() {
        return this._granteeType;
    }
}
exports.ElastictranscoderPipelineContentConfigPermissionsOutputReference = ElastictranscoderPipelineContentConfigPermissionsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
ElastictranscoderPipelineContentConfigPermissionsOutputReference[_b] = { fqn: "@cdktf/provider-aws.elastictranscoderPipeline.ElastictranscoderPipelineContentConfigPermissionsOutputReference", version: "10.0.12" };
class ElastictranscoderPipelineContentConfigPermissionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ElastictranscoderPipelineContentConfigPermissionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ElastictranscoderPipelineContentConfigPermissionsList = ElastictranscoderPipelineContentConfigPermissionsList;
_c = JSII_RTTI_SYMBOL_1;
ElastictranscoderPipelineContentConfigPermissionsList[_c] = { fqn: "@cdktf/provider-aws.elastictranscoderPipeline.ElastictranscoderPipelineContentConfigPermissionsList", version: "10.0.12" };
function elastictranscoderPipelineNotificationsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        completed: cdktf.stringToTerraform(struct.completed),
        error: cdktf.stringToTerraform(struct.error),
        progressing: cdktf.stringToTerraform(struct.progressing),
        warning: cdktf.stringToTerraform(struct.warning),
    };
}
exports.elastictranscoderPipelineNotificationsToTerraform = elastictranscoderPipelineNotificationsToTerraform;
class ElastictranscoderPipelineNotificationsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._completed !== undefined) {
            hasAnyValues = true;
            internalValueResult.completed = this._completed;
        }
        if (this._error !== undefined) {
            hasAnyValues = true;
            internalValueResult.error = this._error;
        }
        if (this._progressing !== undefined) {
            hasAnyValues = true;
            internalValueResult.progressing = this._progressing;
        }
        if (this._warning !== undefined) {
            hasAnyValues = true;
            internalValueResult.warning = this._warning;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._completed = undefined;
            this._error = undefined;
            this._progressing = undefined;
            this._warning = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._completed = value.completed;
            this._error = value.error;
            this._progressing = value.progressing;
            this._warning = value.warning;
        }
    }
    get completed() {
        return this.getStringAttribute('completed');
    }
    set completed(value) {
        this._completed = value;
    }
    resetCompleted() {
        this._completed = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get completedInput() {
        return this._completed;
    }
    get error() {
        return this.getStringAttribute('error');
    }
    set error(value) {
        this._error = value;
    }
    resetError() {
        this._error = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get errorInput() {
        return this._error;
    }
    get progressing() {
        return this.getStringAttribute('progressing');
    }
    set progressing(value) {
        this._progressing = value;
    }
    resetProgressing() {
        this._progressing = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get progressingInput() {
        return this._progressing;
    }
    get warning() {
        return this.getStringAttribute('warning');
    }
    set warning(value) {
        this._warning = value;
    }
    resetWarning() {
        this._warning = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get warningInput() {
        return this._warning;
    }
}
exports.ElastictranscoderPipelineNotificationsOutputReference = ElastictranscoderPipelineNotificationsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
ElastictranscoderPipelineNotificationsOutputReference[_d] = { fqn: "@cdktf/provider-aws.elastictranscoderPipeline.ElastictranscoderPipelineNotificationsOutputReference", version: "10.0.12" };
function elastictranscoderPipelineThumbnailConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket: cdktf.stringToTerraform(struct.bucket),
        storage_class: cdktf.stringToTerraform(struct.storageClass),
    };
}
exports.elastictranscoderPipelineThumbnailConfigToTerraform = elastictranscoderPipelineThumbnailConfigToTerraform;
class ElastictranscoderPipelineThumbnailConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucket !== undefined) {
            hasAnyValues = true;
            internalValueResult.bucket = this._bucket;
        }
        if (this._storageClass !== undefined) {
            hasAnyValues = true;
            internalValueResult.storageClass = this._storageClass;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucket = undefined;
            this._storageClass = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucket = value.bucket;
            this._storageClass = value.storageClass;
        }
    }
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    set bucket(value) {
        this._bucket = value;
    }
    resetBucket() {
        this._bucket = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bucketInput() {
        return this._bucket;
    }
    get storageClass() {
        return this.getStringAttribute('storage_class');
    }
    set storageClass(value) {
        this._storageClass = value;
    }
    resetStorageClass() {
        this._storageClass = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get storageClassInput() {
        return this._storageClass;
    }
}
exports.ElastictranscoderPipelineThumbnailConfigOutputReference = ElastictranscoderPipelineThumbnailConfigOutputReference;
_e = JSII_RTTI_SYMBOL_1;
ElastictranscoderPipelineThumbnailConfigOutputReference[_e] = { fqn: "@cdktf/provider-aws.elastictranscoderPipeline.ElastictranscoderPipelineThumbnailConfigOutputReference", version: "10.0.12" };
function elastictranscoderPipelineThumbnailConfigPermissionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        access: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.access),
        grantee: cdktf.stringToTerraform(struct.grantee),
        grantee_type: cdktf.stringToTerraform(struct.granteeType),
    };
}
exports.elastictranscoderPipelineThumbnailConfigPermissionsToTerraform = elastictranscoderPipelineThumbnailConfigPermissionsToTerraform;
class ElastictranscoderPipelineThumbnailConfigPermissionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._access !== undefined) {
            hasAnyValues = true;
            internalValueResult.access = this._access;
        }
        if (this._grantee !== undefined) {
            hasAnyValues = true;
            internalValueResult.grantee = this._grantee;
        }
        if (this._granteeType !== undefined) {
            hasAnyValues = true;
            internalValueResult.granteeType = this._granteeType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._access = undefined;
            this._grantee = undefined;
            this._granteeType = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._access = value.access;
            this._grantee = value.grantee;
            this._granteeType = value.granteeType;
        }
    }
    get access() {
        return this.getListAttribute('access');
    }
    set access(value) {
        this._access = value;
    }
    resetAccess() {
        this._access = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessInput() {
        return this._access;
    }
    get grantee() {
        return this.getStringAttribute('grantee');
    }
    set grantee(value) {
        this._grantee = value;
    }
    resetGrantee() {
        this._grantee = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get granteeInput() {
        return this._grantee;
    }
    get granteeType() {
        return this.getStringAttribute('grantee_type');
    }
    set granteeType(value) {
        this._granteeType = value;
    }
    resetGranteeType() {
        this._granteeType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get granteeTypeInput() {
        return this._granteeType;
    }
}
exports.ElastictranscoderPipelineThumbnailConfigPermissionsOutputReference = ElastictranscoderPipelineThumbnailConfigPermissionsOutputReference;
_f = JSII_RTTI_SYMBOL_1;
ElastictranscoderPipelineThumbnailConfigPermissionsOutputReference[_f] = { fqn: "@cdktf/provider-aws.elastictranscoderPipeline.ElastictranscoderPipelineThumbnailConfigPermissionsOutputReference", version: "10.0.12" };
class ElastictranscoderPipelineThumbnailConfigPermissionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ElastictranscoderPipelineThumbnailConfigPermissionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ElastictranscoderPipelineThumbnailConfigPermissionsList = ElastictranscoderPipelineThumbnailConfigPermissionsList;
_g = JSII_RTTI_SYMBOL_1;
ElastictranscoderPipelineThumbnailConfigPermissionsList[_g] = { fqn: "@cdktf/provider-aws.elastictranscoderPipeline.ElastictranscoderPipelineThumbnailConfigPermissionsList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/elastictranscoder_pipeline aws_elastictranscoder_pipeline}
*/
class ElastictranscoderPipeline extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/elastictranscoder_pipeline aws_elastictranscoder_pipeline} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ElastictranscoderPipelineConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_elastictranscoder_pipeline',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // content_config - computed: false, optional: true, required: false
        this._contentConfig = new ElastictranscoderPipelineContentConfigOutputReference(this, "content_config");
        // content_config_permissions - computed: false, optional: true, required: false
        this._contentConfigPermissions = new ElastictranscoderPipelineContentConfigPermissionsList(this, "content_config_permissions", true);
        // notifications - computed: false, optional: true, required: false
        this._notifications = new ElastictranscoderPipelineNotificationsOutputReference(this, "notifications");
        // thumbnail_config - computed: false, optional: true, required: false
        this._thumbnailConfig = new ElastictranscoderPipelineThumbnailConfigOutputReference(this, "thumbnail_config");
        // thumbnail_config_permissions - computed: false, optional: true, required: false
        this._thumbnailConfigPermissions = new ElastictranscoderPipelineThumbnailConfigPermissionsList(this, "thumbnail_config_permissions", true);
        this._awsKmsKeyArn = config.awsKmsKeyArn;
        this._id = config.id;
        this._inputBucket = config.inputBucket;
        this._name = config.name;
        this._outputBucket = config.outputBucket;
        this._role = config.role;
        this._contentConfig.internalValue = config.contentConfig;
        this._contentConfigPermissions.internalValue = config.contentConfigPermissions;
        this._notifications.internalValue = config.notifications;
        this._thumbnailConfig.internalValue = config.thumbnailConfig;
        this._thumbnailConfigPermissions.internalValue = config.thumbnailConfigPermissions;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get awsKmsKeyArn() {
        return this.getStringAttribute('aws_kms_key_arn');
    }
    set awsKmsKeyArn(value) {
        this._awsKmsKeyArn = value;
    }
    resetAwsKmsKeyArn() {
        this._awsKmsKeyArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get awsKmsKeyArnInput() {
        return this._awsKmsKeyArn;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get inputBucket() {
        return this.getStringAttribute('input_bucket');
    }
    set inputBucket(value) {
        this._inputBucket = value;
    }
    // Temporarily expose input value. Use with caution.
    get inputBucketInput() {
        return this._inputBucket;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get outputBucket() {
        return this.getStringAttribute('output_bucket');
    }
    set outputBucket(value) {
        this._outputBucket = value;
    }
    resetOutputBucket() {
        this._outputBucket = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get outputBucketInput() {
        return this._outputBucket;
    }
    get role() {
        return this.getStringAttribute('role');
    }
    set role(value) {
        this._role = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleInput() {
        return this._role;
    }
    get contentConfig() {
        return this._contentConfig;
    }
    putContentConfig(value) {
        this._contentConfig.internalValue = value;
    }
    resetContentConfig() {
        this._contentConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get contentConfigInput() {
        return this._contentConfig.internalValue;
    }
    get contentConfigPermissions() {
        return this._contentConfigPermissions;
    }
    putContentConfigPermissions(value) {
        this._contentConfigPermissions.internalValue = value;
    }
    resetContentConfigPermissions() {
        this._contentConfigPermissions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get contentConfigPermissionsInput() {
        return this._contentConfigPermissions.internalValue;
    }
    get notifications() {
        return this._notifications;
    }
    putNotifications(value) {
        this._notifications.internalValue = value;
    }
    resetNotifications() {
        this._notifications.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get notificationsInput() {
        return this._notifications.internalValue;
    }
    get thumbnailConfig() {
        return this._thumbnailConfig;
    }
    putThumbnailConfig(value) {
        this._thumbnailConfig.internalValue = value;
    }
    resetThumbnailConfig() {
        this._thumbnailConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get thumbnailConfigInput() {
        return this._thumbnailConfig.internalValue;
    }
    get thumbnailConfigPermissions() {
        return this._thumbnailConfigPermissions;
    }
    putThumbnailConfigPermissions(value) {
        this._thumbnailConfigPermissions.internalValue = value;
    }
    resetThumbnailConfigPermissions() {
        this._thumbnailConfigPermissions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get thumbnailConfigPermissionsInput() {
        return this._thumbnailConfigPermissions.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            aws_kms_key_arn: cdktf.stringToTerraform(this._awsKmsKeyArn),
            id: cdktf.stringToTerraform(this._id),
            input_bucket: cdktf.stringToTerraform(this._inputBucket),
            name: cdktf.stringToTerraform(this._name),
            output_bucket: cdktf.stringToTerraform(this._outputBucket),
            role: cdktf.stringToTerraform(this._role),
            content_config: elastictranscoderPipelineContentConfigToTerraform(this._contentConfig.internalValue),
            content_config_permissions: cdktf.listMapper(elastictranscoderPipelineContentConfigPermissionsToTerraform, true)(this._contentConfigPermissions.internalValue),
            notifications: elastictranscoderPipelineNotificationsToTerraform(this._notifications.internalValue),
            thumbnail_config: elastictranscoderPipelineThumbnailConfigToTerraform(this._thumbnailConfig.internalValue),
            thumbnail_config_permissions: cdktf.listMapper(elastictranscoderPipelineThumbnailConfigPermissionsToTerraform, true)(this._thumbnailConfigPermissions.internalValue),
        };
    }
}
exports.ElastictranscoderPipeline = ElastictranscoderPipeline;
_h = JSII_RTTI_SYMBOL_1;
ElastictranscoderPipeline[_h] = { fqn: "@cdktf/provider-aws.elastictranscoderPipeline.ElastictranscoderPipeline", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
ElastictranscoderPipeline.tfResourceType = "aws_elastictranscoder_pipeline";
