// @cdktf/provider-aws/lib/dynamodb-table-item/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DynamodbTableItem = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/dynamodb_table_item aws_dynamodb_table_item}
*/
class DynamodbTableItem extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/dynamodb_table_item aws_dynamodb_table_item} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DynamodbTableItemConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_dynamodb_table_item',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._hashKey = config.hashKey;
        this._id = config.id;
        this._item = config.item;
        this._rangeKey = config.rangeKey;
        this._tableName = config.tableName;
    }
    get hashKey() {
        return this.getStringAttribute('hash_key');
    }
    set hashKey(value) {
        this._hashKey = value;
    }
    // Temporarily expose input value. Use with caution.
    get hashKeyInput() {
        return this._hashKey;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get item() {
        return this.getStringAttribute('item');
    }
    set item(value) {
        this._item = value;
    }
    // Temporarily expose input value. Use with caution.
    get itemInput() {
        return this._item;
    }
    get rangeKey() {
        return this.getStringAttribute('range_key');
    }
    set rangeKey(value) {
        this._rangeKey = value;
    }
    resetRangeKey() {
        this._rangeKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rangeKeyInput() {
        return this._rangeKey;
    }
    get tableName() {
        return this.getStringAttribute('table_name');
    }
    set tableName(value) {
        this._tableName = value;
    }
    // Temporarily expose input value. Use with caution.
    get tableNameInput() {
        return this._tableName;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            hash_key: cdktf.stringToTerraform(this._hashKey),
            id: cdktf.stringToTerraform(this._id),
            item: cdktf.stringToTerraform(this._item),
            range_key: cdktf.stringToTerraform(this._rangeKey),
            table_name: cdktf.stringToTerraform(this._tableName),
        };
    }
}
exports.DynamodbTableItem = DynamodbTableItem;
_a = JSII_RTTI_SYMBOL_1;
DynamodbTableItem[_a] = { fqn: "@cdktf/provider-aws.dynamodbTableItem.DynamodbTableItem", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DynamodbTableItem.tfResourceType = "aws_dynamodb_table_item";
