// @cdktf/provider-aws/lib/iot-role-alias/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IotRoleAlias = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/iot_role_alias aws_iot_role_alias}
*/
class IotRoleAlias extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/iot_role_alias aws_iot_role_alias} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options IotRoleAliasConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iot_role_alias',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._alias = config.alias;
        this._credentialDuration = config.credentialDuration;
        this._id = config.id;
        this._roleArn = config.roleArn;
    }
    get alias() {
        return this.getStringAttribute('alias');
    }
    set alias(value) {
        this._alias = value;
    }
    // Temporarily expose input value. Use with caution.
    get aliasInput() {
        return this._alias;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get credentialDuration() {
        return this.getNumberAttribute('credential_duration');
    }
    set credentialDuration(value) {
        this._credentialDuration = value;
    }
    resetCredentialDuration() {
        this._credentialDuration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get credentialDurationInput() {
        return this._credentialDuration;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            alias: cdktf.stringToTerraform(this._alias),
            credential_duration: cdktf.numberToTerraform(this._credentialDuration),
            id: cdktf.stringToTerraform(this._id),
            role_arn: cdktf.stringToTerraform(this._roleArn),
        };
    }
}
exports.IotRoleAlias = IotRoleAlias;
_a = JSII_RTTI_SYMBOL_1;
IotRoleAlias[_a] = { fqn: "@cdktf/provider-aws.iotRoleAlias.IotRoleAlias", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
IotRoleAlias.tfResourceType = "aws_iot_role_alias";
