// @cdktf/provider-aws/lib/db-proxy-target/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DbProxyTarget = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/db_proxy_target aws_db_proxy_target}
*/
class DbProxyTarget extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/db_proxy_target aws_db_proxy_target} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DbProxyTargetConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_db_proxy_target',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._dbClusterIdentifier = config.dbClusterIdentifier;
        this._dbInstanceIdentifier = config.dbInstanceIdentifier;
        this._dbProxyName = config.dbProxyName;
        this._id = config.id;
        this._targetGroupName = config.targetGroupName;
    }
    get dbClusterIdentifier() {
        return this.getStringAttribute('db_cluster_identifier');
    }
    set dbClusterIdentifier(value) {
        this._dbClusterIdentifier = value;
    }
    resetDbClusterIdentifier() {
        this._dbClusterIdentifier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dbClusterIdentifierInput() {
        return this._dbClusterIdentifier;
    }
    get dbInstanceIdentifier() {
        return this.getStringAttribute('db_instance_identifier');
    }
    set dbInstanceIdentifier(value) {
        this._dbInstanceIdentifier = value;
    }
    resetDbInstanceIdentifier() {
        this._dbInstanceIdentifier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dbInstanceIdentifierInput() {
        return this._dbInstanceIdentifier;
    }
    get dbProxyName() {
        return this.getStringAttribute('db_proxy_name');
    }
    set dbProxyName(value) {
        this._dbProxyName = value;
    }
    // Temporarily expose input value. Use with caution.
    get dbProxyNameInput() {
        return this._dbProxyName;
    }
    // endpoint - computed: true, optional: false, required: false
    get endpoint() {
        return this.getStringAttribute('endpoint');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getNumberAttribute('port');
    }
    // rds_resource_id - computed: true, optional: false, required: false
    get rdsResourceId() {
        return this.getStringAttribute('rds_resource_id');
    }
    // target_arn - computed: true, optional: false, required: false
    get targetArn() {
        return this.getStringAttribute('target_arn');
    }
    get targetGroupName() {
        return this.getStringAttribute('target_group_name');
    }
    set targetGroupName(value) {
        this._targetGroupName = value;
    }
    // Temporarily expose input value. Use with caution.
    get targetGroupNameInput() {
        return this._targetGroupName;
    }
    // tracked_cluster_id - computed: true, optional: false, required: false
    get trackedClusterId() {
        return this.getStringAttribute('tracked_cluster_id');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            db_cluster_identifier: cdktf.stringToTerraform(this._dbClusterIdentifier),
            db_instance_identifier: cdktf.stringToTerraform(this._dbInstanceIdentifier),
            db_proxy_name: cdktf.stringToTerraform(this._dbProxyName),
            id: cdktf.stringToTerraform(this._id),
            target_group_name: cdktf.stringToTerraform(this._targetGroupName),
        };
    }
}
exports.DbProxyTarget = DbProxyTarget;
_a = JSII_RTTI_SYMBOL_1;
DbProxyTarget[_a] = { fqn: "@cdktf/provider-aws.dbProxyTarget.DbProxyTarget", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DbProxyTarget.tfResourceType = "aws_db_proxy_target";
