// @cdktf/provider-aws/lib/data-aws-appconfig-environments/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsAppconfigEnvironments = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/appconfig_environments aws_appconfig_environments}
*/
class DataAwsAppconfigEnvironments extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/appconfig_environments aws_appconfig_environments} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsAppconfigEnvironmentsConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_appconfig_environments',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._applicationId = config.applicationId;
        this._id = config.id;
    }
    get applicationId() {
        return this.getStringAttribute('application_id');
    }
    set applicationId(value) {
        this._applicationId = value;
    }
    // Temporarily expose input value. Use with caution.
    get applicationIdInput() {
        return this._applicationId;
    }
    // environment_ids - computed: true, optional: false, required: false
    get environmentIds() {
        return cdktf.Fn.tolist(this.getListAttribute('environment_ids'));
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            application_id: cdktf.stringToTerraform(this._applicationId),
            id: cdktf.stringToTerraform(this._id),
        };
    }
}
exports.DataAwsAppconfigEnvironments = DataAwsAppconfigEnvironments;
_a = JSII_RTTI_SYMBOL_1;
DataAwsAppconfigEnvironments[_a] = { fqn: "@cdktf/provider-aws.dataAwsAppconfigEnvironments.DataAwsAppconfigEnvironments", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsAppconfigEnvironments.tfResourceType = "aws_appconfig_environments";
