// @cdktf/provider-aws/lib/grafana-workspace/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrafanaWorkspace = exports.GrafanaWorkspaceTimeoutsOutputReference = exports.grafanaWorkspaceTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function grafanaWorkspaceTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.grafanaWorkspaceTimeoutsToTerraform = grafanaWorkspaceTimeoutsToTerraform;
class GrafanaWorkspaceTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.GrafanaWorkspaceTimeoutsOutputReference = GrafanaWorkspaceTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
GrafanaWorkspaceTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.grafanaWorkspace.GrafanaWorkspaceTimeoutsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/grafana_workspace aws_grafana_workspace}
*/
class GrafanaWorkspace extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/grafana_workspace aws_grafana_workspace} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options GrafanaWorkspaceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_grafana_workspace',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new GrafanaWorkspaceTimeoutsOutputReference(this, "timeouts");
        this._accountAccessType = config.accountAccessType;
        this._authenticationProviders = config.authenticationProviders;
        this._dataSources = config.dataSources;
        this._description = config.description;
        this._id = config.id;
        this._name = config.name;
        this._notificationDestinations = config.notificationDestinations;
        this._organizationRoleName = config.organizationRoleName;
        this._organizationalUnits = config.organizationalUnits;
        this._permissionType = config.permissionType;
        this._roleArn = config.roleArn;
        this._stackSetName = config.stackSetName;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._timeouts.internalValue = config.timeouts;
    }
    get accountAccessType() {
        return this.getStringAttribute('account_access_type');
    }
    set accountAccessType(value) {
        this._accountAccessType = value;
    }
    // Temporarily expose input value. Use with caution.
    get accountAccessTypeInput() {
        return this._accountAccessType;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get authenticationProviders() {
        return this.getListAttribute('authentication_providers');
    }
    set authenticationProviders(value) {
        this._authenticationProviders = value;
    }
    // Temporarily expose input value. Use with caution.
    get authenticationProvidersInput() {
        return this._authenticationProviders;
    }
    get dataSources() {
        return this.getListAttribute('data_sources');
    }
    set dataSources(value) {
        this._dataSources = value;
    }
    resetDataSources() {
        this._dataSources = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataSourcesInput() {
        return this._dataSources;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    // endpoint - computed: true, optional: false, required: false
    get endpoint() {
        return this.getStringAttribute('endpoint');
    }
    // grafana_version - computed: true, optional: false, required: false
    get grafanaVersion() {
        return this.getStringAttribute('grafana_version');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get notificationDestinations() {
        return this.getListAttribute('notification_destinations');
    }
    set notificationDestinations(value) {
        this._notificationDestinations = value;
    }
    resetNotificationDestinations() {
        this._notificationDestinations = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get notificationDestinationsInput() {
        return this._notificationDestinations;
    }
    get organizationRoleName() {
        return this.getStringAttribute('organization_role_name');
    }
    set organizationRoleName(value) {
        this._organizationRoleName = value;
    }
    resetOrganizationRoleName() {
        this._organizationRoleName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get organizationRoleNameInput() {
        return this._organizationRoleName;
    }
    get organizationalUnits() {
        return this.getListAttribute('organizational_units');
    }
    set organizationalUnits(value) {
        this._organizationalUnits = value;
    }
    resetOrganizationalUnits() {
        this._organizationalUnits = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get organizationalUnitsInput() {
        return this._organizationalUnits;
    }
    get permissionType() {
        return this.getStringAttribute('permission_type');
    }
    set permissionType(value) {
        this._permissionType = value;
    }
    // Temporarily expose input value. Use with caution.
    get permissionTypeInput() {
        return this._permissionType;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    resetRoleArn() {
        this._roleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    // saml_configuration_status - computed: true, optional: false, required: false
    get samlConfigurationStatus() {
        return this.getStringAttribute('saml_configuration_status');
    }
    get stackSetName() {
        return this.getStringAttribute('stack_set_name');
    }
    set stackSetName(value) {
        this._stackSetName = value;
    }
    resetStackSetName() {
        this._stackSetName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get stackSetNameInput() {
        return this._stackSetName;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            account_access_type: cdktf.stringToTerraform(this._accountAccessType),
            authentication_providers: cdktf.listMapper(cdktf.stringToTerraform, false)(this._authenticationProviders),
            data_sources: cdktf.listMapper(cdktf.stringToTerraform, false)(this._dataSources),
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            notification_destinations: cdktf.listMapper(cdktf.stringToTerraform, false)(this._notificationDestinations),
            organization_role_name: cdktf.stringToTerraform(this._organizationRoleName),
            organizational_units: cdktf.listMapper(cdktf.stringToTerraform, false)(this._organizationalUnits),
            permission_type: cdktf.stringToTerraform(this._permissionType),
            role_arn: cdktf.stringToTerraform(this._roleArn),
            stack_set_name: cdktf.stringToTerraform(this._stackSetName),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            timeouts: grafanaWorkspaceTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.GrafanaWorkspace = GrafanaWorkspace;
_b = JSII_RTTI_SYMBOL_1;
GrafanaWorkspace[_b] = { fqn: "@cdktf/provider-aws.grafanaWorkspace.GrafanaWorkspace", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
GrafanaWorkspace.tfResourceType = "aws_grafana_workspace";
