// @cdktf/provider-aws/lib/appmesh-gateway-route/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppmeshGatewayRoute = exports.AppmeshGatewayRouteSpecOutputReference = exports.appmeshGatewayRouteSpecToTerraform = exports.AppmeshGatewayRouteSpecHttpRouteOutputReference = exports.appmeshGatewayRouteSpecHttpRouteToTerraform = exports.AppmeshGatewayRouteSpecHttpRouteMatchOutputReference = exports.appmeshGatewayRouteSpecHttpRouteMatchToTerraform = exports.AppmeshGatewayRouteSpecHttpRouteMatchHostnameOutputReference = exports.appmeshGatewayRouteSpecHttpRouteMatchHostnameToTerraform = exports.AppmeshGatewayRouteSpecHttpRouteActionOutputReference = exports.appmeshGatewayRouteSpecHttpRouteActionToTerraform = exports.AppmeshGatewayRouteSpecHttpRouteActionTargetOutputReference = exports.appmeshGatewayRouteSpecHttpRouteActionTargetToTerraform = exports.AppmeshGatewayRouteSpecHttpRouteActionTargetVirtualServiceOutputReference = exports.appmeshGatewayRouteSpecHttpRouteActionTargetVirtualServiceToTerraform = exports.AppmeshGatewayRouteSpecHttpRouteActionRewriteOutputReference = exports.appmeshGatewayRouteSpecHttpRouteActionRewriteToTerraform = exports.AppmeshGatewayRouteSpecHttpRouteActionRewritePrefixOutputReference = exports.appmeshGatewayRouteSpecHttpRouteActionRewritePrefixToTerraform = exports.AppmeshGatewayRouteSpecHttpRouteActionRewriteHostnameOutputReference = exports.appmeshGatewayRouteSpecHttpRouteActionRewriteHostnameToTerraform = exports.AppmeshGatewayRouteSpecHttp2RouteOutputReference = exports.appmeshGatewayRouteSpecHttp2RouteToTerraform = exports.AppmeshGatewayRouteSpecHttp2RouteMatchOutputReference = exports.appmeshGatewayRouteSpecHttp2RouteMatchToTerraform = exports.AppmeshGatewayRouteSpecHttp2RouteMatchHostnameOutputReference = exports.appmeshGatewayRouteSpecHttp2RouteMatchHostnameToTerraform = exports.AppmeshGatewayRouteSpecHttp2RouteActionOutputReference = exports.appmeshGatewayRouteSpecHttp2RouteActionToTerraform = exports.AppmeshGatewayRouteSpecHttp2RouteActionTargetOutputReference = exports.appmeshGatewayRouteSpecHttp2RouteActionTargetToTerraform = exports.AppmeshGatewayRouteSpecHttp2RouteActionTargetVirtualServiceOutputReference = exports.appmeshGatewayRouteSpecHttp2RouteActionTargetVirtualServiceToTerraform = exports.AppmeshGatewayRouteSpecHttp2RouteActionRewriteOutputReference = exports.appmeshGatewayRouteSpecHttp2RouteActionRewriteToTerraform = exports.AppmeshGatewayRouteSpecHttp2RouteActionRewritePrefixOutputReference = exports.appmeshGatewayRouteSpecHttp2RouteActionRewritePrefixToTerraform = exports.AppmeshGatewayRouteSpecHttp2RouteActionRewriteHostnameOutputReference = exports.appmeshGatewayRouteSpecHttp2RouteActionRewriteHostnameToTerraform = exports.AppmeshGatewayRouteSpecGrpcRouteOutputReference = exports.appmeshGatewayRouteSpecGrpcRouteToTerraform = exports.AppmeshGatewayRouteSpecGrpcRouteMatchOutputReference = exports.appmeshGatewayRouteSpecGrpcRouteMatchToTerraform = exports.AppmeshGatewayRouteSpecGrpcRouteActionOutputReference = exports.appmeshGatewayRouteSpecGrpcRouteActionToTerraform = exports.AppmeshGatewayRouteSpecGrpcRouteActionTargetOutputReference = exports.appmeshGatewayRouteSpecGrpcRouteActionTargetToTerraform = exports.AppmeshGatewayRouteSpecGrpcRouteActionTargetVirtualServiceOutputReference = exports.appmeshGatewayRouteSpecGrpcRouteActionTargetVirtualServiceToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function appmeshGatewayRouteSpecGrpcRouteActionTargetVirtualServiceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        virtual_service_name: cdktf.stringToTerraform(struct.virtualServiceName),
    };
}
exports.appmeshGatewayRouteSpecGrpcRouteActionTargetVirtualServiceToTerraform = appmeshGatewayRouteSpecGrpcRouteActionTargetVirtualServiceToTerraform;
class AppmeshGatewayRouteSpecGrpcRouteActionTargetVirtualServiceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._virtualServiceName !== undefined) {
            hasAnyValues = true;
            internalValueResult.virtualServiceName = this._virtualServiceName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._virtualServiceName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._virtualServiceName = value.virtualServiceName;
        }
    }
    get virtualServiceName() {
        return this.getStringAttribute('virtual_service_name');
    }
    set virtualServiceName(value) {
        this._virtualServiceName = value;
    }
    // Temporarily expose input value. Use with caution.
    get virtualServiceNameInput() {
        return this._virtualServiceName;
    }
}
exports.AppmeshGatewayRouteSpecGrpcRouteActionTargetVirtualServiceOutputReference = AppmeshGatewayRouteSpecGrpcRouteActionTargetVirtualServiceOutputReference;
_a = JSII_RTTI_SYMBOL_1;
AppmeshGatewayRouteSpecGrpcRouteActionTargetVirtualServiceOutputReference[_a] = { fqn: "@cdktf/provider-aws.appmeshGatewayRoute.AppmeshGatewayRouteSpecGrpcRouteActionTargetVirtualServiceOutputReference", version: "10.0.12" };
function appmeshGatewayRouteSpecGrpcRouteActionTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        virtual_service: appmeshGatewayRouteSpecGrpcRouteActionTargetVirtualServiceToTerraform(struct.virtualService),
    };
}
exports.appmeshGatewayRouteSpecGrpcRouteActionTargetToTerraform = appmeshGatewayRouteSpecGrpcRouteActionTargetToTerraform;
class AppmeshGatewayRouteSpecGrpcRouteActionTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // virtual_service - computed: false, optional: false, required: true
        this._virtualService = new AppmeshGatewayRouteSpecGrpcRouteActionTargetVirtualServiceOutputReference(this, "virtual_service");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._virtualService?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.virtualService = this._virtualService?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._virtualService.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._virtualService.internalValue = value.virtualService;
        }
    }
    get virtualService() {
        return this._virtualService;
    }
    putVirtualService(value) {
        this._virtualService.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get virtualServiceInput() {
        return this._virtualService.internalValue;
    }
}
exports.AppmeshGatewayRouteSpecGrpcRouteActionTargetOutputReference = AppmeshGatewayRouteSpecGrpcRouteActionTargetOutputReference;
_b = JSII_RTTI_SYMBOL_1;
AppmeshGatewayRouteSpecGrpcRouteActionTargetOutputReference[_b] = { fqn: "@cdktf/provider-aws.appmeshGatewayRoute.AppmeshGatewayRouteSpecGrpcRouteActionTargetOutputReference", version: "10.0.12" };
function appmeshGatewayRouteSpecGrpcRouteActionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        target: appmeshGatewayRouteSpecGrpcRouteActionTargetToTerraform(struct.target),
    };
}
exports.appmeshGatewayRouteSpecGrpcRouteActionToTerraform = appmeshGatewayRouteSpecGrpcRouteActionToTerraform;
class AppmeshGatewayRouteSpecGrpcRouteActionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // target - computed: false, optional: false, required: true
        this._target = new AppmeshGatewayRouteSpecGrpcRouteActionTargetOutputReference(this, "target");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._target?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.target = this._target?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._target.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._target.internalValue = value.target;
        }
    }
    get target() {
        return this._target;
    }
    putTarget(value) {
        this._target.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get targetInput() {
        return this._target.internalValue;
    }
}
exports.AppmeshGatewayRouteSpecGrpcRouteActionOutputReference = AppmeshGatewayRouteSpecGrpcRouteActionOutputReference;
_c = JSII_RTTI_SYMBOL_1;
AppmeshGatewayRouteSpecGrpcRouteActionOutputReference[_c] = { fqn: "@cdktf/provider-aws.appmeshGatewayRoute.AppmeshGatewayRouteSpecGrpcRouteActionOutputReference", version: "10.0.12" };
function appmeshGatewayRouteSpecGrpcRouteMatchToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        service_name: cdktf.stringToTerraform(struct.serviceName),
    };
}
exports.appmeshGatewayRouteSpecGrpcRouteMatchToTerraform = appmeshGatewayRouteSpecGrpcRouteMatchToTerraform;
class AppmeshGatewayRouteSpecGrpcRouteMatchOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._serviceName !== undefined) {
            hasAnyValues = true;
            internalValueResult.serviceName = this._serviceName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._serviceName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._serviceName = value.serviceName;
        }
    }
    get serviceName() {
        return this.getStringAttribute('service_name');
    }
    set serviceName(value) {
        this._serviceName = value;
    }
    // Temporarily expose input value. Use with caution.
    get serviceNameInput() {
        return this._serviceName;
    }
}
exports.AppmeshGatewayRouteSpecGrpcRouteMatchOutputReference = AppmeshGatewayRouteSpecGrpcRouteMatchOutputReference;
_d = JSII_RTTI_SYMBOL_1;
AppmeshGatewayRouteSpecGrpcRouteMatchOutputReference[_d] = { fqn: "@cdktf/provider-aws.appmeshGatewayRoute.AppmeshGatewayRouteSpecGrpcRouteMatchOutputReference", version: "10.0.12" };
function appmeshGatewayRouteSpecGrpcRouteToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        action: appmeshGatewayRouteSpecGrpcRouteActionToTerraform(struct.action),
        match: appmeshGatewayRouteSpecGrpcRouteMatchToTerraform(struct.match),
    };
}
exports.appmeshGatewayRouteSpecGrpcRouteToTerraform = appmeshGatewayRouteSpecGrpcRouteToTerraform;
class AppmeshGatewayRouteSpecGrpcRouteOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // action - computed: false, optional: false, required: true
        this._action = new AppmeshGatewayRouteSpecGrpcRouteActionOutputReference(this, "action");
        // match - computed: false, optional: false, required: true
        this._match = new AppmeshGatewayRouteSpecGrpcRouteMatchOutputReference(this, "match");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._action?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.action = this._action?.internalValue;
        }
        if (this._match?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.match = this._match?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._action.internalValue = undefined;
            this._match.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._action.internalValue = value.action;
            this._match.internalValue = value.match;
        }
    }
    get action() {
        return this._action;
    }
    putAction(value) {
        this._action.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get actionInput() {
        return this._action.internalValue;
    }
    get match() {
        return this._match;
    }
    putMatch(value) {
        this._match.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get matchInput() {
        return this._match.internalValue;
    }
}
exports.AppmeshGatewayRouteSpecGrpcRouteOutputReference = AppmeshGatewayRouteSpecGrpcRouteOutputReference;
_e = JSII_RTTI_SYMBOL_1;
AppmeshGatewayRouteSpecGrpcRouteOutputReference[_e] = { fqn: "@cdktf/provider-aws.appmeshGatewayRoute.AppmeshGatewayRouteSpecGrpcRouteOutputReference", version: "10.0.12" };
function appmeshGatewayRouteSpecHttp2RouteActionRewriteHostnameToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        default_target_hostname: cdktf.stringToTerraform(struct.defaultTargetHostname),
    };
}
exports.appmeshGatewayRouteSpecHttp2RouteActionRewriteHostnameToTerraform = appmeshGatewayRouteSpecHttp2RouteActionRewriteHostnameToTerraform;
class AppmeshGatewayRouteSpecHttp2RouteActionRewriteHostnameOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._defaultTargetHostname !== undefined) {
            hasAnyValues = true;
            internalValueResult.defaultTargetHostname = this._defaultTargetHostname;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._defaultTargetHostname = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._defaultTargetHostname = value.defaultTargetHostname;
        }
    }
    get defaultTargetHostname() {
        return this.getStringAttribute('default_target_hostname');
    }
    set defaultTargetHostname(value) {
        this._defaultTargetHostname = value;
    }
    // Temporarily expose input value. Use with caution.
    get defaultTargetHostnameInput() {
        return this._defaultTargetHostname;
    }
}
exports.AppmeshGatewayRouteSpecHttp2RouteActionRewriteHostnameOutputReference = AppmeshGatewayRouteSpecHttp2RouteActionRewriteHostnameOutputReference;
_f = JSII_RTTI_SYMBOL_1;
AppmeshGatewayRouteSpecHttp2RouteActionRewriteHostnameOutputReference[_f] = { fqn: "@cdktf/provider-aws.appmeshGatewayRoute.AppmeshGatewayRouteSpecHttp2RouteActionRewriteHostnameOutputReference", version: "10.0.12" };
function appmeshGatewayRouteSpecHttp2RouteActionRewritePrefixToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        default_prefix: cdktf.stringToTerraform(struct.defaultPrefix),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.appmeshGatewayRouteSpecHttp2RouteActionRewritePrefixToTerraform = appmeshGatewayRouteSpecHttp2RouteActionRewritePrefixToTerraform;
class AppmeshGatewayRouteSpecHttp2RouteActionRewritePrefixOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._defaultPrefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.defaultPrefix = this._defaultPrefix;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._defaultPrefix = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._defaultPrefix = value.defaultPrefix;
            this._value = value.value;
        }
    }
    get defaultPrefix() {
        return this.getStringAttribute('default_prefix');
    }
    set defaultPrefix(value) {
        this._defaultPrefix = value;
    }
    resetDefaultPrefix() {
        this._defaultPrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultPrefixInput() {
        return this._defaultPrefix;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.AppmeshGatewayRouteSpecHttp2RouteActionRewritePrefixOutputReference = AppmeshGatewayRouteSpecHttp2RouteActionRewritePrefixOutputReference;
_g = JSII_RTTI_SYMBOL_1;
AppmeshGatewayRouteSpecHttp2RouteActionRewritePrefixOutputReference[_g] = { fqn: "@cdktf/provider-aws.appmeshGatewayRoute.AppmeshGatewayRouteSpecHttp2RouteActionRewritePrefixOutputReference", version: "10.0.12" };
function appmeshGatewayRouteSpecHttp2RouteActionRewriteToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        hostname: appmeshGatewayRouteSpecHttp2RouteActionRewriteHostnameToTerraform(struct.hostname),
        prefix: appmeshGatewayRouteSpecHttp2RouteActionRewritePrefixToTerraform(struct.prefix),
    };
}
exports.appmeshGatewayRouteSpecHttp2RouteActionRewriteToTerraform = appmeshGatewayRouteSpecHttp2RouteActionRewriteToTerraform;
class AppmeshGatewayRouteSpecHttp2RouteActionRewriteOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // hostname - computed: false, optional: true, required: false
        this._hostname = new AppmeshGatewayRouteSpecHttp2RouteActionRewriteHostnameOutputReference(this, "hostname");
        // prefix - computed: false, optional: true, required: false
        this._prefix = new AppmeshGatewayRouteSpecHttp2RouteActionRewritePrefixOutputReference(this, "prefix");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._hostname?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.hostname = this._hostname?.internalValue;
        }
        if (this._prefix?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._hostname.internalValue = undefined;
            this._prefix.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._hostname.internalValue = value.hostname;
            this._prefix.internalValue = value.prefix;
        }
    }
    get hostname() {
        return this._hostname;
    }
    putHostname(value) {
        this._hostname.internalValue = value;
    }
    resetHostname() {
        this._hostname.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hostnameInput() {
        return this._hostname.internalValue;
    }
    get prefix() {
        return this._prefix;
    }
    putPrefix(value) {
        this._prefix.internalValue = value;
    }
    resetPrefix() {
        this._prefix.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix.internalValue;
    }
}
exports.AppmeshGatewayRouteSpecHttp2RouteActionRewriteOutputReference = AppmeshGatewayRouteSpecHttp2RouteActionRewriteOutputReference;
_h = JSII_RTTI_SYMBOL_1;
AppmeshGatewayRouteSpecHttp2RouteActionRewriteOutputReference[_h] = { fqn: "@cdktf/provider-aws.appmeshGatewayRoute.AppmeshGatewayRouteSpecHttp2RouteActionRewriteOutputReference", version: "10.0.12" };
function appmeshGatewayRouteSpecHttp2RouteActionTargetVirtualServiceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        virtual_service_name: cdktf.stringToTerraform(struct.virtualServiceName),
    };
}
exports.appmeshGatewayRouteSpecHttp2RouteActionTargetVirtualServiceToTerraform = appmeshGatewayRouteSpecHttp2RouteActionTargetVirtualServiceToTerraform;
class AppmeshGatewayRouteSpecHttp2RouteActionTargetVirtualServiceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._virtualServiceName !== undefined) {
            hasAnyValues = true;
            internalValueResult.virtualServiceName = this._virtualServiceName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._virtualServiceName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._virtualServiceName = value.virtualServiceName;
        }
    }
    get virtualServiceName() {
        return this.getStringAttribute('virtual_service_name');
    }
    set virtualServiceName(value) {
        this._virtualServiceName = value;
    }
    // Temporarily expose input value. Use with caution.
    get virtualServiceNameInput() {
        return this._virtualServiceName;
    }
}
exports.AppmeshGatewayRouteSpecHttp2RouteActionTargetVirtualServiceOutputReference = AppmeshGatewayRouteSpecHttp2RouteActionTargetVirtualServiceOutputReference;
_j = JSII_RTTI_SYMBOL_1;
AppmeshGatewayRouteSpecHttp2RouteActionTargetVirtualServiceOutputReference[_j] = { fqn: "@cdktf/provider-aws.appmeshGatewayRoute.AppmeshGatewayRouteSpecHttp2RouteActionTargetVirtualServiceOutputReference", version: "10.0.12" };
function appmeshGatewayRouteSpecHttp2RouteActionTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        virtual_service: appmeshGatewayRouteSpecHttp2RouteActionTargetVirtualServiceToTerraform(struct.virtualService),
    };
}
exports.appmeshGatewayRouteSpecHttp2RouteActionTargetToTerraform = appmeshGatewayRouteSpecHttp2RouteActionTargetToTerraform;
class AppmeshGatewayRouteSpecHttp2RouteActionTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // virtual_service - computed: false, optional: false, required: true
        this._virtualService = new AppmeshGatewayRouteSpecHttp2RouteActionTargetVirtualServiceOutputReference(this, "virtual_service");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._virtualService?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.virtualService = this._virtualService?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._virtualService.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._virtualService.internalValue = value.virtualService;
        }
    }
    get virtualService() {
        return this._virtualService;
    }
    putVirtualService(value) {
        this._virtualService.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get virtualServiceInput() {
        return this._virtualService.internalValue;
    }
}
exports.AppmeshGatewayRouteSpecHttp2RouteActionTargetOutputReference = AppmeshGatewayRouteSpecHttp2RouteActionTargetOutputReference;
_k = JSII_RTTI_SYMBOL_1;
AppmeshGatewayRouteSpecHttp2RouteActionTargetOutputReference[_k] = { fqn: "@cdktf/provider-aws.appmeshGatewayRoute.AppmeshGatewayRouteSpecHttp2RouteActionTargetOutputReference", version: "10.0.12" };
function appmeshGatewayRouteSpecHttp2RouteActionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        rewrite: appmeshGatewayRouteSpecHttp2RouteActionRewriteToTerraform(struct.rewrite),
        target: appmeshGatewayRouteSpecHttp2RouteActionTargetToTerraform(struct.target),
    };
}
exports.appmeshGatewayRouteSpecHttp2RouteActionToTerraform = appmeshGatewayRouteSpecHttp2RouteActionToTerraform;
class AppmeshGatewayRouteSpecHttp2RouteActionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // rewrite - computed: false, optional: true, required: false
        this._rewrite = new AppmeshGatewayRouteSpecHttp2RouteActionRewriteOutputReference(this, "rewrite");
        // target - computed: false, optional: false, required: true
        this._target = new AppmeshGatewayRouteSpecHttp2RouteActionTargetOutputReference(this, "target");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._rewrite?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.rewrite = this._rewrite?.internalValue;
        }
        if (this._target?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.target = this._target?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._rewrite.internalValue = undefined;
            this._target.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._rewrite.internalValue = value.rewrite;
            this._target.internalValue = value.target;
        }
    }
    get rewrite() {
        return this._rewrite;
    }
    putRewrite(value) {
        this._rewrite.internalValue = value;
    }
    resetRewrite() {
        this._rewrite.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rewriteInput() {
        return this._rewrite.internalValue;
    }
    get target() {
        return this._target;
    }
    putTarget(value) {
        this._target.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get targetInput() {
        return this._target.internalValue;
    }
}
exports.AppmeshGatewayRouteSpecHttp2RouteActionOutputReference = AppmeshGatewayRouteSpecHttp2RouteActionOutputReference;
_l = JSII_RTTI_SYMBOL_1;
AppmeshGatewayRouteSpecHttp2RouteActionOutputReference[_l] = { fqn: "@cdktf/provider-aws.appmeshGatewayRoute.AppmeshGatewayRouteSpecHttp2RouteActionOutputReference", version: "10.0.12" };
function appmeshGatewayRouteSpecHttp2RouteMatchHostnameToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        exact: cdktf.stringToTerraform(struct.exact),
        suffix: cdktf.stringToTerraform(struct.suffix),
    };
}
exports.appmeshGatewayRouteSpecHttp2RouteMatchHostnameToTerraform = appmeshGatewayRouteSpecHttp2RouteMatchHostnameToTerraform;
class AppmeshGatewayRouteSpecHttp2RouteMatchHostnameOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._exact !== undefined) {
            hasAnyValues = true;
            internalValueResult.exact = this._exact;
        }
        if (this._suffix !== undefined) {
            hasAnyValues = true;
            internalValueResult.suffix = this._suffix;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._exact = undefined;
            this._suffix = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._exact = value.exact;
            this._suffix = value.suffix;
        }
    }
    get exact() {
        return this.getStringAttribute('exact');
    }
    set exact(value) {
        this._exact = value;
    }
    resetExact() {
        this._exact = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exactInput() {
        return this._exact;
    }
    get suffix() {
        return this.getStringAttribute('suffix');
    }
    set suffix(value) {
        this._suffix = value;
    }
    resetSuffix() {
        this._suffix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get suffixInput() {
        return this._suffix;
    }
}
exports.AppmeshGatewayRouteSpecHttp2RouteMatchHostnameOutputReference = AppmeshGatewayRouteSpecHttp2RouteMatchHostnameOutputReference;
_m = JSII_RTTI_SYMBOL_1;
AppmeshGatewayRouteSpecHttp2RouteMatchHostnameOutputReference[_m] = { fqn: "@cdktf/provider-aws.appmeshGatewayRoute.AppmeshGatewayRouteSpecHttp2RouteMatchHostnameOutputReference", version: "10.0.12" };
function appmeshGatewayRouteSpecHttp2RouteMatchToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        prefix: cdktf.stringToTerraform(struct.prefix),
        hostname: appmeshGatewayRouteSpecHttp2RouteMatchHostnameToTerraform(struct.hostname),
    };
}
exports.appmeshGatewayRouteSpecHttp2RouteMatchToTerraform = appmeshGatewayRouteSpecHttp2RouteMatchToTerraform;
class AppmeshGatewayRouteSpecHttp2RouteMatchOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // hostname - computed: false, optional: true, required: false
        this._hostname = new AppmeshGatewayRouteSpecHttp2RouteMatchHostnameOutputReference(this, "hostname");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._prefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        if (this._hostname?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.hostname = this._hostname?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._prefix = undefined;
            this._hostname.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._prefix = value.prefix;
            this._hostname.internalValue = value.hostname;
        }
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
    get hostname() {
        return this._hostname;
    }
    putHostname(value) {
        this._hostname.internalValue = value;
    }
    resetHostname() {
        this._hostname.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hostnameInput() {
        return this._hostname.internalValue;
    }
}
exports.AppmeshGatewayRouteSpecHttp2RouteMatchOutputReference = AppmeshGatewayRouteSpecHttp2RouteMatchOutputReference;
_o = JSII_RTTI_SYMBOL_1;
AppmeshGatewayRouteSpecHttp2RouteMatchOutputReference[_o] = { fqn: "@cdktf/provider-aws.appmeshGatewayRoute.AppmeshGatewayRouteSpecHttp2RouteMatchOutputReference", version: "10.0.12" };
function appmeshGatewayRouteSpecHttp2RouteToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        action: appmeshGatewayRouteSpecHttp2RouteActionToTerraform(struct.action),
        match: appmeshGatewayRouteSpecHttp2RouteMatchToTerraform(struct.match),
    };
}
exports.appmeshGatewayRouteSpecHttp2RouteToTerraform = appmeshGatewayRouteSpecHttp2RouteToTerraform;
class AppmeshGatewayRouteSpecHttp2RouteOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // action - computed: false, optional: false, required: true
        this._action = new AppmeshGatewayRouteSpecHttp2RouteActionOutputReference(this, "action");
        // match - computed: false, optional: false, required: true
        this._match = new AppmeshGatewayRouteSpecHttp2RouteMatchOutputReference(this, "match");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._action?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.action = this._action?.internalValue;
        }
        if (this._match?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.match = this._match?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._action.internalValue = undefined;
            this._match.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._action.internalValue = value.action;
            this._match.internalValue = value.match;
        }
    }
    get action() {
        return this._action;
    }
    putAction(value) {
        this._action.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get actionInput() {
        return this._action.internalValue;
    }
    get match() {
        return this._match;
    }
    putMatch(value) {
        this._match.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get matchInput() {
        return this._match.internalValue;
    }
}
exports.AppmeshGatewayRouteSpecHttp2RouteOutputReference = AppmeshGatewayRouteSpecHttp2RouteOutputReference;
_p = JSII_RTTI_SYMBOL_1;
AppmeshGatewayRouteSpecHttp2RouteOutputReference[_p] = { fqn: "@cdktf/provider-aws.appmeshGatewayRoute.AppmeshGatewayRouteSpecHttp2RouteOutputReference", version: "10.0.12" };
function appmeshGatewayRouteSpecHttpRouteActionRewriteHostnameToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        default_target_hostname: cdktf.stringToTerraform(struct.defaultTargetHostname),
    };
}
exports.appmeshGatewayRouteSpecHttpRouteActionRewriteHostnameToTerraform = appmeshGatewayRouteSpecHttpRouteActionRewriteHostnameToTerraform;
class AppmeshGatewayRouteSpecHttpRouteActionRewriteHostnameOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._defaultTargetHostname !== undefined) {
            hasAnyValues = true;
            internalValueResult.defaultTargetHostname = this._defaultTargetHostname;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._defaultTargetHostname = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._defaultTargetHostname = value.defaultTargetHostname;
        }
    }
    get defaultTargetHostname() {
        return this.getStringAttribute('default_target_hostname');
    }
    set defaultTargetHostname(value) {
        this._defaultTargetHostname = value;
    }
    // Temporarily expose input value. Use with caution.
    get defaultTargetHostnameInput() {
        return this._defaultTargetHostname;
    }
}
exports.AppmeshGatewayRouteSpecHttpRouteActionRewriteHostnameOutputReference = AppmeshGatewayRouteSpecHttpRouteActionRewriteHostnameOutputReference;
_q = JSII_RTTI_SYMBOL_1;
AppmeshGatewayRouteSpecHttpRouteActionRewriteHostnameOutputReference[_q] = { fqn: "@cdktf/provider-aws.appmeshGatewayRoute.AppmeshGatewayRouteSpecHttpRouteActionRewriteHostnameOutputReference", version: "10.0.12" };
function appmeshGatewayRouteSpecHttpRouteActionRewritePrefixToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        default_prefix: cdktf.stringToTerraform(struct.defaultPrefix),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.appmeshGatewayRouteSpecHttpRouteActionRewritePrefixToTerraform = appmeshGatewayRouteSpecHttpRouteActionRewritePrefixToTerraform;
class AppmeshGatewayRouteSpecHttpRouteActionRewritePrefixOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._defaultPrefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.defaultPrefix = this._defaultPrefix;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._defaultPrefix = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._defaultPrefix = value.defaultPrefix;
            this._value = value.value;
        }
    }
    get defaultPrefix() {
        return this.getStringAttribute('default_prefix');
    }
    set defaultPrefix(value) {
        this._defaultPrefix = value;
    }
    resetDefaultPrefix() {
        this._defaultPrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultPrefixInput() {
        return this._defaultPrefix;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.AppmeshGatewayRouteSpecHttpRouteActionRewritePrefixOutputReference = AppmeshGatewayRouteSpecHttpRouteActionRewritePrefixOutputReference;
_r = JSII_RTTI_SYMBOL_1;
AppmeshGatewayRouteSpecHttpRouteActionRewritePrefixOutputReference[_r] = { fqn: "@cdktf/provider-aws.appmeshGatewayRoute.AppmeshGatewayRouteSpecHttpRouteActionRewritePrefixOutputReference", version: "10.0.12" };
function appmeshGatewayRouteSpecHttpRouteActionRewriteToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        hostname: appmeshGatewayRouteSpecHttpRouteActionRewriteHostnameToTerraform(struct.hostname),
        prefix: appmeshGatewayRouteSpecHttpRouteActionRewritePrefixToTerraform(struct.prefix),
    };
}
exports.appmeshGatewayRouteSpecHttpRouteActionRewriteToTerraform = appmeshGatewayRouteSpecHttpRouteActionRewriteToTerraform;
class AppmeshGatewayRouteSpecHttpRouteActionRewriteOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // hostname - computed: false, optional: true, required: false
        this._hostname = new AppmeshGatewayRouteSpecHttpRouteActionRewriteHostnameOutputReference(this, "hostname");
        // prefix - computed: false, optional: true, required: false
        this._prefix = new AppmeshGatewayRouteSpecHttpRouteActionRewritePrefixOutputReference(this, "prefix");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._hostname?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.hostname = this._hostname?.internalValue;
        }
        if (this._prefix?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._hostname.internalValue = undefined;
            this._prefix.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._hostname.internalValue = value.hostname;
            this._prefix.internalValue = value.prefix;
        }
    }
    get hostname() {
        return this._hostname;
    }
    putHostname(value) {
        this._hostname.internalValue = value;
    }
    resetHostname() {
        this._hostname.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hostnameInput() {
        return this._hostname.internalValue;
    }
    get prefix() {
        return this._prefix;
    }
    putPrefix(value) {
        this._prefix.internalValue = value;
    }
    resetPrefix() {
        this._prefix.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix.internalValue;
    }
}
exports.AppmeshGatewayRouteSpecHttpRouteActionRewriteOutputReference = AppmeshGatewayRouteSpecHttpRouteActionRewriteOutputReference;
_s = JSII_RTTI_SYMBOL_1;
AppmeshGatewayRouteSpecHttpRouteActionRewriteOutputReference[_s] = { fqn: "@cdktf/provider-aws.appmeshGatewayRoute.AppmeshGatewayRouteSpecHttpRouteActionRewriteOutputReference", version: "10.0.12" };
function appmeshGatewayRouteSpecHttpRouteActionTargetVirtualServiceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        virtual_service_name: cdktf.stringToTerraform(struct.virtualServiceName),
    };
}
exports.appmeshGatewayRouteSpecHttpRouteActionTargetVirtualServiceToTerraform = appmeshGatewayRouteSpecHttpRouteActionTargetVirtualServiceToTerraform;
class AppmeshGatewayRouteSpecHttpRouteActionTargetVirtualServiceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._virtualServiceName !== undefined) {
            hasAnyValues = true;
            internalValueResult.virtualServiceName = this._virtualServiceName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._virtualServiceName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._virtualServiceName = value.virtualServiceName;
        }
    }
    get virtualServiceName() {
        return this.getStringAttribute('virtual_service_name');
    }
    set virtualServiceName(value) {
        this._virtualServiceName = value;
    }
    // Temporarily expose input value. Use with caution.
    get virtualServiceNameInput() {
        return this._virtualServiceName;
    }
}
exports.AppmeshGatewayRouteSpecHttpRouteActionTargetVirtualServiceOutputReference = AppmeshGatewayRouteSpecHttpRouteActionTargetVirtualServiceOutputReference;
_t = JSII_RTTI_SYMBOL_1;
AppmeshGatewayRouteSpecHttpRouteActionTargetVirtualServiceOutputReference[_t] = { fqn: "@cdktf/provider-aws.appmeshGatewayRoute.AppmeshGatewayRouteSpecHttpRouteActionTargetVirtualServiceOutputReference", version: "10.0.12" };
function appmeshGatewayRouteSpecHttpRouteActionTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        virtual_service: appmeshGatewayRouteSpecHttpRouteActionTargetVirtualServiceToTerraform(struct.virtualService),
    };
}
exports.appmeshGatewayRouteSpecHttpRouteActionTargetToTerraform = appmeshGatewayRouteSpecHttpRouteActionTargetToTerraform;
class AppmeshGatewayRouteSpecHttpRouteActionTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // virtual_service - computed: false, optional: false, required: true
        this._virtualService = new AppmeshGatewayRouteSpecHttpRouteActionTargetVirtualServiceOutputReference(this, "virtual_service");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._virtualService?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.virtualService = this._virtualService?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._virtualService.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._virtualService.internalValue = value.virtualService;
        }
    }
    get virtualService() {
        return this._virtualService;
    }
    putVirtualService(value) {
        this._virtualService.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get virtualServiceInput() {
        return this._virtualService.internalValue;
    }
}
exports.AppmeshGatewayRouteSpecHttpRouteActionTargetOutputReference = AppmeshGatewayRouteSpecHttpRouteActionTargetOutputReference;
_u = JSII_RTTI_SYMBOL_1;
AppmeshGatewayRouteSpecHttpRouteActionTargetOutputReference[_u] = { fqn: "@cdktf/provider-aws.appmeshGatewayRoute.AppmeshGatewayRouteSpecHttpRouteActionTargetOutputReference", version: "10.0.12" };
function appmeshGatewayRouteSpecHttpRouteActionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        rewrite: appmeshGatewayRouteSpecHttpRouteActionRewriteToTerraform(struct.rewrite),
        target: appmeshGatewayRouteSpecHttpRouteActionTargetToTerraform(struct.target),
    };
}
exports.appmeshGatewayRouteSpecHttpRouteActionToTerraform = appmeshGatewayRouteSpecHttpRouteActionToTerraform;
class AppmeshGatewayRouteSpecHttpRouteActionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // rewrite - computed: false, optional: true, required: false
        this._rewrite = new AppmeshGatewayRouteSpecHttpRouteActionRewriteOutputReference(this, "rewrite");
        // target - computed: false, optional: false, required: true
        this._target = new AppmeshGatewayRouteSpecHttpRouteActionTargetOutputReference(this, "target");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._rewrite?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.rewrite = this._rewrite?.internalValue;
        }
        if (this._target?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.target = this._target?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._rewrite.internalValue = undefined;
            this._target.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._rewrite.internalValue = value.rewrite;
            this._target.internalValue = value.target;
        }
    }
    get rewrite() {
        return this._rewrite;
    }
    putRewrite(value) {
        this._rewrite.internalValue = value;
    }
    resetRewrite() {
        this._rewrite.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rewriteInput() {
        return this._rewrite.internalValue;
    }
    get target() {
        return this._target;
    }
    putTarget(value) {
        this._target.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get targetInput() {
        return this._target.internalValue;
    }
}
exports.AppmeshGatewayRouteSpecHttpRouteActionOutputReference = AppmeshGatewayRouteSpecHttpRouteActionOutputReference;
_v = JSII_RTTI_SYMBOL_1;
AppmeshGatewayRouteSpecHttpRouteActionOutputReference[_v] = { fqn: "@cdktf/provider-aws.appmeshGatewayRoute.AppmeshGatewayRouteSpecHttpRouteActionOutputReference", version: "10.0.12" };
function appmeshGatewayRouteSpecHttpRouteMatchHostnameToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        exact: cdktf.stringToTerraform(struct.exact),
        suffix: cdktf.stringToTerraform(struct.suffix),
    };
}
exports.appmeshGatewayRouteSpecHttpRouteMatchHostnameToTerraform = appmeshGatewayRouteSpecHttpRouteMatchHostnameToTerraform;
class AppmeshGatewayRouteSpecHttpRouteMatchHostnameOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._exact !== undefined) {
            hasAnyValues = true;
            internalValueResult.exact = this._exact;
        }
        if (this._suffix !== undefined) {
            hasAnyValues = true;
            internalValueResult.suffix = this._suffix;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._exact = undefined;
            this._suffix = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._exact = value.exact;
            this._suffix = value.suffix;
        }
    }
    get exact() {
        return this.getStringAttribute('exact');
    }
    set exact(value) {
        this._exact = value;
    }
    resetExact() {
        this._exact = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exactInput() {
        return this._exact;
    }
    get suffix() {
        return this.getStringAttribute('suffix');
    }
    set suffix(value) {
        this._suffix = value;
    }
    resetSuffix() {
        this._suffix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get suffixInput() {
        return this._suffix;
    }
}
exports.AppmeshGatewayRouteSpecHttpRouteMatchHostnameOutputReference = AppmeshGatewayRouteSpecHttpRouteMatchHostnameOutputReference;
_w = JSII_RTTI_SYMBOL_1;
AppmeshGatewayRouteSpecHttpRouteMatchHostnameOutputReference[_w] = { fqn: "@cdktf/provider-aws.appmeshGatewayRoute.AppmeshGatewayRouteSpecHttpRouteMatchHostnameOutputReference", version: "10.0.12" };
function appmeshGatewayRouteSpecHttpRouteMatchToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        prefix: cdktf.stringToTerraform(struct.prefix),
        hostname: appmeshGatewayRouteSpecHttpRouteMatchHostnameToTerraform(struct.hostname),
    };
}
exports.appmeshGatewayRouteSpecHttpRouteMatchToTerraform = appmeshGatewayRouteSpecHttpRouteMatchToTerraform;
class AppmeshGatewayRouteSpecHttpRouteMatchOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // hostname - computed: false, optional: true, required: false
        this._hostname = new AppmeshGatewayRouteSpecHttpRouteMatchHostnameOutputReference(this, "hostname");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._prefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        if (this._hostname?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.hostname = this._hostname?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._prefix = undefined;
            this._hostname.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._prefix = value.prefix;
            this._hostname.internalValue = value.hostname;
        }
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
    get hostname() {
        return this._hostname;
    }
    putHostname(value) {
        this._hostname.internalValue = value;
    }
    resetHostname() {
        this._hostname.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hostnameInput() {
        return this._hostname.internalValue;
    }
}
exports.AppmeshGatewayRouteSpecHttpRouteMatchOutputReference = AppmeshGatewayRouteSpecHttpRouteMatchOutputReference;
_x = JSII_RTTI_SYMBOL_1;
AppmeshGatewayRouteSpecHttpRouteMatchOutputReference[_x] = { fqn: "@cdktf/provider-aws.appmeshGatewayRoute.AppmeshGatewayRouteSpecHttpRouteMatchOutputReference", version: "10.0.12" };
function appmeshGatewayRouteSpecHttpRouteToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        action: appmeshGatewayRouteSpecHttpRouteActionToTerraform(struct.action),
        match: appmeshGatewayRouteSpecHttpRouteMatchToTerraform(struct.match),
    };
}
exports.appmeshGatewayRouteSpecHttpRouteToTerraform = appmeshGatewayRouteSpecHttpRouteToTerraform;
class AppmeshGatewayRouteSpecHttpRouteOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // action - computed: false, optional: false, required: true
        this._action = new AppmeshGatewayRouteSpecHttpRouteActionOutputReference(this, "action");
        // match - computed: false, optional: false, required: true
        this._match = new AppmeshGatewayRouteSpecHttpRouteMatchOutputReference(this, "match");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._action?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.action = this._action?.internalValue;
        }
        if (this._match?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.match = this._match?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._action.internalValue = undefined;
            this._match.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._action.internalValue = value.action;
            this._match.internalValue = value.match;
        }
    }
    get action() {
        return this._action;
    }
    putAction(value) {
        this._action.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get actionInput() {
        return this._action.internalValue;
    }
    get match() {
        return this._match;
    }
    putMatch(value) {
        this._match.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get matchInput() {
        return this._match.internalValue;
    }
}
exports.AppmeshGatewayRouteSpecHttpRouteOutputReference = AppmeshGatewayRouteSpecHttpRouteOutputReference;
_y = JSII_RTTI_SYMBOL_1;
AppmeshGatewayRouteSpecHttpRouteOutputReference[_y] = { fqn: "@cdktf/provider-aws.appmeshGatewayRoute.AppmeshGatewayRouteSpecHttpRouteOutputReference", version: "10.0.12" };
function appmeshGatewayRouteSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        grpc_route: appmeshGatewayRouteSpecGrpcRouteToTerraform(struct.grpcRoute),
        http2_route: appmeshGatewayRouteSpecHttp2RouteToTerraform(struct.http2Route),
        http_route: appmeshGatewayRouteSpecHttpRouteToTerraform(struct.httpRoute),
    };
}
exports.appmeshGatewayRouteSpecToTerraform = appmeshGatewayRouteSpecToTerraform;
class AppmeshGatewayRouteSpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // grpc_route - computed: false, optional: true, required: false
        this._grpcRoute = new AppmeshGatewayRouteSpecGrpcRouteOutputReference(this, "grpc_route");
        // http2_route - computed: false, optional: true, required: false
        this._http2Route = new AppmeshGatewayRouteSpecHttp2RouteOutputReference(this, "http2_route");
        // http_route - computed: false, optional: true, required: false
        this._httpRoute = new AppmeshGatewayRouteSpecHttpRouteOutputReference(this, "http_route");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._grpcRoute?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.grpcRoute = this._grpcRoute?.internalValue;
        }
        if (this._http2Route?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.http2Route = this._http2Route?.internalValue;
        }
        if (this._httpRoute?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.httpRoute = this._httpRoute?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._grpcRoute.internalValue = undefined;
            this._http2Route.internalValue = undefined;
            this._httpRoute.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._grpcRoute.internalValue = value.grpcRoute;
            this._http2Route.internalValue = value.http2Route;
            this._httpRoute.internalValue = value.httpRoute;
        }
    }
    get grpcRoute() {
        return this._grpcRoute;
    }
    putGrpcRoute(value) {
        this._grpcRoute.internalValue = value;
    }
    resetGrpcRoute() {
        this._grpcRoute.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get grpcRouteInput() {
        return this._grpcRoute.internalValue;
    }
    get http2Route() {
        return this._http2Route;
    }
    putHttp2Route(value) {
        this._http2Route.internalValue = value;
    }
    resetHttp2Route() {
        this._http2Route.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get http2RouteInput() {
        return this._http2Route.internalValue;
    }
    get httpRoute() {
        return this._httpRoute;
    }
    putHttpRoute(value) {
        this._httpRoute.internalValue = value;
    }
    resetHttpRoute() {
        this._httpRoute.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get httpRouteInput() {
        return this._httpRoute.internalValue;
    }
}
exports.AppmeshGatewayRouteSpecOutputReference = AppmeshGatewayRouteSpecOutputReference;
_z = JSII_RTTI_SYMBOL_1;
AppmeshGatewayRouteSpecOutputReference[_z] = { fqn: "@cdktf/provider-aws.appmeshGatewayRoute.AppmeshGatewayRouteSpecOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/appmesh_gateway_route aws_appmesh_gateway_route}
*/
class AppmeshGatewayRoute extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/appmesh_gateway_route aws_appmesh_gateway_route} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options AppmeshGatewayRouteConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_appmesh_gateway_route',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // spec - computed: false, optional: false, required: true
        this._spec = new AppmeshGatewayRouteSpecOutputReference(this, "spec");
        this._id = config.id;
        this._meshName = config.meshName;
        this._meshOwner = config.meshOwner;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._virtualGatewayName = config.virtualGatewayName;
        this._spec.internalValue = config.spec;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // created_date - computed: true, optional: false, required: false
    get createdDate() {
        return this.getStringAttribute('created_date');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // last_updated_date - computed: true, optional: false, required: false
    get lastUpdatedDate() {
        return this.getStringAttribute('last_updated_date');
    }
    get meshName() {
        return this.getStringAttribute('mesh_name');
    }
    set meshName(value) {
        this._meshName = value;
    }
    // Temporarily expose input value. Use with caution.
    get meshNameInput() {
        return this._meshName;
    }
    get meshOwner() {
        return this.getStringAttribute('mesh_owner');
    }
    set meshOwner(value) {
        this._meshOwner = value;
    }
    resetMeshOwner() {
        this._meshOwner = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get meshOwnerInput() {
        return this._meshOwner;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // resource_owner - computed: true, optional: false, required: false
    get resourceOwner() {
        return this.getStringAttribute('resource_owner');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get virtualGatewayName() {
        return this.getStringAttribute('virtual_gateway_name');
    }
    set virtualGatewayName(value) {
        this._virtualGatewayName = value;
    }
    // Temporarily expose input value. Use with caution.
    get virtualGatewayNameInput() {
        return this._virtualGatewayName;
    }
    get spec() {
        return this._spec;
    }
    putSpec(value) {
        this._spec.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get specInput() {
        return this._spec.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            mesh_name: cdktf.stringToTerraform(this._meshName),
            mesh_owner: cdktf.stringToTerraform(this._meshOwner),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            virtual_gateway_name: cdktf.stringToTerraform(this._virtualGatewayName),
            spec: appmeshGatewayRouteSpecToTerraform(this._spec.internalValue),
        };
    }
}
exports.AppmeshGatewayRoute = AppmeshGatewayRoute;
_0 = JSII_RTTI_SYMBOL_1;
AppmeshGatewayRoute[_0] = { fqn: "@cdktf/provider-aws.appmeshGatewayRoute.AppmeshGatewayRoute", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
AppmeshGatewayRoute.tfResourceType = "aws_appmesh_gateway_route";
