// @cdktf/provider-aws/lib/data-aws-identitystore-group/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsIdentitystoreGroup = exports.DataAwsIdentitystoreGroupFilterList = exports.DataAwsIdentitystoreGroupFilterOutputReference = exports.dataAwsIdentitystoreGroupFilterToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsIdentitystoreGroupFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        attribute_path: cdktf.stringToTerraform(struct.attributePath),
        attribute_value: cdktf.stringToTerraform(struct.attributeValue),
    };
}
exports.dataAwsIdentitystoreGroupFilterToTerraform = dataAwsIdentitystoreGroupFilterToTerraform;
class DataAwsIdentitystoreGroupFilterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._attributePath !== undefined) {
            hasAnyValues = true;
            internalValueResult.attributePath = this._attributePath;
        }
        if (this._attributeValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.attributeValue = this._attributeValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._attributePath = undefined;
            this._attributeValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._attributePath = value.attributePath;
            this._attributeValue = value.attributeValue;
        }
    }
    get attributePath() {
        return this.getStringAttribute('attribute_path');
    }
    set attributePath(value) {
        this._attributePath = value;
    }
    // Temporarily expose input value. Use with caution.
    get attributePathInput() {
        return this._attributePath;
    }
    get attributeValue() {
        return this.getStringAttribute('attribute_value');
    }
    set attributeValue(value) {
        this._attributeValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get attributeValueInput() {
        return this._attributeValue;
    }
}
exports.DataAwsIdentitystoreGroupFilterOutputReference = DataAwsIdentitystoreGroupFilterOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsIdentitystoreGroupFilterOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsIdentitystoreGroup.DataAwsIdentitystoreGroupFilterOutputReference", version: "10.0.12" };
class DataAwsIdentitystoreGroupFilterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsIdentitystoreGroupFilterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsIdentitystoreGroupFilterList = DataAwsIdentitystoreGroupFilterList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsIdentitystoreGroupFilterList[_b] = { fqn: "@cdktf/provider-aws.dataAwsIdentitystoreGroup.DataAwsIdentitystoreGroupFilterList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/identitystore_group aws_identitystore_group}
*/
class DataAwsIdentitystoreGroup extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/identitystore_group aws_identitystore_group} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsIdentitystoreGroupConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_identitystore_group',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // filter - computed: false, optional: false, required: true
        this._filter = new DataAwsIdentitystoreGroupFilterList(this, "filter", true);
        this._groupId = config.groupId;
        this._id = config.id;
        this._identityStoreId = config.identityStoreId;
        this._filter.internalValue = config.filter;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // display_name - computed: true, optional: false, required: false
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    get groupId() {
        return this.getStringAttribute('group_id');
    }
    set groupId(value) {
        this._groupId = value;
    }
    resetGroupId() {
        this._groupId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get groupIdInput() {
        return this._groupId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get identityStoreId() {
        return this.getStringAttribute('identity_store_id');
    }
    set identityStoreId(value) {
        this._identityStoreId = value;
    }
    // Temporarily expose input value. Use with caution.
    get identityStoreIdInput() {
        return this._identityStoreId;
    }
    get filter() {
        return this._filter;
    }
    putFilter(value) {
        this._filter.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            group_id: cdktf.stringToTerraform(this._groupId),
            id: cdktf.stringToTerraform(this._id),
            identity_store_id: cdktf.stringToTerraform(this._identityStoreId),
            filter: cdktf.listMapper(dataAwsIdentitystoreGroupFilterToTerraform, true)(this._filter.internalValue),
        };
    }
}
exports.DataAwsIdentitystoreGroup = DataAwsIdentitystoreGroup;
_c = JSII_RTTI_SYMBOL_1;
DataAwsIdentitystoreGroup[_c] = { fqn: "@cdktf/provider-aws.dataAwsIdentitystoreGroup.DataAwsIdentitystoreGroup", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsIdentitystoreGroup.tfResourceType = "aws_identitystore_group";
