// @cdktf/provider-aws/lib/cloudwatch-event-connection/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudwatchEventConnection = exports.CloudwatchEventConnectionAuthParametersOutputReference = exports.cloudwatchEventConnectionAuthParametersToTerraform = exports.CloudwatchEventConnectionAuthParametersOauthOutputReference = exports.cloudwatchEventConnectionAuthParametersOauthToTerraform = exports.CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersOutputReference = exports.cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersToTerraform = exports.CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersQueryStringList = exports.CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersQueryStringOutputReference = exports.cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersQueryStringToTerraform = exports.CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersHeaderList = exports.CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersHeaderOutputReference = exports.cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersHeaderToTerraform = exports.CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersBodyList = exports.CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersBodyOutputReference = exports.cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersBodyToTerraform = exports.CloudwatchEventConnectionAuthParametersOauthClientParametersOutputReference = exports.cloudwatchEventConnectionAuthParametersOauthClientParametersToTerraform = exports.CloudwatchEventConnectionAuthParametersInvocationHttpParametersOutputReference = exports.cloudwatchEventConnectionAuthParametersInvocationHttpParametersToTerraform = exports.CloudwatchEventConnectionAuthParametersInvocationHttpParametersQueryStringList = exports.CloudwatchEventConnectionAuthParametersInvocationHttpParametersQueryStringOutputReference = exports.cloudwatchEventConnectionAuthParametersInvocationHttpParametersQueryStringToTerraform = exports.CloudwatchEventConnectionAuthParametersInvocationHttpParametersHeaderList = exports.CloudwatchEventConnectionAuthParametersInvocationHttpParametersHeaderOutputReference = exports.cloudwatchEventConnectionAuthParametersInvocationHttpParametersHeaderToTerraform = exports.CloudwatchEventConnectionAuthParametersInvocationHttpParametersBodyList = exports.CloudwatchEventConnectionAuthParametersInvocationHttpParametersBodyOutputReference = exports.cloudwatchEventConnectionAuthParametersInvocationHttpParametersBodyToTerraform = exports.CloudwatchEventConnectionAuthParametersBasicOutputReference = exports.cloudwatchEventConnectionAuthParametersBasicToTerraform = exports.CloudwatchEventConnectionAuthParametersApiKeyOutputReference = exports.cloudwatchEventConnectionAuthParametersApiKeyToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function cloudwatchEventConnectionAuthParametersApiKeyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.cloudwatchEventConnectionAuthParametersApiKeyToTerraform = cloudwatchEventConnectionAuthParametersApiKeyToTerraform;
class CloudwatchEventConnectionAuthParametersApiKeyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._key = undefined;
            this._value = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._key = value.key;
            this._value = value.value;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.CloudwatchEventConnectionAuthParametersApiKeyOutputReference = CloudwatchEventConnectionAuthParametersApiKeyOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersApiKeyOutputReference[_a] = { fqn: "@cdktf/provider-aws.cloudwatchEventConnection.CloudwatchEventConnectionAuthParametersApiKeyOutputReference", version: "10.0.12" };
function cloudwatchEventConnectionAuthParametersBasicToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        password: cdktf.stringToTerraform(struct.password),
        username: cdktf.stringToTerraform(struct.username),
    };
}
exports.cloudwatchEventConnectionAuthParametersBasicToTerraform = cloudwatchEventConnectionAuthParametersBasicToTerraform;
class CloudwatchEventConnectionAuthParametersBasicOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._password !== undefined) {
            hasAnyValues = true;
            internalValueResult.password = this._password;
        }
        if (this._username !== undefined) {
            hasAnyValues = true;
            internalValueResult.username = this._username;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._password = undefined;
            this._username = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._password = value.password;
            this._username = value.username;
        }
    }
    get password() {
        return this.getStringAttribute('password');
    }
    set password(value) {
        this._password = value;
    }
    // Temporarily expose input value. Use with caution.
    get passwordInput() {
        return this._password;
    }
    get username() {
        return this.getStringAttribute('username');
    }
    set username(value) {
        this._username = value;
    }
    // Temporarily expose input value. Use with caution.
    get usernameInput() {
        return this._username;
    }
}
exports.CloudwatchEventConnectionAuthParametersBasicOutputReference = CloudwatchEventConnectionAuthParametersBasicOutputReference;
_b = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersBasicOutputReference[_b] = { fqn: "@cdktf/provider-aws.cloudwatchEventConnection.CloudwatchEventConnectionAuthParametersBasicOutputReference", version: "10.0.12" };
function cloudwatchEventConnectionAuthParametersInvocationHttpParametersBodyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        is_value_secret: cdktf.booleanToTerraform(struct.isValueSecret),
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.cloudwatchEventConnectionAuthParametersInvocationHttpParametersBodyToTerraform = cloudwatchEventConnectionAuthParametersInvocationHttpParametersBodyToTerraform;
class CloudwatchEventConnectionAuthParametersInvocationHttpParametersBodyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._isValueSecret !== undefined) {
            hasAnyValues = true;
            internalValueResult.isValueSecret = this._isValueSecret;
        }
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._isValueSecret = undefined;
            this._key = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._isValueSecret = value.isValueSecret;
            this._key = value.key;
            this._value = value.value;
        }
    }
    get isValueSecret() {
        return this.getBooleanAttribute('is_value_secret');
    }
    set isValueSecret(value) {
        this._isValueSecret = value;
    }
    resetIsValueSecret() {
        this._isValueSecret = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get isValueSecretInput() {
        return this._isValueSecret;
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.CloudwatchEventConnectionAuthParametersInvocationHttpParametersBodyOutputReference = CloudwatchEventConnectionAuthParametersInvocationHttpParametersBodyOutputReference;
_c = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersInvocationHttpParametersBodyOutputReference[_c] = { fqn: "@cdktf/provider-aws.cloudwatchEventConnection.CloudwatchEventConnectionAuthParametersInvocationHttpParametersBodyOutputReference", version: "10.0.12" };
class CloudwatchEventConnectionAuthParametersInvocationHttpParametersBodyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudwatchEventConnectionAuthParametersInvocationHttpParametersBodyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudwatchEventConnectionAuthParametersInvocationHttpParametersBodyList = CloudwatchEventConnectionAuthParametersInvocationHttpParametersBodyList;
_d = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersInvocationHttpParametersBodyList[_d] = { fqn: "@cdktf/provider-aws.cloudwatchEventConnection.CloudwatchEventConnectionAuthParametersInvocationHttpParametersBodyList", version: "10.0.12" };
function cloudwatchEventConnectionAuthParametersInvocationHttpParametersHeaderToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        is_value_secret: cdktf.booleanToTerraform(struct.isValueSecret),
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.cloudwatchEventConnectionAuthParametersInvocationHttpParametersHeaderToTerraform = cloudwatchEventConnectionAuthParametersInvocationHttpParametersHeaderToTerraform;
class CloudwatchEventConnectionAuthParametersInvocationHttpParametersHeaderOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._isValueSecret !== undefined) {
            hasAnyValues = true;
            internalValueResult.isValueSecret = this._isValueSecret;
        }
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._isValueSecret = undefined;
            this._key = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._isValueSecret = value.isValueSecret;
            this._key = value.key;
            this._value = value.value;
        }
    }
    get isValueSecret() {
        return this.getBooleanAttribute('is_value_secret');
    }
    set isValueSecret(value) {
        this._isValueSecret = value;
    }
    resetIsValueSecret() {
        this._isValueSecret = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get isValueSecretInput() {
        return this._isValueSecret;
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.CloudwatchEventConnectionAuthParametersInvocationHttpParametersHeaderOutputReference = CloudwatchEventConnectionAuthParametersInvocationHttpParametersHeaderOutputReference;
_e = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersInvocationHttpParametersHeaderOutputReference[_e] = { fqn: "@cdktf/provider-aws.cloudwatchEventConnection.CloudwatchEventConnectionAuthParametersInvocationHttpParametersHeaderOutputReference", version: "10.0.12" };
class CloudwatchEventConnectionAuthParametersInvocationHttpParametersHeaderList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudwatchEventConnectionAuthParametersInvocationHttpParametersHeaderOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudwatchEventConnectionAuthParametersInvocationHttpParametersHeaderList = CloudwatchEventConnectionAuthParametersInvocationHttpParametersHeaderList;
_f = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersInvocationHttpParametersHeaderList[_f] = { fqn: "@cdktf/provider-aws.cloudwatchEventConnection.CloudwatchEventConnectionAuthParametersInvocationHttpParametersHeaderList", version: "10.0.12" };
function cloudwatchEventConnectionAuthParametersInvocationHttpParametersQueryStringToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        is_value_secret: cdktf.booleanToTerraform(struct.isValueSecret),
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.cloudwatchEventConnectionAuthParametersInvocationHttpParametersQueryStringToTerraform = cloudwatchEventConnectionAuthParametersInvocationHttpParametersQueryStringToTerraform;
class CloudwatchEventConnectionAuthParametersInvocationHttpParametersQueryStringOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._isValueSecret !== undefined) {
            hasAnyValues = true;
            internalValueResult.isValueSecret = this._isValueSecret;
        }
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._isValueSecret = undefined;
            this._key = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._isValueSecret = value.isValueSecret;
            this._key = value.key;
            this._value = value.value;
        }
    }
    get isValueSecret() {
        return this.getBooleanAttribute('is_value_secret');
    }
    set isValueSecret(value) {
        this._isValueSecret = value;
    }
    resetIsValueSecret() {
        this._isValueSecret = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get isValueSecretInput() {
        return this._isValueSecret;
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.CloudwatchEventConnectionAuthParametersInvocationHttpParametersQueryStringOutputReference = CloudwatchEventConnectionAuthParametersInvocationHttpParametersQueryStringOutputReference;
_g = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersInvocationHttpParametersQueryStringOutputReference[_g] = { fqn: "@cdktf/provider-aws.cloudwatchEventConnection.CloudwatchEventConnectionAuthParametersInvocationHttpParametersQueryStringOutputReference", version: "10.0.12" };
class CloudwatchEventConnectionAuthParametersInvocationHttpParametersQueryStringList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudwatchEventConnectionAuthParametersInvocationHttpParametersQueryStringOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudwatchEventConnectionAuthParametersInvocationHttpParametersQueryStringList = CloudwatchEventConnectionAuthParametersInvocationHttpParametersQueryStringList;
_h = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersInvocationHttpParametersQueryStringList[_h] = { fqn: "@cdktf/provider-aws.cloudwatchEventConnection.CloudwatchEventConnectionAuthParametersInvocationHttpParametersQueryStringList", version: "10.0.12" };
function cloudwatchEventConnectionAuthParametersInvocationHttpParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        body: cdktf.listMapper(cloudwatchEventConnectionAuthParametersInvocationHttpParametersBodyToTerraform, true)(struct.body),
        header: cdktf.listMapper(cloudwatchEventConnectionAuthParametersInvocationHttpParametersHeaderToTerraform, true)(struct.header),
        query_string: cdktf.listMapper(cloudwatchEventConnectionAuthParametersInvocationHttpParametersQueryStringToTerraform, true)(struct.queryString),
    };
}
exports.cloudwatchEventConnectionAuthParametersInvocationHttpParametersToTerraform = cloudwatchEventConnectionAuthParametersInvocationHttpParametersToTerraform;
class CloudwatchEventConnectionAuthParametersInvocationHttpParametersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // body - computed: false, optional: true, required: false
        this._body = new CloudwatchEventConnectionAuthParametersInvocationHttpParametersBodyList(this, "body", false);
        // header - computed: false, optional: true, required: false
        this._header = new CloudwatchEventConnectionAuthParametersInvocationHttpParametersHeaderList(this, "header", false);
        // query_string - computed: false, optional: true, required: false
        this._queryString = new CloudwatchEventConnectionAuthParametersInvocationHttpParametersQueryStringList(this, "query_string", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._body?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.body = this._body?.internalValue;
        }
        if (this._header?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.header = this._header?.internalValue;
        }
        if (this._queryString?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.queryString = this._queryString?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._body.internalValue = undefined;
            this._header.internalValue = undefined;
            this._queryString.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._body.internalValue = value.body;
            this._header.internalValue = value.header;
            this._queryString.internalValue = value.queryString;
        }
    }
    get body() {
        return this._body;
    }
    putBody(value) {
        this._body.internalValue = value;
    }
    resetBody() {
        this._body.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bodyInput() {
        return this._body.internalValue;
    }
    get header() {
        return this._header;
    }
    putHeader(value) {
        this._header.internalValue = value;
    }
    resetHeader() {
        this._header.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get headerInput() {
        return this._header.internalValue;
    }
    get queryString() {
        return this._queryString;
    }
    putQueryString(value) {
        this._queryString.internalValue = value;
    }
    resetQueryString() {
        this._queryString.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get queryStringInput() {
        return this._queryString.internalValue;
    }
}
exports.CloudwatchEventConnectionAuthParametersInvocationHttpParametersOutputReference = CloudwatchEventConnectionAuthParametersInvocationHttpParametersOutputReference;
_j = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersInvocationHttpParametersOutputReference[_j] = { fqn: "@cdktf/provider-aws.cloudwatchEventConnection.CloudwatchEventConnectionAuthParametersInvocationHttpParametersOutputReference", version: "10.0.12" };
function cloudwatchEventConnectionAuthParametersOauthClientParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        client_id: cdktf.stringToTerraform(struct.clientId),
        client_secret: cdktf.stringToTerraform(struct.clientSecret),
    };
}
exports.cloudwatchEventConnectionAuthParametersOauthClientParametersToTerraform = cloudwatchEventConnectionAuthParametersOauthClientParametersToTerraform;
class CloudwatchEventConnectionAuthParametersOauthClientParametersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._clientId !== undefined) {
            hasAnyValues = true;
            internalValueResult.clientId = this._clientId;
        }
        if (this._clientSecret !== undefined) {
            hasAnyValues = true;
            internalValueResult.clientSecret = this._clientSecret;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._clientId = undefined;
            this._clientSecret = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._clientId = value.clientId;
            this._clientSecret = value.clientSecret;
        }
    }
    get clientId() {
        return this.getStringAttribute('client_id');
    }
    set clientId(value) {
        this._clientId = value;
    }
    // Temporarily expose input value. Use with caution.
    get clientIdInput() {
        return this._clientId;
    }
    get clientSecret() {
        return this.getStringAttribute('client_secret');
    }
    set clientSecret(value) {
        this._clientSecret = value;
    }
    // Temporarily expose input value. Use with caution.
    get clientSecretInput() {
        return this._clientSecret;
    }
}
exports.CloudwatchEventConnectionAuthParametersOauthClientParametersOutputReference = CloudwatchEventConnectionAuthParametersOauthClientParametersOutputReference;
_k = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersOauthClientParametersOutputReference[_k] = { fqn: "@cdktf/provider-aws.cloudwatchEventConnection.CloudwatchEventConnectionAuthParametersOauthClientParametersOutputReference", version: "10.0.12" };
function cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersBodyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        is_value_secret: cdktf.booleanToTerraform(struct.isValueSecret),
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersBodyToTerraform = cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersBodyToTerraform;
class CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersBodyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._isValueSecret !== undefined) {
            hasAnyValues = true;
            internalValueResult.isValueSecret = this._isValueSecret;
        }
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._isValueSecret = undefined;
            this._key = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._isValueSecret = value.isValueSecret;
            this._key = value.key;
            this._value = value.value;
        }
    }
    get isValueSecret() {
        return this.getBooleanAttribute('is_value_secret');
    }
    set isValueSecret(value) {
        this._isValueSecret = value;
    }
    resetIsValueSecret() {
        this._isValueSecret = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get isValueSecretInput() {
        return this._isValueSecret;
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersBodyOutputReference = CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersBodyOutputReference;
_l = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersBodyOutputReference[_l] = { fqn: "@cdktf/provider-aws.cloudwatchEventConnection.CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersBodyOutputReference", version: "10.0.12" };
class CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersBodyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersBodyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersBodyList = CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersBodyList;
_m = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersBodyList[_m] = { fqn: "@cdktf/provider-aws.cloudwatchEventConnection.CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersBodyList", version: "10.0.12" };
function cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersHeaderToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        is_value_secret: cdktf.booleanToTerraform(struct.isValueSecret),
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersHeaderToTerraform = cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersHeaderToTerraform;
class CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersHeaderOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._isValueSecret !== undefined) {
            hasAnyValues = true;
            internalValueResult.isValueSecret = this._isValueSecret;
        }
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._isValueSecret = undefined;
            this._key = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._isValueSecret = value.isValueSecret;
            this._key = value.key;
            this._value = value.value;
        }
    }
    get isValueSecret() {
        return this.getBooleanAttribute('is_value_secret');
    }
    set isValueSecret(value) {
        this._isValueSecret = value;
    }
    resetIsValueSecret() {
        this._isValueSecret = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get isValueSecretInput() {
        return this._isValueSecret;
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersHeaderOutputReference = CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersHeaderOutputReference;
_o = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersHeaderOutputReference[_o] = { fqn: "@cdktf/provider-aws.cloudwatchEventConnection.CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersHeaderOutputReference", version: "10.0.12" };
class CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersHeaderList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersHeaderOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersHeaderList = CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersHeaderList;
_p = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersHeaderList[_p] = { fqn: "@cdktf/provider-aws.cloudwatchEventConnection.CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersHeaderList", version: "10.0.12" };
function cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersQueryStringToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        is_value_secret: cdktf.booleanToTerraform(struct.isValueSecret),
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersQueryStringToTerraform = cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersQueryStringToTerraform;
class CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersQueryStringOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._isValueSecret !== undefined) {
            hasAnyValues = true;
            internalValueResult.isValueSecret = this._isValueSecret;
        }
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._isValueSecret = undefined;
            this._key = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._isValueSecret = value.isValueSecret;
            this._key = value.key;
            this._value = value.value;
        }
    }
    get isValueSecret() {
        return this.getBooleanAttribute('is_value_secret');
    }
    set isValueSecret(value) {
        this._isValueSecret = value;
    }
    resetIsValueSecret() {
        this._isValueSecret = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get isValueSecretInput() {
        return this._isValueSecret;
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersQueryStringOutputReference = CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersQueryStringOutputReference;
_q = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersQueryStringOutputReference[_q] = { fqn: "@cdktf/provider-aws.cloudwatchEventConnection.CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersQueryStringOutputReference", version: "10.0.12" };
class CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersQueryStringList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersQueryStringOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersQueryStringList = CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersQueryStringList;
_r = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersQueryStringList[_r] = { fqn: "@cdktf/provider-aws.cloudwatchEventConnection.CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersQueryStringList", version: "10.0.12" };
function cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        body: cdktf.listMapper(cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersBodyToTerraform, true)(struct.body),
        header: cdktf.listMapper(cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersHeaderToTerraform, true)(struct.header),
        query_string: cdktf.listMapper(cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersQueryStringToTerraform, true)(struct.queryString),
    };
}
exports.cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersToTerraform = cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersToTerraform;
class CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // body - computed: false, optional: true, required: false
        this._body = new CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersBodyList(this, "body", false);
        // header - computed: false, optional: true, required: false
        this._header = new CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersHeaderList(this, "header", false);
        // query_string - computed: false, optional: true, required: false
        this._queryString = new CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersQueryStringList(this, "query_string", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._body?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.body = this._body?.internalValue;
        }
        if (this._header?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.header = this._header?.internalValue;
        }
        if (this._queryString?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.queryString = this._queryString?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._body.internalValue = undefined;
            this._header.internalValue = undefined;
            this._queryString.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._body.internalValue = value.body;
            this._header.internalValue = value.header;
            this._queryString.internalValue = value.queryString;
        }
    }
    get body() {
        return this._body;
    }
    putBody(value) {
        this._body.internalValue = value;
    }
    resetBody() {
        this._body.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bodyInput() {
        return this._body.internalValue;
    }
    get header() {
        return this._header;
    }
    putHeader(value) {
        this._header.internalValue = value;
    }
    resetHeader() {
        this._header.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get headerInput() {
        return this._header.internalValue;
    }
    get queryString() {
        return this._queryString;
    }
    putQueryString(value) {
        this._queryString.internalValue = value;
    }
    resetQueryString() {
        this._queryString.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get queryStringInput() {
        return this._queryString.internalValue;
    }
}
exports.CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersOutputReference = CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersOutputReference;
_s = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersOutputReference[_s] = { fqn: "@cdktf/provider-aws.cloudwatchEventConnection.CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersOutputReference", version: "10.0.12" };
function cloudwatchEventConnectionAuthParametersOauthToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        authorization_endpoint: cdktf.stringToTerraform(struct.authorizationEndpoint),
        http_method: cdktf.stringToTerraform(struct.httpMethod),
        client_parameters: cloudwatchEventConnectionAuthParametersOauthClientParametersToTerraform(struct.clientParameters),
        oauth_http_parameters: cloudwatchEventConnectionAuthParametersOauthOauthHttpParametersToTerraform(struct.oauthHttpParameters),
    };
}
exports.cloudwatchEventConnectionAuthParametersOauthToTerraform = cloudwatchEventConnectionAuthParametersOauthToTerraform;
class CloudwatchEventConnectionAuthParametersOauthOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // client_parameters - computed: false, optional: true, required: false
        this._clientParameters = new CloudwatchEventConnectionAuthParametersOauthClientParametersOutputReference(this, "client_parameters");
        // oauth_http_parameters - computed: false, optional: false, required: true
        this._oauthHttpParameters = new CloudwatchEventConnectionAuthParametersOauthOauthHttpParametersOutputReference(this, "oauth_http_parameters");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._authorizationEndpoint !== undefined) {
            hasAnyValues = true;
            internalValueResult.authorizationEndpoint = this._authorizationEndpoint;
        }
        if (this._httpMethod !== undefined) {
            hasAnyValues = true;
            internalValueResult.httpMethod = this._httpMethod;
        }
        if (this._clientParameters?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.clientParameters = this._clientParameters?.internalValue;
        }
        if (this._oauthHttpParameters?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.oauthHttpParameters = this._oauthHttpParameters?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._authorizationEndpoint = undefined;
            this._httpMethod = undefined;
            this._clientParameters.internalValue = undefined;
            this._oauthHttpParameters.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._authorizationEndpoint = value.authorizationEndpoint;
            this._httpMethod = value.httpMethod;
            this._clientParameters.internalValue = value.clientParameters;
            this._oauthHttpParameters.internalValue = value.oauthHttpParameters;
        }
    }
    get authorizationEndpoint() {
        return this.getStringAttribute('authorization_endpoint');
    }
    set authorizationEndpoint(value) {
        this._authorizationEndpoint = value;
    }
    // Temporarily expose input value. Use with caution.
    get authorizationEndpointInput() {
        return this._authorizationEndpoint;
    }
    get httpMethod() {
        return this.getStringAttribute('http_method');
    }
    set httpMethod(value) {
        this._httpMethod = value;
    }
    // Temporarily expose input value. Use with caution.
    get httpMethodInput() {
        return this._httpMethod;
    }
    get clientParameters() {
        return this._clientParameters;
    }
    putClientParameters(value) {
        this._clientParameters.internalValue = value;
    }
    resetClientParameters() {
        this._clientParameters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clientParametersInput() {
        return this._clientParameters.internalValue;
    }
    get oauthHttpParameters() {
        return this._oauthHttpParameters;
    }
    putOauthHttpParameters(value) {
        this._oauthHttpParameters.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get oauthHttpParametersInput() {
        return this._oauthHttpParameters.internalValue;
    }
}
exports.CloudwatchEventConnectionAuthParametersOauthOutputReference = CloudwatchEventConnectionAuthParametersOauthOutputReference;
_t = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersOauthOutputReference[_t] = { fqn: "@cdktf/provider-aws.cloudwatchEventConnection.CloudwatchEventConnectionAuthParametersOauthOutputReference", version: "10.0.12" };
function cloudwatchEventConnectionAuthParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        api_key: cloudwatchEventConnectionAuthParametersApiKeyToTerraform(struct.apiKey),
        basic: cloudwatchEventConnectionAuthParametersBasicToTerraform(struct.basic),
        invocation_http_parameters: cloudwatchEventConnectionAuthParametersInvocationHttpParametersToTerraform(struct.invocationHttpParameters),
        oauth: cloudwatchEventConnectionAuthParametersOauthToTerraform(struct.oauth),
    };
}
exports.cloudwatchEventConnectionAuthParametersToTerraform = cloudwatchEventConnectionAuthParametersToTerraform;
class CloudwatchEventConnectionAuthParametersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // api_key - computed: false, optional: true, required: false
        this._apiKey = new CloudwatchEventConnectionAuthParametersApiKeyOutputReference(this, "api_key");
        // basic - computed: false, optional: true, required: false
        this._basic = new CloudwatchEventConnectionAuthParametersBasicOutputReference(this, "basic");
        // invocation_http_parameters - computed: false, optional: true, required: false
        this._invocationHttpParameters = new CloudwatchEventConnectionAuthParametersInvocationHttpParametersOutputReference(this, "invocation_http_parameters");
        // oauth - computed: false, optional: true, required: false
        this._oauth = new CloudwatchEventConnectionAuthParametersOauthOutputReference(this, "oauth");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._apiKey?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.apiKey = this._apiKey?.internalValue;
        }
        if (this._basic?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.basic = this._basic?.internalValue;
        }
        if (this._invocationHttpParameters?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.invocationHttpParameters = this._invocationHttpParameters?.internalValue;
        }
        if (this._oauth?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.oauth = this._oauth?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._apiKey.internalValue = undefined;
            this._basic.internalValue = undefined;
            this._invocationHttpParameters.internalValue = undefined;
            this._oauth.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._apiKey.internalValue = value.apiKey;
            this._basic.internalValue = value.basic;
            this._invocationHttpParameters.internalValue = value.invocationHttpParameters;
            this._oauth.internalValue = value.oauth;
        }
    }
    get apiKey() {
        return this._apiKey;
    }
    putApiKey(value) {
        this._apiKey.internalValue = value;
    }
    resetApiKey() {
        this._apiKey.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get apiKeyInput() {
        return this._apiKey.internalValue;
    }
    get basic() {
        return this._basic;
    }
    putBasic(value) {
        this._basic.internalValue = value;
    }
    resetBasic() {
        this._basic.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get basicInput() {
        return this._basic.internalValue;
    }
    get invocationHttpParameters() {
        return this._invocationHttpParameters;
    }
    putInvocationHttpParameters(value) {
        this._invocationHttpParameters.internalValue = value;
    }
    resetInvocationHttpParameters() {
        this._invocationHttpParameters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get invocationHttpParametersInput() {
        return this._invocationHttpParameters.internalValue;
    }
    get oauth() {
        return this._oauth;
    }
    putOauth(value) {
        this._oauth.internalValue = value;
    }
    resetOauth() {
        this._oauth.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get oauthInput() {
        return this._oauth.internalValue;
    }
}
exports.CloudwatchEventConnectionAuthParametersOutputReference = CloudwatchEventConnectionAuthParametersOutputReference;
_u = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnectionAuthParametersOutputReference[_u] = { fqn: "@cdktf/provider-aws.cloudwatchEventConnection.CloudwatchEventConnectionAuthParametersOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/cloudwatch_event_connection aws_cloudwatch_event_connection}
*/
class CloudwatchEventConnection extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/cloudwatch_event_connection aws_cloudwatch_event_connection} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CloudwatchEventConnectionConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_cloudwatch_event_connection',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // auth_parameters - computed: false, optional: false, required: true
        this._authParameters = new CloudwatchEventConnectionAuthParametersOutputReference(this, "auth_parameters");
        this._authorizationType = config.authorizationType;
        this._description = config.description;
        this._id = config.id;
        this._name = config.name;
        this._authParameters.internalValue = config.authParameters;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get authorizationType() {
        return this.getStringAttribute('authorization_type');
    }
    set authorizationType(value) {
        this._authorizationType = value;
    }
    // Temporarily expose input value. Use with caution.
    get authorizationTypeInput() {
        return this._authorizationType;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // secret_arn - computed: true, optional: false, required: false
    get secretArn() {
        return this.getStringAttribute('secret_arn');
    }
    get authParameters() {
        return this._authParameters;
    }
    putAuthParameters(value) {
        this._authParameters.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get authParametersInput() {
        return this._authParameters.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            authorization_type: cdktf.stringToTerraform(this._authorizationType),
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            auth_parameters: cloudwatchEventConnectionAuthParametersToTerraform(this._authParameters.internalValue),
        };
    }
}
exports.CloudwatchEventConnection = CloudwatchEventConnection;
_v = JSII_RTTI_SYMBOL_1;
CloudwatchEventConnection[_v] = { fqn: "@cdktf/provider-aws.cloudwatchEventConnection.CloudwatchEventConnection", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
CloudwatchEventConnection.tfResourceType = "aws_cloudwatch_event_connection";
