// @cdktf/provider-aws/lib/sagemaker-device-fleet/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SagemakerDeviceFleet = exports.SagemakerDeviceFleetOutputConfigOutputReference = exports.sagemakerDeviceFleetOutputConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function sagemakerDeviceFleetOutputConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        kms_key_id: cdktf.stringToTerraform(struct.kmsKeyId),
        s3_output_location: cdktf.stringToTerraform(struct.s3OutputLocation),
    };
}
exports.sagemakerDeviceFleetOutputConfigToTerraform = sagemakerDeviceFleetOutputConfigToTerraform;
class SagemakerDeviceFleetOutputConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._kmsKeyId !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKeyId = this._kmsKeyId;
        }
        if (this._s3OutputLocation !== undefined) {
            hasAnyValues = true;
            internalValueResult.s3OutputLocation = this._s3OutputLocation;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._kmsKeyId = undefined;
            this._s3OutputLocation = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._kmsKeyId = value.kmsKeyId;
            this._s3OutputLocation = value.s3OutputLocation;
        }
    }
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    resetKmsKeyId() {
        this._kmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
    get s3OutputLocation() {
        return this.getStringAttribute('s3_output_location');
    }
    set s3OutputLocation(value) {
        this._s3OutputLocation = value;
    }
    // Temporarily expose input value. Use with caution.
    get s3OutputLocationInput() {
        return this._s3OutputLocation;
    }
}
exports.SagemakerDeviceFleetOutputConfigOutputReference = SagemakerDeviceFleetOutputConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SagemakerDeviceFleetOutputConfigOutputReference[_a] = { fqn: "@cdktf/provider-aws.sagemakerDeviceFleet.SagemakerDeviceFleetOutputConfigOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/sagemaker_device_fleet aws_sagemaker_device_fleet}
*/
class SagemakerDeviceFleet extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/sagemaker_device_fleet aws_sagemaker_device_fleet} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SagemakerDeviceFleetConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_sagemaker_device_fleet',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // output_config - computed: false, optional: false, required: true
        this._outputConfig = new SagemakerDeviceFleetOutputConfigOutputReference(this, "output_config");
        this._description = config.description;
        this._deviceFleetName = config.deviceFleetName;
        this._enableIotRoleAlias = config.enableIotRoleAlias;
        this._id = config.id;
        this._roleArn = config.roleArn;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._outputConfig.internalValue = config.outputConfig;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get deviceFleetName() {
        return this.getStringAttribute('device_fleet_name');
    }
    set deviceFleetName(value) {
        this._deviceFleetName = value;
    }
    // Temporarily expose input value. Use with caution.
    get deviceFleetNameInput() {
        return this._deviceFleetName;
    }
    get enableIotRoleAlias() {
        return this.getBooleanAttribute('enable_iot_role_alias');
    }
    set enableIotRoleAlias(value) {
        this._enableIotRoleAlias = value;
    }
    resetEnableIotRoleAlias() {
        this._enableIotRoleAlias = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableIotRoleAliasInput() {
        return this._enableIotRoleAlias;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // iot_role_alias - computed: true, optional: false, required: false
    get iotRoleAlias() {
        return this.getStringAttribute('iot_role_alias');
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get outputConfig() {
        return this._outputConfig;
    }
    putOutputConfig(value) {
        this._outputConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get outputConfigInput() {
        return this._outputConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            device_fleet_name: cdktf.stringToTerraform(this._deviceFleetName),
            enable_iot_role_alias: cdktf.booleanToTerraform(this._enableIotRoleAlias),
            id: cdktf.stringToTerraform(this._id),
            role_arn: cdktf.stringToTerraform(this._roleArn),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            output_config: sagemakerDeviceFleetOutputConfigToTerraform(this._outputConfig.internalValue),
        };
    }
}
exports.SagemakerDeviceFleet = SagemakerDeviceFleet;
_b = JSII_RTTI_SYMBOL_1;
SagemakerDeviceFleet[_b] = { fqn: "@cdktf/provider-aws.sagemakerDeviceFleet.SagemakerDeviceFleet", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
SagemakerDeviceFleet.tfResourceType = "aws_sagemaker_device_fleet";
