// @cdktf/provider-aws/lib/data-aws-redshift-cluster/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsRedshiftCluster = exports.DataAwsRedshiftClusterClusterNodesList = exports.DataAwsRedshiftClusterClusterNodesOutputReference = exports.dataAwsRedshiftClusterClusterNodesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsRedshiftClusterClusterNodesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsRedshiftClusterClusterNodesToTerraform = dataAwsRedshiftClusterClusterNodesToTerraform;
class DataAwsRedshiftClusterClusterNodesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // node_role - computed: true, optional: false, required: false
    get nodeRole() {
        return this.getStringAttribute('node_role');
    }
    // private_ip_address - computed: true, optional: false, required: false
    get privateIpAddress() {
        return this.getStringAttribute('private_ip_address');
    }
    // public_ip_address - computed: true, optional: false, required: false
    get publicIpAddress() {
        return this.getStringAttribute('public_ip_address');
    }
}
exports.DataAwsRedshiftClusterClusterNodesOutputReference = DataAwsRedshiftClusterClusterNodesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsRedshiftClusterClusterNodesOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsRedshiftCluster.DataAwsRedshiftClusterClusterNodesOutputReference", version: "10.0.12" };
class DataAwsRedshiftClusterClusterNodesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsRedshiftClusterClusterNodesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsRedshiftClusterClusterNodesList = DataAwsRedshiftClusterClusterNodesList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsRedshiftClusterClusterNodesList[_b] = { fqn: "@cdktf/provider-aws.dataAwsRedshiftCluster.DataAwsRedshiftClusterClusterNodesList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/redshift_cluster aws_redshift_cluster}
*/
class DataAwsRedshiftCluster extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/redshift_cluster aws_redshift_cluster} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsRedshiftClusterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_redshift_cluster',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // cluster_nodes - computed: true, optional: false, required: false
        this._clusterNodes = new DataAwsRedshiftClusterClusterNodesList(this, "cluster_nodes", false);
        this._clusterIdentifier = config.clusterIdentifier;
        this._id = config.id;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // allow_version_upgrade - computed: true, optional: false, required: false
    get allowVersionUpgrade() {
        return this.getBooleanAttribute('allow_version_upgrade');
    }
    // aqua_configuration_status - computed: true, optional: false, required: false
    get aquaConfigurationStatus() {
        return this.getStringAttribute('aqua_configuration_status');
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // automated_snapshot_retention_period - computed: true, optional: false, required: false
    get automatedSnapshotRetentionPeriod() {
        return this.getNumberAttribute('automated_snapshot_retention_period');
    }
    // availability_zone - computed: true, optional: false, required: false
    get availabilityZone() {
        return this.getStringAttribute('availability_zone');
    }
    // availability_zone_relocation_enabled - computed: true, optional: false, required: false
    get availabilityZoneRelocationEnabled() {
        return this.getBooleanAttribute('availability_zone_relocation_enabled');
    }
    // bucket_name - computed: true, optional: false, required: false
    get bucketName() {
        return this.getStringAttribute('bucket_name');
    }
    get clusterIdentifier() {
        return this.getStringAttribute('cluster_identifier');
    }
    set clusterIdentifier(value) {
        this._clusterIdentifier = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterIdentifierInput() {
        return this._clusterIdentifier;
    }
    get clusterNodes() {
        return this._clusterNodes;
    }
    // cluster_parameter_group_name - computed: true, optional: false, required: false
    get clusterParameterGroupName() {
        return this.getStringAttribute('cluster_parameter_group_name');
    }
    // cluster_public_key - computed: true, optional: false, required: false
    get clusterPublicKey() {
        return this.getStringAttribute('cluster_public_key');
    }
    // cluster_revision_number - computed: true, optional: false, required: false
    get clusterRevisionNumber() {
        return this.getStringAttribute('cluster_revision_number');
    }
    // cluster_security_groups - computed: true, optional: false, required: false
    get clusterSecurityGroups() {
        return this.getListAttribute('cluster_security_groups');
    }
    // cluster_subnet_group_name - computed: true, optional: false, required: false
    get clusterSubnetGroupName() {
        return this.getStringAttribute('cluster_subnet_group_name');
    }
    // cluster_type - computed: true, optional: false, required: false
    get clusterType() {
        return this.getStringAttribute('cluster_type');
    }
    // cluster_version - computed: true, optional: false, required: false
    get clusterVersion() {
        return this.getStringAttribute('cluster_version');
    }
    // database_name - computed: true, optional: false, required: false
    get databaseName() {
        return this.getStringAttribute('database_name');
    }
    // default_iam_role_arn - computed: true, optional: false, required: false
    get defaultIamRoleArn() {
        return this.getStringAttribute('default_iam_role_arn');
    }
    // elastic_ip - computed: true, optional: false, required: false
    get elasticIp() {
        return this.getStringAttribute('elastic_ip');
    }
    // enable_logging - computed: true, optional: false, required: false
    get enableLogging() {
        return this.getBooleanAttribute('enable_logging');
    }
    // encrypted - computed: true, optional: false, required: false
    get encrypted() {
        return this.getBooleanAttribute('encrypted');
    }
    // endpoint - computed: true, optional: false, required: false
    get endpoint() {
        return this.getStringAttribute('endpoint');
    }
    // enhanced_vpc_routing - computed: true, optional: false, required: false
    get enhancedVpcRouting() {
        return this.getBooleanAttribute('enhanced_vpc_routing');
    }
    // iam_roles - computed: true, optional: false, required: false
    get iamRoles() {
        return this.getListAttribute('iam_roles');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // kms_key_id - computed: true, optional: false, required: false
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    // log_destination_type - computed: true, optional: false, required: false
    get logDestinationType() {
        return this.getStringAttribute('log_destination_type');
    }
    // log_exports - computed: true, optional: false, required: false
    get logExports() {
        return cdktf.Fn.tolist(this.getListAttribute('log_exports'));
    }
    // maintenance_track_name - computed: true, optional: false, required: false
    get maintenanceTrackName() {
        return this.getStringAttribute('maintenance_track_name');
    }
    // manual_snapshot_retention_period - computed: true, optional: false, required: false
    get manualSnapshotRetentionPeriod() {
        return this.getNumberAttribute('manual_snapshot_retention_period');
    }
    // master_username - computed: true, optional: false, required: false
    get masterUsername() {
        return this.getStringAttribute('master_username');
    }
    // node_type - computed: true, optional: false, required: false
    get nodeType() {
        return this.getStringAttribute('node_type');
    }
    // number_of_nodes - computed: true, optional: false, required: false
    get numberOfNodes() {
        return this.getNumberAttribute('number_of_nodes');
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getNumberAttribute('port');
    }
    // preferred_maintenance_window - computed: true, optional: false, required: false
    get preferredMaintenanceWindow() {
        return this.getStringAttribute('preferred_maintenance_window');
    }
    // publicly_accessible - computed: true, optional: false, required: false
    get publiclyAccessible() {
        return this.getBooleanAttribute('publicly_accessible');
    }
    // s3_key_prefix - computed: true, optional: false, required: false
    get s3KeyPrefix() {
        return this.getStringAttribute('s3_key_prefix');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // vpc_id - computed: true, optional: false, required: false
    get vpcId() {
        return this.getStringAttribute('vpc_id');
    }
    // vpc_security_group_ids - computed: true, optional: false, required: false
    get vpcSecurityGroupIds() {
        return this.getListAttribute('vpc_security_group_ids');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_identifier: cdktf.stringToTerraform(this._clusterIdentifier),
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
}
exports.DataAwsRedshiftCluster = DataAwsRedshiftCluster;
_c = JSII_RTTI_SYMBOL_1;
DataAwsRedshiftCluster[_c] = { fqn: "@cdktf/provider-aws.dataAwsRedshiftCluster.DataAwsRedshiftCluster", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsRedshiftCluster.tfResourceType = "aws_redshift_cluster";
