// @cdktf/provider-aws/lib/accessanalyzer-archive-rule/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccessanalyzerArchiveRule = exports.AccessanalyzerArchiveRuleFilterList = exports.AccessanalyzerArchiveRuleFilterOutputReference = exports.accessanalyzerArchiveRuleFilterToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function accessanalyzerArchiveRuleFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        contains: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.contains),
        criteria: cdktf.stringToTerraform(struct.criteria),
        eq: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.eq),
        exists: cdktf.stringToTerraform(struct.exists),
        neq: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.neq),
    };
}
exports.accessanalyzerArchiveRuleFilterToTerraform = accessanalyzerArchiveRuleFilterToTerraform;
class AccessanalyzerArchiveRuleFilterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._contains !== undefined) {
            hasAnyValues = true;
            internalValueResult.contains = this._contains;
        }
        if (this._criteria !== undefined) {
            hasAnyValues = true;
            internalValueResult.criteria = this._criteria;
        }
        if (this._eq !== undefined) {
            hasAnyValues = true;
            internalValueResult.eq = this._eq;
        }
        if (this._exists !== undefined) {
            hasAnyValues = true;
            internalValueResult.exists = this._exists;
        }
        if (this._neq !== undefined) {
            hasAnyValues = true;
            internalValueResult.neq = this._neq;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._contains = undefined;
            this._criteria = undefined;
            this._eq = undefined;
            this._exists = undefined;
            this._neq = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._contains = value.contains;
            this._criteria = value.criteria;
            this._eq = value.eq;
            this._exists = value.exists;
            this._neq = value.neq;
        }
    }
    get contains() {
        return this.getListAttribute('contains');
    }
    set contains(value) {
        this._contains = value;
    }
    resetContains() {
        this._contains = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containsInput() {
        return this._contains;
    }
    get criteria() {
        return this.getStringAttribute('criteria');
    }
    set criteria(value) {
        this._criteria = value;
    }
    // Temporarily expose input value. Use with caution.
    get criteriaInput() {
        return this._criteria;
    }
    get eq() {
        return this.getListAttribute('eq');
    }
    set eq(value) {
        this._eq = value;
    }
    resetEq() {
        this._eq = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get eqInput() {
        return this._eq;
    }
    get exists() {
        return this.getStringAttribute('exists');
    }
    set exists(value) {
        this._exists = value;
    }
    resetExists() {
        this._exists = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get existsInput() {
        return this._exists;
    }
    get neq() {
        return this.getListAttribute('neq');
    }
    set neq(value) {
        this._neq = value;
    }
    resetNeq() {
        this._neq = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get neqInput() {
        return this._neq;
    }
}
exports.AccessanalyzerArchiveRuleFilterOutputReference = AccessanalyzerArchiveRuleFilterOutputReference;
_a = JSII_RTTI_SYMBOL_1;
AccessanalyzerArchiveRuleFilterOutputReference[_a] = { fqn: "@cdktf/provider-aws.accessanalyzerArchiveRule.AccessanalyzerArchiveRuleFilterOutputReference", version: "10.0.12" };
class AccessanalyzerArchiveRuleFilterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AccessanalyzerArchiveRuleFilterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AccessanalyzerArchiveRuleFilterList = AccessanalyzerArchiveRuleFilterList;
_b = JSII_RTTI_SYMBOL_1;
AccessanalyzerArchiveRuleFilterList[_b] = { fqn: "@cdktf/provider-aws.accessanalyzerArchiveRule.AccessanalyzerArchiveRuleFilterList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/accessanalyzer_archive_rule aws_accessanalyzer_archive_rule}
*/
class AccessanalyzerArchiveRule extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/accessanalyzer_archive_rule aws_accessanalyzer_archive_rule} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options AccessanalyzerArchiveRuleConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_accessanalyzer_archive_rule',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // filter - computed: false, optional: false, required: true
        this._filter = new AccessanalyzerArchiveRuleFilterList(this, "filter", true);
        this._analyzerName = config.analyzerName;
        this._id = config.id;
        this._ruleName = config.ruleName;
        this._filter.internalValue = config.filter;
    }
    get analyzerName() {
        return this.getStringAttribute('analyzer_name');
    }
    set analyzerName(value) {
        this._analyzerName = value;
    }
    // Temporarily expose input value. Use with caution.
    get analyzerNameInput() {
        return this._analyzerName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get ruleName() {
        return this.getStringAttribute('rule_name');
    }
    set ruleName(value) {
        this._ruleName = value;
    }
    // Temporarily expose input value. Use with caution.
    get ruleNameInput() {
        return this._ruleName;
    }
    get filter() {
        return this._filter;
    }
    putFilter(value) {
        this._filter.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            analyzer_name: cdktf.stringToTerraform(this._analyzerName),
            id: cdktf.stringToTerraform(this._id),
            rule_name: cdktf.stringToTerraform(this._ruleName),
            filter: cdktf.listMapper(accessanalyzerArchiveRuleFilterToTerraform, true)(this._filter.internalValue),
        };
    }
}
exports.AccessanalyzerArchiveRule = AccessanalyzerArchiveRule;
_c = JSII_RTTI_SYMBOL_1;
AccessanalyzerArchiveRule[_c] = { fqn: "@cdktf/provider-aws.accessanalyzerArchiveRule.AccessanalyzerArchiveRule", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
AccessanalyzerArchiveRule.tfResourceType = "aws_accessanalyzer_archive_rule";
