// @cdktf/provider-aws/lib/vpc-ipam-pool/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VpcIpamPool = exports.VpcIpamPoolTimeoutsOutputReference = exports.vpcIpamPoolTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function vpcIpamPoolTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.vpcIpamPoolTimeoutsToTerraform = vpcIpamPoolTimeoutsToTerraform;
class VpcIpamPoolTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.VpcIpamPoolTimeoutsOutputReference = VpcIpamPoolTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
VpcIpamPoolTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-aws.vpcIpamPool.VpcIpamPoolTimeoutsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/vpc_ipam_pool aws_vpc_ipam_pool}
*/
class VpcIpamPool extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/vpc_ipam_pool aws_vpc_ipam_pool} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options VpcIpamPoolConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_vpc_ipam_pool',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new VpcIpamPoolTimeoutsOutputReference(this, "timeouts");
        this._addressFamily = config.addressFamily;
        this._allocationDefaultNetmaskLength = config.allocationDefaultNetmaskLength;
        this._allocationMaxNetmaskLength = config.allocationMaxNetmaskLength;
        this._allocationMinNetmaskLength = config.allocationMinNetmaskLength;
        this._allocationResourceTags = config.allocationResourceTags;
        this._autoImport = config.autoImport;
        this._awsService = config.awsService;
        this._description = config.description;
        this._id = config.id;
        this._ipamScopeId = config.ipamScopeId;
        this._locale = config.locale;
        this._publiclyAdvertisable = config.publiclyAdvertisable;
        this._sourceIpamPoolId = config.sourceIpamPoolId;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._timeouts.internalValue = config.timeouts;
    }
    get addressFamily() {
        return this.getStringAttribute('address_family');
    }
    set addressFamily(value) {
        this._addressFamily = value;
    }
    // Temporarily expose input value. Use with caution.
    get addressFamilyInput() {
        return this._addressFamily;
    }
    get allocationDefaultNetmaskLength() {
        return this.getNumberAttribute('allocation_default_netmask_length');
    }
    set allocationDefaultNetmaskLength(value) {
        this._allocationDefaultNetmaskLength = value;
    }
    resetAllocationDefaultNetmaskLength() {
        this._allocationDefaultNetmaskLength = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allocationDefaultNetmaskLengthInput() {
        return this._allocationDefaultNetmaskLength;
    }
    get allocationMaxNetmaskLength() {
        return this.getNumberAttribute('allocation_max_netmask_length');
    }
    set allocationMaxNetmaskLength(value) {
        this._allocationMaxNetmaskLength = value;
    }
    resetAllocationMaxNetmaskLength() {
        this._allocationMaxNetmaskLength = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allocationMaxNetmaskLengthInput() {
        return this._allocationMaxNetmaskLength;
    }
    get allocationMinNetmaskLength() {
        return this.getNumberAttribute('allocation_min_netmask_length');
    }
    set allocationMinNetmaskLength(value) {
        this._allocationMinNetmaskLength = value;
    }
    resetAllocationMinNetmaskLength() {
        this._allocationMinNetmaskLength = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allocationMinNetmaskLengthInput() {
        return this._allocationMinNetmaskLength;
    }
    get allocationResourceTags() {
        return this.getStringMapAttribute('allocation_resource_tags');
    }
    set allocationResourceTags(value) {
        this._allocationResourceTags = value;
    }
    resetAllocationResourceTags() {
        this._allocationResourceTags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allocationResourceTagsInput() {
        return this._allocationResourceTags;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get autoImport() {
        return this.getBooleanAttribute('auto_import');
    }
    set autoImport(value) {
        this._autoImport = value;
    }
    resetAutoImport() {
        this._autoImport = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoImportInput() {
        return this._autoImport;
    }
    get awsService() {
        return this.getStringAttribute('aws_service');
    }
    set awsService(value) {
        this._awsService = value;
    }
    resetAwsService() {
        this._awsService = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get awsServiceInput() {
        return this._awsService;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get ipamScopeId() {
        return this.getStringAttribute('ipam_scope_id');
    }
    set ipamScopeId(value) {
        this._ipamScopeId = value;
    }
    // Temporarily expose input value. Use with caution.
    get ipamScopeIdInput() {
        return this._ipamScopeId;
    }
    // ipam_scope_type - computed: true, optional: false, required: false
    get ipamScopeType() {
        return this.getStringAttribute('ipam_scope_type');
    }
    get locale() {
        return this.getStringAttribute('locale');
    }
    set locale(value) {
        this._locale = value;
    }
    resetLocale() {
        this._locale = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get localeInput() {
        return this._locale;
    }
    // pool_depth - computed: true, optional: false, required: false
    get poolDepth() {
        return this.getNumberAttribute('pool_depth');
    }
    get publiclyAdvertisable() {
        return this.getBooleanAttribute('publicly_advertisable');
    }
    set publiclyAdvertisable(value) {
        this._publiclyAdvertisable = value;
    }
    resetPubliclyAdvertisable() {
        this._publiclyAdvertisable = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get publiclyAdvertisableInput() {
        return this._publiclyAdvertisable;
    }
    get sourceIpamPoolId() {
        return this.getStringAttribute('source_ipam_pool_id');
    }
    set sourceIpamPoolId(value) {
        this._sourceIpamPoolId = value;
    }
    resetSourceIpamPoolId() {
        this._sourceIpamPoolId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceIpamPoolIdInput() {
        return this._sourceIpamPoolId;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            address_family: cdktf.stringToTerraform(this._addressFamily),
            allocation_default_netmask_length: cdktf.numberToTerraform(this._allocationDefaultNetmaskLength),
            allocation_max_netmask_length: cdktf.numberToTerraform(this._allocationMaxNetmaskLength),
            allocation_min_netmask_length: cdktf.numberToTerraform(this._allocationMinNetmaskLength),
            allocation_resource_tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._allocationResourceTags),
            auto_import: cdktf.booleanToTerraform(this._autoImport),
            aws_service: cdktf.stringToTerraform(this._awsService),
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            ipam_scope_id: cdktf.stringToTerraform(this._ipamScopeId),
            locale: cdktf.stringToTerraform(this._locale),
            publicly_advertisable: cdktf.booleanToTerraform(this._publiclyAdvertisable),
            source_ipam_pool_id: cdktf.stringToTerraform(this._sourceIpamPoolId),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            timeouts: vpcIpamPoolTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.VpcIpamPool = VpcIpamPool;
_b = JSII_RTTI_SYMBOL_1;
VpcIpamPool[_b] = { fqn: "@cdktf/provider-aws.vpcIpamPool.VpcIpamPool", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
VpcIpamPool.tfResourceType = "aws_vpc_ipam_pool";
