// @cdktf/provider-aws/lib/lightsail-instance-public-ports/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LightsailInstancePublicPorts = exports.LightsailInstancePublicPortsPortInfoList = exports.LightsailInstancePublicPortsPortInfoOutputReference = exports.lightsailInstancePublicPortsPortInfoToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function lightsailInstancePublicPortsPortInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cidrs: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.cidrs),
        from_port: cdktf.numberToTerraform(struct.fromPort),
        ipv6_cidrs: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.ipv6Cidrs),
        protocol: cdktf.stringToTerraform(struct.protocol),
        to_port: cdktf.numberToTerraform(struct.toPort),
    };
}
exports.lightsailInstancePublicPortsPortInfoToTerraform = lightsailInstancePublicPortsPortInfoToTerraform;
class LightsailInstancePublicPortsPortInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cidrs !== undefined) {
            hasAnyValues = true;
            internalValueResult.cidrs = this._cidrs;
        }
        if (this._fromPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.fromPort = this._fromPort;
        }
        if (this._ipv6Cidrs !== undefined) {
            hasAnyValues = true;
            internalValueResult.ipv6Cidrs = this._ipv6Cidrs;
        }
        if (this._protocol !== undefined) {
            hasAnyValues = true;
            internalValueResult.protocol = this._protocol;
        }
        if (this._toPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.toPort = this._toPort;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._cidrs = undefined;
            this._fromPort = undefined;
            this._ipv6Cidrs = undefined;
            this._protocol = undefined;
            this._toPort = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._cidrs = value.cidrs;
            this._fromPort = value.fromPort;
            this._ipv6Cidrs = value.ipv6Cidrs;
            this._protocol = value.protocol;
            this._toPort = value.toPort;
        }
    }
    get cidrs() {
        return cdktf.Fn.tolist(this.getListAttribute('cidrs'));
    }
    set cidrs(value) {
        this._cidrs = value;
    }
    resetCidrs() {
        this._cidrs = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cidrsInput() {
        return this._cidrs;
    }
    get fromPort() {
        return this.getNumberAttribute('from_port');
    }
    set fromPort(value) {
        this._fromPort = value;
    }
    // Temporarily expose input value. Use with caution.
    get fromPortInput() {
        return this._fromPort;
    }
    get ipv6Cidrs() {
        return cdktf.Fn.tolist(this.getListAttribute('ipv6_cidrs'));
    }
    set ipv6Cidrs(value) {
        this._ipv6Cidrs = value;
    }
    resetIpv6Cidrs() {
        this._ipv6Cidrs = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipv6CidrsInput() {
        return this._ipv6Cidrs;
    }
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    set protocol(value) {
        this._protocol = value;
    }
    // Temporarily expose input value. Use with caution.
    get protocolInput() {
        return this._protocol;
    }
    get toPort() {
        return this.getNumberAttribute('to_port');
    }
    set toPort(value) {
        this._toPort = value;
    }
    // Temporarily expose input value. Use with caution.
    get toPortInput() {
        return this._toPort;
    }
}
exports.LightsailInstancePublicPortsPortInfoOutputReference = LightsailInstancePublicPortsPortInfoOutputReference;
_a = JSII_RTTI_SYMBOL_1;
LightsailInstancePublicPortsPortInfoOutputReference[_a] = { fqn: "@cdktf/provider-aws.lightsailInstancePublicPorts.LightsailInstancePublicPortsPortInfoOutputReference", version: "10.0.12" };
class LightsailInstancePublicPortsPortInfoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new LightsailInstancePublicPortsPortInfoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.LightsailInstancePublicPortsPortInfoList = LightsailInstancePublicPortsPortInfoList;
_b = JSII_RTTI_SYMBOL_1;
LightsailInstancePublicPortsPortInfoList[_b] = { fqn: "@cdktf/provider-aws.lightsailInstancePublicPorts.LightsailInstancePublicPortsPortInfoList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/lightsail_instance_public_ports aws_lightsail_instance_public_ports}
*/
class LightsailInstancePublicPorts extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/lightsail_instance_public_ports aws_lightsail_instance_public_ports} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options LightsailInstancePublicPortsConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_lightsail_instance_public_ports',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // port_info - computed: false, optional: false, required: true
        this._portInfo = new LightsailInstancePublicPortsPortInfoList(this, "port_info", true);
        this._id = config.id;
        this._instanceName = config.instanceName;
        this._portInfo.internalValue = config.portInfo;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instanceName() {
        return this.getStringAttribute('instance_name');
    }
    set instanceName(value) {
        this._instanceName = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceNameInput() {
        return this._instanceName;
    }
    get portInfo() {
        return this._portInfo;
    }
    putPortInfo(value) {
        this._portInfo.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get portInfoInput() {
        return this._portInfo.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            instance_name: cdktf.stringToTerraform(this._instanceName),
            port_info: cdktf.listMapper(lightsailInstancePublicPortsPortInfoToTerraform, true)(this._portInfo.internalValue),
        };
    }
}
exports.LightsailInstancePublicPorts = LightsailInstancePublicPorts;
_c = JSII_RTTI_SYMBOL_1;
LightsailInstancePublicPorts[_c] = { fqn: "@cdktf/provider-aws.lightsailInstancePublicPorts.LightsailInstancePublicPorts", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
LightsailInstancePublicPorts.tfResourceType = "aws_lightsail_instance_public_ports";
