// @cdktf/provider-aws/lib/data-aws-rds-certificate/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsRdsCertificate = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/rds_certificate aws_rds_certificate}
*/
class DataAwsRdsCertificate extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/rds_certificate aws_rds_certificate} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsRdsCertificateConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_rds_certificate',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._latestValidTill = config.latestValidTill;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // certificate_type - computed: true, optional: false, required: false
    get certificateType() {
        return this.getStringAttribute('certificate_type');
    }
    // customer_override - computed: true, optional: false, required: false
    get customerOverride() {
        return this.getBooleanAttribute('customer_override');
    }
    // customer_override_valid_till - computed: true, optional: false, required: false
    get customerOverrideValidTill() {
        return this.getStringAttribute('customer_override_valid_till');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get latestValidTill() {
        return this.getBooleanAttribute('latest_valid_till');
    }
    set latestValidTill(value) {
        this._latestValidTill = value;
    }
    resetLatestValidTill() {
        this._latestValidTill = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get latestValidTillInput() {
        return this._latestValidTill;
    }
    // thumbprint - computed: true, optional: false, required: false
    get thumbprint() {
        return this.getStringAttribute('thumbprint');
    }
    // valid_from - computed: true, optional: false, required: false
    get validFrom() {
        return this.getStringAttribute('valid_from');
    }
    // valid_till - computed: true, optional: false, required: false
    get validTill() {
        return this.getStringAttribute('valid_till');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            latest_valid_till: cdktf.booleanToTerraform(this._latestValidTill),
        };
    }
}
exports.DataAwsRdsCertificate = DataAwsRdsCertificate;
_a = JSII_RTTI_SYMBOL_1;
DataAwsRdsCertificate[_a] = { fqn: "@cdktf/provider-aws.dataAwsRdsCertificate.DataAwsRdsCertificate", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsRdsCertificate.tfResourceType = "aws_rds_certificate";
