// @cdktf/provider-aws/lib/data-aws-ce-cost-category/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsCeCostCategory = exports.DataAwsCeCostCategorySplitChargeRuleList = exports.DataAwsCeCostCategorySplitChargeRuleOutputReference = exports.dataAwsCeCostCategorySplitChargeRuleToTerraform = exports.DataAwsCeCostCategorySplitChargeRuleParameterList = exports.DataAwsCeCostCategorySplitChargeRuleParameterOutputReference = exports.dataAwsCeCostCategorySplitChargeRuleParameterToTerraform = exports.DataAwsCeCostCategoryRuleList = exports.DataAwsCeCostCategoryRuleOutputReference = exports.dataAwsCeCostCategoryRuleToTerraform = exports.DataAwsCeCostCategoryRuleRuleList = exports.DataAwsCeCostCategoryRuleRuleOutputReference = exports.dataAwsCeCostCategoryRuleRuleToTerraform = exports.DataAwsCeCostCategoryRuleRuleTagsList = exports.DataAwsCeCostCategoryRuleRuleTagsOutputReference = exports.dataAwsCeCostCategoryRuleRuleTagsToTerraform = exports.DataAwsCeCostCategoryRuleRuleOrList = exports.DataAwsCeCostCategoryRuleRuleOrOutputReference = exports.dataAwsCeCostCategoryRuleRuleOrToTerraform = exports.DataAwsCeCostCategoryRuleRuleOrTagsList = exports.DataAwsCeCostCategoryRuleRuleOrTagsOutputReference = exports.dataAwsCeCostCategoryRuleRuleOrTagsToTerraform = exports.DataAwsCeCostCategoryRuleRuleOrDimensionList = exports.DataAwsCeCostCategoryRuleRuleOrDimensionOutputReference = exports.dataAwsCeCostCategoryRuleRuleOrDimensionToTerraform = exports.DataAwsCeCostCategoryRuleRuleOrCostCategoryList = exports.DataAwsCeCostCategoryRuleRuleOrCostCategoryOutputReference = exports.dataAwsCeCostCategoryRuleRuleOrCostCategoryToTerraform = exports.DataAwsCeCostCategoryRuleRuleNotList = exports.DataAwsCeCostCategoryRuleRuleNotOutputReference = exports.dataAwsCeCostCategoryRuleRuleNotToTerraform = exports.DataAwsCeCostCategoryRuleRuleNotTagsList = exports.DataAwsCeCostCategoryRuleRuleNotTagsOutputReference = exports.dataAwsCeCostCategoryRuleRuleNotTagsToTerraform = exports.DataAwsCeCostCategoryRuleRuleNotDimensionList = exports.DataAwsCeCostCategoryRuleRuleNotDimensionOutputReference = exports.dataAwsCeCostCategoryRuleRuleNotDimensionToTerraform = exports.DataAwsCeCostCategoryRuleRuleNotCostCategoryList = exports.DataAwsCeCostCategoryRuleRuleNotCostCategoryOutputReference = exports.dataAwsCeCostCategoryRuleRuleNotCostCategoryToTerraform = exports.DataAwsCeCostCategoryRuleRuleDimensionList = exports.DataAwsCeCostCategoryRuleRuleDimensionOutputReference = exports.dataAwsCeCostCategoryRuleRuleDimensionToTerraform = exports.DataAwsCeCostCategoryRuleRuleCostCategoryList = exports.DataAwsCeCostCategoryRuleRuleCostCategoryOutputReference = exports.dataAwsCeCostCategoryRuleRuleCostCategoryToTerraform = exports.DataAwsCeCostCategoryRuleRuleAndList = exports.DataAwsCeCostCategoryRuleRuleAndOutputReference = exports.dataAwsCeCostCategoryRuleRuleAndToTerraform = exports.DataAwsCeCostCategoryRuleRuleAndTagsList = exports.DataAwsCeCostCategoryRuleRuleAndTagsOutputReference = exports.dataAwsCeCostCategoryRuleRuleAndTagsToTerraform = exports.DataAwsCeCostCategoryRuleRuleAndDimensionList = exports.DataAwsCeCostCategoryRuleRuleAndDimensionOutputReference = exports.dataAwsCeCostCategoryRuleRuleAndDimensionToTerraform = exports.DataAwsCeCostCategoryRuleRuleAndCostCategoryList = exports.DataAwsCeCostCategoryRuleRuleAndCostCategoryOutputReference = exports.dataAwsCeCostCategoryRuleRuleAndCostCategoryToTerraform = exports.DataAwsCeCostCategoryRuleInheritedValueList = exports.DataAwsCeCostCategoryRuleInheritedValueOutputReference = exports.dataAwsCeCostCategoryRuleInheritedValueToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsCeCostCategoryRuleInheritedValueToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCeCostCategoryRuleInheritedValueToTerraform = dataAwsCeCostCategoryRuleInheritedValueToTerraform;
class DataAwsCeCostCategoryRuleInheritedValueOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // dimension_key - computed: true, optional: false, required: false
    get dimensionKey() {
        return this.getStringAttribute('dimension_key');
    }
    // dimension_name - computed: true, optional: false, required: false
    get dimensionName() {
        return this.getStringAttribute('dimension_name');
    }
}
exports.DataAwsCeCostCategoryRuleInheritedValueOutputReference = DataAwsCeCostCategoryRuleInheritedValueOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleInheritedValueOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleInheritedValueOutputReference", version: "10.0.12" };
class DataAwsCeCostCategoryRuleInheritedValueList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCeCostCategoryRuleInheritedValueOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCeCostCategoryRuleInheritedValueList = DataAwsCeCostCategoryRuleInheritedValueList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleInheritedValueList[_b] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleInheritedValueList", version: "10.0.12" };
function dataAwsCeCostCategoryRuleRuleAndCostCategoryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCeCostCategoryRuleRuleAndCostCategoryToTerraform = dataAwsCeCostCategoryRuleRuleAndCostCategoryToTerraform;
class DataAwsCeCostCategoryRuleRuleAndCostCategoryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // match_options - computed: true, optional: false, required: false
    get matchOptions() {
        return cdktf.Fn.tolist(this.getListAttribute('match_options'));
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
}
exports.DataAwsCeCostCategoryRuleRuleAndCostCategoryOutputReference = DataAwsCeCostCategoryRuleRuleAndCostCategoryOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleAndCostCategoryOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleAndCostCategoryOutputReference", version: "10.0.12" };
class DataAwsCeCostCategoryRuleRuleAndCostCategoryList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCeCostCategoryRuleRuleAndCostCategoryOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCeCostCategoryRuleRuleAndCostCategoryList = DataAwsCeCostCategoryRuleRuleAndCostCategoryList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleAndCostCategoryList[_d] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleAndCostCategoryList", version: "10.0.12" };
function dataAwsCeCostCategoryRuleRuleAndDimensionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCeCostCategoryRuleRuleAndDimensionToTerraform = dataAwsCeCostCategoryRuleRuleAndDimensionToTerraform;
class DataAwsCeCostCategoryRuleRuleAndDimensionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // match_options - computed: true, optional: false, required: false
    get matchOptions() {
        return cdktf.Fn.tolist(this.getListAttribute('match_options'));
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
}
exports.DataAwsCeCostCategoryRuleRuleAndDimensionOutputReference = DataAwsCeCostCategoryRuleRuleAndDimensionOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleAndDimensionOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleAndDimensionOutputReference", version: "10.0.12" };
class DataAwsCeCostCategoryRuleRuleAndDimensionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCeCostCategoryRuleRuleAndDimensionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCeCostCategoryRuleRuleAndDimensionList = DataAwsCeCostCategoryRuleRuleAndDimensionList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleAndDimensionList[_f] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleAndDimensionList", version: "10.0.12" };
function dataAwsCeCostCategoryRuleRuleAndTagsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCeCostCategoryRuleRuleAndTagsToTerraform = dataAwsCeCostCategoryRuleRuleAndTagsToTerraform;
class DataAwsCeCostCategoryRuleRuleAndTagsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // match_options - computed: true, optional: false, required: false
    get matchOptions() {
        return cdktf.Fn.tolist(this.getListAttribute('match_options'));
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
}
exports.DataAwsCeCostCategoryRuleRuleAndTagsOutputReference = DataAwsCeCostCategoryRuleRuleAndTagsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleAndTagsOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleAndTagsOutputReference", version: "10.0.12" };
class DataAwsCeCostCategoryRuleRuleAndTagsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCeCostCategoryRuleRuleAndTagsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCeCostCategoryRuleRuleAndTagsList = DataAwsCeCostCategoryRuleRuleAndTagsList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleAndTagsList[_h] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleAndTagsList", version: "10.0.12" };
function dataAwsCeCostCategoryRuleRuleAndToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCeCostCategoryRuleRuleAndToTerraform = dataAwsCeCostCategoryRuleRuleAndToTerraform;
class DataAwsCeCostCategoryRuleRuleAndOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // cost_category - computed: true, optional: false, required: false
        this._costCategory = new DataAwsCeCostCategoryRuleRuleAndCostCategoryList(this, "cost_category", false);
        // dimension - computed: true, optional: false, required: false
        this._dimension = new DataAwsCeCostCategoryRuleRuleAndDimensionList(this, "dimension", false);
        // tags - computed: true, optional: false, required: false
        this._tags = new DataAwsCeCostCategoryRuleRuleAndTagsList(this, "tags", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get costCategory() {
        return this._costCategory;
    }
    get dimension() {
        return this._dimension;
    }
    get tags() {
        return this._tags;
    }
}
exports.DataAwsCeCostCategoryRuleRuleAndOutputReference = DataAwsCeCostCategoryRuleRuleAndOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleAndOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleAndOutputReference", version: "10.0.12" };
class DataAwsCeCostCategoryRuleRuleAndList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCeCostCategoryRuleRuleAndOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCeCostCategoryRuleRuleAndList = DataAwsCeCostCategoryRuleRuleAndList;
_k = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleAndList[_k] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleAndList", version: "10.0.12" };
function dataAwsCeCostCategoryRuleRuleCostCategoryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCeCostCategoryRuleRuleCostCategoryToTerraform = dataAwsCeCostCategoryRuleRuleCostCategoryToTerraform;
class DataAwsCeCostCategoryRuleRuleCostCategoryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // match_options - computed: true, optional: false, required: false
    get matchOptions() {
        return cdktf.Fn.tolist(this.getListAttribute('match_options'));
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
}
exports.DataAwsCeCostCategoryRuleRuleCostCategoryOutputReference = DataAwsCeCostCategoryRuleRuleCostCategoryOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleCostCategoryOutputReference[_l] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleCostCategoryOutputReference", version: "10.0.12" };
class DataAwsCeCostCategoryRuleRuleCostCategoryList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCeCostCategoryRuleRuleCostCategoryOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCeCostCategoryRuleRuleCostCategoryList = DataAwsCeCostCategoryRuleRuleCostCategoryList;
_m = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleCostCategoryList[_m] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleCostCategoryList", version: "10.0.12" };
function dataAwsCeCostCategoryRuleRuleDimensionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCeCostCategoryRuleRuleDimensionToTerraform = dataAwsCeCostCategoryRuleRuleDimensionToTerraform;
class DataAwsCeCostCategoryRuleRuleDimensionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // match_options - computed: true, optional: false, required: false
    get matchOptions() {
        return cdktf.Fn.tolist(this.getListAttribute('match_options'));
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
}
exports.DataAwsCeCostCategoryRuleRuleDimensionOutputReference = DataAwsCeCostCategoryRuleRuleDimensionOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleDimensionOutputReference[_o] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleDimensionOutputReference", version: "10.0.12" };
class DataAwsCeCostCategoryRuleRuleDimensionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCeCostCategoryRuleRuleDimensionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCeCostCategoryRuleRuleDimensionList = DataAwsCeCostCategoryRuleRuleDimensionList;
_p = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleDimensionList[_p] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleDimensionList", version: "10.0.12" };
function dataAwsCeCostCategoryRuleRuleNotCostCategoryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCeCostCategoryRuleRuleNotCostCategoryToTerraform = dataAwsCeCostCategoryRuleRuleNotCostCategoryToTerraform;
class DataAwsCeCostCategoryRuleRuleNotCostCategoryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // match_options - computed: true, optional: false, required: false
    get matchOptions() {
        return cdktf.Fn.tolist(this.getListAttribute('match_options'));
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
}
exports.DataAwsCeCostCategoryRuleRuleNotCostCategoryOutputReference = DataAwsCeCostCategoryRuleRuleNotCostCategoryOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleNotCostCategoryOutputReference[_q] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleNotCostCategoryOutputReference", version: "10.0.12" };
class DataAwsCeCostCategoryRuleRuleNotCostCategoryList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCeCostCategoryRuleRuleNotCostCategoryOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCeCostCategoryRuleRuleNotCostCategoryList = DataAwsCeCostCategoryRuleRuleNotCostCategoryList;
_r = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleNotCostCategoryList[_r] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleNotCostCategoryList", version: "10.0.12" };
function dataAwsCeCostCategoryRuleRuleNotDimensionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCeCostCategoryRuleRuleNotDimensionToTerraform = dataAwsCeCostCategoryRuleRuleNotDimensionToTerraform;
class DataAwsCeCostCategoryRuleRuleNotDimensionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // match_options - computed: true, optional: false, required: false
    get matchOptions() {
        return cdktf.Fn.tolist(this.getListAttribute('match_options'));
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
}
exports.DataAwsCeCostCategoryRuleRuleNotDimensionOutputReference = DataAwsCeCostCategoryRuleRuleNotDimensionOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleNotDimensionOutputReference[_s] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleNotDimensionOutputReference", version: "10.0.12" };
class DataAwsCeCostCategoryRuleRuleNotDimensionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCeCostCategoryRuleRuleNotDimensionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCeCostCategoryRuleRuleNotDimensionList = DataAwsCeCostCategoryRuleRuleNotDimensionList;
_t = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleNotDimensionList[_t] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleNotDimensionList", version: "10.0.12" };
function dataAwsCeCostCategoryRuleRuleNotTagsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCeCostCategoryRuleRuleNotTagsToTerraform = dataAwsCeCostCategoryRuleRuleNotTagsToTerraform;
class DataAwsCeCostCategoryRuleRuleNotTagsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // match_options - computed: true, optional: false, required: false
    get matchOptions() {
        return cdktf.Fn.tolist(this.getListAttribute('match_options'));
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
}
exports.DataAwsCeCostCategoryRuleRuleNotTagsOutputReference = DataAwsCeCostCategoryRuleRuleNotTagsOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleNotTagsOutputReference[_u] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleNotTagsOutputReference", version: "10.0.12" };
class DataAwsCeCostCategoryRuleRuleNotTagsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCeCostCategoryRuleRuleNotTagsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCeCostCategoryRuleRuleNotTagsList = DataAwsCeCostCategoryRuleRuleNotTagsList;
_v = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleNotTagsList[_v] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleNotTagsList", version: "10.0.12" };
function dataAwsCeCostCategoryRuleRuleNotToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCeCostCategoryRuleRuleNotToTerraform = dataAwsCeCostCategoryRuleRuleNotToTerraform;
class DataAwsCeCostCategoryRuleRuleNotOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // cost_category - computed: true, optional: false, required: false
        this._costCategory = new DataAwsCeCostCategoryRuleRuleNotCostCategoryList(this, "cost_category", false);
        // dimension - computed: true, optional: false, required: false
        this._dimension = new DataAwsCeCostCategoryRuleRuleNotDimensionList(this, "dimension", false);
        // tags - computed: true, optional: false, required: false
        this._tags = new DataAwsCeCostCategoryRuleRuleNotTagsList(this, "tags", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get costCategory() {
        return this._costCategory;
    }
    get dimension() {
        return this._dimension;
    }
    get tags() {
        return this._tags;
    }
}
exports.DataAwsCeCostCategoryRuleRuleNotOutputReference = DataAwsCeCostCategoryRuleRuleNotOutputReference;
_w = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleNotOutputReference[_w] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleNotOutputReference", version: "10.0.12" };
class DataAwsCeCostCategoryRuleRuleNotList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCeCostCategoryRuleRuleNotOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCeCostCategoryRuleRuleNotList = DataAwsCeCostCategoryRuleRuleNotList;
_x = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleNotList[_x] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleNotList", version: "10.0.12" };
function dataAwsCeCostCategoryRuleRuleOrCostCategoryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCeCostCategoryRuleRuleOrCostCategoryToTerraform = dataAwsCeCostCategoryRuleRuleOrCostCategoryToTerraform;
class DataAwsCeCostCategoryRuleRuleOrCostCategoryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // match_options - computed: true, optional: false, required: false
    get matchOptions() {
        return cdktf.Fn.tolist(this.getListAttribute('match_options'));
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
}
exports.DataAwsCeCostCategoryRuleRuleOrCostCategoryOutputReference = DataAwsCeCostCategoryRuleRuleOrCostCategoryOutputReference;
_y = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleOrCostCategoryOutputReference[_y] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleOrCostCategoryOutputReference", version: "10.0.12" };
class DataAwsCeCostCategoryRuleRuleOrCostCategoryList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCeCostCategoryRuleRuleOrCostCategoryOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCeCostCategoryRuleRuleOrCostCategoryList = DataAwsCeCostCategoryRuleRuleOrCostCategoryList;
_z = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleOrCostCategoryList[_z] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleOrCostCategoryList", version: "10.0.12" };
function dataAwsCeCostCategoryRuleRuleOrDimensionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCeCostCategoryRuleRuleOrDimensionToTerraform = dataAwsCeCostCategoryRuleRuleOrDimensionToTerraform;
class DataAwsCeCostCategoryRuleRuleOrDimensionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // match_options - computed: true, optional: false, required: false
    get matchOptions() {
        return cdktf.Fn.tolist(this.getListAttribute('match_options'));
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
}
exports.DataAwsCeCostCategoryRuleRuleOrDimensionOutputReference = DataAwsCeCostCategoryRuleRuleOrDimensionOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleOrDimensionOutputReference[_0] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleOrDimensionOutputReference", version: "10.0.12" };
class DataAwsCeCostCategoryRuleRuleOrDimensionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCeCostCategoryRuleRuleOrDimensionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCeCostCategoryRuleRuleOrDimensionList = DataAwsCeCostCategoryRuleRuleOrDimensionList;
_1 = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleOrDimensionList[_1] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleOrDimensionList", version: "10.0.12" };
function dataAwsCeCostCategoryRuleRuleOrTagsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCeCostCategoryRuleRuleOrTagsToTerraform = dataAwsCeCostCategoryRuleRuleOrTagsToTerraform;
class DataAwsCeCostCategoryRuleRuleOrTagsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // match_options - computed: true, optional: false, required: false
    get matchOptions() {
        return cdktf.Fn.tolist(this.getListAttribute('match_options'));
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
}
exports.DataAwsCeCostCategoryRuleRuleOrTagsOutputReference = DataAwsCeCostCategoryRuleRuleOrTagsOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleOrTagsOutputReference[_2] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleOrTagsOutputReference", version: "10.0.12" };
class DataAwsCeCostCategoryRuleRuleOrTagsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCeCostCategoryRuleRuleOrTagsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCeCostCategoryRuleRuleOrTagsList = DataAwsCeCostCategoryRuleRuleOrTagsList;
_3 = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleOrTagsList[_3] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleOrTagsList", version: "10.0.12" };
function dataAwsCeCostCategoryRuleRuleOrToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCeCostCategoryRuleRuleOrToTerraform = dataAwsCeCostCategoryRuleRuleOrToTerraform;
class DataAwsCeCostCategoryRuleRuleOrOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // cost_category - computed: true, optional: false, required: false
        this._costCategory = new DataAwsCeCostCategoryRuleRuleOrCostCategoryList(this, "cost_category", false);
        // dimension - computed: true, optional: false, required: false
        this._dimension = new DataAwsCeCostCategoryRuleRuleOrDimensionList(this, "dimension", false);
        // tags - computed: true, optional: false, required: false
        this._tags = new DataAwsCeCostCategoryRuleRuleOrTagsList(this, "tags", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get costCategory() {
        return this._costCategory;
    }
    get dimension() {
        return this._dimension;
    }
    get tags() {
        return this._tags;
    }
}
exports.DataAwsCeCostCategoryRuleRuleOrOutputReference = DataAwsCeCostCategoryRuleRuleOrOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleOrOutputReference[_4] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleOrOutputReference", version: "10.0.12" };
class DataAwsCeCostCategoryRuleRuleOrList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCeCostCategoryRuleRuleOrOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCeCostCategoryRuleRuleOrList = DataAwsCeCostCategoryRuleRuleOrList;
_5 = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleOrList[_5] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleOrList", version: "10.0.12" };
function dataAwsCeCostCategoryRuleRuleTagsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCeCostCategoryRuleRuleTagsToTerraform = dataAwsCeCostCategoryRuleRuleTagsToTerraform;
class DataAwsCeCostCategoryRuleRuleTagsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // match_options - computed: true, optional: false, required: false
    get matchOptions() {
        return cdktf.Fn.tolist(this.getListAttribute('match_options'));
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
}
exports.DataAwsCeCostCategoryRuleRuleTagsOutputReference = DataAwsCeCostCategoryRuleRuleTagsOutputReference;
_6 = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleTagsOutputReference[_6] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleTagsOutputReference", version: "10.0.12" };
class DataAwsCeCostCategoryRuleRuleTagsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCeCostCategoryRuleRuleTagsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCeCostCategoryRuleRuleTagsList = DataAwsCeCostCategoryRuleRuleTagsList;
_7 = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleTagsList[_7] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleTagsList", version: "10.0.12" };
function dataAwsCeCostCategoryRuleRuleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCeCostCategoryRuleRuleToTerraform = dataAwsCeCostCategoryRuleRuleToTerraform;
class DataAwsCeCostCategoryRuleRuleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // and - computed: true, optional: false, required: false
        this._and = new DataAwsCeCostCategoryRuleRuleAndList(this, "and", true);
        // cost_category - computed: true, optional: false, required: false
        this._costCategory = new DataAwsCeCostCategoryRuleRuleCostCategoryList(this, "cost_category", false);
        // dimension - computed: true, optional: false, required: false
        this._dimension = new DataAwsCeCostCategoryRuleRuleDimensionList(this, "dimension", false);
        // not - computed: true, optional: false, required: false
        this._not = new DataAwsCeCostCategoryRuleRuleNotList(this, "not", false);
        // or - computed: true, optional: false, required: false
        this._or = new DataAwsCeCostCategoryRuleRuleOrList(this, "or", true);
        // tags - computed: true, optional: false, required: false
        this._tags = new DataAwsCeCostCategoryRuleRuleTagsList(this, "tags", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get and() {
        return this._and;
    }
    get costCategory() {
        return this._costCategory;
    }
    get dimension() {
        return this._dimension;
    }
    get not() {
        return this._not;
    }
    get or() {
        return this._or;
    }
    get tags() {
        return this._tags;
    }
}
exports.DataAwsCeCostCategoryRuleRuleOutputReference = DataAwsCeCostCategoryRuleRuleOutputReference;
_8 = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleOutputReference[_8] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleOutputReference", version: "10.0.12" };
class DataAwsCeCostCategoryRuleRuleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCeCostCategoryRuleRuleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCeCostCategoryRuleRuleList = DataAwsCeCostCategoryRuleRuleList;
_9 = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleRuleList[_9] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleRuleList", version: "10.0.12" };
function dataAwsCeCostCategoryRuleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCeCostCategoryRuleToTerraform = dataAwsCeCostCategoryRuleToTerraform;
class DataAwsCeCostCategoryRuleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // inherited_value - computed: true, optional: false, required: false
        this._inheritedValue = new DataAwsCeCostCategoryRuleInheritedValueList(this, "inherited_value", false);
        // rule - computed: true, optional: false, required: false
        this._rule = new DataAwsCeCostCategoryRuleRuleList(this, "rule", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get inheritedValue() {
        return this._inheritedValue;
    }
    get rule() {
        return this._rule;
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
    // value - computed: true, optional: false, required: false
    get value() {
        return this.getStringAttribute('value');
    }
}
exports.DataAwsCeCostCategoryRuleOutputReference = DataAwsCeCostCategoryRuleOutputReference;
_10 = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleOutputReference[_10] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleOutputReference", version: "10.0.12" };
class DataAwsCeCostCategoryRuleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCeCostCategoryRuleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCeCostCategoryRuleList = DataAwsCeCostCategoryRuleList;
_11 = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategoryRuleList[_11] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategoryRuleList", version: "10.0.12" };
function dataAwsCeCostCategorySplitChargeRuleParameterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCeCostCategorySplitChargeRuleParameterToTerraform = dataAwsCeCostCategorySplitChargeRuleParameterToTerraform;
class DataAwsCeCostCategorySplitChargeRuleParameterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
}
exports.DataAwsCeCostCategorySplitChargeRuleParameterOutputReference = DataAwsCeCostCategorySplitChargeRuleParameterOutputReference;
_12 = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategorySplitChargeRuleParameterOutputReference[_12] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategorySplitChargeRuleParameterOutputReference", version: "10.0.12" };
class DataAwsCeCostCategorySplitChargeRuleParameterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCeCostCategorySplitChargeRuleParameterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCeCostCategorySplitChargeRuleParameterList = DataAwsCeCostCategorySplitChargeRuleParameterList;
_13 = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategorySplitChargeRuleParameterList[_13] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategorySplitChargeRuleParameterList", version: "10.0.12" };
function dataAwsCeCostCategorySplitChargeRuleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsCeCostCategorySplitChargeRuleToTerraform = dataAwsCeCostCategorySplitChargeRuleToTerraform;
class DataAwsCeCostCategorySplitChargeRuleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // parameter - computed: true, optional: false, required: false
        this._parameter = new DataAwsCeCostCategorySplitChargeRuleParameterList(this, "parameter", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // method - computed: true, optional: false, required: false
    get method() {
        return this.getStringAttribute('method');
    }
    get parameter() {
        return this._parameter;
    }
    // source - computed: true, optional: false, required: false
    get source() {
        return this.getStringAttribute('source');
    }
    // targets - computed: true, optional: false, required: false
    get targets() {
        return cdktf.Fn.tolist(this.getListAttribute('targets'));
    }
}
exports.DataAwsCeCostCategorySplitChargeRuleOutputReference = DataAwsCeCostCategorySplitChargeRuleOutputReference;
_14 = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategorySplitChargeRuleOutputReference[_14] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategorySplitChargeRuleOutputReference", version: "10.0.12" };
class DataAwsCeCostCategorySplitChargeRuleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsCeCostCategorySplitChargeRuleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsCeCostCategorySplitChargeRuleList = DataAwsCeCostCategorySplitChargeRuleList;
_15 = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategorySplitChargeRuleList[_15] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategorySplitChargeRuleList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/ce_cost_category aws_ce_cost_category}
*/
class DataAwsCeCostCategory extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/ce_cost_category aws_ce_cost_category} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsCeCostCategoryConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ce_cost_category',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // rule - computed: true, optional: false, required: false
        this._rule = new DataAwsCeCostCategoryRuleList(this, "rule", true);
        // split_charge_rule - computed: true, optional: false, required: false
        this._splitChargeRule = new DataAwsCeCostCategorySplitChargeRuleList(this, "split_charge_rule", true);
        this._costCategoryArn = config.costCategoryArn;
        this._id = config.id;
        this._tags = config.tags;
    }
    get costCategoryArn() {
        return this.getStringAttribute('cost_category_arn');
    }
    set costCategoryArn(value) {
        this._costCategoryArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get costCategoryArnInput() {
        return this._costCategoryArn;
    }
    // effective_end - computed: true, optional: false, required: false
    get effectiveEnd() {
        return this.getStringAttribute('effective_end');
    }
    // effective_start - computed: true, optional: false, required: false
    get effectiveStart() {
        return this.getStringAttribute('effective_start');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get rule() {
        return this._rule;
    }
    // rule_version - computed: true, optional: false, required: false
    get ruleVersion() {
        return this.getStringAttribute('rule_version');
    }
    get splitChargeRule() {
        return this._splitChargeRule;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cost_category_arn: cdktf.stringToTerraform(this._costCategoryArn),
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
}
exports.DataAwsCeCostCategory = DataAwsCeCostCategory;
_16 = JSII_RTTI_SYMBOL_1;
DataAwsCeCostCategory[_16] = { fqn: "@cdktf/provider-aws.dataAwsCeCostCategory.DataAwsCeCostCategory", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsCeCostCategory.tfResourceType = "aws_ce_cost_category";
