// @cdktf/provider-aws/lib/iam-account-alias/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IamAccountAlias = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/iam_account_alias aws_iam_account_alias}
*/
class IamAccountAlias extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/iam_account_alias aws_iam_account_alias} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options IamAccountAliasConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iam_account_alias',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._accountAlias = config.accountAlias;
        this._id = config.id;
    }
    get accountAlias() {
        return this.getStringAttribute('account_alias');
    }
    set accountAlias(value) {
        this._accountAlias = value;
    }
    // Temporarily expose input value. Use with caution.
    get accountAliasInput() {
        return this._accountAlias;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            account_alias: cdktf.stringToTerraform(this._accountAlias),
            id: cdktf.stringToTerraform(this._id),
        };
    }
}
exports.IamAccountAlias = IamAccountAlias;
_a = JSII_RTTI_SYMBOL_1;
IamAccountAlias[_a] = { fqn: "@cdktf/provider-aws.iamAccountAlias.IamAccountAlias", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
IamAccountAlias.tfResourceType = "aws_iam_account_alias";
