// @cdktf/provider-aws/lib/glue-workflow/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlueWorkflow = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/glue_workflow aws_glue_workflow}
*/
class GlueWorkflow extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/glue_workflow aws_glue_workflow} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options GlueWorkflowConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'aws_glue_workflow',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._defaultRunProperties = config.defaultRunProperties;
        this._description = config.description;
        this._id = config.id;
        this._maxConcurrentRuns = config.maxConcurrentRuns;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get defaultRunProperties() {
        return this.getStringMapAttribute('default_run_properties');
    }
    set defaultRunProperties(value) {
        this._defaultRunProperties = value;
    }
    resetDefaultRunProperties() {
        this._defaultRunProperties = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultRunPropertiesInput() {
        return this._defaultRunProperties;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get maxConcurrentRuns() {
        return this.getNumberAttribute('max_concurrent_runs');
    }
    set maxConcurrentRuns(value) {
        this._maxConcurrentRuns = value;
    }
    resetMaxConcurrentRuns() {
        this._maxConcurrentRuns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxConcurrentRunsInput() {
        return this._maxConcurrentRuns;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            default_run_properties: cdktf.hashMapper(cdktf.stringToTerraform)(this._defaultRunProperties),
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            max_concurrent_runs: cdktf.numberToTerraform(this._maxConcurrentRuns),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
        };
    }
}
exports.GlueWorkflow = GlueWorkflow;
_a = JSII_RTTI_SYMBOL_1;
GlueWorkflow[_a] = { fqn: "@cdktf/provider-aws.glueWorkflow.GlueWorkflow", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
GlueWorkflow.tfResourceType = "aws_glue_workflow";
