// @cdktf/provider-aws/lib/kms-grant/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.KmsGrant = exports.KmsGrantConstraintsList = exports.KmsGrantConstraintsOutputReference = exports.kmsGrantConstraintsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function kmsGrantConstraintsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        encryption_context_equals: cdktf.hashMapper(cdktf.stringToTerraform)(struct.encryptionContextEquals),
        encryption_context_subset: cdktf.hashMapper(cdktf.stringToTerraform)(struct.encryptionContextSubset),
    };
}
exports.kmsGrantConstraintsToTerraform = kmsGrantConstraintsToTerraform;
class KmsGrantConstraintsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._encryptionContextEquals !== undefined) {
            hasAnyValues = true;
            internalValueResult.encryptionContextEquals = this._encryptionContextEquals;
        }
        if (this._encryptionContextSubset !== undefined) {
            hasAnyValues = true;
            internalValueResult.encryptionContextSubset = this._encryptionContextSubset;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._encryptionContextEquals = undefined;
            this._encryptionContextSubset = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._encryptionContextEquals = value.encryptionContextEquals;
            this._encryptionContextSubset = value.encryptionContextSubset;
        }
    }
    get encryptionContextEquals() {
        return this.getStringMapAttribute('encryption_context_equals');
    }
    set encryptionContextEquals(value) {
        this._encryptionContextEquals = value;
    }
    resetEncryptionContextEquals() {
        this._encryptionContextEquals = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encryptionContextEqualsInput() {
        return this._encryptionContextEquals;
    }
    get encryptionContextSubset() {
        return this.getStringMapAttribute('encryption_context_subset');
    }
    set encryptionContextSubset(value) {
        this._encryptionContextSubset = value;
    }
    resetEncryptionContextSubset() {
        this._encryptionContextSubset = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encryptionContextSubsetInput() {
        return this._encryptionContextSubset;
    }
}
exports.KmsGrantConstraintsOutputReference = KmsGrantConstraintsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
KmsGrantConstraintsOutputReference[_a] = { fqn: "@cdktf/provider-aws.kmsGrant.KmsGrantConstraintsOutputReference", version: "10.0.12" };
class KmsGrantConstraintsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new KmsGrantConstraintsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.KmsGrantConstraintsList = KmsGrantConstraintsList;
_b = JSII_RTTI_SYMBOL_1;
KmsGrantConstraintsList[_b] = { fqn: "@cdktf/provider-aws.kmsGrant.KmsGrantConstraintsList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/kms_grant aws_kms_grant}
*/
class KmsGrant extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/kms_grant aws_kms_grant} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options KmsGrantConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_kms_grant',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // constraints - computed: false, optional: true, required: false
        this._constraints = new KmsGrantConstraintsList(this, "constraints", true);
        this._grantCreationTokens = config.grantCreationTokens;
        this._granteePrincipal = config.granteePrincipal;
        this._id = config.id;
        this._keyId = config.keyId;
        this._name = config.name;
        this._operations = config.operations;
        this._retireOnDelete = config.retireOnDelete;
        this._retiringPrincipal = config.retiringPrincipal;
        this._constraints.internalValue = config.constraints;
    }
    get grantCreationTokens() {
        return cdktf.Fn.tolist(this.getListAttribute('grant_creation_tokens'));
    }
    set grantCreationTokens(value) {
        this._grantCreationTokens = value;
    }
    resetGrantCreationTokens() {
        this._grantCreationTokens = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get grantCreationTokensInput() {
        return this._grantCreationTokens;
    }
    // grant_id - computed: true, optional: false, required: false
    get grantId() {
        return this.getStringAttribute('grant_id');
    }
    // grant_token - computed: true, optional: false, required: false
    get grantToken() {
        return this.getStringAttribute('grant_token');
    }
    get granteePrincipal() {
        return this.getStringAttribute('grantee_principal');
    }
    set granteePrincipal(value) {
        this._granteePrincipal = value;
    }
    // Temporarily expose input value. Use with caution.
    get granteePrincipalInput() {
        return this._granteePrincipal;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get keyId() {
        return this.getStringAttribute('key_id');
    }
    set keyId(value) {
        this._keyId = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyIdInput() {
        return this._keyId;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get operations() {
        return cdktf.Fn.tolist(this.getListAttribute('operations'));
    }
    set operations(value) {
        this._operations = value;
    }
    // Temporarily expose input value. Use with caution.
    get operationsInput() {
        return this._operations;
    }
    get retireOnDelete() {
        return this.getBooleanAttribute('retire_on_delete');
    }
    set retireOnDelete(value) {
        this._retireOnDelete = value;
    }
    resetRetireOnDelete() {
        this._retireOnDelete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get retireOnDeleteInput() {
        return this._retireOnDelete;
    }
    get retiringPrincipal() {
        return this.getStringAttribute('retiring_principal');
    }
    set retiringPrincipal(value) {
        this._retiringPrincipal = value;
    }
    resetRetiringPrincipal() {
        this._retiringPrincipal = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get retiringPrincipalInput() {
        return this._retiringPrincipal;
    }
    get constraints() {
        return this._constraints;
    }
    putConstraints(value) {
        this._constraints.internalValue = value;
    }
    resetConstraints() {
        this._constraints.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get constraintsInput() {
        return this._constraints.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            grant_creation_tokens: cdktf.listMapper(cdktf.stringToTerraform, false)(this._grantCreationTokens),
            grantee_principal: cdktf.stringToTerraform(this._granteePrincipal),
            id: cdktf.stringToTerraform(this._id),
            key_id: cdktf.stringToTerraform(this._keyId),
            name: cdktf.stringToTerraform(this._name),
            operations: cdktf.listMapper(cdktf.stringToTerraform, false)(this._operations),
            retire_on_delete: cdktf.booleanToTerraform(this._retireOnDelete),
            retiring_principal: cdktf.stringToTerraform(this._retiringPrincipal),
            constraints: cdktf.listMapper(kmsGrantConstraintsToTerraform, true)(this._constraints.internalValue),
        };
    }
}
exports.KmsGrant = KmsGrant;
_c = JSII_RTTI_SYMBOL_1;
KmsGrant[_c] = { fqn: "@cdktf/provider-aws.kmsGrant.KmsGrant", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
KmsGrant.tfResourceType = "aws_kms_grant";
