// @cdktf/provider-aws/lib/iam-virtual-mfa-device/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IamVirtualMfaDevice = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/iam_virtual_mfa_device aws_iam_virtual_mfa_device}
*/
class IamVirtualMfaDevice extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/iam_virtual_mfa_device aws_iam_virtual_mfa_device} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options IamVirtualMfaDeviceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iam_virtual_mfa_device',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._path = config.path;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._virtualMfaDeviceName = config.virtualMfaDeviceName;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // base_32_string_seed - computed: true, optional: false, required: false
    get base32StringSeed() {
        return this.getStringAttribute('base_32_string_seed');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    resetPath() {
        this._path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
    // qr_code_png - computed: true, optional: false, required: false
    get qrCodePng() {
        return this.getStringAttribute('qr_code_png');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get virtualMfaDeviceName() {
        return this.getStringAttribute('virtual_mfa_device_name');
    }
    set virtualMfaDeviceName(value) {
        this._virtualMfaDeviceName = value;
    }
    // Temporarily expose input value. Use with caution.
    get virtualMfaDeviceNameInput() {
        return this._virtualMfaDeviceName;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            path: cdktf.stringToTerraform(this._path),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            virtual_mfa_device_name: cdktf.stringToTerraform(this._virtualMfaDeviceName),
        };
    }
}
exports.IamVirtualMfaDevice = IamVirtualMfaDevice;
_a = JSII_RTTI_SYMBOL_1;
IamVirtualMfaDevice[_a] = { fqn: "@cdktf/provider-aws.iamVirtualMfaDevice.IamVirtualMfaDevice", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
IamVirtualMfaDevice.tfResourceType = "aws_iam_virtual_mfa_device";
