// @cdktf/provider-aws/lib/mskconnect-worker-configuration/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MskconnectWorkerConfiguration = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/mskconnect_worker_configuration aws_mskconnect_worker_configuration}
*/
class MskconnectWorkerConfiguration extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/mskconnect_worker_configuration aws_mskconnect_worker_configuration} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options MskconnectWorkerConfigurationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_mskconnect_worker_configuration',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._description = config.description;
        this._id = config.id;
        this._name = config.name;
        this._propertiesFileContent = config.propertiesFileContent;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // latest_revision - computed: true, optional: false, required: false
    get latestRevision() {
        return this.getNumberAttribute('latest_revision');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get propertiesFileContent() {
        return this.getStringAttribute('properties_file_content');
    }
    set propertiesFileContent(value) {
        this._propertiesFileContent = value;
    }
    // Temporarily expose input value. Use with caution.
    get propertiesFileContentInput() {
        return this._propertiesFileContent;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            properties_file_content: cdktf.stringToTerraform(this._propertiesFileContent),
        };
    }
}
exports.MskconnectWorkerConfiguration = MskconnectWorkerConfiguration;
_a = JSII_RTTI_SYMBOL_1;
MskconnectWorkerConfiguration[_a] = { fqn: "@cdktf/provider-aws.mskconnectWorkerConfiguration.MskconnectWorkerConfiguration", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
MskconnectWorkerConfiguration.tfResourceType = "aws_mskconnect_worker_configuration";
