// @cdktf/provider-aws/lib/vpc-endpoint-security-group-association/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VpcEndpointSecurityGroupAssociation = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/vpc_endpoint_security_group_association aws_vpc_endpoint_security_group_association}
*/
class VpcEndpointSecurityGroupAssociation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/vpc_endpoint_security_group_association aws_vpc_endpoint_security_group_association} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options VpcEndpointSecurityGroupAssociationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_vpc_endpoint_security_group_association',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._replaceDefaultAssociation = config.replaceDefaultAssociation;
        this._securityGroupId = config.securityGroupId;
        this._vpcEndpointId = config.vpcEndpointId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get replaceDefaultAssociation() {
        return this.getBooleanAttribute('replace_default_association');
    }
    set replaceDefaultAssociation(value) {
        this._replaceDefaultAssociation = value;
    }
    resetReplaceDefaultAssociation() {
        this._replaceDefaultAssociation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get replaceDefaultAssociationInput() {
        return this._replaceDefaultAssociation;
    }
    get securityGroupId() {
        return this.getStringAttribute('security_group_id');
    }
    set securityGroupId(value) {
        this._securityGroupId = value;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdInput() {
        return this._securityGroupId;
    }
    get vpcEndpointId() {
        return this.getStringAttribute('vpc_endpoint_id');
    }
    set vpcEndpointId(value) {
        this._vpcEndpointId = value;
    }
    // Temporarily expose input value. Use with caution.
    get vpcEndpointIdInput() {
        return this._vpcEndpointId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            replace_default_association: cdktf.booleanToTerraform(this._replaceDefaultAssociation),
            security_group_id: cdktf.stringToTerraform(this._securityGroupId),
            vpc_endpoint_id: cdktf.stringToTerraform(this._vpcEndpointId),
        };
    }
}
exports.VpcEndpointSecurityGroupAssociation = VpcEndpointSecurityGroupAssociation;
_a = JSII_RTTI_SYMBOL_1;
VpcEndpointSecurityGroupAssociation[_a] = { fqn: "@cdktf/provider-aws.vpcEndpointSecurityGroupAssociation.VpcEndpointSecurityGroupAssociation", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
VpcEndpointSecurityGroupAssociation.tfResourceType = "aws_vpc_endpoint_security_group_association";
