// @cdktf/provider-aws/lib/data-aws-backup-selection/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsBackupSelection = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/backup_selection aws_backup_selection}
*/
class DataAwsBackupSelection extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/backup_selection aws_backup_selection} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsBackupSelectionConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_backup_selection',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._planId = config.planId;
        this._selectionId = config.selectionId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // iam_role_arn - computed: true, optional: false, required: false
    get iamRoleArn() {
        return this.getStringAttribute('iam_role_arn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get planId() {
        return this.getStringAttribute('plan_id');
    }
    set planId(value) {
        this._planId = value;
    }
    // Temporarily expose input value. Use with caution.
    get planIdInput() {
        return this._planId;
    }
    // resources - computed: true, optional: false, required: false
    get resources() {
        return cdktf.Fn.tolist(this.getListAttribute('resources'));
    }
    get selectionId() {
        return this.getStringAttribute('selection_id');
    }
    set selectionId(value) {
        this._selectionId = value;
    }
    // Temporarily expose input value. Use with caution.
    get selectionIdInput() {
        return this._selectionId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            plan_id: cdktf.stringToTerraform(this._planId),
            selection_id: cdktf.stringToTerraform(this._selectionId),
        };
    }
}
exports.DataAwsBackupSelection = DataAwsBackupSelection;
_a = JSII_RTTI_SYMBOL_1;
DataAwsBackupSelection[_a] = { fqn: "@cdktf/provider-aws.dataAwsBackupSelection.DataAwsBackupSelection", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsBackupSelection.tfResourceType = "aws_backup_selection";
