// @cdktf/provider-aws/lib/grafana-workspace-api-key/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrafanaWorkspaceApiKey = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/grafana_workspace_api_key aws_grafana_workspace_api_key}
*/
class GrafanaWorkspaceApiKey extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/grafana_workspace_api_key aws_grafana_workspace_api_key} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options GrafanaWorkspaceApiKeyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_grafana_workspace_api_key',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._keyName = config.keyName;
        this._keyRole = config.keyRole;
        this._secondsToLive = config.secondsToLive;
        this._workspaceId = config.workspaceId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    get keyName() {
        return this.getStringAttribute('key_name');
    }
    set keyName(value) {
        this._keyName = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyNameInput() {
        return this._keyName;
    }
    get keyRole() {
        return this.getStringAttribute('key_role');
    }
    set keyRole(value) {
        this._keyRole = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyRoleInput() {
        return this._keyRole;
    }
    get secondsToLive() {
        return this.getNumberAttribute('seconds_to_live');
    }
    set secondsToLive(value) {
        this._secondsToLive = value;
    }
    // Temporarily expose input value. Use with caution.
    get secondsToLiveInput() {
        return this._secondsToLive;
    }
    get workspaceId() {
        return this.getStringAttribute('workspace_id');
    }
    set workspaceId(value) {
        this._workspaceId = value;
    }
    // Temporarily expose input value. Use with caution.
    get workspaceIdInput() {
        return this._workspaceId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            key_name: cdktf.stringToTerraform(this._keyName),
            key_role: cdktf.stringToTerraform(this._keyRole),
            seconds_to_live: cdktf.numberToTerraform(this._secondsToLive),
            workspace_id: cdktf.stringToTerraform(this._workspaceId),
        };
    }
}
exports.GrafanaWorkspaceApiKey = GrafanaWorkspaceApiKey;
_a = JSII_RTTI_SYMBOL_1;
GrafanaWorkspaceApiKey[_a] = { fqn: "@cdktf/provider-aws.grafanaWorkspaceApiKey.GrafanaWorkspaceApiKey", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
GrafanaWorkspaceApiKey.tfResourceType = "aws_grafana_workspace_api_key";
