// @cdktf/provider-aws/lib/data-aws-rds-engine-version/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsRdsEngineVersion = exports.DataAwsRdsEngineVersionFilterList = exports.DataAwsRdsEngineVersionFilterOutputReference = exports.dataAwsRdsEngineVersionFilterToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsRdsEngineVersionFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.values),
    };
}
exports.dataAwsRdsEngineVersionFilterToTerraform = dataAwsRdsEngineVersionFilterToTerraform;
class DataAwsRdsEngineVersionFilterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._values !== undefined) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._values = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._values = value.values;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
    set values(value) {
        this._values = value;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
}
exports.DataAwsRdsEngineVersionFilterOutputReference = DataAwsRdsEngineVersionFilterOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsRdsEngineVersionFilterOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsRdsEngineVersion.DataAwsRdsEngineVersionFilterOutputReference", version: "10.0.12" };
class DataAwsRdsEngineVersionFilterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsRdsEngineVersionFilterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsRdsEngineVersionFilterList = DataAwsRdsEngineVersionFilterList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsRdsEngineVersionFilterList[_b] = { fqn: "@cdktf/provider-aws.dataAwsRdsEngineVersion.DataAwsRdsEngineVersionFilterList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/rds_engine_version aws_rds_engine_version}
*/
class DataAwsRdsEngineVersion extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/rds_engine_version aws_rds_engine_version} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsRdsEngineVersionConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_rds_engine_version',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // filter - computed: false, optional: true, required: false
        this._filter = new DataAwsRdsEngineVersionFilterList(this, "filter", true);
        this._defaultOnly = config.defaultOnly;
        this._engine = config.engine;
        this._id = config.id;
        this._includeAll = config.includeAll;
        this._parameterGroupFamily = config.parameterGroupFamily;
        this._preferredVersions = config.preferredVersions;
        this._version = config.version;
        this._filter.internalValue = config.filter;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // default_character_set - computed: true, optional: false, required: false
    get defaultCharacterSet() {
        return this.getStringAttribute('default_character_set');
    }
    get defaultOnly() {
        return this.getBooleanAttribute('default_only');
    }
    set defaultOnly(value) {
        this._defaultOnly = value;
    }
    resetDefaultOnly() {
        this._defaultOnly = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultOnlyInput() {
        return this._defaultOnly;
    }
    get engine() {
        return this.getStringAttribute('engine');
    }
    set engine(value) {
        this._engine = value;
    }
    // Temporarily expose input value. Use with caution.
    get engineInput() {
        return this._engine;
    }
    // engine_description - computed: true, optional: false, required: false
    get engineDescription() {
        return this.getStringAttribute('engine_description');
    }
    // exportable_log_types - computed: true, optional: false, required: false
    get exportableLogTypes() {
        return cdktf.Fn.tolist(this.getListAttribute('exportable_log_types'));
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get includeAll() {
        return this.getBooleanAttribute('include_all');
    }
    set includeAll(value) {
        this._includeAll = value;
    }
    resetIncludeAll() {
        this._includeAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includeAllInput() {
        return this._includeAll;
    }
    get parameterGroupFamily() {
        return this.getStringAttribute('parameter_group_family');
    }
    set parameterGroupFamily(value) {
        this._parameterGroupFamily = value;
    }
    resetParameterGroupFamily() {
        this._parameterGroupFamily = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parameterGroupFamilyInput() {
        return this._parameterGroupFamily;
    }
    get preferredVersions() {
        return this.getListAttribute('preferred_versions');
    }
    set preferredVersions(value) {
        this._preferredVersions = value;
    }
    resetPreferredVersions() {
        this._preferredVersions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get preferredVersionsInput() {
        return this._preferredVersions;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    // supported_character_sets - computed: true, optional: false, required: false
    get supportedCharacterSets() {
        return cdktf.Fn.tolist(this.getListAttribute('supported_character_sets'));
    }
    // supported_feature_names - computed: true, optional: false, required: false
    get supportedFeatureNames() {
        return cdktf.Fn.tolist(this.getListAttribute('supported_feature_names'));
    }
    // supported_modes - computed: true, optional: false, required: false
    get supportedModes() {
        return cdktf.Fn.tolist(this.getListAttribute('supported_modes'));
    }
    // supported_timezones - computed: true, optional: false, required: false
    get supportedTimezones() {
        return cdktf.Fn.tolist(this.getListAttribute('supported_timezones'));
    }
    // supports_global_databases - computed: true, optional: false, required: false
    get supportsGlobalDatabases() {
        return this.getBooleanAttribute('supports_global_databases');
    }
    // supports_log_exports_to_cloudwatch - computed: true, optional: false, required: false
    get supportsLogExportsToCloudwatch() {
        return this.getBooleanAttribute('supports_log_exports_to_cloudwatch');
    }
    // supports_parallel_query - computed: true, optional: false, required: false
    get supportsParallelQuery() {
        return this.getBooleanAttribute('supports_parallel_query');
    }
    // supports_read_replica - computed: true, optional: false, required: false
    get supportsReadReplica() {
        return this.getBooleanAttribute('supports_read_replica');
    }
    // valid_upgrade_targets - computed: true, optional: false, required: false
    get validUpgradeTargets() {
        return cdktf.Fn.tolist(this.getListAttribute('valid_upgrade_targets'));
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    resetVersion() {
        this._version = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
    // version_description - computed: true, optional: false, required: false
    get versionDescription() {
        return this.getStringAttribute('version_description');
    }
    get filter() {
        return this._filter;
    }
    putFilter(value) {
        this._filter.internalValue = value;
    }
    resetFilter() {
        this._filter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            default_only: cdktf.booleanToTerraform(this._defaultOnly),
            engine: cdktf.stringToTerraform(this._engine),
            id: cdktf.stringToTerraform(this._id),
            include_all: cdktf.booleanToTerraform(this._includeAll),
            parameter_group_family: cdktf.stringToTerraform(this._parameterGroupFamily),
            preferred_versions: cdktf.listMapper(cdktf.stringToTerraform, false)(this._preferredVersions),
            version: cdktf.stringToTerraform(this._version),
            filter: cdktf.listMapper(dataAwsRdsEngineVersionFilterToTerraform, true)(this._filter.internalValue),
        };
    }
}
exports.DataAwsRdsEngineVersion = DataAwsRdsEngineVersion;
_c = JSII_RTTI_SYMBOL_1;
DataAwsRdsEngineVersion[_c] = { fqn: "@cdktf/provider-aws.dataAwsRdsEngineVersion.DataAwsRdsEngineVersion", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsRdsEngineVersion.tfResourceType = "aws_rds_engine_version";
