// @cdktf/provider-aws/lib/iam-openid-connect-provider/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IamOpenidConnectProvider = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/iam_openid_connect_provider aws_iam_openid_connect_provider}
*/
class IamOpenidConnectProvider extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/iam_openid_connect_provider aws_iam_openid_connect_provider} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options IamOpenidConnectProviderConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iam_openid_connect_provider',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._clientIdList = config.clientIdList;
        this._id = config.id;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._thumbprintList = config.thumbprintList;
        this._url = config.url;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get clientIdList() {
        return this.getListAttribute('client_id_list');
    }
    set clientIdList(value) {
        this._clientIdList = value;
    }
    // Temporarily expose input value. Use with caution.
    get clientIdListInput() {
        return this._clientIdList;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get thumbprintList() {
        return this.getListAttribute('thumbprint_list');
    }
    set thumbprintList(value) {
        this._thumbprintList = value;
    }
    // Temporarily expose input value. Use with caution.
    get thumbprintListInput() {
        return this._thumbprintList;
    }
    get url() {
        return this.getStringAttribute('url');
    }
    set url(value) {
        this._url = value;
    }
    // Temporarily expose input value. Use with caution.
    get urlInput() {
        return this._url;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            client_id_list: cdktf.listMapper(cdktf.stringToTerraform, false)(this._clientIdList),
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            thumbprint_list: cdktf.listMapper(cdktf.stringToTerraform, false)(this._thumbprintList),
            url: cdktf.stringToTerraform(this._url),
        };
    }
}
exports.IamOpenidConnectProvider = IamOpenidConnectProvider;
_a = JSII_RTTI_SYMBOL_1;
IamOpenidConnectProvider[_a] = { fqn: "@cdktf/provider-aws.iamOpenidConnectProvider.IamOpenidConnectProvider", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
IamOpenidConnectProvider.tfResourceType = "aws_iam_openid_connect_provider";
