// @cdktf/provider-aws/lib/data-aws-api-gateway-export/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsApiGatewayExport = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/api_gateway_export aws_api_gateway_export}
*/
class DataAwsApiGatewayExport extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/api_gateway_export aws_api_gateway_export} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsApiGatewayExportConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_api_gateway_export',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._accepts = config.accepts;
        this._exportType = config.exportType;
        this._id = config.id;
        this._parameters = config.parameters;
        this._restApiId = config.restApiId;
        this._stageName = config.stageName;
    }
    get accepts() {
        return this.getStringAttribute('accepts');
    }
    set accepts(value) {
        this._accepts = value;
    }
    resetAccepts() {
        this._accepts = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get acceptsInput() {
        return this._accepts;
    }
    // body - computed: true, optional: false, required: false
    get body() {
        return this.getStringAttribute('body');
    }
    // content_disposition - computed: true, optional: false, required: false
    get contentDisposition() {
        return this.getStringAttribute('content_disposition');
    }
    // content_type - computed: true, optional: false, required: false
    get contentType() {
        return this.getStringAttribute('content_type');
    }
    get exportType() {
        return this.getStringAttribute('export_type');
    }
    set exportType(value) {
        this._exportType = value;
    }
    // Temporarily expose input value. Use with caution.
    get exportTypeInput() {
        return this._exportType;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get parameters() {
        return this.getStringMapAttribute('parameters');
    }
    set parameters(value) {
        this._parameters = value;
    }
    resetParameters() {
        this._parameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parametersInput() {
        return this._parameters;
    }
    get restApiId() {
        return this.getStringAttribute('rest_api_id');
    }
    set restApiId(value) {
        this._restApiId = value;
    }
    // Temporarily expose input value. Use with caution.
    get restApiIdInput() {
        return this._restApiId;
    }
    get stageName() {
        return this.getStringAttribute('stage_name');
    }
    set stageName(value) {
        this._stageName = value;
    }
    // Temporarily expose input value. Use with caution.
    get stageNameInput() {
        return this._stageName;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            accepts: cdktf.stringToTerraform(this._accepts),
            export_type: cdktf.stringToTerraform(this._exportType),
            id: cdktf.stringToTerraform(this._id),
            parameters: cdktf.hashMapper(cdktf.stringToTerraform)(this._parameters),
            rest_api_id: cdktf.stringToTerraform(this._restApiId),
            stage_name: cdktf.stringToTerraform(this._stageName),
        };
    }
}
exports.DataAwsApiGatewayExport = DataAwsApiGatewayExport;
_a = JSII_RTTI_SYMBOL_1;
DataAwsApiGatewayExport[_a] = { fqn: "@cdktf/provider-aws.dataAwsApiGatewayExport.DataAwsApiGatewayExport", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsApiGatewayExport.tfResourceType = "aws_api_gateway_export";
