// @cdktf/provider-aws/lib/securityhub-standards-control/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecurityhubStandardsControl = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/securityhub_standards_control aws_securityhub_standards_control}
*/
class SecurityhubStandardsControl extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/securityhub_standards_control aws_securityhub_standards_control} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SecurityhubStandardsControlConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_securityhub_standards_control',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._controlStatus = config.controlStatus;
        this._disabledReason = config.disabledReason;
        this._id = config.id;
        this._standardsControlArn = config.standardsControlArn;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // control_id - computed: true, optional: false, required: false
    get controlId() {
        return this.getStringAttribute('control_id');
    }
    get controlStatus() {
        return this.getStringAttribute('control_status');
    }
    set controlStatus(value) {
        this._controlStatus = value;
    }
    // Temporarily expose input value. Use with caution.
    get controlStatusInput() {
        return this._controlStatus;
    }
    // control_status_updated_at - computed: true, optional: false, required: false
    get controlStatusUpdatedAt() {
        return this.getStringAttribute('control_status_updated_at');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get disabledReason() {
        return this.getStringAttribute('disabled_reason');
    }
    set disabledReason(value) {
        this._disabledReason = value;
    }
    resetDisabledReason() {
        this._disabledReason = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get disabledReasonInput() {
        return this._disabledReason;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // related_requirements - computed: true, optional: false, required: false
    get relatedRequirements() {
        return this.getListAttribute('related_requirements');
    }
    // remediation_url - computed: true, optional: false, required: false
    get remediationUrl() {
        return this.getStringAttribute('remediation_url');
    }
    // severity_rating - computed: true, optional: false, required: false
    get severityRating() {
        return this.getStringAttribute('severity_rating');
    }
    get standardsControlArn() {
        return this.getStringAttribute('standards_control_arn');
    }
    set standardsControlArn(value) {
        this._standardsControlArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get standardsControlArnInput() {
        return this._standardsControlArn;
    }
    // title - computed: true, optional: false, required: false
    get title() {
        return this.getStringAttribute('title');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            control_status: cdktf.stringToTerraform(this._controlStatus),
            disabled_reason: cdktf.stringToTerraform(this._disabledReason),
            id: cdktf.stringToTerraform(this._id),
            standards_control_arn: cdktf.stringToTerraform(this._standardsControlArn),
        };
    }
}
exports.SecurityhubStandardsControl = SecurityhubStandardsControl;
_a = JSII_RTTI_SYMBOL_1;
SecurityhubStandardsControl[_a] = { fqn: "@cdktf/provider-aws.securityhubStandardsControl.SecurityhubStandardsControl", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
SecurityhubStandardsControl.tfResourceType = "aws_securityhub_standards_control";
