// @cdktf/provider-aws/lib/glue-crawler/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlueCrawler = exports.GlueCrawlerSchemaChangePolicyOutputReference = exports.glueCrawlerSchemaChangePolicyToTerraform = exports.GlueCrawlerS3TargetList = exports.GlueCrawlerS3TargetOutputReference = exports.glueCrawlerS3TargetToTerraform = exports.GlueCrawlerRecrawlPolicyOutputReference = exports.glueCrawlerRecrawlPolicyToTerraform = exports.GlueCrawlerMongodbTargetList = exports.GlueCrawlerMongodbTargetOutputReference = exports.glueCrawlerMongodbTargetToTerraform = exports.GlueCrawlerLineageConfigurationOutputReference = exports.glueCrawlerLineageConfigurationToTerraform = exports.GlueCrawlerJdbcTargetList = exports.GlueCrawlerJdbcTargetOutputReference = exports.glueCrawlerJdbcTargetToTerraform = exports.GlueCrawlerDynamodbTargetList = exports.GlueCrawlerDynamodbTargetOutputReference = exports.glueCrawlerDynamodbTargetToTerraform = exports.GlueCrawlerDeltaTargetList = exports.GlueCrawlerDeltaTargetOutputReference = exports.glueCrawlerDeltaTargetToTerraform = exports.GlueCrawlerCatalogTargetList = exports.GlueCrawlerCatalogTargetOutputReference = exports.glueCrawlerCatalogTargetToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function glueCrawlerCatalogTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        database_name: cdktf.stringToTerraform(struct.databaseName),
        tables: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.tables),
    };
}
exports.glueCrawlerCatalogTargetToTerraform = glueCrawlerCatalogTargetToTerraform;
class GlueCrawlerCatalogTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._databaseName !== undefined) {
            hasAnyValues = true;
            internalValueResult.databaseName = this._databaseName;
        }
        if (this._tables !== undefined) {
            hasAnyValues = true;
            internalValueResult.tables = this._tables;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._databaseName = undefined;
            this._tables = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._databaseName = value.databaseName;
            this._tables = value.tables;
        }
    }
    get databaseName() {
        return this.getStringAttribute('database_name');
    }
    set databaseName(value) {
        this._databaseName = value;
    }
    // Temporarily expose input value. Use with caution.
    get databaseNameInput() {
        return this._databaseName;
    }
    get tables() {
        return this.getListAttribute('tables');
    }
    set tables(value) {
        this._tables = value;
    }
    // Temporarily expose input value. Use with caution.
    get tablesInput() {
        return this._tables;
    }
}
exports.GlueCrawlerCatalogTargetOutputReference = GlueCrawlerCatalogTargetOutputReference;
_a = JSII_RTTI_SYMBOL_1;
GlueCrawlerCatalogTargetOutputReference[_a] = { fqn: "@cdktf/provider-aws.glueCrawler.GlueCrawlerCatalogTargetOutputReference", version: "10.0.12" };
class GlueCrawlerCatalogTargetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new GlueCrawlerCatalogTargetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.GlueCrawlerCatalogTargetList = GlueCrawlerCatalogTargetList;
_b = JSII_RTTI_SYMBOL_1;
GlueCrawlerCatalogTargetList[_b] = { fqn: "@cdktf/provider-aws.glueCrawler.GlueCrawlerCatalogTargetList", version: "10.0.12" };
function glueCrawlerDeltaTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        connection_name: cdktf.stringToTerraform(struct.connectionName),
        delta_tables: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.deltaTables),
        write_manifest: cdktf.booleanToTerraform(struct.writeManifest),
    };
}
exports.glueCrawlerDeltaTargetToTerraform = glueCrawlerDeltaTargetToTerraform;
class GlueCrawlerDeltaTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._connectionName !== undefined) {
            hasAnyValues = true;
            internalValueResult.connectionName = this._connectionName;
        }
        if (this._deltaTables !== undefined) {
            hasAnyValues = true;
            internalValueResult.deltaTables = this._deltaTables;
        }
        if (this._writeManifest !== undefined) {
            hasAnyValues = true;
            internalValueResult.writeManifest = this._writeManifest;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._connectionName = undefined;
            this._deltaTables = undefined;
            this._writeManifest = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._connectionName = value.connectionName;
            this._deltaTables = value.deltaTables;
            this._writeManifest = value.writeManifest;
        }
    }
    get connectionName() {
        return this.getStringAttribute('connection_name');
    }
    set connectionName(value) {
        this._connectionName = value;
    }
    // Temporarily expose input value. Use with caution.
    get connectionNameInput() {
        return this._connectionName;
    }
    get deltaTables() {
        return cdktf.Fn.tolist(this.getListAttribute('delta_tables'));
    }
    set deltaTables(value) {
        this._deltaTables = value;
    }
    // Temporarily expose input value. Use with caution.
    get deltaTablesInput() {
        return this._deltaTables;
    }
    get writeManifest() {
        return this.getBooleanAttribute('write_manifest');
    }
    set writeManifest(value) {
        this._writeManifest = value;
    }
    // Temporarily expose input value. Use with caution.
    get writeManifestInput() {
        return this._writeManifest;
    }
}
exports.GlueCrawlerDeltaTargetOutputReference = GlueCrawlerDeltaTargetOutputReference;
_c = JSII_RTTI_SYMBOL_1;
GlueCrawlerDeltaTargetOutputReference[_c] = { fqn: "@cdktf/provider-aws.glueCrawler.GlueCrawlerDeltaTargetOutputReference", version: "10.0.12" };
class GlueCrawlerDeltaTargetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new GlueCrawlerDeltaTargetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.GlueCrawlerDeltaTargetList = GlueCrawlerDeltaTargetList;
_d = JSII_RTTI_SYMBOL_1;
GlueCrawlerDeltaTargetList[_d] = { fqn: "@cdktf/provider-aws.glueCrawler.GlueCrawlerDeltaTargetList", version: "10.0.12" };
function glueCrawlerDynamodbTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        path: cdktf.stringToTerraform(struct.path),
        scan_all: cdktf.booleanToTerraform(struct.scanAll),
        scan_rate: cdktf.numberToTerraform(struct.scanRate),
    };
}
exports.glueCrawlerDynamodbTargetToTerraform = glueCrawlerDynamodbTargetToTerraform;
class GlueCrawlerDynamodbTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        if (this._scanAll !== undefined) {
            hasAnyValues = true;
            internalValueResult.scanAll = this._scanAll;
        }
        if (this._scanRate !== undefined) {
            hasAnyValues = true;
            internalValueResult.scanRate = this._scanRate;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._path = undefined;
            this._scanAll = undefined;
            this._scanRate = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._path = value.path;
            this._scanAll = value.scanAll;
            this._scanRate = value.scanRate;
        }
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
    get scanAll() {
        return this.getBooleanAttribute('scan_all');
    }
    set scanAll(value) {
        this._scanAll = value;
    }
    resetScanAll() {
        this._scanAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scanAllInput() {
        return this._scanAll;
    }
    get scanRate() {
        return this.getNumberAttribute('scan_rate');
    }
    set scanRate(value) {
        this._scanRate = value;
    }
    resetScanRate() {
        this._scanRate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scanRateInput() {
        return this._scanRate;
    }
}
exports.GlueCrawlerDynamodbTargetOutputReference = GlueCrawlerDynamodbTargetOutputReference;
_e = JSII_RTTI_SYMBOL_1;
GlueCrawlerDynamodbTargetOutputReference[_e] = { fqn: "@cdktf/provider-aws.glueCrawler.GlueCrawlerDynamodbTargetOutputReference", version: "10.0.12" };
class GlueCrawlerDynamodbTargetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new GlueCrawlerDynamodbTargetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.GlueCrawlerDynamodbTargetList = GlueCrawlerDynamodbTargetList;
_f = JSII_RTTI_SYMBOL_1;
GlueCrawlerDynamodbTargetList[_f] = { fqn: "@cdktf/provider-aws.glueCrawler.GlueCrawlerDynamodbTargetList", version: "10.0.12" };
function glueCrawlerJdbcTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        connection_name: cdktf.stringToTerraform(struct.connectionName),
        exclusions: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.exclusions),
        path: cdktf.stringToTerraform(struct.path),
    };
}
exports.glueCrawlerJdbcTargetToTerraform = glueCrawlerJdbcTargetToTerraform;
class GlueCrawlerJdbcTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._connectionName !== undefined) {
            hasAnyValues = true;
            internalValueResult.connectionName = this._connectionName;
        }
        if (this._exclusions !== undefined) {
            hasAnyValues = true;
            internalValueResult.exclusions = this._exclusions;
        }
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._connectionName = undefined;
            this._exclusions = undefined;
            this._path = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._connectionName = value.connectionName;
            this._exclusions = value.exclusions;
            this._path = value.path;
        }
    }
    get connectionName() {
        return this.getStringAttribute('connection_name');
    }
    set connectionName(value) {
        this._connectionName = value;
    }
    // Temporarily expose input value. Use with caution.
    get connectionNameInput() {
        return this._connectionName;
    }
    get exclusions() {
        return this.getListAttribute('exclusions');
    }
    set exclusions(value) {
        this._exclusions = value;
    }
    resetExclusions() {
        this._exclusions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exclusionsInput() {
        return this._exclusions;
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
}
exports.GlueCrawlerJdbcTargetOutputReference = GlueCrawlerJdbcTargetOutputReference;
_g = JSII_RTTI_SYMBOL_1;
GlueCrawlerJdbcTargetOutputReference[_g] = { fqn: "@cdktf/provider-aws.glueCrawler.GlueCrawlerJdbcTargetOutputReference", version: "10.0.12" };
class GlueCrawlerJdbcTargetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new GlueCrawlerJdbcTargetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.GlueCrawlerJdbcTargetList = GlueCrawlerJdbcTargetList;
_h = JSII_RTTI_SYMBOL_1;
GlueCrawlerJdbcTargetList[_h] = { fqn: "@cdktf/provider-aws.glueCrawler.GlueCrawlerJdbcTargetList", version: "10.0.12" };
function glueCrawlerLineageConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        crawler_lineage_settings: cdktf.stringToTerraform(struct.crawlerLineageSettings),
    };
}
exports.glueCrawlerLineageConfigurationToTerraform = glueCrawlerLineageConfigurationToTerraform;
class GlueCrawlerLineageConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._crawlerLineageSettings !== undefined) {
            hasAnyValues = true;
            internalValueResult.crawlerLineageSettings = this._crawlerLineageSettings;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._crawlerLineageSettings = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._crawlerLineageSettings = value.crawlerLineageSettings;
        }
    }
    get crawlerLineageSettings() {
        return this.getStringAttribute('crawler_lineage_settings');
    }
    set crawlerLineageSettings(value) {
        this._crawlerLineageSettings = value;
    }
    resetCrawlerLineageSettings() {
        this._crawlerLineageSettings = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get crawlerLineageSettingsInput() {
        return this._crawlerLineageSettings;
    }
}
exports.GlueCrawlerLineageConfigurationOutputReference = GlueCrawlerLineageConfigurationOutputReference;
_j = JSII_RTTI_SYMBOL_1;
GlueCrawlerLineageConfigurationOutputReference[_j] = { fqn: "@cdktf/provider-aws.glueCrawler.GlueCrawlerLineageConfigurationOutputReference", version: "10.0.12" };
function glueCrawlerMongodbTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        connection_name: cdktf.stringToTerraform(struct.connectionName),
        path: cdktf.stringToTerraform(struct.path),
        scan_all: cdktf.booleanToTerraform(struct.scanAll),
    };
}
exports.glueCrawlerMongodbTargetToTerraform = glueCrawlerMongodbTargetToTerraform;
class GlueCrawlerMongodbTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._connectionName !== undefined) {
            hasAnyValues = true;
            internalValueResult.connectionName = this._connectionName;
        }
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        if (this._scanAll !== undefined) {
            hasAnyValues = true;
            internalValueResult.scanAll = this._scanAll;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._connectionName = undefined;
            this._path = undefined;
            this._scanAll = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._connectionName = value.connectionName;
            this._path = value.path;
            this._scanAll = value.scanAll;
        }
    }
    get connectionName() {
        return this.getStringAttribute('connection_name');
    }
    set connectionName(value) {
        this._connectionName = value;
    }
    // Temporarily expose input value. Use with caution.
    get connectionNameInput() {
        return this._connectionName;
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
    get scanAll() {
        return this.getBooleanAttribute('scan_all');
    }
    set scanAll(value) {
        this._scanAll = value;
    }
    resetScanAll() {
        this._scanAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scanAllInput() {
        return this._scanAll;
    }
}
exports.GlueCrawlerMongodbTargetOutputReference = GlueCrawlerMongodbTargetOutputReference;
_k = JSII_RTTI_SYMBOL_1;
GlueCrawlerMongodbTargetOutputReference[_k] = { fqn: "@cdktf/provider-aws.glueCrawler.GlueCrawlerMongodbTargetOutputReference", version: "10.0.12" };
class GlueCrawlerMongodbTargetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new GlueCrawlerMongodbTargetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.GlueCrawlerMongodbTargetList = GlueCrawlerMongodbTargetList;
_l = JSII_RTTI_SYMBOL_1;
GlueCrawlerMongodbTargetList[_l] = { fqn: "@cdktf/provider-aws.glueCrawler.GlueCrawlerMongodbTargetList", version: "10.0.12" };
function glueCrawlerRecrawlPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        recrawl_behavior: cdktf.stringToTerraform(struct.recrawlBehavior),
    };
}
exports.glueCrawlerRecrawlPolicyToTerraform = glueCrawlerRecrawlPolicyToTerraform;
class GlueCrawlerRecrawlPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._recrawlBehavior !== undefined) {
            hasAnyValues = true;
            internalValueResult.recrawlBehavior = this._recrawlBehavior;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._recrawlBehavior = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._recrawlBehavior = value.recrawlBehavior;
        }
    }
    get recrawlBehavior() {
        return this.getStringAttribute('recrawl_behavior');
    }
    set recrawlBehavior(value) {
        this._recrawlBehavior = value;
    }
    resetRecrawlBehavior() {
        this._recrawlBehavior = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get recrawlBehaviorInput() {
        return this._recrawlBehavior;
    }
}
exports.GlueCrawlerRecrawlPolicyOutputReference = GlueCrawlerRecrawlPolicyOutputReference;
_m = JSII_RTTI_SYMBOL_1;
GlueCrawlerRecrawlPolicyOutputReference[_m] = { fqn: "@cdktf/provider-aws.glueCrawler.GlueCrawlerRecrawlPolicyOutputReference", version: "10.0.12" };
function glueCrawlerS3TargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        connection_name: cdktf.stringToTerraform(struct.connectionName),
        dlq_event_queue_arn: cdktf.stringToTerraform(struct.dlqEventQueueArn),
        event_queue_arn: cdktf.stringToTerraform(struct.eventQueueArn),
        exclusions: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.exclusions),
        path: cdktf.stringToTerraform(struct.path),
        sample_size: cdktf.numberToTerraform(struct.sampleSize),
    };
}
exports.glueCrawlerS3TargetToTerraform = glueCrawlerS3TargetToTerraform;
class GlueCrawlerS3TargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._connectionName !== undefined) {
            hasAnyValues = true;
            internalValueResult.connectionName = this._connectionName;
        }
        if (this._dlqEventQueueArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.dlqEventQueueArn = this._dlqEventQueueArn;
        }
        if (this._eventQueueArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.eventQueueArn = this._eventQueueArn;
        }
        if (this._exclusions !== undefined) {
            hasAnyValues = true;
            internalValueResult.exclusions = this._exclusions;
        }
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        if (this._sampleSize !== undefined) {
            hasAnyValues = true;
            internalValueResult.sampleSize = this._sampleSize;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._connectionName = undefined;
            this._dlqEventQueueArn = undefined;
            this._eventQueueArn = undefined;
            this._exclusions = undefined;
            this._path = undefined;
            this._sampleSize = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._connectionName = value.connectionName;
            this._dlqEventQueueArn = value.dlqEventQueueArn;
            this._eventQueueArn = value.eventQueueArn;
            this._exclusions = value.exclusions;
            this._path = value.path;
            this._sampleSize = value.sampleSize;
        }
    }
    get connectionName() {
        return this.getStringAttribute('connection_name');
    }
    set connectionName(value) {
        this._connectionName = value;
    }
    resetConnectionName() {
        this._connectionName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get connectionNameInput() {
        return this._connectionName;
    }
    get dlqEventQueueArn() {
        return this.getStringAttribute('dlq_event_queue_arn');
    }
    set dlqEventQueueArn(value) {
        this._dlqEventQueueArn = value;
    }
    resetDlqEventQueueArn() {
        this._dlqEventQueueArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dlqEventQueueArnInput() {
        return this._dlqEventQueueArn;
    }
    get eventQueueArn() {
        return this.getStringAttribute('event_queue_arn');
    }
    set eventQueueArn(value) {
        this._eventQueueArn = value;
    }
    resetEventQueueArn() {
        this._eventQueueArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get eventQueueArnInput() {
        return this._eventQueueArn;
    }
    get exclusions() {
        return this.getListAttribute('exclusions');
    }
    set exclusions(value) {
        this._exclusions = value;
    }
    resetExclusions() {
        this._exclusions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exclusionsInput() {
        return this._exclusions;
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
    get sampleSize() {
        return this.getNumberAttribute('sample_size');
    }
    set sampleSize(value) {
        this._sampleSize = value;
    }
    resetSampleSize() {
        this._sampleSize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sampleSizeInput() {
        return this._sampleSize;
    }
}
exports.GlueCrawlerS3TargetOutputReference = GlueCrawlerS3TargetOutputReference;
_o = JSII_RTTI_SYMBOL_1;
GlueCrawlerS3TargetOutputReference[_o] = { fqn: "@cdktf/provider-aws.glueCrawler.GlueCrawlerS3TargetOutputReference", version: "10.0.12" };
class GlueCrawlerS3TargetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new GlueCrawlerS3TargetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.GlueCrawlerS3TargetList = GlueCrawlerS3TargetList;
_p = JSII_RTTI_SYMBOL_1;
GlueCrawlerS3TargetList[_p] = { fqn: "@cdktf/provider-aws.glueCrawler.GlueCrawlerS3TargetList", version: "10.0.12" };
function glueCrawlerSchemaChangePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        delete_behavior: cdktf.stringToTerraform(struct.deleteBehavior),
        update_behavior: cdktf.stringToTerraform(struct.updateBehavior),
    };
}
exports.glueCrawlerSchemaChangePolicyToTerraform = glueCrawlerSchemaChangePolicyToTerraform;
class GlueCrawlerSchemaChangePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._deleteBehavior !== undefined) {
            hasAnyValues = true;
            internalValueResult.deleteBehavior = this._deleteBehavior;
        }
        if (this._updateBehavior !== undefined) {
            hasAnyValues = true;
            internalValueResult.updateBehavior = this._updateBehavior;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._deleteBehavior = undefined;
            this._updateBehavior = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._deleteBehavior = value.deleteBehavior;
            this._updateBehavior = value.updateBehavior;
        }
    }
    get deleteBehavior() {
        return this.getStringAttribute('delete_behavior');
    }
    set deleteBehavior(value) {
        this._deleteBehavior = value;
    }
    resetDeleteBehavior() {
        this._deleteBehavior = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteBehaviorInput() {
        return this._deleteBehavior;
    }
    get updateBehavior() {
        return this.getStringAttribute('update_behavior');
    }
    set updateBehavior(value) {
        this._updateBehavior = value;
    }
    resetUpdateBehavior() {
        this._updateBehavior = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateBehaviorInput() {
        return this._updateBehavior;
    }
}
exports.GlueCrawlerSchemaChangePolicyOutputReference = GlueCrawlerSchemaChangePolicyOutputReference;
_q = JSII_RTTI_SYMBOL_1;
GlueCrawlerSchemaChangePolicyOutputReference[_q] = { fqn: "@cdktf/provider-aws.glueCrawler.GlueCrawlerSchemaChangePolicyOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/glue_crawler aws_glue_crawler}
*/
class GlueCrawler extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/glue_crawler aws_glue_crawler} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options GlueCrawlerConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_glue_crawler',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // catalog_target - computed: false, optional: true, required: false
        this._catalogTarget = new GlueCrawlerCatalogTargetList(this, "catalog_target", false);
        // delta_target - computed: false, optional: true, required: false
        this._deltaTarget = new GlueCrawlerDeltaTargetList(this, "delta_target", false);
        // dynamodb_target - computed: false, optional: true, required: false
        this._dynamodbTarget = new GlueCrawlerDynamodbTargetList(this, "dynamodb_target", false);
        // jdbc_target - computed: false, optional: true, required: false
        this._jdbcTarget = new GlueCrawlerJdbcTargetList(this, "jdbc_target", false);
        // lineage_configuration - computed: false, optional: true, required: false
        this._lineageConfiguration = new GlueCrawlerLineageConfigurationOutputReference(this, "lineage_configuration");
        // mongodb_target - computed: false, optional: true, required: false
        this._mongodbTarget = new GlueCrawlerMongodbTargetList(this, "mongodb_target", false);
        // recrawl_policy - computed: false, optional: true, required: false
        this._recrawlPolicy = new GlueCrawlerRecrawlPolicyOutputReference(this, "recrawl_policy");
        // s3_target - computed: false, optional: true, required: false
        this._s3Target = new GlueCrawlerS3TargetList(this, "s3_target", false);
        // schema_change_policy - computed: false, optional: true, required: false
        this._schemaChangePolicy = new GlueCrawlerSchemaChangePolicyOutputReference(this, "schema_change_policy");
        this._classifiers = config.classifiers;
        this._configuration = config.configuration;
        this._databaseName = config.databaseName;
        this._description = config.description;
        this._id = config.id;
        this._name = config.name;
        this._role = config.role;
        this._schedule = config.schedule;
        this._securityConfiguration = config.securityConfiguration;
        this._tablePrefix = config.tablePrefix;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._catalogTarget.internalValue = config.catalogTarget;
        this._deltaTarget.internalValue = config.deltaTarget;
        this._dynamodbTarget.internalValue = config.dynamodbTarget;
        this._jdbcTarget.internalValue = config.jdbcTarget;
        this._lineageConfiguration.internalValue = config.lineageConfiguration;
        this._mongodbTarget.internalValue = config.mongodbTarget;
        this._recrawlPolicy.internalValue = config.recrawlPolicy;
        this._s3Target.internalValue = config.s3Target;
        this._schemaChangePolicy.internalValue = config.schemaChangePolicy;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get classifiers() {
        return this.getListAttribute('classifiers');
    }
    set classifiers(value) {
        this._classifiers = value;
    }
    resetClassifiers() {
        this._classifiers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get classifiersInput() {
        return this._classifiers;
    }
    get configuration() {
        return this.getStringAttribute('configuration');
    }
    set configuration(value) {
        this._configuration = value;
    }
    resetConfiguration() {
        this._configuration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get configurationInput() {
        return this._configuration;
    }
    get databaseName() {
        return this.getStringAttribute('database_name');
    }
    set databaseName(value) {
        this._databaseName = value;
    }
    // Temporarily expose input value. Use with caution.
    get databaseNameInput() {
        return this._databaseName;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get role() {
        return this.getStringAttribute('role');
    }
    set role(value) {
        this._role = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleInput() {
        return this._role;
    }
    get schedule() {
        return this.getStringAttribute('schedule');
    }
    set schedule(value) {
        this._schedule = value;
    }
    resetSchedule() {
        this._schedule = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scheduleInput() {
        return this._schedule;
    }
    get securityConfiguration() {
        return this.getStringAttribute('security_configuration');
    }
    set securityConfiguration(value) {
        this._securityConfiguration = value;
    }
    resetSecurityConfiguration() {
        this._securityConfiguration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityConfigurationInput() {
        return this._securityConfiguration;
    }
    get tablePrefix() {
        return this.getStringAttribute('table_prefix');
    }
    set tablePrefix(value) {
        this._tablePrefix = value;
    }
    resetTablePrefix() {
        this._tablePrefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tablePrefixInput() {
        return this._tablePrefix;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get catalogTarget() {
        return this._catalogTarget;
    }
    putCatalogTarget(value) {
        this._catalogTarget.internalValue = value;
    }
    resetCatalogTarget() {
        this._catalogTarget.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get catalogTargetInput() {
        return this._catalogTarget.internalValue;
    }
    get deltaTarget() {
        return this._deltaTarget;
    }
    putDeltaTarget(value) {
        this._deltaTarget.internalValue = value;
    }
    resetDeltaTarget() {
        this._deltaTarget.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deltaTargetInput() {
        return this._deltaTarget.internalValue;
    }
    get dynamodbTarget() {
        return this._dynamodbTarget;
    }
    putDynamodbTarget(value) {
        this._dynamodbTarget.internalValue = value;
    }
    resetDynamodbTarget() {
        this._dynamodbTarget.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dynamodbTargetInput() {
        return this._dynamodbTarget.internalValue;
    }
    get jdbcTarget() {
        return this._jdbcTarget;
    }
    putJdbcTarget(value) {
        this._jdbcTarget.internalValue = value;
    }
    resetJdbcTarget() {
        this._jdbcTarget.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get jdbcTargetInput() {
        return this._jdbcTarget.internalValue;
    }
    get lineageConfiguration() {
        return this._lineageConfiguration;
    }
    putLineageConfiguration(value) {
        this._lineageConfiguration.internalValue = value;
    }
    resetLineageConfiguration() {
        this._lineageConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lineageConfigurationInput() {
        return this._lineageConfiguration.internalValue;
    }
    get mongodbTarget() {
        return this._mongodbTarget;
    }
    putMongodbTarget(value) {
        this._mongodbTarget.internalValue = value;
    }
    resetMongodbTarget() {
        this._mongodbTarget.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get mongodbTargetInput() {
        return this._mongodbTarget.internalValue;
    }
    get recrawlPolicy() {
        return this._recrawlPolicy;
    }
    putRecrawlPolicy(value) {
        this._recrawlPolicy.internalValue = value;
    }
    resetRecrawlPolicy() {
        this._recrawlPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get recrawlPolicyInput() {
        return this._recrawlPolicy.internalValue;
    }
    get s3Target() {
        return this._s3Target;
    }
    putS3Target(value) {
        this._s3Target.internalValue = value;
    }
    resetS3Target() {
        this._s3Target.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get s3TargetInput() {
        return this._s3Target.internalValue;
    }
    get schemaChangePolicy() {
        return this._schemaChangePolicy;
    }
    putSchemaChangePolicy(value) {
        this._schemaChangePolicy.internalValue = value;
    }
    resetSchemaChangePolicy() {
        this._schemaChangePolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get schemaChangePolicyInput() {
        return this._schemaChangePolicy.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            classifiers: cdktf.listMapper(cdktf.stringToTerraform, false)(this._classifiers),
            configuration: cdktf.stringToTerraform(this._configuration),
            database_name: cdktf.stringToTerraform(this._databaseName),
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            role: cdktf.stringToTerraform(this._role),
            schedule: cdktf.stringToTerraform(this._schedule),
            security_configuration: cdktf.stringToTerraform(this._securityConfiguration),
            table_prefix: cdktf.stringToTerraform(this._tablePrefix),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            catalog_target: cdktf.listMapper(glueCrawlerCatalogTargetToTerraform, true)(this._catalogTarget.internalValue),
            delta_target: cdktf.listMapper(glueCrawlerDeltaTargetToTerraform, true)(this._deltaTarget.internalValue),
            dynamodb_target: cdktf.listMapper(glueCrawlerDynamodbTargetToTerraform, true)(this._dynamodbTarget.internalValue),
            jdbc_target: cdktf.listMapper(glueCrawlerJdbcTargetToTerraform, true)(this._jdbcTarget.internalValue),
            lineage_configuration: glueCrawlerLineageConfigurationToTerraform(this._lineageConfiguration.internalValue),
            mongodb_target: cdktf.listMapper(glueCrawlerMongodbTargetToTerraform, true)(this._mongodbTarget.internalValue),
            recrawl_policy: glueCrawlerRecrawlPolicyToTerraform(this._recrawlPolicy.internalValue),
            s3_target: cdktf.listMapper(glueCrawlerS3TargetToTerraform, true)(this._s3Target.internalValue),
            schema_change_policy: glueCrawlerSchemaChangePolicyToTerraform(this._schemaChangePolicy.internalValue),
        };
    }
}
exports.GlueCrawler = GlueCrawler;
_r = JSII_RTTI_SYMBOL_1;
GlueCrawler[_r] = { fqn: "@cdktf/provider-aws.glueCrawler.GlueCrawler", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
GlueCrawler.tfResourceType = "aws_glue_crawler";
