// @cdktf/provider-aws/lib/data-aws-eks-node-group/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsEksNodeGroup = exports.DataAwsEksNodeGroupTaintsList = exports.DataAwsEksNodeGroupTaintsOutputReference = exports.dataAwsEksNodeGroupTaintsToTerraform = exports.DataAwsEksNodeGroupScalingConfigList = exports.DataAwsEksNodeGroupScalingConfigOutputReference = exports.dataAwsEksNodeGroupScalingConfigToTerraform = exports.DataAwsEksNodeGroupResourcesList = exports.DataAwsEksNodeGroupResourcesOutputReference = exports.dataAwsEksNodeGroupResourcesToTerraform = exports.DataAwsEksNodeGroupResourcesAutoscalingGroupsList = exports.DataAwsEksNodeGroupResourcesAutoscalingGroupsOutputReference = exports.dataAwsEksNodeGroupResourcesAutoscalingGroupsToTerraform = exports.DataAwsEksNodeGroupRemoteAccessList = exports.DataAwsEksNodeGroupRemoteAccessOutputReference = exports.dataAwsEksNodeGroupRemoteAccessToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataAwsEksNodeGroupRemoteAccessToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEksNodeGroupRemoteAccessToTerraform = dataAwsEksNodeGroupRemoteAccessToTerraform;
class DataAwsEksNodeGroupRemoteAccessOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ec2_ssh_key - computed: true, optional: false, required: false
    get ec2SshKey() {
        return this.getStringAttribute('ec2_ssh_key');
    }
    // source_security_group_ids - computed: true, optional: false, required: false
    get sourceSecurityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('source_security_group_ids'));
    }
}
exports.DataAwsEksNodeGroupRemoteAccessOutputReference = DataAwsEksNodeGroupRemoteAccessOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataAwsEksNodeGroupRemoteAccessOutputReference[_a] = { fqn: "@cdktf/provider-aws.dataAwsEksNodeGroup.DataAwsEksNodeGroupRemoteAccessOutputReference", version: "10.0.12" };
class DataAwsEksNodeGroupRemoteAccessList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEksNodeGroupRemoteAccessOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEksNodeGroupRemoteAccessList = DataAwsEksNodeGroupRemoteAccessList;
_b = JSII_RTTI_SYMBOL_1;
DataAwsEksNodeGroupRemoteAccessList[_b] = { fqn: "@cdktf/provider-aws.dataAwsEksNodeGroup.DataAwsEksNodeGroupRemoteAccessList", version: "10.0.12" };
function dataAwsEksNodeGroupResourcesAutoscalingGroupsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEksNodeGroupResourcesAutoscalingGroupsToTerraform = dataAwsEksNodeGroupResourcesAutoscalingGroupsToTerraform;
class DataAwsEksNodeGroupResourcesAutoscalingGroupsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataAwsEksNodeGroupResourcesAutoscalingGroupsOutputReference = DataAwsEksNodeGroupResourcesAutoscalingGroupsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataAwsEksNodeGroupResourcesAutoscalingGroupsOutputReference[_c] = { fqn: "@cdktf/provider-aws.dataAwsEksNodeGroup.DataAwsEksNodeGroupResourcesAutoscalingGroupsOutputReference", version: "10.0.12" };
class DataAwsEksNodeGroupResourcesAutoscalingGroupsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEksNodeGroupResourcesAutoscalingGroupsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEksNodeGroupResourcesAutoscalingGroupsList = DataAwsEksNodeGroupResourcesAutoscalingGroupsList;
_d = JSII_RTTI_SYMBOL_1;
DataAwsEksNodeGroupResourcesAutoscalingGroupsList[_d] = { fqn: "@cdktf/provider-aws.dataAwsEksNodeGroup.DataAwsEksNodeGroupResourcesAutoscalingGroupsList", version: "10.0.12" };
function dataAwsEksNodeGroupResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEksNodeGroupResourcesToTerraform = dataAwsEksNodeGroupResourcesToTerraform;
class DataAwsEksNodeGroupResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // autoscaling_groups - computed: true, optional: false, required: false
        this._autoscalingGroups = new DataAwsEksNodeGroupResourcesAutoscalingGroupsList(this, "autoscaling_groups", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get autoscalingGroups() {
        return this._autoscalingGroups;
    }
    // remote_access_security_group_id - computed: true, optional: false, required: false
    get remoteAccessSecurityGroupId() {
        return this.getStringAttribute('remote_access_security_group_id');
    }
}
exports.DataAwsEksNodeGroupResourcesOutputReference = DataAwsEksNodeGroupResourcesOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataAwsEksNodeGroupResourcesOutputReference[_e] = { fqn: "@cdktf/provider-aws.dataAwsEksNodeGroup.DataAwsEksNodeGroupResourcesOutputReference", version: "10.0.12" };
class DataAwsEksNodeGroupResourcesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEksNodeGroupResourcesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEksNodeGroupResourcesList = DataAwsEksNodeGroupResourcesList;
_f = JSII_RTTI_SYMBOL_1;
DataAwsEksNodeGroupResourcesList[_f] = { fqn: "@cdktf/provider-aws.dataAwsEksNodeGroup.DataAwsEksNodeGroupResourcesList", version: "10.0.12" };
function dataAwsEksNodeGroupScalingConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEksNodeGroupScalingConfigToTerraform = dataAwsEksNodeGroupScalingConfigToTerraform;
class DataAwsEksNodeGroupScalingConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // desired_size - computed: true, optional: false, required: false
    get desiredSize() {
        return this.getNumberAttribute('desired_size');
    }
    // max_size - computed: true, optional: false, required: false
    get maxSize() {
        return this.getNumberAttribute('max_size');
    }
    // min_size - computed: true, optional: false, required: false
    get minSize() {
        return this.getNumberAttribute('min_size');
    }
}
exports.DataAwsEksNodeGroupScalingConfigOutputReference = DataAwsEksNodeGroupScalingConfigOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataAwsEksNodeGroupScalingConfigOutputReference[_g] = { fqn: "@cdktf/provider-aws.dataAwsEksNodeGroup.DataAwsEksNodeGroupScalingConfigOutputReference", version: "10.0.12" };
class DataAwsEksNodeGroupScalingConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEksNodeGroupScalingConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEksNodeGroupScalingConfigList = DataAwsEksNodeGroupScalingConfigList;
_h = JSII_RTTI_SYMBOL_1;
DataAwsEksNodeGroupScalingConfigList[_h] = { fqn: "@cdktf/provider-aws.dataAwsEksNodeGroup.DataAwsEksNodeGroupScalingConfigList", version: "10.0.12" };
function dataAwsEksNodeGroupTaintsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataAwsEksNodeGroupTaintsToTerraform = dataAwsEksNodeGroupTaintsToTerraform;
class DataAwsEksNodeGroupTaintsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // effect - computed: true, optional: false, required: false
    get effect() {
        return this.getStringAttribute('effect');
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // value - computed: true, optional: false, required: false
    get value() {
        return this.getStringAttribute('value');
    }
}
exports.DataAwsEksNodeGroupTaintsOutputReference = DataAwsEksNodeGroupTaintsOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataAwsEksNodeGroupTaintsOutputReference[_j] = { fqn: "@cdktf/provider-aws.dataAwsEksNodeGroup.DataAwsEksNodeGroupTaintsOutputReference", version: "10.0.12" };
class DataAwsEksNodeGroupTaintsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataAwsEksNodeGroupTaintsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataAwsEksNodeGroupTaintsList = DataAwsEksNodeGroupTaintsList;
_k = JSII_RTTI_SYMBOL_1;
DataAwsEksNodeGroupTaintsList[_k] = { fqn: "@cdktf/provider-aws.dataAwsEksNodeGroup.DataAwsEksNodeGroupTaintsList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/eks_node_group aws_eks_node_group}
*/
class DataAwsEksNodeGroup extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/eks_node_group aws_eks_node_group} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsEksNodeGroupConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_eks_node_group',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // remote_access - computed: true, optional: false, required: false
        this._remoteAccess = new DataAwsEksNodeGroupRemoteAccessList(this, "remote_access", false);
        // resources - computed: true, optional: false, required: false
        this._resources = new DataAwsEksNodeGroupResourcesList(this, "resources", false);
        // scaling_config - computed: true, optional: false, required: false
        this._scalingConfig = new DataAwsEksNodeGroupScalingConfigList(this, "scaling_config", false);
        // taints - computed: true, optional: false, required: false
        this._taints = new DataAwsEksNodeGroupTaintsList(this, "taints", false);
        this._clusterName = config.clusterName;
        this._id = config.id;
        this._nodeGroupName = config.nodeGroupName;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // ami_type - computed: true, optional: false, required: false
    get amiType() {
        return this.getStringAttribute('ami_type');
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // capacity_type - computed: true, optional: false, required: false
    get capacityType() {
        return this.getStringAttribute('capacity_type');
    }
    get clusterName() {
        return this.getStringAttribute('cluster_name');
    }
    set clusterName(value) {
        this._clusterName = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterNameInput() {
        return this._clusterName;
    }
    // disk_size - computed: true, optional: false, required: false
    get diskSize() {
        return this.getNumberAttribute('disk_size');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // instance_types - computed: true, optional: false, required: false
    get instanceTypes() {
        return this.getListAttribute('instance_types');
    }
    get labels() {
        return this._labels;
    }
    get nodeGroupName() {
        return this.getStringAttribute('node_group_name');
    }
    set nodeGroupName(value) {
        this._nodeGroupName = value;
    }
    // Temporarily expose input value. Use with caution.
    get nodeGroupNameInput() {
        return this._nodeGroupName;
    }
    // node_role_arn - computed: true, optional: false, required: false
    get nodeRoleArn() {
        return this.getStringAttribute('node_role_arn');
    }
    // release_version - computed: true, optional: false, required: false
    get releaseVersion() {
        return this.getStringAttribute('release_version');
    }
    get remoteAccess() {
        return this._remoteAccess;
    }
    get resources() {
        return this._resources;
    }
    get scalingConfig() {
        return this._scalingConfig;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    // subnet_ids - computed: true, optional: false, required: false
    get subnetIds() {
        return cdktf.Fn.tolist(this.getListAttribute('subnet_ids'));
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get taints() {
        return this._taints;
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_name: cdktf.stringToTerraform(this._clusterName),
            id: cdktf.stringToTerraform(this._id),
            node_group_name: cdktf.stringToTerraform(this._nodeGroupName),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
}
exports.DataAwsEksNodeGroup = DataAwsEksNodeGroup;
_l = JSII_RTTI_SYMBOL_1;
DataAwsEksNodeGroup[_l] = { fqn: "@cdktf/provider-aws.dataAwsEksNodeGroup.DataAwsEksNodeGroup", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsEksNodeGroup.tfResourceType = "aws_eks_node_group";
