// @cdktf/provider-aws/lib/msk-configuration/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MskConfiguration = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/msk_configuration aws_msk_configuration}
*/
class MskConfiguration extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/msk_configuration aws_msk_configuration} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options MskConfigurationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_msk_configuration',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._description = config.description;
        this._id = config.id;
        this._kafkaVersions = config.kafkaVersions;
        this._name = config.name;
        this._serverProperties = config.serverProperties;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get kafkaVersions() {
        return cdktf.Fn.tolist(this.getListAttribute('kafka_versions'));
    }
    set kafkaVersions(value) {
        this._kafkaVersions = value;
    }
    resetKafkaVersions() {
        this._kafkaVersions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kafkaVersionsInput() {
        return this._kafkaVersions;
    }
    // latest_revision - computed: true, optional: false, required: false
    get latestRevision() {
        return this.getNumberAttribute('latest_revision');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get serverProperties() {
        return this.getStringAttribute('server_properties');
    }
    set serverProperties(value) {
        this._serverProperties = value;
    }
    // Temporarily expose input value. Use with caution.
    get serverPropertiesInput() {
        return this._serverProperties;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            kafka_versions: cdktf.listMapper(cdktf.stringToTerraform, false)(this._kafkaVersions),
            name: cdktf.stringToTerraform(this._name),
            server_properties: cdktf.stringToTerraform(this._serverProperties),
        };
    }
}
exports.MskConfiguration = MskConfiguration;
_a = JSII_RTTI_SYMBOL_1;
MskConfiguration[_a] = { fqn: "@cdktf/provider-aws.mskConfiguration.MskConfiguration", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
MskConfiguration.tfResourceType = "aws_msk_configuration";
