// @cdktf/provider-aws/lib/backup-global-settings/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BackupGlobalSettings = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/backup_global_settings aws_backup_global_settings}
*/
class BackupGlobalSettings extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/backup_global_settings aws_backup_global_settings} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options BackupGlobalSettingsConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_backup_global_settings',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._globalSettings = config.globalSettings;
        this._id = config.id;
    }
    get globalSettings() {
        return this.getStringMapAttribute('global_settings');
    }
    set globalSettings(value) {
        this._globalSettings = value;
    }
    // Temporarily expose input value. Use with caution.
    get globalSettingsInput() {
        return this._globalSettings;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            global_settings: cdktf.hashMapper(cdktf.stringToTerraform)(this._globalSettings),
            id: cdktf.stringToTerraform(this._id),
        };
    }
}
exports.BackupGlobalSettings = BackupGlobalSettings;
_a = JSII_RTTI_SYMBOL_1;
BackupGlobalSettings[_a] = { fqn: "@cdktf/provider-aws.backupGlobalSettings.BackupGlobalSettings", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
BackupGlobalSettings.tfResourceType = "aws_backup_global_settings";
