// @cdktf/provider-aws/lib/emr-cluster/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmrCluster = exports.EmrClusterMasterInstanceGroupOutputReference = exports.emrClusterMasterInstanceGroupToTerraform = exports.EmrClusterMasterInstanceGroupEbsConfigList = exports.EmrClusterMasterInstanceGroupEbsConfigOutputReference = exports.emrClusterMasterInstanceGroupEbsConfigToTerraform = exports.EmrClusterMasterInstanceFleetOutputReference = exports.emrClusterMasterInstanceFleetToTerraform = exports.EmrClusterMasterInstanceFleetLaunchSpecificationsOutputReference = exports.emrClusterMasterInstanceFleetLaunchSpecificationsToTerraform = exports.EmrClusterMasterInstanceFleetLaunchSpecificationsSpotSpecificationList = exports.EmrClusterMasterInstanceFleetLaunchSpecificationsSpotSpecificationOutputReference = exports.emrClusterMasterInstanceFleetLaunchSpecificationsSpotSpecificationToTerraform = exports.EmrClusterMasterInstanceFleetLaunchSpecificationsOnDemandSpecificationList = exports.EmrClusterMasterInstanceFleetLaunchSpecificationsOnDemandSpecificationOutputReference = exports.emrClusterMasterInstanceFleetLaunchSpecificationsOnDemandSpecificationToTerraform = exports.EmrClusterMasterInstanceFleetInstanceTypeConfigsList = exports.EmrClusterMasterInstanceFleetInstanceTypeConfigsOutputReference = exports.emrClusterMasterInstanceFleetInstanceTypeConfigsToTerraform = exports.EmrClusterMasterInstanceFleetInstanceTypeConfigsEbsConfigList = exports.EmrClusterMasterInstanceFleetInstanceTypeConfigsEbsConfigOutputReference = exports.emrClusterMasterInstanceFleetInstanceTypeConfigsEbsConfigToTerraform = exports.EmrClusterMasterInstanceFleetInstanceTypeConfigsConfigurationsList = exports.EmrClusterMasterInstanceFleetInstanceTypeConfigsConfigurationsOutputReference = exports.emrClusterMasterInstanceFleetInstanceTypeConfigsConfigurationsToTerraform = exports.EmrClusterKerberosAttributesOutputReference = exports.emrClusterKerberosAttributesToTerraform = exports.EmrClusterEc2AttributesOutputReference = exports.emrClusterEc2AttributesToTerraform = exports.EmrClusterCoreInstanceGroupOutputReference = exports.emrClusterCoreInstanceGroupToTerraform = exports.EmrClusterCoreInstanceGroupEbsConfigList = exports.EmrClusterCoreInstanceGroupEbsConfigOutputReference = exports.emrClusterCoreInstanceGroupEbsConfigToTerraform = exports.EmrClusterCoreInstanceFleetOutputReference = exports.emrClusterCoreInstanceFleetToTerraform = exports.EmrClusterCoreInstanceFleetLaunchSpecificationsOutputReference = exports.emrClusterCoreInstanceFleetLaunchSpecificationsToTerraform = exports.EmrClusterCoreInstanceFleetLaunchSpecificationsSpotSpecificationList = exports.EmrClusterCoreInstanceFleetLaunchSpecificationsSpotSpecificationOutputReference = exports.emrClusterCoreInstanceFleetLaunchSpecificationsSpotSpecificationToTerraform = exports.EmrClusterCoreInstanceFleetLaunchSpecificationsOnDemandSpecificationList = exports.EmrClusterCoreInstanceFleetLaunchSpecificationsOnDemandSpecificationOutputReference = exports.emrClusterCoreInstanceFleetLaunchSpecificationsOnDemandSpecificationToTerraform = exports.EmrClusterCoreInstanceFleetInstanceTypeConfigsList = exports.EmrClusterCoreInstanceFleetInstanceTypeConfigsOutputReference = exports.emrClusterCoreInstanceFleetInstanceTypeConfigsToTerraform = exports.EmrClusterCoreInstanceFleetInstanceTypeConfigsEbsConfigList = exports.EmrClusterCoreInstanceFleetInstanceTypeConfigsEbsConfigOutputReference = exports.emrClusterCoreInstanceFleetInstanceTypeConfigsEbsConfigToTerraform = exports.EmrClusterCoreInstanceFleetInstanceTypeConfigsConfigurationsList = exports.EmrClusterCoreInstanceFleetInstanceTypeConfigsConfigurationsOutputReference = exports.emrClusterCoreInstanceFleetInstanceTypeConfigsConfigurationsToTerraform = exports.EmrClusterBootstrapActionList = exports.EmrClusterBootstrapActionOutputReference = exports.emrClusterBootstrapActionToTerraform = exports.EmrClusterAutoTerminationPolicyOutputReference = exports.emrClusterAutoTerminationPolicyToTerraform = exports.EmrClusterStepList = exports.EmrClusterStepOutputReference = exports.emrClusterStepToTerraform = exports.EmrClusterStepHadoopJarStepList = exports.EmrClusterStepHadoopJarStepOutputReference = exports.emrClusterStepHadoopJarStepToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function emrClusterStepHadoopJarStepToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        args: struct.args === undefined ? null : cdktf.listMapper(cdktf.stringToTerraform, false)(struct.args),
        jar: struct.jar === undefined ? null : cdktf.stringToTerraform(struct.jar),
        main_class: struct.mainClass === undefined ? null : cdktf.stringToTerraform(struct.mainClass),
        properties: struct.properties === undefined ? null : cdktf.hashMapper(cdktf.stringToTerraform)(struct.properties),
    };
}
exports.emrClusterStepHadoopJarStepToTerraform = emrClusterStepHadoopJarStepToTerraform;
class EmrClusterStepHadoopJarStepOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._args !== undefined) {
            hasAnyValues = true;
            internalValueResult.args = this._args;
        }
        if (this._jar !== undefined) {
            hasAnyValues = true;
            internalValueResult.jar = this._jar;
        }
        if (this._mainClass !== undefined) {
            hasAnyValues = true;
            internalValueResult.mainClass = this._mainClass;
        }
        if (this._properties !== undefined) {
            hasAnyValues = true;
            internalValueResult.properties = this._properties;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._args = undefined;
            this._jar = undefined;
            this._mainClass = undefined;
            this._properties = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._args = value.args;
            this._jar = value.jar;
            this._mainClass = value.mainClass;
            this._properties = value.properties;
        }
    }
    get args() {
        return this.getListAttribute('args');
    }
    set args(value) {
        this._args = value;
    }
    resetArgs() {
        this._args = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get argsInput() {
        return this._args;
    }
    get jar() {
        return this.getStringAttribute('jar');
    }
    set jar(value) {
        this._jar = value;
    }
    resetJar() {
        this._jar = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get jarInput() {
        return this._jar;
    }
    get mainClass() {
        return this.getStringAttribute('main_class');
    }
    set mainClass(value) {
        this._mainClass = value;
    }
    resetMainClass() {
        this._mainClass = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get mainClassInput() {
        return this._mainClass;
    }
    get properties() {
        return this.getStringMapAttribute('properties');
    }
    set properties(value) {
        this._properties = value;
    }
    resetProperties() {
        this._properties = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get propertiesInput() {
        return this._properties;
    }
}
exports.EmrClusterStepHadoopJarStepOutputReference = EmrClusterStepHadoopJarStepOutputReference;
_a = JSII_RTTI_SYMBOL_1;
EmrClusterStepHadoopJarStepOutputReference[_a] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterStepHadoopJarStepOutputReference", version: "10.0.12" };
class EmrClusterStepHadoopJarStepList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EmrClusterStepHadoopJarStepOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EmrClusterStepHadoopJarStepList = EmrClusterStepHadoopJarStepList;
_b = JSII_RTTI_SYMBOL_1;
EmrClusterStepHadoopJarStepList[_b] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterStepHadoopJarStepList", version: "10.0.12" };
function emrClusterStepToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        action_on_failure: struct.actionOnFailure === undefined ? null : cdktf.stringToTerraform(struct.actionOnFailure),
        hadoop_jar_step: struct.hadoopJarStep === undefined ? null : cdktf.listMapper(emrClusterStepHadoopJarStepToTerraform, false)(struct.hadoopJarStep),
        name: struct.name === undefined ? null : cdktf.stringToTerraform(struct.name),
    };
}
exports.emrClusterStepToTerraform = emrClusterStepToTerraform;
class EmrClusterStepOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // hadoop_jar_step - computed: true, optional: true, required: false
        this._hadoopJarStep = new EmrClusterStepHadoopJarStepList(this, "hadoop_jar_step", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._actionOnFailure !== undefined) {
            hasAnyValues = true;
            internalValueResult.actionOnFailure = this._actionOnFailure;
        }
        if (this._hadoopJarStep?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.hadoopJarStep = this._hadoopJarStep?.internalValue;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._actionOnFailure = undefined;
            this._hadoopJarStep.internalValue = undefined;
            this._name = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._actionOnFailure = value.actionOnFailure;
            this._hadoopJarStep.internalValue = value.hadoopJarStep;
            this._name = value.name;
        }
    }
    get actionOnFailure() {
        return this.getStringAttribute('action_on_failure');
    }
    set actionOnFailure(value) {
        this._actionOnFailure = value;
    }
    resetActionOnFailure() {
        this._actionOnFailure = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get actionOnFailureInput() {
        return this._actionOnFailure;
    }
    get hadoopJarStep() {
        return this._hadoopJarStep;
    }
    putHadoopJarStep(value) {
        this._hadoopJarStep.internalValue = value;
    }
    resetHadoopJarStep() {
        this._hadoopJarStep.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hadoopJarStepInput() {
        return this._hadoopJarStep.internalValue;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
}
exports.EmrClusterStepOutputReference = EmrClusterStepOutputReference;
_c = JSII_RTTI_SYMBOL_1;
EmrClusterStepOutputReference[_c] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterStepOutputReference", version: "10.0.12" };
class EmrClusterStepList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EmrClusterStepOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EmrClusterStepList = EmrClusterStepList;
_d = JSII_RTTI_SYMBOL_1;
EmrClusterStepList[_d] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterStepList", version: "10.0.12" };
function emrClusterAutoTerminationPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        idle_timeout: cdktf.numberToTerraform(struct.idleTimeout),
    };
}
exports.emrClusterAutoTerminationPolicyToTerraform = emrClusterAutoTerminationPolicyToTerraform;
class EmrClusterAutoTerminationPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._idleTimeout !== undefined) {
            hasAnyValues = true;
            internalValueResult.idleTimeout = this._idleTimeout;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._idleTimeout = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._idleTimeout = value.idleTimeout;
        }
    }
    get idleTimeout() {
        return this.getNumberAttribute('idle_timeout');
    }
    set idleTimeout(value) {
        this._idleTimeout = value;
    }
    resetIdleTimeout() {
        this._idleTimeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idleTimeoutInput() {
        return this._idleTimeout;
    }
}
exports.EmrClusterAutoTerminationPolicyOutputReference = EmrClusterAutoTerminationPolicyOutputReference;
_e = JSII_RTTI_SYMBOL_1;
EmrClusterAutoTerminationPolicyOutputReference[_e] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterAutoTerminationPolicyOutputReference", version: "10.0.12" };
function emrClusterBootstrapActionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        args: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.args),
        name: cdktf.stringToTerraform(struct.name),
        path: cdktf.stringToTerraform(struct.path),
    };
}
exports.emrClusterBootstrapActionToTerraform = emrClusterBootstrapActionToTerraform;
class EmrClusterBootstrapActionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._args !== undefined) {
            hasAnyValues = true;
            internalValueResult.args = this._args;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._args = undefined;
            this._name = undefined;
            this._path = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._args = value.args;
            this._name = value.name;
            this._path = value.path;
        }
    }
    get args() {
        return this.getListAttribute('args');
    }
    set args(value) {
        this._args = value;
    }
    resetArgs() {
        this._args = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get argsInput() {
        return this._args;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
}
exports.EmrClusterBootstrapActionOutputReference = EmrClusterBootstrapActionOutputReference;
_f = JSII_RTTI_SYMBOL_1;
EmrClusterBootstrapActionOutputReference[_f] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterBootstrapActionOutputReference", version: "10.0.12" };
class EmrClusterBootstrapActionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EmrClusterBootstrapActionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EmrClusterBootstrapActionList = EmrClusterBootstrapActionList;
_g = JSII_RTTI_SYMBOL_1;
EmrClusterBootstrapActionList[_g] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterBootstrapActionList", version: "10.0.12" };
function emrClusterCoreInstanceFleetInstanceTypeConfigsConfigurationsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        classification: cdktf.stringToTerraform(struct.classification),
        properties: cdktf.hashMapper(cdktf.stringToTerraform)(struct.properties),
    };
}
exports.emrClusterCoreInstanceFleetInstanceTypeConfigsConfigurationsToTerraform = emrClusterCoreInstanceFleetInstanceTypeConfigsConfigurationsToTerraform;
class EmrClusterCoreInstanceFleetInstanceTypeConfigsConfigurationsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._classification !== undefined) {
            hasAnyValues = true;
            internalValueResult.classification = this._classification;
        }
        if (this._properties !== undefined) {
            hasAnyValues = true;
            internalValueResult.properties = this._properties;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._classification = undefined;
            this._properties = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._classification = value.classification;
            this._properties = value.properties;
        }
    }
    get classification() {
        return this.getStringAttribute('classification');
    }
    set classification(value) {
        this._classification = value;
    }
    resetClassification() {
        this._classification = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get classificationInput() {
        return this._classification;
    }
    get properties() {
        return this.getStringMapAttribute('properties');
    }
    set properties(value) {
        this._properties = value;
    }
    resetProperties() {
        this._properties = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get propertiesInput() {
        return this._properties;
    }
}
exports.EmrClusterCoreInstanceFleetInstanceTypeConfigsConfigurationsOutputReference = EmrClusterCoreInstanceFleetInstanceTypeConfigsConfigurationsOutputReference;
_h = JSII_RTTI_SYMBOL_1;
EmrClusterCoreInstanceFleetInstanceTypeConfigsConfigurationsOutputReference[_h] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterCoreInstanceFleetInstanceTypeConfigsConfigurationsOutputReference", version: "10.0.12" };
class EmrClusterCoreInstanceFleetInstanceTypeConfigsConfigurationsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EmrClusterCoreInstanceFleetInstanceTypeConfigsConfigurationsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EmrClusterCoreInstanceFleetInstanceTypeConfigsConfigurationsList = EmrClusterCoreInstanceFleetInstanceTypeConfigsConfigurationsList;
_j = JSII_RTTI_SYMBOL_1;
EmrClusterCoreInstanceFleetInstanceTypeConfigsConfigurationsList[_j] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterCoreInstanceFleetInstanceTypeConfigsConfigurationsList", version: "10.0.12" };
function emrClusterCoreInstanceFleetInstanceTypeConfigsEbsConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        iops: cdktf.numberToTerraform(struct.iops),
        size: cdktf.numberToTerraform(struct.size),
        type: cdktf.stringToTerraform(struct.type),
        volumes_per_instance: cdktf.numberToTerraform(struct.volumesPerInstance),
    };
}
exports.emrClusterCoreInstanceFleetInstanceTypeConfigsEbsConfigToTerraform = emrClusterCoreInstanceFleetInstanceTypeConfigsEbsConfigToTerraform;
class EmrClusterCoreInstanceFleetInstanceTypeConfigsEbsConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._iops !== undefined) {
            hasAnyValues = true;
            internalValueResult.iops = this._iops;
        }
        if (this._size !== undefined) {
            hasAnyValues = true;
            internalValueResult.size = this._size;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._volumesPerInstance !== undefined) {
            hasAnyValues = true;
            internalValueResult.volumesPerInstance = this._volumesPerInstance;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._iops = undefined;
            this._size = undefined;
            this._type = undefined;
            this._volumesPerInstance = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._iops = value.iops;
            this._size = value.size;
            this._type = value.type;
            this._volumesPerInstance = value.volumesPerInstance;
        }
    }
    get iops() {
        return this.getNumberAttribute('iops');
    }
    set iops(value) {
        this._iops = value;
    }
    resetIops() {
        this._iops = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iopsInput() {
        return this._iops;
    }
    get size() {
        return this.getNumberAttribute('size');
    }
    set size(value) {
        this._size = value;
    }
    // Temporarily expose input value. Use with caution.
    get sizeInput() {
        return this._size;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get volumesPerInstance() {
        return this.getNumberAttribute('volumes_per_instance');
    }
    set volumesPerInstance(value) {
        this._volumesPerInstance = value;
    }
    resetVolumesPerInstance() {
        this._volumesPerInstance = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumesPerInstanceInput() {
        return this._volumesPerInstance;
    }
}
exports.EmrClusterCoreInstanceFleetInstanceTypeConfigsEbsConfigOutputReference = EmrClusterCoreInstanceFleetInstanceTypeConfigsEbsConfigOutputReference;
_k = JSII_RTTI_SYMBOL_1;
EmrClusterCoreInstanceFleetInstanceTypeConfigsEbsConfigOutputReference[_k] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterCoreInstanceFleetInstanceTypeConfigsEbsConfigOutputReference", version: "10.0.12" };
class EmrClusterCoreInstanceFleetInstanceTypeConfigsEbsConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EmrClusterCoreInstanceFleetInstanceTypeConfigsEbsConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EmrClusterCoreInstanceFleetInstanceTypeConfigsEbsConfigList = EmrClusterCoreInstanceFleetInstanceTypeConfigsEbsConfigList;
_l = JSII_RTTI_SYMBOL_1;
EmrClusterCoreInstanceFleetInstanceTypeConfigsEbsConfigList[_l] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterCoreInstanceFleetInstanceTypeConfigsEbsConfigList", version: "10.0.12" };
function emrClusterCoreInstanceFleetInstanceTypeConfigsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bid_price: cdktf.stringToTerraform(struct.bidPrice),
        bid_price_as_percentage_of_on_demand_price: cdktf.numberToTerraform(struct.bidPriceAsPercentageOfOnDemandPrice),
        instance_type: cdktf.stringToTerraform(struct.instanceType),
        weighted_capacity: cdktf.numberToTerraform(struct.weightedCapacity),
        configurations: cdktf.listMapper(emrClusterCoreInstanceFleetInstanceTypeConfigsConfigurationsToTerraform, true)(struct.configurations),
        ebs_config: cdktf.listMapper(emrClusterCoreInstanceFleetInstanceTypeConfigsEbsConfigToTerraform, true)(struct.ebsConfig),
    };
}
exports.emrClusterCoreInstanceFleetInstanceTypeConfigsToTerraform = emrClusterCoreInstanceFleetInstanceTypeConfigsToTerraform;
class EmrClusterCoreInstanceFleetInstanceTypeConfigsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // configurations - computed: false, optional: true, required: false
        this._configurations = new EmrClusterCoreInstanceFleetInstanceTypeConfigsConfigurationsList(this, "configurations", true);
        // ebs_config - computed: false, optional: true, required: false
        this._ebsConfig = new EmrClusterCoreInstanceFleetInstanceTypeConfigsEbsConfigList(this, "ebs_config", true);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bidPrice !== undefined) {
            hasAnyValues = true;
            internalValueResult.bidPrice = this._bidPrice;
        }
        if (this._bidPriceAsPercentageOfOnDemandPrice !== undefined) {
            hasAnyValues = true;
            internalValueResult.bidPriceAsPercentageOfOnDemandPrice = this._bidPriceAsPercentageOfOnDemandPrice;
        }
        if (this._instanceType !== undefined) {
            hasAnyValues = true;
            internalValueResult.instanceType = this._instanceType;
        }
        if (this._weightedCapacity !== undefined) {
            hasAnyValues = true;
            internalValueResult.weightedCapacity = this._weightedCapacity;
        }
        if (this._configurations?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.configurations = this._configurations?.internalValue;
        }
        if (this._ebsConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.ebsConfig = this._ebsConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._bidPrice = undefined;
            this._bidPriceAsPercentageOfOnDemandPrice = undefined;
            this._instanceType = undefined;
            this._weightedCapacity = undefined;
            this._configurations.internalValue = undefined;
            this._ebsConfig.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._bidPrice = value.bidPrice;
            this._bidPriceAsPercentageOfOnDemandPrice = value.bidPriceAsPercentageOfOnDemandPrice;
            this._instanceType = value.instanceType;
            this._weightedCapacity = value.weightedCapacity;
            this._configurations.internalValue = value.configurations;
            this._ebsConfig.internalValue = value.ebsConfig;
        }
    }
    get bidPrice() {
        return this.getStringAttribute('bid_price');
    }
    set bidPrice(value) {
        this._bidPrice = value;
    }
    resetBidPrice() {
        this._bidPrice = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bidPriceInput() {
        return this._bidPrice;
    }
    get bidPriceAsPercentageOfOnDemandPrice() {
        return this.getNumberAttribute('bid_price_as_percentage_of_on_demand_price');
    }
    set bidPriceAsPercentageOfOnDemandPrice(value) {
        this._bidPriceAsPercentageOfOnDemandPrice = value;
    }
    resetBidPriceAsPercentageOfOnDemandPrice() {
        this._bidPriceAsPercentageOfOnDemandPrice = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bidPriceAsPercentageOfOnDemandPriceInput() {
        return this._bidPriceAsPercentageOfOnDemandPrice;
    }
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    set instanceType(value) {
        this._instanceType = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceTypeInput() {
        return this._instanceType;
    }
    get weightedCapacity() {
        return this.getNumberAttribute('weighted_capacity');
    }
    set weightedCapacity(value) {
        this._weightedCapacity = value;
    }
    resetWeightedCapacity() {
        this._weightedCapacity = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get weightedCapacityInput() {
        return this._weightedCapacity;
    }
    get configurations() {
        return this._configurations;
    }
    putConfigurations(value) {
        this._configurations.internalValue = value;
    }
    resetConfigurations() {
        this._configurations.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get configurationsInput() {
        return this._configurations.internalValue;
    }
    get ebsConfig() {
        return this._ebsConfig;
    }
    putEbsConfig(value) {
        this._ebsConfig.internalValue = value;
    }
    resetEbsConfig() {
        this._ebsConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ebsConfigInput() {
        return this._ebsConfig.internalValue;
    }
}
exports.EmrClusterCoreInstanceFleetInstanceTypeConfigsOutputReference = EmrClusterCoreInstanceFleetInstanceTypeConfigsOutputReference;
_m = JSII_RTTI_SYMBOL_1;
EmrClusterCoreInstanceFleetInstanceTypeConfigsOutputReference[_m] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterCoreInstanceFleetInstanceTypeConfigsOutputReference", version: "10.0.12" };
class EmrClusterCoreInstanceFleetInstanceTypeConfigsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EmrClusterCoreInstanceFleetInstanceTypeConfigsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EmrClusterCoreInstanceFleetInstanceTypeConfigsList = EmrClusterCoreInstanceFleetInstanceTypeConfigsList;
_o = JSII_RTTI_SYMBOL_1;
EmrClusterCoreInstanceFleetInstanceTypeConfigsList[_o] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterCoreInstanceFleetInstanceTypeConfigsList", version: "10.0.12" };
function emrClusterCoreInstanceFleetLaunchSpecificationsOnDemandSpecificationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        allocation_strategy: cdktf.stringToTerraform(struct.allocationStrategy),
    };
}
exports.emrClusterCoreInstanceFleetLaunchSpecificationsOnDemandSpecificationToTerraform = emrClusterCoreInstanceFleetLaunchSpecificationsOnDemandSpecificationToTerraform;
class EmrClusterCoreInstanceFleetLaunchSpecificationsOnDemandSpecificationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._allocationStrategy !== undefined) {
            hasAnyValues = true;
            internalValueResult.allocationStrategy = this._allocationStrategy;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._allocationStrategy = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._allocationStrategy = value.allocationStrategy;
        }
    }
    get allocationStrategy() {
        return this.getStringAttribute('allocation_strategy');
    }
    set allocationStrategy(value) {
        this._allocationStrategy = value;
    }
    // Temporarily expose input value. Use with caution.
    get allocationStrategyInput() {
        return this._allocationStrategy;
    }
}
exports.EmrClusterCoreInstanceFleetLaunchSpecificationsOnDemandSpecificationOutputReference = EmrClusterCoreInstanceFleetLaunchSpecificationsOnDemandSpecificationOutputReference;
_p = JSII_RTTI_SYMBOL_1;
EmrClusterCoreInstanceFleetLaunchSpecificationsOnDemandSpecificationOutputReference[_p] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterCoreInstanceFleetLaunchSpecificationsOnDemandSpecificationOutputReference", version: "10.0.12" };
class EmrClusterCoreInstanceFleetLaunchSpecificationsOnDemandSpecificationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EmrClusterCoreInstanceFleetLaunchSpecificationsOnDemandSpecificationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EmrClusterCoreInstanceFleetLaunchSpecificationsOnDemandSpecificationList = EmrClusterCoreInstanceFleetLaunchSpecificationsOnDemandSpecificationList;
_q = JSII_RTTI_SYMBOL_1;
EmrClusterCoreInstanceFleetLaunchSpecificationsOnDemandSpecificationList[_q] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterCoreInstanceFleetLaunchSpecificationsOnDemandSpecificationList", version: "10.0.12" };
function emrClusterCoreInstanceFleetLaunchSpecificationsSpotSpecificationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        allocation_strategy: cdktf.stringToTerraform(struct.allocationStrategy),
        block_duration_minutes: cdktf.numberToTerraform(struct.blockDurationMinutes),
        timeout_action: cdktf.stringToTerraform(struct.timeoutAction),
        timeout_duration_minutes: cdktf.numberToTerraform(struct.timeoutDurationMinutes),
    };
}
exports.emrClusterCoreInstanceFleetLaunchSpecificationsSpotSpecificationToTerraform = emrClusterCoreInstanceFleetLaunchSpecificationsSpotSpecificationToTerraform;
class EmrClusterCoreInstanceFleetLaunchSpecificationsSpotSpecificationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._allocationStrategy !== undefined) {
            hasAnyValues = true;
            internalValueResult.allocationStrategy = this._allocationStrategy;
        }
        if (this._blockDurationMinutes !== undefined) {
            hasAnyValues = true;
            internalValueResult.blockDurationMinutes = this._blockDurationMinutes;
        }
        if (this._timeoutAction !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeoutAction = this._timeoutAction;
        }
        if (this._timeoutDurationMinutes !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeoutDurationMinutes = this._timeoutDurationMinutes;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._allocationStrategy = undefined;
            this._blockDurationMinutes = undefined;
            this._timeoutAction = undefined;
            this._timeoutDurationMinutes = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._allocationStrategy = value.allocationStrategy;
            this._blockDurationMinutes = value.blockDurationMinutes;
            this._timeoutAction = value.timeoutAction;
            this._timeoutDurationMinutes = value.timeoutDurationMinutes;
        }
    }
    get allocationStrategy() {
        return this.getStringAttribute('allocation_strategy');
    }
    set allocationStrategy(value) {
        this._allocationStrategy = value;
    }
    // Temporarily expose input value. Use with caution.
    get allocationStrategyInput() {
        return this._allocationStrategy;
    }
    get blockDurationMinutes() {
        return this.getNumberAttribute('block_duration_minutes');
    }
    set blockDurationMinutes(value) {
        this._blockDurationMinutes = value;
    }
    resetBlockDurationMinutes() {
        this._blockDurationMinutes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get blockDurationMinutesInput() {
        return this._blockDurationMinutes;
    }
    get timeoutAction() {
        return this.getStringAttribute('timeout_action');
    }
    set timeoutAction(value) {
        this._timeoutAction = value;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutActionInput() {
        return this._timeoutAction;
    }
    get timeoutDurationMinutes() {
        return this.getNumberAttribute('timeout_duration_minutes');
    }
    set timeoutDurationMinutes(value) {
        this._timeoutDurationMinutes = value;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutDurationMinutesInput() {
        return this._timeoutDurationMinutes;
    }
}
exports.EmrClusterCoreInstanceFleetLaunchSpecificationsSpotSpecificationOutputReference = EmrClusterCoreInstanceFleetLaunchSpecificationsSpotSpecificationOutputReference;
_r = JSII_RTTI_SYMBOL_1;
EmrClusterCoreInstanceFleetLaunchSpecificationsSpotSpecificationOutputReference[_r] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterCoreInstanceFleetLaunchSpecificationsSpotSpecificationOutputReference", version: "10.0.12" };
class EmrClusterCoreInstanceFleetLaunchSpecificationsSpotSpecificationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EmrClusterCoreInstanceFleetLaunchSpecificationsSpotSpecificationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EmrClusterCoreInstanceFleetLaunchSpecificationsSpotSpecificationList = EmrClusterCoreInstanceFleetLaunchSpecificationsSpotSpecificationList;
_s = JSII_RTTI_SYMBOL_1;
EmrClusterCoreInstanceFleetLaunchSpecificationsSpotSpecificationList[_s] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterCoreInstanceFleetLaunchSpecificationsSpotSpecificationList", version: "10.0.12" };
function emrClusterCoreInstanceFleetLaunchSpecificationsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        on_demand_specification: cdktf.listMapper(emrClusterCoreInstanceFleetLaunchSpecificationsOnDemandSpecificationToTerraform, true)(struct.onDemandSpecification),
        spot_specification: cdktf.listMapper(emrClusterCoreInstanceFleetLaunchSpecificationsSpotSpecificationToTerraform, true)(struct.spotSpecification),
    };
}
exports.emrClusterCoreInstanceFleetLaunchSpecificationsToTerraform = emrClusterCoreInstanceFleetLaunchSpecificationsToTerraform;
class EmrClusterCoreInstanceFleetLaunchSpecificationsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // on_demand_specification - computed: false, optional: true, required: false
        this._onDemandSpecification = new EmrClusterCoreInstanceFleetLaunchSpecificationsOnDemandSpecificationList(this, "on_demand_specification", false);
        // spot_specification - computed: false, optional: true, required: false
        this._spotSpecification = new EmrClusterCoreInstanceFleetLaunchSpecificationsSpotSpecificationList(this, "spot_specification", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._onDemandSpecification?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.onDemandSpecification = this._onDemandSpecification?.internalValue;
        }
        if (this._spotSpecification?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.spotSpecification = this._spotSpecification?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._onDemandSpecification.internalValue = undefined;
            this._spotSpecification.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._onDemandSpecification.internalValue = value.onDemandSpecification;
            this._spotSpecification.internalValue = value.spotSpecification;
        }
    }
    get onDemandSpecification() {
        return this._onDemandSpecification;
    }
    putOnDemandSpecification(value) {
        this._onDemandSpecification.internalValue = value;
    }
    resetOnDemandSpecification() {
        this._onDemandSpecification.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get onDemandSpecificationInput() {
        return this._onDemandSpecification.internalValue;
    }
    get spotSpecification() {
        return this._spotSpecification;
    }
    putSpotSpecification(value) {
        this._spotSpecification.internalValue = value;
    }
    resetSpotSpecification() {
        this._spotSpecification.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get spotSpecificationInput() {
        return this._spotSpecification.internalValue;
    }
}
exports.EmrClusterCoreInstanceFleetLaunchSpecificationsOutputReference = EmrClusterCoreInstanceFleetLaunchSpecificationsOutputReference;
_t = JSII_RTTI_SYMBOL_1;
EmrClusterCoreInstanceFleetLaunchSpecificationsOutputReference[_t] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterCoreInstanceFleetLaunchSpecificationsOutputReference", version: "10.0.12" };
function emrClusterCoreInstanceFleetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        target_on_demand_capacity: cdktf.numberToTerraform(struct.targetOnDemandCapacity),
        target_spot_capacity: cdktf.numberToTerraform(struct.targetSpotCapacity),
        instance_type_configs: cdktf.listMapper(emrClusterCoreInstanceFleetInstanceTypeConfigsToTerraform, true)(struct.instanceTypeConfigs),
        launch_specifications: emrClusterCoreInstanceFleetLaunchSpecificationsToTerraform(struct.launchSpecifications),
    };
}
exports.emrClusterCoreInstanceFleetToTerraform = emrClusterCoreInstanceFleetToTerraform;
class EmrClusterCoreInstanceFleetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // instance_type_configs - computed: false, optional: true, required: false
        this._instanceTypeConfigs = new EmrClusterCoreInstanceFleetInstanceTypeConfigsList(this, "instance_type_configs", true);
        // launch_specifications - computed: false, optional: true, required: false
        this._launchSpecifications = new EmrClusterCoreInstanceFleetLaunchSpecificationsOutputReference(this, "launch_specifications");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._targetOnDemandCapacity !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetOnDemandCapacity = this._targetOnDemandCapacity;
        }
        if (this._targetSpotCapacity !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetSpotCapacity = this._targetSpotCapacity;
        }
        if (this._instanceTypeConfigs?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.instanceTypeConfigs = this._instanceTypeConfigs?.internalValue;
        }
        if (this._launchSpecifications?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.launchSpecifications = this._launchSpecifications?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._name = undefined;
            this._targetOnDemandCapacity = undefined;
            this._targetSpotCapacity = undefined;
            this._instanceTypeConfigs.internalValue = undefined;
            this._launchSpecifications.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._name = value.name;
            this._targetOnDemandCapacity = value.targetOnDemandCapacity;
            this._targetSpotCapacity = value.targetSpotCapacity;
            this._instanceTypeConfigs.internalValue = value.instanceTypeConfigs;
            this._launchSpecifications.internalValue = value.launchSpecifications;
        }
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // provisioned_on_demand_capacity - computed: true, optional: false, required: false
    get provisionedOnDemandCapacity() {
        return this.getNumberAttribute('provisioned_on_demand_capacity');
    }
    // provisioned_spot_capacity - computed: true, optional: false, required: false
    get provisionedSpotCapacity() {
        return this.getNumberAttribute('provisioned_spot_capacity');
    }
    get targetOnDemandCapacity() {
        return this.getNumberAttribute('target_on_demand_capacity');
    }
    set targetOnDemandCapacity(value) {
        this._targetOnDemandCapacity = value;
    }
    resetTargetOnDemandCapacity() {
        this._targetOnDemandCapacity = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetOnDemandCapacityInput() {
        return this._targetOnDemandCapacity;
    }
    get targetSpotCapacity() {
        return this.getNumberAttribute('target_spot_capacity');
    }
    set targetSpotCapacity(value) {
        this._targetSpotCapacity = value;
    }
    resetTargetSpotCapacity() {
        this._targetSpotCapacity = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetSpotCapacityInput() {
        return this._targetSpotCapacity;
    }
    get instanceTypeConfigs() {
        return this._instanceTypeConfigs;
    }
    putInstanceTypeConfigs(value) {
        this._instanceTypeConfigs.internalValue = value;
    }
    resetInstanceTypeConfigs() {
        this._instanceTypeConfigs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instanceTypeConfigsInput() {
        return this._instanceTypeConfigs.internalValue;
    }
    get launchSpecifications() {
        return this._launchSpecifications;
    }
    putLaunchSpecifications(value) {
        this._launchSpecifications.internalValue = value;
    }
    resetLaunchSpecifications() {
        this._launchSpecifications.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get launchSpecificationsInput() {
        return this._launchSpecifications.internalValue;
    }
}
exports.EmrClusterCoreInstanceFleetOutputReference = EmrClusterCoreInstanceFleetOutputReference;
_u = JSII_RTTI_SYMBOL_1;
EmrClusterCoreInstanceFleetOutputReference[_u] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterCoreInstanceFleetOutputReference", version: "10.0.12" };
function emrClusterCoreInstanceGroupEbsConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        iops: cdktf.numberToTerraform(struct.iops),
        size: cdktf.numberToTerraform(struct.size),
        throughput: cdktf.numberToTerraform(struct.throughput),
        type: cdktf.stringToTerraform(struct.type),
        volumes_per_instance: cdktf.numberToTerraform(struct.volumesPerInstance),
    };
}
exports.emrClusterCoreInstanceGroupEbsConfigToTerraform = emrClusterCoreInstanceGroupEbsConfigToTerraform;
class EmrClusterCoreInstanceGroupEbsConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._iops !== undefined) {
            hasAnyValues = true;
            internalValueResult.iops = this._iops;
        }
        if (this._size !== undefined) {
            hasAnyValues = true;
            internalValueResult.size = this._size;
        }
        if (this._throughput !== undefined) {
            hasAnyValues = true;
            internalValueResult.throughput = this._throughput;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._volumesPerInstance !== undefined) {
            hasAnyValues = true;
            internalValueResult.volumesPerInstance = this._volumesPerInstance;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._iops = undefined;
            this._size = undefined;
            this._throughput = undefined;
            this._type = undefined;
            this._volumesPerInstance = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._iops = value.iops;
            this._size = value.size;
            this._throughput = value.throughput;
            this._type = value.type;
            this._volumesPerInstance = value.volumesPerInstance;
        }
    }
    get iops() {
        return this.getNumberAttribute('iops');
    }
    set iops(value) {
        this._iops = value;
    }
    resetIops() {
        this._iops = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iopsInput() {
        return this._iops;
    }
    get size() {
        return this.getNumberAttribute('size');
    }
    set size(value) {
        this._size = value;
    }
    // Temporarily expose input value. Use with caution.
    get sizeInput() {
        return this._size;
    }
    get throughput() {
        return this.getNumberAttribute('throughput');
    }
    set throughput(value) {
        this._throughput = value;
    }
    resetThroughput() {
        this._throughput = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get throughputInput() {
        return this._throughput;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get volumesPerInstance() {
        return this.getNumberAttribute('volumes_per_instance');
    }
    set volumesPerInstance(value) {
        this._volumesPerInstance = value;
    }
    resetVolumesPerInstance() {
        this._volumesPerInstance = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumesPerInstanceInput() {
        return this._volumesPerInstance;
    }
}
exports.EmrClusterCoreInstanceGroupEbsConfigOutputReference = EmrClusterCoreInstanceGroupEbsConfigOutputReference;
_v = JSII_RTTI_SYMBOL_1;
EmrClusterCoreInstanceGroupEbsConfigOutputReference[_v] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterCoreInstanceGroupEbsConfigOutputReference", version: "10.0.12" };
class EmrClusterCoreInstanceGroupEbsConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EmrClusterCoreInstanceGroupEbsConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EmrClusterCoreInstanceGroupEbsConfigList = EmrClusterCoreInstanceGroupEbsConfigList;
_w = JSII_RTTI_SYMBOL_1;
EmrClusterCoreInstanceGroupEbsConfigList[_w] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterCoreInstanceGroupEbsConfigList", version: "10.0.12" };
function emrClusterCoreInstanceGroupToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        autoscaling_policy: cdktf.stringToTerraform(struct.autoscalingPolicy),
        bid_price: cdktf.stringToTerraform(struct.bidPrice),
        instance_count: cdktf.numberToTerraform(struct.instanceCount),
        instance_type: cdktf.stringToTerraform(struct.instanceType),
        name: cdktf.stringToTerraform(struct.name),
        ebs_config: cdktf.listMapper(emrClusterCoreInstanceGroupEbsConfigToTerraform, true)(struct.ebsConfig),
    };
}
exports.emrClusterCoreInstanceGroupToTerraform = emrClusterCoreInstanceGroupToTerraform;
class EmrClusterCoreInstanceGroupOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // ebs_config - computed: false, optional: true, required: false
        this._ebsConfig = new EmrClusterCoreInstanceGroupEbsConfigList(this, "ebs_config", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._autoscalingPolicy !== undefined) {
            hasAnyValues = true;
            internalValueResult.autoscalingPolicy = this._autoscalingPolicy;
        }
        if (this._bidPrice !== undefined) {
            hasAnyValues = true;
            internalValueResult.bidPrice = this._bidPrice;
        }
        if (this._instanceCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.instanceCount = this._instanceCount;
        }
        if (this._instanceType !== undefined) {
            hasAnyValues = true;
            internalValueResult.instanceType = this._instanceType;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._ebsConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.ebsConfig = this._ebsConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._autoscalingPolicy = undefined;
            this._bidPrice = undefined;
            this._instanceCount = undefined;
            this._instanceType = undefined;
            this._name = undefined;
            this._ebsConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._autoscalingPolicy = value.autoscalingPolicy;
            this._bidPrice = value.bidPrice;
            this._instanceCount = value.instanceCount;
            this._instanceType = value.instanceType;
            this._name = value.name;
            this._ebsConfig.internalValue = value.ebsConfig;
        }
    }
    get autoscalingPolicy() {
        return this.getStringAttribute('autoscaling_policy');
    }
    set autoscalingPolicy(value) {
        this._autoscalingPolicy = value;
    }
    resetAutoscalingPolicy() {
        this._autoscalingPolicy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoscalingPolicyInput() {
        return this._autoscalingPolicy;
    }
    get bidPrice() {
        return this.getStringAttribute('bid_price');
    }
    set bidPrice(value) {
        this._bidPrice = value;
    }
    resetBidPrice() {
        this._bidPrice = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bidPriceInput() {
        return this._bidPrice;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get instanceCount() {
        return this.getNumberAttribute('instance_count');
    }
    set instanceCount(value) {
        this._instanceCount = value;
    }
    resetInstanceCount() {
        this._instanceCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instanceCountInput() {
        return this._instanceCount;
    }
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    set instanceType(value) {
        this._instanceType = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceTypeInput() {
        return this._instanceType;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get ebsConfig() {
        return this._ebsConfig;
    }
    putEbsConfig(value) {
        this._ebsConfig.internalValue = value;
    }
    resetEbsConfig() {
        this._ebsConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ebsConfigInput() {
        return this._ebsConfig.internalValue;
    }
}
exports.EmrClusterCoreInstanceGroupOutputReference = EmrClusterCoreInstanceGroupOutputReference;
_x = JSII_RTTI_SYMBOL_1;
EmrClusterCoreInstanceGroupOutputReference[_x] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterCoreInstanceGroupOutputReference", version: "10.0.12" };
function emrClusterEc2AttributesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        additional_master_security_groups: cdktf.stringToTerraform(struct.additionalMasterSecurityGroups),
        additional_slave_security_groups: cdktf.stringToTerraform(struct.additionalSlaveSecurityGroups),
        emr_managed_master_security_group: cdktf.stringToTerraform(struct.emrManagedMasterSecurityGroup),
        emr_managed_slave_security_group: cdktf.stringToTerraform(struct.emrManagedSlaveSecurityGroup),
        instance_profile: cdktf.stringToTerraform(struct.instanceProfile),
        key_name: cdktf.stringToTerraform(struct.keyName),
        service_access_security_group: cdktf.stringToTerraform(struct.serviceAccessSecurityGroup),
        subnet_id: cdktf.stringToTerraform(struct.subnetId),
        subnet_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.subnetIds),
    };
}
exports.emrClusterEc2AttributesToTerraform = emrClusterEc2AttributesToTerraform;
class EmrClusterEc2AttributesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._additionalMasterSecurityGroups !== undefined) {
            hasAnyValues = true;
            internalValueResult.additionalMasterSecurityGroups = this._additionalMasterSecurityGroups;
        }
        if (this._additionalSlaveSecurityGroups !== undefined) {
            hasAnyValues = true;
            internalValueResult.additionalSlaveSecurityGroups = this._additionalSlaveSecurityGroups;
        }
        if (this._emrManagedMasterSecurityGroup !== undefined) {
            hasAnyValues = true;
            internalValueResult.emrManagedMasterSecurityGroup = this._emrManagedMasterSecurityGroup;
        }
        if (this._emrManagedSlaveSecurityGroup !== undefined) {
            hasAnyValues = true;
            internalValueResult.emrManagedSlaveSecurityGroup = this._emrManagedSlaveSecurityGroup;
        }
        if (this._instanceProfile !== undefined) {
            hasAnyValues = true;
            internalValueResult.instanceProfile = this._instanceProfile;
        }
        if (this._keyName !== undefined) {
            hasAnyValues = true;
            internalValueResult.keyName = this._keyName;
        }
        if (this._serviceAccessSecurityGroup !== undefined) {
            hasAnyValues = true;
            internalValueResult.serviceAccessSecurityGroup = this._serviceAccessSecurityGroup;
        }
        if (this._subnetId !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetId = this._subnetId;
        }
        if (this._subnetIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetIds = this._subnetIds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._additionalMasterSecurityGroups = undefined;
            this._additionalSlaveSecurityGroups = undefined;
            this._emrManagedMasterSecurityGroup = undefined;
            this._emrManagedSlaveSecurityGroup = undefined;
            this._instanceProfile = undefined;
            this._keyName = undefined;
            this._serviceAccessSecurityGroup = undefined;
            this._subnetId = undefined;
            this._subnetIds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._additionalMasterSecurityGroups = value.additionalMasterSecurityGroups;
            this._additionalSlaveSecurityGroups = value.additionalSlaveSecurityGroups;
            this._emrManagedMasterSecurityGroup = value.emrManagedMasterSecurityGroup;
            this._emrManagedSlaveSecurityGroup = value.emrManagedSlaveSecurityGroup;
            this._instanceProfile = value.instanceProfile;
            this._keyName = value.keyName;
            this._serviceAccessSecurityGroup = value.serviceAccessSecurityGroup;
            this._subnetId = value.subnetId;
            this._subnetIds = value.subnetIds;
        }
    }
    get additionalMasterSecurityGroups() {
        return this.getStringAttribute('additional_master_security_groups');
    }
    set additionalMasterSecurityGroups(value) {
        this._additionalMasterSecurityGroups = value;
    }
    resetAdditionalMasterSecurityGroups() {
        this._additionalMasterSecurityGroups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get additionalMasterSecurityGroupsInput() {
        return this._additionalMasterSecurityGroups;
    }
    get additionalSlaveSecurityGroups() {
        return this.getStringAttribute('additional_slave_security_groups');
    }
    set additionalSlaveSecurityGroups(value) {
        this._additionalSlaveSecurityGroups = value;
    }
    resetAdditionalSlaveSecurityGroups() {
        this._additionalSlaveSecurityGroups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get additionalSlaveSecurityGroupsInput() {
        return this._additionalSlaveSecurityGroups;
    }
    get emrManagedMasterSecurityGroup() {
        return this.getStringAttribute('emr_managed_master_security_group');
    }
    set emrManagedMasterSecurityGroup(value) {
        this._emrManagedMasterSecurityGroup = value;
    }
    resetEmrManagedMasterSecurityGroup() {
        this._emrManagedMasterSecurityGroup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get emrManagedMasterSecurityGroupInput() {
        return this._emrManagedMasterSecurityGroup;
    }
    get emrManagedSlaveSecurityGroup() {
        return this.getStringAttribute('emr_managed_slave_security_group');
    }
    set emrManagedSlaveSecurityGroup(value) {
        this._emrManagedSlaveSecurityGroup = value;
    }
    resetEmrManagedSlaveSecurityGroup() {
        this._emrManagedSlaveSecurityGroup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get emrManagedSlaveSecurityGroupInput() {
        return this._emrManagedSlaveSecurityGroup;
    }
    get instanceProfile() {
        return this.getStringAttribute('instance_profile');
    }
    set instanceProfile(value) {
        this._instanceProfile = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceProfileInput() {
        return this._instanceProfile;
    }
    get keyName() {
        return this.getStringAttribute('key_name');
    }
    set keyName(value) {
        this._keyName = value;
    }
    resetKeyName() {
        this._keyName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyNameInput() {
        return this._keyName;
    }
    get serviceAccessSecurityGroup() {
        return this.getStringAttribute('service_access_security_group');
    }
    set serviceAccessSecurityGroup(value) {
        this._serviceAccessSecurityGroup = value;
    }
    resetServiceAccessSecurityGroup() {
        this._serviceAccessSecurityGroup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceAccessSecurityGroupInput() {
        return this._serviceAccessSecurityGroup;
    }
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    set subnetId(value) {
        this._subnetId = value;
    }
    resetSubnetId() {
        this._subnetId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdInput() {
        return this._subnetId;
    }
    get subnetIds() {
        return cdktf.Fn.tolist(this.getListAttribute('subnet_ids'));
    }
    set subnetIds(value) {
        this._subnetIds = value;
    }
    resetSubnetIds() {
        this._subnetIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdsInput() {
        return this._subnetIds;
    }
}
exports.EmrClusterEc2AttributesOutputReference = EmrClusterEc2AttributesOutputReference;
_y = JSII_RTTI_SYMBOL_1;
EmrClusterEc2AttributesOutputReference[_y] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterEc2AttributesOutputReference", version: "10.0.12" };
function emrClusterKerberosAttributesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ad_domain_join_password: cdktf.stringToTerraform(struct.adDomainJoinPassword),
        ad_domain_join_user: cdktf.stringToTerraform(struct.adDomainJoinUser),
        cross_realm_trust_principal_password: cdktf.stringToTerraform(struct.crossRealmTrustPrincipalPassword),
        kdc_admin_password: cdktf.stringToTerraform(struct.kdcAdminPassword),
        realm: cdktf.stringToTerraform(struct.realm),
    };
}
exports.emrClusterKerberosAttributesToTerraform = emrClusterKerberosAttributesToTerraform;
class EmrClusterKerberosAttributesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._adDomainJoinPassword !== undefined) {
            hasAnyValues = true;
            internalValueResult.adDomainJoinPassword = this._adDomainJoinPassword;
        }
        if (this._adDomainJoinUser !== undefined) {
            hasAnyValues = true;
            internalValueResult.adDomainJoinUser = this._adDomainJoinUser;
        }
        if (this._crossRealmTrustPrincipalPassword !== undefined) {
            hasAnyValues = true;
            internalValueResult.crossRealmTrustPrincipalPassword = this._crossRealmTrustPrincipalPassword;
        }
        if (this._kdcAdminPassword !== undefined) {
            hasAnyValues = true;
            internalValueResult.kdcAdminPassword = this._kdcAdminPassword;
        }
        if (this._realm !== undefined) {
            hasAnyValues = true;
            internalValueResult.realm = this._realm;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._adDomainJoinPassword = undefined;
            this._adDomainJoinUser = undefined;
            this._crossRealmTrustPrincipalPassword = undefined;
            this._kdcAdminPassword = undefined;
            this._realm = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._adDomainJoinPassword = value.adDomainJoinPassword;
            this._adDomainJoinUser = value.adDomainJoinUser;
            this._crossRealmTrustPrincipalPassword = value.crossRealmTrustPrincipalPassword;
            this._kdcAdminPassword = value.kdcAdminPassword;
            this._realm = value.realm;
        }
    }
    get adDomainJoinPassword() {
        return this.getStringAttribute('ad_domain_join_password');
    }
    set adDomainJoinPassword(value) {
        this._adDomainJoinPassword = value;
    }
    resetAdDomainJoinPassword() {
        this._adDomainJoinPassword = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get adDomainJoinPasswordInput() {
        return this._adDomainJoinPassword;
    }
    get adDomainJoinUser() {
        return this.getStringAttribute('ad_domain_join_user');
    }
    set adDomainJoinUser(value) {
        this._adDomainJoinUser = value;
    }
    resetAdDomainJoinUser() {
        this._adDomainJoinUser = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get adDomainJoinUserInput() {
        return this._adDomainJoinUser;
    }
    get crossRealmTrustPrincipalPassword() {
        return this.getStringAttribute('cross_realm_trust_principal_password');
    }
    set crossRealmTrustPrincipalPassword(value) {
        this._crossRealmTrustPrincipalPassword = value;
    }
    resetCrossRealmTrustPrincipalPassword() {
        this._crossRealmTrustPrincipalPassword = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get crossRealmTrustPrincipalPasswordInput() {
        return this._crossRealmTrustPrincipalPassword;
    }
    get kdcAdminPassword() {
        return this.getStringAttribute('kdc_admin_password');
    }
    set kdcAdminPassword(value) {
        this._kdcAdminPassword = value;
    }
    // Temporarily expose input value. Use with caution.
    get kdcAdminPasswordInput() {
        return this._kdcAdminPassword;
    }
    get realm() {
        return this.getStringAttribute('realm');
    }
    set realm(value) {
        this._realm = value;
    }
    // Temporarily expose input value. Use with caution.
    get realmInput() {
        return this._realm;
    }
}
exports.EmrClusterKerberosAttributesOutputReference = EmrClusterKerberosAttributesOutputReference;
_z = JSII_RTTI_SYMBOL_1;
EmrClusterKerberosAttributesOutputReference[_z] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterKerberosAttributesOutputReference", version: "10.0.12" };
function emrClusterMasterInstanceFleetInstanceTypeConfigsConfigurationsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        classification: cdktf.stringToTerraform(struct.classification),
        properties: cdktf.hashMapper(cdktf.stringToTerraform)(struct.properties),
    };
}
exports.emrClusterMasterInstanceFleetInstanceTypeConfigsConfigurationsToTerraform = emrClusterMasterInstanceFleetInstanceTypeConfigsConfigurationsToTerraform;
class EmrClusterMasterInstanceFleetInstanceTypeConfigsConfigurationsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._classification !== undefined) {
            hasAnyValues = true;
            internalValueResult.classification = this._classification;
        }
        if (this._properties !== undefined) {
            hasAnyValues = true;
            internalValueResult.properties = this._properties;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._classification = undefined;
            this._properties = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._classification = value.classification;
            this._properties = value.properties;
        }
    }
    get classification() {
        return this.getStringAttribute('classification');
    }
    set classification(value) {
        this._classification = value;
    }
    resetClassification() {
        this._classification = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get classificationInput() {
        return this._classification;
    }
    get properties() {
        return this.getStringMapAttribute('properties');
    }
    set properties(value) {
        this._properties = value;
    }
    resetProperties() {
        this._properties = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get propertiesInput() {
        return this._properties;
    }
}
exports.EmrClusterMasterInstanceFleetInstanceTypeConfigsConfigurationsOutputReference = EmrClusterMasterInstanceFleetInstanceTypeConfigsConfigurationsOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
EmrClusterMasterInstanceFleetInstanceTypeConfigsConfigurationsOutputReference[_0] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterMasterInstanceFleetInstanceTypeConfigsConfigurationsOutputReference", version: "10.0.12" };
class EmrClusterMasterInstanceFleetInstanceTypeConfigsConfigurationsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EmrClusterMasterInstanceFleetInstanceTypeConfigsConfigurationsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EmrClusterMasterInstanceFleetInstanceTypeConfigsConfigurationsList = EmrClusterMasterInstanceFleetInstanceTypeConfigsConfigurationsList;
_1 = JSII_RTTI_SYMBOL_1;
EmrClusterMasterInstanceFleetInstanceTypeConfigsConfigurationsList[_1] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterMasterInstanceFleetInstanceTypeConfigsConfigurationsList", version: "10.0.12" };
function emrClusterMasterInstanceFleetInstanceTypeConfigsEbsConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        iops: cdktf.numberToTerraform(struct.iops),
        size: cdktf.numberToTerraform(struct.size),
        type: cdktf.stringToTerraform(struct.type),
        volumes_per_instance: cdktf.numberToTerraform(struct.volumesPerInstance),
    };
}
exports.emrClusterMasterInstanceFleetInstanceTypeConfigsEbsConfigToTerraform = emrClusterMasterInstanceFleetInstanceTypeConfigsEbsConfigToTerraform;
class EmrClusterMasterInstanceFleetInstanceTypeConfigsEbsConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._iops !== undefined) {
            hasAnyValues = true;
            internalValueResult.iops = this._iops;
        }
        if (this._size !== undefined) {
            hasAnyValues = true;
            internalValueResult.size = this._size;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._volumesPerInstance !== undefined) {
            hasAnyValues = true;
            internalValueResult.volumesPerInstance = this._volumesPerInstance;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._iops = undefined;
            this._size = undefined;
            this._type = undefined;
            this._volumesPerInstance = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._iops = value.iops;
            this._size = value.size;
            this._type = value.type;
            this._volumesPerInstance = value.volumesPerInstance;
        }
    }
    get iops() {
        return this.getNumberAttribute('iops');
    }
    set iops(value) {
        this._iops = value;
    }
    resetIops() {
        this._iops = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iopsInput() {
        return this._iops;
    }
    get size() {
        return this.getNumberAttribute('size');
    }
    set size(value) {
        this._size = value;
    }
    // Temporarily expose input value. Use with caution.
    get sizeInput() {
        return this._size;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get volumesPerInstance() {
        return this.getNumberAttribute('volumes_per_instance');
    }
    set volumesPerInstance(value) {
        this._volumesPerInstance = value;
    }
    resetVolumesPerInstance() {
        this._volumesPerInstance = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumesPerInstanceInput() {
        return this._volumesPerInstance;
    }
}
exports.EmrClusterMasterInstanceFleetInstanceTypeConfigsEbsConfigOutputReference = EmrClusterMasterInstanceFleetInstanceTypeConfigsEbsConfigOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
EmrClusterMasterInstanceFleetInstanceTypeConfigsEbsConfigOutputReference[_2] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterMasterInstanceFleetInstanceTypeConfigsEbsConfigOutputReference", version: "10.0.12" };
class EmrClusterMasterInstanceFleetInstanceTypeConfigsEbsConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EmrClusterMasterInstanceFleetInstanceTypeConfigsEbsConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EmrClusterMasterInstanceFleetInstanceTypeConfigsEbsConfigList = EmrClusterMasterInstanceFleetInstanceTypeConfigsEbsConfigList;
_3 = JSII_RTTI_SYMBOL_1;
EmrClusterMasterInstanceFleetInstanceTypeConfigsEbsConfigList[_3] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterMasterInstanceFleetInstanceTypeConfigsEbsConfigList", version: "10.0.12" };
function emrClusterMasterInstanceFleetInstanceTypeConfigsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bid_price: cdktf.stringToTerraform(struct.bidPrice),
        bid_price_as_percentage_of_on_demand_price: cdktf.numberToTerraform(struct.bidPriceAsPercentageOfOnDemandPrice),
        instance_type: cdktf.stringToTerraform(struct.instanceType),
        weighted_capacity: cdktf.numberToTerraform(struct.weightedCapacity),
        configurations: cdktf.listMapper(emrClusterMasterInstanceFleetInstanceTypeConfigsConfigurationsToTerraform, true)(struct.configurations),
        ebs_config: cdktf.listMapper(emrClusterMasterInstanceFleetInstanceTypeConfigsEbsConfigToTerraform, true)(struct.ebsConfig),
    };
}
exports.emrClusterMasterInstanceFleetInstanceTypeConfigsToTerraform = emrClusterMasterInstanceFleetInstanceTypeConfigsToTerraform;
class EmrClusterMasterInstanceFleetInstanceTypeConfigsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // configurations - computed: false, optional: true, required: false
        this._configurations = new EmrClusterMasterInstanceFleetInstanceTypeConfigsConfigurationsList(this, "configurations", true);
        // ebs_config - computed: false, optional: true, required: false
        this._ebsConfig = new EmrClusterMasterInstanceFleetInstanceTypeConfigsEbsConfigList(this, "ebs_config", true);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bidPrice !== undefined) {
            hasAnyValues = true;
            internalValueResult.bidPrice = this._bidPrice;
        }
        if (this._bidPriceAsPercentageOfOnDemandPrice !== undefined) {
            hasAnyValues = true;
            internalValueResult.bidPriceAsPercentageOfOnDemandPrice = this._bidPriceAsPercentageOfOnDemandPrice;
        }
        if (this._instanceType !== undefined) {
            hasAnyValues = true;
            internalValueResult.instanceType = this._instanceType;
        }
        if (this._weightedCapacity !== undefined) {
            hasAnyValues = true;
            internalValueResult.weightedCapacity = this._weightedCapacity;
        }
        if (this._configurations?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.configurations = this._configurations?.internalValue;
        }
        if (this._ebsConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.ebsConfig = this._ebsConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._bidPrice = undefined;
            this._bidPriceAsPercentageOfOnDemandPrice = undefined;
            this._instanceType = undefined;
            this._weightedCapacity = undefined;
            this._configurations.internalValue = undefined;
            this._ebsConfig.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._bidPrice = value.bidPrice;
            this._bidPriceAsPercentageOfOnDemandPrice = value.bidPriceAsPercentageOfOnDemandPrice;
            this._instanceType = value.instanceType;
            this._weightedCapacity = value.weightedCapacity;
            this._configurations.internalValue = value.configurations;
            this._ebsConfig.internalValue = value.ebsConfig;
        }
    }
    get bidPrice() {
        return this.getStringAttribute('bid_price');
    }
    set bidPrice(value) {
        this._bidPrice = value;
    }
    resetBidPrice() {
        this._bidPrice = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bidPriceInput() {
        return this._bidPrice;
    }
    get bidPriceAsPercentageOfOnDemandPrice() {
        return this.getNumberAttribute('bid_price_as_percentage_of_on_demand_price');
    }
    set bidPriceAsPercentageOfOnDemandPrice(value) {
        this._bidPriceAsPercentageOfOnDemandPrice = value;
    }
    resetBidPriceAsPercentageOfOnDemandPrice() {
        this._bidPriceAsPercentageOfOnDemandPrice = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bidPriceAsPercentageOfOnDemandPriceInput() {
        return this._bidPriceAsPercentageOfOnDemandPrice;
    }
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    set instanceType(value) {
        this._instanceType = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceTypeInput() {
        return this._instanceType;
    }
    get weightedCapacity() {
        return this.getNumberAttribute('weighted_capacity');
    }
    set weightedCapacity(value) {
        this._weightedCapacity = value;
    }
    resetWeightedCapacity() {
        this._weightedCapacity = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get weightedCapacityInput() {
        return this._weightedCapacity;
    }
    get configurations() {
        return this._configurations;
    }
    putConfigurations(value) {
        this._configurations.internalValue = value;
    }
    resetConfigurations() {
        this._configurations.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get configurationsInput() {
        return this._configurations.internalValue;
    }
    get ebsConfig() {
        return this._ebsConfig;
    }
    putEbsConfig(value) {
        this._ebsConfig.internalValue = value;
    }
    resetEbsConfig() {
        this._ebsConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ebsConfigInput() {
        return this._ebsConfig.internalValue;
    }
}
exports.EmrClusterMasterInstanceFleetInstanceTypeConfigsOutputReference = EmrClusterMasterInstanceFleetInstanceTypeConfigsOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
EmrClusterMasterInstanceFleetInstanceTypeConfigsOutputReference[_4] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterMasterInstanceFleetInstanceTypeConfigsOutputReference", version: "10.0.12" };
class EmrClusterMasterInstanceFleetInstanceTypeConfigsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EmrClusterMasterInstanceFleetInstanceTypeConfigsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EmrClusterMasterInstanceFleetInstanceTypeConfigsList = EmrClusterMasterInstanceFleetInstanceTypeConfigsList;
_5 = JSII_RTTI_SYMBOL_1;
EmrClusterMasterInstanceFleetInstanceTypeConfigsList[_5] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterMasterInstanceFleetInstanceTypeConfigsList", version: "10.0.12" };
function emrClusterMasterInstanceFleetLaunchSpecificationsOnDemandSpecificationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        allocation_strategy: cdktf.stringToTerraform(struct.allocationStrategy),
    };
}
exports.emrClusterMasterInstanceFleetLaunchSpecificationsOnDemandSpecificationToTerraform = emrClusterMasterInstanceFleetLaunchSpecificationsOnDemandSpecificationToTerraform;
class EmrClusterMasterInstanceFleetLaunchSpecificationsOnDemandSpecificationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._allocationStrategy !== undefined) {
            hasAnyValues = true;
            internalValueResult.allocationStrategy = this._allocationStrategy;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._allocationStrategy = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._allocationStrategy = value.allocationStrategy;
        }
    }
    get allocationStrategy() {
        return this.getStringAttribute('allocation_strategy');
    }
    set allocationStrategy(value) {
        this._allocationStrategy = value;
    }
    // Temporarily expose input value. Use with caution.
    get allocationStrategyInput() {
        return this._allocationStrategy;
    }
}
exports.EmrClusterMasterInstanceFleetLaunchSpecificationsOnDemandSpecificationOutputReference = EmrClusterMasterInstanceFleetLaunchSpecificationsOnDemandSpecificationOutputReference;
_6 = JSII_RTTI_SYMBOL_1;
EmrClusterMasterInstanceFleetLaunchSpecificationsOnDemandSpecificationOutputReference[_6] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterMasterInstanceFleetLaunchSpecificationsOnDemandSpecificationOutputReference", version: "10.0.12" };
class EmrClusterMasterInstanceFleetLaunchSpecificationsOnDemandSpecificationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EmrClusterMasterInstanceFleetLaunchSpecificationsOnDemandSpecificationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EmrClusterMasterInstanceFleetLaunchSpecificationsOnDemandSpecificationList = EmrClusterMasterInstanceFleetLaunchSpecificationsOnDemandSpecificationList;
_7 = JSII_RTTI_SYMBOL_1;
EmrClusterMasterInstanceFleetLaunchSpecificationsOnDemandSpecificationList[_7] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterMasterInstanceFleetLaunchSpecificationsOnDemandSpecificationList", version: "10.0.12" };
function emrClusterMasterInstanceFleetLaunchSpecificationsSpotSpecificationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        allocation_strategy: cdktf.stringToTerraform(struct.allocationStrategy),
        block_duration_minutes: cdktf.numberToTerraform(struct.blockDurationMinutes),
        timeout_action: cdktf.stringToTerraform(struct.timeoutAction),
        timeout_duration_minutes: cdktf.numberToTerraform(struct.timeoutDurationMinutes),
    };
}
exports.emrClusterMasterInstanceFleetLaunchSpecificationsSpotSpecificationToTerraform = emrClusterMasterInstanceFleetLaunchSpecificationsSpotSpecificationToTerraform;
class EmrClusterMasterInstanceFleetLaunchSpecificationsSpotSpecificationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._allocationStrategy !== undefined) {
            hasAnyValues = true;
            internalValueResult.allocationStrategy = this._allocationStrategy;
        }
        if (this._blockDurationMinutes !== undefined) {
            hasAnyValues = true;
            internalValueResult.blockDurationMinutes = this._blockDurationMinutes;
        }
        if (this._timeoutAction !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeoutAction = this._timeoutAction;
        }
        if (this._timeoutDurationMinutes !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeoutDurationMinutes = this._timeoutDurationMinutes;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._allocationStrategy = undefined;
            this._blockDurationMinutes = undefined;
            this._timeoutAction = undefined;
            this._timeoutDurationMinutes = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._allocationStrategy = value.allocationStrategy;
            this._blockDurationMinutes = value.blockDurationMinutes;
            this._timeoutAction = value.timeoutAction;
            this._timeoutDurationMinutes = value.timeoutDurationMinutes;
        }
    }
    get allocationStrategy() {
        return this.getStringAttribute('allocation_strategy');
    }
    set allocationStrategy(value) {
        this._allocationStrategy = value;
    }
    // Temporarily expose input value. Use with caution.
    get allocationStrategyInput() {
        return this._allocationStrategy;
    }
    get blockDurationMinutes() {
        return this.getNumberAttribute('block_duration_minutes');
    }
    set blockDurationMinutes(value) {
        this._blockDurationMinutes = value;
    }
    resetBlockDurationMinutes() {
        this._blockDurationMinutes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get blockDurationMinutesInput() {
        return this._blockDurationMinutes;
    }
    get timeoutAction() {
        return this.getStringAttribute('timeout_action');
    }
    set timeoutAction(value) {
        this._timeoutAction = value;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutActionInput() {
        return this._timeoutAction;
    }
    get timeoutDurationMinutes() {
        return this.getNumberAttribute('timeout_duration_minutes');
    }
    set timeoutDurationMinutes(value) {
        this._timeoutDurationMinutes = value;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutDurationMinutesInput() {
        return this._timeoutDurationMinutes;
    }
}
exports.EmrClusterMasterInstanceFleetLaunchSpecificationsSpotSpecificationOutputReference = EmrClusterMasterInstanceFleetLaunchSpecificationsSpotSpecificationOutputReference;
_8 = JSII_RTTI_SYMBOL_1;
EmrClusterMasterInstanceFleetLaunchSpecificationsSpotSpecificationOutputReference[_8] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterMasterInstanceFleetLaunchSpecificationsSpotSpecificationOutputReference", version: "10.0.12" };
class EmrClusterMasterInstanceFleetLaunchSpecificationsSpotSpecificationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EmrClusterMasterInstanceFleetLaunchSpecificationsSpotSpecificationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EmrClusterMasterInstanceFleetLaunchSpecificationsSpotSpecificationList = EmrClusterMasterInstanceFleetLaunchSpecificationsSpotSpecificationList;
_9 = JSII_RTTI_SYMBOL_1;
EmrClusterMasterInstanceFleetLaunchSpecificationsSpotSpecificationList[_9] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterMasterInstanceFleetLaunchSpecificationsSpotSpecificationList", version: "10.0.12" };
function emrClusterMasterInstanceFleetLaunchSpecificationsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        on_demand_specification: cdktf.listMapper(emrClusterMasterInstanceFleetLaunchSpecificationsOnDemandSpecificationToTerraform, true)(struct.onDemandSpecification),
        spot_specification: cdktf.listMapper(emrClusterMasterInstanceFleetLaunchSpecificationsSpotSpecificationToTerraform, true)(struct.spotSpecification),
    };
}
exports.emrClusterMasterInstanceFleetLaunchSpecificationsToTerraform = emrClusterMasterInstanceFleetLaunchSpecificationsToTerraform;
class EmrClusterMasterInstanceFleetLaunchSpecificationsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // on_demand_specification - computed: false, optional: true, required: false
        this._onDemandSpecification = new EmrClusterMasterInstanceFleetLaunchSpecificationsOnDemandSpecificationList(this, "on_demand_specification", false);
        // spot_specification - computed: false, optional: true, required: false
        this._spotSpecification = new EmrClusterMasterInstanceFleetLaunchSpecificationsSpotSpecificationList(this, "spot_specification", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._onDemandSpecification?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.onDemandSpecification = this._onDemandSpecification?.internalValue;
        }
        if (this._spotSpecification?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.spotSpecification = this._spotSpecification?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._onDemandSpecification.internalValue = undefined;
            this._spotSpecification.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._onDemandSpecification.internalValue = value.onDemandSpecification;
            this._spotSpecification.internalValue = value.spotSpecification;
        }
    }
    get onDemandSpecification() {
        return this._onDemandSpecification;
    }
    putOnDemandSpecification(value) {
        this._onDemandSpecification.internalValue = value;
    }
    resetOnDemandSpecification() {
        this._onDemandSpecification.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get onDemandSpecificationInput() {
        return this._onDemandSpecification.internalValue;
    }
    get spotSpecification() {
        return this._spotSpecification;
    }
    putSpotSpecification(value) {
        this._spotSpecification.internalValue = value;
    }
    resetSpotSpecification() {
        this._spotSpecification.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get spotSpecificationInput() {
        return this._spotSpecification.internalValue;
    }
}
exports.EmrClusterMasterInstanceFleetLaunchSpecificationsOutputReference = EmrClusterMasterInstanceFleetLaunchSpecificationsOutputReference;
_10 = JSII_RTTI_SYMBOL_1;
EmrClusterMasterInstanceFleetLaunchSpecificationsOutputReference[_10] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterMasterInstanceFleetLaunchSpecificationsOutputReference", version: "10.0.12" };
function emrClusterMasterInstanceFleetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        target_on_demand_capacity: cdktf.numberToTerraform(struct.targetOnDemandCapacity),
        target_spot_capacity: cdktf.numberToTerraform(struct.targetSpotCapacity),
        instance_type_configs: cdktf.listMapper(emrClusterMasterInstanceFleetInstanceTypeConfigsToTerraform, true)(struct.instanceTypeConfigs),
        launch_specifications: emrClusterMasterInstanceFleetLaunchSpecificationsToTerraform(struct.launchSpecifications),
    };
}
exports.emrClusterMasterInstanceFleetToTerraform = emrClusterMasterInstanceFleetToTerraform;
class EmrClusterMasterInstanceFleetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // instance_type_configs - computed: false, optional: true, required: false
        this._instanceTypeConfigs = new EmrClusterMasterInstanceFleetInstanceTypeConfigsList(this, "instance_type_configs", true);
        // launch_specifications - computed: false, optional: true, required: false
        this._launchSpecifications = new EmrClusterMasterInstanceFleetLaunchSpecificationsOutputReference(this, "launch_specifications");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._targetOnDemandCapacity !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetOnDemandCapacity = this._targetOnDemandCapacity;
        }
        if (this._targetSpotCapacity !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetSpotCapacity = this._targetSpotCapacity;
        }
        if (this._instanceTypeConfigs?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.instanceTypeConfigs = this._instanceTypeConfigs?.internalValue;
        }
        if (this._launchSpecifications?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.launchSpecifications = this._launchSpecifications?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._name = undefined;
            this._targetOnDemandCapacity = undefined;
            this._targetSpotCapacity = undefined;
            this._instanceTypeConfigs.internalValue = undefined;
            this._launchSpecifications.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._name = value.name;
            this._targetOnDemandCapacity = value.targetOnDemandCapacity;
            this._targetSpotCapacity = value.targetSpotCapacity;
            this._instanceTypeConfigs.internalValue = value.instanceTypeConfigs;
            this._launchSpecifications.internalValue = value.launchSpecifications;
        }
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // provisioned_on_demand_capacity - computed: true, optional: false, required: false
    get provisionedOnDemandCapacity() {
        return this.getNumberAttribute('provisioned_on_demand_capacity');
    }
    // provisioned_spot_capacity - computed: true, optional: false, required: false
    get provisionedSpotCapacity() {
        return this.getNumberAttribute('provisioned_spot_capacity');
    }
    get targetOnDemandCapacity() {
        return this.getNumberAttribute('target_on_demand_capacity');
    }
    set targetOnDemandCapacity(value) {
        this._targetOnDemandCapacity = value;
    }
    resetTargetOnDemandCapacity() {
        this._targetOnDemandCapacity = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetOnDemandCapacityInput() {
        return this._targetOnDemandCapacity;
    }
    get targetSpotCapacity() {
        return this.getNumberAttribute('target_spot_capacity');
    }
    set targetSpotCapacity(value) {
        this._targetSpotCapacity = value;
    }
    resetTargetSpotCapacity() {
        this._targetSpotCapacity = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetSpotCapacityInput() {
        return this._targetSpotCapacity;
    }
    get instanceTypeConfigs() {
        return this._instanceTypeConfigs;
    }
    putInstanceTypeConfigs(value) {
        this._instanceTypeConfigs.internalValue = value;
    }
    resetInstanceTypeConfigs() {
        this._instanceTypeConfigs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instanceTypeConfigsInput() {
        return this._instanceTypeConfigs.internalValue;
    }
    get launchSpecifications() {
        return this._launchSpecifications;
    }
    putLaunchSpecifications(value) {
        this._launchSpecifications.internalValue = value;
    }
    resetLaunchSpecifications() {
        this._launchSpecifications.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get launchSpecificationsInput() {
        return this._launchSpecifications.internalValue;
    }
}
exports.EmrClusterMasterInstanceFleetOutputReference = EmrClusterMasterInstanceFleetOutputReference;
_11 = JSII_RTTI_SYMBOL_1;
EmrClusterMasterInstanceFleetOutputReference[_11] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterMasterInstanceFleetOutputReference", version: "10.0.12" };
function emrClusterMasterInstanceGroupEbsConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        iops: cdktf.numberToTerraform(struct.iops),
        size: cdktf.numberToTerraform(struct.size),
        throughput: cdktf.numberToTerraform(struct.throughput),
        type: cdktf.stringToTerraform(struct.type),
        volumes_per_instance: cdktf.numberToTerraform(struct.volumesPerInstance),
    };
}
exports.emrClusterMasterInstanceGroupEbsConfigToTerraform = emrClusterMasterInstanceGroupEbsConfigToTerraform;
class EmrClusterMasterInstanceGroupEbsConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._iops !== undefined) {
            hasAnyValues = true;
            internalValueResult.iops = this._iops;
        }
        if (this._size !== undefined) {
            hasAnyValues = true;
            internalValueResult.size = this._size;
        }
        if (this._throughput !== undefined) {
            hasAnyValues = true;
            internalValueResult.throughput = this._throughput;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._volumesPerInstance !== undefined) {
            hasAnyValues = true;
            internalValueResult.volumesPerInstance = this._volumesPerInstance;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._iops = undefined;
            this._size = undefined;
            this._throughput = undefined;
            this._type = undefined;
            this._volumesPerInstance = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._iops = value.iops;
            this._size = value.size;
            this._throughput = value.throughput;
            this._type = value.type;
            this._volumesPerInstance = value.volumesPerInstance;
        }
    }
    get iops() {
        return this.getNumberAttribute('iops');
    }
    set iops(value) {
        this._iops = value;
    }
    resetIops() {
        this._iops = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iopsInput() {
        return this._iops;
    }
    get size() {
        return this.getNumberAttribute('size');
    }
    set size(value) {
        this._size = value;
    }
    // Temporarily expose input value. Use with caution.
    get sizeInput() {
        return this._size;
    }
    get throughput() {
        return this.getNumberAttribute('throughput');
    }
    set throughput(value) {
        this._throughput = value;
    }
    resetThroughput() {
        this._throughput = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get throughputInput() {
        return this._throughput;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get volumesPerInstance() {
        return this.getNumberAttribute('volumes_per_instance');
    }
    set volumesPerInstance(value) {
        this._volumesPerInstance = value;
    }
    resetVolumesPerInstance() {
        this._volumesPerInstance = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumesPerInstanceInput() {
        return this._volumesPerInstance;
    }
}
exports.EmrClusterMasterInstanceGroupEbsConfigOutputReference = EmrClusterMasterInstanceGroupEbsConfigOutputReference;
_12 = JSII_RTTI_SYMBOL_1;
EmrClusterMasterInstanceGroupEbsConfigOutputReference[_12] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterMasterInstanceGroupEbsConfigOutputReference", version: "10.0.12" };
class EmrClusterMasterInstanceGroupEbsConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EmrClusterMasterInstanceGroupEbsConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EmrClusterMasterInstanceGroupEbsConfigList = EmrClusterMasterInstanceGroupEbsConfigList;
_13 = JSII_RTTI_SYMBOL_1;
EmrClusterMasterInstanceGroupEbsConfigList[_13] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterMasterInstanceGroupEbsConfigList", version: "10.0.12" };
function emrClusterMasterInstanceGroupToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bid_price: cdktf.stringToTerraform(struct.bidPrice),
        instance_count: cdktf.numberToTerraform(struct.instanceCount),
        instance_type: cdktf.stringToTerraform(struct.instanceType),
        name: cdktf.stringToTerraform(struct.name),
        ebs_config: cdktf.listMapper(emrClusterMasterInstanceGroupEbsConfigToTerraform, true)(struct.ebsConfig),
    };
}
exports.emrClusterMasterInstanceGroupToTerraform = emrClusterMasterInstanceGroupToTerraform;
class EmrClusterMasterInstanceGroupOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // ebs_config - computed: false, optional: true, required: false
        this._ebsConfig = new EmrClusterMasterInstanceGroupEbsConfigList(this, "ebs_config", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bidPrice !== undefined) {
            hasAnyValues = true;
            internalValueResult.bidPrice = this._bidPrice;
        }
        if (this._instanceCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.instanceCount = this._instanceCount;
        }
        if (this._instanceType !== undefined) {
            hasAnyValues = true;
            internalValueResult.instanceType = this._instanceType;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._ebsConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.ebsConfig = this._ebsConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bidPrice = undefined;
            this._instanceCount = undefined;
            this._instanceType = undefined;
            this._name = undefined;
            this._ebsConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bidPrice = value.bidPrice;
            this._instanceCount = value.instanceCount;
            this._instanceType = value.instanceType;
            this._name = value.name;
            this._ebsConfig.internalValue = value.ebsConfig;
        }
    }
    get bidPrice() {
        return this.getStringAttribute('bid_price');
    }
    set bidPrice(value) {
        this._bidPrice = value;
    }
    resetBidPrice() {
        this._bidPrice = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bidPriceInput() {
        return this._bidPrice;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get instanceCount() {
        return this.getNumberAttribute('instance_count');
    }
    set instanceCount(value) {
        this._instanceCount = value;
    }
    resetInstanceCount() {
        this._instanceCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instanceCountInput() {
        return this._instanceCount;
    }
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    set instanceType(value) {
        this._instanceType = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceTypeInput() {
        return this._instanceType;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get ebsConfig() {
        return this._ebsConfig;
    }
    putEbsConfig(value) {
        this._ebsConfig.internalValue = value;
    }
    resetEbsConfig() {
        this._ebsConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ebsConfigInput() {
        return this._ebsConfig.internalValue;
    }
}
exports.EmrClusterMasterInstanceGroupOutputReference = EmrClusterMasterInstanceGroupOutputReference;
_14 = JSII_RTTI_SYMBOL_1;
EmrClusterMasterInstanceGroupOutputReference[_14] = { fqn: "@cdktf/provider-aws.emrCluster.EmrClusterMasterInstanceGroupOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/emr_cluster aws_emr_cluster}
*/
class EmrCluster extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/emr_cluster aws_emr_cluster} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options EmrClusterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_emr_cluster',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // step - computed: true, optional: true, required: false
        this._step = new EmrClusterStepList(this, "step", false);
        // auto_termination_policy - computed: false, optional: true, required: false
        this._autoTerminationPolicy = new EmrClusterAutoTerminationPolicyOutputReference(this, "auto_termination_policy");
        // bootstrap_action - computed: false, optional: true, required: false
        this._bootstrapAction = new EmrClusterBootstrapActionList(this, "bootstrap_action", false);
        // core_instance_fleet - computed: false, optional: true, required: false
        this._coreInstanceFleet = new EmrClusterCoreInstanceFleetOutputReference(this, "core_instance_fleet");
        // core_instance_group - computed: false, optional: true, required: false
        this._coreInstanceGroup = new EmrClusterCoreInstanceGroupOutputReference(this, "core_instance_group");
        // ec2_attributes - computed: false, optional: true, required: false
        this._ec2Attributes = new EmrClusterEc2AttributesOutputReference(this, "ec2_attributes");
        // kerberos_attributes - computed: false, optional: true, required: false
        this._kerberosAttributes = new EmrClusterKerberosAttributesOutputReference(this, "kerberos_attributes");
        // master_instance_fleet - computed: false, optional: true, required: false
        this._masterInstanceFleet = new EmrClusterMasterInstanceFleetOutputReference(this, "master_instance_fleet");
        // master_instance_group - computed: false, optional: true, required: false
        this._masterInstanceGroup = new EmrClusterMasterInstanceGroupOutputReference(this, "master_instance_group");
        this._additionalInfo = config.additionalInfo;
        this._applications = config.applications;
        this._autoscalingRole = config.autoscalingRole;
        this._configurations = config.configurations;
        this._configurationsJson = config.configurationsJson;
        this._customAmiId = config.customAmiId;
        this._ebsRootVolumeSize = config.ebsRootVolumeSize;
        this._id = config.id;
        this._keepJobFlowAliveWhenNoSteps = config.keepJobFlowAliveWhenNoSteps;
        this._listStepsStates = config.listStepsStates;
        this._logEncryptionKmsKeyId = config.logEncryptionKmsKeyId;
        this._logUri = config.logUri;
        this._name = config.name;
        this._releaseLabel = config.releaseLabel;
        this._scaleDownBehavior = config.scaleDownBehavior;
        this._securityConfiguration = config.securityConfiguration;
        this._serviceRole = config.serviceRole;
        this._step.internalValue = config.step;
        this._stepConcurrencyLevel = config.stepConcurrencyLevel;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._terminationProtection = config.terminationProtection;
        this._visibleToAllUsers = config.visibleToAllUsers;
        this._autoTerminationPolicy.internalValue = config.autoTerminationPolicy;
        this._bootstrapAction.internalValue = config.bootstrapAction;
        this._coreInstanceFleet.internalValue = config.coreInstanceFleet;
        this._coreInstanceGroup.internalValue = config.coreInstanceGroup;
        this._ec2Attributes.internalValue = config.ec2Attributes;
        this._kerberosAttributes.internalValue = config.kerberosAttributes;
        this._masterInstanceFleet.internalValue = config.masterInstanceFleet;
        this._masterInstanceGroup.internalValue = config.masterInstanceGroup;
    }
    get additionalInfo() {
        return this.getStringAttribute('additional_info');
    }
    set additionalInfo(value) {
        this._additionalInfo = value;
    }
    resetAdditionalInfo() {
        this._additionalInfo = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get additionalInfoInput() {
        return this._additionalInfo;
    }
    get applications() {
        return cdktf.Fn.tolist(this.getListAttribute('applications'));
    }
    set applications(value) {
        this._applications = value;
    }
    resetApplications() {
        this._applications = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get applicationsInput() {
        return this._applications;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get autoscalingRole() {
        return this.getStringAttribute('autoscaling_role');
    }
    set autoscalingRole(value) {
        this._autoscalingRole = value;
    }
    resetAutoscalingRole() {
        this._autoscalingRole = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoscalingRoleInput() {
        return this._autoscalingRole;
    }
    // cluster_state - computed: true, optional: false, required: false
    get clusterState() {
        return this.getStringAttribute('cluster_state');
    }
    get configurations() {
        return this.getStringAttribute('configurations');
    }
    set configurations(value) {
        this._configurations = value;
    }
    resetConfigurations() {
        this._configurations = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get configurationsInput() {
        return this._configurations;
    }
    get configurationsJson() {
        return this.getStringAttribute('configurations_json');
    }
    set configurationsJson(value) {
        this._configurationsJson = value;
    }
    resetConfigurationsJson() {
        this._configurationsJson = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get configurationsJsonInput() {
        return this._configurationsJson;
    }
    get customAmiId() {
        return this.getStringAttribute('custom_ami_id');
    }
    set customAmiId(value) {
        this._customAmiId = value;
    }
    resetCustomAmiId() {
        this._customAmiId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get customAmiIdInput() {
        return this._customAmiId;
    }
    get ebsRootVolumeSize() {
        return this.getNumberAttribute('ebs_root_volume_size');
    }
    set ebsRootVolumeSize(value) {
        this._ebsRootVolumeSize = value;
    }
    resetEbsRootVolumeSize() {
        this._ebsRootVolumeSize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ebsRootVolumeSizeInput() {
        return this._ebsRootVolumeSize;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get keepJobFlowAliveWhenNoSteps() {
        return this.getBooleanAttribute('keep_job_flow_alive_when_no_steps');
    }
    set keepJobFlowAliveWhenNoSteps(value) {
        this._keepJobFlowAliveWhenNoSteps = value;
    }
    resetKeepJobFlowAliveWhenNoSteps() {
        this._keepJobFlowAliveWhenNoSteps = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keepJobFlowAliveWhenNoStepsInput() {
        return this._keepJobFlowAliveWhenNoSteps;
    }
    get listStepsStates() {
        return cdktf.Fn.tolist(this.getListAttribute('list_steps_states'));
    }
    set listStepsStates(value) {
        this._listStepsStates = value;
    }
    resetListStepsStates() {
        this._listStepsStates = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get listStepsStatesInput() {
        return this._listStepsStates;
    }
    get logEncryptionKmsKeyId() {
        return this.getStringAttribute('log_encryption_kms_key_id');
    }
    set logEncryptionKmsKeyId(value) {
        this._logEncryptionKmsKeyId = value;
    }
    resetLogEncryptionKmsKeyId() {
        this._logEncryptionKmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get logEncryptionKmsKeyIdInput() {
        return this._logEncryptionKmsKeyId;
    }
    get logUri() {
        return this.getStringAttribute('log_uri');
    }
    set logUri(value) {
        this._logUri = value;
    }
    resetLogUri() {
        this._logUri = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get logUriInput() {
        return this._logUri;
    }
    // master_public_dns - computed: true, optional: false, required: false
    get masterPublicDns() {
        return this.getStringAttribute('master_public_dns');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get releaseLabel() {
        return this.getStringAttribute('release_label');
    }
    set releaseLabel(value) {
        this._releaseLabel = value;
    }
    // Temporarily expose input value. Use with caution.
    get releaseLabelInput() {
        return this._releaseLabel;
    }
    get scaleDownBehavior() {
        return this.getStringAttribute('scale_down_behavior');
    }
    set scaleDownBehavior(value) {
        this._scaleDownBehavior = value;
    }
    resetScaleDownBehavior() {
        this._scaleDownBehavior = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scaleDownBehaviorInput() {
        return this._scaleDownBehavior;
    }
    get securityConfiguration() {
        return this.getStringAttribute('security_configuration');
    }
    set securityConfiguration(value) {
        this._securityConfiguration = value;
    }
    resetSecurityConfiguration() {
        this._securityConfiguration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityConfigurationInput() {
        return this._securityConfiguration;
    }
    get serviceRole() {
        return this.getStringAttribute('service_role');
    }
    set serviceRole(value) {
        this._serviceRole = value;
    }
    // Temporarily expose input value. Use with caution.
    get serviceRoleInput() {
        return this._serviceRole;
    }
    get step() {
        return this._step;
    }
    putStep(value) {
        this._step.internalValue = value;
    }
    resetStep() {
        this._step.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get stepInput() {
        return this._step.internalValue;
    }
    get stepConcurrencyLevel() {
        return this.getNumberAttribute('step_concurrency_level');
    }
    set stepConcurrencyLevel(value) {
        this._stepConcurrencyLevel = value;
    }
    resetStepConcurrencyLevel() {
        this._stepConcurrencyLevel = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get stepConcurrencyLevelInput() {
        return this._stepConcurrencyLevel;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get terminationProtection() {
        return this.getBooleanAttribute('termination_protection');
    }
    set terminationProtection(value) {
        this._terminationProtection = value;
    }
    resetTerminationProtection() {
        this._terminationProtection = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get terminationProtectionInput() {
        return this._terminationProtection;
    }
    get visibleToAllUsers() {
        return this.getBooleanAttribute('visible_to_all_users');
    }
    set visibleToAllUsers(value) {
        this._visibleToAllUsers = value;
    }
    resetVisibleToAllUsers() {
        this._visibleToAllUsers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get visibleToAllUsersInput() {
        return this._visibleToAllUsers;
    }
    get autoTerminationPolicy() {
        return this._autoTerminationPolicy;
    }
    putAutoTerminationPolicy(value) {
        this._autoTerminationPolicy.internalValue = value;
    }
    resetAutoTerminationPolicy() {
        this._autoTerminationPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoTerminationPolicyInput() {
        return this._autoTerminationPolicy.internalValue;
    }
    get bootstrapAction() {
        return this._bootstrapAction;
    }
    putBootstrapAction(value) {
        this._bootstrapAction.internalValue = value;
    }
    resetBootstrapAction() {
        this._bootstrapAction.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bootstrapActionInput() {
        return this._bootstrapAction.internalValue;
    }
    get coreInstanceFleet() {
        return this._coreInstanceFleet;
    }
    putCoreInstanceFleet(value) {
        this._coreInstanceFleet.internalValue = value;
    }
    resetCoreInstanceFleet() {
        this._coreInstanceFleet.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get coreInstanceFleetInput() {
        return this._coreInstanceFleet.internalValue;
    }
    get coreInstanceGroup() {
        return this._coreInstanceGroup;
    }
    putCoreInstanceGroup(value) {
        this._coreInstanceGroup.internalValue = value;
    }
    resetCoreInstanceGroup() {
        this._coreInstanceGroup.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get coreInstanceGroupInput() {
        return this._coreInstanceGroup.internalValue;
    }
    get ec2Attributes() {
        return this._ec2Attributes;
    }
    putEc2Attributes(value) {
        this._ec2Attributes.internalValue = value;
    }
    resetEc2Attributes() {
        this._ec2Attributes.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ec2AttributesInput() {
        return this._ec2Attributes.internalValue;
    }
    get kerberosAttributes() {
        return this._kerberosAttributes;
    }
    putKerberosAttributes(value) {
        this._kerberosAttributes.internalValue = value;
    }
    resetKerberosAttributes() {
        this._kerberosAttributes.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kerberosAttributesInput() {
        return this._kerberosAttributes.internalValue;
    }
    get masterInstanceFleet() {
        return this._masterInstanceFleet;
    }
    putMasterInstanceFleet(value) {
        this._masterInstanceFleet.internalValue = value;
    }
    resetMasterInstanceFleet() {
        this._masterInstanceFleet.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get masterInstanceFleetInput() {
        return this._masterInstanceFleet.internalValue;
    }
    get masterInstanceGroup() {
        return this._masterInstanceGroup;
    }
    putMasterInstanceGroup(value) {
        this._masterInstanceGroup.internalValue = value;
    }
    resetMasterInstanceGroup() {
        this._masterInstanceGroup.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get masterInstanceGroupInput() {
        return this._masterInstanceGroup.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            additional_info: cdktf.stringToTerraform(this._additionalInfo),
            applications: cdktf.listMapper(cdktf.stringToTerraform, false)(this._applications),
            autoscaling_role: cdktf.stringToTerraform(this._autoscalingRole),
            configurations: cdktf.stringToTerraform(this._configurations),
            configurations_json: cdktf.stringToTerraform(this._configurationsJson),
            custom_ami_id: cdktf.stringToTerraform(this._customAmiId),
            ebs_root_volume_size: cdktf.numberToTerraform(this._ebsRootVolumeSize),
            id: cdktf.stringToTerraform(this._id),
            keep_job_flow_alive_when_no_steps: cdktf.booleanToTerraform(this._keepJobFlowAliveWhenNoSteps),
            list_steps_states: cdktf.listMapper(cdktf.stringToTerraform, false)(this._listStepsStates),
            log_encryption_kms_key_id: cdktf.stringToTerraform(this._logEncryptionKmsKeyId),
            log_uri: cdktf.stringToTerraform(this._logUri),
            name: cdktf.stringToTerraform(this._name),
            release_label: cdktf.stringToTerraform(this._releaseLabel),
            scale_down_behavior: cdktf.stringToTerraform(this._scaleDownBehavior),
            security_configuration: cdktf.stringToTerraform(this._securityConfiguration),
            service_role: cdktf.stringToTerraform(this._serviceRole),
            step: cdktf.listMapper(emrClusterStepToTerraform, false)(this._step.internalValue),
            step_concurrency_level: cdktf.numberToTerraform(this._stepConcurrencyLevel),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            termination_protection: cdktf.booleanToTerraform(this._terminationProtection),
            visible_to_all_users: cdktf.booleanToTerraform(this._visibleToAllUsers),
            auto_termination_policy: emrClusterAutoTerminationPolicyToTerraform(this._autoTerminationPolicy.internalValue),
            bootstrap_action: cdktf.listMapper(emrClusterBootstrapActionToTerraform, true)(this._bootstrapAction.internalValue),
            core_instance_fleet: emrClusterCoreInstanceFleetToTerraform(this._coreInstanceFleet.internalValue),
            core_instance_group: emrClusterCoreInstanceGroupToTerraform(this._coreInstanceGroup.internalValue),
            ec2_attributes: emrClusterEc2AttributesToTerraform(this._ec2Attributes.internalValue),
            kerberos_attributes: emrClusterKerberosAttributesToTerraform(this._kerberosAttributes.internalValue),
            master_instance_fleet: emrClusterMasterInstanceFleetToTerraform(this._masterInstanceFleet.internalValue),
            master_instance_group: emrClusterMasterInstanceGroupToTerraform(this._masterInstanceGroup.internalValue),
        };
    }
}
exports.EmrCluster = EmrCluster;
_15 = JSII_RTTI_SYMBOL_1;
EmrCluster[_15] = { fqn: "@cdktf/provider-aws.emrCluster.EmrCluster", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
EmrCluster.tfResourceType = "aws_emr_cluster";
