// @cdktf/provider-aws/lib/qldb-stream/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.QldbStream = exports.QldbStreamKinesisConfigurationOutputReference = exports.qldbStreamKinesisConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function qldbStreamKinesisConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        aggregation_enabled: cdktf.booleanToTerraform(struct.aggregationEnabled),
        stream_arn: cdktf.stringToTerraform(struct.streamArn),
    };
}
exports.qldbStreamKinesisConfigurationToTerraform = qldbStreamKinesisConfigurationToTerraform;
class QldbStreamKinesisConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._aggregationEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.aggregationEnabled = this._aggregationEnabled;
        }
        if (this._streamArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.streamArn = this._streamArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._aggregationEnabled = undefined;
            this._streamArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._aggregationEnabled = value.aggregationEnabled;
            this._streamArn = value.streamArn;
        }
    }
    get aggregationEnabled() {
        return this.getBooleanAttribute('aggregation_enabled');
    }
    set aggregationEnabled(value) {
        this._aggregationEnabled = value;
    }
    resetAggregationEnabled() {
        this._aggregationEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get aggregationEnabledInput() {
        return this._aggregationEnabled;
    }
    get streamArn() {
        return this.getStringAttribute('stream_arn');
    }
    set streamArn(value) {
        this._streamArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get streamArnInput() {
        return this._streamArn;
    }
}
exports.QldbStreamKinesisConfigurationOutputReference = QldbStreamKinesisConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
QldbStreamKinesisConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.qldbStream.QldbStreamKinesisConfigurationOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/qldb_stream aws_qldb_stream}
*/
class QldbStream extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/qldb_stream aws_qldb_stream} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options QldbStreamConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_qldb_stream',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // kinesis_configuration - computed: false, optional: false, required: true
        this._kinesisConfiguration = new QldbStreamKinesisConfigurationOutputReference(this, "kinesis_configuration");
        this._exclusiveEndTime = config.exclusiveEndTime;
        this._id = config.id;
        this._inclusiveStartTime = config.inclusiveStartTime;
        this._ledgerName = config.ledgerName;
        this._roleArn = config.roleArn;
        this._streamName = config.streamName;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._kinesisConfiguration.internalValue = config.kinesisConfiguration;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get exclusiveEndTime() {
        return this.getStringAttribute('exclusive_end_time');
    }
    set exclusiveEndTime(value) {
        this._exclusiveEndTime = value;
    }
    resetExclusiveEndTime() {
        this._exclusiveEndTime = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exclusiveEndTimeInput() {
        return this._exclusiveEndTime;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get inclusiveStartTime() {
        return this.getStringAttribute('inclusive_start_time');
    }
    set inclusiveStartTime(value) {
        this._inclusiveStartTime = value;
    }
    // Temporarily expose input value. Use with caution.
    get inclusiveStartTimeInput() {
        return this._inclusiveStartTime;
    }
    get ledgerName() {
        return this.getStringAttribute('ledger_name');
    }
    set ledgerName(value) {
        this._ledgerName = value;
    }
    // Temporarily expose input value. Use with caution.
    get ledgerNameInput() {
        return this._ledgerName;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get streamName() {
        return this.getStringAttribute('stream_name');
    }
    set streamName(value) {
        this._streamName = value;
    }
    // Temporarily expose input value. Use with caution.
    get streamNameInput() {
        return this._streamName;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get kinesisConfiguration() {
        return this._kinesisConfiguration;
    }
    putKinesisConfiguration(value) {
        this._kinesisConfiguration.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get kinesisConfigurationInput() {
        return this._kinesisConfiguration.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            exclusive_end_time: cdktf.stringToTerraform(this._exclusiveEndTime),
            id: cdktf.stringToTerraform(this._id),
            inclusive_start_time: cdktf.stringToTerraform(this._inclusiveStartTime),
            ledger_name: cdktf.stringToTerraform(this._ledgerName),
            role_arn: cdktf.stringToTerraform(this._roleArn),
            stream_name: cdktf.stringToTerraform(this._streamName),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            kinesis_configuration: qldbStreamKinesisConfigurationToTerraform(this._kinesisConfiguration.internalValue),
        };
    }
}
exports.QldbStream = QldbStream;
_b = JSII_RTTI_SYMBOL_1;
QldbStream[_b] = { fqn: "@cdktf/provider-aws.qldbStream.QldbStream", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
QldbStream.tfResourceType = "aws_qldb_stream";
