// @cdktf/provider-aws/lib/api-gateway-request-validator/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiGatewayRequestValidator = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/api_gateway_request_validator aws_api_gateway_request_validator}
*/
class ApiGatewayRequestValidator extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/api_gateway_request_validator aws_api_gateway_request_validator} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ApiGatewayRequestValidatorConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_api_gateway_request_validator',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._name = config.name;
        this._restApiId = config.restApiId;
        this._validateRequestBody = config.validateRequestBody;
        this._validateRequestParameters = config.validateRequestParameters;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get restApiId() {
        return this.getStringAttribute('rest_api_id');
    }
    set restApiId(value) {
        this._restApiId = value;
    }
    // Temporarily expose input value. Use with caution.
    get restApiIdInput() {
        return this._restApiId;
    }
    get validateRequestBody() {
        return this.getBooleanAttribute('validate_request_body');
    }
    set validateRequestBody(value) {
        this._validateRequestBody = value;
    }
    resetValidateRequestBody() {
        this._validateRequestBody = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get validateRequestBodyInput() {
        return this._validateRequestBody;
    }
    get validateRequestParameters() {
        return this.getBooleanAttribute('validate_request_parameters');
    }
    set validateRequestParameters(value) {
        this._validateRequestParameters = value;
    }
    resetValidateRequestParameters() {
        this._validateRequestParameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get validateRequestParametersInput() {
        return this._validateRequestParameters;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            rest_api_id: cdktf.stringToTerraform(this._restApiId),
            validate_request_body: cdktf.booleanToTerraform(this._validateRequestBody),
            validate_request_parameters: cdktf.booleanToTerraform(this._validateRequestParameters),
        };
    }
}
exports.ApiGatewayRequestValidator = ApiGatewayRequestValidator;
_a = JSII_RTTI_SYMBOL_1;
ApiGatewayRequestValidator[_a] = { fqn: "@cdktf/provider-aws.apiGatewayRequestValidator.ApiGatewayRequestValidator", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
ApiGatewayRequestValidator.tfResourceType = "aws_api_gateway_request_validator";
