// @cdktf/provider-aws/lib/kinesis-stream/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.KinesisStream = exports.KinesisStreamTimeoutsOutputReference = exports.kinesisStreamTimeoutsToTerraform = exports.KinesisStreamStreamModeDetailsOutputReference = exports.kinesisStreamStreamModeDetailsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function kinesisStreamStreamModeDetailsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        stream_mode: cdktf.stringToTerraform(struct.streamMode),
    };
}
exports.kinesisStreamStreamModeDetailsToTerraform = kinesisStreamStreamModeDetailsToTerraform;
class KinesisStreamStreamModeDetailsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._streamMode !== undefined) {
            hasAnyValues = true;
            internalValueResult.streamMode = this._streamMode;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._streamMode = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._streamMode = value.streamMode;
        }
    }
    get streamMode() {
        return this.getStringAttribute('stream_mode');
    }
    set streamMode(value) {
        this._streamMode = value;
    }
    // Temporarily expose input value. Use with caution.
    get streamModeInput() {
        return this._streamMode;
    }
}
exports.KinesisStreamStreamModeDetailsOutputReference = KinesisStreamStreamModeDetailsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
KinesisStreamStreamModeDetailsOutputReference[_a] = { fqn: "@cdktf/provider-aws.kinesisStream.KinesisStreamStreamModeDetailsOutputReference", version: "10.0.12" };
function kinesisStreamTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.kinesisStreamTimeoutsToTerraform = kinesisStreamTimeoutsToTerraform;
class KinesisStreamTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.KinesisStreamTimeoutsOutputReference = KinesisStreamTimeoutsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
KinesisStreamTimeoutsOutputReference[_b] = { fqn: "@cdktf/provider-aws.kinesisStream.KinesisStreamTimeoutsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/kinesis_stream aws_kinesis_stream}
*/
class KinesisStream extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/kinesis_stream aws_kinesis_stream} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options KinesisStreamConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_kinesis_stream',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // stream_mode_details - computed: false, optional: true, required: false
        this._streamModeDetails = new KinesisStreamStreamModeDetailsOutputReference(this, "stream_mode_details");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new KinesisStreamTimeoutsOutputReference(this, "timeouts");
        this._arn = config.arn;
        this._encryptionType = config.encryptionType;
        this._enforceConsumerDeletion = config.enforceConsumerDeletion;
        this._id = config.id;
        this._kmsKeyId = config.kmsKeyId;
        this._name = config.name;
        this._retentionPeriod = config.retentionPeriod;
        this._shardCount = config.shardCount;
        this._shardLevelMetrics = config.shardLevelMetrics;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._streamModeDetails.internalValue = config.streamModeDetails;
        this._timeouts.internalValue = config.timeouts;
    }
    get arn() {
        return this.getStringAttribute('arn');
    }
    set arn(value) {
        this._arn = value;
    }
    resetArn() {
        this._arn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get arnInput() {
        return this._arn;
    }
    get encryptionType() {
        return this.getStringAttribute('encryption_type');
    }
    set encryptionType(value) {
        this._encryptionType = value;
    }
    resetEncryptionType() {
        this._encryptionType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encryptionTypeInput() {
        return this._encryptionType;
    }
    get enforceConsumerDeletion() {
        return this.getBooleanAttribute('enforce_consumer_deletion');
    }
    set enforceConsumerDeletion(value) {
        this._enforceConsumerDeletion = value;
    }
    resetEnforceConsumerDeletion() {
        this._enforceConsumerDeletion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enforceConsumerDeletionInput() {
        return this._enforceConsumerDeletion;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    resetKmsKeyId() {
        this._kmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get retentionPeriod() {
        return this.getNumberAttribute('retention_period');
    }
    set retentionPeriod(value) {
        this._retentionPeriod = value;
    }
    resetRetentionPeriod() {
        this._retentionPeriod = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get retentionPeriodInput() {
        return this._retentionPeriod;
    }
    get shardCount() {
        return this.getNumberAttribute('shard_count');
    }
    set shardCount(value) {
        this._shardCount = value;
    }
    resetShardCount() {
        this._shardCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get shardCountInput() {
        return this._shardCount;
    }
    get shardLevelMetrics() {
        return cdktf.Fn.tolist(this.getListAttribute('shard_level_metrics'));
    }
    set shardLevelMetrics(value) {
        this._shardLevelMetrics = value;
    }
    resetShardLevelMetrics() {
        this._shardLevelMetrics = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get shardLevelMetricsInput() {
        return this._shardLevelMetrics;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get streamModeDetails() {
        return this._streamModeDetails;
    }
    putStreamModeDetails(value) {
        this._streamModeDetails.internalValue = value;
    }
    resetStreamModeDetails() {
        this._streamModeDetails.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get streamModeDetailsInput() {
        return this._streamModeDetails.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            arn: cdktf.stringToTerraform(this._arn),
            encryption_type: cdktf.stringToTerraform(this._encryptionType),
            enforce_consumer_deletion: cdktf.booleanToTerraform(this._enforceConsumerDeletion),
            id: cdktf.stringToTerraform(this._id),
            kms_key_id: cdktf.stringToTerraform(this._kmsKeyId),
            name: cdktf.stringToTerraform(this._name),
            retention_period: cdktf.numberToTerraform(this._retentionPeriod),
            shard_count: cdktf.numberToTerraform(this._shardCount),
            shard_level_metrics: cdktf.listMapper(cdktf.stringToTerraform, false)(this._shardLevelMetrics),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            stream_mode_details: kinesisStreamStreamModeDetailsToTerraform(this._streamModeDetails.internalValue),
            timeouts: kinesisStreamTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.KinesisStream = KinesisStream;
_c = JSII_RTTI_SYMBOL_1;
KinesisStream[_c] = { fqn: "@cdktf/provider-aws.kinesisStream.KinesisStream", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
KinesisStream.tfResourceType = "aws_kinesis_stream";
