// @cdktf/provider-aws/lib/fsx-ontap-volume/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.FsxOntapVolume = exports.FsxOntapVolumeTimeoutsOutputReference = exports.fsxOntapVolumeTimeoutsToTerraform = exports.FsxOntapVolumeTieringPolicyOutputReference = exports.fsxOntapVolumeTieringPolicyToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function fsxOntapVolumeTieringPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cooling_period: cdktf.numberToTerraform(struct.coolingPeriod),
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.fsxOntapVolumeTieringPolicyToTerraform = fsxOntapVolumeTieringPolicyToTerraform;
class FsxOntapVolumeTieringPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._coolingPeriod !== undefined) {
            hasAnyValues = true;
            internalValueResult.coolingPeriod = this._coolingPeriod;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._coolingPeriod = undefined;
            this._name = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._coolingPeriod = value.coolingPeriod;
            this._name = value.name;
        }
    }
    get coolingPeriod() {
        return this.getNumberAttribute('cooling_period');
    }
    set coolingPeriod(value) {
        this._coolingPeriod = value;
    }
    resetCoolingPeriod() {
        this._coolingPeriod = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get coolingPeriodInput() {
        return this._coolingPeriod;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
}
exports.FsxOntapVolumeTieringPolicyOutputReference = FsxOntapVolumeTieringPolicyOutputReference;
_a = JSII_RTTI_SYMBOL_1;
FsxOntapVolumeTieringPolicyOutputReference[_a] = { fqn: "@cdktf/provider-aws.fsxOntapVolume.FsxOntapVolumeTieringPolicyOutputReference", version: "10.0.12" };
function fsxOntapVolumeTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.fsxOntapVolumeTimeoutsToTerraform = fsxOntapVolumeTimeoutsToTerraform;
class FsxOntapVolumeTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.FsxOntapVolumeTimeoutsOutputReference = FsxOntapVolumeTimeoutsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
FsxOntapVolumeTimeoutsOutputReference[_b] = { fqn: "@cdktf/provider-aws.fsxOntapVolume.FsxOntapVolumeTimeoutsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/fsx_ontap_volume aws_fsx_ontap_volume}
*/
class FsxOntapVolume extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/fsx_ontap_volume aws_fsx_ontap_volume} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options FsxOntapVolumeConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_fsx_ontap_volume',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // tiering_policy - computed: false, optional: true, required: false
        this._tieringPolicy = new FsxOntapVolumeTieringPolicyOutputReference(this, "tiering_policy");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new FsxOntapVolumeTimeoutsOutputReference(this, "timeouts");
        this._id = config.id;
        this._junctionPath = config.junctionPath;
        this._name = config.name;
        this._securityStyle = config.securityStyle;
        this._sizeInMegabytes = config.sizeInMegabytes;
        this._storageEfficiencyEnabled = config.storageEfficiencyEnabled;
        this._storageVirtualMachineId = config.storageVirtualMachineId;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._volumeType = config.volumeType;
        this._tieringPolicy.internalValue = config.tieringPolicy;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // file_system_id - computed: true, optional: false, required: false
    get fileSystemId() {
        return this.getStringAttribute('file_system_id');
    }
    // flexcache_endpoint_type - computed: true, optional: false, required: false
    get flexcacheEndpointType() {
        return this.getStringAttribute('flexcache_endpoint_type');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get junctionPath() {
        return this.getStringAttribute('junction_path');
    }
    set junctionPath(value) {
        this._junctionPath = value;
    }
    // Temporarily expose input value. Use with caution.
    get junctionPathInput() {
        return this._junctionPath;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // ontap_volume_type - computed: true, optional: false, required: false
    get ontapVolumeType() {
        return this.getStringAttribute('ontap_volume_type');
    }
    get securityStyle() {
        return this.getStringAttribute('security_style');
    }
    set securityStyle(value) {
        this._securityStyle = value;
    }
    resetSecurityStyle() {
        this._securityStyle = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityStyleInput() {
        return this._securityStyle;
    }
    get sizeInMegabytes() {
        return this.getNumberAttribute('size_in_megabytes');
    }
    set sizeInMegabytes(value) {
        this._sizeInMegabytes = value;
    }
    // Temporarily expose input value. Use with caution.
    get sizeInMegabytesInput() {
        return this._sizeInMegabytes;
    }
    get storageEfficiencyEnabled() {
        return this.getBooleanAttribute('storage_efficiency_enabled');
    }
    set storageEfficiencyEnabled(value) {
        this._storageEfficiencyEnabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get storageEfficiencyEnabledInput() {
        return this._storageEfficiencyEnabled;
    }
    get storageVirtualMachineId() {
        return this.getStringAttribute('storage_virtual_machine_id');
    }
    set storageVirtualMachineId(value) {
        this._storageVirtualMachineId = value;
    }
    // Temporarily expose input value. Use with caution.
    get storageVirtualMachineIdInput() {
        return this._storageVirtualMachineId;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // uuid - computed: true, optional: false, required: false
    get uuid() {
        return this.getStringAttribute('uuid');
    }
    get volumeType() {
        return this.getStringAttribute('volume_type');
    }
    set volumeType(value) {
        this._volumeType = value;
    }
    resetVolumeType() {
        this._volumeType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumeTypeInput() {
        return this._volumeType;
    }
    get tieringPolicy() {
        return this._tieringPolicy;
    }
    putTieringPolicy(value) {
        this._tieringPolicy.internalValue = value;
    }
    resetTieringPolicy() {
        this._tieringPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tieringPolicyInput() {
        return this._tieringPolicy.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            junction_path: cdktf.stringToTerraform(this._junctionPath),
            name: cdktf.stringToTerraform(this._name),
            security_style: cdktf.stringToTerraform(this._securityStyle),
            size_in_megabytes: cdktf.numberToTerraform(this._sizeInMegabytes),
            storage_efficiency_enabled: cdktf.booleanToTerraform(this._storageEfficiencyEnabled),
            storage_virtual_machine_id: cdktf.stringToTerraform(this._storageVirtualMachineId),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            volume_type: cdktf.stringToTerraform(this._volumeType),
            tiering_policy: fsxOntapVolumeTieringPolicyToTerraform(this._tieringPolicy.internalValue),
            timeouts: fsxOntapVolumeTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.FsxOntapVolume = FsxOntapVolume;
_c = JSII_RTTI_SYMBOL_1;
FsxOntapVolume[_c] = { fqn: "@cdktf/provider-aws.fsxOntapVolume.FsxOntapVolume", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
FsxOntapVolume.tfResourceType = "aws_fsx_ontap_volume";
