// @cdktf/provider-aws/lib/sagemaker-notebook-instance/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SagemakerNotebookInstance = exports.SagemakerNotebookInstanceInstanceMetadataServiceConfigurationOutputReference = exports.sagemakerNotebookInstanceInstanceMetadataServiceConfigurationToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function sagemakerNotebookInstanceInstanceMetadataServiceConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        minimum_instance_metadata_service_version: cdktf.stringToTerraform(struct.minimumInstanceMetadataServiceVersion),
    };
}
exports.sagemakerNotebookInstanceInstanceMetadataServiceConfigurationToTerraform = sagemakerNotebookInstanceInstanceMetadataServiceConfigurationToTerraform;
class SagemakerNotebookInstanceInstanceMetadataServiceConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._minimumInstanceMetadataServiceVersion !== undefined) {
            hasAnyValues = true;
            internalValueResult.minimumInstanceMetadataServiceVersion = this._minimumInstanceMetadataServiceVersion;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._minimumInstanceMetadataServiceVersion = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._minimumInstanceMetadataServiceVersion = value.minimumInstanceMetadataServiceVersion;
        }
    }
    get minimumInstanceMetadataServiceVersion() {
        return this.getStringAttribute('minimum_instance_metadata_service_version');
    }
    set minimumInstanceMetadataServiceVersion(value) {
        this._minimumInstanceMetadataServiceVersion = value;
    }
    resetMinimumInstanceMetadataServiceVersion() {
        this._minimumInstanceMetadataServiceVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get minimumInstanceMetadataServiceVersionInput() {
        return this._minimumInstanceMetadataServiceVersion;
    }
}
exports.SagemakerNotebookInstanceInstanceMetadataServiceConfigurationOutputReference = SagemakerNotebookInstanceInstanceMetadataServiceConfigurationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SagemakerNotebookInstanceInstanceMetadataServiceConfigurationOutputReference[_a] = { fqn: "@cdktf/provider-aws.sagemakerNotebookInstance.SagemakerNotebookInstanceInstanceMetadataServiceConfigurationOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/sagemaker_notebook_instance aws_sagemaker_notebook_instance}
*/
class SagemakerNotebookInstance extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/sagemaker_notebook_instance aws_sagemaker_notebook_instance} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SagemakerNotebookInstanceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_sagemaker_notebook_instance',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // instance_metadata_service_configuration - computed: false, optional: true, required: false
        this._instanceMetadataServiceConfiguration = new SagemakerNotebookInstanceInstanceMetadataServiceConfigurationOutputReference(this, "instance_metadata_service_configuration");
        this._acceleratorTypes = config.acceleratorTypes;
        this._additionalCodeRepositories = config.additionalCodeRepositories;
        this._defaultCodeRepository = config.defaultCodeRepository;
        this._directInternetAccess = config.directInternetAccess;
        this._id = config.id;
        this._instanceType = config.instanceType;
        this._kmsKeyId = config.kmsKeyId;
        this._lifecycleConfigName = config.lifecycleConfigName;
        this._name = config.name;
        this._platformIdentifier = config.platformIdentifier;
        this._roleArn = config.roleArn;
        this._rootAccess = config.rootAccess;
        this._securityGroups = config.securityGroups;
        this._subnetId = config.subnetId;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._volumeSize = config.volumeSize;
        this._instanceMetadataServiceConfiguration.internalValue = config.instanceMetadataServiceConfiguration;
    }
    get acceleratorTypes() {
        return cdktf.Fn.tolist(this.getListAttribute('accelerator_types'));
    }
    set acceleratorTypes(value) {
        this._acceleratorTypes = value;
    }
    resetAcceleratorTypes() {
        this._acceleratorTypes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get acceleratorTypesInput() {
        return this._acceleratorTypes;
    }
    get additionalCodeRepositories() {
        return cdktf.Fn.tolist(this.getListAttribute('additional_code_repositories'));
    }
    set additionalCodeRepositories(value) {
        this._additionalCodeRepositories = value;
    }
    resetAdditionalCodeRepositories() {
        this._additionalCodeRepositories = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get additionalCodeRepositoriesInput() {
        return this._additionalCodeRepositories;
    }
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get defaultCodeRepository() {
        return this.getStringAttribute('default_code_repository');
    }
    set defaultCodeRepository(value) {
        this._defaultCodeRepository = value;
    }
    resetDefaultCodeRepository() {
        this._defaultCodeRepository = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultCodeRepositoryInput() {
        return this._defaultCodeRepository;
    }
    get directInternetAccess() {
        return this.getStringAttribute('direct_internet_access');
    }
    set directInternetAccess(value) {
        this._directInternetAccess = value;
    }
    resetDirectInternetAccess() {
        this._directInternetAccess = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get directInternetAccessInput() {
        return this._directInternetAccess;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    set instanceType(value) {
        this._instanceType = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceTypeInput() {
        return this._instanceType;
    }
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    set kmsKeyId(value) {
        this._kmsKeyId = value;
    }
    resetKmsKeyId() {
        this._kmsKeyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyIdInput() {
        return this._kmsKeyId;
    }
    get lifecycleConfigName() {
        return this.getStringAttribute('lifecycle_config_name');
    }
    set lifecycleConfigName(value) {
        this._lifecycleConfigName = value;
    }
    resetLifecycleConfigName() {
        this._lifecycleConfigName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lifecycleConfigNameInput() {
        return this._lifecycleConfigName;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // network_interface_id - computed: true, optional: false, required: false
    get networkInterfaceId() {
        return this.getStringAttribute('network_interface_id');
    }
    get platformIdentifier() {
        return this.getStringAttribute('platform_identifier');
    }
    set platformIdentifier(value) {
        this._platformIdentifier = value;
    }
    resetPlatformIdentifier() {
        this._platformIdentifier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get platformIdentifierInput() {
        return this._platformIdentifier;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get rootAccess() {
        return this.getStringAttribute('root_access');
    }
    set rootAccess(value) {
        this._rootAccess = value;
    }
    resetRootAccess() {
        this._rootAccess = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rootAccessInput() {
        return this._rootAccess;
    }
    get securityGroups() {
        return cdktf.Fn.tolist(this.getListAttribute('security_groups'));
    }
    set securityGroups(value) {
        this._securityGroups = value;
    }
    resetSecurityGroups() {
        this._securityGroups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupsInput() {
        return this._securityGroups;
    }
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    set subnetId(value) {
        this._subnetId = value;
    }
    resetSubnetId() {
        this._subnetId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdInput() {
        return this._subnetId;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // url - computed: true, optional: false, required: false
    get url() {
        return this.getStringAttribute('url');
    }
    get volumeSize() {
        return this.getNumberAttribute('volume_size');
    }
    set volumeSize(value) {
        this._volumeSize = value;
    }
    resetVolumeSize() {
        this._volumeSize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumeSizeInput() {
        return this._volumeSize;
    }
    get instanceMetadataServiceConfiguration() {
        return this._instanceMetadataServiceConfiguration;
    }
    putInstanceMetadataServiceConfiguration(value) {
        this._instanceMetadataServiceConfiguration.internalValue = value;
    }
    resetInstanceMetadataServiceConfiguration() {
        this._instanceMetadataServiceConfiguration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instanceMetadataServiceConfigurationInput() {
        return this._instanceMetadataServiceConfiguration.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            accelerator_types: cdktf.listMapper(cdktf.stringToTerraform, false)(this._acceleratorTypes),
            additional_code_repositories: cdktf.listMapper(cdktf.stringToTerraform, false)(this._additionalCodeRepositories),
            default_code_repository: cdktf.stringToTerraform(this._defaultCodeRepository),
            direct_internet_access: cdktf.stringToTerraform(this._directInternetAccess),
            id: cdktf.stringToTerraform(this._id),
            instance_type: cdktf.stringToTerraform(this._instanceType),
            kms_key_id: cdktf.stringToTerraform(this._kmsKeyId),
            lifecycle_config_name: cdktf.stringToTerraform(this._lifecycleConfigName),
            name: cdktf.stringToTerraform(this._name),
            platform_identifier: cdktf.stringToTerraform(this._platformIdentifier),
            role_arn: cdktf.stringToTerraform(this._roleArn),
            root_access: cdktf.stringToTerraform(this._rootAccess),
            security_groups: cdktf.listMapper(cdktf.stringToTerraform, false)(this._securityGroups),
            subnet_id: cdktf.stringToTerraform(this._subnetId),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            volume_size: cdktf.numberToTerraform(this._volumeSize),
            instance_metadata_service_configuration: sagemakerNotebookInstanceInstanceMetadataServiceConfigurationToTerraform(this._instanceMetadataServiceConfiguration.internalValue),
        };
    }
}
exports.SagemakerNotebookInstance = SagemakerNotebookInstance;
_b = JSII_RTTI_SYMBOL_1;
SagemakerNotebookInstance[_b] = { fqn: "@cdktf/provider-aws.sagemakerNotebookInstance.SagemakerNotebookInstance", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
SagemakerNotebookInstance.tfResourceType = "aws_sagemaker_notebook_instance";
