// @cdktf/provider-aws/lib/api-gateway-authorizer/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiGatewayAuthorizer = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/api_gateway_authorizer aws_api_gateway_authorizer}
*/
class ApiGatewayAuthorizer extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/api_gateway_authorizer aws_api_gateway_authorizer} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ApiGatewayAuthorizerConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_api_gateway_authorizer',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._authorizerCredentials = config.authorizerCredentials;
        this._authorizerResultTtlInSeconds = config.authorizerResultTtlInSeconds;
        this._authorizerUri = config.authorizerUri;
        this._id = config.id;
        this._identitySource = config.identitySource;
        this._identityValidationExpression = config.identityValidationExpression;
        this._name = config.name;
        this._providerArns = config.providerArns;
        this._restApiId = config.restApiId;
        this._type = config.type;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get authorizerCredentials() {
        return this.getStringAttribute('authorizer_credentials');
    }
    set authorizerCredentials(value) {
        this._authorizerCredentials = value;
    }
    resetAuthorizerCredentials() {
        this._authorizerCredentials = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authorizerCredentialsInput() {
        return this._authorizerCredentials;
    }
    get authorizerResultTtlInSeconds() {
        return this.getNumberAttribute('authorizer_result_ttl_in_seconds');
    }
    set authorizerResultTtlInSeconds(value) {
        this._authorizerResultTtlInSeconds = value;
    }
    resetAuthorizerResultTtlInSeconds() {
        this._authorizerResultTtlInSeconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authorizerResultTtlInSecondsInput() {
        return this._authorizerResultTtlInSeconds;
    }
    get authorizerUri() {
        return this.getStringAttribute('authorizer_uri');
    }
    set authorizerUri(value) {
        this._authorizerUri = value;
    }
    resetAuthorizerUri() {
        this._authorizerUri = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authorizerUriInput() {
        return this._authorizerUri;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get identitySource() {
        return this.getStringAttribute('identity_source');
    }
    set identitySource(value) {
        this._identitySource = value;
    }
    resetIdentitySource() {
        this._identitySource = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get identitySourceInput() {
        return this._identitySource;
    }
    get identityValidationExpression() {
        return this.getStringAttribute('identity_validation_expression');
    }
    set identityValidationExpression(value) {
        this._identityValidationExpression = value;
    }
    resetIdentityValidationExpression() {
        this._identityValidationExpression = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get identityValidationExpressionInput() {
        return this._identityValidationExpression;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get providerArns() {
        return cdktf.Fn.tolist(this.getListAttribute('provider_arns'));
    }
    set providerArns(value) {
        this._providerArns = value;
    }
    resetProviderArns() {
        this._providerArns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get providerArnsInput() {
        return this._providerArns;
    }
    get restApiId() {
        return this.getStringAttribute('rest_api_id');
    }
    set restApiId(value) {
        this._restApiId = value;
    }
    // Temporarily expose input value. Use with caution.
    get restApiIdInput() {
        return this._restApiId;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            authorizer_credentials: cdktf.stringToTerraform(this._authorizerCredentials),
            authorizer_result_ttl_in_seconds: cdktf.numberToTerraform(this._authorizerResultTtlInSeconds),
            authorizer_uri: cdktf.stringToTerraform(this._authorizerUri),
            id: cdktf.stringToTerraform(this._id),
            identity_source: cdktf.stringToTerraform(this._identitySource),
            identity_validation_expression: cdktf.stringToTerraform(this._identityValidationExpression),
            name: cdktf.stringToTerraform(this._name),
            provider_arns: cdktf.listMapper(cdktf.stringToTerraform, false)(this._providerArns),
            rest_api_id: cdktf.stringToTerraform(this._restApiId),
            type: cdktf.stringToTerraform(this._type),
        };
    }
}
exports.ApiGatewayAuthorizer = ApiGatewayAuthorizer;
_a = JSII_RTTI_SYMBOL_1;
ApiGatewayAuthorizer[_a] = { fqn: "@cdktf/provider-aws.apiGatewayAuthorizer.ApiGatewayAuthorizer", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
ApiGatewayAuthorizer.tfResourceType = "aws_api_gateway_authorizer";
