// @cdktf/provider-aws/lib/data-aws-db-instance/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataAwsDbInstance = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/d/db_instance aws_db_instance}
*/
class DataAwsDbInstance extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/d/db_instance aws_db_instance} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataAwsDbInstanceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_db_instance',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._dbInstanceIdentifier = config.dbInstanceIdentifier;
        this._id = config.id;
        this._tags = config.tags;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // address - computed: true, optional: false, required: false
    get address() {
        return this.getStringAttribute('address');
    }
    // allocated_storage - computed: true, optional: false, required: false
    get allocatedStorage() {
        return this.getNumberAttribute('allocated_storage');
    }
    // auto_minor_version_upgrade - computed: true, optional: false, required: false
    get autoMinorVersionUpgrade() {
        return this.getBooleanAttribute('auto_minor_version_upgrade');
    }
    // availability_zone - computed: true, optional: false, required: false
    get availabilityZone() {
        return this.getStringAttribute('availability_zone');
    }
    // backup_retention_period - computed: true, optional: false, required: false
    get backupRetentionPeriod() {
        return this.getNumberAttribute('backup_retention_period');
    }
    // ca_cert_identifier - computed: true, optional: false, required: false
    get caCertIdentifier() {
        return this.getStringAttribute('ca_cert_identifier');
    }
    // db_cluster_identifier - computed: true, optional: false, required: false
    get dbClusterIdentifier() {
        return this.getStringAttribute('db_cluster_identifier');
    }
    // db_instance_arn - computed: true, optional: false, required: false
    get dbInstanceArn() {
        return this.getStringAttribute('db_instance_arn');
    }
    // db_instance_class - computed: true, optional: false, required: false
    get dbInstanceClass() {
        return this.getStringAttribute('db_instance_class');
    }
    get dbInstanceIdentifier() {
        return this.getStringAttribute('db_instance_identifier');
    }
    set dbInstanceIdentifier(value) {
        this._dbInstanceIdentifier = value;
    }
    // Temporarily expose input value. Use with caution.
    get dbInstanceIdentifierInput() {
        return this._dbInstanceIdentifier;
    }
    // db_instance_port - computed: true, optional: false, required: false
    get dbInstancePort() {
        return this.getNumberAttribute('db_instance_port');
    }
    // db_name - computed: true, optional: false, required: false
    get dbName() {
        return this.getStringAttribute('db_name');
    }
    // db_parameter_groups - computed: true, optional: false, required: false
    get dbParameterGroups() {
        return this.getListAttribute('db_parameter_groups');
    }
    // db_security_groups - computed: true, optional: false, required: false
    get dbSecurityGroups() {
        return this.getListAttribute('db_security_groups');
    }
    // db_subnet_group - computed: true, optional: false, required: false
    get dbSubnetGroup() {
        return this.getStringAttribute('db_subnet_group');
    }
    // enabled_cloudwatch_logs_exports - computed: true, optional: false, required: false
    get enabledCloudwatchLogsExports() {
        return this.getListAttribute('enabled_cloudwatch_logs_exports');
    }
    // endpoint - computed: true, optional: false, required: false
    get endpoint() {
        return this.getStringAttribute('endpoint');
    }
    // engine - computed: true, optional: false, required: false
    get engine() {
        return this.getStringAttribute('engine');
    }
    // engine_version - computed: true, optional: false, required: false
    get engineVersion() {
        return this.getStringAttribute('engine_version');
    }
    // hosted_zone_id - computed: true, optional: false, required: false
    get hostedZoneId() {
        return this.getStringAttribute('hosted_zone_id');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // iops - computed: true, optional: false, required: false
    get iops() {
        return this.getNumberAttribute('iops');
    }
    // kms_key_id - computed: true, optional: false, required: false
    get kmsKeyId() {
        return this.getStringAttribute('kms_key_id');
    }
    // license_model - computed: true, optional: false, required: false
    get licenseModel() {
        return this.getStringAttribute('license_model');
    }
    // master_username - computed: true, optional: false, required: false
    get masterUsername() {
        return this.getStringAttribute('master_username');
    }
    // monitoring_interval - computed: true, optional: false, required: false
    get monitoringInterval() {
        return this.getNumberAttribute('monitoring_interval');
    }
    // monitoring_role_arn - computed: true, optional: false, required: false
    get monitoringRoleArn() {
        return this.getStringAttribute('monitoring_role_arn');
    }
    // multi_az - computed: true, optional: false, required: false
    get multiAz() {
        return this.getBooleanAttribute('multi_az');
    }
    // network_type - computed: true, optional: false, required: false
    get networkType() {
        return this.getStringAttribute('network_type');
    }
    // option_group_memberships - computed: true, optional: false, required: false
    get optionGroupMemberships() {
        return this.getListAttribute('option_group_memberships');
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getNumberAttribute('port');
    }
    // preferred_backup_window - computed: true, optional: false, required: false
    get preferredBackupWindow() {
        return this.getStringAttribute('preferred_backup_window');
    }
    // preferred_maintenance_window - computed: true, optional: false, required: false
    get preferredMaintenanceWindow() {
        return this.getStringAttribute('preferred_maintenance_window');
    }
    // publicly_accessible - computed: true, optional: false, required: false
    get publiclyAccessible() {
        return this.getBooleanAttribute('publicly_accessible');
    }
    // replicate_source_db - computed: true, optional: false, required: false
    get replicateSourceDb() {
        return this.getStringAttribute('replicate_source_db');
    }
    // resource_id - computed: true, optional: false, required: false
    get resourceId() {
        return this.getStringAttribute('resource_id');
    }
    // storage_encrypted - computed: true, optional: false, required: false
    get storageEncrypted() {
        return this.getBooleanAttribute('storage_encrypted');
    }
    // storage_type - computed: true, optional: false, required: false
    get storageType() {
        return this.getStringAttribute('storage_type');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // timezone - computed: true, optional: false, required: false
    get timezone() {
        return this.getStringAttribute('timezone');
    }
    // vpc_security_groups - computed: true, optional: false, required: false
    get vpcSecurityGroups() {
        return this.getListAttribute('vpc_security_groups');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            db_instance_identifier: cdktf.stringToTerraform(this._dbInstanceIdentifier),
            id: cdktf.stringToTerraform(this._id),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
        };
    }
}
exports.DataAwsDbInstance = DataAwsDbInstance;
_a = JSII_RTTI_SYMBOL_1;
DataAwsDbInstance[_a] = { fqn: "@cdktf/provider-aws.dataAwsDbInstance.DataAwsDbInstance", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DataAwsDbInstance.tfResourceType = "aws_db_instance";
