// @cdktf/provider-aws/lib/ssoadmin-customer-managed-policy-attachment/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SsoadminCustomerManagedPolicyAttachment = exports.SsoadminCustomerManagedPolicyAttachmentCustomerManagedPolicyReferenceOutputReference = exports.ssoadminCustomerManagedPolicyAttachmentCustomerManagedPolicyReferenceToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function ssoadminCustomerManagedPolicyAttachmentCustomerManagedPolicyReferenceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        path: cdktf.stringToTerraform(struct.path),
    };
}
exports.ssoadminCustomerManagedPolicyAttachmentCustomerManagedPolicyReferenceToTerraform = ssoadminCustomerManagedPolicyAttachmentCustomerManagedPolicyReferenceToTerraform;
class SsoadminCustomerManagedPolicyAttachmentCustomerManagedPolicyReferenceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._name = undefined;
            this._path = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._name = value.name;
            this._path = value.path;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    resetPath() {
        this._path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
}
exports.SsoadminCustomerManagedPolicyAttachmentCustomerManagedPolicyReferenceOutputReference = SsoadminCustomerManagedPolicyAttachmentCustomerManagedPolicyReferenceOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SsoadminCustomerManagedPolicyAttachmentCustomerManagedPolicyReferenceOutputReference[_a] = { fqn: "@cdktf/provider-aws.ssoadminCustomerManagedPolicyAttachment.SsoadminCustomerManagedPolicyAttachmentCustomerManagedPolicyReferenceOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/ssoadmin_customer_managed_policy_attachment aws_ssoadmin_customer_managed_policy_attachment}
*/
class SsoadminCustomerManagedPolicyAttachment extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/ssoadmin_customer_managed_policy_attachment aws_ssoadmin_customer_managed_policy_attachment} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SsoadminCustomerManagedPolicyAttachmentConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_ssoadmin_customer_managed_policy_attachment',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // customer_managed_policy_reference - computed: false, optional: false, required: true
        this._customerManagedPolicyReference = new SsoadminCustomerManagedPolicyAttachmentCustomerManagedPolicyReferenceOutputReference(this, "customer_managed_policy_reference");
        this._id = config.id;
        this._instanceArn = config.instanceArn;
        this._permissionSetArn = config.permissionSetArn;
        this._customerManagedPolicyReference.internalValue = config.customerManagedPolicyReference;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instanceArn() {
        return this.getStringAttribute('instance_arn');
    }
    set instanceArn(value) {
        this._instanceArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceArnInput() {
        return this._instanceArn;
    }
    get permissionSetArn() {
        return this.getStringAttribute('permission_set_arn');
    }
    set permissionSetArn(value) {
        this._permissionSetArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get permissionSetArnInput() {
        return this._permissionSetArn;
    }
    get customerManagedPolicyReference() {
        return this._customerManagedPolicyReference;
    }
    putCustomerManagedPolicyReference(value) {
        this._customerManagedPolicyReference.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get customerManagedPolicyReferenceInput() {
        return this._customerManagedPolicyReference.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            instance_arn: cdktf.stringToTerraform(this._instanceArn),
            permission_set_arn: cdktf.stringToTerraform(this._permissionSetArn),
            customer_managed_policy_reference: ssoadminCustomerManagedPolicyAttachmentCustomerManagedPolicyReferenceToTerraform(this._customerManagedPolicyReference.internalValue),
        };
    }
}
exports.SsoadminCustomerManagedPolicyAttachment = SsoadminCustomerManagedPolicyAttachment;
_b = JSII_RTTI_SYMBOL_1;
SsoadminCustomerManagedPolicyAttachment[_b] = { fqn: "@cdktf/provider-aws.ssoadminCustomerManagedPolicyAttachment.SsoadminCustomerManagedPolicyAttachment", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
SsoadminCustomerManagedPolicyAttachment.tfResourceType = "aws_ssoadmin_customer_managed_policy_attachment";
