// @cdktf/provider-aws/lib/iot-provisioning-template/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IotProvisioningTemplate = exports.IotProvisioningTemplatePreProvisioningHookOutputReference = exports.iotProvisioningTemplatePreProvisioningHookToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function iotProvisioningTemplatePreProvisioningHookToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        payload_version: cdktf.stringToTerraform(struct.payloadVersion),
        target_arn: cdktf.stringToTerraform(struct.targetArn),
    };
}
exports.iotProvisioningTemplatePreProvisioningHookToTerraform = iotProvisioningTemplatePreProvisioningHookToTerraform;
class IotProvisioningTemplatePreProvisioningHookOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._payloadVersion !== undefined) {
            hasAnyValues = true;
            internalValueResult.payloadVersion = this._payloadVersion;
        }
        if (this._targetArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetArn = this._targetArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._payloadVersion = undefined;
            this._targetArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._payloadVersion = value.payloadVersion;
            this._targetArn = value.targetArn;
        }
    }
    get payloadVersion() {
        return this.getStringAttribute('payload_version');
    }
    set payloadVersion(value) {
        this._payloadVersion = value;
    }
    resetPayloadVersion() {
        this._payloadVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get payloadVersionInput() {
        return this._payloadVersion;
    }
    get targetArn() {
        return this.getStringAttribute('target_arn');
    }
    set targetArn(value) {
        this._targetArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get targetArnInput() {
        return this._targetArn;
    }
}
exports.IotProvisioningTemplatePreProvisioningHookOutputReference = IotProvisioningTemplatePreProvisioningHookOutputReference;
_a = JSII_RTTI_SYMBOL_1;
IotProvisioningTemplatePreProvisioningHookOutputReference[_a] = { fqn: "@cdktf/provider-aws.iotProvisioningTemplate.IotProvisioningTemplatePreProvisioningHookOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/iot_provisioning_template aws_iot_provisioning_template}
*/
class IotProvisioningTemplate extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/iot_provisioning_template aws_iot_provisioning_template} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options IotProvisioningTemplateConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_iot_provisioning_template',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // pre_provisioning_hook - computed: false, optional: true, required: false
        this._preProvisioningHook = new IotProvisioningTemplatePreProvisioningHookOutputReference(this, "pre_provisioning_hook");
        this._description = config.description;
        this._enabled = config.enabled;
        this._id = config.id;
        this._name = config.name;
        this._provisioningRoleArn = config.provisioningRoleArn;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._templateBody = config.templateBody;
        this._preProvisioningHook.internalValue = config.preProvisioningHook;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // default_version_id - computed: true, optional: false, required: false
    get defaultVersionId() {
        return this.getNumberAttribute('default_version_id');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get provisioningRoleArn() {
        return this.getStringAttribute('provisioning_role_arn');
    }
    set provisioningRoleArn(value) {
        this._provisioningRoleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get provisioningRoleArnInput() {
        return this._provisioningRoleArn;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get templateBody() {
        return this.getStringAttribute('template_body');
    }
    set templateBody(value) {
        this._templateBody = value;
    }
    // Temporarily expose input value. Use with caution.
    get templateBodyInput() {
        return this._templateBody;
    }
    get preProvisioningHook() {
        return this._preProvisioningHook;
    }
    putPreProvisioningHook(value) {
        this._preProvisioningHook.internalValue = value;
    }
    resetPreProvisioningHook() {
        this._preProvisioningHook.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get preProvisioningHookInput() {
        return this._preProvisioningHook.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            enabled: cdktf.booleanToTerraform(this._enabled),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            provisioning_role_arn: cdktf.stringToTerraform(this._provisioningRoleArn),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            template_body: cdktf.stringToTerraform(this._templateBody),
            pre_provisioning_hook: iotProvisioningTemplatePreProvisioningHookToTerraform(this._preProvisioningHook.internalValue),
        };
    }
}
exports.IotProvisioningTemplate = IotProvisioningTemplate;
_b = JSII_RTTI_SYMBOL_1;
IotProvisioningTemplate[_b] = { fqn: "@cdktf/provider-aws.iotProvisioningTemplate.IotProvisioningTemplate", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
IotProvisioningTemplate.tfResourceType = "aws_iot_provisioning_template";
