// @cdktf/provider-aws/lib/api-gateway-usage-plan-key/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiGatewayUsagePlanKey = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/api_gateway_usage_plan_key aws_api_gateway_usage_plan_key}
*/
class ApiGatewayUsagePlanKey extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/api_gateway_usage_plan_key aws_api_gateway_usage_plan_key} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ApiGatewayUsagePlanKeyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_api_gateway_usage_plan_key',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._keyId = config.keyId;
        this._keyType = config.keyType;
        this._usagePlanId = config.usagePlanId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get keyId() {
        return this.getStringAttribute('key_id');
    }
    set keyId(value) {
        this._keyId = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyIdInput() {
        return this._keyId;
    }
    get keyType() {
        return this.getStringAttribute('key_type');
    }
    set keyType(value) {
        this._keyType = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyTypeInput() {
        return this._keyType;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get usagePlanId() {
        return this.getStringAttribute('usage_plan_id');
    }
    set usagePlanId(value) {
        this._usagePlanId = value;
    }
    // Temporarily expose input value. Use with caution.
    get usagePlanIdInput() {
        return this._usagePlanId;
    }
    // value - computed: true, optional: false, required: false
    get value() {
        return this.getStringAttribute('value');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            key_id: cdktf.stringToTerraform(this._keyId),
            key_type: cdktf.stringToTerraform(this._keyType),
            usage_plan_id: cdktf.stringToTerraform(this._usagePlanId),
        };
    }
}
exports.ApiGatewayUsagePlanKey = ApiGatewayUsagePlanKey;
_a = JSII_RTTI_SYMBOL_1;
ApiGatewayUsagePlanKey[_a] = { fqn: "@cdktf/provider-aws.apiGatewayUsagePlanKey.ApiGatewayUsagePlanKey", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
ApiGatewayUsagePlanKey.tfResourceType = "aws_api_gateway_usage_plan_key";
