// @cdktf/provider-aws/lib/licensemanager-license-configuration/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LicensemanagerLicenseConfiguration = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/licensemanager_license_configuration aws_licensemanager_license_configuration}
*/
class LicensemanagerLicenseConfiguration extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/licensemanager_license_configuration aws_licensemanager_license_configuration} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options LicensemanagerLicenseConfigurationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_licensemanager_license_configuration',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._description = config.description;
        this._id = config.id;
        this._licenseCount = config.licenseCount;
        this._licenseCountHardLimit = config.licenseCountHardLimit;
        this._licenseCountingType = config.licenseCountingType;
        this._licenseRules = config.licenseRules;
        this._name = config.name;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get licenseCount() {
        return this.getNumberAttribute('license_count');
    }
    set licenseCount(value) {
        this._licenseCount = value;
    }
    resetLicenseCount() {
        this._licenseCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get licenseCountInput() {
        return this._licenseCount;
    }
    get licenseCountHardLimit() {
        return this.getBooleanAttribute('license_count_hard_limit');
    }
    set licenseCountHardLimit(value) {
        this._licenseCountHardLimit = value;
    }
    resetLicenseCountHardLimit() {
        this._licenseCountHardLimit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get licenseCountHardLimitInput() {
        return this._licenseCountHardLimit;
    }
    get licenseCountingType() {
        return this.getStringAttribute('license_counting_type');
    }
    set licenseCountingType(value) {
        this._licenseCountingType = value;
    }
    // Temporarily expose input value. Use with caution.
    get licenseCountingTypeInput() {
        return this._licenseCountingType;
    }
    get licenseRules() {
        return this.getListAttribute('license_rules');
    }
    set licenseRules(value) {
        this._licenseRules = value;
    }
    resetLicenseRules() {
        this._licenseRules = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get licenseRulesInput() {
        return this._licenseRules;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // owner_account_id - computed: true, optional: false, required: false
    get ownerAccountId() {
        return this.getStringAttribute('owner_account_id');
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            license_count: cdktf.numberToTerraform(this._licenseCount),
            license_count_hard_limit: cdktf.booleanToTerraform(this._licenseCountHardLimit),
            license_counting_type: cdktf.stringToTerraform(this._licenseCountingType),
            license_rules: cdktf.listMapper(cdktf.stringToTerraform, false)(this._licenseRules),
            name: cdktf.stringToTerraform(this._name),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
        };
    }
}
exports.LicensemanagerLicenseConfiguration = LicensemanagerLicenseConfiguration;
_a = JSII_RTTI_SYMBOL_1;
LicensemanagerLicenseConfiguration[_a] = { fqn: "@cdktf/provider-aws.licensemanagerLicenseConfiguration.LicensemanagerLicenseConfiguration", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
LicensemanagerLicenseConfiguration.tfResourceType = "aws_licensemanager_license_configuration";
