// @cdktf/provider-aws/lib/db-proxy/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DbProxy = exports.DbProxyTimeoutsOutputReference = exports.dbProxyTimeoutsToTerraform = exports.DbProxyAuthList = exports.DbProxyAuthOutputReference = exports.dbProxyAuthToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dbProxyAuthToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        auth_scheme: cdktf.stringToTerraform(struct.authScheme),
        description: cdktf.stringToTerraform(struct.description),
        iam_auth: cdktf.stringToTerraform(struct.iamAuth),
        secret_arn: cdktf.stringToTerraform(struct.secretArn),
        username: cdktf.stringToTerraform(struct.username),
    };
}
exports.dbProxyAuthToTerraform = dbProxyAuthToTerraform;
class DbProxyAuthOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._authScheme !== undefined) {
            hasAnyValues = true;
            internalValueResult.authScheme = this._authScheme;
        }
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._iamAuth !== undefined) {
            hasAnyValues = true;
            internalValueResult.iamAuth = this._iamAuth;
        }
        if (this._secretArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.secretArn = this._secretArn;
        }
        if (this._username !== undefined) {
            hasAnyValues = true;
            internalValueResult.username = this._username;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._authScheme = undefined;
            this._description = undefined;
            this._iamAuth = undefined;
            this._secretArn = undefined;
            this._username = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._authScheme = value.authScheme;
            this._description = value.description;
            this._iamAuth = value.iamAuth;
            this._secretArn = value.secretArn;
            this._username = value.username;
        }
    }
    get authScheme() {
        return this.getStringAttribute('auth_scheme');
    }
    set authScheme(value) {
        this._authScheme = value;
    }
    resetAuthScheme() {
        this._authScheme = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authSchemeInput() {
        return this._authScheme;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get iamAuth() {
        return this.getStringAttribute('iam_auth');
    }
    set iamAuth(value) {
        this._iamAuth = value;
    }
    resetIamAuth() {
        this._iamAuth = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iamAuthInput() {
        return this._iamAuth;
    }
    get secretArn() {
        return this.getStringAttribute('secret_arn');
    }
    set secretArn(value) {
        this._secretArn = value;
    }
    resetSecretArn() {
        this._secretArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secretArnInput() {
        return this._secretArn;
    }
    get username() {
        return this.getStringAttribute('username');
    }
    set username(value) {
        this._username = value;
    }
    resetUsername() {
        this._username = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get usernameInput() {
        return this._username;
    }
}
exports.DbProxyAuthOutputReference = DbProxyAuthOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DbProxyAuthOutputReference[_a] = { fqn: "@cdktf/provider-aws.dbProxy.DbProxyAuthOutputReference", version: "10.0.12" };
class DbProxyAuthList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DbProxyAuthOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DbProxyAuthList = DbProxyAuthList;
_b = JSII_RTTI_SYMBOL_1;
DbProxyAuthList[_b] = { fqn: "@cdktf/provider-aws.dbProxy.DbProxyAuthList", version: "10.0.12" };
function dbProxyTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.dbProxyTimeoutsToTerraform = dbProxyTimeoutsToTerraform;
class DbProxyTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.DbProxyTimeoutsOutputReference = DbProxyTimeoutsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DbProxyTimeoutsOutputReference[_c] = { fqn: "@cdktf/provider-aws.dbProxy.DbProxyTimeoutsOutputReference", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/db_proxy aws_db_proxy}
*/
class DbProxy extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/db_proxy aws_db_proxy} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DbProxyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_db_proxy',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // auth - computed: false, optional: false, required: true
        this._auth = new DbProxyAuthList(this, "auth", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new DbProxyTimeoutsOutputReference(this, "timeouts");
        this._debugLogging = config.debugLogging;
        this._engineFamily = config.engineFamily;
        this._id = config.id;
        this._idleClientTimeout = config.idleClientTimeout;
        this._name = config.name;
        this._requireTls = config.requireTls;
        this._roleArn = config.roleArn;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._vpcSecurityGroupIds = config.vpcSecurityGroupIds;
        this._vpcSubnetIds = config.vpcSubnetIds;
        this._auth.internalValue = config.auth;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    get debugLogging() {
        return this.getBooleanAttribute('debug_logging');
    }
    set debugLogging(value) {
        this._debugLogging = value;
    }
    resetDebugLogging() {
        this._debugLogging = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get debugLoggingInput() {
        return this._debugLogging;
    }
    // endpoint - computed: true, optional: false, required: false
    get endpoint() {
        return this.getStringAttribute('endpoint');
    }
    get engineFamily() {
        return this.getStringAttribute('engine_family');
    }
    set engineFamily(value) {
        this._engineFamily = value;
    }
    // Temporarily expose input value. Use with caution.
    get engineFamilyInput() {
        return this._engineFamily;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get idleClientTimeout() {
        return this.getNumberAttribute('idle_client_timeout');
    }
    set idleClientTimeout(value) {
        this._idleClientTimeout = value;
    }
    resetIdleClientTimeout() {
        this._idleClientTimeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idleClientTimeoutInput() {
        return this._idleClientTimeout;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get requireTls() {
        return this.getBooleanAttribute('require_tls');
    }
    set requireTls(value) {
        this._requireTls = value;
    }
    resetRequireTls() {
        this._requireTls = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get requireTlsInput() {
        return this._requireTls;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get vpcSecurityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('vpc_security_group_ids'));
    }
    set vpcSecurityGroupIds(value) {
        this._vpcSecurityGroupIds = value;
    }
    resetVpcSecurityGroupIds() {
        this._vpcSecurityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcSecurityGroupIdsInput() {
        return this._vpcSecurityGroupIds;
    }
    get vpcSubnetIds() {
        return cdktf.Fn.tolist(this.getListAttribute('vpc_subnet_ids'));
    }
    set vpcSubnetIds(value) {
        this._vpcSubnetIds = value;
    }
    // Temporarily expose input value. Use with caution.
    get vpcSubnetIdsInput() {
        return this._vpcSubnetIds;
    }
    get auth() {
        return this._auth;
    }
    putAuth(value) {
        this._auth.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get authInput() {
        return this._auth.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            debug_logging: cdktf.booleanToTerraform(this._debugLogging),
            engine_family: cdktf.stringToTerraform(this._engineFamily),
            id: cdktf.stringToTerraform(this._id),
            idle_client_timeout: cdktf.numberToTerraform(this._idleClientTimeout),
            name: cdktf.stringToTerraform(this._name),
            require_tls: cdktf.booleanToTerraform(this._requireTls),
            role_arn: cdktf.stringToTerraform(this._roleArn),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            vpc_security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(this._vpcSecurityGroupIds),
            vpc_subnet_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(this._vpcSubnetIds),
            auth: cdktf.listMapper(dbProxyAuthToTerraform, true)(this._auth.internalValue),
            timeouts: dbProxyTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.DbProxy = DbProxy;
_d = JSII_RTTI_SYMBOL_1;
DbProxy[_d] = { fqn: "@cdktf/provider-aws.dbProxy.DbProxy", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
DbProxy.tfResourceType = "aws_db_proxy";
