// @cdktf/provider-aws/lib/sqs-queue-redrive-allow-policy/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SqsQueueRedriveAllowPolicy = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/sqs_queue_redrive_allow_policy aws_sqs_queue_redrive_allow_policy}
*/
class SqsQueueRedriveAllowPolicy extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/sqs_queue_redrive_allow_policy aws_sqs_queue_redrive_allow_policy} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SqsQueueRedriveAllowPolicyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_sqs_queue_redrive_allow_policy',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._queueUrl = config.queueUrl;
        this._redriveAllowPolicy = config.redriveAllowPolicy;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get queueUrl() {
        return this.getStringAttribute('queue_url');
    }
    set queueUrl(value) {
        this._queueUrl = value;
    }
    // Temporarily expose input value. Use with caution.
    get queueUrlInput() {
        return this._queueUrl;
    }
    get redriveAllowPolicy() {
        return this.getStringAttribute('redrive_allow_policy');
    }
    set redriveAllowPolicy(value) {
        this._redriveAllowPolicy = value;
    }
    // Temporarily expose input value. Use with caution.
    get redriveAllowPolicyInput() {
        return this._redriveAllowPolicy;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            queue_url: cdktf.stringToTerraform(this._queueUrl),
            redrive_allow_policy: cdktf.stringToTerraform(this._redriveAllowPolicy),
        };
    }
}
exports.SqsQueueRedriveAllowPolicy = SqsQueueRedriveAllowPolicy;
_a = JSII_RTTI_SYMBOL_1;
SqsQueueRedriveAllowPolicy[_a] = { fqn: "@cdktf/provider-aws.sqsQueueRedriveAllowPolicy.SqsQueueRedriveAllowPolicy", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
SqsQueueRedriveAllowPolicy.tfResourceType = "aws_sqs_queue_redrive_allow_policy";
