// @cdktf/provider-aws/lib/appstream-stack/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppstreamStack = exports.AppstreamStackUserSettingsList = exports.AppstreamStackUserSettingsOutputReference = exports.appstreamStackUserSettingsToTerraform = exports.AppstreamStackStorageConnectorsList = exports.AppstreamStackStorageConnectorsOutputReference = exports.appstreamStackStorageConnectorsToTerraform = exports.AppstreamStackApplicationSettingsOutputReference = exports.appstreamStackApplicationSettingsToTerraform = exports.AppstreamStackAccessEndpointsList = exports.AppstreamStackAccessEndpointsOutputReference = exports.appstreamStackAccessEndpointsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function appstreamStackAccessEndpointsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        endpoint_type: cdktf.stringToTerraform(struct.endpointType),
        vpce_id: cdktf.stringToTerraform(struct.vpceId),
    };
}
exports.appstreamStackAccessEndpointsToTerraform = appstreamStackAccessEndpointsToTerraform;
class AppstreamStackAccessEndpointsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._endpointType !== undefined) {
            hasAnyValues = true;
            internalValueResult.endpointType = this._endpointType;
        }
        if (this._vpceId !== undefined) {
            hasAnyValues = true;
            internalValueResult.vpceId = this._vpceId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._endpointType = undefined;
            this._vpceId = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._endpointType = value.endpointType;
            this._vpceId = value.vpceId;
        }
    }
    get endpointType() {
        return this.getStringAttribute('endpoint_type');
    }
    set endpointType(value) {
        this._endpointType = value;
    }
    // Temporarily expose input value. Use with caution.
    get endpointTypeInput() {
        return this._endpointType;
    }
    get vpceId() {
        return this.getStringAttribute('vpce_id');
    }
    set vpceId(value) {
        this._vpceId = value;
    }
    resetVpceId() {
        this._vpceId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpceIdInput() {
        return this._vpceId;
    }
}
exports.AppstreamStackAccessEndpointsOutputReference = AppstreamStackAccessEndpointsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
AppstreamStackAccessEndpointsOutputReference[_a] = { fqn: "@cdktf/provider-aws.appstreamStack.AppstreamStackAccessEndpointsOutputReference", version: "10.0.12" };
class AppstreamStackAccessEndpointsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AppstreamStackAccessEndpointsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AppstreamStackAccessEndpointsList = AppstreamStackAccessEndpointsList;
_b = JSII_RTTI_SYMBOL_1;
AppstreamStackAccessEndpointsList[_b] = { fqn: "@cdktf/provider-aws.appstreamStack.AppstreamStackAccessEndpointsList", version: "10.0.12" };
function appstreamStackApplicationSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        settings_group: cdktf.stringToTerraform(struct.settingsGroup),
    };
}
exports.appstreamStackApplicationSettingsToTerraform = appstreamStackApplicationSettingsToTerraform;
class AppstreamStackApplicationSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._settingsGroup !== undefined) {
            hasAnyValues = true;
            internalValueResult.settingsGroup = this._settingsGroup;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._settingsGroup = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._settingsGroup = value.settingsGroup;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get settingsGroup() {
        return this.getStringAttribute('settings_group');
    }
    set settingsGroup(value) {
        this._settingsGroup = value;
    }
    resetSettingsGroup() {
        this._settingsGroup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get settingsGroupInput() {
        return this._settingsGroup;
    }
}
exports.AppstreamStackApplicationSettingsOutputReference = AppstreamStackApplicationSettingsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
AppstreamStackApplicationSettingsOutputReference[_c] = { fqn: "@cdktf/provider-aws.appstreamStack.AppstreamStackApplicationSettingsOutputReference", version: "10.0.12" };
function appstreamStackStorageConnectorsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        connector_type: cdktf.stringToTerraform(struct.connectorType),
        domains: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.domains),
        resource_identifier: cdktf.stringToTerraform(struct.resourceIdentifier),
    };
}
exports.appstreamStackStorageConnectorsToTerraform = appstreamStackStorageConnectorsToTerraform;
class AppstreamStackStorageConnectorsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._connectorType !== undefined) {
            hasAnyValues = true;
            internalValueResult.connectorType = this._connectorType;
        }
        if (this._domains !== undefined) {
            hasAnyValues = true;
            internalValueResult.domains = this._domains;
        }
        if (this._resourceIdentifier !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourceIdentifier = this._resourceIdentifier;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._connectorType = undefined;
            this._domains = undefined;
            this._resourceIdentifier = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._connectorType = value.connectorType;
            this._domains = value.domains;
            this._resourceIdentifier = value.resourceIdentifier;
        }
    }
    get connectorType() {
        return this.getStringAttribute('connector_type');
    }
    set connectorType(value) {
        this._connectorType = value;
    }
    // Temporarily expose input value. Use with caution.
    get connectorTypeInput() {
        return this._connectorType;
    }
    get domains() {
        return this.getListAttribute('domains');
    }
    set domains(value) {
        this._domains = value;
    }
    resetDomains() {
        this._domains = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get domainsInput() {
        return this._domains;
    }
    get resourceIdentifier() {
        return this.getStringAttribute('resource_identifier');
    }
    set resourceIdentifier(value) {
        this._resourceIdentifier = value;
    }
    resetResourceIdentifier() {
        this._resourceIdentifier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourceIdentifierInput() {
        return this._resourceIdentifier;
    }
}
exports.AppstreamStackStorageConnectorsOutputReference = AppstreamStackStorageConnectorsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
AppstreamStackStorageConnectorsOutputReference[_d] = { fqn: "@cdktf/provider-aws.appstreamStack.AppstreamStackStorageConnectorsOutputReference", version: "10.0.12" };
class AppstreamStackStorageConnectorsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AppstreamStackStorageConnectorsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AppstreamStackStorageConnectorsList = AppstreamStackStorageConnectorsList;
_e = JSII_RTTI_SYMBOL_1;
AppstreamStackStorageConnectorsList[_e] = { fqn: "@cdktf/provider-aws.appstreamStack.AppstreamStackStorageConnectorsList", version: "10.0.12" };
function appstreamStackUserSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        action: cdktf.stringToTerraform(struct.action),
        permission: cdktf.stringToTerraform(struct.permission),
    };
}
exports.appstreamStackUserSettingsToTerraform = appstreamStackUserSettingsToTerraform;
class AppstreamStackUserSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._action !== undefined) {
            hasAnyValues = true;
            internalValueResult.action = this._action;
        }
        if (this._permission !== undefined) {
            hasAnyValues = true;
            internalValueResult.permission = this._permission;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._action = undefined;
            this._permission = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._action = value.action;
            this._permission = value.permission;
        }
    }
    get action() {
        return this.getStringAttribute('action');
    }
    set action(value) {
        this._action = value;
    }
    // Temporarily expose input value. Use with caution.
    get actionInput() {
        return this._action;
    }
    get permission() {
        return this.getStringAttribute('permission');
    }
    set permission(value) {
        this._permission = value;
    }
    // Temporarily expose input value. Use with caution.
    get permissionInput() {
        return this._permission;
    }
}
exports.AppstreamStackUserSettingsOutputReference = AppstreamStackUserSettingsOutputReference;
_f = JSII_RTTI_SYMBOL_1;
AppstreamStackUserSettingsOutputReference[_f] = { fqn: "@cdktf/provider-aws.appstreamStack.AppstreamStackUserSettingsOutputReference", version: "10.0.12" };
class AppstreamStackUserSettingsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AppstreamStackUserSettingsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AppstreamStackUserSettingsList = AppstreamStackUserSettingsList;
_g = JSII_RTTI_SYMBOL_1;
AppstreamStackUserSettingsList[_g] = { fqn: "@cdktf/provider-aws.appstreamStack.AppstreamStackUserSettingsList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/appstream_stack aws_appstream_stack}
*/
class AppstreamStack extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/appstream_stack aws_appstream_stack} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options AppstreamStackConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_appstream_stack',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // access_endpoints - computed: false, optional: true, required: false
        this._accessEndpoints = new AppstreamStackAccessEndpointsList(this, "access_endpoints", true);
        // application_settings - computed: false, optional: true, required: false
        this._applicationSettings = new AppstreamStackApplicationSettingsOutputReference(this, "application_settings");
        // storage_connectors - computed: false, optional: true, required: false
        this._storageConnectors = new AppstreamStackStorageConnectorsList(this, "storage_connectors", true);
        // user_settings - computed: false, optional: true, required: false
        this._userSettings = new AppstreamStackUserSettingsList(this, "user_settings", true);
        this._description = config.description;
        this._displayName = config.displayName;
        this._embedHostDomains = config.embedHostDomains;
        this._feedbackUrl = config.feedbackUrl;
        this._id = config.id;
        this._name = config.name;
        this._redirectUrl = config.redirectUrl;
        this._tags = config.tags;
        this._tagsAll = config.tagsAll;
        this._accessEndpoints.internalValue = config.accessEndpoints;
        this._applicationSettings.internalValue = config.applicationSettings;
        this._storageConnectors.internalValue = config.storageConnectors;
        this._userSettings.internalValue = config.userSettings;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // arn - computed: true, optional: false, required: false
    get arn() {
        return this.getStringAttribute('arn');
    }
    // created_time - computed: true, optional: false, required: false
    get createdTime() {
        return this.getStringAttribute('created_time');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    set displayName(value) {
        this._displayName = value;
    }
    resetDisplayName() {
        this._displayName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get displayNameInput() {
        return this._displayName;
    }
    get embedHostDomains() {
        return cdktf.Fn.tolist(this.getListAttribute('embed_host_domains'));
    }
    set embedHostDomains(value) {
        this._embedHostDomains = value;
    }
    resetEmbedHostDomains() {
        this._embedHostDomains = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get embedHostDomainsInput() {
        return this._embedHostDomains;
    }
    get feedbackUrl() {
        return this.getStringAttribute('feedback_url');
    }
    set feedbackUrl(value) {
        this._feedbackUrl = value;
    }
    resetFeedbackUrl() {
        this._feedbackUrl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get feedbackUrlInput() {
        return this._feedbackUrl;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get redirectUrl() {
        return this.getStringAttribute('redirect_url');
    }
    set redirectUrl(value) {
        this._redirectUrl = value;
    }
    resetRedirectUrl() {
        this._redirectUrl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get redirectUrlInput() {
        return this._redirectUrl;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get tagsAll() {
        return this.getStringMapAttribute('tags_all');
    }
    set tagsAll(value) {
        this._tagsAll = value;
    }
    resetTagsAll() {
        this._tagsAll = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsAllInput() {
        return this._tagsAll;
    }
    get accessEndpoints() {
        return this._accessEndpoints;
    }
    putAccessEndpoints(value) {
        this._accessEndpoints.internalValue = value;
    }
    resetAccessEndpoints() {
        this._accessEndpoints.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessEndpointsInput() {
        return this._accessEndpoints.internalValue;
    }
    get applicationSettings() {
        return this._applicationSettings;
    }
    putApplicationSettings(value) {
        this._applicationSettings.internalValue = value;
    }
    resetApplicationSettings() {
        this._applicationSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get applicationSettingsInput() {
        return this._applicationSettings.internalValue;
    }
    get storageConnectors() {
        return this._storageConnectors;
    }
    putStorageConnectors(value) {
        this._storageConnectors.internalValue = value;
    }
    resetStorageConnectors() {
        this._storageConnectors.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get storageConnectorsInput() {
        return this._storageConnectors.internalValue;
    }
    get userSettings() {
        return this._userSettings;
    }
    putUserSettings(value) {
        this._userSettings.internalValue = value;
    }
    resetUserSettings() {
        this._userSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get userSettingsInput() {
        return this._userSettings.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            display_name: cdktf.stringToTerraform(this._displayName),
            embed_host_domains: cdktf.listMapper(cdktf.stringToTerraform, false)(this._embedHostDomains),
            feedback_url: cdktf.stringToTerraform(this._feedbackUrl),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            redirect_url: cdktf.stringToTerraform(this._redirectUrl),
            tags: cdktf.hashMapper(cdktf.stringToTerraform)(this._tags),
            tags_all: cdktf.hashMapper(cdktf.stringToTerraform)(this._tagsAll),
            access_endpoints: cdktf.listMapper(appstreamStackAccessEndpointsToTerraform, true)(this._accessEndpoints.internalValue),
            application_settings: appstreamStackApplicationSettingsToTerraform(this._applicationSettings.internalValue),
            storage_connectors: cdktf.listMapper(appstreamStackStorageConnectorsToTerraform, true)(this._storageConnectors.internalValue),
            user_settings: cdktf.listMapper(appstreamStackUserSettingsToTerraform, true)(this._userSettings.internalValue),
        };
    }
}
exports.AppstreamStack = AppstreamStack;
_h = JSII_RTTI_SYMBOL_1;
AppstreamStack[_h] = { fqn: "@cdktf/provider-aws.appstreamStack.AppstreamStack", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
AppstreamStack.tfResourceType = "aws_appstream_stack";
