// @cdktf/provider-aws/lib/wafregional-sql-injection-match-set/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.WafregionalSqlInjectionMatchSet = exports.WafregionalSqlInjectionMatchSetSqlInjectionMatchTupleList = exports.WafregionalSqlInjectionMatchSetSqlInjectionMatchTupleOutputReference = exports.wafregionalSqlInjectionMatchSetSqlInjectionMatchTupleToTerraform = exports.WafregionalSqlInjectionMatchSetSqlInjectionMatchTupleFieldToMatchOutputReference = exports.wafregionalSqlInjectionMatchSetSqlInjectionMatchTupleFieldToMatchToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function wafregionalSqlInjectionMatchSetSqlInjectionMatchTupleFieldToMatchToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        data: cdktf.stringToTerraform(struct.data),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.wafregionalSqlInjectionMatchSetSqlInjectionMatchTupleFieldToMatchToTerraform = wafregionalSqlInjectionMatchSetSqlInjectionMatchTupleFieldToMatchToTerraform;
class WafregionalSqlInjectionMatchSetSqlInjectionMatchTupleFieldToMatchOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._data !== undefined) {
            hasAnyValues = true;
            internalValueResult.data = this._data;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._data = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._data = value.data;
            this._type = value.type;
        }
    }
    get data() {
        return this.getStringAttribute('data');
    }
    set data(value) {
        this._data = value;
    }
    resetData() {
        this._data = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataInput() {
        return this._data;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.WafregionalSqlInjectionMatchSetSqlInjectionMatchTupleFieldToMatchOutputReference = WafregionalSqlInjectionMatchSetSqlInjectionMatchTupleFieldToMatchOutputReference;
_a = JSII_RTTI_SYMBOL_1;
WafregionalSqlInjectionMatchSetSqlInjectionMatchTupleFieldToMatchOutputReference[_a] = { fqn: "@cdktf/provider-aws.wafregionalSqlInjectionMatchSet.WafregionalSqlInjectionMatchSetSqlInjectionMatchTupleFieldToMatchOutputReference", version: "10.0.12" };
function wafregionalSqlInjectionMatchSetSqlInjectionMatchTupleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        text_transformation: cdktf.stringToTerraform(struct.textTransformation),
        field_to_match: wafregionalSqlInjectionMatchSetSqlInjectionMatchTupleFieldToMatchToTerraform(struct.fieldToMatch),
    };
}
exports.wafregionalSqlInjectionMatchSetSqlInjectionMatchTupleToTerraform = wafregionalSqlInjectionMatchSetSqlInjectionMatchTupleToTerraform;
class WafregionalSqlInjectionMatchSetSqlInjectionMatchTupleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // field_to_match - computed: false, optional: false, required: true
        this._fieldToMatch = new WafregionalSqlInjectionMatchSetSqlInjectionMatchTupleFieldToMatchOutputReference(this, "field_to_match");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._textTransformation !== undefined) {
            hasAnyValues = true;
            internalValueResult.textTransformation = this._textTransformation;
        }
        if (this._fieldToMatch?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.fieldToMatch = this._fieldToMatch?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._textTransformation = undefined;
            this._fieldToMatch.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._textTransformation = value.textTransformation;
            this._fieldToMatch.internalValue = value.fieldToMatch;
        }
    }
    get textTransformation() {
        return this.getStringAttribute('text_transformation');
    }
    set textTransformation(value) {
        this._textTransformation = value;
    }
    // Temporarily expose input value. Use with caution.
    get textTransformationInput() {
        return this._textTransformation;
    }
    get fieldToMatch() {
        return this._fieldToMatch;
    }
    putFieldToMatch(value) {
        this._fieldToMatch.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get fieldToMatchInput() {
        return this._fieldToMatch.internalValue;
    }
}
exports.WafregionalSqlInjectionMatchSetSqlInjectionMatchTupleOutputReference = WafregionalSqlInjectionMatchSetSqlInjectionMatchTupleOutputReference;
_b = JSII_RTTI_SYMBOL_1;
WafregionalSqlInjectionMatchSetSqlInjectionMatchTupleOutputReference[_b] = { fqn: "@cdktf/provider-aws.wafregionalSqlInjectionMatchSet.WafregionalSqlInjectionMatchSetSqlInjectionMatchTupleOutputReference", version: "10.0.12" };
class WafregionalSqlInjectionMatchSetSqlInjectionMatchTupleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new WafregionalSqlInjectionMatchSetSqlInjectionMatchTupleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.WafregionalSqlInjectionMatchSetSqlInjectionMatchTupleList = WafregionalSqlInjectionMatchSetSqlInjectionMatchTupleList;
_c = JSII_RTTI_SYMBOL_1;
WafregionalSqlInjectionMatchSetSqlInjectionMatchTupleList[_c] = { fqn: "@cdktf/provider-aws.wafregionalSqlInjectionMatchSet.WafregionalSqlInjectionMatchSetSqlInjectionMatchTupleList", version: "10.0.12" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/aws/r/wafregional_sql_injection_match_set aws_wafregional_sql_injection_match_set}
*/
class WafregionalSqlInjectionMatchSet extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/aws/r/wafregional_sql_injection_match_set aws_wafregional_sql_injection_match_set} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options WafregionalSqlInjectionMatchSetConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'aws_wafregional_sql_injection_match_set',
            terraformGeneratorMetadata: {
                providerName: 'aws',
                providerVersion: '4.39.0',
                providerVersionConstraint: '~> 4.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // sql_injection_match_tuple - computed: false, optional: true, required: false
        this._sqlInjectionMatchTuple = new WafregionalSqlInjectionMatchSetSqlInjectionMatchTupleList(this, "sql_injection_match_tuple", true);
        this._id = config.id;
        this._name = config.name;
        this._sqlInjectionMatchTuple.internalValue = config.sqlInjectionMatchTuple;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get sqlInjectionMatchTuple() {
        return this._sqlInjectionMatchTuple;
    }
    putSqlInjectionMatchTuple(value) {
        this._sqlInjectionMatchTuple.internalValue = value;
    }
    resetSqlInjectionMatchTuple() {
        this._sqlInjectionMatchTuple.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sqlInjectionMatchTupleInput() {
        return this._sqlInjectionMatchTuple.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            sql_injection_match_tuple: cdktf.listMapper(wafregionalSqlInjectionMatchSetSqlInjectionMatchTupleToTerraform, true)(this._sqlInjectionMatchTuple.internalValue),
        };
    }
}
exports.WafregionalSqlInjectionMatchSet = WafregionalSqlInjectionMatchSet;
_d = JSII_RTTI_SYMBOL_1;
WafregionalSqlInjectionMatchSet[_d] = { fqn: "@cdktf/provider-aws.wafregionalSqlInjectionMatchSet.WafregionalSqlInjectionMatchSet", version: "10.0.12" };
// =================
// STATIC PROPERTIES
// =================
WafregionalSqlInjectionMatchSet.tfResourceType = "aws_wafregional_sql_injection_match_set";
